{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Supplementary Material\n",
    "\n",
    "> **\"Background-dependent selection of genetic variation in heterogeneous populations\"**\n",
    "\n",
    "> Ignacio Vázquez-García, Francisco Salinas, Jing Li, Andrej Fischer, Benjamin Barré, Johan Hallin, Anders Bergström, Elisa Alonso-Pérez, Jonas Warringer, Ville Mustonen, Gianni Liti\n",
    "\n",
    "## Figure 3 (+ Supp. Figs.)\n",
    "\n",
    "This IPython notebook is provided for reproduction of Figure 3 of the paper. It can be viewed by copying its URL to nbviewer and it can be run by opening it in binder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# load external dependencies\n",
    "import matplotlib.gridspec as gridspec\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import matplotlib.lines as lines\n",
    "import matplotlib.patches as patches\n",
    "%matplotlib inline\n",
    "\n",
    "import json,matplotlib\n",
    "s = json.load( open(\"matplotlibrc.json\") )\n",
    "matplotlib.rcParams.update(s)\n",
    "\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "import pandas as pd # (*) Pandas for data manipulation\n",
    "\n",
    "# load internal dependencies\n",
    "import colors,plot,model,utils,config\n",
    "from utils import sgd_length\n",
    "\n",
    "# absolute file path for root directory of git repo\n",
    "dir_repo = utils.get_git_path()\n",
    "dir_data = dir_repo+\"/data/\"\n",
    "dir_paper = dir_repo+\"/manuscript/main/\"\n",
    "dir_supp = dir_repo+\"/manuscript/supp/\"\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import warnings\n",
    "\n",
    "with warnings.catch_warnings():\n",
    "    warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data import"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Length distribution of homozygosity tracts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HU</th>\n",
       "      <th>RM</th>\n",
       "      <th>YPD</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50kb_bin_center</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>0.731240</td>\n",
       "      <td>0.745737</td>\n",
       "      <td>0.749937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>0.140123</td>\n",
       "      <td>0.153088</td>\n",
       "      <td>0.159649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>0.053599</td>\n",
       "      <td>0.056461</td>\n",
       "      <td>0.053509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>0.033691</td>\n",
       "      <td>0.021599</td>\n",
       "      <td>0.021700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>0.013017</td>\n",
       "      <td>0.009852</td>\n",
       "      <td>0.008454</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       HU        RM       YPD\n",
       "50kb_bin_center                              \n",
       "25               0.731240  0.745737  0.749937\n",
       "75               0.140123  0.153088  0.159649\n",
       "125              0.053599  0.056461  0.053509\n",
       "175              0.033691  0.021599  0.021700\n",
       "225              0.013017  0.009852  0.008454"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load data\n",
    "loh_length_df = pd.read_csv(dir_data+'seq/loh/homozygosity_length.tsv', sep='\\t')\n",
    "loh_length_df = loh_length_df.set_index(\"50kb_bin_center\")\n",
    "loh_length_df = loh_length_df.reindex(columns=['HU','RM','YPD'])\n",
    "\n",
    "loh_length_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Examples of loss-of-heterozygosity in clonal genome sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>selection</th>\n",
       "      <th>population</th>\n",
       "      <th>set</th>\n",
       "      <th>clone</th>\n",
       "      <th>lineage</th>\n",
       "      <th>type</th>\n",
       "      <th>chr_arabic</th>\n",
       "      <th>chr_roman</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>pos</th>\n",
       "      <th>genotype</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HU</td>\n",
       "      <td>WAxNA_F12_1_HU_3</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>subclone A</td>\n",
       "      <td>copy_number</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>0</td>\n",
       "      <td>230218</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RM</td>\n",
       "      <td>WAxNA_F12_1_RM_3</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>subclone A</td>\n",
       "      <td>copy_number</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>0</td>\n",
       "      <td>230218</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RM</td>\n",
       "      <td>WAxNA_F12_1_RM_3</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>bulk</td>\n",
       "      <td>copy_number</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>0</td>\n",
       "      <td>230218</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RM</td>\n",
       "      <td>WAxNA_F12_1_RM_2</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>bulk</td>\n",
       "      <td>copy_number</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>0</td>\n",
       "      <td>230218</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RM</td>\n",
       "      <td>WAxNA_F12_1_RM_2</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>bulk</td>\n",
       "      <td>copy_number</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>0</td>\n",
       "      <td>230218</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  selection        population  set clone     lineage         type  chr_arabic  \\\n",
       "0        HU  WAxNA_F12_1_HU_3    4     6  subclone A  copy_number           1   \n",
       "1        RM  WAxNA_F12_1_RM_3    8     2  subclone A  copy_number           1   \n",
       "2        RM  WAxNA_F12_1_RM_3    7     1        bulk  copy_number           1   \n",
       "3        RM  WAxNA_F12_1_RM_2    6     6        bulk  copy_number           1   \n",
       "4        RM  WAxNA_F12_1_RM_2    5     5        bulk  copy_number           1   \n",
       "\n",
       "  chr_roman  start     end  length  pos  genotype gene  \n",
       "0         I      0  230218     NaN    0         2       \n",
       "1         I      0  230218     NaN    0         2       \n",
       "2         I      0  230218     NaN    0         2       \n",
       "3         I      0  230218     NaN    0         2       \n",
       "4         I      0  230218     NaN    0         2       "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load data\n",
    "seq_loh_df = utils.load_data(dir_data+'seq/loh/seq_loh.pickle')\n",
    "\n",
    "seq_loh_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ac</th>\n",
       "      <th>af</th>\n",
       "      <th>alt</th>\n",
       "      <th>an</th>\n",
       "      <th>allele</th>\n",
       "      <th>amino_acids</th>\n",
       "      <th>biotype</th>\n",
       "      <th>baseqranksum</th>\n",
       "      <th>cds_position</th>\n",
       "      <th>chr_arabic</th>\n",
       "      <th>...</th>\n",
       "      <th>clone</th>\n",
       "      <th>population</th>\n",
       "      <th>coverage</th>\n",
       "      <th>substitution</th>\n",
       "      <th>assignment</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>consequence_short</th>\n",
       "      <th>chr_roman</th>\n",
       "      <th>pos</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>[A]</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>R/I</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.723</td>\n",
       "      <td>101</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>WAxNA_F12_1_HU_1</td>\n",
       "      <td>28.795032</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>bulk</td>\n",
       "      <td>driver</td>\n",
       "      <td>non-synonymous</td>\n",
       "      <td>VII</td>\n",
       "      <td>4595191</td>\n",
       "      <td>de novo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>[C]</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.34</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>WAxNA_F12_1_HU_1</td>\n",
       "      <td>28.795032</td>\n",
       "      <td>A&gt;C</td>\n",
       "      <td>bulk</td>\n",
       "      <td>passenger</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IV</td>\n",
       "      <td>2849276</td>\n",
       "      <td>de novo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>[A]</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>C/F</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-1.488</td>\n",
       "      <td>1883</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>WAxNA_F12_1_HU_1</td>\n",
       "      <td>28.795032</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>bulk</td>\n",
       "      <td>passenger</td>\n",
       "      <td>non-synonymous</td>\n",
       "      <td>X</td>\n",
       "      <td>5980969</td>\n",
       "      <td>de novo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>[T]</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.653</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>WAxNA_F12_1_HU_1</td>\n",
       "      <td>28.795032</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>bulk</td>\n",
       "      <td>passenger</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XIII</td>\n",
       "      <td>8949297</td>\n",
       "      <td>de novo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>[C]</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>C/W</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.674</td>\n",
       "      <td>477</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>WAxNA_F12_1_HU_1</td>\n",
       "      <td>28.795032</td>\n",
       "      <td>A&gt;C</td>\n",
       "      <td>bulk</td>\n",
       "      <td>passenger</td>\n",
       "      <td>non-synonymous</td>\n",
       "      <td>XIII</td>\n",
       "      <td>9157827</td>\n",
       "      <td>de novo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 82 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  ac   af  alt  an allele amino_acids         biotype baseqranksum  \\\n",
       "0  1  0.5  [A]   2      A         R/I  protein_coding        1.723   \n",
       "1  1  0.5  [C]   2    NaN         NaN             NaN        -0.34   \n",
       "2  1  0.5  [A]   2      A         C/F  protein_coding       -1.488   \n",
       "3  1  0.5  [T]   2    NaN         NaN             NaN       -0.653   \n",
       "4  1  0.5  [C]   2      C         C/W  protein_coding        1.674   \n",
       "\n",
       "  cds_position  chr_arabic   ...    clone        population   coverage  \\\n",
       "0          101           7   ...        1  WAxNA_F12_1_HU_1  28.795032   \n",
       "1          NaN           4   ...        1  WAxNA_F12_1_HU_1  28.795032   \n",
       "2         1883          10   ...        1  WAxNA_F12_1_HU_1  28.795032   \n",
       "3          NaN          13   ...        1  WAxNA_F12_1_HU_1  28.795032   \n",
       "4          477          13   ...        1  WAxNA_F12_1_HU_1  28.795032   \n",
       "\n",
       "   substitution assignment mutation_type consequence_short chr_roman      pos  \\\n",
       "0           C>A       bulk        driver    non-synonymous       VII  4595191   \n",
       "1           A>C       bulk     passenger               NaN        IV  2849276   \n",
       "2           C>A       bulk     passenger    non-synonymous         X  5980969   \n",
       "3           C>T       bulk     passenger               NaN      XIII  8949297   \n",
       "4           A>C       bulk     passenger    non-synonymous      XIII  9157827   \n",
       "\n",
       "      type  \n",
       "0  de novo  \n",
       "1  de novo  \n",
       "2  de novo  \n",
       "3  de novo  \n",
       "4  de novo  \n",
       "\n",
       "[5 rows x 82 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load data\n",
    "seq_mut_df = utils.load_data(dir_data+'seq/de-novo/seq_de_novo_snps_indels.pickle')\n",
    "\n",
    "seq_mut_df = seq_mut_df.rename(columns={'chrom':'chr_arabic', 'pos':'site'})\n",
    "seq_mut_df['chr_roman'] = seq_mut_df['chr_arabic'].apply(utils.int_to_roman)\n",
    "seq_mut_df = utils.chr_to_gw(seq_mut_df)\n",
    "\n",
    "seq_mut_df = seq_mut_df[(seq_mut_df.time==32)&\\\n",
    "                        (seq_mut_df.clone!='')].reset_index(drop=True)\n",
    "\n",
    "# remap genotype values\n",
    "seq_mut_df['genotype'] = seq_mut_df['genotype'].map({'0/0': 0, '0/1': 1, '1/0': 1, '1/1': 2})\n",
    "\n",
    "seq_mut_df['type'] = 'de novo'\n",
    "\n",
    "seq_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>chr_arabic</th>\n",
       "      <th colspan=\"10\" halign=\"left\">1</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"10\" halign=\"left\">16</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>chr_roman</th>\n",
       "      <th colspan=\"10\" halign=\"left\">I</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"10\" halign=\"left\">XVI</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>0</th>\n",
       "      <th>2718</th>\n",
       "      <th>32682</th>\n",
       "      <th>32947</th>\n",
       "      <th>33867</th>\n",
       "      <th>33918</th>\n",
       "      <th>34245</th>\n",
       "      <th>49871</th>\n",
       "      <th>50204</th>\n",
       "      <th>51599</th>\n",
       "      <th>...</th>\n",
       "      <th>12044507</th>\n",
       "      <th>12044750</th>\n",
       "      <th>12045073</th>\n",
       "      <th>12045651</th>\n",
       "      <th>12046112</th>\n",
       "      <th>12048293</th>\n",
       "      <th>12050111</th>\n",
       "      <th>12051171</th>\n",
       "      <th>12056269</th>\n",
       "      <th>12071326</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>BDH2</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>...</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>non-coding</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>selection</th>\n",
       "      <th>population</th>\n",
       "      <th>set</th>\n",
       "      <th>clone</th>\n",
       "      <th>lineage</th>\n",
       "      <th>type</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">HU</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">WAxNA_F12_1_HU_1_T32</th>\n",
       "      <th rowspan=\"4\" valign=\"top\">1</th>\n",
       "      <th></th>\n",
       "      <th>subclone A</th>\n",
       "      <th>consensus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">1</th>\n",
       "      <th rowspan=\"3\" valign=\"top\">subclone A</th>\n",
       "      <th>copy_number</th>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>de novo</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>loh</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <th></th>\n",
       "      <th>bulk</th>\n",
       "      <th>consensus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 10733 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "chr_arabic                                                            1   \\\n",
       "chr_roman                                                              I   \n",
       "pos                                                             0          \n",
       "gene                                                                       \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus        NaN   \n",
       "                                   1     subclone A copy_number        2   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh              NaN   \n",
       "                               2         bulk       consensus        NaN   \n",
       "\n",
       "chr_arabic                                                                \\\n",
       "chr_roman                                                                  \n",
       "pos                                                             2718       \n",
       "gene                                                                       \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus          2   \n",
       "                                   1     subclone A copy_number      NaN   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh                0   \n",
       "                               2         bulk       consensus          2   \n",
       "\n",
       "chr_arabic                                                                \\\n",
       "chr_roman                                                                  \n",
       "pos                                                             32682      \n",
       "gene                                                                       \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus          2   \n",
       "                                   1     subclone A copy_number      NaN   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh                0   \n",
       "                               2         bulk       consensus        NaN   \n",
       "\n",
       "chr_arabic                                                                \\\n",
       "chr_roman                                                                  \n",
       "pos                                                             32947      \n",
       "gene                                                                       \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus          1   \n",
       "                                   1     subclone A copy_number      NaN   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh              NaN   \n",
       "                               2         bulk       consensus        NaN   \n",
       "\n",
       "chr_arabic                                                                \\\n",
       "chr_roman                                                                  \n",
       "pos                                                             33867      \n",
       "gene                                                                       \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus          1   \n",
       "                                   1     subclone A copy_number      NaN   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh              NaN   \n",
       "                               2         bulk       consensus        NaN   \n",
       "\n",
       "chr_arabic                                                                \\\n",
       "chr_roman                                                                  \n",
       "pos                                                             33918      \n",
       "gene                                                                BDH2   \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus        NaN   \n",
       "                                   1     subclone A copy_number      NaN   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh              NaN   \n",
       "                               2         bulk       consensus        NaN   \n",
       "\n",
       "chr_arabic                                                                \\\n",
       "chr_roman                                                                  \n",
       "pos                                                             34245      \n",
       "gene                                                                       \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus          2   \n",
       "                                   1     subclone A copy_number      NaN   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh                0   \n",
       "                               2         bulk       consensus        NaN   \n",
       "\n",
       "chr_arabic                                                                \\\n",
       "chr_roman                                                                  \n",
       "pos                                                             49871      \n",
       "gene                                                                       \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus        NaN   \n",
       "                                   1     subclone A copy_number      NaN   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh              NaN   \n",
       "                               2         bulk       consensus        NaN   \n",
       "\n",
       "chr_arabic                                                                \\\n",
       "chr_roman                                                                  \n",
       "pos                                                             50204      \n",
       "gene                                                                       \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus        NaN   \n",
       "                                   1     subclone A copy_number      NaN   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh              NaN   \n",
       "                               2         bulk       consensus        NaN   \n",
       "\n",
       "chr_arabic                                                                \\\n",
       "chr_roman                                                                  \n",
       "pos                                                             51599      \n",
       "gene                                                                       \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus        NaN   \n",
       "                                   1     subclone A copy_number      NaN   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh              NaN   \n",
       "                               2         bulk       consensus        NaN   \n",
       "\n",
       "chr_arabic                                                        ...     \\\n",
       "chr_roman                                                         ...      \n",
       "pos                                                               ...      \n",
       "gene                                                              ...      \n",
       "selection population           set clone lineage    type          ...      \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus     ...      \n",
       "                                   1     subclone A copy_number   ...      \n",
       "                                                    de novo       ...      \n",
       "                                                    loh           ...      \n",
       "                               2         bulk       consensus     ...      \n",
       "\n",
       "chr_arabic                                                            16  \\\n",
       "chr_roman                                                            XVI   \n",
       "pos                                                             12044507   \n",
       "gene                                                                       \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus        NaN   \n",
       "                                   1     subclone A copy_number      NaN   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh              NaN   \n",
       "                               2         bulk       consensus          0   \n",
       "\n",
       "chr_arabic                                                                \\\n",
       "chr_roman                                                                  \n",
       "pos                                                             12044750   \n",
       "gene                                                                       \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus        NaN   \n",
       "                                   1     subclone A copy_number      NaN   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh              NaN   \n",
       "                               2         bulk       consensus        NaN   \n",
       "\n",
       "chr_arabic                                                                \\\n",
       "chr_roman                                                                  \n",
       "pos                                                             12045073   \n",
       "gene                                                                       \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus        NaN   \n",
       "                                   1     subclone A copy_number      NaN   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh              NaN   \n",
       "                               2         bulk       consensus        NaN   \n",
       "\n",
       "chr_arabic                                                                \\\n",
       "chr_roman                                                                  \n",
       "pos                                                             12045651   \n",
       "gene                                                                       \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus        NaN   \n",
       "                                   1     subclone A copy_number      NaN   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh              NaN   \n",
       "                               2         bulk       consensus        NaN   \n",
       "\n",
       "chr_arabic                                                                \\\n",
       "chr_roman                                                                  \n",
       "pos                                                             12046112   \n",
       "gene                                                                       \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus        NaN   \n",
       "                                   1     subclone A copy_number      NaN   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh              NaN   \n",
       "                               2         bulk       consensus        NaN   \n",
       "\n",
       "chr_arabic                                                                \\\n",
       "chr_roman                                                                  \n",
       "pos                                                             12048293   \n",
       "gene                                                                       \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus        NaN   \n",
       "                                   1     subclone A copy_number      NaN   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh              NaN   \n",
       "                               2         bulk       consensus          0   \n",
       "\n",
       "chr_arabic                                                                \\\n",
       "chr_roman                                                                  \n",
       "pos                                                             12050111   \n",
       "gene                                                                       \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus        NaN   \n",
       "                                   1     subclone A copy_number      NaN   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh              NaN   \n",
       "                               2         bulk       consensus          1   \n",
       "\n",
       "chr_arabic                                                                \\\n",
       "chr_roman                                                                  \n",
       "pos                                                             12051171   \n",
       "gene                                                                       \n",
       "selection population           set clone lineage    type                   \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus          2   \n",
       "                                   1     subclone A copy_number      NaN   \n",
       "                                                    de novo          NaN   \n",
       "                                                    loh                0   \n",
       "                               2         bulk       consensus          1   \n",
       "\n",
       "chr_arabic                                                                  \\\n",
       "chr_roman                                                                    \n",
       "pos                                                               12056269   \n",
       "gene                                                            non-coding   \n",
       "selection population           set clone lineage    type                     \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus          NaN   \n",
       "                                   1     subclone A copy_number        NaN   \n",
       "                                                    de novo            NaN   \n",
       "                                                    loh                NaN   \n",
       "                               2         bulk       consensus          NaN   \n",
       "\n",
       "chr_arabic                                                                \n",
       "chr_roman                                                                 \n",
       "pos                                                             12071326  \n",
       "gene                                                                      \n",
       "selection population           set clone lineage    type                  \n",
       "HU        WAxNA_F12_1_HU_1_T32 1         subclone A consensus        NaN  \n",
       "                                   1     subclone A copy_number        2  \n",
       "                                                    de novo          NaN  \n",
       "                                                    loh              NaN  \n",
       "                               2         bulk       consensus        NaN  \n",
       "\n",
       "[5 rows x 10733 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns=['selection','population','clone']\n",
    "seq_mut_fixed_df = pd.merge(seq_mut_df, \n",
    "                            seq_loh_df[['selection','population','clone','set','lineage']].drop_duplicates(columns), \n",
    "                            on=columns, how='left')\n",
    "seq_loh_fixed_df = pd.merge(seq_loh_df, \n",
    "                            seq_mut_df[['selection','population','clone']].drop_duplicates(columns), \n",
    "                            on=columns, how='left')\n",
    "\n",
    "seq_dn_df = pd.concat([seq_loh_fixed_df, seq_mut_fixed_df], axis=0).reset_index(drop=True)\n",
    "\n",
    "seq_dn_df['population'] = seq_dn_df['population']+'_T32'\n",
    "\n",
    "seq_dn_df = pd.pivot_table(seq_dn_df, \n",
    "                           index=['selection','population','set','clone','lineage','type'], \n",
    "                           columns=['chr_arabic','chr_roman','pos','gene'])['genotype']\n",
    "seq_dn_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fluctuation assay"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"6\" halign=\"left\">biological_rep</th>\n",
       "      <th colspan=\"4\" halign=\"left\">technical_rep</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"4\" halign=\"left\">Total 5-FOA</th>\n",
       "      <th colspan=\"6\" halign=\"left\">LOH rate</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"3\" halign=\"left\">mean</th>\n",
       "      <th colspan=\"3\" halign=\"left\">sem</th>\n",
       "      <th colspan=\"3\" halign=\"left\">mean</th>\n",
       "      <th>sem</th>\n",
       "      <th>...</th>\n",
       "      <th>mean</th>\n",
       "      <th colspan=\"3\" halign=\"left\">sem</th>\n",
       "      <th colspan=\"3\" halign=\"left\">mean</th>\n",
       "      <th colspan=\"3\" halign=\"left\">sem</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>background</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>...</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>environment</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HU</th>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0.188982</td>\n",
       "      <td>...</td>\n",
       "      <td>237500</td>\n",
       "      <td>37796.447301</td>\n",
       "      <td>459910.705619</td>\n",
       "      <td>86473.571520</td>\n",
       "      <td>0.016617</td>\n",
       "      <td>0.002876</td>\n",
       "      <td>0.004080</td>\n",
       "      <td>0.004539</td>\n",
       "      <td>0.000354</td>\n",
       "      <td>0.000876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RM</th>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0.188982</td>\n",
       "      <td>...</td>\n",
       "      <td>3662500</td>\n",
       "      <td>329628.782009</td>\n",
       "      <td>434293.093511</td>\n",
       "      <td>302924.731340</td>\n",
       "      <td>0.000352</td>\n",
       "      <td>0.000990</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.000021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YPD</th>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0.188982</td>\n",
       "      <td>...</td>\n",
       "      <td>4237500</td>\n",
       "      <td>957396.021956</td>\n",
       "      <td>316712.676795</td>\n",
       "      <td>342228.776865</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>0.000009</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            biological_rep                                            \\\n",
       "                      mean                   sem                       \n",
       "background           WA/WA NA/NA WA/NA     WA/WA     NA/NA     WA/NA   \n",
       "environment                                                            \n",
       "HU                     2.5   2.5   2.5  0.422577  0.422577  0.422577   \n",
       "RM                     2.5   2.5   2.5  0.422577  0.422577  0.422577   \n",
       "YPD                    2.5   2.5   2.5  0.422577  0.422577  0.422577   \n",
       "\n",
       "            technical_rep                          ...    Total 5-FOA  \\\n",
       "                     mean                   sem    ...           mean   \n",
       "background          WA/WA NA/NA WA/NA     WA/WA    ...          WA/NA   \n",
       "environment                                        ...                  \n",
       "HU                    1.5   1.5   1.5  0.188982    ...         237500   \n",
       "RM                    1.5   1.5   1.5  0.188982    ...        3662500   \n",
       "YPD                   1.5   1.5   1.5  0.188982    ...        4237500   \n",
       "\n",
       "                                                          LOH rate            \\\n",
       "                       sem                                    mean             \n",
       "background           WA/WA          NA/NA          WA/NA     WA/WA     NA/NA   \n",
       "environment                                                                    \n",
       "HU            37796.447301  459910.705619   86473.571520  0.016617  0.002876   \n",
       "RM           329628.782009  434293.093511  302924.731340  0.000352  0.000990   \n",
       "YPD          957396.021956  316712.676795  342228.776865  0.000090  0.000209   \n",
       "\n",
       "                                                     \n",
       "                            sem                      \n",
       "background      WA/NA     WA/WA     NA/NA     WA/NA  \n",
       "environment                                          \n",
       "HU           0.004080  0.004539  0.000354  0.000876  \n",
       "RM           0.000044  0.000181  0.000114  0.000021  \n",
       "YPD          0.000038  0.000022  0.000037  0.000009  \n",
       "\n",
       "[3 rows x 42 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read csv file containing the competition assay data\n",
    "loh_fluctuation_df = pd.read_csv(dir_data+'fluctuation/fluctuation_assay.csv')\n",
    "\n",
    "# dilution factors\n",
    "loh_fluctuation_df['5-FOA+'] = loh_fluctuation_df['5FOA']*200/10\n",
    "loh_fluctuation_df['Total 5-FOA'] = loh_fluctuation_df['YPD']*1000000*2/40\n",
    "\n",
    "# rate\n",
    "loh_fluctuation_df['LOH rate'] = loh_fluctuation_df['5-FOA+']/loh_fluctuation_df['Total 5-FOA']\n",
    "\n",
    "loh_fluctuation_df = loh_fluctuation_df.groupby(['background','environment'])\\\n",
    ".agg([np.mean, stats.sem])\n",
    "\n",
    "# order columns\n",
    "loh_fluctuation_df = loh_fluctuation_df.ix[['WA/WA','NA/NA','WA/NA']]\n",
    "loh_fluctuation_df = loh_fluctuation_df.unstack('background')\n",
    "\n",
    "loh_fluctuation_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 3 - Loss of heterozygosity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAUwAAAEWCAYAAAANe67OAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJztnXl4VOX1+D8nMQQCAUT2XRERcUOUH4oKtlJacStaKbao\nLUWLICq4V2SxbiBYKKLWii1W+VqlreBGUVkUFxYRCyKi7IvsCUsIJHB+f7x3cAiT5E4yy53kfJ5n\nnpm5y/uee+edc9/lLKKqGIZhGKWTlmwBDMMwUgVTmIZhGD4xhWkYhuETU5iGYRg+MYVpGIbhE1OY\nhmEYPolKYYrIJBE5LCKd4yWQYRhGUPGtMEXkOOAqYCtwTdwkMiosItLFe+CGXqtE5NJky2UYfomm\nhxlq2MOAnnGQxagcKHAi0ASYBzyWXHEMwz/RKMxrgDeAqUBTETkv2spE5HoRGRTteUaFYxfwPbAd\nOFiWAkSkg4jcHlOpDKMUjvNzkIikA1cDNYHegADXAguKOb4lMAbogxvGN/R2bQOWi8jxqrqrPIIb\nKYsAa3FtLwvoF/Gg4tvQl8BCVV0kIu3iLaxhhOO3h/kjoDpwNnAW8AhOYRbHbmAj7s9xmqo+BTT2\n6mttyrJSo8CFwBm4djRGRCTCcRHbkKq+r6q5CZPWMMLwqzCvBear6nJV/QZ4C2gpImdHOlhVd+KG\nW+B6EQCHVXWyqk4srhIR0eHDh+vs2bMV98eyV7BesUCAPO8FkO5tO4ri2lCRYyZHrMDaUSq8UhJf\nQ3JVvaXI989wDb3UU4ElInIvsMZPXV27dqVr165+DjVSEwW+9d43A/1U9XApx0fVhsDakREfJEjh\n3UREX377Nd7cuIBhV/WnTb2WyRbJOJpIQ+fAISIapHadCqzZvpdteYVHvtfLOo6WdWvEs8qUaEtF\nCZynzz9XzuHL+UsZ8cYzyRYlqeQdymPezg/JO5R3zL4xY8YAMGfOHBYtWnTUvieeeCIh8gWd2bNn\nJ1uElGJbXiGPLsw78gpXngBDhw5l//799OjRg8OHD3Pvvfdy4YUXHtk/efJkPvnkE/74xz/ywQcf\nMGrUKPbt28ef/vQnJk+OOHOSkvgakieSJ3oPYcQbzzDsqv7JFiWpLM5dxHvbZgDQuc5FR+3bvn07\nTzzxBB988AHdunXj/fffp379+uzevZucnBz69+9P7969eeeddxARunbtytq1aykoKCA3N5c2bdqw\nc+dOevbsyYsvvkhhYSHNmjVj/fr1FBQU8Ktf/YpXXnmF7du38+STT5Ke7mf2xajIZGVl8fHHH9Om\nTRs+++wz2rVrxymnnMKCBQvIzs7mtNNO47PPPiM7O5vly5dz8OBBqlevjoiwdu3aZIsfMwLXw9y8\nbA2v/O6JSj8cb1+rA5fW6077Wh2O2VevXj3uvfdeHnjgAQAyMjLYvHkzgwYNolq1arRq1Ypq1arR\nrVs3OnfuTNOmTZk0aRI33ngjVapUYfXq1VSpUgVVpaCggPT0dK6//noWLVpEgwYNWLt2LVWrViU3\nN5d9+/Yl+tJjgs1fxpZTTz2VadOm0bdvXyZMmMC8efPIycnh3XffZdGiRXTo0IEdO3ZQo4Ybxmdm\nZvLVV1+xYsUKli5dyu7du5N8BbEhcArTGrojKz2LznUuIis9q8TjDh8+zAUXXECTJk0YO3YshYWF\npKWl0bZtW9555x0+/vhjVq1axZAhQ3j++efp2LEj6enptGnThokTJ7J+/XrS0lwzOOOMM9iwYQMN\nGjRg69at1KpVi/Xr1yficmOODcmjo17WcTxwbtaRV72sowefF154Idu3b6ddu3Zs376dM888kyFD\nhlBQUMChQ4cQEXbu3Mn5559PRkYGrVu3ZsaMGYwfP55HH32UadOmJenKYkvgFn1mzZplStMn+fn5\nPPbYY4wYMcL3OePGjeO6666jUaNGZakyJSbqbdEnJUiJtlSUwCnMIMljHENKNHJrRylBSrSlogRu\nSG5DKSMWWDsy4oH1MFOQxx9/nLS0NLZt28btt99O06ZNy1XeqlWryMzMpEmTJqUdmhK9AmtH0bNr\nxUryN24+8r1qk0Yc36b1ke8333wzffv2JT8/HxFhzJgxvPHGG+Tn5zN9+nQ+/PBDxo8fz5gxYxgy\nZAg9e/Zk7NixtGzZsrgqU6ItFSVwZkWGI+fAPqZ9O58rT+5I7czqR+2rUqUKgwcPZvny5UyaNIld\nu3bRo0cPZsyYQVpaGr/61a+YOXMmVapUYc+ePdSoUYMvv/yS7Oxs2rVrx9VXX81f//pXRIRdu3bR\nsWNHWrRowciRI6lVqxajRo1K0lUbySJ/42ZW3DDgyPc2k5+GMIXZpk0bpk+fTrdu3fj000+59NJL\n2bJlC1988QWqSnp6OsuXL0dE2Lp1Kz/96U+ZNm0agwZVrOBkNiQPKNO+nc+4z99i2rfzj9l38OBB\nxo0bx9ChQ0lLSyM3N5e8vDyuvvpqWrVqhYiQmZnJxo0bqVKlCoMGDaJt27a0bduW7t278/TTT3Pq\nqaciIvzxj38kOzsbEaFbt25lXQwKHNaOYs8vfvELpk6dynvvvUdubi5vvPEG33//PQ0bNuT6669n\nypQpqCrTp09nx44dLFmyJNkix5zAKUxbIXdceXJHbj+nB1ee3PGYfVWqVOH2229n8ODBvPnmm9Sq\nVYt169YxdepU1q9fz6pVq6hWrRo5OTlHTIZC1KlThxUrVnDZZZdRWFjIgw8+SF6e8yYSESrKUNba\nUXRUbdKINpOfPvKq2uToB6eIcNZZZ/HNN99wySWX8OCDD7JhwwZCgabS09Pp3r07mzZtYtu2bdx/\n//107tyZDRs2JONy4obNYVYQ5s6dS40aNTjnnHNKPO7bb7/lgw8+4Oabby5LNSkx72TtKCVIibZU\nFFOYRjSkRCM3e96UICXaUlECNyS3uScjFpiyNOKB9TCNaEiJXoG1o+hZ8816tm3aeeR7vcZ1aHlK\ns3hWmRJtqSiB62FGYu3O7xkw+QnW7vw+2aIkjJy9u5k0Yyo5e48NWvDJJ58wYsQIbrvtNrZt23bM\n/r///e/s3LnzmO2hsHCGUZRtm3byWN9nj7zClSeUHt5t5MiRXHXVVcAP7ezOO++scCPGlLDDfGz6\ni3w07yMO6WGevfH+ZIuTEP41byZjp04C4Lfdj04DP3fuXIYNG8bu3bsZPHgw9evXp0uXLsydO5cq\nVapQWFjI0qVLGTNmDMOGDePtt9+matWqx6yYG4ZfSgvv9rOf/Yxt27Yxc+ZMRITCwkJOPvlk5s2b\nV6GmR1LiH9T9ojNofHZDul90RrJFSRg9O3dj8DW/pWfnbsfsy8jIAODrr7/mnHPO4YILLmDVqlX0\n7NmT7OxsWrduTbt27bjwwguZNWsW9913H4WFhceUU5Hx27N59dVXufjii3n11VfjK1CKU1p4t3r1\n6tG1a1c+/vhjCgsLmTlzJps3b+bLL79MtugxJeFzmCLSAbhQVcdF2Bdx7invUB6LcxfRvlaHUsOd\nVQY++ugj3n33XfLz88nNzaV+/fpccsklHH/88cyZM4dTTz2VqlWr8sUXX3DRRRfx7rvvkp2dzeHD\nhxk8eHB5qk6Jeado5zCvvfZaXn/99ThKFHxKm8Pctm0bd9xxBy+//DLdu3fnyiuvZMCAAQwbNoxW\nrVrRtWtXFi5cSPv27bn//vs5++yzue+++3j//ffJzMw8avjukRJtqSgxV5g+8knnisgNkTL+2WR9\n4EmJRm4KMyVIibZUlHgMycuVT3r48OEVbqLYCB5rtu9lwbocFqzLYXd+4ZHPa7bvTbZoRoCJ+aKP\nqu4UkTLlkwZLj2okhlDSL4B1ew4d+fzAuVm0TKJcRrCJ56JPmfNJG4ZhBJG4mBWp6kjvY9T5NWfP\nnm1K0ygVEbkBNz/+Z1Xdn2x5jMpBSpgVGZUPEWkpIlNFJEtEeovInd7rxyJSC/gvkEGR6Z5o2fbp\ndAp272Dbp9NjIrdRsQmcwvTbu8w5sI/Jy2aRcyA108AapVLa4uFQ4CvK2YbrdbqCM4e+Tr1OV5Rb\nYKPiEziF6XeF/KUl7zFm2mReWvJefAUykoKq7gSKXTxU1QGq+u/ihuNmbWHEg5RwjYzExnXLObR6\nKxvXLYeOVyVbHCN+lGnxcPjw4fGSx6jEBE5h+h2SD+r2G6qkH8fvf9QnvgIZScMWD42gETiF6ZfG\ntRow8ud3JVsMwzAqESk7h2mkLiJyjGNxrLHepREPAqcwy9PQ8w7lMW/nh+QdyoudQEZMEJF6IjJE\nRN4ATop3ffbgNeJB4BRmeVicu4j3ts1gce6iZItiHMs8YAnwfkmusbHCephGPAicwixPz6B9rQ5c\nWq877Wt1iJ1ARqxoD7QAOonIr+NdmfUwjXgQuEWf8vQMstKz6FznotgJY8QMVd0nIsuAhUB6vOuz\nHqYRDwLXwywPlTH3T6ogIg8DJ+J6mpfGu7549DAtOrtRoRTmqDf/zpyP5jLqzb8nWxTjWPJVdYqq\n/g2oEu/K4tHD7NWrF/Xr16dXr14xL9tIDQKnMMvTM7jn8hvpcuHF3HP5jbETyIgV2WGfayZNCsMo\nBxVqDrNFnYY8fcO9sRPGiCVviMifcK6O/4p3ZbH09AnPd7M7Zy8LZi8BEpK72wgYgVOYRoWlBTBJ\nVROSRjCWQ/JQzm6AdQc2Hfl8/wu/N4VZyQjckDyWmCF7oGgD/DTZQhhGeUi4whSRG0TkHhGpFu+6\nzJA9UJwCnCoil4nIZckWpjQirYhvK1xHgeazrXBdEiUzkknMh+SlpdnFRcruSzkjZfshZMBuhuyB\n4C/eu5KAFKtlncOstW0jm7/9kovrNeIfGVW4uF4j1uS5kJv1jmtOveOax1hSI5VIRprdmETK9sPB\nQuW7zYUcLLRc58lGVeeo6hzgIu89rpR5DnPLFlbcMIAVNwxg/9crWXHDAA4fOBhT2YzUxZfSEpGm\nfgsMUqTsad/OZ9znbzHt2/klHmdznQnlw2QL4Ic5eoAcDjNHDyRbFCNA+B2S9xKR+sAcVX3b5zll\nTrMbqxXOLo3a8pnOp0ujtke25R3KY3HuItrX6kBWutPnoblOwFwr44SIXA0cBLqLSHtVHZdsmUqi\ni2TShcxki2EEDL/D4g+BA8DFIjK2tINVdaSq5qnqZFV9QlWfKZeUZeT9BfOY89Fc3l8w78i2SAtB\nFrQjIbQB2qjq7ckWxDDKit8eZmvgUaAWUDV+4sTWfq5n525HvUPkhSAL2pEQNuIiFTUH6sS7MktR\nYcQDvz3M84BDwH2qujaO8sQ0aELtGjX5bfdrqF3jB0+8kHIMDceNxKCq/1DVgaq6TlWHxbs+U5ZG\nPIhmpboDbg7KMAyjUuJ3SP4IzktjfBxlAaxnUFERkdG4hUAAVPWeJIpjGGXCbw/zj7g8LH3jKAsQ\n/0jZOQf2MXnZLHIO7ItrPcbRqOrdwH+AZcDOJItjGGXCbw9zMfBSPAUJEe8eZsg2E+CGdpfEtS7j\nGLrh5sL3JFsQwygLfnuYDYAh3iuuxLuH2aVRWzppvaNsM42EUR1nnubbEaI4ROR3InJf+UUyDP/4\nVZjbcT2DzXGUBYh/DzOSbaaRMMYDzwLTSjtQRFqKyFQRyRKR3iJyp/f6sYjUAl4G6sVbYMMIx6/C\nTAPWA3XjKEtC6Nm5G4Ov+e1RtpllxVwq/eMt+ryAiyVwpY9TSotJ8BtgWyKiXiUTyyMULPwqzBrA\nr0hAwOF4D8kj2WaWFQsf5x9v0WcokAf8w8fxpcUkmKiqjyciJkEy2LViJZs/mMvF9RpRy4uatPmD\nuexasTLZolVq/CrAL71X3MP+BMWsKOfAPqZ9O58rT+5I7czqEY+x8HFRczku9N+dwBKf5yQ9JkEy\nyN+4mRU3DGCOHmAz+/lLn9/RRTJpM/lpaNM62eJVWvwqzL24TH89AL/BN1IaP6vp5lIZNTWAM3Au\ntqWiqiO9j5OjrSiVlWU4FgQkWPhVmDW890rzy115csej3o2Y8BTw/4BSA7iUF/MlN+KB3znMPd7r\n4TjKAsR/DtMvtTOrc0O7S4odjhvR4S36/BWnMAclWRzDKBN+e5gPAd8AbUVkQTzd2pLRK8jZu5t/\nzZtJz87dYrIYZETke2AL8JKqbol3Zda7NOKB3x7mZFW9FXijIvoA/2veTMZOncS/5s08sq2oyVAk\nE6JIbpab87YzcsEkNudtpzgqozmSqo4BngN+JCKD411fUEYqRsXCbw+ziog8geslVDgixc0sGoU9\nUlT2SAtDzy+bxhtffwXAQ+f9NmJ9lTXCu6ruEZFaqhr3OUzrYRrxwK/C3Ads8l4VjpBtZjhFTYYi\nmRBFWhjq1+7Ko94jUVnNkbx5zPYiclJFHKmUl1dffZWnn36aAQMGcHG9RskWx4iA3yF5Q2AX0Kq8\nFQbRBzjS0LpooOFIGSgjLQw1yqrLQ+f9lkZZxTtFVdYgxp7x+vREKMtUHJL36tWL+vXr06tXr2SL\nYhSDX4VZD/g5PlwjU9EH2E92Sb8ZKI1SeVVEfh3vSmxIbsQDv0PyN4ExqrrDx7FFfYCHishozxcY\nEbkVzwe4OLe2ROPH5tLsMsuHiJwN9MYZrj+eZHEMo0yU2sMUkfOBZ4AHRGRUacenog+wH5tLs8ss\nN38B5gEzVHVuvCtLxSG5EXz8DMl/ArysqkOinHuqlD7AkQiCGVGyZVDVjsB+4FSvTcSVVGpDa7bv\nZcG6HBasy2F3fiEL1uWQX3Ao2WIZEfAzJK8CHBKRLABVLfUfVx4f4IpIEMyIki2DiGQCJwDvAjYR\nHMa2vEIeXej+Vuv2HOLRhXk8UTXucW6MMuCnh3kQSAfuBu6KrzgVk/a1OnBpve5JNSMKgAyPAGuB\nXKB/vCuzIbkRD0rtYarqiEQIEiIoQ6lYuksGIapRAGTYqaqfAIjIhfGuLCjtyKhYRJOXPCEEpWcQ\nyV3SKBd1RCRNRNKAOskWxjDKQuAUZlB6BrFMZWEAbpV8lPd6IcmyGGXA0mUEUGEGpYcZy1QWBgA/\nwgVvuUtVlyVbGCN6zBMpATl6oiUoPUwj5jQAagIfJqIyCyAcO3atWEn+Rpcw9sCuHDZ/4MxoqzZp\nxPGVLF2G9TB9krN3N5NmTCVn7+7ylxXBdz3StgpGR+B8EbnV8/aKK6YsY0cov9CKGwaw/+uVRz6H\nlGhlInAKM6gNPZaLQJH80iuBr/qtwB3AHOCtJMsSSLZ9Op2C3TvY9un0ZIsSkTl6gBwOM0cPJFuU\npBG4IXlQiRQzs6xE8kuv6L7qqroWQEQeCnNsMMKo1+kK6nW6wvv2dVJliYQlZAtgDzOoQ/JYLgJF\n8kuvRL7q05ItgGGUlcD1MIM6JDfKh4jcCBQA54rIldbLNFKRwPUwjQpLI6Ceqg7GhQA0jJTDFKaR\nKJYCDUWkeSwK87yG4p4byDDCCZzCDOocplE+VPVNVb1fVdep6p9KO76UyP21gctxqZ+NGGMePcUT\nOIVpc5iGR9HI/U8BjVX1fVXNAWoDrUQk4ipZMgJRVxTMo6d4ArfoYxjgIveLSEmR+0uMtTp8+PA4\nSWZUZkxhGkGnTJH7zTUyetZs38u2vEKAI5HfAepa9PcjJFxheuG9nvRWSw2jWCxyf2KJFPkdsOjv\nYcRcYYpIS2AM0Ae4CpfTHOBLYBFwMTZZb8QZ610a8SAeiz7lmqxPJWIZkMOILbbgU3aC7tOeTGKu\nMH2k2Z2sqnerasSwPKm0umlR2YNLRe9hxtP0p16nKzhz6Othfu1GiHiaFVX4NLsWlT24pMpDt6yY\n6U9yiMuiT2WZrA8F5DCCR6o8dI3UwgzXjQpJRe9hGskhcArTGroRC+zBmzgqkyulGa4bRgqxK60q\nG2YvAWB3zl4WeJ/rNa5Dy1OaJUWmXr168dprr1WK+dTAKUzrGRixoKJ6+uzalc+Yu14EYN2BTTzW\n91kA7n/h90lTmJWJwCnMitrQjcRibSj+BLG3G28CpzCtoRtGalAZe7uBW/QxDMMIKoFTmKm+Sm7u\nksEg1dtRKrGtcB0Fms+2wnXJFiXu2JA8xoTcJQEzak8iqd6OUol6xzWn3nExyTwSeAKnMFOdWOYv\nNwwjWNiQPMbEMn+5YRRHZRoGB4nA9TBtKGUYpVOZhsFBInA9TMOIBak+UjGCiSlMo0JiIxUjHgRO\nYVrPwDCMoBI4hWk9A8MwgkrgFKZhGEZQMYVpGIbhk4QrTBHpISI9El1vkDD3yfJj7chIBjFXmCLS\nUkSmikiWiPQWkTu9149FpDZQC6gmIhLrusOJ5eJRrMuKVbbJoF5jLAhKOzKMcJKRl3w30ByoGoe6\njxBUZTJ79uyYZZsM6jXGiEC0I8M4ClWN+Qt4CKgOjPG+P+HzPLVXsF/xaC/WjirnK5FtKVaveLpG\nKlHmJVdVG14ZRbF2ZAQG8Z7IhmEYRimYWZFhGIZPTGEahmH4JBDh3USkN9AQwFsNjfb8q4AzgZbA\nTuAw8Cjwe9wc2HJVnR5FeZcBV3llpZW1LBG5APh/wKkxKOtGoDFwMrC1rGWJSEtgDHAHcGukcoB8\n4BwgU1VH+pGvIiAiw4DRqponIq8Bs1R1omfvuUdV5yZQjkPAAZyVQDdV7SYirYA3VbVtgmR4GSgA\n+gN7gWrAUuAU3H9MgM2q+ny85QkKgVCYQFtVfUhERpfx/C2q+rCIHMKZmjQBeuBW4kZ55fpVTB29\nj9cC7bzyylQW8HNgC/A/4PVylgVQ2zv/7HKUtRvYhHsgjCumnP3e73GviByvqruikDHVGSwiBUCV\nJMqgwFhPcX8KrBeRE4GrgVkJkuFR4EmcUswLk+dJXBsa430fkyB5AkFQFGZ17/1wGc/f5P2Q9+F+\n4HRgP5Dh7T8URVk/BXKBmt755SmrgareLSI5wL/LWVYHVR0kIrm4361MZanqThHZjjP8jlTOYX74\nPSRKGSsCIcUwCajhbauG69UnCgFuFZFs3O8zFfiFJ8feRAigqgUi8l/cPTgVGCAi1YDJuI5AiErV\nPoKiMKMyG4nASOBbIBMYjBvKPAycJSL3A+/5LSg0BBWRM3DD1oNlLQv4j4jch2vwg8tZ1nzvHrUC\n7ixnWQr8ncj3aiaQISIPAQdUtTL5b4abjGwHqnv3pDbugZdIOSZ6ivvnuB7dicAbwI8SKMeeMHme\nVtU8ABG5GtcTL8SNnioNZlZkGIbhE1slNwzD8IkpTMMwDJ+YwjQMw/CJKUzDMAyfmMI0DMPwSUoo\nTBEZJiJZ3ueyGreXte7jRaRzEuodHfb5yPX7PVdEukSKSC4i/UXkEs+bCe+4a8L3l1d242i83+8R\n7/2KYo5pJSIx9+ARkRtF5LRYl1tZCYodph9CHhineA3rRpyB7zPA34APcQbX6cAoYAjO2HgpcALQ\nHZgDnAbcjbM93Avswtm2jQMuwRnqjsPZKO7BGeqeDZwpIhcBHb26WqnqFBG50iszD5jhyZUGjAfu\nAdbiXBqrA18Cl+K8NfK8+lt48r0LXINzS/ynV19TVd2As4MTzzh/O7DAq+c9nH3ePKATzj30z977\nR8DVIvJTnFfGGhGpj7NVPQTUFZEnvHtytYh0Av4P0LB6jdixC2c7e4eInACcDqwA5gNdvP07RWQU\n8LS3bQfu9+4AfI7zPJsO9MZ5Ii0DmgKP4Jw28nBt52GczWZtoBGu/XyViIus6KRED9NjrKo+AXyD\nc+WbALwCdAUWAlNwCmkezjXyc1V9HKeg5gLrVfU5nHH8LcBrQG1VnQBsxrkKjsApw7Nw7oMH+MHr\nZYmqfgg0AC7HGaMDfI1rkGnAFbhGvhlnYP4x8Kx3XJ6qPgusVtWngbbedWwDCoFsnMJ9E6jv1RdS\nWgK0x3nnbML5zG9R1cmefD9V1T/iFPFO7z7sAv7lXWeo59IK2OCV18eTfR/wH9xD4kfe/WlT8k9h\nlIGJqjoeWKyqf8N5VE0BrgTq4R6s4Nr3QmCfqj6Je1jnq+qL3jkKvOUd8zrO17uL974e9xt/69XR\nGKdUFybg+ioFqaIwtcjn/wIDgV/iPFNCLpWh96rAuZ5nzNvAq8BWEbkU+AdwtaouBPJF5BacgnoX\nGApcCKwGjvde1b0694hIO5wSVFU96NXVAdcrbO6V0RgXSGQNcAEwANeYQ7KFX8t/gWbe5z1h+wTX\no2wWds4XXj3NvLLDy5krIkOAbrheTE1P9kNh5YHr0TT3zn0K5/LWBPenHYhT1s2BlRix5h4R+QPu\nwQyAqu7F/U6Lwo47pKrbgeNF5HacsivqXVLUhXgebmTUDFhV5PituFGREQMqnaePiPwG2K6q00Wk\nL25aolBVX/Bxbhou0s+TqrqxlGMzgdtwD6V5qjqv/NIXW9f5uB5oc+ABVS3WJ19E+qvqMyXs/73X\nEzYMowiVTmEahmGUlVQZkhuGYSQdU5iGYRg+MYVpGIbhE1OYhmEYPjGFaRiG4RNTmIZhGD4xhWkY\nhuETU5iGYRg+MYVpGIbhE1OYhmEYPvGlMEXkKhH5SkT2i8gnItI+3oIZhmEEjVIVpoi0wsVJfAEX\n3WYN8G8RkZLOMwzDqGj46WFeB2xV1TGquha42dtmw/kKgIhM8t4nisjlInKWiAzyts0XkbMjnPOl\niJzqfR7tvd8jIo8UOe4tL8r4QyJSU0SGh0V6f1hEHhSRAWHHXyoiTxcpI1R+9bCyjom+LyK3icjd\nRbbdLSLve/VeLCJ/FpERIvJzb9Q0VESe96JQGUap+GkozXCBSQFQ1T2qOl9VD5VwjpE6LBeR5rjA\nx+figtG+KyLdgD8CvUWkhog8KyJ3edHuPwBuF5F0XIT2NFwMzqoikhFW9iFgP/AdLhjzd2H73sQF\ngW4dtu07XDzPcNqIyD24CPhFCQ+19TRu9FP9yE7V0biI8qNUdS4uKnk9r44tqvowLsao7/QfRuXG\nj8LchAsyC4CINBGR39pTucLwX5wi+dj73lJVvwF64iK7n4ELgLwEOFlVl3vfx+HSgABchgtgmw5c\nG1b216o6SlVfVtUDwLqwfU2B24GHQhtUdTUu+jxFywDGet/346LTww+BkQG6q+prQG6R8wVARBrj\n0orcClyZbKTyAAAcBUlEQVQEbPJSfkz2AvkaRqn4UXr/BBp7vYsTcTl0+pYUpNZIHVR1CS7Nx/u4\nvDHpItIQWOOlVHgcNwVzMvCtt+Cnqvo1LicRwKWqeoeq3oFLc1EiIlIdp2z3Az8tKlJJ4uLa470i\nMhSXiiPEKV6E/f3FlLcD+L2IPIjrdY4EcoBLPHkMo1R8BRAWkctxf5zmwCdAP1VdV/JZhmEYFQuL\nuG4YhuETm4c0DMPwiSlMwzAMn5jCNAzD8IkpTMMwDJ+YwjQMw/CJKUzDMAyfmMI0DMPwiSlMwzAM\nn5jCNAzD8IkpTMMwDJ+YwjQMw/CJKUzDMAyfmMI0DMPwyXHJrNyLzj0SGAHkAXfjInA/D9RR1Zle\n5O+zcNG42+AiwE8AfgycDdzmnVcIbFbVF8shzxTgC+A3uMCyj4rIu8DDwI3ALly08TVAfy/+Y6Dx\nYkR+rKoflnBM+O+wQ1Wri0hnXPTy3+DuyfG4WJQP4SKp5wKnAP29c2sB33pF7gVOAOoCN6nqtmLq\nXQVswEV6fx/4PfAHoKFXXwNcvMwXgCqq2qVsd8EfXp6qxap6toi8ClwD/AsXpX0jLnDyauByYAEu\njmgdYAWu/R2HC7w8HliOaycLiWFb8QIh/wX3O/QAagOTVHWZiMzBBXOuhmuzd6nqvhLKagAMANoB\nVwIP4u77LNz/KxeYDlyN+61/p6r5Ucr5HHCmV+4UYJCq3igid3llp8T/KERSFaaqFojIfuBO4EUg\nA/gepwjf8A47XVWf9NIUnAJk4pRmI2A+7k82TlVzvQDHZUJE6gKfqeqfRCQT2C0iFwDbcZHHR6rq\nBhEZASzGKc+4IyKP4/54dXAjgv/gFNQq4AFVPSgivXH3ZCkumrgCbwGX4BLXFYjISTglBrANp8xO\nUNUHi/wO/+cp2YuBm4DTvHtyP055/hmn4FrjUjtchwvE2wD30NsAtMAF6e1VgrJsgYut2gd4z5On\nC+7e3qSqnUVkKi4G6+3Ao+W4jX45BReJvTYu0vzTuOjuBcAy3L0dA3T0ti3CXeteYCruXlfx9hUA\nH+JSc8SyrVyNC/R8OfAS7jdSEbka90BrhVPqu0tSlgCqukVE3sFF0/8w7Hc+Bdfu9+AyLryAy+UV\nTdDwkJydgI9wD5BLge+9/1roviTkfxQrgjAkX4R7Ci9U1cdxP0oDVd3uKa5DInImTmn8Gffj/j9c\nQOOGQHVPWTbANday0h6XP+Zx3B9fcIp7CdDcU5btcEp6P7ClHHVFwzZVfdb7/BROkc3DpZToJCIP\nAM28NA57gZnevh/h7tFK71wF/u59bqiqT+Hy2YRYhLu2xar6BPA/3PWfKiKP4f44LXC9qe24P+Zh\nYBowCPegewY43ZNvADC8hOs6Gzgfp4jOwCmiP+AUZlsR+dS7nlU4hXmPr7tVPs4G3saNWKbiUrN8\niOtpdcfdw4txvcvGwD7cg6oAF/39Iq+Mh73UHOtwbSkmbUVETsb1BtsB+72UHnuB3bieZTNcGo49\n+FdEZ+NSi7QJ+53x5H8G9z/4LTBFVYvmW/Ij5+u4B+oA3EP/GWAY7ndP5P8oJgRBYbbC9Uou8no2\ns4HTPEVwHm4Y3h2XiK03roF8C/wO9ydfJSIP456Aq8shxxm4P+XTuEbSEPjU29dGRAbjnupv44YY\nX0YqJA6EIjwfh0vr8F7YtjWq+ijuqX0vToH9DKeI3gEeACL1uneJyK243nqIVrg/S6gB78fdk7uB\nibi2sg2nSJrh7tVmXE/zEVzv9Te4oflNuBw/a0q4rrOAfsATuJHFdNyooQ7uN+iF+x1e8Y69sISy\nYkUTnLJv49XXArgV19P6Gveg6IK7vqa4e/UB8BNcwrh/4B68fUTkARHJJrZtpaeqDgD+BnQRkTu8\n+n6MG04/ArTyktP5Um64+90U93uGfucunvy/xo1mCnE9/Wjl/DvwJO5e/hdAVdfg7u8XRHFvRORR\nL03OU8XsHxJpezR4/6GSj7GI65UPEemB+wN8p6r/TbY8qYiIZKhqQbLlqCyIyKPAVtxDfDWu07AA\n10NejevNFuI6UafjOgp7cFND5+Mevl/i1kK64h5+oSmqHbi54Aaq+kBJciR1DrM0RORafsgQmA/8\nnyZIw3tzfuGLDPNVdVki6i4L3vxt17BNC1R1aaRjVfWtKMtuh5u3CzFHVVeJyHX8kNY2T1Vfjabc\nInWcjesthXhPVdcXd3wiEZHuuLnaEK+GP2hE5ATgCn7IYrlCVT8mzniZW6/H9dABdqnqf8pY1plA\nh7BN78cqb5eInIMbJYR4V1U3R1nMdm+Odaj3vTFwtqqOEpEauDTM43BrHqO9ZHf5OKXaEPiMH3rJ\nGfwwRXUj0FhVHxORhyiFQCtM3CTxDbhhwlRVnZLAugtxc3DNcUPVotkNg0YBbm6oBW6o/LMYlp2G\nG+41wqXKvcjb3h33hwV41XuVlarAY7jFo1W4qY+goLg/VjrOQuCVIvsPAnfhejk7cHN+iZKrJ26q\n6CBuRbpMChM3//8oTrmsxk3pxIoauPn0+rg59ellKKOuN2TuistpnwEs94biq4HXcItu33nTTQ1x\n7Wg7bgrgJNz9uc3btyms7I3elFvp+lBVA/3C5aN+Dm/6IMF1t8AthJyZ7PvgU95mnrxnxaHsM72y\nW4RtE9xC3MRY/D64odOnuEWppN/PIrJ1x82fdytmf03c8O+qBMuVhpvTfDQGZZ2H64k1ioOcXXCL\nkfXifD8uxq1n9IpH+TaHaRiG4ZMgrJIbhmGkBIGawxQR6+4aRiVBVaX0o47l5//a9BeccT3AN//u\n2fhmABG5GTeXvxK3Ul5dVceIyJU4+96HcPOnnwF/Ba5T1f3R1B0ohQnw85N+H3G7jh6K3P3wUdsG\nHfiI8ZkXHvkcIrRNRw/l9kHdj3wvSvg5Rc+NVH+IcDlC2+Xuh33JE35u0frHZ154ZFtROSKVV9q2\nSNvDZQmnuPsUOi60P3SNRcsr6fugAx8xbvyMI9de2n0vKlu4jEUZN37Gkc+h38DvdUcqr7iybx/U\nnXHjZxz1XpLMka61OFmLky9cxkjXE+k/ESoj/PcqSvj20LWEy1f0e0ltO/yawq87RPh54fL+e9Wz\nlINTONqCJUS+92oLDASmeRYeX+BceJfibG1r4hwzrgL+L5qKbUhuGEaq8Q1ugW2O9znEFlwPsw1u\ngUlwRvTrcPEAXsK5OV+J8+C7iCjx1cP0ltxrAV9rYk17DMMwjiI0BC+Kqoa6u/+OsK+f9zHcmydq\nMzi/Q/KTcHZmz3tGvGtxNmmf41wUp+Cc7LNxNoAFRfZ1xPmLr/YEfkhVF0UrrGEYRjLxOyRfh/OV\n7YQz3FVc2KsTccah5+OCDHyO80Euui8f1/3NwwUFOC9mV2AYhpEgfPUw1UXCAdfLDOfzEk4rbt+v\n/dRpGIYRidlNzzhqlbzrhv+FVslfwOmoHji/8DEicpl37FZVfUVEplOG1fEQtuhjGEaqEVol78IP\nihNcpLNf4FyET/BWyOviwvC1FJHW/LA6XiYC5eljdpiGUXkoqx1mCT3M63HrK3OBUbgpwDm4CEVj\nca69G4CL1YWfK5PQgXk5cY5m1qxZx2yLluLKiEXZ5SHa+hN9HbGSrzxlxur8eN2jrl27aqR2Wxbi\n2daDVq53z5Kuc6J92ZDcKBffffcdK1asiElZu3fvZv78+eTn+0obY6QQFeW3jbunT/uXVp6JizZU\nCLyLMy1qvbhP6xHxrtuIL6+//jqff/45VapUoUWLFpx4YplTKrFnzx4GDRrEeeedx7PPPstf//pX\n0tLseV4RiPTbpiqJcI18ALhrcZ/W89q/tLI/MBqnQI0UZ/78+fTp0wcR4W9/+1u5FOY333xDhw4d\n6NKlCxs3biQnJ4c6derEUFojWaxYsYJzzz2XLl26sGHDBnJzc8tVXs9W/Y+aw/zXd8+U5kses5Xy\nRPmSH27/0krB2WPej3NPMlKcHj16MH78eDIyMujRo0e5yjr99NN55pln2LhxI/n5+Rx//PExktJI\nNmeccQbPPvssGzZs4MCBA9SuXbu8RUbrS17cSnlUfuSQGIX5GD8kV/o3LrugUQHo0qULnTp14vDh\nw1SrVo3Zs2eXuazMzEyef/55du/eTc2aNXEpwo2KQBx+22+K+bwFl/r5KF9yjl4pP4DzI7+YICrM\nxX1aLwE6x7seIzlkZmaWfpBPRIRatWrFrDwjOMTytw0NwYuiJfiSe/QK+1ymdCo2q24Y5WDYsGEh\nkzijEhC4eJhbthyd133v3r3HbIuW4sqIRdnlIdr6E30dsZKvPGXG6vyg3KN4l5UK15nKBE5hNmjQ\n4KjvNWrUOGZbtBRXRizKLg/R1p/o64iVfOUpM1bnB+UexbusVLjOVCbRdpgzcfmbAa5Z3Kd1TPIe\nGxWDgoIC1q1bR/PmzcnIyCj9BMNIMImYwwzZYV6EC/32G1wq1dMTULeRIhQUFHDLLbfwz3/+k1tu\nuYWCgoJki2QYx5CoRZ+QHWYOcAOwdXGf1m8nqG4jBVi3bh2tW7fmiiuuoHXr1qxfvz7ZIhkBRUSe\nE5HzRSRLROaFbR9S5LgbReQEEbk4VnUn2g7zDaAPsLz9SysXLu7T+p0E1G+kAM2bN2flypVMnz6d\nlStX0qxZs2SLZASXlcC5QANcUPOQsuwoIg1xmR4UqAqc5m3vhrPLnA78FshV1XuirTgZdpiPx7tO\nI/XIyMjgueeeY/369TRr1szmMI2SUFwHrClwt7ctA5c+Nx3n1XOy9/0rXNaHVcACnMH6f4HmZanY\n7DCNwJCRkcFJJ51kytIoDcV5+GQAiMiduNS5gss/Jt6+ncBZwGGc988vcesneMdETeDMiswOM/rj\noylnxIgRDBs2rNh94Vx22WUxt8N86623aNu2re8yy1JHLM9LZLlmh+kPVR3rfSxuSu/DIt8/KPK9\npNQ6JRI4hWl2mNEfH005zzzzDBMnTix2XzjXXXddzO0wFy5caHaYcSwrFa4zlbEhuREYCgoK+O67\n78ykyIiIiGR7q+PZyZIh0YbrM4ArcXMK1yzu03pjvOs3UoOQHWbr1q1ZuXIlzz33nM1lGkcQkezu\n3bt/OHDgwNMmTJjwlYhcpKp7vH034VbMuwEPAa1xiz/ZuDw+B1T187CyWgAtVXWO9/1XqvqyHzkS\nbbi+ARdAuCVOgRoGYHaYRqmcPnDgwNMuv/zyjIEDB7YF2oXt24hLeDYT6IhbFMoHlgF1gLYicq+I\n3CMiV+C8DTuIyEgR6eQd74tEBxCuAryJi0N3ITA1QfUbAcfsMI1SWDphwoSvgLYTJkxYjlOGIZrj\nRq0ZOAW5E5itqptEpAdOIYaewFVxnbUsb1t6NEIk2nD9LeBlnJAPJ6BuI0UwO0yjJFR1j4hcNGPG\njHbAstBw3Nv3QgnnveWj+HmlH+KwAMJGYAjZYRpGJDwl+WmpB8YRWyVPMCJy5BXN8bGipGC3RXMw\nF7XLjAXF2YDGAktrUbERkb+IyGzv9ZdkyBA4O8yKbrjev3//qOoPHb9ly5aEX8e5556bUgGE+/fv\nX+y+VDDoNsP1UomY/ExEfgVsB6rhEp7ledkhfa9++yVwCrOiG66HG4f7MQwPHT9x4sSEX0e0RubJ\nDiBcklF+Khh0m+F6uTjBU5K/AaqLyFWU0V+8JBJth/kusBp4YXGf1tXiXbcRG1SVoUOHsmfPHo4/\n/niGDRsW8+Hv4cOHue+++zhw4ACNGjXivvvui2n55UFVGT58ODk5OdSqVYsRI0bY8D85FJctsjmw\nVkSuw+mZA6r6hohUj7UAibbD3IYzKt2cgHqNGPHtt9+SlZVFv379OHz4MBs3xt7fYMmSJTRs2JB+\n/fqxY8cOcnJyYl5HWVm/fj0iQr9+/cjIyGD16tXJFqlSoqo3q2pX73Vz2PbHVPUVVf2nqr6kqq94\n21+JtQyJDiBcHfc0OKH9Syu7J6huo5w0bNiQlStXsmrVKtatW8cJJ5wQ8zqaNWvGsmXLWLVqFVu3\nbqVGjRoxr6Os1K1bl9WrV7Nq1SpWrVpF/fr1ky2SkSQSoTAfA57EJVbPWNyndV9gx+I+rWeUfJoR\nFLKzsxkxYgS7du3ikUceoVq12M+m1K1blz/84Q/s2rWL0aNHc9xxwZlez8rK4pFHHmHXrl2MHDky\nUMrcSCxJscNc3Ke1GdulGM2bN+fGG2+Max0nnXRSYO0wmzZtGvfrN4JPcB7jURCacFdVROSo96L7\ng0a4TLNnz47q+HgTfi8BZs2albC6Y0EQf28jtojIebgAPmnAP1R1eQnH3ghMV9Wdsao/cArTjx1m\nuG1i//79ufXWW/n++++PHBe+v7gyStqeKJIRQLgkQvcydP9iJV+0x8Tj/FSwTzQ7TF90wblV/x64\nSkSuwblbj8GloXjK26e4qOrtRKQubrG5Li4A0E6gCS62xUuqusZv5YFTmH7sMMNtE8M/R9pfXBkl\nbU8UyQggXBKJCCAcTzvMeNabiHLNDtMXqqoHRWQyTvHdArTA+YOvBn4EvAhcBhzC5fT5DVAA5Kjq\neBEZDJwDvOedu8Zv5YF2jSwsLOTNN99k3LhxHDx4MOH1L1u2jOHDh/Ppp0l1X600fPzxxwwfPpzl\ny4sdZSWNr776iuHDh/PJJ58kW5SjOHjwIOPHj2fChAkUFlaKiIkLRGQYMAgYhkuEtpYfQrTNxWWN\nrONtPwun5z4GjheR23FRipZ6566LpvJEG67PxKXZ3QDcvLhP65UlnTt27FjOP/98qlatymOPPRZX\nP+Si5Ofn8+STTzJw4EAmTZpEy5YtadiwYcLqr2xs2rSJKVOmcNNNNzFq1Ciee+45qlSpkmyxADhw\n4ACjR49m4MCBvPjii7Ro0YLGjRsnWywAHn/8cdq3b09hYSGjR4/m/PPPT7ZIcUVV5+KUYlHGhn1+\nIPRBRKoCXVT1MxHprKrjylN/og3XC4C9uKdBqROxO3bsONI4c3Nz4yzm0eTn51OnTh0yMzNp1KgR\nu3btSmj9lY2dO3fSuHFjMjMzqV27Nvn5+ckW6QgHDhygdu3aZGZm0qRJk0C1hdzcXJo2bUrz5s3Z\nuTNmaxsVBlXNV9Vh3uexpR1fGok2XD8AXAN8TQQn+qIMGjSIp556ihdeeIGBAwfGW8ajqF27Nqee\neiqTJk2iWrVq5cp0aJTO6aefTnp6OpMmTeL000+nZs2ayRbpCDVr1qRdu3ZMmjSJ9PR02rVrV/pJ\nCeK2227jhRde4MUXX2TQoEHJFifhJDrPT6IDCL+Ni7K+D/hDaSc2adKEfv360bVr17gKWBz9+vVL\nSr2VlXvuuSfZIhTL7373u2SLEJGWLVsyYcKEI9+/++67JEqTWEQku3Pnzh/27t37tClTphzJ8xMe\npUhEagE1VTUmOU+SYbhukdYNw4gFp/fu3fu0Ll26ZABt582b1w4XYPhEEXkIp9/eARrxQ4qKchE4\ns6Ky2GEWPS9odpihQLzDhg07KijvZZddFig7zO+///6o7wsWLIhavltvvbXExTmzw4xvWalwnTFk\n6ZQpU74C2k6ZMiU8z88qXKLF9d7nhiLSSlXL3/0uGmU7mS8nztHMmjXrmG24RaNjPkfaX1wZJW2P\nNZHkBaKuP9HXURb5Iv2G5SkzVucH5R7Fu6xUuE5VDbWTWOiMbKATkB2L8kp7BdoO04gOVWXPnj0p\n6yKY6vLv27ePQ4cOJVuMMpOK919V96jqpxqWFC2eJMMO8ye4VJc/WdyntdlBxAhV5a677uK4445j\n586dTJgwgczMzGSL5ZuDBw8yYMAA6tSpQ0FBAWPGjEmpIL1jx45l/fr17Nixg9GjRwfFK8Y3hYWF\nDBw4kFq1apGXl8e4ceNISwt+f8pbHT8dWJoIpZloO0zF5Q5OB1Ln35wC7N69m4yMDPr06cNZZ53F\nkiVLki1SVPzvf//jjDPOoE+fPlStWjVQto5+WLt2LX379qV3795Mnz492eJEzTfffMPJJ59Mnz59\nqFu3Lps3Bz/Gt6csPwTmAB+GTIu8HD9xIVGLPofbv7QyDegPDAQuBS7AmRgZMSA7O5utW7cyZ84c\nFi5cyA033JBskaLilFNOYfz48aSnp/P9999Tu3btZIsUFVWrVuW9997jq6++YsiQIckWJ2patmzJ\n0qVLycrKYs2aNakSJPl04DRcJ6wtUHSVfB0uMdo+4CRgpZYzKVqi7TDHAffhepoPJaDuSkNaWhoT\nJ07kiy++4Ne//nWgDL/9kJ2dzfjx41m+fDl9+/ZNieFgOI899hiLFi3il7/8ZUq60GZlZfHnP/+Z\nZcuWcdNNN5GRkZFskfywFBdcoy1QdJUcXIaHQlyQja4AInKKqn5DGUmGHeYT8a6zslK1alU6deqU\nbDHKTK1atVJW/rS0NM4777xki1EusrOzU+r+qzNSvwjXs1wWmsPUyLl8YpLfJ3B2mIZhGH7xlGTC\nwolJkEwIRCQ4whiGEVdUNXXMIDwCpTANwzCCTGrNrBuGYSQRU5iGYRg+MYVpGIbhk0AqTBG5UURO\n83nca1GU+ZqINBeRAeWUb5CIzPQ+XyEi15anvLByJ4pIFRFJF5FJpew/7OceFXeeiIwuo1xHzheR\nFpHupYj0FZF/ikhNEXlcRE70UcctInKlFwz29ihkCz/vDr/3pJQy2kVzf/yWV9z9iqKcM0XkKW9f\n1O2uSHl5JbW18pYnIjeU5bcIOhXBrCiaVSst8l5WcgFEpArQHbijnOWF+CdwFS6Nx7RS9g8uY7mD\ncelHo7kHxZ0fehWlAOek8BowSVVXl1aBqj4nIk/jFiJv9StY2HlplNG8pJgyytxGSiivuPvlt5xD\nUPZ2F36PgR6U3NbKW15qeU74JJA9zAQQC3OGEcD1wC5VjUm6PlWdjXMZ/TFwjENy2P5LAd+BBoo5\nz/c9KOX8SOUI7qGyC/Dl4ygu0kYaUEVE0v3KFnZeBmXsAMSiDB/llXS//JajOHfiMrW78HuM88Eu\ntq3FoLyUMxnyQ1B7mPG2dSpv+YpTBrfgch7Hko1ANVUtLk7YBiAL14uLhqLnRXsPijs/UjlZuB7m\nT4CHRORMVf2ylPLvA/6Kc2UbCgz3KVfovEPAf3D5qKMlFmWUVl5ZRjfHlKOqH3nTBWVpd0Xv8QYg\nq4S2VubyKqpNtdlhGoZh+KSyDskNwzCixhSmYRiGT0xhGoZh+MQUpmEYhk9MYRqGYfgkqGZFRgog\nIu2AG3CmRjnAO6q6rOSzDCN1MbMio8yIyBhVHeJ9vhk4BefhMRh4E5cA71pgKy6dwPXAXOBHwGJg\nIXAekIczdM7DGUHvw3mMnAicAMwAfuHtWwY0BR7B2QHmAa+aojYSgQ3JjfIQbjyfAfwd+AhoDiwC\nDgKfq+rjOC+hTar6LLAWeAE4A2ioqn8CWgCf41IwK3Chqj6BM+RvilPAC4HXvXq7eO/rgVbxvUzD\ncJjCNMrDFBF50svQV4jn64zrLR7C9QbPE5F7gbeBw95+DTtup4gMBL7jh95mC2ChiAwBGuIUbNGh\n0DygBtAMKNVX3TBigQ3JDcMwfGI9TMMwDJ+YwjQMw/CJKUzDMAyfmMI0DMPwiSlMwzAMn5jCNAzD\n8Mn/B7eFjz+O/CxiAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x123bdc690>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plt.figure(figsize=(4, 4), dpi=300)\n",
    "\n",
    "grid = gridspec.GridSpec(2, 2, height_ratios=[2, 1], hspace=0.7, wspace=0.4)\n",
    "\n",
    "gs = {}\n",
    "gs['length'] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,0])\n",
    "gs['fluctuation'] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,1])\n",
    "gs['loh'] = gridspec.GridSpecFromSubplotSpec(7, 1, subplot_spec=grid[1,:], hspace=0)\n",
    "\n",
    "### top left panel ###\n",
    "ax1 = plt.subplot(gs['length'][:])\n",
    "ax1.text(-0.185, 1.055, 'A', transform=ax1.transAxes,\n",
    "         fontsize=9, fontweight='bold', va='top', ha='right')\n",
    "\n",
    "colors = [config.selection['color'][e] for e in loh_length_df.columns]\n",
    "\n",
    "loh_length_df.rename(columns=config.selection['short_label'])\\\n",
    ".plot(ax=ax1, logy=True, color=colors, style='.', marker='o', ms=2., mec='none', legend=False)\n",
    "\n",
    "utils.simple_axes(ax1)\n",
    "ax1.set_xlim(0,1.1E3)\n",
    "ax1.set_xlabel('Homozygosity segment length (kb)')\n",
    "ax1.set_ylabel('Frequency')\n",
    "\n",
    "ax1.legend(frameon=False, loc='upper right', \n",
    "           borderaxespad=0., prop={'size':5},\n",
    "           handlelength=0.75)\n",
    "\n",
    "### top right panel ###\n",
    "ax2 = plt.subplot(gs['fluctuation'][:])\n",
    "ax2.text(-0.2, 1.05, 'B', transform=ax2.transAxes,\n",
    "         fontsize=9, fontweight='bold', va='top', ha='right')\n",
    "\n",
    "colors = [config.background['color'][b] for b in loh_fluctuation_df['LOH rate','mean'].columns] \n",
    "loh_fluctuation_df['LOH rate','mean'].plot(ax=ax2, kind='bar', yerr = loh_fluctuation_df['LOH rate','sem'], \n",
    "                  color=colors, legend=False,\n",
    "                  error_kw=dict(ecolor='0.1', lw=.75, capsize=.75, capthick=.75))\n",
    "\n",
    "utils.simple_axes(ax2)\n",
    "ax2.set_yscale('log')\n",
    "ax2.set_xlabel('Environment')\n",
    "ax2.set_xticklabels(loh_fluctuation_df.index.get_level_values('environment'), minor=False, rotation=0)\n",
    "ax2.set_ylabel('5-FOA+/total 5-FOA')\n",
    "\n",
    "ax2.legend(frameon=False, loc='upper right', \n",
    "           borderaxespad=0., prop={'size':5},\n",
    "           handlelength=0.75)\n",
    "\n",
    "### bottom panel ###\n",
    "ax3 = plt.subplot(gs['loh'][:1])\n",
    "\n",
    "# make a color map of fixed colors\n",
    "cmap = mpl.colors.ListedColormap([config.background['color']['NA/NA'], \n",
    "                                  config.background['color']['WA/NA'], \n",
    "                                  config.background['color']['WA/WA']])\n",
    "bounds=[0,1,2]\n",
    "norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "\n",
    "sdn = seq_dn_df.ix['HU','WAxNA_F12_1_HU_3_T32']\n",
    "\n",
    "indexer = [slice(None)]*len(sdn.index.names)\n",
    "indexer[sdn.index.names.index('type')] = ['consensus']\n",
    "gc = sdn.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "\n",
    "x = gc.columns.get_level_values('pos').values\n",
    "y = np.arange(len(gc.index))\n",
    "data = gc.values\n",
    "\n",
    "plot.heatmap(np.r_[x, x.max()+1], np.r_[y, y.max()+1], data, \n",
    "             ax3, '', '', '', [], [], cmap=cmap, vmin=0, vmax=2)\n",
    "\n",
    "ax3.set_title('WAxNA F12 1 HU 3 T32', fontsize=6, y=3.5)\n",
    "\n",
    "ylabels = ['Consensus']\n",
    "ax3.set_yticks(np.arange(len(ylabels)) + 0.5, minor=True)\n",
    "ax3.set_yticklabels(ylabels, fontweight='bold', va='center', minor=True)\n",
    "\n",
    "# show de novo mutations\n",
    "indexer = [slice(None)]*len(sdn.index.names)\n",
    "indexer[sdn.index.names.index('type')] = ['de novo']\n",
    "gn_all = sdn.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "        \n",
    "labels = gn_all.columns.get_level_values('gene')\n",
    "loc = zip(gn_all.columns.get_level_values('pos'), [-.25]*gn_all.shape[1])\n",
    "\n",
    "for l, xy in zip(labels, loc):\n",
    "    trans = ax3.get_xaxis_transform() # x in data units, y in axes fraction\n",
    "    ann = ax3.annotate(l, xy=xy, xytext=(0, 4), textcoords='offset points',\n",
    "                       arrowprops=dict(arrowstyle='wedge,tail_width=0.7', color='black'),\n",
    "                       fontsize=5, style=('italic' if l!='non-coding' else 'normal'), \n",
    "                       weight=('bold' if l in ['RNR2','RNR4','FPR1','TOR1'] else 'normal'),\n",
    "                       annotation_clip=False, va='bottom', ha='center')\n",
    "\n",
    "ax4 = plt.subplot(gs['loh'][1:])\n",
    "\n",
    "ax4.text(-0.085, 1.8, 'C', transform=ax4.transAxes,\n",
    "         fontsize=9, fontweight='bold', va='top', ha='right')\n",
    "\n",
    "# copy number\n",
    "indexer = [slice(None)]*len(sdn.index.names)\n",
    "indexer[sdn.index.names.index('type')] = ['copy_number']\n",
    "gc = sdn.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "\n",
    "x = gc.columns.get_level_values('pos').values\n",
    "y = np.arange(len(gc.index.get_level_values('clone')))\n",
    "data = gc.values\n",
    "\n",
    "cmap = mpl.colors.ListedColormap(['k','w','k'])\n",
    "bounds = [1,2,3]\n",
    "norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "\n",
    "plot.heatmap(np.r_[x, x.max()+1], np.r_[y, y.max()+1], data, \n",
    "             ax4, '', '', '', [], [], cmap=cmap, vmin=data.min(), vmax=data.max())\n",
    "\n",
    "# LOH\n",
    "indexer = [slice(None)]*len(sdn.index.names)\n",
    "indexer[sdn.index.names.index('type')] = ['loh']\n",
    "gl = sdn.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "\n",
    "x = gl.columns.get_level_values('pos').values\n",
    "y = np.arange(len(gl.index.get_level_values('clone')))\n",
    "data = gl.values\n",
    "\n",
    "# make a color map of fixed colors\n",
    "# cmap = mpl.colors.ListedColormap([background['color']['NA'],\n",
    "#                                   'w',\n",
    "#                                   background['color']['WA']])\n",
    "cmap = mpl.colors.ListedColormap(['k','w','k'])\n",
    "bounds = [-1,0,1]\n",
    "norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "\n",
    "plot.heatmap(np.r_[x, x.max()+1], np.r_[y, y.max()+1], data, \n",
    "             ax4, '', '', '', [], [], cmap=cmap, vmin=data.min(), vmax=data.max())\n",
    "\n",
    "lineage = gl.index.get_level_values('lineage').unique()[0]\n",
    "ax4.tick_params(axis='y', colors=config.lineages[lineage]['fill'], width=5, which='both')\n",
    "\n",
    "# de novo mutations\n",
    "indexer = [slice(None)]*len(sdn.index.names)\n",
    "indexer[sdn.index.names.index('type')] = ['de novo']\n",
    "gn = sdn.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "            \n",
    "for i,(k,g) in enumerate(gn.groupby(level='clone')):\n",
    "    g = g.dropna(axis=1)\n",
    "    x = g.columns.get_level_values('pos').values\n",
    "    y = [i+.5]*len(x)\n",
    "    data = g.values.flatten()\n",
    "    ax4.scatter(x, y, \n",
    "                facecolors=[config.genotype['color'][int(gt)] for gt in data],\n",
    "                edgecolors='k', s=8, zorder=3)\n",
    "\n",
    "ylabels = gl.index.get_level_values('clone')\n",
    "ax4.set_yticks(np.arange(len(ylabels)) + 0.5)\n",
    "ax4.set_yticklabels('C' + ylabels, fontweight='bold', va='center')\n",
    "[ax4.axhline(g, lw=0.5, ls=\"-\", color=\"lightgray\") for g in np.arange(len(ylabels))]\n",
    "\n",
    "xlabels = sgd_length.chr_start + (sgd_length.chr_end - sgd_length.chr_start)/2.\n",
    "xstart = sgd_length.chr_start\n",
    "ax4.set_xticks(xlabels)\n",
    "ax4.set_xticklabels(sgd_length.chr_arabic.apply(utils.int_to_roman))\n",
    "xgrid=[xst+1. for xst in list(set(xstart))]\n",
    "[ax4.axvline(g, lw=0.5, ls=\"-\", color=\"gray\") for g in xgrid]\n",
    "\n",
    "ax4.set_xlabel('Chromosome')\n",
    "\n",
    "# set legend    \n",
    "wt_artist = patches.Rectangle((0,0), width=1, height=1, color=config.background['color']['WA/WA'])\n",
    "hom_artist = patches.Rectangle((0,0), width=1, height=1, color=config.background['color']['NA/NA'])\n",
    "het_artist = patches.Rectangle((0,0), width=1, height=1, color=config.background['color']['WA/NA'])\n",
    "\n",
    "leg1 = ax3.legend([wt_artist,hom_artist,het_artist], ['WA/WA','NA/NA','WA/NA'], \n",
    "                  bbox_to_anchor=(1.1, 0.5), ncol=1,\n",
    "                  frameon=False, loc='center',\n",
    "                  borderaxespad=0, handlelength=0.75, \n",
    "                  title='Background', prop={'size':4})\n",
    "\n",
    "snp_indel_artist = lines.Line2D((0,1),(0,0), mec='k', mfc=config.genotype['color'][0], \n",
    "                                marker='o', markersize=3, linestyle='')\n",
    "loh_artist = patches.Rectangle((0,0), width=1, height=1, color='k')\n",
    "    \n",
    "leg2 = ax4.legend([snp_indel_artist,loh_artist], ['SNP/indel','LOH'], \n",
    "                  bbox_to_anchor=(1.1, 0.6), ncol=1,\n",
    "                  frameon=False, loc='center',\n",
    "                  borderaxespad=0, handlelength=0.75, \n",
    "                  title='Mutation', prop={'size':4})\n",
    "    \n",
    "het_artist = lines.Line2D((0,1),(0,0), mec='k', mfc=config.genotype['color'][1], \n",
    "                          marker='o', markersize=3, linestyle='')\n",
    "hom_artist = lines.Line2D((0,1),(0,0), mec='k', mfc=config.genotype['color'][2],\n",
    "                          marker='o', markersize=3, linestyle='')\n",
    "    \n",
    "leg3 = ax4.legend([het_artist,hom_artist], ['het.','hom.'], \n",
    "                  bbox_to_anchor=(1.1, 0), ncol=1,\n",
    "                  frameon=False, loc='lower center',\n",
    "                  borderaxespad=0, handlelength=0.75, \n",
    "                  title='Genotype', prop={'size':4})\n",
    "\n",
    "ax4.add_artist(leg2)\n",
    "        \n",
    "for leg in [leg1,leg2,leg3]:\n",
    "    plt.setp(leg.get_title(),fontsize=5)\n",
    "\n",
    "# axes limits\n",
    "for ax in fig.get_axes():\n",
    "    ax.xaxis.label.set_size(6)\n",
    "    ax.yaxis.label.set_size(6)\n",
    "    ax.tick_params(axis='both', which='major', size=2, labelsize=6)\n",
    "    ax.tick_params(axis='both', which='minor', size=1, labelsize=4)\n",
    "\n",
    "fig.savefig(dir_paper+'figures/figure3/figure3.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_paper+'figures/figure3/figure3.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fig. 3:** Pervasive loss of heterozygosity. (**A**) The length distribution of homozygous segments, in bins corresponding to 50-kb increments, shows evidence of long homozygosity tracts above 300 kb in hydroxyurea and rapamycin. Ancestral haploid isolates are used to compare a set of *in silico* diploid genomes to evolved diploid isolates. Only unrelated isolate backgrounds were included. (**B**) Background- and environment-dependent rates of loss-of-heterozygosity were measured in a fluctuation assay by loss of the *URA3* marker. 5-FOA+ colonies indicate loss of the marker. The mean number of tandard error of the mean. (**C**) Parental genotype (WA/WA=blue, het=white, NA/NA=red) of six related isolates from population WAxNA F12 1 HU 3, carrying a shared *RNR2* heterozygous mutation and private regions with LOH (black). Chromosome II has undergone extensive LOH in isolates C5 and C6 and these individuals have 15% faster growth rate compared to C1-C4."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure S6 - Genome instability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(4, 18), dpi=300)\n",
    "\n",
    "indexer = [slice(None)]*len(seq_dn_df.index.names)\n",
    "indexer[seq_dn_df.index.names.index('type')] = ['consensus','loh']\n",
    "seq_dn_df_sliced = seq_dn_df.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "\n",
    "shape = pd.DataFrame({k: \n",
    "x.shape for k,x in seq_dn_df_sliced.groupby(level=['selection','population'])\n",
    "})\n",
    "\n",
    "ygrid = shape.ix[0].values\n",
    "grid = gridspec.GridSpec(len(ygrid), 1, height_ratios=ygrid, hspace=0.9)\n",
    "\n",
    "gs = {}\n",
    "\n",
    "for ii, ((e,p), sdf) in enumerate(seq_dn_df.groupby(level=['selection','population'])):\n",
    "\n",
    "    idx = 0\n",
    "    \n",
    "    nrows = ygrid[ii]\n",
    "    ncols = 1\n",
    "    gs[(e,p)] = gridspec.GridSpecFromSubplotSpec(nrows, ncols,\n",
    "                                                 subplot_spec=grid[ii],\n",
    "                                                 hspace=0, wspace=0)\n",
    "    \n",
    "    for jj, (s, jdf) in enumerate(sdf.groupby(level='set')):\n",
    "        \n",
    "        # consensus genotype\n",
    "        indexer = [slice(None)]*len(jdf.index.names)\n",
    "        indexer[jdf.index.names.index('type')] = ['consensus']\n",
    "        gc = jdf.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "        \n",
    "        print e,p,s\n",
    "        ax1 = plt.subplot(gs[(e,p)][idx:idx+len(gc)])\n",
    "\n",
    "        idx += len(gc)\n",
    "        \n",
    "        # set title\n",
    "        ax1.set_title(p)\n",
    "        \n",
    "        if len(gc) > 0:\n",
    "            x = gc.columns.get_level_values('pos').values\n",
    "            y = np.arange(len(gc.index))\n",
    "            data = gc.values\n",
    "            \n",
    "            # make a color map of fixed colors\n",
    "            cmap = mpl.colors.ListedColormap([config.background['color']['NA'], \n",
    "                                              config.background['color']['WAxNA'], \n",
    "                                              config.background['color']['WA']])\n",
    "            bounds = [0,1,2]\n",
    "            norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "            \n",
    "            title = (p.replace('_',' ') if jj==0 else '')\n",
    "            xlabel= ''\n",
    "            ylabel= ''\n",
    "            xticklabels = []\n",
    "            yticklabels = []\n",
    "    \n",
    "            plot.heatmap(np.r_[x, x.max()+1], np.r_[y, y.max()+1], data, \n",
    "                         ax1, '', xlabel, ylabel, xticklabels, yticklabels, cmap=cmap, vmin=data.min(), vmax=data.max())\n",
    "        \n",
    "            ax1.set_title(title, fontsize=6)\n",
    "              \n",
    "        # LOH\n",
    "        indexer = [slice(None)]*len(jdf.index.names)\n",
    "        indexer[jdf.index.names.index('type')] = ['loh']\n",
    "        gl = jdf.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "        \n",
    "        ax2 = plt.subplot(gs[(e,p)][idx:idx+len(gl)], sharex=ax1)\n",
    "        idx += len(gl)\n",
    "        \n",
    "        if len(gl) > 0:\n",
    "            x = gl.columns.get_level_values('pos').values\n",
    "            y = np.arange(len(gl.index))\n",
    "            data = gl.values\n",
    "    \n",
    "            # make a color map of fixed colors\n",
    "#             cmap = mpl.colors.ListedColormap([background['color']['NA'],\n",
    "#                                               'w',\n",
    "#                                               background['color']['WA']])\n",
    "            cmap = mpl.colors.ListedColormap(['k','w','k'])\n",
    "            bounds = [-1,0,1]\n",
    "            norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "\n",
    "            plot.heatmap(np.r_[x, x.max()+1], np.r_[y, y.max()+1], data, \n",
    "                         ax2, '', '', '', [], [], cmap=cmap, vmin=-1, vmax=1)\n",
    "            \n",
    "            lineage = gl.index.get_level_values('lineage').unique()[0]\n",
    "            ax2.tick_params(axis='y', colors=config.lineages[lineage]['fill'], width=5, which='both')\n",
    "\n",
    "        # copy number\n",
    "        indexer = [slice(None)]*len(jdf.index.names)\n",
    "        indexer[jdf.index.names.index('type')] = ['copy_number']\n",
    "        gc = jdf.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "                    \n",
    "        x = gc.columns.get_level_values('pos').values\n",
    "        y = np.arange(len(gc.index.get_level_values('clone')))\n",
    "        data = gc.values\n",
    "        data = np.ma.masked_array(data, data==2)\n",
    "\n",
    "        cmap = mpl.colors.ListedColormap(['r','w','g'])\n",
    "        hatches = ['x','','/']\n",
    "        bounds = [1,2,3]\n",
    "        norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "\n",
    "        plot.heatmap(np.r_[x, x.max()+1], np.r_[y, y.max()+1], data, \n",
    "                     ax2, '', '', '', [], [], cmap=cmap, vmin=1, vmax=3)\n",
    "            \n",
    "        # de novo mutations\n",
    "        indexer = [slice(None)]*len(jdf.index.names)\n",
    "        indexer[jdf.index.names.index('type')] = ['de novo']\n",
    "        gn = jdf.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "        \n",
    "        if len(gn) > 0:\n",
    "            \n",
    "            for i,(k,g) in enumerate(gn.groupby(level='clone')):\n",
    "                g = g.dropna(axis=1)\n",
    "                x = g.columns.get_level_values('pos').values\n",
    "                y = [i+.5]*len(x)\n",
    "                data = g.values.flatten()\n",
    "                ax2.scatter(x, y, \n",
    "                            facecolors=[config.genotype['color'][gt] for gt in data],\n",
    "                            edgecolors='k', s=8, zorder=3)\n",
    "\n",
    "        # set grid\n",
    "        xlabels = sgd_length.chr_start + (sgd_length.chr_end - sgd_length.chr_start)/2.\n",
    "        xstart = sgd_length.chr_start\n",
    "        if ii<(len(seq_loh_df.index)-1):\n",
    "            ax2.set_xticks([])\n",
    "            ax2.set_xticklabels([])\n",
    "        else:\n",
    "            ax2.set_xticks(xlabels, minor=False)\n",
    "            ax2.set_xticklabels(sgd_length.chr_arabic.apply(utils.int_to_roman), minor=False)\n",
    "            ax2.set_xlabel('Chromosome')\n",
    "        xgrid=[xst+1. for xst in list(set(xstart))]\n",
    "        [ax2.axvline(g, lw=0.5, ls=\"-\", color=\"gray\") for g in xgrid]\n",
    "    \n",
    "        ylabels = gl.index.get_level_values('clone')\n",
    "        ax2.set_yticks(np.arange(len(ylabels)) + 0.5, minor=False)\n",
    "        ax2.set_yticklabels('C' + ylabels, fontweight='bold', va='center', minor=False)\n",
    "        [ax2.axhline(g, lw=0.5, ls=\"-\", color=\"lightgray\") for g in np.arange(len(ylabels))]\n",
    "    \n",
    "    # show de novo mutations\n",
    "    indexer = [slice(None)]*len(sdf.index.names)\n",
    "    indexer[sdf.index.names.index('type')] = ['de novo']\n",
    "    gn_all = sdf.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "        \n",
    "    genes = gn_all.columns.get_level_values('gene')\n",
    "    loc = zip(gn_all.columns.get_level_values('pos'), [gn.shape[0]+.25]*gn_all.shape[1])\n",
    "\n",
    "    for l, xy in zip(genes, loc):     \n",
    "        trans = ax2.get_xaxis_transform() # x in data units, y in axes fraction\n",
    "        ann = ax2.annotate(l, xy=xy, xytext=(0, -4), textcoords='offset points',\n",
    "                           arrowprops=dict(arrowstyle='wedge,tail_width=0.7', color='black'),\n",
    "                           fontsize=5, annotation_clip=False, va='top', ha='center',\n",
    "                           style=('italic' if l!='non-coding' else 'normal'), \n",
    "                           weight=('bold' if l in ['RNR2','RNR4','FPR1','TOR1'] else 'normal'))\n",
    "\n",
    "# set panel labels\n",
    "ax1 = plt.subplot(gs['HU','WAxNA_F12_1_HU_1_T32'][(0,0)])\n",
    "ax1.text(-0.05, 4,  chr(ord('A')), transform=ax1.transAxes,\n",
    "         fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "ax1.text(0, 4, 'Selection: %s' % selection['long_label']['HU'], \n",
    "         transform=ax1.transAxes, fontsize=8, va='center', ha='left')\n",
    "\n",
    "ax1 = plt.subplot(gs['RM','WAxNA_F12_1_RM_1_T32'][(0,0)])\n",
    "ax1.text(-0.05, 4,  chr(ord('B')), transform=ax1.transAxes,\n",
    "         fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "ax1.text(0, 4, 'Selection: %s' % selection['long_label']['RM'],\n",
    "         transform=ax1.transAxes, fontsize=8, va='center', ha='left')\n",
    "\n",
    "# set legend\n",
    "for k, xy in zip([('HU','WAxNA_F12_2_HU_1_T32'),('RM','WAxNA_F12_1_RM_3_T32')],[(2,0),(2,0)]):\n",
    "    ax3 = plt.subplot(gs[k][(0,0)])\n",
    "    \n",
    "    wt_artist = patches.Rectangle((0,0), width=1, height=1, color=background['color']['WA/WA'])\n",
    "    het_artist = patches.Rectangle((0,0), width=1, height=1, color=background['color']['WA/NA'])\n",
    "    hom_artist = patches.Rectangle((0,0), width=1, height=1, color=background['color']['NA/NA'])\n",
    "    \n",
    "    leg1 = ax3.legend([wt_artist,hom_artist,het_artist], ['WA/WA','NA/NA','WA/NA'], \n",
    "                      bbox_to_anchor=(1.1, 2.5), ncol=1,\n",
    "                      frameon=False, loc='center',\n",
    "                      borderaxespad=0, handlelength=0.75, \n",
    "                      title='Background', prop={'size':5})\n",
    "    \n",
    "    ax4 = plt.subplot(gs[k][(2,0)])\n",
    "    \n",
    "    snp_indel_artist = lines.Line2D((0,1),(0,0), mec='k', mfc='w', marker='o', markersize=4, linestyle='')\n",
    "    loh_artist = patches.Rectangle((0,0), width=1, height=1, color='k')\n",
    "    cn_gain_artist = patches.Rectangle((0,0), width=1, height=1, color='w', hatch='+')\n",
    "    cn_loss_artist = patches.Rectangle((0,0), width=1, height=1, color='w', hatch='o')\n",
    "    \n",
    "    leg2 = ax4.legend([snp_indel_artist,loh_artist,cn_gain_artist,cn_loss_artist], \n",
    "                      ['SNP/indel','LOH','Copy gain','Copy loss'], \n",
    "                      bbox_to_anchor=(1.1, 0.0), ncol=1,\n",
    "                      frameon=False, loc='center',\n",
    "                      borderaxespad=0, handlelength=0.75, \n",
    "                      title='Mutation', prop={'size':5})\n",
    "    \n",
    "    het_artist = lines.Line2D((0,1),(0,0), mec='k', mfc=genotype['color'][1], \n",
    "                              marker='o', markersize=4, linestyle='')\n",
    "    hom_artist = lines.Line2D((0,1),(0,0), mec='k', mfc=genotype['color'][2],\n",
    "                              marker='o', markersize=4, linestyle='')\n",
    "    \n",
    "    leg3 = ax4.legend([het_artist,hom_artist], ['het.','hom.'], \n",
    "                      bbox_to_anchor=(1.1, -4), ncol=1,\n",
    "                      frameon=False, loc='center',\n",
    "                      borderaxespad=0, handlelength=0.75, \n",
    "                      title='Genotype', prop={'size':5})\n",
    "    \n",
    "    ax4.add_artist(leg2)\n",
    "    \n",
    "    for leg in [leg1,leg2,leg3]:\n",
    "        plt.setp(leg.get_title(),fontsize=6)\n",
    "    \n",
    "# axes limits\n",
    "for ax in fig.get_axes():\n",
    "    ax.xaxis.label.set_size(6)\n",
    "    ax.yaxis.label.set_size(6)\n",
    "    ax.tick_params(axis='both', which='major', size=2, labelsize=6)\n",
    "    ax.tick_params(axis='both', which='minor', size=1, labelsize=4)\n",
    "\n",
    "fig.savefig(dir_supp+'figures/supp_figure_seq_loh/supp_figure_seq_loh.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_supp+'figures/supp_figure_seq_loh/supp_figure_seq_loh.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fig. S6:** Genome instability in sequences of evolved clones. SNPs, indels and chromosome-level aberrations were detected by whole-genome sequencing in single-cell derived clones from evolved populations, after $t = 32$ days in (**A**) hydroxurea or (**B**) rapamycin (see Table S1). Chromosomes are shown on the $x$-axis; clone isolates are listed on the left, colored by lineage (see Fig. \\ref{fig:subclonal_dynamics_cross}).\n",
    "The consensus shows the majority genotype across population isolates with sequence identity greater than 80%. WA/WA (in blue) and NA/NA (in red) represent homozygous diploid genotypes and WA/NA (in purple) represents a heterozygous genotype. Individual cells with shared background genotype carry *de novo* SNPs and indels (circles), *de novo* variation in copy number (???) and *de novo* mis-segregations with loss-of-heterozygosity (blocks). Driver and passenger mutations are listed along the bottom (drivers are in boldface). Populations marked by $\\circ$ indicate cross-contamination during the selection phase, but any derived events are independent."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  },
  "latex_metadata": {
   "affiliation": "Wellcome Trust Sanger Institute, Hinxton, Cambridge CB10 1SA, United Kingdom; Department of Applied Mathematics and Theoretical Physics, University of Cambridge, Cambridge CB3 0WA, United Kingdom",
   "author": "Ignacio Vazquez-Garcia",
   "title": "Background-dependent selection of genetic variation in heterogeneous populations"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
