{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 3 - Loss of heterozygosity (+ Supp. Figs.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# load external dependencies\n",
    "import matplotlib.gridspec as gridspec\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import matplotlib.lines as lines\n",
    "import matplotlib.patches as patches\n",
    "from matplotlib.collections import PatchCollection\n",
    "%matplotlib inline\n",
    "\n",
    "import json, matplotlib\n",
    "s = json.load( open(\"matplotlibrc.json\") )\n",
    "matplotlib.rcParams.update(s)\n",
    "\n",
    "import glob, os, re\n",
    "import numpy as np\n",
    "import pandas as pd # (*) Pandas for data manipulation\n",
    "from scipy import stats\n",
    "\n",
    "# load internal dependencies\n",
    "import colors,plot,stats,utils\n",
    "from config import *\n",
    "from utils import sgd_length\n",
    "from paths import * # absolute file path for data directory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data import"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Length distribution of homozygosity tracts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HU</th>\n",
       "      <th>RM</th>\n",
       "      <th>YPD</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50kb_bin_center</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>0.732980</td>\n",
       "      <td>0.748970</td>\n",
       "      <td>0.750893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>0.139939</td>\n",
       "      <td>0.152117</td>\n",
       "      <td>0.159386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>0.052950</td>\n",
       "      <td>0.055077</td>\n",
       "      <td>0.052793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>0.033283</td>\n",
       "      <td>0.020607</td>\n",
       "      <td>0.021708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>0.012859</td>\n",
       "      <td>0.010116</td>\n",
       "      <td>0.008484</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       HU        RM       YPD\n",
       "50kb_bin_center                              \n",
       "25               0.732980  0.748970  0.750893\n",
       "75               0.139939  0.152117  0.159386\n",
       "125              0.052950  0.055077  0.052793\n",
       "175              0.033283  0.020607  0.021708\n",
       "225              0.012859  0.010116  0.008484"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load data\n",
    "loh_length_df = pd.read_csv(dir_data+'seq/loh/homozygosity_length.tsv', sep='\\t')\n",
    "\n",
    "loh_length_df = loh_length_df.set_index(\"50kb_bin_center\")\n",
    "\n",
    "loh_length_df = loh_length_df.reindex(columns=['HU','RM','YPD'])\n",
    "\n",
    "loh_length_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Examples of loss-of-heterozygosity in clonal genome sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>selection</th>\n",
       "      <th>population</th>\n",
       "      <th>set</th>\n",
       "      <th>clone</th>\n",
       "      <th>lineage</th>\n",
       "      <th>type</th>\n",
       "      <th>chr_arabic</th>\n",
       "      <th>chr_roman</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>pos</th>\n",
       "      <th>genotype</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HU</td>\n",
       "      <td>WAxNA_F12_1_HU_1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>subclone A</td>\n",
       "      <td>consensus</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>2718</td>\n",
       "      <td>32682</td>\n",
       "      <td>48</td>\n",
       "      <td>2718</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16528</th>\n",
       "      <td>RM</td>\n",
       "      <td>WAxNA_F12_1_RM_3</td>\n",
       "      <td>14</td>\n",
       "      <td>C3</td>\n",
       "      <td>subclone A</td>\n",
       "      <td>loh_pm</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>2718</td>\n",
       "      <td>88212</td>\n",
       "      <td>327</td>\n",
       "      <td>2718</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16809</th>\n",
       "      <td>HU</td>\n",
       "      <td>WAxNA_F12_2_HU_1</td>\n",
       "      <td>1</td>\n",
       "      <td>C1</td>\n",
       "      <td>subclone A</td>\n",
       "      <td>loh_pm</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>2718</td>\n",
       "      <td>32682</td>\n",
       "      <td>48</td>\n",
       "      <td>2718</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17098</th>\n",
       "      <td>HU</td>\n",
       "      <td>WAxNA_F12_2_HU_1</td>\n",
       "      <td>2</td>\n",
       "      <td>C2</td>\n",
       "      <td>bulk</td>\n",
       "      <td>loh_pm</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>2718</td>\n",
       "      <td>56480</td>\n",
       "      <td>195</td>\n",
       "      <td>2718</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17401</th>\n",
       "      <td>HU</td>\n",
       "      <td>WAxNA_F12_2_HU_2</td>\n",
       "      <td>4</td>\n",
       "      <td>C1</td>\n",
       "      <td>bulk</td>\n",
       "      <td>loh_pm</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>2718</td>\n",
       "      <td>58539</td>\n",
       "      <td>202</td>\n",
       "      <td>2718</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      selection        population  set clone     lineage       type  \\\n",
       "0            HU  WAxNA_F12_1_HU_1    1        subclone A  consensus   \n",
       "16528        RM  WAxNA_F12_1_RM_3   14    C3  subclone A     loh_pm   \n",
       "16809        HU  WAxNA_F12_2_HU_1    1    C1  subclone A     loh_pm   \n",
       "17098        HU  WAxNA_F12_2_HU_1    2    C2        bulk     loh_pm   \n",
       "17401        HU  WAxNA_F12_2_HU_2    4    C1        bulk     loh_pm   \n",
       "\n",
       "       chr_arabic chr_roman  start    end  length   pos  genotype gene  \n",
       "0               1         I   2718  32682      48  2718         2       \n",
       "16528           1         I   2718  88212     327  2718         0       \n",
       "16809           1         I   2718  32682      48  2718         0       \n",
       "17098           1         I   2718  56480     195  2718         0       \n",
       "17401           1         I   2718  58539     202  2718         0       "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "types = ('*consensus*.txt','*loh_pm*.txt') # the tuple of file types'*consensus*.txt',*gt*.txt',\n",
    "files = []\n",
    "for t in types:\n",
    "    files.extend(glob.glob(dir_data + 'seq/loh/'+t))\n",
    "\n",
    "seq_loh_df = pd.DataFrame()\n",
    "\n",
    "for file_ in files:\n",
    "    \n",
    "    df = pd.read_csv(file_, index_col=None, names=['chr_arabic','start','length','end','genotype'], sep='\\t')\n",
    "    \n",
    "    f = re.split(r'[.-]+', os.path.basename(file_))\n",
    "\n",
    "    df['background'] = f[0].split('_')[0]\n",
    "    df['cross'] = f[0].split('_')[1]\n",
    "    df['cross_rep'] = f[0].split('_')[2]\n",
    "    df['selection'] = f[0].split('_')[3]\n",
    "    df['selection_rep'] = f[0].split('_')[4]\n",
    "    \n",
    "    df['type'] = f[1]\n",
    "    df['set'] = int(f[2].replace('set',''))#int(f[2].replace('set',''))\n",
    "\n",
    "    if f[1]=='consensus':\n",
    "        df['population'] = '_'.join(f[0].split('_')[:-1])\n",
    "        df['clone'] = \"\"\n",
    "    else:\n",
    "        df['population'] = '_'.join(f[0].split('_')[:-2])\n",
    "        df['clone'] = f[0].split('_')[6]\n",
    "    df['lineage'] = f[3]\n",
    "\n",
    "#     if f[1]!='gt':\n",
    "#         df['set'] = int(f[2].replace('set',''))\n",
    "#     else:\n",
    "#         df['set'] = \"\"\n",
    "    \n",
    "    seq_loh_df = seq_loh_df.append(df, ignore_index=True)\n",
    "\n",
    "seq_loh_df = utils.chr_to_gw(seq_loh_df)\n",
    "\n",
    "start = seq_loh_df[['selection','population','set','clone','lineage','type','chr_arabic','chr_roman','start','end','length','pos_start','genotype']]\\\n",
    ".rename(columns={'pos_start':'pos'})\n",
    "end = seq_loh_df[['selection','population','set','clone','lineage','type','chr_arabic','chr_roman','start','end','length','pos_end','genotype']]\\\n",
    ".rename(columns={'pos_end':'pos'})\n",
    "\n",
    "seq_loh_df = pd.concat([start, end]).sort_values('pos')\n",
    "\n",
    "# seq_loh_df = seq_loh_df[seq_loh_df.length > 10]\n",
    "\n",
    "seq_loh_df['lineage'] = seq_loh_df['lineage'].replace('A','subclone A')\n",
    "seq_loh_df['lineage'] = seq_loh_df['lineage'].replace('B','subclone B')\n",
    "seq_loh_df['lineage'] = seq_loh_df['lineage'].replace('C','subclone C')\n",
    "seq_loh_df['lineage'] = seq_loh_df['lineage'].replace('D','subclone D')\n",
    "\n",
    "# dummy\n",
    "seq_loh_df['gene'] = ''\n",
    "\n",
    "seq_loh_df['genotype'] = seq_loh_df['genotype'].astype(int)\n",
    "\n",
    "seq_loh_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ac</th>\n",
       "      <th>af</th>\n",
       "      <th>alt</th>\n",
       "      <th>an</th>\n",
       "      <th>allele</th>\n",
       "      <th>amino_acids</th>\n",
       "      <th>biotype</th>\n",
       "      <th>baseqranksum</th>\n",
       "      <th>cds_position</th>\n",
       "      <th>chr_arabic</th>\n",
       "      <th>...</th>\n",
       "      <th>population</th>\n",
       "      <th>coverage</th>\n",
       "      <th>missing</th>\n",
       "      <th>substitution</th>\n",
       "      <th>assignment</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>consequence_short</th>\n",
       "      <th>chr_roman</th>\n",
       "      <th>pos</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>[A]</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>R/I</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.31</td>\n",
       "      <td>101</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>WAxNA_F12_1_HU_1</td>\n",
       "      <td>28.795032</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>bulk</td>\n",
       "      <td>driver</td>\n",
       "      <td>non-synonymous</td>\n",
       "      <td>VII</td>\n",
       "      <td>4595191</td>\n",
       "      <td>de novo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[C]</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>R/G</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>None</td>\n",
       "      <td>100</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>WAxNA_F12_1_HU_1</td>\n",
       "      <td>27.116307</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>bulk</td>\n",
       "      <td>driver</td>\n",
       "      <td>non-synonymous</td>\n",
       "      <td>VII</td>\n",
       "      <td>4595192</td>\n",
       "      <td>de novo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>[A]</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>R/I</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0.3</td>\n",
       "      <td>101</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>WAxNA_F12_1_HU_2</td>\n",
       "      <td>28.993576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>bulk</td>\n",
       "      <td>driver</td>\n",
       "      <td>non-synonymous</td>\n",
       "      <td>VII</td>\n",
       "      <td>4595191</td>\n",
       "      <td>de novo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[A]</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>R/I</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>None</td>\n",
       "      <td>101</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>WAxNA_F12_1_HU_2</td>\n",
       "      <td>28.719772</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>bulk</td>\n",
       "      <td>driver</td>\n",
       "      <td>non-synonymous</td>\n",
       "      <td>VII</td>\n",
       "      <td>4595191</td>\n",
       "      <td>de novo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>[C]</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>Y/H</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0.289</td>\n",
       "      <td>505</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>WAxNA_F12_1_HU_3</td>\n",
       "      <td>31.647430</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>bulk</td>\n",
       "      <td>driver</td>\n",
       "      <td>non-synonymous</td>\n",
       "      <td>X</td>\n",
       "      <td>6225369</td>\n",
       "      <td>de novo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 80 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  ac   af  alt an allele amino_acids         biotype baseqranksum  \\\n",
       "0  1  0.5  [A]  2      A         R/I  protein_coding         1.31   \n",
       "1  2  1.0  [C]  2      C         R/G  protein_coding         None   \n",
       "2  1  0.5  [A]  2      A         R/I  protein_coding          0.3   \n",
       "3  2  1.0  [A]  2      A         R/I  protein_coding         None   \n",
       "4  1  0.5  [C]  2      C         Y/H  protein_coding        0.289   \n",
       "\n",
       "  cds_position  chr_arabic   ...           population   coverage missing  \\\n",
       "0          101           7   ...     WAxNA_F12_1_HU_1  28.795032     NaN   \n",
       "1          100           7   ...     WAxNA_F12_1_HU_1  27.116307     NaN   \n",
       "2          101           7   ...     WAxNA_F12_1_HU_2  28.993576     NaN   \n",
       "3          101           7   ...     WAxNA_F12_1_HU_2  28.719772     NaN   \n",
       "4          505          10   ...     WAxNA_F12_1_HU_3  31.647430     NaN   \n",
       "\n",
       "   substitution assignment mutation_type consequence_short chr_roman      pos  \\\n",
       "0           C>A       bulk        driver    non-synonymous       VII  4595191   \n",
       "1           T>C       bulk        driver    non-synonymous       VII  4595192   \n",
       "2           C>A       bulk        driver    non-synonymous       VII  4595191   \n",
       "3           C>A       bulk        driver    non-synonymous       VII  4595191   \n",
       "4           T>C       bulk        driver    non-synonymous         X  6225369   \n",
       "\n",
       "      type  \n",
       "0  de novo  \n",
       "1  de novo  \n",
       "2  de novo  \n",
       "3  de novo  \n",
       "4  de novo  \n",
       "\n",
       "[5 rows x 80 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load data\n",
    "seq_mut_df = utils.load_data(dir_data+'seq/de-novo/seq_dn_df_consensus.pickle')\n",
    "\n",
    "seq_mut_df = seq_mut_df.rename(columns={'chrom':'chr_arabic', 'pos':'site'})\n",
    "seq_mut_df['chr_roman'] = seq_mut_df['chr_arabic'].apply(utils.int_to_roman)\n",
    "seq_mut_df = utils.chr_to_gw(seq_mut_df)\n",
    "\n",
    "seq_mut_df = seq_mut_df[(seq_mut_df.time=='32')&\\\n",
    "                        (seq_mut_df.clone!='')].reset_index(drop=True)\n",
    "\n",
    "# Annotate zygosity\n",
    "seq_mut_df['genotype'] = seq_mut_df['genotype'].map({'0/0': 0, '0/1': 1, '1/0': 1, '1/1': 2})\n",
    "\n",
    "seq_mut_df['type'] = 'de novo'\n",
    "\n",
    "seq_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "columns=['selection','population','clone']\n",
    "seq_mut_fixed_df = pd.merge(seq_mut_df, \n",
    "                            seq_loh_df[['selection','population','clone','set','lineage']].drop_duplicates(columns), \n",
    "                            on=columns, how='left')\n",
    "seq_loh_fixed_df = pd.merge(seq_loh_df, \n",
    "                            seq_mut_df[['selection','population','clone']].drop_duplicates(columns), \n",
    "                            on=columns, how='left')\n",
    "\n",
    "seq_dn_df = pd.concat([seq_loh_fixed_df, seq_mut_fixed_df], axis=0).reset_index(drop=True)\n",
    "\n",
    "seq_dn_df['population'] = seq_dn_df['population']+'_T32'\n",
    "\n",
    "seq_dn_df = pd.pivot_table(seq_dn_df, \n",
    "                           index=['selection','population','set','clone','lineage','type'], \n",
    "                           columns=['chr_arabic','chr_roman','pos','gene'])['genotype']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fluctuation assay"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"6\" halign=\"left\">biological_rep</th>\n",
       "      <th colspan=\"4\" halign=\"left\">technical_rep</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"4\" halign=\"left\">Total 5-FOA</th>\n",
       "      <th colspan=\"6\" halign=\"left\">LOH rate</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"3\" halign=\"left\">mean</th>\n",
       "      <th colspan=\"3\" halign=\"left\">sem</th>\n",
       "      <th colspan=\"3\" halign=\"left\">mean</th>\n",
       "      <th>sem</th>\n",
       "      <th>...</th>\n",
       "      <th>mean</th>\n",
       "      <th colspan=\"3\" halign=\"left\">sem</th>\n",
       "      <th colspan=\"3\" halign=\"left\">mean</th>\n",
       "      <th colspan=\"3\" halign=\"left\">sem</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>background</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>...</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>environment</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HU</th>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0.188982</td>\n",
       "      <td>...</td>\n",
       "      <td>237500</td>\n",
       "      <td>37796.447301</td>\n",
       "      <td>459910.705619</td>\n",
       "      <td>86473.571520</td>\n",
       "      <td>0.016617</td>\n",
       "      <td>0.002876</td>\n",
       "      <td>0.004080</td>\n",
       "      <td>0.004539</td>\n",
       "      <td>0.000354</td>\n",
       "      <td>0.000876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RM</th>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0.188982</td>\n",
       "      <td>...</td>\n",
       "      <td>3662500</td>\n",
       "      <td>329628.782009</td>\n",
       "      <td>434293.093511</td>\n",
       "      <td>302924.731340</td>\n",
       "      <td>0.000352</td>\n",
       "      <td>0.000990</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.000021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YPD</th>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0.188982</td>\n",
       "      <td>...</td>\n",
       "      <td>4237500</td>\n",
       "      <td>957396.021956</td>\n",
       "      <td>316712.676795</td>\n",
       "      <td>342228.776865</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>0.000009</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            biological_rep                                            \\\n",
       "                      mean                   sem                       \n",
       "background           WA/WA NA/NA WA/NA     WA/WA     NA/NA     WA/NA   \n",
       "environment                                                            \n",
       "HU                     2.5   2.5   2.5  0.422577  0.422577  0.422577   \n",
       "RM                     2.5   2.5   2.5  0.422577  0.422577  0.422577   \n",
       "YPD                    2.5   2.5   2.5  0.422577  0.422577  0.422577   \n",
       "\n",
       "            technical_rep                          ...    Total 5-FOA  \\\n",
       "                     mean                   sem    ...           mean   \n",
       "background          WA/WA NA/NA WA/NA     WA/WA    ...          WA/NA   \n",
       "environment                                        ...                  \n",
       "HU                    1.5   1.5   1.5  0.188982    ...         237500   \n",
       "RM                    1.5   1.5   1.5  0.188982    ...        3662500   \n",
       "YPD                   1.5   1.5   1.5  0.188982    ...        4237500   \n",
       "\n",
       "                                                          LOH rate            \\\n",
       "                       sem                                    mean             \n",
       "background           WA/WA          NA/NA          WA/NA     WA/WA     NA/NA   \n",
       "environment                                                                    \n",
       "HU            37796.447301  459910.705619   86473.571520  0.016617  0.002876   \n",
       "RM           329628.782009  434293.093511  302924.731340  0.000352  0.000990   \n",
       "YPD          957396.021956  316712.676795  342228.776865  0.000090  0.000209   \n",
       "\n",
       "                                                     \n",
       "                            sem                      \n",
       "background      WA/NA     WA/WA     NA/NA     WA/NA  \n",
       "environment                                          \n",
       "HU           0.004080  0.004539  0.000354  0.000876  \n",
       "RM           0.000044  0.000181  0.000114  0.000021  \n",
       "YPD          0.000038  0.000022  0.000037  0.000009  \n",
       "\n",
       "[3 rows x 42 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read csv file containing the competition assay data\n",
    "# loh_fluctuation_df = pd.ExcelFile(dir_data+'fluctuation/LOH_assay_raw_data_IVG_new.xlsx')\n",
    "# loh_fluctuation_df = loh_fluctuation_df.parse('New', keep_default_na=False, na_values='NaN')\n",
    "loh_fluctuation_df = pd.read_csv(dir_data+'fluctuation/fluctuation_assay.csv')\n",
    "\n",
    "loh_fluctuation_df['5-FOA+'] = loh_fluctuation_df['5FOA']*200/10\n",
    "loh_fluctuation_df['Total 5-FOA'] = loh_fluctuation_df['YPD']*1000000*2/40\n",
    "\n",
    "loh_fluctuation_df['LOH rate'] = loh_fluctuation_df['5-FOA+']/loh_fluctuation_df['Total 5-FOA']\n",
    "\n",
    "from scipy import stats\n",
    "loh_fluctuation_df = loh_fluctuation_df.groupby(['background','environment'])\\\n",
    ".agg([np.mean, stats.sem])\n",
    "\n",
    "loh_fluctuation_df = loh_fluctuation_df.ix[['WA/WA','NA/NA','WA/NA']]\n",
    "loh_fluctuation_df = loh_fluctuation_df.unstack('background')\n",
    "\n",
    "loh_fluctuation_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAUYAAAEWCAYAAAAaWT4HAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJztnXmYFcW1wH9nhhmGTdxAQFkUB0QiiigiIOJCMOAeI88F\nTWJCxBBQCG6RXUU2DUaRuD7BxBg0iWDcUEEUFQERnoqA7JswyACyM8N5f1Tf4XKZ5e63753z+77+\nbi/Vp0531z1d1XXqlKgqhmEYxiGyUq2AYRiG3zDDaBiGEYIZRsMwjBDMMBqGYYRghtEwDCMEM4yG\nYRghRGQYReQzETkoIk8mSiEjc/HKTvCyXUTeFpGGqdbNMIKRcP0YRaQxsBJQoACor+YEaUSAiBwE\nVgAvAtnAacD1wBuqemUqdTOMYKpEkPZ/cEbxVeA64GLg/UQoZWQ0K1R1RGBDRC4FrMZo+IpImtI9\ngPXAfYB42xEhIjeKSN9IzzMyilNEZIi3/As4BleDjAgRaSMi/eKvnmGEWWMUkXzgLGC8qq4QkUXA\nNSLSW1WLS0nfBBgH9ASuAup5hwqAxSJyjKoWxkF/I/04GRgctL0XVy4Oo5wytAiYp6rzRaRlQjU1\nKi3h1hhv8H5bishTOIN6LNCljPQ7cLVLAU5X1ceABl5++WYUKzXvq2q2qmbjjOQuYKKI5IakK7UM\nqer7qro9uSoblY1wDeP1uO+LlwC9gBbe/lKb06q6FdjibVb3fg+q6iRVnVBWJiKiQ4cO1ZkzZ6qX\nny3+WuKKqq4BvseVkVohx0otQyFpJpUm18qR7xffU2FTWkTOAE4HXlLVW4L2rwGuEpEcVT1QxukK\nLBSRe4BV4SjUuXNnOnfuHE5SIz05RUSGeOutgJ/gmsY/lJE+4jIEVo6MGFHVchdgBFAMXBey/xlv\n/5UVyQh3AbR+i5O0/uknqVE2Y8eOVVXVmTNn6rx58w479sgjjyQy61ifb3HIUgj8B2gUq+yQfBJ5\nDzKSlQU/6uerC0uWlQU/Hnb8gQce0N27d2u3bt20uLhY7777bu3QoUPJ8RdffFE/+eQTHTFihL7/\n/vs6atQo3blzpz722GP64osvhmYXt2edqKXCGqOqDgIGlbL/t8BvYzXMRyBxl5iW3PKS66iddPOt\nRxzbsmULo0aN4oMPPqBLly68//771K1blx07drBt2zZ69+7NDTfcwFtvvYWI0LlzZ1avXs2BAwfY\nvn07zZs3Z+vWrVx77bW88MILFBUV0bBhQ9auXcuBAwe46aab+Pvf/86WLVsYO3Ys2dnZcbkmdd8V\nk8LMmTOtxhgBBbuLeHje7pLt+8+pTpOg49WrV+eTTz6hefPmzJkzh5YtW9KsWTPmzp1LrVq1OP30\n05kzZw61atVi8eLF7N+/nxo1aiAirF69OunXEyu+GxL49ycns+HrtalWw9fUqVOHe+65h/vvvx+A\nnJwcNm7cSN++falWrRpNmzalWrVqdOnShQ4dOnDSSSfx/PPPc+utt5Kbm8vKlSvJzc1FVTlw4ADZ\n2dnceOONzJ8/nxNOOIHVq1eTl5fH9u3b2bVrV4qvNjrMKMaX0047jalTp3LbbbfxxBNPMHv2bLZt\n28bbb7/N/PnzadOmDT/88AM1a9YEoGrVqnzzzTcsWbKEr776ih07dqT4CiLDd4bRcEy6+dZSa4uh\nHDx4kPbt23PiiSfy6KOPUlRURFZWFi1atOCtt97ik08+YcWKFQwYMIBnnnmGtm3bkp2dTfPmzZkw\nYQJr164lK8sVgzPOOIN169ZxwgknsHnzZmrXrs3aten5kpo5c2aqVcgoOnbsyJYtW2jZsiVbtmyh\nVatWDBgwgAMHDlBcXIyIsHXrVs4//3xycnLIz8/nnXfe4fHHH+fhhx9m6tSpqb6EiAh7SGAyEBH1\nkz5+Z+/evYwcOZJhw4aFfc748eO5/vrrqV+/fjRZpsWHDitHkbNqy04KdheVbNepXoUmx9dMVHa+\nL0e+M4wzZsywZpB/8X2BBitHaYDvy5HvDKOf9DGOwPcFGqwcpQG+L0f2jdHISOwboxELvjOMVqAr\n5pFHHmH06NEMHDiQdevWxSxvxYoVrF+/Pg6a+QdrRkdG4ZJlbPxgVslSuGTZYcd79erFnDlz+PDD\nD5k1axZXXXUV4L5zT5kyhb59XWyYcePGAXDttdeyatWqpF5DPIkk7FhSsALteHbZMwD8Jv9IV9Hc\n3Fz69+/P4sWLef755yksLKR79+688847ZGVlcdNNNzF9+nRyc3P58ccfqVmzJosWLaJWrVq0bNmS\nq6++mmeffRYRobCwkLZt29K4cWOGDx9O7dq1GT16dLIv10gxe9dvZMktvy/Zbj7pSWief2i7eXOm\nTZtGly5d+Oyzz7j00kvZtGkTX375JapKdnY2ixcvRkTYvHkzl112GVOnTi0xmOmG72qMRsXs37+f\n8ePHM2jQILKysti+fTu7d+/m6quvpmnTpogIVatWZf369eTm5tK3b19atGhBixYt6Nq1K08++SSn\nnXYaIsKDDz5IrVq1EBG6dOkSbW+177CWR/z5xS9+wWuvvcZ7773H9u3bef311/n++++pV68eN954\nIy+//DKqyrRp0/jhhx9YuHBhqlWOGt8ZRivQjt/k/7bU2iK4GmO/fv3o378/b7zxBrVr12bNmjW8\n9tprrF27lhUrVlCtWjW2bdtW4qMY4Nhjj2XJkiV069aNoqIiHnjgAXbvdiMeRIRM6bSwlkdk5J1Y\nn+aTnixZ8k48/AUpIpx55pksXbqUiy66iAceeIB169Yh4vpRsrOz6dq1Kxs2bKCgoID77ruPDh06\nxOVTTyqwXukMYdasWdSsWZOzzz673HTfffcdH3zwAb169YomG9/3JoKVozTA9+XIDKMRCb4v0GDl\nKA3wfTnyXVPaMOKBfZIxYsF3htEKdMV8+umnDBs2jD/84Q8UFBwxKwAvvvgiW7duPWJ/wJWiMmDf\nGCNj1dK1zJ25sGRZtfTwMfKDBg1iz549dO/enYMHD3LPPffQsWPHkuPDhw8vceEJlLO77rorbf/P\naeGuc9555wEwZ86cJGuTOsq75lmzZjFkyBB27NhB//79qVu3LhdeeCGzZs0iNzeXoqIivvrqK8aN\nG8eQIUN48803ycvLO6IjxjACFGzYysjbJpZs3/fc7TRpdmjyxorCjv3sZz+joKCA6dOnIyIUFRVx\n6qmnMnv27LR8SaXFP6Vg72YK9m5OtRq+IScnB4Bvv/2Ws88+m/bt27NixQquvfZaatWqRX5+Pi1b\ntqRjx47MmDGDe++9l6KiogqkGkbZVBR2rE6dOnTu3JlPPvmEoqIipk+fzsaNG1m0aFGqVY+KtDCM\nPSb2osfEqHpR05Y5c+aUWUNu27YtDzzwAP/85z9ZsGABn376Kfn5h5xxjz/+eBYuXEh2djadOnVi\n5MiR5OXlJUv1tOGVV16hU6dOvPLKK6lWxfeUFXasqKiI4uJisrKyEBFuueUWvvjiCxYuXMiDDz5I\nr169+Pjjj1OtfsQkvVdaRNoAHVV1fCnHSu1NLC+atZFUfN+bCJFF17nuuut49dVXE6+Uz1m1dC0F\nGw59l67T4NjDmtJxxvflKO6GMYz5gLeLyC1aygxv5mbhe3xfoCGycmSGMSX4vhwloilt8wEbhpHW\nxL1XWlW3ikhU8wEDDB061Ka+NBJKcLTqHXuLmLtmG5DwqNVGGpFIdx2bD9jwJcEz4q35sbhkPXRm\nPKPykhDDqKrDvdUya4ZlYUbRMIxU4zt3nXT1lDeSh4jcIiJ3i0i1VOtiZCa+M4yGISJNROQ1Eaku\nIjeIyF3ecomI1AbeBXII+XYdKQWfTePAjh8o+GxaXPQ2Moe0GBJoVDpCPRsGicgYz8MBEXkYeI8Y\nX+x12l1BnXZXxKyskXn4rsYYTlO6QcuGNGiZMOdTI8Wo6lagTM8GVf29qv5bVfeUJWPo0KH2WcaI\nGqsxGn4mKs8GcIbRMKIlLWuMG75ey4av11aYzkhfVHW4qu5W1UmqOkpVn4rkfKstGrHgO8NoNUbD\nMFKN7wyjvemNeGAvWCMWfGcYoy3Qt7z0YkkUHsO/iEjHilPFjr1gjVjwnWE0Mg8RqSMiA0TkdeCU\nZORpNUYjFnxnGKN900+6+VaL1+hfZgMLgffLCyAST6zGaMSC7wyjvekzktZAY6CdiNycjAytHBmx\n4Ds/RiPzUNVdIvI1MA/ITrU+hlERvqsxRsspZ57MKWeenGo1jFIQkRHAybia46XJyDMRTWmbI6by\n4DvDaN+GMpK9qvqyqv4vkJuMDBPRlO7Rowd169alR48ecZdt+AvfNaWjLdArFq6MryJGPKkVtH5U\nyrSIkuCJonZs28ncmQuBhE8YZaQQ3xlGIyN5XUT+jBv7/K9kZDhz5sy41RqDJ6Nfs29DyXropPRG\n5mCG0UgGjYHnVTVps69br7QRC777xhhPbDSMb2gOXJZqJcLFOlmMpBpGC0lfaWkGnCYi3USkWzIy\njKUTr6xOloKiNRzQvRQUrYlRO8PvxLUpLSJNgHFAT+AqoJ53aBHOh+1d4DZiDEkfLjYSxjc87f0q\nSZpsPZqmdO2C9Wz8zrX29xVuY+MHswAo2us60utUaUSdKo3ipqPhX+L9jTEpIenD5dllzwDwm/zf\nJiM7owxU9UMAEfmTqj6Uan3KZNMmlvTqB8Ae3cGSW34PQM5fnkylVkYKCMtAichJ4aRLx5D09h0y\nqXyUagXC4UPdxzYO8qHuS7UqRooIt8bYQ0TqAh+q6pthpI86JH3nzp3j1qP4zM3PAvCbOYdqjAEj\naM3s5CEiVwP7ga4i0lpVx6dap/K4UKpyIVVTrYaRQsI1jB8BlwOdRORSVe1fVkJVHe6tJiWKSqyY\ngUwKzYH9qtpPRPqlWhnDqIhwDWM+8DBQG8hLnDrx9T+bM2fOEfvMEKaE9bjIOo2AY5ORYTwdvI3K\nR7idIOcCxcC9qro6gfrYWOkMRFVfUtU+qrpGVYckI08zikYsRNI73Ab3ncgwDCOjCbcp/RBu5MLj\nCdQFsDd9JiIiY3AdcgCo6t0pVMcwKiTcGuODuLk6bkugLkDim9LPLnumxL/RSA6qOhD4D/A1sDXF\n6hhGhYRbY1wATE6kIgGsxpixdMF9p/4x1YoYRkWEW2M8ARjgLWnNMzc/W+LfaCSVGsA+IKzBAuUh\nIr8RkXtjV8kwSidcw7gF97bfmEBdAOuVzmAeByYCUytKKCJNROQ1EakuIjeIyF3ecomI1Ab+BtRJ\ntMJG5SVcw5gFrAWOT6AuQOKb0nPmzCnVv9FIHF7ny3PAIODKME4JHXP/GNBAVd9X1e3Ar4CCyhCl\nyUKgpYZwDWNN4CYssO1h2Djr8PA6XwYBu4GXwkhf0Zj7Car6iJ/G3MebwiXL2PjBLDrVqU/tnFw6\n1anPxg9mUbhkWapVqxSEa+gWeYtWlDBW/DBiwaLyJITLcSHp7gIWhnmOL8bcp4K96zeWRPcJjvTT\nfNKT0Dw/lapVCsI1jDtxs7t1B8IJIhE16VSYbXhhRNQEzsANK62QWMfcp1M5Ko/gSD8XigW2SBbh\nGsaa3m+leDJWU0wIjwHnAY8mIzM/tDzigUX6SQ3hGsaA79mIRCliZC5e50sr3OeY84CBic4zE4yi\nkTrC7XwZDPwP8DcRGZ1AfdL6g7lRJt8D04GxXkdMwrFyZMRCuDXGSao6SUTuVNU/J1KhZL/pzzvv\nPKD0EGVGfFDVcSJSC7hcROqrasKb01ZjNGIh3BpjroiMIkkTGaWa0txwStsXOu76vk9Hct+nI6OS\nn+mo6o9A7WQYRbAaoxEb4RrGXcAGYF0CdUkJ5vCdHLzvjNcl+lNMgHSrMZojt78ItyldD/gBaBpL\nZiLyG+B4VX0kFjmJpjQ3nNL2hfZejzz/vqjlZzqqOlBE+iVrvpd065Xu0aMHU6ZMOWIuayM1hFtj\nrANcQwVDAtNxjGu4YcgsXFlceEVEbk5GRulkFA3/Ea5hfAP4jar+sYJ0NsbVOAIROcv7Rv08sCbV\n+hhGRVTYlBaR84GngPdEpLi86MuqulVEyh3jWlF+Q4cOTepwrnCduc3pOyaexgU73qCqs5KRYbo1\npQ1/EU6N8afA31R1QAQh6SvtGFfjSFS1LbAHOM0rEwnHypARC+F0vuQCxSJSHUBVd5eX2Ma4Hk7A\nLSeZHS6pyLM8RKQqcBzwNvB5itXxFau27KRgdxEAO/YWMXfNNgCOP1CcSrUqPeHUGPcD2bhhXBV9\nY4wZ8z/LSB4CVgPbgd4p1sVXFOwu4uF5u3l43m7W/Fhcsn6gOOGBrIxyqLDGqKrDkqGIn4jnaJhU\n1Nr8UlMMYquqfgogIh2TkaF9YzRiwXeBZ60wZyTHikigdXJsMjK0cmTEgu8Mox/e9DYSJu48DQRG\nvDyXSkUMIxzC9WNMGqk2ikZCuBh4XVX/qKpfp1oZIzoq07BF3xlG63zJSE4Azk+1EkZs9OjRg7p1\n61aKYYu+a0pbjTEjaQvsF5E7IDxH/1jxwyeZTKFwyTL2rnczJ+8r3MbGD5yPft6J9TkmQ+ef8Z1h\n9GuBjmdPdWmTbWX4BFx3eL81cfMHJRw/lqF0pTJOzOW7prQV6MxDVVer6mrg596vkYYET8yV6fiu\nxuhX4tlTXVqtMENriqFMTbUCfqXgs2kc2PEDBZ9No067K1KtTqlUpom5fGcY/dqUNqJHRG4FDgDn\niMiVQcNGDY867a7wrUGsjFhT2kgG9YE6qtofF5rOMHyN7wyjkZF8BdQTkUapVsQwwsEMo5FwVPUN\nVb1PVdfEY5ZJEckSkaRMqmVUTnxnGM3B26hgioyjgcuBpanWM1OpTCNcysJ3htG+MRqUP0XGNuBo\noKmI1EilkplKZRrhUha+M4yGoapbgfKmyJikqgNVdVdZMoYOHWqtDyNqfOeuYxhB2BQZScSiiR/C\nDKPhS2yKjOQTiCYOlEQTBxiVV/miiSe1KW29iUaysGa0EQtxrTGKSBNgHNATuAqo5x1aBMwHOmG9\niYZh+Jx41xitN9HwBZnelE6kS03wuO3KSlwNY2XrTTzvvPNKwpEZ/iJdylC0JNKlpk67K2g16NVK\nPXY7Ud8YrTfRSClWhoxYiHuvdKy9iemETZplGJmJ7xy87U1vxINMb0obicV3htEKtBEP7AWbPDJx\nbLU5eBtGGlGYlce6mQsB2LFtJ3O99ToNjqVJs4Yp0alHjx5MmTIlo8ZW+84w2pveiAeZGgm+sHAv\n4/74AgBr9m1g5G0TAbjvudtTZhgzEWtKGxlJJhpFv1GYlcfcmQuZO3NhSe111dK1qVYrLvjOMFqB\nNoz0oLBwLyNvm8jI2yay5ltXey3YsDXVasUF3xlGwzCMVOM7w2hNaSMeWDlKHgVFazigeykoWpNq\nVeKGdb4YGYmVo+RRp0oj6lTJrHnOfFdjTGds7LRhZAa+M4zWBDKMisnE5qufsKZ0HLGx00ayyMTm\nq5/wXY3RMOKBtTyMWDDDaGQk6dzyMFKP7wxjZXjTWyeNYfgb3xlGe9MbhpFqfNf5UhmwThrD8De+\nqzEahmGkmmTPK91dRLonM08j87ByZCSauBpGEWkiIq+JSHURuUFE7vKWS0TkaKA2UE1EJJ75lkY8\nO3FMVnLxUzkyKifJnld6B9AIyItzvkfgV6NRGWTFAd+UI6OSoqpxXYDBQA1gnLc9KoJz1RZ/L/Eu\nL1aOKueSrHIU7ZKoXmklinmlVdWaRkYwVo6MlCDeG9YwDMPwMHcdwzCMEMwwGoZhhOCLkS8icgNQ\nD8DrgYz0/KuAVkATYCtwEHgYuB33nWqxqk6LQF434CpPVla0skSkPXAecFocZN0KNABOBTZHI0tE\nmgDjgDuBO0qTAewFzgaqqurwcHTLBERkCDBGVXeLyBRghqpO8Pwlf1TVWUnWpRjYh+uZ76KqXUSk\nKfCGqrZIkg5/Aw4AvYGdQDXgK6AZ7j8mwEZVfSbR+iQbXxhGoIWqDhaRMVGev0lVR4hIMc6N40Sg\nO673a7QnN1wD1NZbvQ5o6cmLShZwDbAJ+D/g1RhlARztnX9WlLJ2ABtwRn98GTL2eM/iHhE5RlUL\nI9Av3ekvIgeA3BTrocCjnpH+DFgrIicDVwMzkqTDw8BYnPHbHaTPWFw5Gudtj0uSPknFL4axhvd7\nMMrzN3gP7F7cg8wG9gA53vHiCGRdBmwHjvLOj0XWCao6UES2Af+OUVYbVe0rIttxzy1iWaq6VUS2\n4BykS5NxkEPPQiLULxMI/PmfB2p6+6rhaujJRIA7RKQW7hm9BvzC02VnMhRQ1QMi8i7uPpwG/F5E\nqgGTcC/8ABlZRvxiGCN2yQhhOPAdUBXoj2uCjADOFJH7gPfCFRRoPorIGbgm5/5oZQH/EZF7cQW7\nf4yyPvfuUVPgrhhkKfAipd+n6UCOiAwG9qnqjgj0S3eC3TO2ADW8e3I07qWWbF0meEb6GlwN7WTg\ndeDiJOrxY5A+T6rqbgARuRpXuy7CtYYyDnPXMQzDCMF6pQ3DMEIww2gYhhGCGUbDMIwQzDAahmGE\nYIbRMAwjBN8bRhEZIiLVvfVoHcCjzfsYEemQgnzHBK2XXH+454rIhaVFuBaR3iJykTeyBy/dz4OP\nx6q7cSTeM3zI+72ijDRNRSTuI1pE5FYROT3ecjMdv/gxVkRgREIzr/DcinOCfQr4X+AjnGNyNjAa\nGIBzyv0KOA7oCnwInA4MxPnu7QQKcX5h44GLcM6s43E+fj/inFnPAlqJyAVAWy+vpqr6sohc6cnc\nDbzj6ZUFPA7cDazGDeOrASwCLsWNXNjt5d/Y0+9t4Oe44Xj/9PI7SVXX4XzIxHNg3wLM9fJ5D+fb\nNhtohxsS+Rfv92PgahG5DDdCYZWI1MX5eRYDx4vIKO+eXC0i7YB/ABqUrxFfCnG+p3eKyHHAT4Al\nwOfAhd7xrSIyGnjS2/cD7pm3Ab7AjcSaBtyAG53zNXAS8BBucMNuXPkZgfN5PBqojytD3yTjIjMF\n39cYPR5V1VHAUtwQtieAvwOdgXnAyzjDMxs3HPALVX0EZ4hmAWtV9a84B/LfAVOAo1X1CWAjbojc\nMJzROxM3bG4fh0aBLFTVj4ATgMtxDtsA3+IKXRZwBa4gb8Q5YX8CTPTS7VbVicBKVX0SaOFdRwFQ\nBNTCGdY3gLpefgHjJEBr3GiVDbjx4JtUdZKn32Wq+iDO4G717kMh8C/vOgO1kKbAOk9eT0/3XcB/\ncC+Di73707z8R2FEyQRVfRxYoKr/ixtl9DJwJVAH9xIFV8bnAbtUdSzuxbxXVV/wzlHgv16aV3Fj\nmS/0ftfinvN3Xh4NcMZzXhKuL6NIB8OoIevvAn2A/8GN1AgMIwz85gHneKNE3gReATaLyKXAS8DV\nqjoP2Csiv8MZoreBQUBHYCVwjLfU8PL8UURa4oydqup+L682uFpeI09GA1wwjFVAe+D3uAIb0C34\nWt4FGnrrPwYdE1wNsWHQOV96+TT0ZAfLmSUiA4AuuBrJUZ7uxUHywNVOGnnnPoYb5nUi7o/ZB2eU\nGwHLMBLB3SLyJ9xLGABV3Yl7VvOD0hWr6hbgGBHphzNqoaMwQofOzsa1dhoCK0LSb8a1dIwIqFQj\nX0TkV8AWVZ0mIrfhPiUUqepzYZybhYtMM1ZV11eQtirwB9yLZ7aqzo5d+zLzOh9Xo2wE3K+qZY43\nF5HeqvpUOcdv92q2hlGpqVSG0TAMIxzSoSltGIaRVMwwGoZhhGCG0TAMIwQzjIZhGCGYYTQMwwjB\nDKNhGEYIZhgNwzBCMMNoGIYRghlGwzCMEMwwVmK8aUIRkQkicrmInCkifb19n4vIWaWcs0hETvPW\nx3i/d4vIQyHp/uuF2RosIkeJyNCgcGcjROQBEfl9UPpLReTJEBkB+TWCZB0Rgk5E/iAiA0P2DRSR\n9718O4nIX0RkmIhcIyJXicggEXnGG+ppGIcRVqEQkWyvgK0WkX0i8p1XyHIqPtvwMYtFpBEuKtA5\nuCgtb4tIF+BB4AYRqSkiE0Xkj17Itw+AfiKSjQtTloULXJEXUh6KcXNWL8dFKloedOwNXISk/KB9\ny3FBMIJpLiJ348LAhRI8lvVJXHCNGiUHVcfgwqqNVtVZuLBcdbw8NqnqCFxgjrBjXRqVh3Dflk8D\ng3FhrR4HNuGi0ZQZkMBIC97FGYxPvO0mqroUuBYX3uwMXHSghcCpqrrY2x6Pi3kJ0A0X2SUbuC5I\n9reqOlpV/6aq+4A1QcdOAvrhyhQAqroSF4KNUBnAo972HlyINjgUNQigq6pOAbaHnC8AItIAF0Pz\nDuACYIMX33KSF+HGMA6jQsMoIs2AX+HCVrVT1YGq2gFYAPxSROolWEcjQajqQlxMy/dxgVCzvee5\nyosd+AhwPXAq8J2ItHan6be4QLsAl6rqnap6J2FMBi8iNXBGdQ9wWahK5amLC8J6j4gMwsWdDNDM\nCzO3pwx5PwC3i8gDuFrkcGAbcJGnj2EcRoXRdbyYhU8Bw1R1WMixrPLCXBmGYaQj4TSlj8O9eTeE\nHjCjaBhGJhKOYSzEfaupE7xTROp4E/hkJ0QzwzCMFBGOYfzQ+70h0OvoGcMPcSHZzd3BMIyMokKj\npqrf4OZKOR343PMV+ww3adJIVT2QWBUNwzCSS1hTG3g1xEHALRya7OkJVZ2QUO0MwzBSgM35YhiG\nEYJ9HzQMwwjBDKNhGEYIZhgNwzBCMMNoGIYRghlGwzCMEKqkWgGj8jFs2LBTgV7A7jBPqQZMGjJk\nyNeJ08owDmGG0UgF1wLDhwwZElbIr2HDhmXhQpSZYTSSgi8MoxcstTXQGBfz8QCQq6qTRaQHLn7f\nIqAd8A5u1M0S3GgcARap6vw46HEy8HPgRuAxYDUuJmF9YDPwNq6mM1hVw63tpAwvQvYrqroljLSB\nZ1AXmAzcgIt9uAY4HvdMqgP/AZ4DPvfSfg/kAO2BSbgQZTtVdVR52QHFQff7f4AXcM8eXE2yJfAN\ncBbwyNChQ8O97JjwQqtdBswEPgJeBFYAXb31W3Ajv07HxY9sBPwDWI+7H2uAb4Ffk6By4oVPmw0c\nCxynqk+NtWxsAAAXsElEQVSLyFvA5biXzjZVnR6mrIbAn3BBYhbhnnVd7xoW48rEAS/eZTR6fo4L\nILwLuAP3rLsB5wND/Po/8oVhVNU1ItIHF8lnHq7g1RSRY3A3tAD3R6mFizK9BKgKfIqL/ByvKD9n\nefJq4wxBR2AW7g+6GReQdVEqHqYXb1A9ParhgrIewMVFPB4Xi7CGt2zGPdvjgN+JyEpcINlc4GhV\nHRcqP+gZHIsLRFwdOMbL4ysvv7OAEcC/cPfodODfwB9xz6Q2bgz90WFeVvD9PujpeAyw0stvCdBY\nVfcOGzasTCFx5nTcC/F8XPzHmbiXxG7cPZ2Mu//tcIajBs6QfA60wJWV5SSonHgvsGzcC2gmcI5X\neVgANAV2AF+GK09V14rIfKABrgLwtSd7B+4+dAJmxKBnE1x5+AwX/f123Mu1pl+NIvir8+U7nBHc\nC1yCqz2ejKsdrsUVwK9xRvIY75yLgZviGP6spqr+GxegtTXOwDTE1VjPVtUv4pRPNKwF1uFqZ0W4\nP0WOqr6FMyqKe5453nIwaP9M3FQDNSj/JbIUVxOZhQvkuhD38hGcoQhE2f4prqa0x0tzFO6PdBzQ\nARf0NhwC93sS8DHueW9T1Tc8XQd4+SeTKrjn3QFXHq/CGejPcIb8Vpwh34+79u9xgX5r4wxmVS9i\neaL4g6dPPZwRr+PpuAo3VcQJqloQocxcDj1rcBG1wA0D3oGrMUer5wk4A9sINxNAR1VdFYW8pGJD\nAisRInIdUOwZo5QxbNiwO3DNtiNifJbBMcBFQ4YMGZs4rcLDgjNXDvxUYzwCb/KtlSJS5H0LSWbe\nnURknafDEU1PvyEiF3j6rhKRx0pLo6qvRmMURaSPiBSKyAYRedrbJyKyPuj51I9A5DPAibhPFR2B\njrNnzx79+uuvf/n6668vmDVr1l+CjwHNgL9Eqne88GYTVBEp4vBJvQLHT/XuwSoRWZJEvaqIyDbv\nGewTkdoxyLrJKz+BTyrx1PO3IrLWW34VT9mJwtc1RhGpivsw/LSqrktB/u1w35LGq59vlIeInItr\nAsZdXxH5KXCSqj4ftC8P93yeUtVwa3/l5XENkKeqL8cqK96IyJ3AucDNpd1bEWmCa2Y/nMxQfCJS\nE7gPeCycTrYKZF0JHKWqL8VFucNl98C1Vl4tJ83DuA64E1X1rlKODyjt+3iEetxTQcegS5cG/3fD\nMCoBnmHcjPMEWYn7Pj4X9213Ja4DtggXIPsnuG/SP+I6/M7H81DB1eo7A1tw373BTYh2NO4b7P0V\n6eLrprRhGJWKLar6Z1xHErie8rO8KXTfwrlMfQf8xJs3fAOuM3YuznC+jDOU4DogFediBdBAVcfj\nOncrxBfuOgFExKqvhlEJUFUpZffxnltaZ+ANnHFbLCIDcIZvCnAasFxE7sD1zK/A1QybA6cAf8X1\niNfj8M699SLSnzBtnq+a0iKi15xye5nHdcwgl27gCPru+5jHq3Y87DfA41U7lqz33fcx4x9/57Bz\nAvtLI/jc0Lxl4Igy9QnOrzR9SksbrEeoXqF6hLs/9LqCdQm+XwF9AjoF35vSrjtURmnyA/e5PH1L\n0zE0XSDPYB0B+vXtesR5gXyDjwdk6ZhBh+0LlleWrNLklpc+OK9ggu9ZaLrge1TRdffr2/WI8lya\nrGAZpeUfqkfosdB7GHztZf13SitHwXoG5xfQrV/frly0/quyDGNYiEgnnIHcrqqvRCunPMKzns7S\n1ga+9eOHccMwKg+qOgvna5swwm1Kn4Ib3fCMiHTFOWxm4xx5f4Nr27fDjUzZiBstEXysLW5Ey0rg\nHtxQqZiH8BmGYSSCcA3jGuBBnPGbhfuo+S/gIlwb/3zcKIFluJ6fqaUcuwA3cqUQ5/ZghtEwjFK5\n5l8bnsb5rwIs/fe1DXoBiEgvXMVrGa5nuoaqjvNcjRbggo1MA+YAzwLXq+qeSPMPyzB6vULgao3B\nlDf0q6xjN4eTp2EYlZpmuLgIoez1lhZAH2CqiLTEjQ/vjRvXfyJumOr/4YZL/iPSzM1dxzCMdGIT\nrsbYHPgE57t4raquwY0bn4wLXHElbvz3BdFk4it3HcMwDI+lpa2raqC7/Iihrar6W2/1nqDdUfVa\n+85dJ9U6GIaReGJx10kGvmtKq2rJMmPGjMO2o1nKkxEP+YnSLdz0ibyGeOgXT32jPTeR96hz585H\nlNtkX58fy0J5SzpgTWkjJg4ePEhWVnzer4E/jYivKxNGFET6bGeedMZhvdKd1/1foFf6OVwncHfc\nuOdxItLNS7tZVf8uItOIsjc6QEINY+vJy1rhhugU4aYFWAnkL+iZn7RwzEbi6NdvMFOnLqRhw1ze\nfvvFik8oh88/n88vfzmYoiJhzJjfc9VVP4uTlkaqmTNnHr/85RCKi4WxY8OOaFZWr/RM4BfAQGCK\n1yN9PM4lsImI5BNDb3SARDel7wf+uKBn/gW4cEJjcD6QRpqjqkybtpBVq87mo49q8d57EUe/P4wx\nY55l8eJzWLbsXCZM+GectDT8wNixz/Htt+7ZPvlkzCP4inH+0Kfg9Ujjgk68CZxJjL3RAZLRlD7Y\nevIywfke3YfrSjfSHBGhceNqrFy5ipNP3sTZZ5/Jd999F7W8Sy5px3vv/ZcDB3I499xWcdTUSDUX\nXxx4tlVo2/Ys3n03rNPK6pX+u7eaXco5PYLWY7LAiTaMI4EJuJrpv3HzYxgZwptvvsCMGTM588xW\nnHjiiTEZxttvv5Vzzz2DPXv20LFjhzhqaaSa3r1vpW3bQ8/2wQfvqfCcwDfFVJFQw7igZ/5CXERp\nIwOpVq0a3brF71tgmzZnx02W4S/S7dlar7RhGL7j2qa9D+uV/tfypyoaKx3XnmnfGcZNmzaVrO/c\nufOw7WgoT0Y85MdCpPmXlj6R1xAP/WKVGY9zE3mPBgwYwD/+8Y+4yI9VTz+VhTgQ6VjpuPZM+84w\nnnDCCSXrNWvWPGw7GsqTEQ/5sRBp/qWlT+Q1xEO/WGXG41w/3aNEykqX64yRTUB1QsZK4+Z9WQ48\nCuzD9Ux3wo+GMcSPcTpwhXfo5wt65q9JZN5GerF9+3b69BkMwBNPDKd27ahnAjUyg4jHSnvEpWc6\n0TXGgB/j7NaTlw0EfgX8DvdtwAyjUUKfPoN56aU8b2swkyePT6k+RmoJfFNMFckYKx3wY9wG3AJs\nXtAz/80k5GukJeb/b6SeZPoxvg70BBa3nrxs3oKe+W8lOG8jjXjiieG44MuBdcNIHcn2Y3wkkfkZ\n6Uvt2rWt+Wz4Bt+FHTMMwwAQkb+KyPkiUl1EZgftHxCS7lYROc6bVjUu+M5dx0gsgbBPwXHxRARV\nPSIk1IwZsQWGKI1hw4aVxDA0jApYBpwDnICbiC9gFNuKSD3cLKQK5AGne/u74Fx4pgG/xs09fXek\nGfvOMJqDd2TpI5XRu3dv4PD73Lt3bzZt2lRyLJ76hXLOOedklIN3PGWbg/cRKK5VexIuzBhADm4G\nwGycQ/ep3vY3uBlJVwBzcT6M7wKNosnYd4bRHLwjSx+pjKeeegqACRMmHLZvwoQJJccCXH/99XF3\n8J43b545eCdIVrpcZwQozofxNAARuQs3+98PHAo7loMLaXgmcBDn+N0UN73zyV6aiEmmg/c7uFhp\nB3EO3usTmbeRXpiDtxGKqj7qrZblwfJRyPYHIdvlTe9cLskMVLsOF6i2Cc5QGkYJAQfvl17KKzGQ\nhpEqkungnQvMwI1d7JiEfI20xBy8Kwt5eXkNOnTo8EaHDh3eyMvLa5BqfYJJpoP3f4G/4T6ajkhw\nvkaaYQ7elY82bdo8PXny5O4APXv2fBq4PHBMRH6J643ugisY+TjbUQv4HNinql8EpW8MNFHVD73t\nm1T1b9HqZoFqDV9gDt5GCOtxRnA60BbXwbIXF2PhWOBozzVHgcVAYyBXRC7Bzf8SU9PDd73ShmFU\nDubPn9/Lqykyf/780KARjXAdtTk4Q7gVmKmqG0SkO87wrfXS5uH6Lap7+0qbDyYizDAmmdIcrOOZ\nviJKkxPYF3rsoosuirsz9pAhQ+Iqz0hf9u7du4Gg5nMwqvpcWeep6n/DED+74iRl4zvDmOkO3sEO\n1uHkX1H6RF5DpM7YlTGCtzl4Zya+M4yZ7uAd7GAdTv4VpU/kNUTqjG0RvFMrK12uMx1IpoP328BK\n4LkFPfOrJTJfI77s2LGDX//6bnbs2M1f/jKI5s3z457H1q1b+fWv72HfvgNMnDicxo2jGsmVEHbt\n2sWvfjWQwsKd/PnP99GyZYtUq5QxiMhhk16palIC1IpIbeAoVV1b2vFkRvDuhety35jgPI04c999\nI3nttaOAuvTpM4Lp0yfFPY8BA0bw+uvHATn06TOMadPK/MSUdP70p0eYMqU6UIc+fR5ixoyXUq1S\nJlHqpFci8gBQH3gKuAzXS32G9zsL6Oqt5wPbcXYlH/gWaAW8DHTDddyswg0lzMMNJ9yNmx+mHoc6\ncA4jmQ7eNXA9Tce1nrysaxLyNeLEscfWpkqV3cA+atbMqzB9NBx99FFkZe0G9lCrVvWE5BEtxx13\nDDk57vpr1MhNtTqZxlLcRFYfcvg8Lytxk10prmf6GOAroBBnCAP7vgMKgGq4cdHVPTnZwGe44BK7\nvWMFwI9eXlmU49KTTAfvfy/omX9b68nLLlrQM/+dCs4zfMSQIX/k4MGxbNmyldGjxyQkj1Gj/kRO\nzih27drDqFFDE5JHtNx/fz+Kih5lw4bNjB49LtXqZBRlNZ1DnLO/LiXJyOANEemIm2N6cki6T8rI\nutw5p5Lu4L2gZ/4piczTiD9VqlThoYfuTWgeubm5jB49KKF5REt2djbDhg2sOKGRMlT1Y+DjeMmz\nCN6GYRgh+M5dJxw/xmHDhh3mKBy6HbwvGX6MpeVfFt9//z0Qvh9jRenjcQ0B/YcNG3bY/gEDBpgf\nYxJlmx+jQ0TOxYUozAJeUtXF5aS9FZimqlvjqoSq+mZx6hxixowZWhqh6UK3g/eVJaOiY5FQWv7h\nEGn+paWPxzUE9Md9jC5Z4qFfNGnifW68nnOiZccqK12u0ytv5dmBP+KicfUF7gUewAWd/RcwFjgR\nF4hmuPd7AXANcDcw2jvvZuAeYBAuuEREtsia0oZh+A1V1f3AJOBhYBMuSMRsXGfKxcALwGZgA67n\nuSlwANimqo8DdYGzge+9cyMioYax9eRlrVpPXvZp68nLPmo9edng1pOXLWs9edmM1pOXheUhPGHC\n3zj11Ku5+OIb2Lt3byJVPYJ169bRps2VNGt2Jc88Y35riWbBgoX85CdX0KLFFbz55nupVucwNm7c\nSJs2V9Os2ZVMnBh/H85Y6N37fk499Wouu6wnRUUZE/95rogMwdX8huDmfFnNIfeaWbiJsI719p+J\ns2WfAMeISD+cf+JX3rnl9kCXRjIdvO8DduIursLvAarKxx+vYePG9ixfvor33vuAyy/vlmB1DzFm\nzF/54ovTgWo899x/+e1vb05a3pWRhx+eyNdfnw0I48e/RLdul6ZapRJcWWgOVOeFF97i9ttvSbVK\nAOzdu5f//ncxa9e2ZvXq5XTvvphLL/XPfYsWVZ2FM36hPBq0fn9gRUTygAtVdY6IdFDVmOPXJdPB\nex/wc5xn+hGe7qGICMcf7yILNW68iTPPPCPBah7OT3/aieOPX0JOziqaNaub1LwrIxdeeC5HHbWE\natWW07p1s4pPSCJdu15AnTpLqVJlFfn5dVKtTglVq1alfn0F1tG48RZOOumkVKuUElR1r6oO8dYf\nrSh9OCTTwftN4DXclId/CufkUaPuYteuvZx77tk0bNgwcVqWQvfuXXj//XqsWrWG7t0vS2relZE+\nfX7N2We3YNeuPVx66UWpVucwuna9hPffP4GVK1f7qiyICNOnT+Kdd96jffvzWLZsWapVijt5eXkN\n2rRpUxKz0QtVlnCS7eAd0ZQG1apV42c/+1l8lYqAVq3OoFWr5NZUKzPt25+fahXK5IwzfsIZZ/wk\n1WocwVFHHcUvfnEtQEYaxgqmPyiZvqCioBCR4js/RsMwjDA5WUQG4+zYW7igE5lpGMNx8O7du/dh\n+0O3g/clw8H7+++/L1dOsAN4wIm6Iufz0kiUg3dA/4AzeYC5c+eag3cSZZuD95FUMP3BCtx0zGu9\n9Xoi0lRVl8eccaSOj4lcSFMH74oI1g/PeTqa/BPl4B1JfrGmNwfvxMlKl+ukAgdvPyzJDFQ7Hfgp\nLibaTxf0zI/vEB6DDRs2sGnTJs4666ySuWLSieXLl7N//35atEi/QLBbt25l+fLlnHXWWeTk5KRa\nnahYtWoVu3btomXLlqlWJeUk2l0n4Md4AYfiqmUDVROcb6Xjs8/m0q5dLy64YAy/+11iI+EkgsmT\n/0n79gPp0GEIDz/8eKrViYjNmzfTvv3NXHDBeC6//FepVicqXn11Ku3a3UnHjsMZPDgxoeWiQUQa\niMgb3tIgWfkmy48xC+gNjAI+BdonId9KxauvvsnatS3Ztas5X34ZsaN/ypk6dQabN7eisLAlH374\nZarViYh58+azZEk99u3LZ+nSnRw8eDDVKkXMa6+9y6ZNZ7Jt2+l8/HFp4Q9TxtNAd295OviAiNyU\nqEyT6cc4HjcgXIHBCc630tG7d0/efLMPO3fmcsUVnVOtTsTccceNLFgwkuLibG6+2R8jS8Llggs6\n0rbtRNav30anTieTlZV+IQj69r2F+fOHsn9/FW644fpUqxMugV7pNbgI3ruAU4Blenig24hJth/j\nqETmV5lp2vQUFi2ayr59+6hRo0aq1YmYiy66gG++OY/i4mKqVUuvudJq1arFp5/+m127dlGrVq1U\nqxMV55/flq+/fp2ioiK/3f9eHKopltYrDW7alCLgPaAzgIg0U9WlRInv3HWM6KlSpQpVqqTvI83N\nTd/5VLKystLWKAbIycnxXceRqm4gyKk75NjfS9ld2r6ISb86v2EYRoIR51bkD0TEP8oYhpEwVNXX\n/mS+MoyGYRh+wJrShmEYIZhhNAzDCMGXhlFEbhWR08NMNyUCmVNEpJGI/D5G/fqKyHRv/QoRuS4W\neUFyJ4hIrohki8jzFRw/GOY9OuIcEYloaENZMkSkcWn3UkRuE5F/ishRIvKIiJwcRh6/E5ErReR8\nLzR9uLoFn3dnOPckDBktI71H4cgr636FKaOViDzmHYuqzIXI3F1eWYtVnojcEs2z8Avp69txiEg+\nkmrIb7RsBxCRXKArcGeM8gL8E7gKNwXE1AqO949CZn9AiPz6y5IRWEI5gHPonwI8r6orK8pAVf8q\nIk/ivnvfEa5iQedlAZ+Fe14YMqIuI+XIK+t+hSOjGGIrc8H3GDeSpLyyFqu8o6KR6Rd8WWNMAvHo\nERsG3AgUqmpcZiFS1Zm44ZKXANPKOX4p8GOEMoPPiej6K5BRmizBvTwKgaPDyUNc1IssIFdEssPV\nLei8HKJ80cdDRhjyyrtf4chQXAT8qMtc8D0GPqKcshYHeb7uda4Iv9YYE91VHqt8xf3pfwfEO2rA\neqCaqhaXcXwdUB1XKwuX0HOiuf6yZJQmqzquxvhTYLCItFLVRRXIvxd4FjeCYRAwNEy9AucVA/8B\nopkxLR4yKpIXaWvlCBmq+rHXxI+2zIXe43VA9XLKWtTy0t31ztx1DMMwQqisTWnDMIwyMcNoGIYR\nghlGwzCMEMwwGoZhhGCG0TAMIwS/uusYPkdEWgK34Nx3tgFvqaqvYuIbRrSYu44RFSIyTlUHeOu9\ngGa40Q79gTdwE6FdB2wGvsI5Js8CLgYWAPOAc4HdOGfg3ThH4V240RMnA8cB7wC/8I59DZwEPITz\nodsNvGIG2Yg31pQ2oiXYwTwHeBH4GGgEzAf2A1+o6iO4ETMbVHUisBp4DjgDqKeqfwYaA1/gptZV\noKOqjsI5u5+EM7TzgFe9fC/0ftcCTRN7mUZlxAyjES0vi8hYbzKiIryxvLjaXzGudneuiNwDvAkE\nps7ToHRbRaQPsJxDtcfGwDwRGQDUwxnS0GbNbKAm0BCocBy2YUSKNaUNwzBCsBqjYRhGCGYYDcMw\nQjDDaBiGEYIZRsMwjBDMMBqGYYRghtEwDCMEM4yGYRgh/D9oTlgRA+JDDQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x112fae590>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plt.figure(figsize=(4, 4), dpi=300)\n",
    "\n",
    "grid = gridspec.GridSpec(2, 2, height_ratios=[2, 1], hspace=0.6, wspace=0.4)\n",
    "\n",
    "gs = {}\n",
    "gs['length'] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,0])\n",
    "gs['fluctuation'] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,1])\n",
    "gs['loh'] = gridspec.GridSpecFromSubplotSpec(7, 1, subplot_spec=grid[1,:], hspace=0)\n",
    "\n",
    "### top left panel ###\n",
    "ax1 = plt.subplot(gs['length'][:])\n",
    "ax1.text(-0.175, 1.05, 'A', transform=ax1.transAxes,\n",
    "         fontsize=12, fontweight='bold', va='top', ha='right')\n",
    "\n",
    "colors = [dict_environment['color'][e] for e in loh_length_df.columns]\n",
    "\n",
    "loh_length_df.rename(columns=dict_environment['short_label'])\\\n",
    ".plot(ax=ax1, logy=True, color=colors, style='.', marker='o', ms=2., mec='none', legend=False)\n",
    "\n",
    "utils.simple_axes(ax1)\n",
    "ax1.set_xlim(0,1.1E3)\n",
    "ax1.set_xlabel('Homozygosity segment length (kb)')\n",
    "ax1.set_ylabel('Frequency')\n",
    "\n",
    "ax1.legend(frameon=False, loc='upper right', \n",
    "           borderaxespad=0., prop={'size':5},\n",
    "           handlelength=0.75)\n",
    "\n",
    "### top right panel ###\n",
    "ax2 = plt.subplot(gs['fluctuation'][:])\n",
    "ax2.text(-0.175, 1.05, 'B', transform=ax2.transAxes,\n",
    "         fontsize=12, fontweight='bold', va='top', ha='right')\n",
    "\n",
    "colors = [dict_background['color'][b] for b in loh_fluctuation_df['LOH rate','mean'].columns] \n",
    "loh_fluctuation_df['LOH rate','mean'].plot(ax=ax2, kind='bar', yerr = loh_fluctuation_df['LOH rate','sem'], \n",
    "                  color=colors, legend=False,\n",
    "                  error_kw=dict(ecolor='0.1', lw=.75, capsize=.75, capthick=.75))\n",
    "\n",
    "utils.simple_axes(ax2)\n",
    "ax2.set_yscale('log')\n",
    "ax2.set_xlabel('Environment')\n",
    "ax2.set_xticklabels(loh_fluctuation_df.index.get_level_values('environment'), minor=False, rotation=0)\n",
    "ax2.set_ylabel('5-FOA+/total 5-FOA')\n",
    "\n",
    "ax2.legend(frameon=False, loc='upper right', \n",
    "           borderaxespad=0., prop={'size':5},\n",
    "           handlelength=0.75)\n",
    "\n",
    "### bottom panel ###\n",
    "ax3 = plt.subplot(gs['loh'][:1])\n",
    "\n",
    "# make a color map of fixed colors\n",
    "cmap = mpl.colors.ListedColormap([dict_background['color']['NA'], \n",
    "                                  dict_background['color']['WAxNA'], \n",
    "                                  dict_background['color']['WA']])\n",
    "bounds=[0,1,2]\n",
    "norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "\n",
    "sdn = seq_dn_df.ix['HU','WAxNA_F12_1_HU_3_T32']\n",
    "\n",
    "indexer = [slice(None)]*len(sdn.index.names)\n",
    "indexer[sdn.index.names.index('type')] = ['consensus']\n",
    "gc = sdn.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "\n",
    "x = gc.columns.get_level_values('pos').values\n",
    "y = np.arange(len(gc.index))\n",
    "data = gc.values\n",
    "\n",
    "plot.heatmap(np.r_[x, x.max()+1], np.r_[y, y.max()+1], data, \n",
    "             ax3, '', '', '', [], [], cmap=cmap, vmin=0, vmax=2)\n",
    "\n",
    "ax3.set_title('WAxNA F12 1 HU 3 T32', fontsize=6, y=3)\n",
    "\n",
    "ylabels = ['Consensus']\n",
    "ax3.set_yticks(np.arange(len(ylabels)) + 0.5, minor=True)\n",
    "ax3.set_yticklabels(ylabels, fontweight='bold', va='center', minor=True)\n",
    "\n",
    "# show de novo mutations\n",
    "indexer = [slice(None)]*len(sdn.index.names)\n",
    "indexer[sdn.index.names.index('type')] = ['de novo']\n",
    "gn_all = sdn.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "        \n",
    "labels = gn_all.columns.get_level_values('gene')\n",
    "loc = zip(gn_all.columns.get_level_values('pos'), [-.25]*gn_all.shape[1])\n",
    "\n",
    "for l, xy in zip(labels, loc):\n",
    "    trans = ax3.get_xaxis_transform() # x in data units, y in axes fraction\n",
    "    ann = ax3.annotate(l, xy=xy, xytext=(0, 2), textcoords='offset points',\n",
    "                       arrowprops=dict(arrowstyle='wedge,tail_width=0.7', color='black'),\n",
    "                       fontsize=4, style=('italic' if l!='non-coding' else 'normal'), \n",
    "                       annotation_clip=False, va='bottom', ha='center', \n",
    "                       bbox=(dict(facecolor='none', edgecolor='gray', boxstyle='round,pad=.2')\\\n",
    "                             if l in ['RNR2','RNR4','FPR1','TOR1'] else \\\n",
    "                             dict(facecolor='none', edgecolor='none', boxstyle='round,pad=.2')))\n",
    "\n",
    "ax4 = plt.subplot(gs['loh'][1:])\n",
    "\n",
    "ax4.text(-0.075, 1.65, 'C', transform=ax4.transAxes,\n",
    "         fontsize=12, fontweight='bold', va='top', ha='right')\n",
    "\n",
    "indexer = [slice(None)]*len(sdn.index.names)\n",
    "indexer[sdn.index.names.index('type')] = ['loh_pm']\n",
    "gl = sdn.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "\n",
    "x = gl.columns.get_level_values('pos').values\n",
    "y = np.arange(len(gl.index.get_level_values('clone')))\n",
    "data = gl.values\n",
    "\n",
    "# make a color map of fixed colors\n",
    "# cmap = mpl.colors.ListedColormap([dict_background['color']['NA'],\n",
    "#                                   'w',\n",
    "#                                   dict_background['color']['WA']])\n",
    "cmap = mpl.colors.ListedColormap(['k','w','k'])\n",
    "bounds=[-1,0,1]\n",
    "norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "\n",
    "plot.heatmap(np.r_[x, x.max()+1], np.r_[y, y.max()+1], data, ax4, '', '', '', [], [], cmap=cmap, vmin=-1, vmax=1)\n",
    "\n",
    "lineage = gl.index.get_level_values('lineage').unique()[0]\n",
    "ax4.tick_params(axis='y', colors=dict_lineages[lineage]['fill'], width=5, which='both')\n",
    "\n",
    "# show de novo mutations\n",
    "indexer = [slice(None)]*len(sdn.index.names)\n",
    "indexer[sdn.index.names.index('type')] = ['de novo']\n",
    "gn = sdn.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "            \n",
    "for i,(k,g) in enumerate(gn.groupby(level='clone')):\n",
    "    g = g.dropna(axis=1)\n",
    "    x = g.columns.get_level_values('pos').values\n",
    "    y = [i+.5]*len(x)\n",
    "    data = g.values.flatten()\n",
    "    ax4.scatter(x, y, \n",
    "                facecolors=[dict_genotype['color'][int(gt)] for gt in data],\n",
    "                edgecolors='k', s=8, zorder=3)\n",
    "\n",
    "ylabels = gl.index.get_level_values('clone')\n",
    "ax4.set_yticks(np.arange(len(ylabels)) + 0.5)\n",
    "ax4.set_yticklabels(ylabels, fontweight='bold', va='center')\n",
    "[ax4.axhline(g, lw=0.5, ls=\"-\", color=\"lightgray\") for g in np.arange(len(ylabels))]\n",
    "\n",
    "xlabels = sgd_length.chr_start + (sgd_length.chr_end - sgd_length.chr_start)/2.\n",
    "xstart = sgd_length.chr_start\n",
    "ax4.set_xticks(xlabels)\n",
    "ax4.set_xticklabels(sgd_length.chr_arabic.apply(utils.int_to_roman))\n",
    "xgrid=[xst+1. for xst in list(set(xstart))]\n",
    "[ax4.axvline(g, lw=0.5, ls=\"-\", color=\"gray\") for g in xgrid]\n",
    "\n",
    "ax4.set_xlabel('Chromosome')\n",
    "\n",
    "# set legend    \n",
    "wt_artist = patches.Rectangle((0,0), width=1, height=1, color=dict_background['color']['WA/WA'])\n",
    "hom_artist = patches.Rectangle((0,0), width=1, height=1, color=dict_background['color']['NA/NA'])\n",
    "het_artist = patches.Rectangle((0,0), width=1, height=1, color=dict_background['color']['WA/NA'])\n",
    "\n",
    "leg1 = ax3.legend([wt_artist,hom_artist,het_artist], ['WA/WA','NA/NA','WA/NA'], \n",
    "                  bbox_to_anchor=(1.075, 0), ncol=1,\n",
    "                  frameon=False, loc='center',\n",
    "                  borderaxespad=0, handlelength=0.75, \n",
    "                  title='Background', prop={'size':4})\n",
    "\n",
    "snp_indel_artist = lines.Line2D((0,1),(0,0), mec='k', mfc=dict_genotype['color'][1], \n",
    "                                marker='o', markersize=3, linestyle='')\n",
    "loh_artist = patches.Rectangle((0,0), width=1, height=1, color='k')\n",
    "    \n",
    "leg2 = ax4.legend([snp_indel_artist,loh_artist], ['SNP/indel','rearrangement'], \n",
    "                  bbox_to_anchor=(1.075, 0.55), ncol=1,\n",
    "                  frameon=False, loc='center',\n",
    "                  borderaxespad=0, handlelength=0.75, \n",
    "                  title='Mutation', prop={'size':4})\n",
    "    \n",
    "het_artist = lines.Line2D((0,1),(0,0), mec='k', mfc=dict_genotype['color'][1], \n",
    "                          marker='o', markersize=3, linestyle='')\n",
    "hom_artist = lines.Line2D((0,1),(0,0), mec='k', mfc=dict_genotype['color'][2],\n",
    "                          marker='o', markersize=3, linestyle='')\n",
    "    \n",
    "leg3 = ax4.legend([het_artist,hom_artist], ['het.','hom.'], \n",
    "                  bbox_to_anchor=(1.075, 0), ncol=1,\n",
    "                  frameon=False, loc='lower center',\n",
    "                  borderaxespad=0, handlelength=0.75, \n",
    "                  title='Genotype', prop={'size':4})\n",
    "\n",
    "ax4.add_artist(leg2)\n",
    "        \n",
    "for leg in [leg1,leg2,leg3]:\n",
    "    plt.setp(leg.get_title(),fontsize=5)\n",
    "\n",
    "# axes limits\n",
    "for ax in fig.get_axes():\n",
    "    ax.xaxis.label.set_size(6)\n",
    "    ax.yaxis.label.set_size(6)\n",
    "    ax.tick_params(axis='both', which='major', size=2, labelsize=6)\n",
    "    ax.tick_params(axis='both', which='minor', size=1, labelsize=4)\n",
    "\n",
    "fig.savefig(dir_paper+'figures/figure3/figure3.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_paper+'figures/figure3/figure3.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fig. 4: Pervasive loss of heterozygosity. (A) The length distribution of homozygous segments, in bins corresponding to 25-kb increments, shows evidence of long homozygosity tracts above 300 kb in hydroxyurea and rapamycin. Ancestral haploid isolates are used to compare a set of *in silico* diploid genomes to evolved diploid isolates. Only unrelated isolate backgrounds were included. (B) Background- and environment-dependent rates of loss-of-heterozygosity were measured in a fluctuation assay by loss of the *URA3* marker. 5-FOA+ colonies indicate loss of the marker. The mean number of tandard error of the mean. (C) Parental genotype (WA/WA=blue, het=white, NA/NA=red) of six related isolates from population WAxNA F12 1 HU 3, carrying a shared *RNR2* heterozygous mutation and private regions with LOH (black). Chromosome II has undergone extensive LOH in isolates C5 and C6 and these individuals have 15% faster growth rate compared to C1-C4."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(4, 14), dpi=300)\n",
    "\n",
    "indexer = [slice(None)]*len(seq_dn_df.index.names)\n",
    "indexer[seq_dn_df.index.names.index('type')] = ['consensus','loh_pm']\n",
    "seq_dn_df_sliced = seq_dn_df.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "\n",
    "shape = pd.DataFrame({k: \n",
    "x.shape for k,x in seq_dn_df_sliced.groupby(level=['selection','population'])\n",
    "})\n",
    "\n",
    "ygrid = shape.ix[0].values\n",
    "grid = gridspec.GridSpec(len(ygrid), 1, height_ratios=ygrid, hspace=0.7)\n",
    "\n",
    "gs = {}\n",
    "\n",
    "###  ###\n",
    "\n",
    "for ii, ((e,p), sdf) in enumerate(seq_dn_df.groupby(level=['selection','population'])):\n",
    "\n",
    "    idx = 0\n",
    "    \n",
    "    nrows = ygrid[ii]\n",
    "    ncols = 1\n",
    "    gs[(e,p)] = gridspec.GridSpecFromSubplotSpec(nrows, ncols,\n",
    "                                                 subplot_spec=grid[ii],\n",
    "                                                 hspace=0, wspace=0)\n",
    "    \n",
    "    for jj, (s, jdf) in enumerate(sdf.groupby(level='set')):\n",
    "        \n",
    "        indexer = [slice(None)]*len(jdf.index.names)\n",
    "        indexer[jdf.index.names.index('type')] = ['consensus']\n",
    "        gc = jdf.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "        \n",
    "        print e,p,s\n",
    "        ax1 = plt.subplot(gs[(e,p)][idx:idx+len(gc)])\n",
    "\n",
    "        idx += len(gc)\n",
    "        \n",
    "        # set title\n",
    "        ax1.set_title(p)\n",
    "        \n",
    "        if len(gc) > 0:\n",
    "            x = gc.columns.get_level_values('pos').values\n",
    "            y = np.arange(len(gc.index))\n",
    "            data = gc.values\n",
    "            \n",
    "            # make a color map of fixed colors\n",
    "            cmap = mpl.colors.ListedColormap([dict_background['color']['NA'], \n",
    "                                              dict_background['color']['WAxNA'], \n",
    "                                              dict_background['color']['WA']])\n",
    "            bounds = [0,1,2]\n",
    "            norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "            \n",
    "            title = (p.replace('_',' ') if jj==0 else '')\n",
    "            xlabel= ''\n",
    "            ylabel= ''\n",
    "            xticklabels = []\n",
    "            yticklabels = []\n",
    "    \n",
    "            plot.heatmap(np.r_[x, x.max()+1], np.r_[y, y.max()+1], data, \n",
    "                         ax1, '', xlabel, ylabel, xticklabels, yticklabels, cmap=cmap, vmin=0, vmax=2)\n",
    "        \n",
    "            ax1.set_title(title, fontsize=6)\n",
    "\n",
    "        indexer = [slice(None)]*len(jdf.index.names)\n",
    "        indexer[jdf.index.names.index('type')] = ['loh_pm']\n",
    "        gl = jdf.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "        \n",
    "        ax2 = plt.subplot(gs[(e,p)][idx:idx+len(gl)], sharex=ax1)\n",
    " \n",
    "        idx += len(gl)\n",
    "        \n",
    "        if len(gl) > 0:\n",
    "            x = gl.columns.get_level_values('pos').values\n",
    "            y = np.arange(len(gl.index))#.get_level_values('isolate')))\n",
    "            data = gl.values\n",
    "    \n",
    "            # make a color map of fixed colors\n",
    "#             cmap = mpl.colors.ListedColormap([dict_background['color']['NA'],\n",
    "#                                               'w',\n",
    "#                                               dict_background['color']['WA']])\n",
    "            cmap = mpl.colors.ListedColormap(['k','w','k'])\n",
    "            bounds=[-1,0,1]\n",
    "            norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "\n",
    "            plot.heatmap(np.r_[x, x.max()+1], np.r_[y, y.max()+1], data, \n",
    "                         ax2, '', '', '', [], [], cmap=cmap, vmin=-1, vmax=1)\n",
    "            \n",
    "            lineage = gl.index.get_level_values('lineage').unique()[0]\n",
    "            ax2.tick_params(axis='y', colors=dict_lineages[lineage]['fill'], width=5, which='both')\n",
    "\n",
    "        # show de novo mutations\n",
    "        indexer = [slice(None)]*len(jdf.index.names)\n",
    "        indexer[jdf.index.names.index('type')] = ['de novo']\n",
    "        gn = jdf.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "        \n",
    "        if len(gn) > 0:\n",
    "            \n",
    "            for i,(k,g) in enumerate(gn.groupby(level='clone')):\n",
    "                g = g.dropna(axis=1)\n",
    "                x = g.columns.get_level_values('pos').values\n",
    "                y = [i+.5]*len(x)\n",
    "                data = g.values.flatten()\n",
    "                ax2.scatter(x, y, \n",
    "                            facecolors=[dict_genotype['color'][gt] for gt in data],\n",
    "                            edgecolors='k', s=8, zorder=3)\n",
    "\n",
    "        # set grid\n",
    "        xlabels = sgd_length.chr_start + (sgd_length.chr_end - sgd_length.chr_start)/2.\n",
    "        xstart = sgd_length.chr_start\n",
    "        if ii<(len(seq_loh_df.index)-1):\n",
    "            ax2.set_xticks([])\n",
    "            ax2.set_xticklabels([])\n",
    "        else:\n",
    "            ax2.set_xticks(xlabels, minor=False)\n",
    "            ax2.set_xticklabels(sgd_length.chr_arabic.apply(utils.int_to_roman), minor=False)\n",
    "            ax2.set_xlabel('Chromosome')\n",
    "        xgrid=[xst+1. for xst in list(set(xstart))]\n",
    "        [ax2.axvline(g, lw=0.5, ls=\"-\", color=\"gray\") for g in xgrid]\n",
    "    \n",
    "        ylabels = gl.index.get_level_values('clone')\n",
    "        ax2.set_yticks(np.arange(len(ylabels)) + 0.5, minor=False)\n",
    "        ax2.set_yticklabels(ylabels, fontweight='bold', va='center', minor=False)\n",
    "        [ax2.axhline(g, lw=0.5, ls=\"-\", color=\"lightgray\") for g in np.arange(len(ylabels))]\n",
    "    \n",
    "    # show de novo mutations\n",
    "    indexer = [slice(None)]*len(sdf.index.names)\n",
    "    indexer[sdf.index.names.index('type')] = ['de novo']\n",
    "    gn_all = sdf.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "        \n",
    "    genes = gn_all.columns.get_level_values('gene')\n",
    "    loc = zip(gn_all.columns.get_level_values('pos'), [gn.shape[0]+.25]*gn_all.shape[1])\n",
    "\n",
    "    for l, xy in zip(genes, loc):     \n",
    "        trans = ax2.get_xaxis_transform() # x in data units, y in axes fraction\n",
    "        ann = ax2.annotate(l, xy=xy, xytext=(0, -2), textcoords='offset points',\n",
    "                           arrowprops=dict(arrowstyle='wedge,tail_width=0.7', color='black'),\n",
    "                           fontsize=4, style=('italic' if l!='non-coding' else 'normal'), \n",
    "                           annotation_clip=False, va='top', ha='center', \n",
    "                           bbox=(dict(facecolor='none', edgecolor='gray', boxstyle='round,pad=.2')\\\n",
    "                                if l in ['RNR2','RNR4','FPR1','TOR1'] else \\\n",
    "                                 dict(facecolor='none', edgecolor='none', boxstyle='round,pad=.2')))\n",
    "        \n",
    "#     # highlight haploids\n",
    "#     if p in ['WAxNA_F12_1_RM_1_T32','WAxNA_F12_2_RM_1_T32']:\n",
    "#         ax2 = plt.subplot(gs[('RM',p)][(1,0)], sharex=ax1)\n",
    "#         ax2.text(1.025, 0.5, 'aneuploidy', fontsize=5, va='center', ha='left')\n",
    "#     # highlight haploids\n",
    "#     if p in ['WAxNA_F12_1_RM_1_T32','WAxNA_F12_2_RM_1_T32']:\n",
    "# #     ax2 = plt.subplot(gs[('RM','WAxNA_F12_1_RM_1_T32')][(1,0)])\n",
    "#         ax2.text(1.025, 0.5, 'aneuploidy', fontsize=5, va='center', ha='left')\n",
    "#     ax2 = plt.subplot(gs[('RM','WAxNA_F12_2_RM_1_T32')][(1,0)])\n",
    "#     ax2.text(1.025, 0.5, 'aneuploidy', transform=ax2.transAxes,\n",
    "#              fontsize=5, va='center', ha='left')\n",
    "\n",
    "# set panel labels\n",
    "ax1 = plt.subplot(gs['HU','WAxNA_F12_1_HU_1_T32'][(0,0)])\n",
    "ax1.text(-0.05, 3,  chr(ord('A')), transform=ax1.transAxes,\n",
    "         fontsize=12, fontweight='bold', va='center', ha='right')\n",
    "ax1.text(0, 3,  dict_environment['short_label']['HU'], transform=ax1.transAxes,\n",
    "         fontsize=8, va='center', ha='left')\n",
    "\n",
    "ax1 = plt.subplot(gs['RM','WAxNA_F12_1_RM_1_T32'][(0,0)])\n",
    "ax1.text(-0.05, 3,  chr(ord('B')), transform=ax1.transAxes,\n",
    "         fontsize=12, fontweight='bold', va='center', ha='right')\n",
    "ax1.text(0, 3,  dict_environment['short_label']['RM'], transform=ax1.transAxes,\n",
    "         fontsize=8, va='center', ha='left')\n",
    "\n",
    "# highlight diploids\n",
    "# ax1 = plt.subplot(gs[('RM','WAxNA_F12_1_RM_1_T32')][(0,0)])\n",
    "# ax2 = plt.subplot(gs[('RM','WAxNA_F12_1_RM_1_T32')][1])\n",
    "# ax2.text(1.025, 0.5, 'aneuploidy', transform=ax2.transAxes, fontsize=5, va='center', ha='left')\n",
    "# ax2 = plt.subplot(gs[('RM','WAxNA_F12_2_RM_1_T32')][1])\n",
    "# ax2.text(1.025, 0.5, 'aneuploidy', transform=ax2.transAxes, fontsize=5, va='center', ha='left')\n",
    "\n",
    "# set legend\n",
    "for k, xy in zip([('HU','WAxNA_F12_2_HU_1_T32'),('RM','WAxNA_F12_1_RM_3_T32')],[(2,0),(2,0)]):\n",
    "    ax3 = plt.subplot(gs[k][(0,0)])\n",
    "    \n",
    "    wt_artist = patches.Rectangle((0,0), width=1, height=1, color=dict_background['color']['WA/WA'])\n",
    "    het_artist = patches.Rectangle((0,0), width=1, height=1, color=dict_background['color']['WA/NA'])\n",
    "    hom_artist = patches.Rectangle((0,0), width=1, height=1, color=dict_background['color']['NA/NA'])\n",
    "    \n",
    "    leg1 = ax3.legend([wt_artist,het_artist,hom_artist], ['WA/WA','WA/NA','NA/NA'], \n",
    "                      bbox_to_anchor=(1.1, 1.5), ncol=1,\n",
    "                      frameon=False, loc='center',\n",
    "                      borderaxespad=0, handlelength=0.75, \n",
    "                      title='Background', prop={'size':5})\n",
    "    \n",
    "    ax4 = plt.subplot(gs[k][(2,0)])\n",
    "\n",
    "    het_artist = lines.Line2D((0,1),(0,0), mec='k', mfc=dict_genotype['color'][1], \n",
    "                              marker='o', markersize=4, linestyle='')\n",
    "    hom_artist = lines.Line2D((0,1),(0,0), mec='k', mfc=dict_genotype['color'][2],\n",
    "                              marker='o', markersize=4, linestyle='')\n",
    "    \n",
    "    leg2 = ax4.legend([het_artist,hom_artist], ['het.','hom.'], \n",
    "                      bbox_to_anchor=(1.1, -1.5), ncol=1,\n",
    "                      frameon=False, loc='center',\n",
    "                      borderaxespad=0, handlelength=0.75, \n",
    "                      title='Genotype', prop={'size':5})\n",
    "        \n",
    "    for leg in [leg1,leg2]:\n",
    "        plt.setp(leg.get_title(),fontsize=6)\n",
    "    \n",
    "# axes limits\n",
    "for ax in fig.get_axes():\n",
    "    ax.xaxis.label.set_size(6)\n",
    "    ax.yaxis.label.set_size(6)\n",
    "    ax.tick_params(axis='both', which='major', size=2, labelsize=6)\n",
    "    ax.tick_params(axis='both', which='minor', size=1, labelsize=4)\n",
    "\n",
    "fig.savefig(dir_supp+'figures/figureSX_seq_loh/figureSX_seq_loh.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_supp+'figures/figureSX_seq_loh/figureSX_seq_loh.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
