{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 5 - Background-averaged fitness effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "# load external dependencies\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "%matplotlib inline\n",
    "\n",
    "import math\n",
    "\n",
    "import json, matplotlib\n",
    "s = json.load( open(\"matplotlibrc.json\") )\n",
    "matplotlib.rcParams.update(s)\n",
    "\n",
    "from itertools import *\n",
    "\n",
    "import random\n",
    "import pandas as pd # (*) Pandas for data manipulation\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import os,glob\n",
    "\n",
    "# load internal dependencies\n",
    "import colors,plot,stats,utils\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "# absolute file path for data directory\n",
    "from paths import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Spores"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this experiment, we choose a set of ancestral and evolved clones. All of these are diploid cells which undergo sporulation. We can import the plate layout and the growth parameters. These two dataframes can be merged by the plate index, and row and column indices within the plates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Layout</th>\n",
       "      <th>Plate</th>\n",
       "      <th>Row</th>\n",
       "      <th>Column</th>\n",
       "      <th>env_evo</th>\n",
       "      <th>env_test</th>\n",
       "      <th>ID_YJH</th>\n",
       "      <th>strain</th>\n",
       "      <th>ID</th>\n",
       "      <th>background</th>\n",
       "      <th>...</th>\n",
       "      <th>tetrad</th>\n",
       "      <th>spore</th>\n",
       "      <th>mating</th>\n",
       "      <th>auxotrophy</th>\n",
       "      <th>genotype_short</th>\n",
       "      <th>genotype_long</th>\n",
       "      <th>aminoacid</th>\n",
       "      <th>contamination</th>\n",
       "      <th>doubling_time</th>\n",
       "      <th>growth_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>HU</td>\n",
       "      <td>HU</td>\n",
       "      <td>1903</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA_F12_2_HU_1_T34_C1</td>\n",
       "      <td>evolved</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>MATα</td>\n",
       "      <td>lys2-</td>\n",
       "      <td>+</td>\n",
       "      <td>RNR4</td>\n",
       "      <td></td>\n",
       "      <td>no</td>\n",
       "      <td>0.197277</td>\n",
       "      <td>0.872196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>HU</td>\n",
       "      <td>HU</td>\n",
       "      <td>1903</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA_F12_2_HU_1_T34_C1</td>\n",
       "      <td>evolved</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>MATα</td>\n",
       "      <td>lys2-</td>\n",
       "      <td>+</td>\n",
       "      <td>RNR4</td>\n",
       "      <td></td>\n",
       "      <td>no</td>\n",
       "      <td>0.482655</td>\n",
       "      <td>0.715659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>HU</td>\n",
       "      <td>HU</td>\n",
       "      <td>1903</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA_F12_2_HU_1_T34_C1</td>\n",
       "      <td>evolved</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>MATα</td>\n",
       "      <td>lys2-</td>\n",
       "      <td>+</td>\n",
       "      <td>RNR4</td>\n",
       "      <td></td>\n",
       "      <td>no</td>\n",
       "      <td>0.517250</td>\n",
       "      <td>0.698703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>HU</td>\n",
       "      <td>HU</td>\n",
       "      <td>1903</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA_F12_2_HU_1_T34_C1</td>\n",
       "      <td>evolved</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>MATα</td>\n",
       "      <td>lys2-</td>\n",
       "      <td>+</td>\n",
       "      <td>RNR4</td>\n",
       "      <td></td>\n",
       "      <td>no</td>\n",
       "      <td>0.554730</td>\n",
       "      <td>0.680785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>HU</td>\n",
       "      <td>HU</td>\n",
       "      <td>1909*</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA_F12_2_HU_1_T34_C1</td>\n",
       "      <td>evolved</td>\n",
       "      <td>...</td>\n",
       "      <td>B</td>\n",
       "      <td>4</td>\n",
       "      <td>MATα</td>\n",
       "      <td>lys2-</td>\n",
       "      <td>-</td>\n",
       "      <td>RNR4*</td>\n",
       "      <td>Arg34Ile</td>\n",
       "      <td>no</td>\n",
       "      <td>0.276690</td>\n",
       "      <td>0.825483</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Layout  Plate  Row  Column env_evo env_test ID_YJH strain  \\\n",
       "0       0      0    0       0      HU       HU   1903  spore   \n",
       "1       0      0    0       1      HU       HU   1903  spore   \n",
       "2       0      0    0       2      HU       HU   1903  spore   \n",
       "3       0      0    0       3      HU       HU   1903  spore   \n",
       "4       0      0    0       4      HU       HU  1909*  spore   \n",
       "\n",
       "                        ID background     ...     tetrad spore mating  \\\n",
       "0  WAxNA_F12_2_HU_1_T34_C1    evolved     ...          A     2   MATα   \n",
       "1  WAxNA_F12_2_HU_1_T34_C1    evolved     ...          A     2   MATα   \n",
       "2  WAxNA_F12_2_HU_1_T34_C1    evolved     ...          A     2   MATα   \n",
       "3  WAxNA_F12_2_HU_1_T34_C1    evolved     ...          A     2   MATα   \n",
       "4  WAxNA_F12_2_HU_1_T34_C1    evolved     ...          B     4   MATα   \n",
       "\n",
       "  auxotrophy genotype_short genotype_long aminoacid contamination  \\\n",
       "0      lys2-              +          RNR4                      no   \n",
       "1      lys2-              +          RNR4                      no   \n",
       "2      lys2-              +          RNR4                      no   \n",
       "3      lys2-              +          RNR4                      no   \n",
       "4      lys2-              -         RNR4*  Arg34Ile            no   \n",
       "\n",
       "  doubling_time growth_rate  \n",
       "0      0.197277    0.872196  \n",
       "1      0.482655    0.715659  \n",
       "2      0.517250    0.698703  \n",
       "3      0.554730    0.680785  \n",
       "4      0.276690    0.825483  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load data\n",
    "spores_abs_df = utils.load_data(dir_data+'pheno/pheno_spores_abs.pickle')\n",
    "spores_norm_df = utils.load_data(dir_data+'pheno/pheno_spores_norm.pickle')\n",
    "\n",
    "# filter out blank positions, contamination and constructs\n",
    "spores_norm_df = spores_norm_df[(spores_norm_df['contamination']=='no')&\\\n",
    "                                ~(spores_norm_df['strain'].isin(['control','']))&\\\n",
    "                                ~(spores_norm_df['genotype_long'].isin([u'fpr1Δ',u'tor1Δ']))]\n",
    "\n",
    "   \n",
    "# remove NaNs\n",
    "# spores_norm_df = spores_norm_df[np.isfinite(spores_norm_df['growth_rate'])]\n",
    "\n",
    "spores_norm_df.head()  # show dataframe header to stdout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#spore_abs_df.to_csv(dir_data+'spores/20150402_spores_abs_pheno_filtered_with_labels.csv', encoding='utf-8', index=False, sep='\\t')\n",
    "#spore_norm_df.to_csv(dir_data+'spores/20150402_spores_norm_pheno_filtered_with_labels.csv', encoding='utf-8', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hybrids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Layout</th>\n",
       "      <th>Plate</th>\n",
       "      <th>Row</th>\n",
       "      <th>Column</th>\n",
       "      <th>env_evo</th>\n",
       "      <th>env_test</th>\n",
       "      <th>MATa</th>\n",
       "      <th>strain_MATa</th>\n",
       "      <th>ID_MATa</th>\n",
       "      <th>background_MATa</th>\n",
       "      <th>...</th>\n",
       "      <th>background</th>\n",
       "      <th>population</th>\n",
       "      <th>gene</th>\n",
       "      <th>genotype_short</th>\n",
       "      <th>genotype_long</th>\n",
       "      <th>auxotrophy</th>\n",
       "      <th>tetrad</th>\n",
       "      <th>spore</th>\n",
       "      <th>strain</th>\n",
       "      <th>contamination</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31488</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>20</td>\n",
       "      <td>RM</td>\n",
       "      <td>RM</td>\n",
       "      <td>lys_218_R1 a ura-</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA</td>\n",
       "      <td>ancestral</td>\n",
       "      <td>...</td>\n",
       "      <td>(ancestral, evolved)</td>\n",
       "      <td>(WAxNA, WAxNA)</td>\n",
       "      <td>(, TOR1)</td>\n",
       "      <td>(+, -)</td>\n",
       "      <td>(, TOR1*)</td>\n",
       "      <td>(lys2-, ura3-)</td>\n",
       "      <td>(A, A)</td>\n",
       "      <td>(1, 1)</td>\n",
       "      <td>(spore, spore)</td>\n",
       "      <td>(no, no)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31489</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>21</td>\n",
       "      <td>RM</td>\n",
       "      <td>RM</td>\n",
       "      <td>lys_218_R1 a ura-</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA</td>\n",
       "      <td>ancestral</td>\n",
       "      <td>...</td>\n",
       "      <td>(ancestral, evolved)</td>\n",
       "      <td>(WAxNA, WAxNA)</td>\n",
       "      <td>(, TOR1)</td>\n",
       "      <td>(+, -)</td>\n",
       "      <td>(, TOR1*)</td>\n",
       "      <td>(lys2-, ura3-)</td>\n",
       "      <td>(A, A)</td>\n",
       "      <td>(1, 1)</td>\n",
       "      <td>(spore, spore)</td>\n",
       "      <td>(no, no)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31490</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>RM</td>\n",
       "      <td>RM</td>\n",
       "      <td>lys_218_R1 a ura-</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA</td>\n",
       "      <td>ancestral</td>\n",
       "      <td>...</td>\n",
       "      <td>(ancestral, evolved)</td>\n",
       "      <td>(WAxNA, WAxNA)</td>\n",
       "      <td>(, TOR1)</td>\n",
       "      <td>(+, -)</td>\n",
       "      <td>(, TOR1*)</td>\n",
       "      <td>(lys2-, ura3-)</td>\n",
       "      <td>(A, A)</td>\n",
       "      <td>(1, 1)</td>\n",
       "      <td>(spore, spore)</td>\n",
       "      <td>(no, no)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31491</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>20</td>\n",
       "      <td>RM</td>\n",
       "      <td>YNB</td>\n",
       "      <td>lys_218_R1 a ura-</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA</td>\n",
       "      <td>ancestral</td>\n",
       "      <td>...</td>\n",
       "      <td>(ancestral, evolved)</td>\n",
       "      <td>(WAxNA, WAxNA)</td>\n",
       "      <td>(, TOR1)</td>\n",
       "      <td>(+, -)</td>\n",
       "      <td>(, TOR1*)</td>\n",
       "      <td>(lys2-, ura3-)</td>\n",
       "      <td>(A, A)</td>\n",
       "      <td>(1, 1)</td>\n",
       "      <td>(spore, spore)</td>\n",
       "      <td>(no, no)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31492</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>21</td>\n",
       "      <td>RM</td>\n",
       "      <td>YNB</td>\n",
       "      <td>lys_218_R1 a ura-</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA</td>\n",
       "      <td>ancestral</td>\n",
       "      <td>...</td>\n",
       "      <td>(ancestral, evolved)</td>\n",
       "      <td>(WAxNA, WAxNA)</td>\n",
       "      <td>(, TOR1)</td>\n",
       "      <td>(+, -)</td>\n",
       "      <td>(, TOR1*)</td>\n",
       "      <td>(lys2-, ura3-)</td>\n",
       "      <td>(A, A)</td>\n",
       "      <td>(1, 1)</td>\n",
       "      <td>(spore, spore)</td>\n",
       "      <td>(no, no)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Layout  Plate  Row  Column env_evo env_test               MATa  \\\n",
       "31488       1      0    8      20      RM       RM  lys_218_R1 a ura-   \n",
       "31489       1      0    8      21      RM       RM  lys_218_R1 a ura-   \n",
       "31490       1      0    9      20      RM       RM  lys_218_R1 a ura-   \n",
       "31491       1      1    8      20      RM      YNB  lys_218_R1 a ura-   \n",
       "31492       1      1    8      21      RM      YNB  lys_218_R1 a ura-   \n",
       "\n",
       "      strain_MATa ID_MATa background_MATa      ...                 background  \\\n",
       "31488       spore   WAxNA       ancestral      ...       (ancestral, evolved)   \n",
       "31489       spore   WAxNA       ancestral      ...       (ancestral, evolved)   \n",
       "31490       spore   WAxNA       ancestral      ...       (ancestral, evolved)   \n",
       "31491       spore   WAxNA       ancestral      ...       (ancestral, evolved)   \n",
       "31492       spore   WAxNA       ancestral      ...       (ancestral, evolved)   \n",
       "\n",
       "           population      gene genotype_short genotype_long      auxotrophy  \\\n",
       "31488  (WAxNA, WAxNA)  (, TOR1)         (+, -)     (, TOR1*)  (lys2-, ura3-)   \n",
       "31489  (WAxNA, WAxNA)  (, TOR1)         (+, -)     (, TOR1*)  (lys2-, ura3-)   \n",
       "31490  (WAxNA, WAxNA)  (, TOR1)         (+, -)     (, TOR1*)  (lys2-, ura3-)   \n",
       "31491  (WAxNA, WAxNA)  (, TOR1)         (+, -)     (, TOR1*)  (lys2-, ura3-)   \n",
       "31492  (WAxNA, WAxNA)  (, TOR1)         (+, -)     (, TOR1*)  (lys2-, ura3-)   \n",
       "\n",
       "       tetrad   spore          strain contamination  \n",
       "31488  (A, A)  (1, 1)  (spore, spore)      (no, no)  \n",
       "31489  (A, A)  (1, 1)  (spore, spore)      (no, no)  \n",
       "31490  (A, A)  (1, 1)  (spore, spore)      (no, no)  \n",
       "31491  (A, A)  (1, 1)  (spore, spore)      (no, no)  \n",
       "31492  (A, A)  (1, 1)  (spore, spore)      (no, no)  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load data\n",
    "hybrids_abs_df = utils.load_data(dir_data+'pheno/pheno_hybrids_abs.pickle')\n",
    "hybrids_norm_df = utils.load_data(dir_data+'pheno/pheno_hybrids_norm.pickle')\n",
    "\n",
    "# filter out contamination\n",
    "# hybrids_norm_df = hybrids_norm_df[hybrids_norm_df.contamination.isin([('no','no')])]\n",
    "# hybrids_norm_df['Generation Time'] = hybrids_norm_df['Generation Time']\\\n",
    "# .where(hybrids_norm_df.contamination.isin([('yes','no'),('no','yes'),('yes','yes')]), np.nan)\n",
    "\n",
    "# filter out blank positions, contamination and constructs\n",
    "hybrids_norm_df = hybrids_norm_df[(hybrids_norm_df['contamination']==('no','no'))&\\\n",
    "                                  ~(hybrids_norm_df['strain'].isin([('control','control'),('','')]))&\\\n",
    "                                  ~(hybrids_norm_df['genotype_long'].isin([u'fpr1Δ',u'tor1Δ']))]\n",
    "\n",
    "# remove NaNs\n",
    "# hybrids_norm_df = hybrids_norm_df[np.isfinite(hybrids_norm_df['growth_rate'])]\n",
    "\n",
    "hybrids_norm_df.head()  # show dataframe header to stdout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#hybrid_abs_df.to_csv(dir_data+'hybrids/20150402_hybrids_abs_pheno_filtered_with_labels.txt', index=False, sep='\\t')\n",
    "#hybrid_norm_df.to_csv(dir_data+'hybrids/20150402_hybrids_norm_pheno_filtered_with_labels.txt', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_evo</th>\n",
       "      <th>env_test</th>\n",
       "      <th>run</th>\n",
       "      <th>Index</th>\n",
       "      <th>Plate</th>\n",
       "      <th>Row</th>\n",
       "      <th>Column</th>\n",
       "      <th>ID</th>\n",
       "      <th>ID_plate</th>\n",
       "      <th>background</th>\n",
       "      <th>...</th>\n",
       "      <th>assignment</th>\n",
       "      <th>isolate</th>\n",
       "      <th>gene</th>\n",
       "      <th>genotype_short</th>\n",
       "      <th>genotype_long</th>\n",
       "      <th>aminoacid</th>\n",
       "      <th>mating</th>\n",
       "      <th>auxotrophy</th>\n",
       "      <th>doubling_time</th>\n",
       "      <th>growth_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RM</td>\n",
       "      <td>RM</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>P1_A1</td>\n",
       "      <td>1</td>\n",
       "      <td>(ancestral, ancestral)</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>(, )</td>\n",
       "      <td>(+, +)</td>\n",
       "      <td>(,)</td>\n",
       "      <td></td>\n",
       "      <td>MATa/α</td>\n",
       "      <td></td>\n",
       "      <td>-0.037316</td>\n",
       "      <td>1.026203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RM</td>\n",
       "      <td>RM</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>P2_A1</td>\n",
       "      <td>2</td>\n",
       "      <td>(ancestral, ancestral)</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>(, )</td>\n",
       "      <td>(+, +)</td>\n",
       "      <td>(,)</td>\n",
       "      <td></td>\n",
       "      <td>MATa/α</td>\n",
       "      <td></td>\n",
       "      <td>0.101343</td>\n",
       "      <td>0.932165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RM</td>\n",
       "      <td>RM</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>P6_A1</td>\n",
       "      <td>6</td>\n",
       "      <td>(evolved, evolved)</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>(, )</td>\n",
       "      <td>(+, +)</td>\n",
       "      <td>(,)</td>\n",
       "      <td></td>\n",
       "      <td>MATa/α</td>\n",
       "      <td></td>\n",
       "      <td>0.039118</td>\n",
       "      <td>0.973249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RM</td>\n",
       "      <td>RM</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>P7_A1</td>\n",
       "      <td>7</td>\n",
       "      <td>(evolved, evolved)</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>(, )</td>\n",
       "      <td>(+, +)</td>\n",
       "      <td>(,)</td>\n",
       "      <td></td>\n",
       "      <td>MATa/α</td>\n",
       "      <td></td>\n",
       "      <td>-0.802399</td>\n",
       "      <td>1.743999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RM</td>\n",
       "      <td>RM</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>P1_A2</td>\n",
       "      <td>1</td>\n",
       "      <td>(ancestral, ancestral)</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>(, )</td>\n",
       "      <td>(+, +)</td>\n",
       "      <td>(,)</td>\n",
       "      <td></td>\n",
       "      <td>MATa/α</td>\n",
       "      <td></td>\n",
       "      <td>0.235418</td>\n",
       "      <td>0.849439</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  env_evo env_test run  Index  Plate  Row  Column     ID  ID_plate  \\\n",
       "0      RM       RM   1      0      0    0       0  P1_A1         1   \n",
       "1      RM       RM   1      1      0    0       1  P2_A1         2   \n",
       "2      RM       RM   1      2      0    0       2  P6_A1         6   \n",
       "3      RM       RM   1      3      0    0       3  P7_A1         7   \n",
       "4      RM       RM   1      4      0    0       4  P1_A2         1   \n",
       "\n",
       "               background     ...     assignment isolate  gene genotype_short  \\\n",
       "0  (ancestral, ancestral)     ...                      1  (, )         (+, +)   \n",
       "1  (ancestral, ancestral)     ...                      1  (, )         (+, +)   \n",
       "2      (evolved, evolved)     ...                      1  (, )         (+, +)   \n",
       "3      (evolved, evolved)     ...                      1  (, )         (+, +)   \n",
       "4  (ancestral, ancestral)     ...                      2  (, )         (+, +)   \n",
       "\n",
       "  genotype_long  aminoacid  mating auxotrophy doubling_time growth_rate  \n",
       "0           (,)             MATa/α                -0.037316    1.026203  \n",
       "1           (,)             MATa/α                 0.101343    0.932165  \n",
       "2           (,)             MATa/α                 0.039118    0.973249  \n",
       "3           (,)             MATa/α                -0.802399    1.743999  \n",
       "4           (,)             MATa/α                 0.235418    0.849439  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load data\n",
    "clone_norm_df = utils.load_data(dir_data+'pheno/pheno_populations_norm.pickle')\n",
    "\n",
    "# filter out strains used for spatial control\n",
    "clone_norm_df = clone_norm_df[(clone_norm_df.background == 'ancestral')|\\\n",
    "                              (clone_norm_df.background == 'evolved')]\n",
    "\n",
    "clone_norm_df['genotype_short'] = map(lambda x: tuple(x.strip(\"/\").split(\"/\")),clone_norm_df['genotype_short'])\n",
    "clone_norm_df['genotype_long'] = map(lambda x: tuple(x.strip(\"/\").split(\"/\")),clone_norm_df['genotype_long'])\n",
    "\n",
    "clone_norm_df['population'] = zip(clone_norm_df.population, clone_norm_df.population)\n",
    "clone_norm_df['background'] = zip(clone_norm_df.background, clone_norm_df.background)\n",
    "clone_norm_df['gene'] = zip(clone_norm_df.gene, clone_norm_df.gene)\n",
    "# clone_norm_df['genotype'] = zip(clone_norm_df.genotype, clone_norm_df.genotype)\n",
    "\n",
    "# remove NaNs\n",
    "# clone_norm_df = clone_norm_df[np.isfinite(clone_norm_df['growth_rate'])]\n",
    "\n",
    "clone_norm_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summary of technical replicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# import colors\n",
    "\n",
    "# pt = pd.pivot_table(spore_norm_df,\n",
    "#                index=[\"env_evo\",\"env_test\"],\n",
    "#                values=[\"Generation Time\"],\n",
    "#                columns=['gene','genotype','tetrad','spore'],\n",
    "#                aggfunc=[np.mean,np.std])\n",
    "\n",
    "# groups = spore_norm_df.groupby(['env_evo','env_test',\n",
    "#                                 'gene','genotype','tetrad','spore'])\n",
    "\n",
    "# norm_filtered_avg_df = groups['Generation Time'].agg([np.mean, np.std, 'count'])\n",
    "\n",
    "\n",
    "# #pg.describe(include=['Lag','Yield'])\n",
    "# # norm_filtered_avg_df.groupby(['env_evo'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# population attributes\n",
    "sp_bg_dict = {\n",
    "'position':\n",
    "{'HU': \n",
    " {'parents': 0, 'ancestral': 1, 'evolved': 2},\n",
    "'RM':\n",
    " {'parents': 0, 'ancestral': 1, 'evolved': 2}\n",
    " }\n",
    "}\n",
    "\n",
    "# background attributes\n",
    "sp_cl_dict = {\n",
    "'position':\n",
    "{'HU': \n",
    " {'WA': 0, 'NA': 1, 'WAxNA': 2},\n",
    "'RM':\n",
    " {'WA': 0, 'NA': 1, 'WAxNA': 2},\n",
    " },\n",
    "'color':\n",
    "{'HU':\n",
    " {'WA': colors.fte_colors[0], 'NA': colors.fte_colors[1], 'WAxNA': colors.fte_colors[2]},\n",
    "'RM':\n",
    " {'WA': colors.fte_colors[0], 'NA': colors.fte_colors[1], 'WAxNA': colors.fte_colors[2]},\n",
    "}\n",
    "}\n",
    "\n",
    "# gene attributes\n",
    "sp_gn_dict = {\n",
    "'position':\n",
    "{'HU': \n",
    " {'RNR2': 0, 'RNR4': 1},\n",
    "'RM':\n",
    " {'no driver': 0, 'FPR1': 1, 'TOR1': 2},\n",
    " }\n",
    "}\n",
    "\n",
    "# genotype attributes\n",
    "sp_gt_long_dict = {\n",
    "'position':\n",
    " {'HU': \n",
    "  {'': 0, 'RNR2': 1, 'RNR2*': 2, 'RNR4': 3, 'RNR4*': 4},\n",
    "  'RM': \n",
    "  {'': 0, 'no driver': 1, 'FPR1*': 2, 'TOR1': 3, 'TOR1*': 4}\n",
    "  },\n",
    "'color':\n",
    "  {'HU': \n",
    "  {'': colors.fte_colors[0], 'RNR2': colors.fte_colors[1], 'RNR2*': colors.fte_colors[2], 'RNR4': colors.fte_colors[1], 'RNR4*': colors.fte_colors[2]},\n",
    "  'RM': \n",
    "  {'': colors.fte_colors[0], 'no driver': colors.fte_colors[1], 'FPR1*': colors.fte_colors[2], 'TOR1': colors.fte_colors[1], 'TOR1*': colors.fte_colors[2]}\n",
    "  }\n",
    "}\n",
    "\n",
    "# genotype attributes\n",
    "sp_gt_short_dict = {\n",
    "'position':\n",
    " {'HU': \n",
    "  {'+': 0, '-': 1},\n",
    "  'RM': \n",
    "  {'+': 0, '-': 1}\n",
    "  },\n",
    "'color':\n",
    "  {'HU': \n",
    "  {'+': colors.fte_colors[0], '-': colors.fte_colors[1]},\n",
    "  'RM': \n",
    "  {'+': colors.fte_colors[0], '-': colors.fte_colors[1]}\n",
    "  }\n",
    "}\n",
    "\n",
    "# collection attributes\n",
    "hy_bg_dict = {\n",
    "'position':\n",
    "{'HU': \n",
    " {('parents','parents'): 0, ('ancestral','ancestral'): 1, \n",
    "  ('ancestral','evolved'): 2, ('evolved','evolved'): 3},\n",
    "'RM':\n",
    " {('parents','parents'): 0, ('ancestral','ancestral'): 1, \n",
    "  ('ancestral','evolved'): 2, ('evolved','evolved'): 3}\n",
    " }\n",
    "}\n",
    "\n",
    "# background attributes\n",
    "hy_cl_dict = {\n",
    "'position':\n",
    "{'HU': \n",
    " {('WA','WA'): 0, ('NA','NA'): 1, ('WAxNA','WAxNA'): 2},\n",
    "'RM':\n",
    " {('WA','WA'): 0, ('NA','NA'): 1, ('WAxNA','WAxNA'): 2},\n",
    " },\n",
    "'color':\n",
    "{'HU':\n",
    " {('WA','WA'): colors.fte_colors[0], ('NA','NA'): colors.fte_colors[1], ('WAxNA','WAxNA'): colors.fte_colors[2]},\n",
    "'RM':\n",
    " {('WA','WA'): colors.fte_colors[0], ('NA','NA'): colors.fte_colors[1], ('WAxNA','WAxNA'): colors.fte_colors[2]},\n",
    "}\n",
    "}\n",
    "\n",
    "# gene attributes\n",
    "hy_gn_dict = {\n",
    "'position':\n",
    "{'HU': \n",
    " {('','RNR2'): 0, ('','RNR4'): 1,\n",
    "  ('RNR2','RNR2'): 0, ('RNR4','RNR4'): 1, ('RNR2','RNR4'): 2},\n",
    "'RM':\n",
    " {('','no driver'): 0, ('','FPR1'): 1, ('','TOR1'): 2,\n",
    "  ('no driver','no driver'): 0, ('no driver','TOR1'): 1, ('FPR1','FPR1'): 2, ('TOR1','TOR1'): 3, ('FPR1','TOR1'): 4},\n",
    " }\n",
    "}\n",
    "\n",
    "# genotype attributes\n",
    "hy_gt_long_dict = {\n",
    "'position':\n",
    "{'HU': \n",
    " {('',''): 0,\n",
    "  ('','RNR2'): 1, ('','RNR2*'): 2, ('','RNR4'): 3, ('','RNR4*'): 4,\n",
    "  ('RNR2','RNR2'): 5, ('RNR2','RNR2*'): 6, ('RNR2*','RNR2*'): 7, \n",
    "  ('RNR4','RNR4'): 8, ('RNR4','RNR4*'): 9, ('RNR4*','RNR4*'): 10, \n",
    "  ('RNR2','RNR4'): 11, ('RNR2*','RNR4'): 12, ('RNR2','RNR4*'): 13, ('RNR2*','RNR4*'): 14},\n",
    "'RM':\n",
    " {('',''): 0,\n",
    "  ('','FPR1'): 1, ('','FPR1*'): 2, ('','TOR1'): 3, ('','TOR1*'): 4,\n",
    "  ('no driver','no driver'): 5, ('FPR1','FPR1'): 6, ('FPR1*','FPR1*'): 7,\n",
    "  ('TOR1','TOR1'): 8, ('TOR1','TOR1*'): 9, ('TOR1*','TOR1*'): 10,\n",
    "  ('FPR1','TOR1'): 11, ('FPR1*','TOR1'): 12, ('FPR1','TOR1*'): 13, ('FPR1*','TOR1*'): 14},\n",
    " },\n",
    "'color':\n",
    "{'HU':\n",
    " {('',''): colors.fte_colors[0],\n",
    "  ('','RNR2'): colors.fte_colors[1], ('','RNR2*'): colors.fte_colors[2], ('','RNR4'): colors.fte_colors[1], ('','RNR4*'): colors.fte_colors[2],\n",
    "  ('RNR2','RNR2'): colors.fte_colors[1], ('RNR2','RNR2*'): colors.fte_colors[2], ('RNR2*','RNR2*'): colors.fte_colors[3],\n",
    "  ('RNR4','RNR4'): colors.fte_colors[1], ('RNR4','RNR4*'): colors.fte_colors[2], ('RNR4*','RNR4*'): colors.fte_colors[3],\n",
    "  ('RNR2','RNR4'): colors.fte_colors[1], ('RNR2*','RNR4'): colors.fte_colors[2], ('RNR2','RNR4*'): colors.fte_colors[2], ('RNR2*','RNR4*'): colors.fte_colors[3]},\n",
    "'RM':\n",
    " {('',''): colors.fte_colors[0],\n",
    "  ('','FPR1'): colors.fte_colors[1], ('','FPR1*'): colors.fte_colors[2], ('','TOR1'): colors.fte_colors[1], ('','TOR1*'): colors.fte_colors[2],\n",
    "  ('no driver','no driver'): colors.fte_colors[1], ('FPR1','FPR1'): colors.fte_colors[1], ('FPR1*','FPR1*'): colors.fte_colors[3],\n",
    "  ('TOR1','TOR1'): colors.fte_colors[1], ('TOR1','TOR1*'): colors.fte_colors[2], ('TOR1*','TOR1*'): colors.fte_colors[3],\n",
    "  ('FPR1','TOR1'): colors.fte_colors[1], ('FPR1*','TOR1'): colors.fte_colors[2], ('FPR1','TOR1*'): colors.fte_colors[2], ('FPR1*','TOR1*'): colors.fte_colors[3]},\n",
    "}\n",
    "}\n",
    "\n",
    "# genotype attributes\n",
    "hy_gt_short_dict = {\n",
    "'position':\n",
    "{'HU': \n",
    " {('+','+'): 0, ('+','-'): 1, ('-','-'): 2},\n",
    "'RM':\n",
    " {('+','+'): 0, ('+','-'): 1, ('-','-'): 2},\n",
    " },\n",
    "'color':\n",
    "{'HU':\n",
    " {('+','+'): colors.fte_colors[0], ('+','-'): colors.fte_colors[1], ('-','-'): colors.fte_colors[2]},\n",
    "'RM':\n",
    " {('+','+'): colors.fte_colors[0], ('+','-'): colors.fte_colors[1], ('-','-'): colors.fte_colors[2]},\n",
    "}\n",
    "}\n",
    "\n",
    "# factor attributes\n",
    "factor_dict = {\n",
    "    'color': {\n",
    "        'background': colors.fte_colors[0], \n",
    "        'de_novo_gene': colors.fte_colors[1], \n",
    "        'de_novo_genotype': colors.fte_colors[2],\n",
    "        'auxotrophy': colors.fte_colors[3],\n",
    "        'tetrad': colors.fte_colors[4],\n",
    "        'spore': colors.fte_colors[4]\n",
    "#         'background': \"#348ABD\", \n",
    "#         'de_novo_gene': \"#A60628\", \n",
    "#         'de_novo_genotype': \"#7A68A6\",\n",
    "#         'auxotrophy': \"#467821\",\n",
    "#         'tetrad': \"#CF4457\",\n",
    "#         'spore': \"#188487\"\n",
    "        },\n",
    "    'hatch': {\n",
    "        'spores': \" \", \n",
    "        'hybrids': \"/\"\n",
    "        },\n",
    "    'position': {\n",
    "        'background': 0, \n",
    "        'de_novo_gene': 1, \n",
    "        'de_novo_genotype': 2,\n",
    "        'auxotrophy': 3,\n",
    "        'tetrad': 4,\n",
    "        'spore': 5     \n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Spores and hybrids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this experiment, we cross the mutant allele to a sample of wild-derived lines with standing variation, and we evaluate phenotypes of F1 genotypes. Phenotypic variation can then be attributed to allelic complementation (that is, dominance effects) or to non-allelic complementation (that is, epistasis)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "dict_background = {'parents':0, 'ancestral':1, 'evolved':2}\n",
    "dict_population = {'WA':0, 'NA':1, 'WAxNA':2}\n",
    "dict_gene = {'RNR2':0, 'RNR4':1, 'no driver':0, 'FPR1':1, 'TOR1':2}\n",
    "dict_genotype = {'+':0,'-':1}\n",
    "\n",
    "labels = {\n",
    "    'Generation Time':'doubling_time',\n",
    "    'gene':'de_novo_gene',\n",
    "    'genotype_short':'de_novo_genotype'\n",
    "}\n",
    "\n",
    "# environment attributes\n",
    "environments = {u'HU': u'Hydroxyurea (YNB+HU 10 mg/ml)', \n",
    "                u'RM': u'Rapamycin (YNB+RM 0.025 μg/ml)', \n",
    "                u'YNB': u'Control (YNB)'}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ensemble by background and genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/ipykernel/__main__.py:114: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/ipykernel/__main__.py:128: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAzUAAAFiCAYAAADP+PaKAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzs3Xl4VOX1wPHvmckelkBYEkBUVBRBURSR4IL7vrVV675U\na9XWVrHa1vorWrdqFZfa2taqiNZW3EVRrAoKgiIqKu4bCAQkhC0JWWbm/P64N2EMk8x2Z8lwPs8z\nT+7c5dwzcyfJfefdRFUxxhhjjDHGmK7Kl+kEjDHGGGOMMSYZVqgxxhhjjDHGdGlWqDHGGGOMMcZ0\naVaoMcYYY4wxxnRpVqgxxhhjjDHGdGlWqDHGGGOMMcZ0aVaoMcYYY4wxKSUiZ4nISHd5Uif77S8i\nx6Yvs++d+2ARGSMit4lILxE5w83neRH5kbvPJHfdvSJylYhcGnb8LiLiD3s+yf25v4gcF/662y33\ndeNe4j7fXUT+KCJ/FpHiBF7HdiJyUifbJ4nI0NbXlCvyMp2AMcYYY4zZIpwqIuOBYSJSCZwK3Anc\nALwIjAN6Ay+LyD3AG8DbwHnAt8By4CtgL2Ag8BfgIlX9vYjcASwA3gME2A0Y68aYB1wALAY2AK8A\nxwA7Ab9T1fVufger6m9E5BNgEvCBqk4RkeOBfUVkmrufuo9ewJdhr+8HwBJgnft8d7egsj3wsnsM\nYTGcBdVVbiHneHfVKcBvgCrgUOBpABF5BHgVGO2+L2NV9WwRuRZYBpysqge6r+3vIjITeAoYCiwC\n+qjqNc4p9TMRORd4LNKF6oqspsYYY4wxxqTDw6p6B/CRqlYD5cA5wEM4N+LXAc/jFEoaVfVBoBR4\nU1UnAaNVdT6wK7BCVZcD69xaiVdwCgo+93jCYnQHegBrgT5ACRACmoBdwvLLB1DVdcA3wL/Ctk0C\nfsWmwsizqno5sJeIdBORXwJ7Az8TkaPdfd5R1TuBx93nnc14L2F5a9i68GNqVPUfQJOq/h34RkS6\nA33d50vc/bqp6kacQtntzkvSu3EKYa1x215vrrCaGmOMMcYYk2qRbugfBm5U1X+4tSAXAYOAOTiF\nDoAPgB+LyADgLRE5EngJp7ZnO+A+4HlVHSMiI4AzgTU4NRetMT4EVgFF7vIwYCPgdx+t6jvKXVW/\nEREJW3eMe/4vVLUOuENEegJ/C6v5af/6nxaRPwAtwP9aN7hNzM4FRojI/4D/ABOBYuAPYTFa3J/N\n7s8QTgFllYj8FBggIjsCH7fbvyls/9ZccN+DnCGqnRUaOzlQ5BtgsPt0mKp+6lVSxqRDu89wCOcP\n3lzgclX9KlN5GWOMMV4QkXE4zZjG4tR4LMOpCblMVQMZzs0HXI3TLGuDqi5MMM4FwBpVfTTG/UdG\nOpeInIVzD9BDVd9OJBevicj+wNequqSD7Wep6mT3Ou+MUws1A1ikqo1RYu+I81rne513piRUUyMi\n++LcDLaWiM4Afu9VUsakSWub2GnA18BhOO1ZewIHZTAvY4wxJiki8mNgCk5zrIXAfGAbnL4lvwci\n1SZ4ef7f4zTham1e1h2nH8ihODUiq4B3cPrQHCUiJcABODUda2I9j6r+XUROFJH/4rzOapzX1gCs\nds/3Fk4/lBeBo0VkLXA5zvtzBDAdt59Jki87ISJyNk4N1UxgCE5TudZapW4isgcwHOe9KgVeB7bC\n6VeEqs7Bqd2KWS5WRiTap+YMnJvBd3GqvU71LCNj0u9fqvor4Nfu8x0zmYwxxhiTDLc5019w7vOm\nqOooVb1AVQ/D6RzfkK5UgBeA7YC1qvoG0IjTaT2kqs/i9PP4Bqc/SA3f7+MSq9YCwFIgyKa+KOrG\n/Q6nALUPzkADglPI6+0eNzShV+edfJz8TgRew2k90tpUrB7nvvs1nNezEKfZWGjzMFu2uJufiUgB\nTim4DDgQeMJdHq+qr3ueoTEpIiJf49Q4PodTU3MIToHmFlX9TZyxzgLeU9WFIjJJVS/tYL/9gZ6q\n+kxy2cdPRA7GGfXlZOCPwNE4f+yvBO5T1cfc0VeewvkD+jXQ4HbORER2wencGXSfT1LVS93X1Po3\n4NLwbe5yX+B3OFXod4rI7jgjxBQDV7udGeN5HdsBe3TU1MB9DX8DdlXVnBnVxRhjYuX+vZ+BcxO8\nk6p+noEcznQ76RuTFok0PzsKp1S9QlVnuR27TncfVqgxXdFRYcsBnG90EmFDVW5iQ1UaY0zm9Atb\nXpyJBKxAY9ItkeZnrU3PWm+AnsS5yTpRRHJqaDizxTheVf04QzEC3OLWBsTLhqrsmA1VaYwx6RP+\n5dzWqT6ZiPxERI4XkcNF5AoRmZpErJtFZISIXJpMnHQSkcvcx7MiUiDORKPXuxNo9nQn3jw5i/K6\n1t1+uIicme68UiWuQo07VN0R7tPzRCSE0/wMnM7Vx3iYmzHp0noT/A6bhnOM959AR0NVHu2OstI6\nVOWB7r7hQ1XuJc6MxOFDVfYMG6pygqo+jdOH7VSc38HwGJ4MVcmm9wGcoSovxR2q0i2szcXpwDkt\nUgzcoSpF5HdEHqpyPxEZzqahKo8N3w8bqtIYY7zyBs6wxgC/Dx+KWEQGS9is9x7y4XxRVw/MFJEd\nEoyzCKfvS0OScdJJcb7InAuc5K4L4XTzWAd8itOSIlvy8rkVEYUZyCll4upT495Y3IPT/OTVsE07\n4zQBeVpVT/A0Q2NSpF2fmq+APXCaiG0Ahqjq6iRitw1Vqaqzk4gT11CVUWINJYuGqoyVDVVpjDHx\nE5FTgck4hY33cUYAG4gzumf/DuZSSfRcBwI7AN1wvry7D9hGVW9NINaZOF8yXorTuiGhOOnk5vwM\nm5qCF7iblgJ1OK02/qaq72dRXqtwRlrbRVWvS2deqRJvoWYWTun5ZlX9bdj6/XAKOc3AgHiG4jMm\nU8IKNa3W4tSc/EFVZ2UmK2OMMcYbIrIPzmAwrfPULMUpKEzI9Dw1xngt4ck3jTHGGGOMMSYbJDpP\njTHGGGOMMW1EpEREBkff02J5LVvzSicr1BhjjDHGGC/0xWnqZrHSL1vzSptE5qkxxhhjjDGmjYiM\nxpnUebCI1HcwUqbF+n68HwLfJTt5fbbmlW7Wp8YYY4wxxiTNbf60t0cjduZ8LC9la17pZIUaY4wx\nxhiTMBHf/uSX/BZn6gBQXUhLww2qobiG1RdnPoRDupF3pQ+2dSZk0/kbCN6oqu/FGcuXl5d3RGVF\nxZV5ef5BwWCIYCg4Z9my5Tep6qI4Y/mLioqO7du37+V5fn9lIBjUYDD46vLly29R1U/jjJVXXFz0\nw77l5Zf58/L6BQKBUCAQeLF6xco/q+pXccYqyEdOLsT3S0HKFQ2GYFoDwdtUdUn0CN+LVZSPnFaI\n7+eClCkaCKJPbCR0hzsZePZTVXvYwx726PSBM3nXz4D9gePiOG5ShHXHAIXt9wGuzvTrtEfuPoA/\n4MxdFGlbh59r4KR2zzf7TIdt+32kfdvH6Cyf8H1xJgH+Jc68ImMy/R7awx7tH4CP/JKHpGzb2rwd\nj9f83c7RvJFnq3/o0SrdB9SQX/LnOGL5u+F/aldftzUX5A/SqwqG6O/yt9Wz8gboYCmq6YZ/Yhyx\n8isrKmac/5Nz1y567x1tXL9GN66r1ZemT9Mxo0evqqyomBBHrKKKiorXf/Hzn6//5OOPdGN9ndZv\nWK/PT3tWd99tt+/69+v30zhidaus6P/Wr3/58w1fffiONtVWa8OqpfrMow/riJ2HfdenvPepccQq\nK8G/cA961p/BIL2ArfV8Buth9A2Vkb+iEN+xccTqW4L/470oazjTjXUeg/UQ+gR7kledjxyU6c9a\nLA+rqTEmh4jIhUA+8AVOp8EdcObfeQ3YDqh0H1cDjwGP4EzKdgbO7NMzgQNwJthdqO7EoSIyAbgf\n2AVnVuJ64EWcybv2c8/TOlHa/cAwVZ0oIpOAe4FTcWYxfgT4EXAHMAFYBvxIVQ8RkZ8Ar2qc31QZ\nEwsR+QNQC+yJ89n7rapOEJFbcSbgPQ1n1u9Pgd2A1Ti/D+cCtwC/BT7G+ay/DgwDvlC3qYeI7AFs\nrapPuJ/7j4HDVfUH7vP3+P7v4+HA7cAl7rnKVPVGd9/7cH4n1wPrVfVOEblJVX+TyvfImHhJfvGv\nfb2H/sE/YI/S9ttUleCS2et03eKfabD5P9FidZe860f7e/6qyl9WEinW44GVa77Ujae3aOj5aLEG\nDhjwlysuv+wnPzv/vKL224LBIMf+4MTama+9dlwgEIg6OfaggQMfuvbaa0489ZRTCtpva2lp4bDD\nj6iZP3/+oS2BwLvRYw14+o5bbjzymCMO26xPe2NjI/sffsyq997/cF+Nofanm+S9fADl4wdSvNmg\nXy2EeIIV362lZU9V/TaGWPMOpu9eFRRK+21NhHiC6pXrCeyiqquixcokG/3MmNyyAKdQ0wdQ4CXg\nLuA4YE9Vvc3dZwTODdnDwKHACmA5TsHnLaAQ6B0Wt0JVa93ll4AbgCqcydyagDJgAFCtqv8FCkWk\n9Y/2EcCdwN3AkWEx+6jq33EKYLjn39qD98CYjkwBnsQpkLwnIicAXwNB4HVVvRmn1kaBh1T1Q0Bw\nJp1+HKewITiT9LbgfHHQamtguYj0A9aq6j+AlWHb2/8+4sbqpap/w6kNbXU4zu/L1LB1+cm9dGO8\nJSJ+xP9zX8XIzQo07nb8A0f3xJ//hxhiFfqRs/f29dysQNMa69C8Pr1K8F0XQ6xuxcXFP/zpT87d\nrEAD4Pf7uePWW3pXVlTcFEOs3j3Leh50yo9/vFmBBiA/P59Jt93Wp6Ky4sYYYg2o6N9vTKQCDUBR\nURG33XRd30EDBlwbQ6ztepI3PFKBBiAfH2Pp1bcE/+9jiLVrHwqGRCrQABTiYzRlfUrwXx4tVqbF\nVKgRkX1F5GkRWSoiIRE5M4ZjRojITBFpEJFvReTq5NM1xkTRG2gAdsS5YToSp1bmGeAtEbkM2APn\npizkHjMN6AEU4dTalAGNwE5hcReJyHbucgjnBk2AXYE6wI/z96RCRM7H6a8XcPd7Hufb6Ivdc+Gu\nrxaRH+EUpMD5FvsDT94FYzbnNM/f9Nl9BPg9Ts0iwD4iciVOLQps+v1QnJqZw4CT3ef9cD734b8j\n7wE7qep3QLH72e4fFgO+//vYur5WRC7Cqa1pXfcCcDrOaEa4XxCsT+K1G5MKg6W4d6H4Oh5IV/KK\nkLziMhGJWFgJs+NWvqJ8n0S8rwagh+RRIL5KkU52cow87JCDC32+jm9xt99+O4qLi7aNEgdg9HHH\nHtuts1OOHLkrBfkFO8cQq+qkHxxf1ukOY/bCn+cfHS2QwP470K1fZ/tsRZH4nJYXnSpADtuB0r6d\n7bMtJX6Bo6LFyrRYh3TuhnOzMRl4MNrOItId5xupmTg3UMOAB0SkTlUnJZaqMSYaVX2hddn98uFh\nVX3fXfVmu90vc4/5FJgYtv59NvcwcIJ+f1SV8G+THnHP+bGq/jMsn8vcxasiHPdH9+dj7s81qloT\n4dzGJE1VWz93TwOIyA+A+1S1HpjlPlo9FXZc62f4d+7P1v+Bb7SL/5WIjHOXW5uJPdYao5Pfx+vb\nxWk9X1sBX0T2x6klMiabhIihC4OiwqYvCTqMFSKGWLH1mAgFAoGoe4ZCMQULBQLBTndUVUKxJRYK\nBjuP1Rovlljum99haUu/96OT80HUpEIoEv0aZlxMNTWqOl1Vf6+qTxDDG4TzDVMxcJaqfuwe9yfc\nmyhjTOqp6oNhN1DJxmrRGIaJDLshS+QcUxI91ph4qeoTqnq3xzE7/Awn8/uoqrNiaRdvTJot0cY1\nDRps6XAHbWmAQNMKVW2MEuuTpaHGpkAnN/S12kILoa81+l3/uy++9L+mQCDQ4Q7vf/ABTc1NsYyA\nNu/Jp56q7+yU8958k5aWlrdiiPXaw/99bG1nsV56ZaYGAoFXogVS+N8n1HXav+VrGoJB9NlosVrQ\naR9T911n+3xJfXOA7B8qOlV9avbGaZ/cHLbuRWCAiFibeWOMMcaYLkxVlVDgxuCyN9dFulFXDRH8\n9o01BBp/G0OsQBDufDVYuyFSrKAqzwdWra4neGUMsRqbW1ruu/HmW+ojbW9sbOTiX166ekX1ilhi\nbWhoaHjszrvuaoi0vb6+nl9deumqZcuWRe27oqo1a9eue+n+Kf+OWMBbt349l//u/1Ytr17xx0jb\n28Va2kBw3hfUN0favpEgb7F21UZCN8cQ67N1tHywmIaIpcB6AixgXU0jobuixcq0VBVqKvh+B0nc\n5+JuM8YYY4wxXZgGGv+pG5b9PfjF8zWhDctRDaEaIrRuiQY+e3aVNtRcp6HA9Fhi1RP888ehuof/\nHaiuWRzaSEiVgCqfhOpC/wosXVWjLb8LxDjDfXV19f/dN/nBp44/8eTVc998k1AoRGNjI48+9nhw\nzD77f/f1199cEstoZQDLly+/9I477nzxlFNPq317wQJCoRANDQ08/O9/B8aMrVr57bdLz1fVT2KJ\ntay6+rxrb7pl1tkXXLzmvfc/IBQKsWFDHfdP+XfLmPGHrlheveI0VV0cS6x6gqfMZc2bM6lZU0sz\nitJEiEVsaH6C6uoGgj9y+/jFEuuE16h953VWr1tDC4rSSJAPWN/4FCuWbSR4jKquiyVWJsU9pLOI\nbAAuVtUO+9aIyIvAt6p6Xti6rYDFwFhVfbPd/jautNmMqkbrDJg1RCSGGvHU2rBhA927d894jFwg\nIrG2a04Lu7apZe+NyZAu8z8O7F7NRJZN92qxDhQQrxVsGvWlVX+c/jgrIh2QrhuIiRMnMnHixJw6\nVy6+puiDm5j26urqkr4x8yKG8Z5d29Sy98aY2OTvdo4ncRb+7WyGNHzjSayig86icYM3lQhF3XtS\nsz5iq7W41X84hz7fLkg6Tggf60cdTb8Sf9KxmgMhZtz9GB/eNTnpWMUVfbn863lJx/FSqpqfzQX2\nFZHwcb0PBZbHWq1mjDHGGJMuqooumO08sqim1hgTm5hqakSkFNgep6rUBwwWkZFArap+KyI3AqNV\n9WD3kH8D/4czjPP1OHNmXAlEnYDJGGOMMSbt3plD6EnnG2wfwB77ZDQdY0x8Ym1+tifwKpuGc77G\nfUwGzsXp/N82iZGqrheRQ3BmRJ4PrAFuUdXbPco7YePHj8+5c+XiazLGGGNMeolIN3x5J+Mv3B0N\nrifQ+F9VXZjpvIyJRazz1MxSVZ+q+ts9znW3n6Oq27U7ZpGqjlfVElUdqKrXpeIFxCsXCwC5+Jq2\ndAsXpvZ/SEfxH300c8PQz5o1iyVLlnS4ffLk5NsAZ1qmrmu2vnddLV+zSU7+vo4ah++Es/CdcBaM\nGpfpbNJO8kuupLDH576K3f/i33q/i/1bjfut9Bz8khSUviMiAzKdnzHRpGqgAGNMO9dddx3bbLMN\n999/P6effjobNmxgzz33ZMaMGZSWltK3b1/Ky8tRVebOnUtDQwOvvvoqF154Ib169Yoa/9lnn+Xl\nl19m5MiRVFZW0qNHD0pKSigvL2fGjBnstddezJ8/n8MOO4xp06ZRVlbGn//8Z8444wymT5/OEUcc\nkZYBGh544AGWLl3K+PHj+eqrr1i/fj2NjY2MGDGCuro6FixYwKJFi6itraW+vp59992Xb7/9lmXL\nlqU8t0Sk+rpOnTqVKVOmsPfee2f1dQ3P97HHHsv6z6GJTa79vnZGRLbYJmeSX3K5dB/wW//gfXqK\nbPq+29dzcF9tqOkb+Prl2bREnKrFmKyRqoECjDERqCqHH344X375JWVlZVRVVVFUVMTw4cPx+Xwc\nc8wxrFmzhm222YbBgwfTp08fPvjgg5hiNzU1MWLECAYNGoTf70dVERFEhMGDB9OvXz9GjRrF7Nmz\nGTJkCKrK6NGjqa2tZcSIEXz22WcpfvWOlpYWRo0axdSpU9lvv/3w+Xz4fM6fotLSUqZMmcJ+++2H\niDBy5EiKi4vbtmerVF7XxsZGdtppp6y/ruH5doXPoYlNV/h9VVVmL6tn9rJ66+CfABEpwue/tH2B\npm17SR98/UdaTY3JenHPU5OSJLJgjg+TXdx5QrrM17WxfIYffPBBzjzzzJTlUF1dTWVlZcZj5IJ4\n5qlJ9XUFu7apZu9N5qgqTSudYWEL+++dUC3d7GX1TF60BoCzhvdin4GlnuaYQin/HyciOwDHquqt\nnexztK//yP/4K0d1+MZpKEDg/Sk2pHMctpQhnbPpXs2anxmTJqm+8TWZYdfVmMQ1rZxH3ecPtT0v\nqhibwWxyi4hMBWYC33Sw/WhVnQb0IL+kuNNYPrtdNNkvu9t0GGNMF/XMM88wadIkAoFATpzHK10t\nX9OxbLmW4waUcNbwXpw1vBfjBpRkNJcsM9/9eW74ShEZLCInAz8UkZOARbpheW1ngbTRm5oQY1LJ\nit7GdEBE9gWqgJWq+oCIHAMMBWaoamwdIswWq76+nvz8fEKhUE6cxytdLV/TMS+uZWH/vSMux0NE\nulKTs3R6HygFXmm3vgaYB+wMzFPVJVJQWq1N6/tIYY/NgqgqweoFa4DoI5sYk0FWqDGmY2NV9U8i\n8mv3+bHAB0BLBnMyXcQpp5ySU+fxSlfL13TMi2spItbkzEMi0gvYRVVfA4pw5hesC99HVRuAxSIy\nTVWdcbkDjScGvnjxZf/gfQZKt4q2vk0aaCS4/O112rBqKvDTdL4WY+JlhRpjOtb69WNrJ7iAqt4p\nIhOBie13njBhAlVVVVRVVaUpvc1VV1dnRYxckA3XM5xd29Sy98akW4oGp7gQpybmNaDcXfd5pB1V\ndX7bcij4qYiMCS557WbEf4AUdBdCzaItDasJBW7UQNNDImKFGpPVrFBjTMfmicgVgF9EBgJvichl\nwLuRdr711g4Hl0kLGyHLW6m4ngsXLmTkyJFxHxfrdeko/uTJkzn00EOz4trOmjWLbbfdlsGDB0fc\nPnnyZM4666y05mSf+/jNnDmTbetXM7iiP4wat9moZZm4jgaAV4ElAKr6r3gOVNVlwGkikq/NG/oA\nG1V1bQpyNCYlrFBjTAdUdTYwO2zV/ZnKJVcErzoPAP/192Y4E+/Y5JuRbUmTNuayjq7j8EJh/Ydz\nmR8I8vHgF1jTs49dxwwTkZE4zaS/Af6eaBxVbQGs6tJ0OTb6mTHGJMkm39xcV5i00UTX6XUsyOPh\nDz5nv1Ej7TpmAVVdCMwFGjOdizGZYJNvmqyUi5NvplpXaH7WVWpqbPLNLYu9N/FTVXhnjvMkQvMz\nExPP3zQROUZVn/U6rhtbtzroEk9iTb/+RwzeuNSTWD0OOJX1tau9idW7nBW13rS4a/j0bXovT36g\nVBU/63c+gD7FyU++2RJSXp/yAp/eNzXpWMX9y7n4zeey6l7Nmp8ZY0wSbPJNsyUSEdhjn0ynYTZX\nLiJnquqDqQj+xnXHeBKnoWdf1g4c4kksgPp13hRqAEJfL/QkTlOLsCpUlnQcRZCWABtIfj6oQEgZ\nMm44Q7fffOjuuPPKL4QTn0s6jpesUGOMMcZ0AapK08p5gDOni9WOeCtHap/6AotSFrzuW0/iLO+7\nLd3LensSC6C8tMC7WA3edCeqW++j4J1ZSccJ+fIoGL4LfX1NScdqBhq+W0re2zOTz6s0+YKR12Ju\n+CoiF4nIVyKyUUTeFpFOv6IRkSNFZK6IrBeRVSLylIjskHzKxhhjzJanaeU86j5/iLrPH2or3Bjv\n6ILZhB68k9CDd6ILZkc/IDuVArtmOgljMiGmQo2InAzcDlwH7Aa8AUwXkUEd7L8d8CQw093/IJxJ\noLKrnsoYY4wxBuDLj6BmhfP48qNMZ5OoGap6U6aTMCYTYm1+dilwn6re5z6/REQOx5nk6aoI+++G\nU2D6XWvvaRG5CXhZRHqram2SeRtjjDEdypGmRN9T2H/viMvGI9sNhz4Vm5a7pj1FZPtE+tSI80uy\nZ7kvbwKwsyAbN2rwoXoNTVbV9d6naoy3ohZqRCQf2AO4pd2mGUBHU22/AdQB54vIvTjVoWcDb1mB\nxhhjTMq9M4fQk5MBt0lCDnRqFxGKKsZmOo2cJXuMo63sO2pcRnNJwgZgaLwHiUh+L1/ek0Pzivc+\nuqS8fNu8Iho0xBtN60c+17D6N8U+/49TkKsxnoqlpqYP4AdWtlu/EqdZ2WZUtVpEjgSeAu7G+Z/y\nDnBE4qkaY4wxxqSGFyO6hUIh6j59AIBuO56diTl7TgdeiPeg3r68f55Q0ufAQ4t7F7eu6yk+jiju\nXTiusMeAq9d+8+iKYLOniRrjtZSMfiYi2+IUaO4HHgG6A9cCU4EDIh0zceLEtuXx48czfvz4VKRm\nstTMmTOZOXNmptMwxuSKUeM2dRrtut+6mzTyosli3acPUP/lI23Peww716v0YnU/cc5/IyLlW/kL\nDzukqFdxpO09fHmcUdq/3y3rvRn5zJhUiaVQUwMEgf7t1vcHVnRwzAXAUlX9TesKETkD+FZEqlT1\njfYHhBdqzJanfUH2mmuuyVwyJqd1lQlATXJEhOATDwDgz4GmZyb1dMEcdMqdAIiC7NklPzfbALTO\nVeOOOnusqt7ayTEH7F/Us1dnhbhRBd3SXuVkTLyiFmpUtUVEFgCHAI+HbToEp+YlEh9OQShcKGyb\nMcYYY0z2+HKRM/JZ63IChZpuO54dcTldVPW61mURmYozCu03kfYVkaNVdRrgzxdfp9PVd/1hNsyW\nINYCxm3A2SLyExHZSUTuACqBewBE5EYR+V/Y/s8Ao0TkahHZXkRG4VSJLgEWeJi/McYYk9VCoRCh\n//7deYRC0Q8wGaFDhkFpdyjt7iwnwOfz0WPYufQYdm4m+tO09wYQAI4MXykig92pOn4oIicBc+c0\nrlvdWaBPAxtTl6UxHonpN05VHwV+hTN887s4o54doapL3V0qgG3D9p8NnAwcizNAwPNAI3C4qtpv\nhjHGmC3H1H+izzyMPvMwTP1nprMxHRDxQXEpFJc6y13fSziVLJPara8B5uF80TxPVZfUhFoWLmyu\ni1jibtEQD9StqEltqsYkL+aBAlT1HtyamQjbzomw7nG+31zNGGOMMSYriYB279m2nAPOBBYBo4AP\nW1eqagPV/V0iAAAgAElEQVSwWESmqeoSgNpQ4KS/blg+67CiXjscXNyrpIcvj5AqH7U0MKV+5ao1\nwZZfAw9k4kUYE6uUjH5mjDHGGIf+6DxYvmTTsslOOTZinqpeEWX7/LDldSIy5umNq099uXHtJX7o\nFQINoq/WhgJ/UtVPReSBlCdtTBKsUGOMMcak0jtvwMfvbVoevW9m8zEReTFPDUDN6xcB0GffvyYd\nK51UtQmn//P9mc7FmERYocYYk1LB43fvdJ3/qXfTmY4xcVFV5ixvAGDcgJKE5i5h9gtQs3LTshVq\nspYN+W5M12WFGtOliUg/wKeqHc2ZZIwxCZuzvIHJi9a0Pd9nYGn8QSoHQ37BpuUkdNVaAGOMSTUr\n1JiurgrIp92cSSJSbCPtGdP1eFIzkmXkpPM3TfRx4vkZzcWYZKwqHehJnKaWFtavXRN9xxitrmvy\nLlZRP0/iBFAYMSbpOEHxEQj6WEVh8jmFoKXvQEIjq5KOpfnJ5+M1K9SYrq4JKA5fISIVwDYispWq\ndjRBbFQisi9OoWmlqj4gIvnAY8D5qvpdMkkbYyLzpGbEQ+MGlERcjofP5yP4vtMn23/yBZ7klWmq\nCu/McZ6MGpcThU8T3VZVh2U6hYgqh+zoWawBow/0LJZJLyvUmC5JRMpx5kf6SlU/bbd5ONAHqE3y\nNGNV9U8i8mv3+WnA9CRjGmO2QM99sYRxg/rjX7uWOXPmcNRRR2U6peS8M4fQk5MBd8I7DzrYm+z3\n93nfeBLnoH7KgMoKT2KVFBezsaHek1jFJaU0r/zak1jLPvmIXp+/kXSckM9Hw4Hn0qe8d9Kxmptb\nWDXzeQo/mJ10rGBJd7a+4tak43jJCjWmS1LV1cBqcezkrvvE/fmyiJwBlIvIfqr6WoKnaZ2IrPUr\nyF2A/sBK4Mn2O0+YMIGqqiqqqpKv1k1UdXV1VsQIF60i3+vzRVPW3AzA2ijnzYbrGS4br20qDBHl\nmEp1l9dRXb0+4VixXmvo+L2ZXxPg8cUtANTW5jO6T2L/Nkf3683F02fT/PW53HDDDQlfC21q7jTf\ndCmoraXEfX8bamtpTiIfXXQ1ADL8j57klox4PjMdifUaVVZWJnyOSETkLEABVPVBT4Mb0wVYocZ0\ndScBJarafgjKDUCATQWTRMwTkSsAv4gMVNUJInImMCfSzrfemtlvLKqrq5P+J+lFjPbk6fc63a5P\neXu+aIIFToftaK8z09czXLZe21RQVXqr06emsjK5PjWxXuvO3pvewTo2frPaWe7Vi8rKbgnlsrog\nH5/PT3FxMRUVFZSVlSUUp+YL5zX1yfC11IpjoLfzzXFRks3PsuU1Qeyfmc5k6vWo6mQR2QsYmtYT\nG5MlrFBjurpVwPcauovIqUA9sFZVE65jdY+d3W6dfftlTAplW58a7WA5XnOWruSuw8biv/qunGh+\n5tWcLsZzRwAfJnKgiJT2KOt1YXFJ6flFRUWlLS0tBAKBBau/W3FtIBBY4HGexnjOCjWmS1PVV1qX\nReQEVX1SVf+dyZzM9wWO2y3TKZg0yrV5PkSEnoX+tuVEHbW9M5Szv6ysyxdoTFabDsTd+UJE+vbp\nX/H62RdfNuTwH5yUX9qtO6rKpx8uHHj7tVeNK+/X/9oU5GqMp6xQY3LJdiKyi6p+kOlETHaxCUC7\njqrKYhatbmxbzjSv8sm5gQLIvQJsjjgDmB/vQX0rKh/741/+NXTknmPaSu4iwk677MZfHnm6/KKT\nj7m6dpUN+mmymy/TCRjjoY+AXiJyTqYTMcYk5o3lDby8uI6XF9fxhjtfTUbzqd7I2yucxxvViU99\nNW5Qf37x4lwuvvhixo0b52GGxnzPn4G4BscRke0GbTNkp/ACTbi8vDx+efV1fTzJzpgUskKNyRmq\n+ryqvhZh0ABjTBexqLaJFfUBVtQHWFTr3YR6xuQ6ERkJnAPsF+P+R7uLow884thOm6wN322PJLMz\nJvViLtSIyEUi8pWIbBSRt0Ukag9BEfmViHwsIo0iskxEbkguXWOMMbls596FVJTmUVGax869Mz9j\n9bgBJZw1vBdnDe+V8OSbsGmggLvvvps5cyIOoGgSpKrMXlbP7GX1zqSgWyhVXQhMI0qfGhEZLCIn\nAz8UkZOA5o0b6wOdHRMIdLrZmKwQU58a98N/O/AznOFsLwami8gwVV3awTG3AUcCl+OMxNETyPx4\njSZnicgQYHvgM1X9JsPpmCySF21Y6TTlYaLbZ2BpW4f8ZAoRXhERT0Zgs4ECUifbRszLsMOAd6Ls\nUwPMA3Z2f65/buoja0+/4JKKjgbDmPXic0HA72Wixngt1pqaS4H7VPU+Vf1UVS8BqoELI+0sIjsC\nPweOVdVpqvqNqi5U1Re8SduYiM4ChgCHZzoRY0xiWgsR4YUbY0zM5uPcn3VIVRtUdTEwTVWXqOra\nhvq6J6b87Y76SPuvWlnNX2+6xkYJMFkvak2NiOQDewC3tNs0A+hoqu1jgS+BI0VkOk7haRbwa1Vd\nlXi6xnTKBzQCFZlOxAvRbui25GYW8bJhpY3JXeE1etlQu5dhewFLgM6rpwFVbRslbdWK6kumPvDP\nnvPnzDr09Asu6bvDziOor9vAjKcfq3/mP1Nq19auPo7oNUDGZFQszc/64FQ5rmy3fiVwUAfHDAG2\nAU4GznTX3Qo8A4yNO0tjYjMd5zP3aKYTyUXnzXBamt576KAMZ2K8pqrMcUcaGzegxGpIssjyJyP/\nywxfP+CEuelKx2S/GSTQolZVg8DpIrLjki+/+KU/L28XoH5Nbc3kxoaGJ1S1yf4umFQSkXJVXZ1M\njFTNU+MDCoDTVfVLABE5A/hUREaHfzvQauLEiW3L48ePZ/z48SlKzWSjmTNnMnPmzISPF5EewK7A\n/wEX4PT96tKsJsaki/VJMF2ZfX6/p5wkugmq6qfARd6lY0zMKoCUF2pqgCDQv936/sCKDo6pBgKt\nBRoAVf1cRILAYCJMDBVeqDFbnvYF2WuuuSbeEJcCOwD9gM88S8wYY4yJIlKNWvt1aapRKwGsSsVk\nHRHpBuygqu+6zw8E+qrqf0XEBwRE5AigEJgLDFXV1+M5R9RCjaq2iMgC4BDg8bBNhwBTOzhsDpAn\nItuq6tdu8tvhNGNbHE+CxsRCVa9xfykGYPMvGRMX65NgurJxA0q4452atuV4BI/fPep6/1PvJp5c\n+vXA+SLamGyzp6rODHte4j4AdlTVj0VkFBDAKZgvEpERqvphrCeItfnZbcCDIjIfp8ByIc7wzPcA\niMiNwGhVPdjd/384HcruE5FL3eQmAXNV9e1YkzMmTpOAMpzC8+kZzsWYLsOrYYtzlfUny24iQs9C\nf9vyFq4SiDjVhhfG9Gj0JE5gfQPV9Wui7xij6q+/8CzW0hU1nsRp0GI2bpV8N3IFfKu/o7lubdKx\nQqrUlvSDXT0YJNbnw+kunxhVnRb+1F33SPg+IhLXKD8xFWpU9VER6Q1chfML8yFwRNgcNRXAtmH7\nqztT7Z04o55txOm8NiGe5IyJ0+dAE7Ah04kYYxKTS4MWRKoFaL+ui9UCmOw3khTOJ7PbzjulKnRS\nhgwf6V2skaM9i2UcIlIOrBKRI3HmR/pSVZ9s3a6qn3RwaFy1jjEPFKCq9+DWzETYdk6EdStxRqIy\nJl1eA0px/qibHGXfmue2ZDp952BTImPi9RhQlKrgTTXeVAIt/+xTen30iiexys67noanbvckVsnx\nv2LDlImexPqiaRAb7r0v+UD5eWxz/RWUr1yUdKjmvGKmv17LvNsfSDpWt8q+3LBssy7yHRmhqrPc\nws2LqvpBjMd9JiI7AQOBRaraUV9+IHWjnxmTCfsBDwHJ/+Ybz4TfRAavOs9Zd/29mUrHeMwKEsZk\nD1XtqK+zMZn0oYiMALZT1fsBRORUVf13lON2UtVX3WZoW4vIOlXd2NHOVqgxuaQc+CVO28xrM5yL\nMSYBNmhB7sumuXfyno46R2VM4yMP/MG86HG6yCj97ihVg4AG4Fu1+QVMklR1tYjs1lqgcdWLyPHA\nWmArVZ0S4dDW9se1wGBVfbOz81ihxuQMVY17HGhjTHaxQQu+L5dulk12E5Htt6roe9fOQwbvNmL7\nrX1r1tfpx19/W9+vd9kdq9asuyvT+ZkuLyQiElZIXo3TZaAb0NJ+Z7ep2ncAqjojlhNYocYYY4xJ\nAa9qAbyUSwMxeCFwXFyDK3Vo2RN7exInU/Lz8kYOGVjxwiN/+k3F7jtt17a+fmMj1/z94ev++8Jr\nVdU1tRnM0OSAt4HdcUZHRlVnA4jIOGB9hP139nyeGmO6CrdUPwzwqeprmc7HmK7EBmDITl7fLCcz\nEANEH9Etkf5T/3t7DaOHdQdg/scbOHjPXnHHMMmp6NPrkZf+fkPFVhV9v7e+tLiIm3/1k+41a9Yf\n9tBz3nTsN1smVd2AW6Bpt35OB/vHVaABK9SY3HIb8CQeffkpIvsCVcBKVX1ARH4BbA/8rZPhB00W\nssEKTCaE1wKsbQnwi/eXAXDXrgMpy8/Mv19VZV1TsG05G4we1p2r/vE1ANf/dNsoexuvicieJx26\nX7/2BZpwV//0lDIr1BgviciJOHMLPgr8EPgCWIFTm7NYVaO3vW3HZl43OUFEeuBUbc4EXvUo7FhV\n/RPQD0BV7wKewpmryRhjuhzpYNlsedz5BAGG7jdqRKfVY9sOrEhDRmYLUwS8gtPC5mVAVPUznPkG\nv0skoNXUmFxxKdATZ/SzMvd5skLhT0SkPzBOVa+LtPOECROoqqqiqqrKg1Mnprq6OmUxmpuakj5H\nWXMzAGuTiJFsHrHmkA3XM1wqry0k/r72S/K8Xkoml1S/N3d8sYrrhlW0Lf9hWGVCMSOJJ0btmgDF\nEnSX11Dtj2+u4mjvcSy5dDb4wZOvrwZg+fL43peu8PltH6eyMjPfj4nIYGAscLiIlABrlqz4rgGn\nw3ZEdQ0djqJrTETdJU/rvj935kpVDS8dNwIHAe8CuwHbiEgQp6Ym5glwwlmhxuSK24FTgEeAn3kU\nc56IXAH4RWQQ8E/gJREZpqoft9/51ltv9ei0iamuru70n2QsfSY6i1FQ6ByfzD/iYEFB0jGSzSPW\nHDJ9PcNFu7ZexEj0fY1luud4YybavyfRXNLx3gzrUcQVi5YD8MMBZZ3mE255DOeJJ6+j+4eoVqdP\nzdHDe+HzxddgI9p77NWNerxxEv670K4PUKTmqbFE9Po6pVgNMA9nZvd5wIr/vvjauv+74NRuBfn5\nEQ/411MzGknhpJ4m99QR5FL/puakk4Jf9w/f3smcSrMTPacVakyu+BVOf5d+wGdeBHRH5gj/5TrC\ni7hdye5TPu903btn7JDOdEwWysYRvrLRmF4l/HaR8039zcMHxHxc+/laal6/CIA++/41oTzeqN7I\n2yucb92HlxdlZPjs9oMf5MJAAeHXKdlrlGqq2gAsFpFpqroEoKJP7+tP/c2frv/3TVf2al+wmfPu\nIr1tyhNLgKEZSNd0YSX+sEausXzrlCQr1JicoKrXiMhVwP+AtzKdj/FOpIJV+/VWuDJe8uozFz5A\nxbwTDmDPrTciI/Zk/o9+xJCTT04+0QR9UtuUsXNHYgMFZIaqtjXxWVFT+7d+vcuCI3544R9OOWJ8\nj9HDh3ZbX9+g9z8147svvl3+WfWq2h8AqzKYrumCSvLCaoKbU38+K9SYXHI7cCXwAE7HM2NMink1\nz0eu61aQz9+P3Af/1XcxZ07EEUzTYtyAEipK89qW4xWtZs5q5bqu72rX/kNEHrj1wceP6tmtdJem\n5pYN6+sbnlLVr4Etfk4jE7/v1dSkgRVqTC6ZDUwBRmY6EWOMCXfU9oMB8JeVcdRRR2UsDxGhZ6G/\nbTkb/PPZaq48bau25ct/vFWGM9pyqWozztQIT2Y6F9P1dc9L7yDLVqgxuWQf4DfAAmCXDOdispTN\nT5O7rH9P6qWiZm6HQcX8cfISAI4a2zvm46L1+YMtu2mqiJwMbACWqur7mc7HbHlK/Okt1MR8NhG5\nSES+EpGNIvK2iOwT43E7iMgGEVmfeJrGxOT3OAWa7BiD15g0OG/G0rbRwoxJp+dWrGNtS4C1LQGe\nW7Eu4TijhnZj4Rd1LPyijlFDOxxV2MSvEfgKuDDTiZgtU4nf1/ZIh5hqatzS/u04Q+XOAS4GprtD\n23b431RE8nGG2J0J7J90tsZ0bhKwL86wk/FNvmCM6fKsf096jSsv5RfvLwPgrl0Hxnxc+xHdnp94\nKLvvPIjCvqP4puxHjDkhc4Mo5JiZqrpORO73KuDSL7/0JE5Ti9K01RhPYgF87d/as1hf5XvTJTdQ\nWkThTz0oTwrUrIXa4DbJx1If21UNZvBO1ySfVl4eN5x7Wqf7FBSnt0FYrGe7FLhPVe9zn18iIofj\nlP6v6uS4m4GFwGtYocak3uXAE8CvgSsynIsxxrR57osljBvUH//atcyZMyej/WqSET6iGxPOgMUb\nkZFj8N99N/6yso4P7ERJST6TrhhPedWtGR1EIQddLk7HqVe8Clj29nOexFm1zRj8z//Hk1gADX+9\ny7NYS6+43pM4PS/5KUtuujv5QAUFbPuLs1l1Z/JNp/Mq+1F6+IF8d+8jycfq1yfqPvklWVaocWtb\n9gBuabdpBp008xGRo4AjcWYGPTGJHI2J1XpVnSciB2U6EWNM1/XemTFMx3FGfL1zxg3qzy9enIus\nuJi77/bgRicLzFm6krsOG9s2oluiBbVDxzrfspdleBCFHOQDQkB8s9ga45H84siTuaZKLEWoPoAf\nWNlu/Uog4s2jiAwA/gEcp6oN2TLCisl560Xkz4A3XyWZrJCKG0zjsMlVTTK8GtFtxtzFjBlRQV6c\ntVj2tyEqTyaiNiZR+SUdF2pE5ESgDHgU+DHO9JzvAMOBV1R1WbznS1W90BTgr6r6tvs8aqlm4sSJ\nbcvjx49n/PjxKUnMZKeZM2cyc+bMZMOEgOvcn8YDUW8atuwbBkO75khA8KrznPUZGGXOq1x2e9D7\ne0GvajVy0ZgRFVw56XUKH86dWqws8QhwEhCI9QC3udpYEdlBVRuAl1W1NlUJmtyW13nzsyKcppHD\ngAZ33TjgC6A0ofPFsE8NTumpf7v1/YEVHRxzALCviEx0nwvgE5Fm4CJV3ew/THihxmx52hdkr7km\noU5s5cAl7vK1yWdlskEqbjCNSbdsmacml9jfhqgeB+bi3MNFVdat9LhB/cpvP2iPET2qdtmxV+2G\nuuZ/z5i9ZlC/8nnLVtWekdpUTS7qrKYGZ3S+g4B3gRKcL6TfA/oB25NATWPUQo2qtojIAuAQnF+Q\nVocAUzs4bES758cDvwNGA8vjTdKYGLVOUtEiIuWqujqj2eQAu2kwxqTaPVPf56rz96L3mOu5/fbb\n7UtO77yK82Vf+z7Rm+lT1uOkXbYf/Nenbvp1efeS4tbVhZf9+OiKZ2a/ffQvbrtv9orVa1OZq8lB\nBd0KOtymquFliPlenC/W5me3AQ+KyHycIZ0vBCqBewBE5EZgtKoe7Cb6UfjBIjIaCKnqx14kbUwH\nDsL5I34kcCDOaGhbNOszYUxuSbT/STZPVPmzE3d1mp89/Ttrfua9GuBY4MGOdhAR/6B+5X9+9uYr\ny0uKCjfbfuw+e+Z9+NW321/zr46+xzYmsvzijgs1qRBToUZVHxWR3jjDN1cCHwJHhM1RUwFsm5oU\njYnZauAZYBcg7g5mxpjMyeab7mxi/U9MrFT1thh3PeCE/Ub3iFSgaXXhCYeUWqHGxCuvpOPPVErO\nF+uOqnoPbs1MhG3nRDl2MjA5vtSMidscnFrFaTjzI5ksk0jn8fY3sufNcL5LufdQG6U0WTYQRGT2\nmUuvNz9cwZ8u3bdtnhrrc5R2/YcOHtC9sx16de+WrlxMDskrztJCjTHZTlX/B/wv03kYk2pWq7Hl\nSrQAkM3DHyc6T034Z9wKnvERkX1V9XX36TcLPvlqLcfRu6P9l9esSVNmJpdYocYYY0xa2EAQXY9N\nVGk80i9s+Y3/zX9/Tc3a9b37lPWIuPMtDz+zHoi80ZgO5JUUpfd8aT2bMWkgIsNVdVGm8zDGGABV\nhQ3r2paTmZC6z75/Teg4K8BGp6r4fnB223IuTxyuqo+HLWtJUeE5B//yj48//acr+m5d0bdtv2Aw\nxJ1Tpzc8MevNt4CDM5Gr6bqspsaY5PXKdAK5wpp3mHTJ5uZRSXtnDvQo27S8xz6ZzcdE9s4cQk86\n3X99kPB1SrTgmSwROQsYCKCqN8RzbENj0+v5ef7DDvrFtf8c3L/PoD2HbVe4dkN98NV3FjUGgsGH\nV6xeexXQkoq8Te6yQo0xCRCRXcOers9YIsakQU4XAIzpwlSVppXzACjsv3daa3tUdbKInIYziWHc\nWgLBd4E9RWTgnA8+3Q5nlvf3VDUA5HTNlUkNf7E1PzMmEbsDfpzPdDPwfrIBRWRfoApYqaoPiMi5\nOO2QZ6nq3GTjp4ONbmU6k001cTndPGrUOOebf3c5E7LpWmctD65T08p51H3+UNvzooqxyecVn2+B\njckEUNVl2LQIxgO+4tIOt4nIiUAZ8CjwY5zC+EfAAGBpIvdZVqgxOcH9hupaoBrw6quBsar6JxH5\ntfu8XFVvEpErgC5RqDHZJZZvOlWjFzRzugCQg0TEmpx1AV5cJ1Ul1Lyp/1QG7ItNaWCyhBSVdLa5\nCHgFGIZTKwiwrao+JCJnkMB9lhVqTC4RnG+oKjyK11qF33onqu1+fs+ECROoqqqiqqrKo9PHr7q6\n+nvPo938tt+/o3UAzU1NnW5Pl3Tlkanr2dHrivf1xnNtwbv3tay5GYC1Kbg+8eYWTy7peG+yRbZd\na21qTjofT16TKgWL3nbiDd8TEmhupWtqoSUAwJo1tayl43wqKysTy7Nz9cB2qQhsTLyksLizzY3A\nQcC7QAnOPdfHbg3Ol4mczwo1JpdMB07Gqcr0wjy3VsYvIgOB1SJyJfBapJ1vvfVWj06bmOrq6gj/\nJDefzyRc+/0jx3AUFC6NeEy6pSuPVFzPRL+53fy6dH5dIb5rC969r8GCgoTivHvGpmWvmkfFmku6\n3ptskelr3V7NF06cPknE8eI16YLZhGY9C4Cvd28kgVqbjdKbDdXOrVX33r0prkj7Z6aWDv5HeaF+\n3AmexPEV9qDgsrjGMujY3VPpd9213sSa/grDH73Dk1Dr1jUx7OZfR98xGhFK9tiNAccckHSokELD\nqu/ou4cH5d68fDiu84+ar5NCjapODXs6P/mErFBjcoSIDAO2B/4MXAnMSTamqs4GZoetuj/ZmMaY\nLY+qOqOeAYwalzMdrv3X35vpFDy12dDbCQXpYDl9FgG/ch+eq2j+zpM41T360r9fv+g7xqiyaYVn\nsSo2fO1JnHXLYeODDyQfKC+PvmP3oH9J8n83moOw9P1v8M97OelYoW49o+4jhZ02P/OcFWpMrrgQ\nmAzcC5ya4VyMMWYTj4YKzlWZGgJ5c9LBchwRRPAV9GxbTicRGYzTBHtSWk9sTAekqNPmZ56zQo3J\nFY04Y+jPxxmnf3Vm0zGpYiM2mVjlWk2CSS0R0O4925YTUdh/74jLabIt0HoXuTjdJzemvSh9ajxn\nhRqTKxbhDOv8ObAbHgzpbL7PChPGJCgLhnQ2MfDgOolIJoZxBkBVZ4nI5Tgdrl9IJIaI9CvCd14+\nvt0Vrasn+KA60xgkNPeN2bKpvyCt54u5UCMiFwGXA5W4bTbdPgeR9t0fuBTYC+gJfAHcrqrWJ8Gk\nhKpOznQO2cjmpshNdl3TK9n3NtuGdLbPSmTZdp0S1EgCk2+KiJTgv6Wc/DN2pUd5Hwr8zYT4jPrj\nlrBxlYgclYJcTY7TvMK0ni+mQo2InAzcDvwMpwP2xcB0ERmmqksjHFKF8035n3DmDTkc+IeIbFTV\n/3iSuTHGGGOMCbeeBEY/K8H/x+0oOX8svXpIWH+iCop6raWl13OsfKWOoJd5mi2A5ns1bWBsYq2p\nuRS4T1Xvc59fIiKH43TOvqr9zqp6Y7tV94jIAcAPASvUGGOM6ZDVJBiTsK9w7tl+GesBIlLag7xz\n2xdoWpWRz56UVcy0rqomTllXUyMi+cAewC3tNs3AqZGJVQ/g2zj2N8Z4yG4UjTG5yv6+tRkPvBjn\nMQcPpbRnpAJNq+0pzbdCjYlbFvap6QP4gZXt1q/EmQk0KhE5GjiQ+ApBxhhjjDEmdktxRgCNSkSO\nVtVpQEkR/k6/UvcnOMS12bLFUlMjIkOBMcB3qvqiiBwJ9ALeBI4G5qnqvFjOl/LRz0RkHPAw8AtV\nXZDq8xljjDHGbIlU9YFo+7jz2YwFDheREmDhYhpqh9O9b0fHrKbZuyTNFiOU13FNjYicBuwNHAoM\nA05zN5UD/wZOBPoBMbdhi6VQUwMEgf7t1vcHOp3CVUT2AZ4Dfq+q/+hs34kTJ7Ytjx8/nvHjx8eQ\nmskVM2fOZObMmZlOw5gtljXfMWaLUQPMA3bG+RZ8STfJ+2Y1zX3L2fwmVFHmsaYW6J3mPE0X1xTQ\nDrep6sPAw25NzanAKhHZBWeewR8DbwFbATsAs2I5X9RCjaq2iMgC4BDg8bBNhwBTOzpORPYDpgFX\nq+pd0c4TXqgxW572Bdlrrrkmc8kYY4wxOUpVG4DFIjJNVZcA1BP80XS+e62KXlttQ4nP5zY320CA\nN6hdU0vL34gwMJQxnWkKRh9dXFU/Az4LW/VB2HL7/vydirX52W3AgyIyH2dI5wtx5qu5B0BEbgRG\nq+rB7vPxOAWau4H/iEhrLU9QVWviSdAYY4wxxnhLVeeHLS8RkT3msOaquaw5sRBfXhCVAPrVRoLX\nBJy+DlaoMXFpCnZcU5MKMRVqVPVREemNU0qvBD4Ejgibo6YC2DbskLOAYpzJOi8PW78YGJJs0sYY\nY7KT//p7M52CMSYBqroauAy4TEQKgRZVjXsiT2Naddb8LBViHihAVe/BrZmJsO2cCM/PibSvMcYY\nY2RCNCkAACAASURBVIxJLRE5010crKrXxXOsqjalICWzhdkYSG+ZOOWjnxljjEkd6+BvjOnAdzgj\nR30QbUdjUqGuOZjW81mhxhhjjDEmx6jqC+4E6jt5FbPeV+RJnEAgSP3GjZ7EAqj3FXsXK7+7N4G6\ng2+HHZIOI3n5tAShviX5lAIKwdKeULl10rG0uDTqPnXNgaTPEw8r1BhjjDHG5BgRuRbwAa94FfM/\ndzznSZx9zjiC0OzPPYkF8OFDMY34G5NZD73hSZytDtmHloLkR8HWPD/13y5m4xcfJR+rqJiWskpC\nwZKkY0koeox6q6kxxhhjjDFJ8gMhwLM2qu9NedqTOPv8cF/yPpoffccYrXvyec9iLf7vNE/iDNh5\ne9Y9NT35QAUF9Bm9A/LOnKRDaY9e6DYjaX5tZtKxpFevqPtYTY0xxhhjjEnWZ9F3MSZ1rE+NMVnA\nHcL8cqChddQYETkOGAl8rqqPZDI/Y4wxJopHgJOA9H5dboxrfWN6P3q+tJ7NmK5jPPBPYLmIlAGo\n6tM4E9HacFNbIBGJ+jDGmCzyODAYiKlXuIj48sR3eP/8wpf75xd+3i+/4KNeefl/FJGK1KZpctW6\nhua2R0dEZKiInCEih7nPdxaRK0UkT0SOF5GYp4ixmhpjXG5NzPGAAh8D7wEStt0HXAHcHOn4CRMm\nUFVVRVVVVRqyjay6ujorYuSCRK5nKt87u7apZe+NSbfKyspUn+JVoBy4JdqOIlLcOy9/xrhuZbv8\noHe/ngMLimgOhZhTt/a3U2qWn9/dn/eTVCdrcs+6jR0P2SYipwF7A4cCw4DTAFT1IxEZraoBEZkL\nHBzr+axQY4zLrYl5Gtqan00AGoCQ+w3CfjgdL8cCL7U//tZbb21bjuVbe1VvZ9qtrq5O+p+kFzFy\nRfj1BO+vVzzs2qaWvTcmh9UAxwIPdrZTn7z8B3/ab9CY8T1657euK/D5OKBHb/+Y0p79f7n4k/vS\n3T/CdH21dR3X0Kjqw8DDIjIU+H/27js8qjJ74Pj3TBIS0ggJgYCK2LCAoigWEETWgg3FLhZ2164/\n11XXinUVd0XRFcWyRcWyrqKiLgtiQQSxIXZEAQsohBoCKaTNnN8f904yhJSZzJ3MJDmf55knd+7c\ne9535s5k5tz7ljHAOhHZG9gI7CMi3YG7cVrNhMWSGmMaoKpFwLiQVbPcmzHGGJPwVPX+cLYTkYJd\nU9MPDU1oQqUnJXFxjx263/SLd0Mwm46hqSs1Qaq6hG0HtbjG/RvRFUJLaoyJgXie1TfGGGMiMOyI\nLrl5TW2wf7pHE1KaDqU8jKTGSzZQgDHGGGNMByIiQ+NdB9P+VW2pqb21BktqjDHGGGM6lu4hy/Pe\n2VRU1NTGn5eXxLg6pj2qrKipvbUGS2qMMcYYYzoQVX05ZLlwo7/6w7mbNzbYVmhLwM+ja35Z23q1\nM+1F1Zbq2ltrCDupEZHLRORHEdkiIp+KyKHNbN9fROaISLmI/CIit0RfXWOMMcYY46X1NdVnP7L2\nlwWPrPll86qqSgCqAgHe21wU+L+fF68pqqm+MM5VNG1Qa1+pCWugABE5A/gbcAkwH7gcmCkie6rq\nrw1sn4Uz5O0cYH+c8aefEpFSVX3Ao7ob02aUl5ezfv16evfuHbcYiVAHLyRCHYwxpj1R1XIRGfa/\n4nXHfFBSfK3AdgHUX6P6ykZ/zUOqusomGDaRqtpS2arlhXul5irgCVV9QlW/V9U/AIXApY1sfw7Q\nGRirqotV9RXgHuDqqGscpTlz5rS7strjc2pv1q1bx4cffhjXGIlQBy/Esw7l5eWsWLECgMmTJwOw\ndu1aNmzY0KoxEq0ujcXyIka86pNIdQlavnw5M2fObPH+9pwij9US0T6neFFVf1UgMH1NdeVhq6sr\nd11bXbV7UU31jaq6Kt51M21Tdfmm2ltraDapEZEUnKst9ScbfBNobKrtg4F5qho6684soJeI7NiS\ninqlPSYA7fE5tScLFizgiSee4I033mD69OlxiZEIdQB4+eWXmTdvXov2TYQ6hCZUSUlJACxevJhl\ny5a1aoxEq0tjsbyIEa/6JFJdgubMmcPw4cNbvL89p8hjtUS0z8mY9qKqZGPtrTWE0/ysG84s6mvq\nrV8D/KaRfQqAXxrYXtzHlkdQR2PatEGDBtGjRw8++ugjjj/++LjESIQ6AJxyyikt2i8R6rBgwQKm\nT5/O8uXLWbJkCb/5zW945513ePXVV+nUqROVlZUMGzYs5jESrS71Yy1dupQRI0bw5JNPMnDgQAYM\nGBBxjHjXJ5HqEvTyyy+zYsUKfv75Z+bOncvFF18c9r72nNrOczKmPakq39zsNiLSFzgIWKuqs0Rk\nL+B44D7gBmC1qj4RTnk2+aYxraB3795R9wGJNkYi1MEL8apDaEJVXFzMwQcfjM/nIyUlhbS0NA48\n8MBWiZFodWks1uDBjV3IT/z6JFJdgk488UQ2btzInnvuyZ577hnx/vacIo8Vj+eU6AaMOcGbQBmZ\n+Pfa35tYPEOXE0d6E+qRb9jxtOM8CZW8XYE39UpOQrt1x7ffIVGHkrR0fLnbkTp0ePT1Sk+HF6c1\nuUl1E0mNiJyN07LrKJy+92cDqOq3InKAqgZEJB3oFG6VpLmZz93mZ+XAmaFDAIrIw0A/VT28gX2m\nALmqekLIugOAj4GdVXV5ve1t+nWzDVVtM70SRUTbUn1N05o7nu4/3E+jLCPqGIlWl/ZYn0Spi4h0\nB1JVtX4riFath5ex7Dm1LfZbzTSkse9KEVkN9AhZtUZVCxrYri9wILAeWAlsxOnLfz9OolOqqo+E\nU5dmkxq3wI+AL1T1kpB13wNTVfXmBra/BPgr0D3Yr0ZEbgIuVdUdwqmYMcYYY4wxxoQj3NHP7gd+\nKyLni8geIvIg0BN4DEBE/iIib4ds/2+cqztPiUg/ETkZuB6Y6GHdjTHGGGOMMSa8PjWq+qKI5ALj\ncJKZb4BjQuaoKQB2Ctl+s4gcCUwGFuBcSrpXVf/mZeWNMcYYY4wxJqzmZ8YYY4wxxhiTqFpl9DMR\nuQz4E85VnkXAH1X1/Sa27w88jNNxaAPwd1W90+uyROQwnM5IBwJdgGXA31T1Sa+fU8h+uwGfAaqq\n2V4/p5B9/ghcjHMFbQMwRVVv8rIcETkWuAXoB1QC84FrVXVpM+UMdcvZH+gF/FZVn25mnxa/J0zs\niEhwQt3DgVOAs4C+QABnOMbhQJqqvhCHOvhV9VYRGYnTv6/J91hHFuZx7Izzv+tInM+74vzf/Brn\nc9zZXb8USAUGquq9HtYl9Hj2AD4Msy5fA39V1cYmi45KIr12XmlrxyAcMThOcX9OxpithdunpsVE\n5Azgb8BdwL7AB8BMEdm+ke2zcCb6LMT50XslcK2IXOV1WTiTh36F8w+uH/Ao8HcROdPjcoL7pQDP\nA3Oaey7RlCUi9wOXANcCewDHAnO9LEdEdgGmuc9lX5w5i9KA/4XxtDJxvhD+gNP3qknRvCdMzCmQ\nh/Pj5nR3XQDnKvAm4HugJE518LmfudQYl98ehHMcNwN743x2OwEVOKPV9KHuu6Q7zpxk7wKhky97\nUZf6xzOcuqzF+YH6UQvrEk194/HaeaWtHYNweHmcEuU5GWNCxDypwbkS8oSqPqGq36vqH3B+nDZ2\nduMcnLMhY1V1saq+AtwDXN3I9i0uS1X/oqq3quqHqvqzqj4GvIKT5Hj5nIImAF8CL4XxXFpUlojs\nDvwfMEpVp7vP60tVfcPj57QvzvvnJlX9UVW/whnxbhe3/1WjVHWmqt7sHttw2j9G856IG3dgjZNE\nZKSIXCciU1sYZ4KI9BeRq1oaI4Y2APcCfqAG53guBWaLyIE4Ceivje8e2zrgJPUAEU9s09zxcx8f\nKiJni8iFIuITkQY/Z00dQxG5WkSOFJHzmooRY+Eex+DndTqQ4y6/hnOWuxznjPfhwJ9x/n94Whfq\njucOYdZlOM5JlP7ufAexUL++Q4EsoEREHsI50RP62hXgzM0wBHgd50RS6Gv3GlAlIjeJyO/d98Q0\n9+QO4LQyEJEbROQ89/6jbiuAWD2nRD8G4fDyPT6cxHhOxpgQMe1TI60wx000ZTUSZybwi6pe5GU5\nInIczihy+wGnAQ811/ysha/ftcDvgceBK3ASj/dwmoWt87CcnsC3OKPa/RPIAB4CdlfVsGeIEpES\n4PKmmga19D0RbyJyPs4gGYLzQwbgzeaa5zUQZyxOUqc4TUYjjmEi19jxw/nhWeneUoBSnDPAK3Ga\ney5R1bfrxdrmGIbEUZwz3zXA6sZimLYhjPdNg8fb3TYLp+nvIOBnoCtOs94ROFcN1gG74FxJ2B2n\nadQjwO+AFFWdEOvnZ4wxiSrWV2q6AUk4l9RDraHun319BY1sH/oF4VVZWxGR43G+PB73shwR6QX8\nHThbVZttbhVNWcDOOJfKzwDOw7nKsQfOGUHPylHVQpxmbXfifFEX4zTh82i64a209D0Rbz8B+TjH\n41icHzCjWhBHgfdx+hO1NIaJXEPH7ySc/5s+nKR6L5zPThHO/46fcJpi1lf/GIbGgbozwU3FMG1D\nc+8baPh4p+I0h/oEpwlUkbv9KJyrXgOAdKDaXf8/nKtCo3GSqF1i/LyMMSahtcpAAW2BiAwBngOu\nUNWFHod/BngkZIbiWM4878NpC3yOqv4AICLnAt+LyCBVXeBFISKyE/Aq8CROP6EsnCYnU3HOSHZ4\nqjobp7kGRDFHU8hVrPOjrpQJW5jH78aQ5VfcvzMbiBXOMRwbsrxNDNM2RPC5b+p4/9b9+27IulfY\n1jEhy0+FUT1jjGm3Yn2lZj1O+9Ue9db3wDl71ZDVjWyvTezT0rIAEJFDgRnAzar696a2bWE5hwO3\niUi1iFTjNNfKFJEqEbnA47IKgZpgQgPgNlXy03i/gpaUczHwq6re4PbZeR84FzhMRAY3sk9LtfQ9\nYYwxxhhjOoCYJjWqWg0sxBklJNSROO2EG/IhMFREOoWsOwpY1VTfiRaWhYgMw0loblXVhxrbLspy\n+uN0rB/g3m7F6cMyAOfKhpdlzQeS3SspQO1IZUlAg69fC8vx4SRCoQIhj3mpRe8JY4wxxhjTMbTG\n6Gf3A791RwraQ0QexJkH5TEAEfmLiIR2iP03zg/+p0Skn4icjNMZPZzmOxGVJSLDcRKaR4H/iEgP\n99bNy3JU9dvQG06H4oA7ktcmL8sC3sYZZ/8JEdlXRPYD/gV8GNL8zYtyXgcGisgtIrKriAzEaYq2\nAidBapSIZIjIABEJjqDW272/QyNlRfOeMMYYY4wx7VzM+9So6oviDPE7DudH8jfAMaoaHOa1AGf0\nl+D2m0XkSGAysACnA+S9qvo3r8vCadPcGWciyD+FrF+O0+Heq3JarAWvn7oDHkzCGfVsC87IO9d4\nXM774sxtcwPOfDjlOGP2j1TVLc08rQNw2ooHh967w71NwRm5zbP3hDHGGGOMaf9iOqSzMcYYY4wx\nxsRaazQ/M8YYY4wxxpiYsaTGGGOMMcYY06ZZUmOMMcYYY4xp0yypMcYYY4wxxrRpltQYY4wxxhhj\n2rSYD+lsjDHGGGOM8Y6IPIwzsbqq6mXxrk8isKTGGGOMMcbEhIiMBfoBacCjqro4xuUdAqxQ1ZWN\nPP5H4AngTlW9UkTuBh4AvgP2xZmnMMddrgR2AP6tqvPd/Qeq6mfu8mHAucAGYKaqzhGR9TjP1wd8\npar5IWXfBmwG1qnqs+66TODPQCnwMuAHjgP6APcB/YHDgR9VdVLIU/lAVf8tImMaeZ6/Bd5Q1dUN\nPHaY+xyzgLmquqKRl7NNsaTGGGOMMcbE0rNACjBIRApxEoGdcCbVvhb4ENgduAcYiZNIbAKWAGNw\nfvCvB1KBr9zHXwEGABXAQHebb3CSgXIRmQD8DxiiqpcDiIgA+e6k3s+KyIvAI6q6TkReAa4Gprl1\nVveWDawJeS7nAp+F3H8dmA9cDswBXgPODnkMt+wcwK+qD4jI/e5rAnCE+1w+AP6iqtcD34jIeTiT\noZfiJEIpIiK67QSTIiKdgXvd53+wqv4W2B7YXUTuaOD12+A+txeBm4DbaQesT40xxhhjjIml04G7\ngdlAZ3fdeuAQnN+izwLvAQcCme5jB+H88H4TeAioUtXbgSHAv4DfAb9R1beBAlV9RFXnAuLGX6uq\n/waqQurRFSgBUNUFwEpVneM+Vgp8DIwI2f5R4I/A6SKyh4hcCewnIn8QkX3dbUa69Znh3t8MpLu3\nTfVeh2BCEmhkvQKIyJFAjqq+r6rvqOqtwI/AYTRsb+BTVX3M3T8fWEvjr5+6r0EVzuvdLlhSY4wx\nxhhjYiV4RWAMcAOwI87vz+DND1wCHI+TVOyO0+wr2Jqo2o1R7d4XVS3FufKzwF23SkQuFZFh1EsQ\nQv6iqkVs/SM+9KqHqup/qEumAC4DrgDmq+p3qvog8JmqTlLVL9xt3gDOAC4WEZ+77yM4CVFo8GIg\nyW3+Fmy+dinwFnAycBvwb7f53K3u4/1E5DARuQ44EvgyJKSELH+FcxXsYvf1PBF41d1mm9evNoBI\nN+AX2gnZ9iqWMR2HiPwM9K63WoH9VPWr1q+RMcYYE7228v3mNsWaoqpfNrtx3T674yQcV6uqP8Ly\nrgIeV9XyeuvHquqUSGLFmtv35aeG+ryIyEycxM+vqiNF5DSgF1AIrHavWjUX/yTgLVUt87jqcWF9\nakxHF2wzOx34IWTdurjVyBhjjIleXL/fRORm4GecZmLP4nRK/xQ4Cihz67EBpx/KmSKSjtMh/lFV\n3dhUbFX9HrgypKzTgFNxrmQU4jQBK3fjHwV8AgwCZgFdgHwR+RPwDHAMMJOtr9q0KrdT//Y4r8XO\nOP140nD6yGSKyP44gw/kAhnAPJzXdEdVvRtAVadGWq6qvupB9ROGJTXGOP6lqq83v5lpTTZkpTHG\nRC2e32+C0zxrF2CJqn7gNhFbBHRT1f+6o6P9DKzA6UuzN9DsVYZ6ggnAryHlBpO6FTj9Sz4DDsXp\nmyI4Tddy3f36tvD5eSUFp36n4YzEdix1/W7KgIuAv+E0K/sS2MK2/XI6PGt+Zjo0EfkJ5/L8/6g7\nk4WqXhVhHBuysq5sz4asFJExwSEr3Q6f9Z/nb+lgQ1YaY0w4vPp+i6L881T16dYoyxiwKzXGBB0X\nsqxAS/7p25CVNmSlMcYkGi++3yJmCY1pbRGNfiYiP4tIQET8IlIiIj+JyAsicmCsKmhMKzlJVZPc\nW0uTfRuy0mFDVhpjTOLw4vutSSJyvoicJCIjReQ6EYm4f0dIrAki0l9EroomTqy4LTOCyxE1WxOR\nq0XkchE5WUROEZH7ReQGEckVkQdiUFcRkeNE5AkRuUVELhSRPUXkRvc7vreI3Ol1ufES6ZDOoZ3O\nXsA5e3wa8L6InOJx3YxpTdL8Jk2yISuxISuNMSYBRfv9Fi4fTif2MmCOiOzWwjiLcPq+lEcZp0VE\n5FoR+b2I7CUifxKRYCuIQSJyNrC9iFwvIvsBQ0VkjIiMFZHRInKze3LwIfe7pz7FOblYidMcXHG+\n74uALxrYPirq+B+wGPgrTl+lTJxjFXCbZS/zutx4iahPTUj7zNGq+rr74+ZZ4Eycph47qGpFTGpq\nTAw01uYY+KeqLvKwnPtV9eoI94lmyMrRwKz6Q1a2Ne6QlcuA3WzISmOMCV/932wxLmsEsBvOD+YR\nOH06+6jqxBbEOg/npNhVOK0DWhSnpdzy5+D0G90TqFDVR9xWSbvgNCv/VVWfdq/aCFCM00JjM84J\nuMMban7nbq/As6oacMuqwkk2rsEZanpJDJ7TjTiDKKThjJzWHefk6is4zbMfUNVfG4/QNkSV1Ljr\ndsfJABU4SlXfiUVFjYmFkPd0fTH/EjDGGGNixf1+2wE42b7PTEfgRdvK5SHL3T2IZ0yrUdWd4l0H\nY4wxxmv2/WY6mkj71DRkx5DltR7EM8YYY4wxbYyIpItIQ60fLFaMJWq9WlNUSY2IJFE3TOoGnCFf\njTHGGGNMx5OPM+KnxWp9iVqvVtPS5mcXiMiJOMOm9sUZcegSGyTAGGOMMabjEZFBOCN89haRMlWd\nbrGajXcKztQM86KMk5D1am0tHShAgS3AOuAj4H5V/TQmNTTGGGOMMQnPbf50sKq+aLFaV6LWqzVF\ndKXGOp0ZY4wxxphQIpKKL/lkUjKOBALiS6pEA/9T1ZoWxOqckd759B2363lYQLUmJSWlpKam5s1I\npzZwY2XgSzmLlMwhoJUiviLQ2apaf3LocGJlp2ZknZ3dvddBGghs8SUlrdFAYK5GcnWgLlZOZpec\n89IzMvf3+/0lG9et/Xcg4P+whbG6ZfmSxnZPThngVzaKyI/AwhbG6pGG73fJyF4BWF+Of4qqftn8\nnokhois1xpiOSUTycCbaXQzkqOprYe73gKpeVW/dCcCbqloZuo2I3KKq7WZmY5NYROQ2nLkYNjfw\n2GE08r4WkdNDz3w29J4OeexmVb2r/rb1YzRVn9BtRWRH4CTgG6BUVT+O5Dkb0xokOXU0vpTJvrzd\nsiVr+wwIoJtWbA5s/HEz/qpz1V89J9xY+bldz8nKzLj3/LNOyR4x5MD06uoaXpn59qaX/vdW8fqi\njadXVFZ+Ena9UjpfjC/lDl+3PbIlo0dn1I8W/1Qc2LRiI9VbTlINfBVurMyu3a5Jzci6duDxY3K2\n232f1OrKChbNmb7xh0/nri8vLjrBX1P9fVh1EpGu+d1vzczqcunx55zfddd+AzptKStl9msvbvji\nw3lri9auPk5Vfwo3Vm5S8r3ZvuRzTurSLW+nTp2TywJ+ZpUUrf+momzVBn/1saq6MsxYvnSSJqeT\nNHpvsrp1JSWpggCLKV23hsrl5fiPU9WEHwzMiyGdjTEJQkQuBVJwJqzMx5kMrRhnYq9dgJ7u7Rbg\nJeB5nInCzgU24kw4djiwCfhSVd93Q/8WeBLYGzhKRIYAs3CaoA5zywlOlPYksKeq3u7WqR8wBgi4\n5e0HzBWRccBKoL9bxioR2VlVf/T2VTGm1lgROQBnkrsbVfUaEZmIM/nuKHfete+BfXEGv5kDHCwi\n84Abcedkcye33RNYFpKA7A986y4/4G4bHB30EBHpzNafR3W3vdktK0dV/+JuuxjnM7kZ2Kyq74jI\nXwFLakxCkaSUEZKW+4+kXY7Kk6SUugcyemT7uu+dXbPsjRdFfEeEk0Dkdu1y4n799njwtSceyu2c\nlla7fsig/bpcd+nvu4w44/evi8hQVV3abL1SOp8r6fl/SepzeFfxJdU9kFmQ4+u+d07NsjfeEJGD\nVPWX5mJl5uZf3nvvA28+4U9/zfEl1f1s7r33oK4bC1d0/c+4C94Wkf3D+dGf273HDQcMO+Kqy2+f\n0MXnqxura8DBh+atWPZ93m0XjZkjIgNUtbi5WHlJKX8ZltHlogtze2WJSO36/TpndVtWWd7t9jU/\nzRWRvcOZhDuDpMl7kHnO/nTJFOpi7UDn/NVU5r/FumCs6uZixZMXQzobYxLHQpykphvOj6a3gIeA\nE4EDVPV+d5v+OD/IngOOAlYDq3ASn0+AVCA3JG6Bqha5y28BdwODgQygEsgBegGFqvoCkCoiwf/+\nxwCTgMnAsSExu6nq4zgJGG75oUPEG+O1Z4BpOAnJF25y8hPgB+ap6gTgMOpm/P4GZ7bwQ4GXcWZJ\nF+BrnAFy8kNi74iTmHcHilX178CakMfrfx5xY3VV1UeBvJBtR+J8XqaGrAv5xWhMgkhKvT9pp8O3\nTmhckpxG8o6H5ZOSPrG5MCIiGZ07T/jPI/dtldAE9cjP41/33dlj+549/hpGLB+S9OekHQ/bOqEJ\nPt4pk6QdDulJSvqfw4iVnJSccsPxV9+9VUIT1LVnb0ZccG2vzNz8G8OIlZbSKfX/Lrvtnq0SmqDe\nu+7O2Vdc17NLbt4fwoiV3dnnO69+QhO0a2o6p3bp3ivTl3RBGLHyU/GddAA5WyU0QQWksheZ26cg\nY5qLFW+W1BjTvuQC5cDuOD+YjsW5KvM68ImIXA3sj/OjLNimeDqQDaThXLXJASqAPULiLhKRXdzl\nAM4PNAH2AUqBJJz/JwUiciFO09Yad7sZwB+Ay92ycNcXisipOIkUOGexv/bkVTBmW8rW793ngZtx\nriwCHCoi1+NcRYG6z4cC84CjgTPc+91x3vehn5EvgD3cs7Wd3fd2j5AYsPXnMbi+SEQuw7laE1z3\nBnAOzmhGuCcItmk2Z0w8iUiBdMoskJT0xrfp3BV8yXuLSHNJ+S777Ll7Tm5Ol0Y3OHjgPqR26jQ4\njKoNkMwemQ0lWrX1yuwF4jsyjFiH7HrQ8IyklE6NbrDbQSN8vqSU0WHEGjH8+JOzk5K2TbSCDjtu\ndEqn1LRzmwskcMzIrNy8hhKaoKOzctM6i++i5mIlI6f0Jyu/qW32IisjFd//NRcr3qz5mTHtiKq+\nEVwWkfOA51Q1eNm/ftOVq919vqduvilwEpv6ngNG1+sXEHqW63m3zMWq+o+Q+lztLo5rYL9g/5mX\n3L8bVXV9A2UbEzVVDb7vXgMQkZOBJ1S1DHjPvQW9GrJf8D18k/v3affvB/Xi/+g2y0RVb3BXvxSM\n0cTncXy9OMHyahN8t8/PE+E9U2NaTQbJ215VqU+SOqk6V/+barqU2b1bbrMn2tNSU0VEpJlO8JmS\n3LnJ37ciApLUeEYQEiujS17jGQ3gS0oiKSUlrFhdcrulNrVBSqdO+JrKelxJSHYXX3KT9ersS0Kc\n1725WF3TaLrMVHyo0zIjodmVGmPaKVV9OuQHVLSxqsMZJjLkB1lLynimpfsaEylVfUVVJ3scs9H3\ncDSfR1V9L5y2/8a0spW6pSjQVH6hgRq0urwaKGsm1k8fffZlTVOxNpeUsrm0tDSMUb2+D5QWVja1\ngdZUQKBmYzNxAL796YsPS5vaYPO6QvzVVevCiPXNwnmzm+wr8+tPywj4/SuaC1SDfvnplpIN0ihj\nLQAAIABJREFUTW2zrLIche+ai1VF4JNf2dLka7GaSnCaric0S2qMMcYYY0xEVLWCgP9VLf6x0WGb\nA+sXVxDwP9ZcIqKqm0pKy9793ztzGx1q+b7HnyorLSu/P4x6raWmYmGgpLDRMgOrvyjBX3l3GLGW\nb1qz8rvCJY23jJ77zEObyorX3x5GrG9/XrJ4+fKlDQ+Upqo8NXH8xg1rCm8JI9Yn31SUrVlV3XDu\nFlDln0WFG9b5q5uPBbOXs6WolIYPYwDlIzauL8PfbB+keLOkxphGiMiJInKriJzp3v+9iNwgIofE\nu27GGGNM3NWUX+tfueBr/7rFlRqo+1Gs/mr8q78sD6xd9DH+invDCbVy9dpLL73pzsVPvDCtprKy\nqnZ98abN3DzhodInX3x17sZNmx8Pq17V5WP9K+YuDRT94A+dkkZrKvGv/LgksGn5LAI1/wknVMn6\n1WNeHn/lT9/NfzMQ8Nc9x7LiDcx86LZNP3/x4TR/dfX/wom1YU3habdfPGbFJ3PeUr+/btqdonVr\nmHjd5cVLv/niGb/fPy+cWMX+6pNvXP3DyoXlJQRCcsZ1NVXctfbnjb9UVzyqql80F0dVtRz/ya+z\npnAVFSh1sTZTzUzWFpXhnxDOqHPxZvPUGNMEEckELlXVe0XkWvfvde4oScYYY0yHJiKpJKddj/h+\nJykZaaiiNeWlBPyP4K+cFMmkmSKS3r1b3rhOKcnnbFfQo1N1dbWs3VBUXFa+ZdLGTZsfi2TSTBHJ\nJrnzLYjvDElJT1H1CzWVGwhUT8Rf9WQkk1OKSNfM3O53+JKSRmfmdU+uqayQLSWb1laVl/61srz0\n+Qhj5ef1KLjTl5R8Ql73Al9FebmUbi5eWbZ50/jystJXwo3jxurVLSnlriRhZF5SipQFAlIe8K/Y\nHKi5vSIQmBFhrB0zSBrvQ0akkyRVBKSawI9bCNxSo4F3IokVLwmR1DTf58t0QOF0uottBUR8OB3o\nJ6hqqYj8SVXvCyY39bbVq6++msGDBzN4cDiDs5hE1qtXL+x4GmNiqWfPnnH/nouEiNgPNbMNVU2Y\n97ElNSZRxf1DIiLjcYYqLsMZeegonKFc56rqh/W2jft7uLCwkJ49e8Y9RnsgIsT7eEL0x8OOp8Ne\nB5Og4v49FwkR0arCZc1vGIan53zPpfe81PyGYaj+4kn6XNjsODZh+fkfp9PrzMc8iTX+wgN5ccmW\nqOMk++DhY3age7duUceqqqpi4b2TWfnYs1HHSunRjTO/nZdQSY0N6WxMI1R1XL1VTza4oTHGGGOM\niauESWqWLFkCQEVaFzYFIp84uW/FalLTKiGp3OuqJYwSfz5bqrZel52dTUFBgedljRw5kk8++YRA\nYOvmqz6fjwMPPJA33nijkT2NMcYY01aJSHegD7AFWBRJPxZj4ilhkprXX3cmWN73hLO59pMmh95u\n0Iw+awjkbqFiyZ3Nb9xGlew6iRmz3ttq3ahRo2KS1FRUVLDPPvswZ86crdYPHz6ciooKz8szW/vy\nyy8ZMGBAq8d/8cUXufLKK2NWblPee+89dtppJ3r37t3g41OmTGHs2LGtXCvvxPKYJuLxrK+xOrb1\n49qRtPfPaEcn4tublPTHJb3bztK5W4r6K/1atrZCUtKfpGbLHfGunzHNiWlSIyLXq+o9InIQMNxd\nvl5V74llucYkorvuuos+ffrw5JNPcs4551BSUsIBBxzAm2++SUZGBvn5+eTl5aGqfPjhh5SXl/Pu\nu+9y6aWX0rVr12bj//e//+Wdd95hwIAB9OzZk+zsbNLT08nLy+PNN9/kwAMPZMGCBRx99NFMnz6d\nnJwc7rvvPs4991xmzpzJMccc48yyHGNPPfUUv/76K8OHD+fHH39k8+bNVFRU0L9/f0pLS1m4cCGL\nFi2iqKiIsrIyhg4dyi+//MLKlStjXrdIxfKYTp06lWeeeYaDDz44YY/n1KlTeemllxL+PWci054+\noyY84ks+hNTsV5N3PqK7pGbXrlcNEFjz9TWB9YsHURN9/xBjYinW89QkicgfgDOgduDr+Pe+NZ44\n9dRTOfvss+nevTujR4/m5Zdf5sgjj+SII47g+eefZ6+99uLGG29kzJgxjB49mmuuuaZ2n59++ine\n1Y8LVWXkyJH88MMP5OTkMHjwYNLS0ujXrx8+n48TTjiBjRs30qdPH3r37k23bt34+uvGJ/0KVVlZ\nSf/+/dl+++1JSkpCVRERRITevXvTvXt3Bg4cyPvvv8/OO++MqjJo0CCKioro379/bRPQWKuurmbg\nwIFMnTqVYcOG4fP58Pmcf0UZGRk888wzDBs2DBFhwIABdO7cufbxRBSrY1pRUcEee+yR0Mcz+EM3\nketoItfePqOmaSIiJKdOSd71mK0SGucxH0kFAzJ8XXoPiVP1jAlbTEc/q3el5jBVndDQlRoR0Xvv\ndUbIdZqflUZc1ow+v5LazpufFTfS/Kxv376elzV8+HCABpufBddPmDCBd955h1WrVjFz5kyuvPJK\niouLmTp1Kueeey5+v5833niD3XbbjcMPP5zKykr69evH7NmzmTx5MrvssktTVWhTp2/DGf3s6aef\n5rzzzotZHWz0M++EO/pZoh9TO54Oex1Mgor595yI7AaMUtWJTWxziHTddXryjkNzG9tGq8qo+fZF\nbPSz8NnoZ60v1qdWQn8VlIrIH4HIMxaTkNasWYOI8Ouvv3LuuecyZswYkpOTOf300xk7dmztmbtD\nDjmEDRs2MGjQINasWQPA4sWL41n1uIjlj18TH3ZMjTGJSkSm4kxF8HMjjx/vLu4sGfk5TcbqlOFt\n5YyJgZj2qQnOuq6qHwMfx7Is4620tDQ++eQTcnK2/j8XHP0MYOJE58TPsccey4wZzsS1p5xySu22\np59+OuCczTbGGGNMq1rg/v098HJwpYj0Bg4BRopIOrCWys2lQPa2IRzqr2rsIWMSRsKMfmYSSyRD\nNgcTGmOM4/XXX+eHH37giiuuIDk5Nv9mW6OMaLWFOprw2fFsc74CMoDZ9davBz4C9nL/rgoU/7zJ\n13NgtvgaPq6B9d+VA+kxrKsxUbOefaZBI0eOJDc3l5ycnK1uubm5jBw5Mt7VMyahlZWVkZKSss08\nT22tjGi1hTqa8NnxbHPScLoBbNXsX1XLVXU5MF1VV6hqDeof5//x7SIN1GwTJLD5V39g3bcdc3Qf\n06YkzKmWUaNGAVCR5mPS0LyI9/dVVONLqyT9gPu8rlrC8Pu71L5OQdnZjV4tjko489Qcc8wxzJw5\nc6vHH3zwQTIzMzn//PO3Wn/XXXdx8803N7qfMe3JWWed1S7KiFZbqGN7oKrw2XznzsAhMRsm245n\n4hORrsDeqjoXCP6YWtrQtqq6oHa5esszkpxaVbP4lXt92TtkSUb3HPVX1QSKlm6gZstn1Gw5E9gU\n+2dgTMslTFIT/QhejQ7a0W402YsvDvx+P2eeeSYvvfQSixcv5oorruCYY47h0Ucf5bHHHuOtt95i\nr7324vbbb2f+/PnceuutfPPNN3z33Xdcf/31/Prrr5x66qmMHj063k/FmJjriJNv2mSNreSz+QSm\nTQHc5hf7HxqTYux4tgmX4jQvm6uq/4pkR62pfEFEpgY2fHcYRUv3QP0lwFuqugawOaVMwkuYpMa0\nPT6fj//85z+MGDGC6667jszMTHw+H6NHj6a8vJy3336bffbZh4suuohp06Yxf/583nnnHQYPHsyI\nESN4+OGHKSkpiffTMKbFOvrkm0E2WWP7YsezTXsXWNHSnVU14MZ417MaGdNKEiapWbJkCRVpXdgU\nSAEgRaBLCmyqhmp3YOi+FavJKtmw1X5V+ZmQVB5xeSopkJwNNZsRrW52+xJ/PlsaGPwjOzubgoKC\niMtvD4JDNqempvLdd9/x9NNP88EHHzBt2jQArr/+ep544gnAOcMzZMgQLrnkEgoLC1m+fDm5ubks\nXLjQhsU1bVro5Jt9+/Zl8ODBzJ07l379+rF+/XpOOOEEpkyZss3km8OGDWsybujkm8FyEnViy9DJ\nGq+66ipmzJix1WSNf//73/njH//Ia6+9ZpM1xsLAIXUdZAdGP0eiHc+2SUQGAKNwhnB+PL61Mab1\nxXTyzbAr4U6+GZx4c0BeCrcNzOKOz0r4ckNdwjGjz690n3j1VvtuGj8+4gk3fTl702mvm6j69m4C\nxeHN1t7QxJcQu8kv4y2cyTeDZs+ezT333MOsWbO8rEKbus4dzuSbsWaTb4bHP+4CAJLG/7PRbWzy\nzbYhEAjA1H84d067sNEf1u39dTBtluffcyIyCuiqqlNiEFtt8s3w2eSbrS+mV2pE5HpVvUdEDgKG\nu8vXq+o9je3TWELjlZYkNB1ROPPUBI0YMYIRI0a0ZvWMSQh2lTHOpv4Dff05wP11eMbFca1OvLTW\nQAGmTdBYJDRBhWuLPInTM78Lk24Y40msS898kltO3NGTWOf/A+48e19PYvXPT2OfnlnRB1Jl3aYy\n1pZWRh8rEKDb4QeQv/9u0cdKSoIT50Ufx0Oxbn6WJCJ/APoAq911jZ7+7OQTS2gSRCTz1BhjjImj\nVhoowLQJeSJynqrGZNbrgmJvrtS8u6kHNz+9oPkNw/SXmd715/rLG6s8ifPyH7ojWR4MYhUIULxx\nA79UpkUdqhN++q36iaT5b0YdS7O6RB3Da7FOavyqOsm9UnOYu67BU0gFBQV0TfNZQmOMMSY8p11Y\n94Vy2oXxrIkxiSIfWBTvShgTD7FOakKvypSKyB+pNwlU0IgRI1hWFrCExhhjTFh8Pl+HbXK2FY8H\nCogXVaVyzUcApPY42JrRtUwGsA8wI94VMaa1xTSpUdUJ7t+PgY+b2nb27NkUDPqN53WIdUJjI74Y\nY4yJJxFpF03OKtd8ROnSug7MaQWHxLE2bdabqvpBS3cWkWSfT47IzkjvX1FVXVJRWfU/Vf3Vywoa\nEysJM6Tz6tWr8Xpg5FgnNAUFBaSmpnoe1xhjjDGmBQ4QkV1b0qcmPzfnvN4F+XeffMShWQfvs2d2\ncUlp4KnX3lq3fY9u36xcu+G0WFTWGC+FldS0ZBSzeGuNhGbEiBGUl0c+R45pG0TkMOASVT3LvT8R\nZ1Kz51V1bVwrZ4wx7Uxqj4MbXDYRKQEinmeiR17X8/ffc7d7X5p4c9e01E7B1b7zR4/sMfuTL/J/\ne/N98wvXezPymTGxEu6VmohGMWuJUaNGUZHmY9LQvEa38VVUUzlu0tbrMjJJP+C+rdbVTqxZvZa0\nXcd6Uj+/vwujRo1yyvT5SE1Npby8nMzMTE/im8Sjqu+5iXzQepz2yoE4VckYY2IiEfqziIg1OYve\nOUBEw5eKSMoOBfl3vHz/LV1TO6Vs8/iIA/f1XXnO6D43/O1fXtXRmJgIN6kJexSzllq1ahXDhzd3\ncmHbofHSvaxEE3Ka38S0T7Xvc1X9i4hkA+cBD9ff8JprrmHw4MEMHjy4Neu3lcLCwoSIkchyqqoA\nKG7meSbC8YToj0d7P57hatevgyqdFn0KQFW/A6AFCYlu/AwKX3XuFBUhXQd6WUPTgBhNCPskkf8+\nG3H6UcOyG0pogi48+ZjOltSYRBduUhP2KGYtFZyp3phEISL7AINF5EbgaeB4YBfguYa2nzhxYivW\nbltezJreEWZe93dymlY09zzjfTwh+uPREY5nONr766AL3yfw3n8B8OXmIi0YNKBCcind4Hw2MnNz\nSStov69XO9cHIDhXjYjsBoxS1W3+oYnI8ao6HcjbafuCjKaCZme21ilkY1ourKQmklHMWmrJkiV1\nd1JqqKgpi2h/Te3F5kDn2vt9K1aTmlYJSU6flxJ/Pluqmo6RnZ1NQYHXwxWYtkpVvwJODFn1eLzq\nYowxjVFVKNlUu9ySZhTWn6V9UNW7gssiMhWYA/wcuo2I9AYOAUaKSDrww/zPv9148anHNdr+f0Wh\ndSM1iS9hRj97/fXXa5eHHjWI5959IKL9jz3yQW5cuKH2/ow+awjkbqFiyZ0AlOw6iRmz3msyxqhR\noyypMcaYNkJV4bP5zp2BQzrwvCbSyHIEEaw/S3v0AVADHAu8HLJ+PfARsJf795e5C79ev2rdhrxe\n+Q3nNeP/8fwmIPGmkDcmhOeTrIjIn0UkWUQCIjJWRA4SkcNE5BERuUZEbva6TGOMMR1P4NN5BB7/\ni3P7dF68qxM3IkBWF8jq0pLuNKb9egsny93qLLGqlqvqcmC6qq5QVV1fvOmcIy68fs2iH5ZvFaCi\nsopbH3m6dOb7C95tvWob0zKxuFLzDPACcAOwPbDUXe93y9scgzKNMcZ0NO/PgvVr6pYHDYtvfeJl\n4JC6M5QDh8SzJiaxnAcsAgYC39R/UFUXBJcrKqs+7ZSSPPyEK259NDc7a499+u6cVFxSqp9/t2xL\ndU3NI2s2FN+H8zvOmITleVKjqktFJPhBeQH4A85lz19UdYKIXO91mcYYYzqgnr0hpVPdsonK+nmX\nAdBt6CNxronxgqpeF8n2VdU13wGHi0jXL5f82BsoB5apqgIduHmnaSti1adGAVR1mYhkuPd3FJFr\ngC0xKtMYY0wHIqdfWNeF5LQL41qXuPpsPoFpUwC3TXkLRj8zJkhVNwIb410PYyIVk6RGVe8JWT7f\nXZwbi7KMMcZ0TD6fD864ON7VaDfe/HA5B/UvILm4mPnz53PcccfFu0rGGBO2hBn9zJiWEJHugE9V\nV8e7LiZx+U/ar8l1Sa9+3prVMcZbHvWpOah/Adc/MI/U5y5n8uTJnlStI7JR+YyJD0tqTFs3GEgB\npoauFJHOqmpNHROMf9wFACSN/2eca2JM+yEi+F95CoAka3oWf+24OWDqYWPiXYUGLZt8csLFGmDn\nBVpdwiQ1o0aNqruTUsMFx46LaH9NTWbS0Kza+76KanxplaQfcB8Afn+XrctoQHZ2dkRlmoRQCXQO\nXSEiBUAfEdlBVac2vJsxpj2wRNk7H3+zmnuuGkre4IlxaX6mqlSu+QhwJgC1KxyJp2axNz0JVvly\n6dlZPYmV0ntvtpSXexKrc3o6RSWRTf7emLLVKyhIj/49HAgEWF+dTH5matSxqmoCfPjwcyx/9Nmo\nY3Xq0Y3zvns/6jheSpikpm/fvh5HzN3qXo7H0U18iUgeUAD8qKrf13u4H9ANKGr1ihljTBt11CE7\nApCTkxOX/jSVaz6idGndj602OxlonIbYFpGx1A3U9HSrFWxMgvB88k1jWoOqblDVRcASEdlDRPYI\neewdoBOQJyIddOIKY4wx8SAiyP6HOrdWvNqkqlOA74BAqxVqTAJJmCs1S5YsqbuTUkNuZQ1JmSmU\n1iSzparpS5TZ2dkUFBTEuIYmQZ0OpKvqk/XWlwA12D93AyS/9kWTj3vTCMIYE43UHgezedHDtcum\nRY6hgYk2wyUiOxUU9Pi/tNTUAQHVsrVr1z1VUVExXVWrPayjMTGRMEnN66+/Xrs84tgh5Psy0bJC\nSqqymDHrvSb3HTVqlCU1Hdc6ID10hYiMAcqAYlVNrAafxhhjGiQi+Dp1qV02LTKT+u3vwyAivp4F\nBf8YfPBBJ1x95RXdBuyzj2zevJn/vDj18OdeeLEoJSXl+BjU1RhPeZ7UiMifgT8DVcDvcC6FpgFn\nAD8AnVX1rsb2LygoICs9B9ashcwKIKuxTY1BVWcHl0VktKpOU9V/x7NOJvHUnLhvvKtgjDGt4Vxg\nQaQ79erVc+Lvzjv39JtvuC6zLqHcjjtvvzXrwvN/n3Xksce/sXzFL97W1BiPxaJPzTPAC8ANwPZA\nkrvej5NEbW5sx4KCAkaMGEFJeTFaXRmDqpl2bhcR2TvelTDGGGPi5D4inOxcRLIzMzLO2DqhqdN7\nh+2587ZbrTmMSXieJzWqupS6swQvAGe6y7+o6j3UG343KJjQzJ49m+qaKq+rZTqGb4GuIvK7eFfE\nGNNxrJ93GevnXRbvapgOTkQG4LSQiXSAnN+cc9aZXZtq8jf6xBOSGn3QmAQRq9HPgkMKLgMy3Ps7\nisg1QIMTIgYTmtWrbWJ40zKqOkNV5zYwaIAxxhjTavzjLqidQ6m1qOqXwHTC7FMjIsF+MmnZ2dlN\ndkdISUmJsnbGxF5MBgpwr8gEl893F5u8HGoJjTHGGGNMVI4GPmtqAxHpDRwCjBSRdODT1/47veiS\nC8/v3tg+n33e9AiSxiSChJmnxhIaEy0R2VlEjhKRPvGuizHGdDSqii5837mpDZQeJwuAwma2WQ98\nBKwAPlLVpUuWLl26YOHCBjf2+/1cd9PNG7ytpjHeS5ghnY3xwFicf+Y7A49FG0xEDgMuUdWz3Pu/\nB7oD76nqh9HGN8aYeFs17ZAm1/UaHcG/us/mE5g2BXDPmO5/aJS1My1wIE6y0uilFVUtB5aLyHRV\nXQGwclXhKWecfd68W2+6cafTTz05OT3dmSnhm0Xf8qcbbtqw7Icf7wEmtEL9jWmxhElqRo0aVXcn\npYZAShXJnZUsMrZ+rAHZ2dkxrp1pI3xABeDJKC2q+p6IHBSyKk9V/yoi1wGW1BhjTIhAIABrV9Uu\nW8/yuHiTMOcTVtUFIctrRGT/m++488rx90z4bUZ6RnpVdZVWVVd/XVi4+vaampqPRMSSGhMzIpKn\nqlFdEUyYpKZv374Nrs9p5XqYNm0mznxIL3oYM3Q4GK331xhjTNCyxbBxfd3yoEgH4TIeyKOF31Gq\nWgLc5d6MaW0FQPtIaoyJhohkA/sAtwIXA/M9iLkPMFhEbgSeBjaIyPU0MujFNddcw+DBgxk8eHC0\nRbdYYWFzTalbJ0ao7pceu806/0n71S6vfXSGp+U1WIdmHm/oOSfC8YToj4fXxzMRtOQ9FcvXQSur\nYl5Gc3KqnDoUe1yHSJ5T1k/fk+avAaDip+8paUFdEuG19EI4x6Nnz56xKDqdrU/GGZMQRCQT2E1V\nP3fvjwDyVfUFEfEBNSJyDJCK0xqmr6rOi6SMhElqlvz0LRU1ZQ0+pqm92BxocHqbRvWtWE1WSdMJ\nX1V+JiSVRxTXhKfEn88Wd7qh7OxsCgpiPm/XVcBuOL9fl3gRUFW/Ak4MWdXkUNETJ070otgWKyws\njPpL0osY9fmbeTxGX+xR1yHexxOiPx6xOJ6JINLjGevXYf2yTgB0i+Nr7e/k1CHS57mqmccjiec/\ncjT8/D0AnY8cTWYLXo9EeC290NLj4YFsmv+IGBMPB6jqnJD76e4NYHdVXSwiA4EanMR8kYj0V9Vv\nwi0gYZKajSXreO7dBxp87NgjH+TGhZFdkZrRZw2pE69ucpuK8eOpWHJnRHFNeEp2ncSMWe8BTn+p\nWCc1qnqHm+n3IoFG9TOJIenVz2uXg3NHJI3/Z7yqY4wnQq9ONbYu9L3fkO1O/qjJxyMZxEz2HwI/\nun1g9x8S/o7GSz2BX2MVfFPAm/lqqlTZ5OE86yUlJZ7FKiv1JlaV38+myuh/jqgGqKipYfOW6C/A\nVfsDdO6zPbm/ib4FQkqXLvDd+y3eX1Wnh9511z0fuo2I7BtJzIRJaozxwAM43bCSgHPiXBfjSn6t\n6fkNrIOSMe2DfP4Bga8/BcC3az8b/Sw+BkDsxmjI67ftaHmJoHuPHp7F2qFnzFuWdDgikgesE5Fj\ngb2AH1R1WvBxVf2ukV0juuoYs6RGRG7CGYkK4GJV3V1Ebge+VVUvO3IbE7QUqAS8O2VjjDHt2MpX\nDo53FYy3XgLSgndEpJuqrvcqeNWHL3sS59+rcrjmXx97EmvDjJvIHHqNJ7FK501kv3EzPYn10Jm7\nsIZMDyIFyK4pY2lFp6gjpRFgvwX/peq/r0YdS7p25cAXpzW/oaO/O6JsHjBLVb8Oc78lIrIHsB2w\nSFWbnNQylldqalT1fje5WSAihxPyQTMmBuYCGThnqkyCqDkxoqvHxpg4efvTjQzaMwuABYtLOOKA\nrpEFGDikru3vwPCbn3k6V04Hp6pT6626TUQWqOrTcamQMY5vRKQ/sIuqPgkgImNU9d/N7LeHqr7r\nNkPbUUQ2qeqWxjaOZVKTLCK3ACnA18AFwHPQcNpaUVHR0GpjIjEMeBZYFO+KGGNiJ9GaNL754XIO\n6l9AcnEx8+fP57jjjmvlGnijvNLPtY/8AMDxg/PiXBvjkclAw6MwGdNKVHWDiOwbTGhcZSJyElAM\n7KCqzzSwa7AjURHQW1WbvLwX6ys1d4vI1UAnYBzOd81BDW2clmYXcUzU8oArcd5nf45zXYwxHUR5\neTVXTZhDav6FnHrqqRHvr6rMX+WMxDmkVzoi4XUIbi65g8gSvP12y2T80ysAuGXsjhHs6Za1cD76\nzCQAREEO6Fh9apobuKG5QRti5FTgc+CX5jYUkZT09PRT83Jzr05OTu5WU1MTqPHXzCosXH2fqv4Y\n+6qadi4gIqJaO/zIBpzWNZlAdf2N3aZqawFU9c1wCohZUqOqE9y/99d7aHmsyjQdm6reEe86GGNi\nL9GaNO6/V3f+/PhH+H5ewH333Rfx/vNXlTNl0cba+4dul+Fl9cL22ZJSBuyaWbu8Q/cITzb+sAjW\nr65b7mBJTYJahDPMc5NEJKugoMec3557zu4XXXB+Rq+ePfH7/bz9zuxLx912+yn5+d3+2Ap1Ne3b\np8B+wGcAqvo+gIgMATY3sP1ebXaeGmOMMaatCO3z8dHH69lnpxSghBmTjuDEQ7u1Sj8QL5K70Hp2\nWXocD95cQN7gicyfP59ekTaj22Uv6FZQt2wSQTawqbmNtuvV6z8P/+3+fY4ZeXTt78KkpCSOPupI\nDhs2tPuw3xz54Pr1UU32bjo4VS3BTWjqrW9wsvRIExpIoKSma1Y+Fxw7rsHHNDWZSUOzIornq6im\nctykprfJyCT9gMjPqpnm+f1dGDXKma8gO7vZk0SecC9V7gn4VHVuqxRqjOnwMtKSuPeynQGng32k\nhvRKb3C5rdF9D4FXn65bNomgGmeqg0aJyPaDDth//9CEJlRaWhr33fPX/KOOPT4mFTRaHYQ9AAAg\nAElEQVQdk4ichvPefBEYA3yPM4JtL+BXVY34zFDCJDV9d/L6rE5us1u03a+OxNfkf9DYuR+Yhk19\nYkybccGbzjyB/zxq+zjXpOVCRwmLeMQwQETi1uQs1EH9C7j+gXmkPnc5kydPjjzApFth0Wd1y1ff\n7W0FTUREpC/QGWeG9qYMOfO005r80TR0SPSTNRpTTxowG+dk9ADgJ6CXqj4rIucCbTepMSYaIpKN\n015zDhCIb22MMbEU2uHaP+4CZ934f7ZqHbY7+aMmH9cITq20dKAAr737yS9UVfuRqipmzZrFGWec\nEVmAonWggbrlMHn5WsZToo3Kp6pLRCRDVRscoUBEjndndQ/U+GuarJ62lYNgEkaWJGvp1nNnrlHV\n0JlNK4Df4Axk8R2wG/C5ewXnh5aUmTBJzVcr1rApkFJ7P0Vgl+oNpJdsQGq2GRQhYlX5mZBUXntf\nJQWSs6FmM6LRx68v0eOX+PPZUrXt+uzsbAoK2uRsulcBXXBGP8tx7xtjTMJLlIECVCCqfGqvgfDt\n53XLJhEcJSJ7h85TIyK9gUOAkSKSDrz77L//U3Tl/13e6Jf/rLfeVuqG1zWmWaX4uVz61N6frD/3\nCH283pxKC7woM2GSmrVbAlz7idMJbUBeCrcNzCJQuJa0v/6fJ/Erxo+nYsmdAPhy9qbTXjdR9e3d\nBIrDndQ0fG0hfsmuk5gx671t1o8aNaqtJjV/A84CngcuiXNdjDHt3MpXDq5d3lRWw7i//wTA+It2\noktG63y1en3FasSgHZj1/s906tSJo48+OvIAPoGUTnXLYQp9LduyRBuVT0TG4vzOq9+2cz3wEbAX\n8JGqrttuu17vTHnm2VPHnntOav04GzcWc92N49YCPeo/ZkxTUpNC/g801wjSA77mN2ldwYTmjs9K\nCFQ3cCkhSm0h4YhrfF/CvSXC9UdgCM6VGhs23BjTahYsLmH8RTsx/qKdWjxQwNh+XRnbr2tcBwr4\n+JvV3HPVUCZPnsz8+Q0OSNQk2bV/g8smPlR1CvAzTl+F0PXlqrocmK6qKwBWrSo8//Y7x8/9/UWX\nFH351dcEAgE2b97Mv56aUnXIYcML16xZc1brPwPT1qX6pPbWGmJ2OklEbsJpLwdwsaruLiK3A9+q\n6osN7ROa0Hy5oRoiG/CsWW0+4Yhx/IKCAlJTtzlJ0yao6h0iMg54G/gk3vUxdRKh/4MxsRTtQAGJ\n4qhDnAk3c3JyOC7S4ZwB2X8I2mO72mWTEJbQSLMxVV0QslwpIke/MPWlwfPen3+dqvYVoaKsrPy5\noo0b/6WqG+PV18u0Xa2VzATF8hp5jare7yY3C0TkcJyRDhrUySdbJzReS05r0wlHayQ0I0aMoLy8\nvPmNE9ffgOuBp3BG0zDGmIQ3f2UZkz535wDRPA7dPjO+FWohEYGsLnXLYQqdK2f9vMsA6Db0EW8r\n10GFJi5hbKvAfODE2NXIdCStndTEsq1RsojcgpPIfA1cADQ6d0jXVF/MEpqa3Qeg6fltNuForYRm\n9uzZBAJteuCw94FinKEBjTExpKrowvedm42MFJVFRZWsLqthdVkNi4oq410d046IyOUiclm862E6\nptQkqb21hlgmNTWqeifOj0wBxgGLGtt4Y2UgZgnNlstvQ8rXtcmEozUTmtWrV3sev5UdCnQFFsa7\nIsa0e5/NJzBtCoFpU+CzyPtfmDq7d0l2kkRVdu+SMOP3tEmqyvsry3h/ZZkl247ONNFKxphY6uzz\n1d5aQ8z+e6rqBPfv/fUearATd1XA+38+wYSm8+Q7KDvvdM/jW0KTcG4GPgYmxLsixrR3qgolm2qX\nrbV9y725vIwyd5qQN5eXMXzHLnGuUdsVz+GxE7T/4AxgR6+CFabkexJn915pPHzZoZ7EOmsGPP6n\n4z2Jdfa8iVz/G29GgE0rLWa7QHH0gVRJzUony4N+5j4NILvtTqdzfhd9sKQkeHFak5u01hWaoHZ7\nSig0oUn+/kvA26SmrSc0SUkp7S2hAXgAGIpzViryIYiMMRGQRpY7Bi/7gSwrrqqdmHFZsfejfpoO\n7WCcAXQ8kb3Um6uyOvBE9tmllyexAE4c7t1w2tk9vBm5Wha8TeXTz0Yfp1MKPSc/SF5uTtSxqmtq\nWPrmm5S89FrUsXzd8prdJuEGCmjJKGYt0b2zj0lDt36BfBXVVI6bFHEsTU7Bn9mFlA1r8J96Pn7A\nl5FJ+gH3eVLX2okvq9eStutYT2K2dnzVHMrLyznwwAO3eiw7O9vz8lrRn4BXgGuB66INJiJDgcE4\ns+A+JSITgRXA8/r/7d15fBzVlejx31Frs2TLNpZ3A2YTMQQTHDAgAnZMCAmPZ5JHPNkGmCQEJplJ\nZmI/SDImTNjCkGDIhwdZCBn2zATiCSHsGCMWExMTzL7IJNjGIFu2JFuyrL3P++OWpLbc6rW6u7r7\nfD8ffdRL1b2nqlqtunXrnqvanG75xuQzEdChQeE5DibPHTy+jA27+oYem9RFpsTOZXrsAJkELATu\niLNcQkr7/RnzJQKlIf9uSSot9e8afVh8iquvD21vT7sYLS+jpEQoLQ2lXVY4rGhPD+Fd6cdFWfzv\nqsA1akgyi1mq5h4QrWW8n2/l21fb3vI36WhM7aq6VkRO9am8E1X1GhG5yHu+A6gG8jqbgjG+mHfS\n8KDMeZa+Nx1nHDSOddu6hx6b1IlIVm85CzpV/UmuYzDFqyJGw1VElgATgN8CtcBXgbeAVuBJVe1K\ntr5EGjWDWczKGM5idjfga87JhoYGFi5c6GeRpvi0i8i1wIM+lTfYeBEAVb1aRGqAc4EbRy68bNky\n6uvrqa+v96n65DU1NQWijNFM6HW31uzMYB1+xRCE4wnpH49MHk9mHOJ+p3ELa2+Pu/KbapyJHs9M\n7gft6U2rjtbWHnoGwt7jVprKOpMuw4+/rXS3w4840o1BVXmhZQCAYyeFkkot7adE9sP06dOzFU5S\nRCQEjAe6VTWv53kwuRWqiNm7VAmsBo4ADsDNqbQFqMF9/jLSqOlX1R+JyFKgHJfFTIHjk60slhkz\nZtDY2Lj3i2X9dPcn/+Wediz9ZVTtTm9IRu/ksRCK/l3QMTCZrhi3TdfU1DBtmj8D1YpMGLgS/3pS\n1orIxUBIRGYCZwKH4Br1+1ixYoVP1aamqakp7X+SfpQRy0B5OZDjf+Y/cXdhjImzWK6PJ6R/PDJ5\nPFV1OOvZvJNSPnksr9gCpP6ZSOQzlenP9Y53XAy1Kdbx4Kvv0dnvbj97sDnEZz6SfDm+/G1NT39Q\ne7pxpLsvn32/kz82uUQB++03no/NyE2vTSC+65IkItNmzZh21ewDZn1q1owZJR27d8sBs2Zs3dXe\nccWu9o6VuY7P5J84jZpu4FRgPa4hcziwDpgFzAaSvloWt1GTbBazVN1///37vHbyJ4/j7iev97Oa\nhPzwo9+k4trvplVG91VX0d14RdT3Og69gYcefWrUdRcvXmyNmtRMAr7tPb483cJU9Vnc3DeDfplu\nmcYUDC+lM3hzA3zUn0xGyQpAdqm0TR4TGrqVb/KY9O+bL2bhcJgPdvcNPTaJEZFD9p85veHnK66e\n9YkFJ1PipeDd1rx96sU/vOqWmdOnnZLjEE0eKi0f/ftMVe+NeBo5SeybKdeX6orGBNBL3u8+EZmk\nqi05jcbsoxBOQI3x2xX1UxFxuUcuP3FKjqPJb2+29tDWMzD0+JT9bYySiNSpamOsZWZOn/a7B397\n56wPHXbIXq9PnTKZ2266fsLZ511wzgdbt2U0TlN44vTU+M73Ro2IXI67St4LfAU36KcS+DzwV2CM\nql7pd73G4LoxnwTOABbhsqEZYzLBEgX4JhQKsW3PwNDjlMqwCwaAuzd+MD92sU+9KSLX4y72zQWW\nxVjuqDNP/8TMkQ2aiPf50Q++N/GBR33LDG2KROmY7PadZKK2O3GZDL6Huy9ug/f6gFefD3nkjImq\nBbgfOAp4P8exGFPQRCRnt5wZM5oj96tgYmVo6HGRu0JVW0Uk3oQwR52+aEHMdLOjNXiMiaW0PM8b\nNaq6QUQG7437LW6Mw0rgPVX9sYikN1jFmNGtAa4DHgBeznEsJkPOf8wNLL/lk7NyHIkJ0KzppkCk\nOoHpICkpYcbYsqHHRe4CcRk83gX+O/INETlZVZ/xnnZs276jhxizX3R3+zM/jSkueX/7mcf1AKu+\nIyLV3vMDRWQZKaRoMyYRqroKH2dONsYEm18Z2NKV7om48Y9NvrmXKbjzr2h3LkQO3lp1972/37l8\n6beqRrv98a57/6cXlwHXmIQVRKNGVa+JePw17+HTmajLGGNMkQpIBjYTHDb55l7WAG+p6usj31DV\nlRGPu2ZMm3rTP120/OKfXfuj8SUjerhefeMtrlpxwwe4NLvGJKy0EBo1xuSSiBwZ7Us838S76qxa\n7MNgTapUlTUfuHm0TppRlbMeDmMiqSo929YCUDH1BPtcJklEzgNmAqjqj3C9NLNF5E1VjZnfumlb\n89UPrXqy7JgFp3/jG189d+K8uR8ub9+9m9v+696WNWvXffB+09Yz8XkqD1P4QjFSOmdCYBo1ixcv\n3vfFsn7OP2N51mMJ95fRs/yGtMooqR5L1bHXRn1vYGB89O311NTUpFW3YWKuAzAmyNZ8sIfbX28b\nep63V7YtA1tB6dm2lt0b7hp6XjntxBxGk39U9XYR+TLDE1BPB3YDfw/cEWddBS4TkZ9eevW159SM\nG3fMwMBA+/tNW38DvKCqao1Mk6xQRVlW6wtMo6auri7XIfgq1p28E7IWRfEQkbkRTwsiw17QemJs\ngL7xwzF3boj7+vpzDkuoLMvAZsw+3mN47HIFbhxMa6Irq+ou4MYMxGWKUKh89EaNiCzBnRLfA3wU\n+AjwPDAD2KKqf0q2vsA0aoxJ0zFACPeZ7gVeyW04xgSXDaY2QVQx9YSoj01STsbL/qmq1+U4FlPk\n4vTUVAKrgTm46V/qgYNU9S4ROQfI30ZN47tv0BUaT3t4DHXdWxnb1c7A2PGEdu9C+vsA6J08FkJ7\n6BiYTFdvevWVlJRQUVFBT08P4XD0W01ramqYNm1aehWZrPC63S8HmnB/KMaYUdhgauOngc8cE/O1\n0H3rEypHROyWs/R1AjapjAmEUGXMhHnduEnT1wNH4MaAbfR6cP6aSn2BadS0dWynrbSa7/+lhUcO\na6NvxgzG3HQZpW8PTzfSfdVVdDdeQcehN/DQo0+lXNe0adNYtGgRq1evZuvWraMut3jxYmvU5BfB\ndbvbQSsgft6uZEwhCkpqazMsh3M3tZLBbLNNs473pZzerj72tHT6UhZAU4xzuWTtLx2+lFN+3DzG\nzb0+7XLCQFdPL03NLemXFQ5TcuSR1Fy8NO2yCJXA04/GXiRGT42q3jvipdiFJSAwjZpBR08qo3TC\nhH0aNH5JtEFj8tLDwOdx92caH0RrUFhjwpiAsdTWeyny8X+vA//q/fiu/JH/jr9QAnrO/R6dlZN8\nKQtgxsSxvpXVV+1PXBNL+5k+Jf0e8XA4zPaBcmonpJ9Eqrevn3efX8/mX9wVf+E4yqfWxl0m1pia\nTPB9ul0RuVxESkUkLCLnicjxIrJARH4mIstE5JLR1q0oLeHf542jf+dOa9CYpIjIHOBQ4FrgizkO\nxxhjTJ4aWH4+A8vPz3UYSRORA3B3K6TfPWCMD0LlZUM/2ZCJnpo7gd8C3wNm4Qb/AAx49UXNTFVW\nWk7tmHIue7GDq8elOWAmCmvQFLxvALcDtwBfynEsgZTK1cuXzo2TlfCcYGVoy6SgzRuUq6vRfo2f\nMP71hOox9cg7rw8/9ic8k38OAsZ4j21OGZNz2U7p7HtPjapuANZ5T38LfMF7/J6qXsPwH9xeqitr\n2NbRxcstfX6HZA2a4tAN9OE+ezNzHIsxxmSNrH8OffUF9NUXkPXP5TockyOq+hTwYdyg65SJyKEi\n8mnvLpsKf6IzxaikvGzoJxsyNaZGAVT1HRGp9p4fKCLLGM6fvpfO7nb68e+eyEHWoCkar+PSOm/A\n5Tq3lM4++MgdjbkOITCCMG+Qn2OcinzcgTGFqpvhyTeTUlFS8rGaktJfzKsYO3lORVVVR3ig/89d\nHV21pWX3tQz0Z2SMjilsBTH5ptcjM/j4a97DmNk4+vp7fY/GGjTFQ1Vvz3UMxhiTE/NOGr7tYt5J\nuYzE5F47KWQ/qyoJnTartOI3l9XOrp0YGj4R/fp4nfBIZ+tXftPefFRbuN/POE0RyFYPzaDAZT/z\nizVojCmM7GVxx/VA0YztCcIYp9I/vBTz/eI4EsEiIkWf8cwM+RvwHeBfEl1BRKQ2VPbLH00+uHZs\nSWjke3x67KTKpv7eo/9n9w6fQzWFTsqye/diYBo1E8dNpjJUyg0nj6Oku4+e5Tfss0xJ9Viqjr2W\ngYHxLF68eNSyBifW3LNnD/Pnz085ppqa9NPnGZNLQTgJNiaSNVILTxAauulewImXAAPyJgnGQpKf\n7+PkYyvHTRjZoIn02XG146xRY5Il5dmdCz0wjZq6gyLHte0XdZkq7/eENOppaGhg4cKFaZQQvLoK\ncZtEZKGqNmS8otgxnAzUA9tU9TYR+SowBXhKVf+Uy9iKiY3rGRaEfdF/1kdyHYLvcjW+yC46mAzY\nQoLJckTkTFV9AJh1cFnl+FjLRt6SZkyipKw8q/UFplGTLYXYACjEbcJdbWrIRkUxnKiq14jIRd7z\nSar6HyJyMZAXjZp0T4Ijr27awHLjhyA0zKAwbs0MikJs6OYrVb0t3jLefDYnAp8SkSqgaVNfdwcw\nasNm14CNpzHJi3X7mYgswfVT3AOcDfwV2B9oBZ5U1aiJxWIJTKMm3hwQfrrssssKrq4C3KZ/B36Y\njYpiGMwgM/jh1BG/97Js2TLq6+upr6/PeGCjaWpqSnv50cro7elJqY5syERMQTiekNy2ZeV4/vyh\noYcTrvsuADuXXjP8vg/HIhPH04+/Db/Fa+AF8W8tmilx3g/CdsSLId42JFJGsqZPn+5reUnYAazF\npX5eC2x5vrtj53nhgfFVo9yCdv/uHbshAylqTUGL01NTCawG5gCrcPMsbQHG4RrY+dmoUVWbK8wE\n0VqvVyYkIjOBFhH5LqNkllmxYkVWgxupqakpyj/Jfa9GRxq5fPQynPKKLVHXybT15+z9PFs9Rrk+\nnhDteKR3PKP1Tpyxavdez5PpoRgoL49ab2yxtyH58uLz428jM4IQQ/oG4ryfD/tS4owLAtD78uN4\nxKOqe4BNIvKAqm4GGFsS+vYl29/9z8smz540rqQ0clkaunb2Pt7Z9iZwXI5CNnlKymMmCugGTgXW\n46bkmA08DMz3Hied5SsQjRpjgkhVnwWejXjp1lzFYgzY7YCFplCOZ+QA+oHl57vXrrolqzHY+KTk\nqergROnsDg/cX1US+uK3tr1zU135mAlHlFfX7A4P9D3btaujW8NPtIb7LwD25DBck49i9NSo6r2j\nvJXyPcrWqDHGGJM1I3uCbJC+CYJCHhckIrNUdUu85faEBx4H6kRk7nNd7XW4RszTqrrbKyfDkZpC\nU7QpnY0x/iuUK8GFIN4JgWrmT6ItnbIpVEFJQBEkInIF8A4wF1iW6Hqq+grwSqbiMsWjaLKfZStd\nbpR6vgUcCvxcVd/yq55R6ioDfgd8XVWbM1jP/wbqgMdU9VW/6hmlrouAfqBZVe/2ua4FwD+q6he9\n53mbQjmRK1rZOIk1xkRnJ8GmCPxKVTeLyKRcB2KKUzH11GQrXe5gPRcDqOr/E5GPA9MBXxs17LtN\nX8YNevLbyHoWA68CfVmoqwuXoWJdjHVSoqpPicjxES/lXQplY0YThEZsZAzWc2cKSbq90kEYF5QB\nnxGRMNAD/CrXwZjio1me3yiXjZqk0uX6UI+rTGQqcJKqXulzPZF1DW7TUcBUYBvw+wzW06+qN4jI\nD/E/DfLIuiao6jIRWc7eg+j9EtnFkanPRMYF4QR2JDt5LSx2PI0xcVQDJcAzfhXYfMgCX8oJv72B\nntYXfCkLYPU9j/lWVuejD/pSzpbaiby7o9WXssZMmcSWlrb0CyotpXTuEUy7dGnaRYXKy+CbcU4D\ni6hRk1S6XJ/qmYW7WvG4iMxR1TczWNdM7+T/XGBNJusB/iwiS3Fp8fw2sq5uEVmG/71ciMhcoF5E\nvg/cQeY+EyaP2cl8cPhxJduOpzEZcx9ubpkOvwrceMm1vpQz9dKlvPxvP/alLID7/vFS38pav/QK\nX8qZc8k/89LlN6RdTkl5OccuO59N192cdlnl06fC6Qt54YY70i6revrkuMtoSXabGTlr1GQrXW6U\nej6diXpGqQtVTf+TE7+ejKUajlKXP99o0et6BTgr4iVLoWyMyQhLopEZBXDLVkFsg+djuIuCp5CB\nC5HGxFVEPTXGGGNyxE7ghwVhXwQhhkJh+3LINOBvuMkMjck6DRVJT40xxhhjjMmYg4B/JYlzPRGp\nKkU+V0HJMWG0vYvwPar6euZCNAWtZPSeGhFZAkwA7gG+gBvD/QYwA9iSStZba9QYYxJiVz+NMSav\nrFTVhEe9V0loaQ2lF81h7PjJlI/pRXmL3d8YK6WbOhlYnMlATWGK01NTCawG5uAmegU4SFXvEpFz\nSCHrrTVqjDHGGGMKz0OJLlgtpUtnUvmDj1M7oSQiCelBVE1upmfyo2x/dg8DGQnSFK44KZ27gVNx\nSa6qcD01b3o9OH9NpT5r1BhjjDHGFBhNcG4BEakYR2jZyAbNoClUcAzjZ67Bn/TEpojEyH6mqvdG\nPPVl7sMSPwoxsYnIZd7v/yUi86K8/6nsR2WMMcYYwycOY2xNtAbNoMOpzu7U8KYgaKh86CcbrKcm\nO5pEpBb4KNAmIp8Bfg0sB24EjhaRPcCJuEFSc4H3gVnAH4EzgT8D+wHTgQ5V/XXWt8IYY4wxhWZ8\nNaGqWAuU2TVwk4LW7nD8hXxkn9Ls+D3wJWAAUFxGEgVWe3OzAIwDmnENn35VvQ3oA04DrlPVx3E5\n51u9ZY0xxhhjkiYiJ0c8fWML3THvLdtFX4YjMoWopbt/6CcbrFGTBaq6DVgINOAaM4PduJFN2EOB\nfiDkLYP3+zFgmYichptEaxLQmPGgjTHGGFOopgw+UNWXmunZ2h6j4fJndrZlJSpTUDbv7B76yQa7\n/SxLVPX/eA/XADd5jzd7710zyjo/9h6+Eu19Y4wxxphkqerKyOddDCz5I9ueWETtjGlUIN74mh4G\neJ6d7Vvp+T3w1VzEavLX5l1dWa3PGjXGGGOMMUVsQPUtETlhNTuuDSEn11Aqvajspr+1H/1JD+Hb\nsEaNSVLj1o6s1me3nxmTBXv27GHz5s05LSMIMfghVzFE1nvTTa6ztbm5mZaWlqyVEYQYYpVTjDEA\nbNq0iYcffjjp9YKwHyKluh1BiCFI+9KvfZFtqvpeh/Z/fhf9B71H9/xt9By5W/uP6NaBWxNND21M\npLe3dgz9ZIM1aozJgu3bt/OnPyU9Oa6vZQQhhpUrV/LMM8/kZQyR9YZCIQDefPNN3nnnnayVEYQY\nYpVTjDEANDQ0sHDhwqTXC8J+iJTqdgQhhiDtS7/2Ra6oao+qvqeqybfwjYmwo7lz6Ccb7PYzY6IQ\nkf2A/wvsUdUrvdf+BXch4ClVfTHRstatW8cDDzzA5s2bqa6u5swzz0w6nnTLCEIMAGeffXbS6wQh\nhsF6N23aRGNjI6eeeipPPPEE9913H+Xl5fT09HDKKadktIwgxBCtnA0bNrBo0SJuvfVW5s2bx9FH\nH100MYBrJG/evJmNGzfy9NNPc+GFFya0XhD2gx/bEYQYgrQv/doXxhSCtubdWa0vkI2at99+++bO\nzs46gOrq6sbDDz/8glzH9Py7zbdVhuTkrV3h7WPL5JWTDp4yFJOInKeqt49cR0QWABtVdVMqdY5W\nbiJaNj53s/bsqAOQitrGSbPrI+M9EJitqk+lUnYyvLp2ququTNfls4XAr4CPi8gEVd0JtAAzSPLv\n5rjjjmPq1KmsXbs2pZNwP8oIQgx+yFUMkfXu3LmTE044gZKSEsrKyqisrGT+/PkZLyMIMcQqp76+\nPuV18zUGgLPOOou2tjbmzJnDnDlzEl4vCPvBj+0IQgxB2pd+7YugKhlT6U9BoRJCVT6V1Q5lVWP8\nKWsPvsUlpaW+lBWqqEBKQ77s+9CYCigro9SH/VWWQDxt774VdxkRqQOOB5pV9VEROQI4U1V/LCI/\nAX6lqgll/ZUg3ib54osvNqxevXoBwKJFi56aN2/eQgAR+TbuhPJdoAc40lulGdgO1AFVwPWq6mvK\nhZc3b//r9a91HvxySx+Xzhv31GePnLZw8D0RuQTYCHwFuAtoB/bg5qU5DFgFnAUcANwLHO6tWga8\n5G3TbNz8M+/hTp5PBA5S1W+mEu+Ot+9v6H3j6gUA5XMvf772kNNO8GL9KlAJ/D1wq6r+ynv9n0fU\nDbAJN+nnUV7M3ap6o7f853EpqMfi5t3pxB2Dwe7qA7163vf2xQRgBy6d9XSg29s/s3Af2G2pbKef\nROQs4DO4VNpvAiuBjwMrvUbN4HLLVfWqEesqcBnQoKoNWQvaZESs4ykix6rqC2mWn1YZQYjBj3IK\nIQYRmQJUqOp7aZQRhP2Q9nYEIYYg7Eu/YwkK73vRmL2oqkR7XUS2AlMjXtqmqtMi3v8ycALwSWAO\n8GVVvdN77zxVvd077/+Lqq5JJJZA9tTEsB74I/APgDA8z4uq6kMichgu/bHvs0Rt7Qpvf7ml7+BR\n3lYvnkeAQ3ANFfV+unEn95HLPgIswt3KtAlYgmscbFLVP4jIeQzPVZOWkglHIRVTJ4+oX4E/AJE3\nG7dH1D24Xx/GpZ++ANdgjDY17OA2vg7UquofI8oIA695624CjgH+Agx+qMuAF4HedLfTD6r6B9x+\nGbz9bBmuQRYWkdNxcwgdC2yIsm7UP2qTn2IdTz9OUtItIwgx+FFOIcSgqs3prBrkzKQAAAbqSURB\nVO9HDH6U48d2BCGGIOxLv2MJCvs/Z5IR2YAZ5f27gbu9npovAdtF5CigDZgrIvvjzh3n4KZDiSuQ\nPTWxbj8TkXNV9Y5sx7Tmb803N3eF6wCmjClpjLz9LBEicjywSFWvzkiAI7RsfO5mHeicKxVTJ2u4\n+5naA+b/wyhx1eMat+sS6d2KWP5dXC/Ga4mML/FuxevE9aRtVNXcptAyxhhjjDEFI5CNGmOMMcYY\nY4xJlKV0NsYYY4wxxuS1QI6peaVx3c3tnW11ADXVExvn1h2X8K1eInK0qr4c4/2UMpK1r1tzs7Q1\n1wHoxCmNNcedtFdM2cwologNGzbc1tHRMRv2voUvcv/E2lfefnrXbhMzxhhjjDFBF8hGTXtnW909\nT/9sAcDfnTKc/EtElgCfA27FZVT4G1AP3I3LLtYF7C8iO3FzjKzDDWLfjcuu8D5u4PppItKBG8C/\nA3hOVdtixSRtzXVVP7t8AcCeb16693sRGcVEJJMZxQ4hwexuHR0dsyMyyEW+9UUR+S7QAEwUkW7g\na8CNwIXe64cDtbhxM8akTESW4jIVNuGSaZyEy1Z4M/ADVf1OmuULcAZwNu7zuhV4FpfFrhO4D/i6\nqv4gnXqKUSLHTkQW4hKAvO29dzrQCuyH+/+yArgYuCXyAkkyx83LStgFfAq4BfgY7nt9Fe573rfP\nU7b3Q4L158W+SmI/9eL+j98JHIf7/9YIfAGXEfNK4CZV/boPMYlX/4G484Am3DFaghuMvByXffME\nEVkO/FxVW2OUd7OqDl4g/KSqPpZujMYY/wSyURNDJa5R8iFc+uOtuC9RcF+GYdyXvuAaNNtxjQD1\n3hvM/PUYbh6STcAxqvpgmnFlK6OYkn52t43ALlwGr1rcLYiv4RqJf8Glye7Hp+xrxUxEvob7h90N\nzAWOU9UlKZTzY+AO4DSgPpUy0iUiF+G2YwxuQtIbE1xVcScTPbjPXj1uLF+riLyUblzqBgU+6OW1\n/ylwMFCD+1yHVXWziCQ/RTzxj5/3fiMuVXsV8GvgIVX91IhyRj1+3ongq7g053dFWz8TEjyeiRy7\nSbh91I/LFNiPS4dej7s40wk8sU/ByR23kKo+JiLTgWrgSWAzsNiLMa3PUwL7Ipn9cCruf9DBwP64\nE/tHIvbD+SLyDKMcb+8C2Spcyv2twG241PLnAoP76vPAP+G+s58Cvs/wBaiM7qs4EtpPqrpSRA4F\ntuD+vz0uIufg/heWqGqviDzrS0CqKiJv4D6X63DnPLtw5wiqqq+KyCve4hOAn4rIZty0DCd6yy/B\nTT3wH7iLpnfhsnZ+SEQ+B+z0tvezwP24xtm/BeWuDWOKSSAbNTXVExsHe2hqqicOTbgzmL96hEe9\n378Y8frGyCciMnlE1rQ7vKtGzyQSk06c0jjYQ6MTp+w1CZCq3jqirsEMYfd674+cQPNc7/eKEcs/\nx3BGsb0yvEUpI6bq6urGwR6a6urqyH04cj+BOwkB96Vv/FWCO7noBBpE5DBV3ScldByv46647kmj\njHRtw/UwnkH0hvhoWnEnOw+ralhEXgZ6RWQGLmVjnSY4qVYcpbiTiUrc3/QzwOEiMsmrZ5aqbkmh\n3H2OH+5vtMd7bTbuRG4M7ur8/SLyCVVdFVHGPscvoowdXhn9MdbPhESOZ7xjdyTuqnu1dyJ9IC6L\nYoOIHAD8DredH8adMEbroYh33PbHndzjxfIacA6uEfI4br+n+3mKty8S3g/AG942bcYd2zG4i2jX\n4hoh24lyvHFp7sfhPitLcL0JE3G95quAv8P17t+K6/1oBcpxn88twJki8l9Z2FexJLKfBi9Coqrb\nReQgETkbNw2CAFu89K5z491KnoQW3PGYj2uMvIxrDE72YponIqfijk2Nql4iIjczPC3BUMjeT6eq\nXiciv8Tt+w955b+Bu1D6mjVojMkNy35mTIaIyCLc5KtjcfMS/Sdu3NWKJMs5F/fP+DvAQ6mUYZI3\nyvE7GOjAnZw14Hpc/4I7+TkRWA2coqrfjyhn5PGLLGMAdyX/J7iT1X3WN/kjgc9M1OON6+EvBW7H\nXaC73StnIq5XfSHuolctrnewEfg5cA/uRPtEVb0wC5tojDGBZY0aY8yoBhNgAE0+X9U1OWDH07H9\nYIwxhccaNcaYqCISYHwDd+X4ZdxV4/m4+/sHE1lsU9WHcxWnSYwdT8f2gzHGFCabp8YYM5rBe8gH\nb5VZBVwBPMhwIouwnfjlDTueju0HY4wpQNZTY4wxxhhjjMlr1lNjjDHGGGOMyWvWqDHGGGOMMcbk\nNWvUGGOMMcYYY/KaNWqMMcYYY4wxec0aNcYYY4wxxpi8Zo0aY4wxxhhjTF6zRo0xxhhjjDEmr/1/\nInJ64JrGDnoAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x11372e090>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# create groups by environment\n",
    "groups_sp = spores_norm_df.groupby(['env_evo','env_test'])\n",
    "groups_hy = hybrids_norm_df.groupby(['env_evo','env_test'])\n",
    "groups_cl = clone_norm_df.groupby(['env_evo','env_test'])\n",
    "\n",
    "param='growth_rate'\n",
    "\n",
    "fig = plt.figure(figsize=(12, 5), dpi=300)\n",
    "\n",
    "import matplotlib.gridspec as gridspec\n",
    "grid = gridspec.GridSpec(2, 3, width_ratios=[1,3,1], hspace=0.4, wspace=0.2)\n",
    "\n",
    "gs = {}\n",
    "gs['schematic'] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,0])\n",
    "gs['variance'] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,0])\n",
    "gs[('HU','heatmap')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,2])\n",
    "gs[('HU','barplot')] = gridspec.GridSpecFromSubplotSpec(1, 24, subplot_spec=grid[0,1], wspace=0)\n",
    "gs[('RM','heatmap')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,2])\n",
    "gs[('RM','barplot')] = gridspec.GridSpecFromSubplotSpec(1, 24, subplot_spec=grid[1,1], wspace=0)\n",
    "\n",
    "\n",
    "### analysis of variance ###\n",
    "ax = plt.subplot(gs['schematic'][:])\n",
    "    \n",
    "ax.text(-0.1, 1.2, chr(ord('A')), transform=ax.transAxes,\n",
    "        fontsize=12, fontweight='bold', va='top', ha='right')\n",
    "\n",
    "### analysis of variance ###\n",
    "ax0 = plt.subplot(gs['variance'][:])\n",
    "    \n",
    "ax0.text(-0.1, 1.2, chr(ord('D')), transform=ax0.transAxes,\n",
    "         fontsize=12, fontweight='bold', va='top', ha='right')\n",
    "\n",
    "S = spores_norm_df[spores_norm_df.background.isin(['ancestral','evolved'])]\n",
    "S = S.rename(columns=labels, copy=True)\n",
    "S = S[np.isfinite(S['growth_rate'])]\n",
    "\n",
    "H = hybrids_norm_df[hybrids_norm_df.background.isin([('ancestral','ancestral'),('ancestral','evolved'),('evolved','evolved')])]\n",
    "H = H.rename(columns=labels, copy=True)\n",
    "H = H[np.isfinite(H['growth_rate'])]\n",
    "\n",
    "formula = 'growth_rate ~ background + de_novo_gene + de_novo_genotype + auxotrophy + tetrad + C(spore)'\n",
    "\n",
    "tsdf = S.groupby(['env_evo','env_test']).apply(stats.variance_vectors, formula)\n",
    "thdf = H.groupby(['env_evo','env_test']).apply(stats.variance_vectors, formula)\n",
    "    \n",
    "df = pd.concat([tsdf, thdf], keys=['spores','hybrids'], names=['type'])\n",
    "    \n",
    "# fix columns\n",
    "df = df.rename(columns={'C(spore)':'spore'})\n",
    "\n",
    "df = df.swaplevel('type', 'env_evo')\n",
    "df = df.sortlevel(['env_evo','env_test'], ascending=False, axis=0)\n",
    "    \n",
    "utils.simple_axes(ax0)\n",
    "    \n",
    "df.plot(ax=ax0, kind='barh', stacked=True, \n",
    "        color=[factor_dict['color'][x] for x in df.columns],\n",
    "        align='center', legend=False, width=0.75)\n",
    "\n",
    "bars = ax0.patches\n",
    "hatches = ''.join([factor_dict['hatch'][x] for x in df.index.get_level_values('type')]*len(df)*2)\n",
    "\n",
    "for bar, hatch in zip(bars, hatches):\n",
    "    bar.set_hatch(hatch)\n",
    "\n",
    "ax0.set_xlabel('Variance', fontsize=5, y=0.9)\n",
    "ax0.set_ylabel('')\n",
    "\n",
    "ystart, yend, ylabels = plot.set_custom_labels(df.index, 0)\n",
    "ax0.set_yticks(ylabels.values(), minor=True)\n",
    "ax0.set_yticklabels(ylabels.keys(), minor=True)\n",
    "\n",
    "ylabels = df.index.get_level_values('env_test')\n",
    "ax0.set_yticks(np.arange(len(ylabels)), minor=False)\n",
    "ax0.set_yticklabels(ylabels, minor=False)\n",
    "\n",
    "ax0.tick_params(axis='both', which='major', size=2, labelsize=5)\n",
    "ax0.tick_params(axis='both', which='minor', size=1, labelsize=5, pad=18)\n",
    "\n",
    "# Shrink current axis's height by 10% on the bottom\n",
    "box = ax0.get_position()\n",
    "ax0.set_position([box.x0, box.y0 + box.height * 0.2,\n",
    "                 box.width, box.height * 0.8])\n",
    "\n",
    "# Put a legend below current axis\n",
    "leg0_1 = ax0.legend(bbox_to_anchor=(0.5, -0.1), frameon=False, prop={'size':4}, \n",
    "                    handlelength=0.75, borderaxespad=3, loc='upper center', ncol=3)\n",
    "\n",
    "#Create custom artists\n",
    "s_artist = plt.Line2D((0,1),(0,0), color='k', marker='s', mfc='none', ls='')\n",
    "h_artist = plt.Line2D((0,1),(0,0), color='k', marker='s', mfc='none', ls='')\n",
    "\n",
    "#Create legend from custom artist/label lists\n",
    "leg0_2 = ax0.legend([s_artist, h_artist], ['spores', 'hybrids'], \n",
    "                    frameon=False, prop={'size':5}, handlelength=0.5, loc='upper right')\n",
    "\n",
    "ax0.add_artist(leg0_1)\n",
    "ax0.add_artist(leg0_2)\n",
    "\n",
    "for (ii, (env_evo, env_test)) in enumerate([('HU','HU'),('RM','RM')]):#,('HU','YNB'),('RM','RM'),('RM','YNB')]):\n",
    "    \n",
    "    # group by evolution and test environment\n",
    "    gsp = groups_sp.get_group((env_evo, env_test))\n",
    "    ghy = groups_hy.get_group((env_evo, env_test))\n",
    "    gcl = groups_cl.get_group((env_evo, env_test))\n",
    "\n",
    "    # apply sorting ranks to reorder rows\n",
    "    gsp.loc[:,u'rank_background'] = gsp[u'background'].map(dict_background)\n",
    "    gsp.loc[:,u'rank_population'] = gsp[u'population'].map(dict_population)\n",
    "    gsp.loc[:,u'rank_genotype'] = gsp[u'genotype_short'].map(dict_genotype)\n",
    "    gsp.loc[:,u'rank_gene'] = gsp[u'gene'].map(dict_gene)\n",
    "    gsp.sort_values(by=[u'rank_background',u'rank_population',u'rank_genotype',u'rank_gene'],\n",
    "                    ascending=True,inplace=True)\n",
    "    \n",
    "    ghy.loc[:,u'rank_background_MATa'] = ghy[u'background_MATa'].map(dict_background)\n",
    "    ghy.loc[:,u'rank_background_MATα'] = ghy[u'background_MATα'].map(dict_background)\n",
    "    ghy.loc[:,u'rank_population_MATa'] = ghy[u'population_MATa'].map(dict_population)\n",
    "    ghy.loc[:,u'rank_population_MATα'] = ghy[u'population_MATα'].map(dict_population)\n",
    "    ghy.loc[:,u'rank_genotype_MATa'] = ghy[u'genotype_short_MATa'].map(dict_genotype)\n",
    "    ghy.loc[:,u'rank_genotype_MATα'] = ghy[u'genotype_short_MATα'].map(dict_genotype)\n",
    "    ghy.loc[:,u'rank_gene_MATa'] = ghy[u'gene_MATa'].map(dict_gene)\n",
    "    ghy.loc[:,u'rank_gene_MATα'] = ghy[u'gene_MATα'].map(dict_gene)\n",
    "    ghy.sort_values(by=[u'rank_background_MATa',u'rank_background_MATα',\n",
    "                        u'rank_population_MATa',u'rank_population_MATα',\n",
    "                        u'rank_gene_MATa',u'rank_gene_MATα',\n",
    "                        u'rank_genotype_MATa',u'rank_genotype_MATα'],\n",
    "                    ascending=True, inplace=True)\n",
    "\n",
    "    ### heatmap ###\n",
    "    ax1 = plt.subplot(gs[(env_evo,'heatmap')][0])\n",
    "    \n",
    "    ax1.text(-0.1, 1.2,  chr(3*ii + ord('C')), transform=ax1.transAxes,\n",
    "             fontsize=12, fontweight='bold', va='top', ha='right')\n",
    "    \n",
    "    # spores\n",
    "    S = gsp\n",
    "    S = S.groupby([u'mating',u'background',u'population',u'genotype_short',u'gene'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    S_ref = np.nanmedian(S.ix[:,'ancestral'].values)\n",
    "    \n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    \n",
    "    plot.heatmap_spores(S-S_ref, ax1, title, \n",
    "                        xlabel, ylabel, xticklabels, yticklabels, \n",
    "                        cmap=plt.cm.RdBu_r, radius=0.25)\n",
    "    \n",
    "    # hybrids    \n",
    "    H = ghy\n",
    "    H = H.groupby([u'background_MATa',u'population_MATa',u'gene_MATa',u'genotype_short_MATa',\n",
    "                   u'background_MATα',u'population_MATα',u'gene_MATα',u'genotype_short_MATα'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    H = H.unstack(level=[u'background_MATα',u'population_MATα',u'gene_MATα',u'genotype_short_MATα'])\n",
    "    H_ref = np.nanmedian(H.loc['ancestral','ancestral'].values)\n",
    "    \n",
    "    title = environments[env_test]\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    legend_title = r'$\\langle r_{bg}^{a/\\alpha}\\rangle_k$'\n",
    "    \n",
    "    plot.heatmap_hybrids(H-H_ref, ax1, title,\n",
    "                         xlabel, ylabel, xticklabels, yticklabels,\n",
    "                         fold=False, cmap=plt.cm.RdBu_r, pad=30, \n",
    "                         legend_title=legend_title)\n",
    "    \n",
    "    ax1.xaxis.set_ticks_position('top')\n",
    "    ax1.yaxis.set_ticks_position('left')\n",
    "    \n",
    "    ## genotype\n",
    "    # labels\n",
    "    xlabels = H.columns.get_level_values(u'genotype_short_MATα')\n",
    "    ax1.set_xticks([x+0.5 for x in np.arange(len(xlabels))], minor=True)\n",
    "    ax1.set_xticklabels(xlabels, minor=True, fontsize=5)\n",
    "    \n",
    "    ylabels = H.index.get_level_values(u'genotype_short_MATa')\n",
    "    ax1.set_yticks([y+0.5 for y in np.arange(len(ylabels))], minor=True)\n",
    "    ax1.set_yticklabels(ylabels, minor=True, fontsize=5, rotation=90)\n",
    "    \n",
    "    ## gene\n",
    "    # labels\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(v+0.5, 1.15), xycoords=(\"data\", \"axes fraction\"),\n",
    "                     ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "    ystart, yend, ylabels = plot.set_custom_labels(H.index, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(-0.15, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                     ha='left', va='center', annotation_clip=False, fontsize=4, rotation=90)\n",
    "        \n",
    "    ## population\n",
    "    # labels\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, 1)\n",
    "    xlabels = {k: v for k,v in xlabels.items() if k != u'WAxNA'}\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(v+0.5, 1.15), xycoords=(\"data\", \"axes fraction\"),\n",
    "                     ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "    ystart, yend, ylabels = plot.set_custom_labels(H.index, 1)\n",
    "    ylabels = {k: v for k,v in ylabels.items() if k != u'WAxNA'}\n",
    "    for k, v in ylabels.iteritems():\n",
    "        ax1.annotate(k, xy=(-0.15, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                     ha='left', va='center', annotation_clip=False, fontsize=4, rotation=90)\n",
    "    \n",
    "    ## background\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, 0)\n",
    "    ax1.set_xticks([x+0.5 for x in xlabels.values()], minor=False)\n",
    "    ax1.set_xticklabels(xlabels.keys(), minor=False, fontsize=4)\n",
    "    xgrid=[xst+1. for xst in list(set(xstart.values()))]\n",
    "    [ax1.axvline(g, lw=1, ls=\"-\", color=\"black\") for g in xgrid]\n",
    "    \n",
    "    ystart, yend, ylabels= plot.set_custom_labels(H.columns, 0)\n",
    "    ax1.set_yticks([y+0.5 for y in ylabels.values()], minor=False)\n",
    "    ax1.set_yticklabels(ylabels.keys(), minor=False, fontsize=4, rotation=90)\n",
    "    ygrid=[yst+1. for yst in list(set(ystart.values()))]\n",
    "    [ax1.axhline(g, lw=1, ls=\"-\", color=\"black\") for g in ygrid]\n",
    "    \n",
    "    ax1.tick_params(axis='both', which='minor', size=0, pad=9)\n",
    "    ax1.tick_params(axis='both', which='major', size=0, pad=20)\n",
    "    \n",
    "    # filter by dictionary\n",
    "    gsp = gsp[(gsp['background'].isin(sp_bg_dict['position'][env_evo].keys())) &\n",
    "              (gsp['genotype_short'].isin(sp_gt_short_dict['position'][env_evo].keys())) &\n",
    "              (gsp['population'].isin(sp_cl_dict['position'][env_evo].keys()))]\n",
    "    ghy = ghy[(ghy['background'].isin(hy_bg_dict['position'][env_evo].keys())) &\n",
    "              (ghy['genotype_short'].isin(hy_gt_short_dict['position'][env_evo].keys())) &\n",
    "              (ghy['population'].isin(hy_cl_dict['position'][env_evo].keys()))]\n",
    "    \n",
    "    # apply sorting ranks to reorder rows\n",
    "    gsp.loc[:,'rank_background'] = gsp['background'].map(sp_bg_dict['position'][env_evo])\n",
    "    gsp.loc[:,'rank_population'] = gsp['population'].map(sp_cl_dict['position'][env_evo])\n",
    "    gsp.loc[:,'rank_gene'] = gsp['gene'].map(sp_gn_dict['position'][env_evo])\n",
    "    gsp.loc[:,'rank_genotype'] = gsp['genotype_short'].map(sp_gt_short_dict['position'][env_evo])\n",
    "    gsp.sort_values(['rank_background','rank_population','rank_gene','rank_genotype'],\n",
    "                    ascending=True,inplace=True)\n",
    "\n",
    "    ghy.loc[:,'rank_background'] = ghy['background'].map(hy_bg_dict['position'][env_evo])\n",
    "    ghy.loc[:,'rank_population'] = ghy['population'].map(hy_cl_dict['position'][env_evo])\n",
    "    ghy.loc[:,'rank_gene'] = ghy['gene'].map(hy_gn_dict['position'][env_evo])\n",
    "    ghy.loc[:,'rank_genotype'] = ghy['genotype_short'].map(hy_gt_short_dict['position'][env_evo])\n",
    "    ghy.sort_values(['rank_background','rank_population','rank_gene','rank_genotype'],\n",
    "                    ascending=True,inplace=True)\n",
    "    \n",
    "    ### spores barplot ###\n",
    "    ax2 = plt.subplot(gs[(env_evo,'barplot')][:7])\n",
    "\n",
    "    ax2.text(-0.1, 1.15,  chr(3*ii + ord('B')), transform=ax2.transAxes,\n",
    "             fontsize=12, fontweight='bold', va='center', ha='right')\n",
    "    \n",
    "    ax2.text(0, 1.15,  environments[env_test], transform=ax2.transAxes,\n",
    "             fontsize=6, va='center', ha='left')\n",
    "\n",
    "    # set reference as mean phenotype of the ancestral hybrid\n",
    "    ref = np.nanmedian(gsp.groupby(['background']).get_group('ancestral')[param].values)#median()\n",
    "    gsp.loc[:,param] = gsp[param] - ref\n",
    "    \n",
    "    gsp = gsp.groupby(['background','population','gene','genotype_short','tetrad','spore'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['background','population','gene','genotype_short'])\n",
    "    \n",
    "    ax2, bp = gsp.plot(ax=ax2, kind='box', by='population', return_type='both',\n",
    "                       labels=gsp.columns.get_level_values('population'))\n",
    "\n",
    "    plot.boxplot_custom(bp, ax2,\n",
    "                        colors=[sp_gt_short_dict['color'][env_evo][x] for x in gsp.columns.get_level_values('genotype_short')],\n",
    "                        hatches=['']*10)\n",
    "    \n",
    "    ax2.set_title('spores (haploid)', fontsize=6)\n",
    "    ax2.set_xlabel('', fontsize=6)\n",
    "    ax2.set_ylabel(r'Rel. growth rate $r_{bgk}^{a/\\alpha}$', fontsize=6)\n",
    "\n",
    "    ### hybrids barplot ###\n",
    "    ax3 = plt.subplot(gs[(env_evo,'barplot')][7:], sharey=ax2)\n",
    "    \n",
    "    # set reference as mean phenotype of the ancestral hybrid\n",
    "    ref = np.nanmedian(ghy.groupby(['background']).get_group(('ancestral','ancestral'))[param].values)\n",
    "    ghy.loc[:,param] = ghy[param] - ref\n",
    "    \n",
    "    ghy = ghy.groupby(['background','population','gene','genotype_short','tetrad','spore'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['background','population','gene','genotype_short'])\n",
    "    \n",
    "    ax3, bp = ghy.plot(ax=ax3, kind='box', by='population', return_type='both', \n",
    "                       labels=ghy.columns.get_level_values('population'))\n",
    "    \n",
    "    plot.boxplot_custom(bp, ax3, \n",
    "                        colors=[hy_gt_short_dict['color'][env_evo][x] for x in ghy.columns.get_level_values('genotype_short')],\n",
    "                        hatches=['']*30)\n",
    "    \n",
    "    # set reference as mean phenotype of the ancestral hybrid\n",
    "    ref = np.nanmedian(gcl.groupby(['background']).get_group(('ancestral','ancestral'))[param].values)\n",
    "    gcl.loc[:,param] = gcl[param] - ref\n",
    "    \n",
    "    gcl = gcl.groupby(['background','population','gene','genotype_short','isolate'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['background','population','gene','genotype_short'])\n",
    "\n",
    "    gcl = gcl.iloc[:, gcl.columns.get_level_values('background').isin([(u'evolved',u'evolved')])]\n",
    "\n",
    "    gcl = gcl.reindex(columns=ghy.columns)\n",
    "    \n",
    "    for i,d in enumerate(gcl):\n",
    "        y = gcl[d]\n",
    "        x = [i+1]*len(y)\n",
    "        ax3.plot(x, y, markersize=3, marker=\"x\", markeredgecolor='k', linestyle=\"None\")\n",
    "\n",
    "    ax3.set_title('hybrids (diploid)', fontsize=6)\n",
    "    ax3.set_xlabel('', fontsize=6)\n",
    "    ax3.set_ylabel('', fontsize=6)\n",
    "        \n",
    "    # Grid\n",
    "    for ax, g in zip([ax2,ax3],[gsp, ghy]):\n",
    "        \n",
    "        ## background\n",
    "        xstart, xend, xlabels = plot.set_custom_labels(g.columns, 0)\n",
    "        # labels\n",
    "        for k, v in xlabels.iteritems():\n",
    "            ax.annotate('\\nx\\n'.join(k) if isinstance(k, tuple) else k, \n",
    "                        xy=(v+1, 0.97), xycoords=(\"data\", \"axes fraction\"),\n",
    "                        ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "        # grid\n",
    "        xgrid=[xst+1.5 for xst in list(set(xstart.values()))]\n",
    "        [ax.axvline(x, lw=1.0, ls=\"-\", color=\"0.9\") for x in xgrid]\n",
    "        \n",
    "        ## population\n",
    "        # labels\n",
    "        di = {u'no driver': u'no\\ndriver'}\n",
    "        xstart, xend, xlabels = plot.set_custom_labels(g.columns, 1)\n",
    "        xlabels = {('x'.join(k) if isinstance(k, tuple) else k): v for k,v in xlabels.items() if k != u'WAxNA'}\n",
    "        for k, v in xlabels.iteritems():\n",
    "            ax.annotate(k, xy=(v+1, -0.05), xycoords=(\"data\", \"axes fraction\"),\n",
    "                        ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "\n",
    "        ## gene\n",
    "        di = {u'no driver': u'no\\ndriver', u'/': u'',\n",
    "              u'/RNR2': u'RNR2/RNR2', u'/RNR4': u'RNR4/RNR4', \n",
    "              u'/no driver': u'no\\ndriver', u'/FPR1': u'FPR1/FPR1', u'/TOR1': u'TOR1/TOR1',\n",
    "              u'no driver/no driver': u'no\\ndriver'}\n",
    "        xstart, xend, xlabels = plot.set_custom_labels(g.columns, 2)\n",
    "        ax.set_xticks([x+1 for x in xlabels.values()], minor=False)\n",
    "        xlabels = ['/'.join(k) if isinstance(k, tuple) else k for k in xlabels.keys()]\n",
    "        xlabels = [di[x] if x in di else x for x in xlabels]\n",
    "        ax.set_xticklabels(xlabels, minor=False, fontsize=4, va='top')\n",
    "\n",
    "        ## genotype\n",
    "        # labels\n",
    "        xlabels = g.columns.get_level_values('genotype_short')\n",
    "        ax.set_xticks([x+1 for x in np.arange(len(xlabels))], minor=True)\n",
    "        ax.set_xticklabels(['/'.join(x) if isinstance(x, tuple) else x for x in xlabels], \n",
    "                           minor=True, fontsize=5)\n",
    "                \n",
    "        ax.tick_params(axis='x', which='minor', size=0, pad=-5)\n",
    "        ax.tick_params(axis='x', which='major', size=0, pad=5)\n",
    "        ax.tick_params(axis='y', which='major', labelsize=5)\n",
    "    \n",
    "    # set spine and tick properties\n",
    "    for ax in fig.get_axes():\n",
    "        for loc in ['top','bottom','left','right']:\n",
    "            ax.spines[loc].set_linewidth(1)\n",
    "        \n",
    "# # show legend\n",
    "# leg1 = ax3.legend(bbox_to_anchor=(1.05, 0.5), frameon=False,\n",
    "#                   loc='upper left', borderaxespad=0.,\n",
    "#                   handlelength=0.75, title='Background', prop={'size':6})\n",
    "    \n",
    "# leg2 = ax3.legend(bbox_to_anchor=(1.05, 0.5), frameon=False,\n",
    "#                   loc='center left', borderaxespad=0.,\n",
    "#                   handlelength=0.75, title='Genotype', prop={'size':6})\n",
    "\n",
    "fig.savefig(dir_paper+'figures/figure5/figure5.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_paper+'figures/figure5/figure5.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "groups_sp = spores_norm_df.groupby(['env_evo','env_test'])\n",
    "groups_hy = hybrids_norm_df.groupby(['env_evo','env_test'])\n",
    "groups_cl = clone_norm_df.groupby(['env_evo','env_test'])\n",
    "\n",
    "param='growth_rate'\n",
    "\n",
    "fig = plt.figure(figsize=(12, 5), dpi=300)\n",
    "\n",
    "import matplotlib.gridspec as gridspec\n",
    "grid = gridspec.GridSpec(2, 3, width_ratios=[1,3,1], hspace=0.3, wspace=0.2)\n",
    "\n",
    "gs = {}\n",
    "gs['variance'] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,0])\n",
    "gs[('HU','heatmap')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,2])\n",
    "gs[('HU','barplot')] = gridspec.GridSpecFromSubplotSpec(1, 24, subplot_spec=grid[0,1], wspace=0)\n",
    "gs[('RM','heatmap')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,2])\n",
    "gs[('RM','barplot')] = gridspec.GridSpecFromSubplotSpec(1, 24, subplot_spec=grid[1,1], wspace=0)\n",
    "\n",
    "for (ii, (env_evo, env_test)) in enumerate([('HU','YNB'),('RM','YNB')]):#,('HU','YNB'),('RM','RM'),('RM','YNB')]):\n",
    "    \n",
    "    # group by evolution and test environment\n",
    "    gsp = groups_sp.get_group((env_evo, env_test))\n",
    "    ghy = groups_hy.get_group((env_evo, env_test))\n",
    "    gcl = groups_cl.get_group((env_evo, env_test))\n",
    "\n",
    "    # apply sorting ranks to reorder rows\n",
    "    gsp.loc[:,u'rank_background'] = gsp[u'background'].map(dict_background)\n",
    "    gsp.loc[:,u'rank_population'] = gsp[u'population'].map(dict_population)\n",
    "    gsp.loc[:,u'rank_genotype'] = gsp[u'genotype_short'].map(dict_genotype)\n",
    "    gsp.loc[:,u'rank_gene'] = gsp[u'gene'].map(dict_gene)\n",
    "    gsp.sort_values(by=[u'rank_background',u'rank_population',u'rank_genotype',u'rank_gene'],\n",
    "                    ascending=True,inplace=True)\n",
    "    \n",
    "    ghy.loc[:,u'rank_background_MATa'] = ghy[u'background_MATa'].map(dict_background)\n",
    "    ghy.loc[:,u'rank_background_MATα'] = ghy[u'background_MATα'].map(dict_background)\n",
    "    ghy.loc[:,u'rank_population_MATa'] = ghy[u'population_MATa'].map(dict_population)\n",
    "    ghy.loc[:,u'rank_population_MATα'] = ghy[u'population_MATα'].map(dict_population)\n",
    "    ghy.loc[:,u'rank_genotype_MATa'] = ghy[u'genotype_short_MATa'].map(dict_genotype)\n",
    "    ghy.loc[:,u'rank_genotype_MATα'] = ghy[u'genotype_short_MATα'].map(dict_genotype)\n",
    "    ghy.loc[:,u'rank_gene_MATa'] = ghy[u'gene_MATa'].map(dict_gene)\n",
    "    ghy.loc[:,u'rank_gene_MATα'] = ghy[u'gene_MATα'].map(dict_gene)\n",
    "    ghy.sort_values(by=[u'rank_background_MATa',u'rank_background_MATα',\n",
    "                        u'rank_population_MATa',u'rank_population_MATα',\n",
    "                        u'rank_gene_MATa',u'rank_gene_MATα',\n",
    "                        u'rank_genotype_MATa',u'rank_genotype_MATα'],\n",
    "                    ascending=True, inplace=True)\n",
    "\n",
    "    ### heatmap ###\n",
    "    ax1 = plt.subplot(gs[(env_evo,'heatmap')][0])\n",
    "    \n",
    "    ax1.text(-0.1, 1.2,  chr(ii + ord('C')), transform=ax1.transAxes,\n",
    "             fontsize=12, fontweight='bold', va='top', ha='right')\n",
    "    \n",
    "    # spores\n",
    "    S = gsp\n",
    "    S = S.groupby([u'mating',u'background',u'population',u'genotype_short',u'gene'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    S_ref = S.ix[:,'ancestral'].mean()\n",
    "    \n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    \n",
    "    heatmap_spores(S-S_ref, ax1, title, \n",
    "                   xlabel, ylabel, xticklabels, yticklabels, \n",
    "                   cmap=plt.cm.RdBu_r, radius=0.25)\n",
    "    \n",
    "    # hybrids    \n",
    "    H = ghy\n",
    "    H = H.groupby([u'background_MATa',u'population_MATa',u'gene_MATa',u'genotype_short_MATa',\n",
    "                   u'background_MATα',u'population_MATα',u'gene_MATα',u'genotype_short_MATα'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    H = H.unstack(level=[u'background_MATα',u'population_MATα',u'gene_MATα',u'genotype_short_MATα'])\n",
    "    H_ref = H.loc['ancestral','ancestral'].loc['WAxNA','WAxNA'].values.squeeze()#.loc['',''].loc['+','+']\n",
    "\n",
    "    title = environments[env_test]\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    \n",
    "    heatmap_hybrids(H-H_ref, ax1, title,\n",
    "                    xlabel, ylabel, xticklabels, yticklabels, \n",
    "                    fold=False, cmap=plt.cm.RdBu_r, pad=30)\n",
    "    \n",
    "    ax1.xaxis.set_ticks_position('top')\n",
    "    ax1.yaxis.set_ticks_position('left')\n",
    "    \n",
    "    ## genotype\n",
    "    # labels\n",
    "    xlabels = H.columns.get_level_values(u'genotype_short_MATα')\n",
    "    ax1.set_xticks([x+0.5 for x in np.arange(len(xlabels))], minor=True)\n",
    "    ax1.set_xticklabels(xlabels, minor=True, fontsize=5)\n",
    "    \n",
    "    ylabels = H.index.get_level_values(u'genotype_short_MATa')\n",
    "    ax1.set_yticks([y+0.5 for y in np.arange(len(ylabels))], minor=True)\n",
    "    ax1.set_yticklabels(ylabels, minor=True, fontsize=5, rotation=90)\n",
    "    \n",
    "    ## gene\n",
    "    # labels\n",
    "    xstart, xend, xlabels = set_custom_labels(H.columns, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(v+0.5, 1.15), xycoords=(\"data\", \"axes fraction\"),\n",
    "                     ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "    ystart, yend, ylabels = set_custom_labels(H.index, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(-0.15, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                     ha='left', va='center', annotation_clip=False, fontsize=4, rotation=90)\n",
    "        \n",
    "    ## population\n",
    "    # labels\n",
    "    xstart, xend, xlabels = set_custom_labels(H.columns, 1)\n",
    "    xlabels = {k: v for k,v in xlabels.items() if k != u'WAxNA'}\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(v+0.5, 1.15), xycoords=(\"data\", \"axes fraction\"),\n",
    "                     ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "    ystart, yend, ylabels = set_custom_labels(H.index, 1)\n",
    "    ylabels = {k: v for k,v in ylabels.items() if k != u'WAxNA'}\n",
    "    for k, v in ylabels.iteritems():\n",
    "        ax1.annotate(k, xy=(-0.15, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                     ha='left', va='center', annotation_clip=False, fontsize=4, rotation=90)\n",
    "    \n",
    "    ## background\n",
    "    xstart, xend, xlabels = set_custom_labels(H.columns, 0)\n",
    "    ax1.set_xticks([x+0.5 for x in xlabels.values()], minor=False)\n",
    "    ax1.set_xticklabels(xlabels.keys(), minor=False, fontsize=5)\n",
    "    xgrid=[xst+1. for xst in list(set(xstart.values()))]\n",
    "    [ax1.axvline(g, lw=1, ls=\"-\", color=\"black\") for g in xgrid]\n",
    "    \n",
    "    ystart, yend, ylabels= set_custom_labels(H.columns, 0)\n",
    "    ax1.set_yticks([y+0.5 for y in ylabels.values()], minor=False)\n",
    "    ax1.set_yticklabels(ylabels.keys(), minor=False, fontsize=5, rotation=90)\n",
    "    ygrid=[yst+1. for yst in list(set(ystart.values()))]\n",
    "    [ax1.axhline(g, lw=1, ls=\"-\", color=\"black\") for g in ygrid]\n",
    "    \n",
    "    ax1.tick_params(axis='both', which='minor', size=0, pad=9)\n",
    "    ax1.tick_params(axis='both', which='major', size=0, pad=20)\n",
    "    \n",
    "    # filter by dictionary\n",
    "    gsp = gsp[(gsp['background'].isin(sp_bg_dict['position'][env_evo].keys())) &\n",
    "              (gsp['genotype_short'].isin(sp_gt_short_dict['position'][env_evo].keys())) &\n",
    "              (gsp['population'].isin(sp_cl_dict['position'][env_evo].keys()))]\n",
    "    ghy = ghy[(ghy['background'].isin(hy_bg_dict['position'][env_evo].keys())) &\n",
    "              (ghy['genotype_short'].isin(hy_gt_short_dict['position'][env_evo].keys())) &\n",
    "              (ghy['population'].isin(hy_cl_dict['position'][env_evo].keys()))]\n",
    "    \n",
    "    # apply sorting ranks to reorder rows\n",
    "    gsp.loc[:,'rank_background'] = gsp['background'].map(sp_bg_dict['position'][env_evo])\n",
    "    gsp.loc[:,'rank_population'] = gsp['population'].map(sp_cl_dict['position'][env_evo])\n",
    "    gsp.loc[:,'rank_gene'] = gsp['gene'].map(sp_gn_dict['position'][env_evo])\n",
    "    gsp.loc[:,'rank_genotype'] = gsp['genotype_short'].map(sp_gt_short_dict['position'][env_evo])\n",
    "    gsp.sort_values(['rank_background','rank_population','rank_gene','rank_genotype'],\n",
    "                    ascending=True,inplace=True)\n",
    "\n",
    "    ghy.loc[:,'rank_background'] = ghy['background'].map(hy_bg_dict['position'][env_evo])\n",
    "    ghy.loc[:,'rank_population'] = ghy['population'].map(hy_cl_dict['position'][env_evo])\n",
    "    ghy.loc[:,'rank_gene'] = ghy['gene'].map(hy_gn_dict['position'][env_evo])\n",
    "    ghy.loc[:,'rank_genotype'] = ghy['genotype_short'].map(hy_gt_short_dict['position'][env_evo])\n",
    "    ghy.sort_values(['rank_background','rank_population','rank_gene','rank_genotype'],\n",
    "                    ascending=True,inplace=True)\n",
    "    \n",
    "    ### spores barplot ###\n",
    "    ax2 = plt.subplot(gs[(env_evo,'barplot')][:7])\n",
    "\n",
    "    ax2.text(-0.1, 1.2,  chr(ii + ord('A')), transform=ax2.transAxes,\n",
    "             fontsize=12, fontweight='bold', va='top', ha='right')\n",
    "    \n",
    "    gsp = gsp.groupby(['background','population','gene','genotype_short','tetrad','spore'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['background','population','gene','genotype_short'])\n",
    "    \n",
    "    ax2, bp = gsp.plot(ax=ax2, kind='box', by='population', return_type='both',\n",
    "                       labels=gsp.columns.get_level_values('population'))\n",
    "\n",
    "    boxplot_custom(bp, ax2,\n",
    "                   colors=[sp_gt_short_dict['color'][env_evo][x] for x in gsp.columns.get_level_values('genotype_short')],\n",
    "                   hatches=['/']*10)\n",
    "    \n",
    "    ax2.set_title('spores (haploid)', fontsize=6)\n",
    "    ax2.set_xlabel('', fontsize=6)\n",
    "    ax2.set_ylabel(r'Rel. doubling time $\\langle\\tau\\rangle_{b,g,k}^{a/\\alpha}$', fontsize=6)\n",
    "\n",
    "    ### hybrids barplot ###\n",
    "    ax3 = plt.subplot(gs[(env_evo,'barplot')][7:], sharey=ax2)\n",
    "    \n",
    "    ghy = ghy.groupby(['background','population','gene','genotype_short','tetrad','spore'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['background','population','gene','genotype_short'])\n",
    "    \n",
    "    ax3, bp = ghy.plot(ax=ax3, kind='box', by='population', return_type='both', \n",
    "                       labels=ghy.columns.get_level_values('population'))\n",
    "    \n",
    "    gcl = gcl.groupby(['background','population','gene','genotype_short','isolate'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['background','population','gene','genotype_short'])\n",
    "        \n",
    "#     ax3, bp = gcl.plot(ax=ax3, kind='box', by='population', return_type='both', \n",
    "#                        labels=gcl.columns.get_level_values('population'))\n",
    "\n",
    "    boxplot_custom(bp, ax3, \n",
    "                   colors=[hy_gt_short_dict['color'][env_evo][x] for x in ghy.columns.get_level_values('genotype_short')],\n",
    "                   hatches=['/']*30)\n",
    "    \n",
    "    ax3.set_title('hybrids (diploid)', fontsize=6)\n",
    "    ax3.set_xlabel('', fontsize=6)\n",
    "    ax3.set_ylabel('', fontsize=6)\n",
    "        \n",
    "    # Grid\n",
    "    for ax, g in zip([ax2,ax3],[gsp, ghy]):\n",
    "        \n",
    "        ## population\n",
    "        # labels\n",
    "#         xstart, xend, xlabels = set_custom_labels(g.columns, 1)\n",
    "#         ax.set_xticks([x+1 for x in xlabels.values()], minor=False)\n",
    "#         ax.set_xticklabels(xlabels.keys(), minor=False, fontsize=5)\n",
    "        \n",
    "        ## background\n",
    "        xstart, xend, xlabels = set_custom_labels(g.columns, 0)\n",
    "        # labels\n",
    "        for k, v in xlabels.iteritems():\n",
    "            ax.annotate('\\nx\\n'.join(k) if isinstance(k, tuple) else k, \n",
    "                        xy=(v+1, 0.97), xycoords=(\"data\", \"axes fraction\"),\n",
    "                        ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "        # grid\n",
    "        xgrid=[xst+1.5 for xst in list(set(xstart.values()))]\n",
    "        [ax.axvline(x, lw=1.0, ls=\"-\", color=\"0.9\") for x in xgrid]\n",
    "        \n",
    "        ## population\n",
    "        # labels\n",
    "        di = {u'no driver': u'no\\ndriver'}\n",
    "        xstart, xend, xlabels = set_custom_labels(g.columns, 1)\n",
    "#         xlabels = {k: v for k,v in xlabels.items() if k != u'WAxNA'}\n",
    "        xlabels = {('x'.join(k) if isinstance(k, tuple) else k): v for k,v in xlabels.items()}\n",
    "        for k, v in xlabels.iteritems():\n",
    "            ax.annotate(k, xy=(v+1, -0.1), xycoords=(\"data\", \"axes fraction\"),\n",
    "                        ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "\n",
    "        ## gene\n",
    "        di = {u'no driver': u'no\\ndriver', u'/': u'',\n",
    "              u'/RNR2': u'RNR2/RNR2', u'/RNR4': u'RNR4/RNR4', \n",
    "              u'/no driver': u'no\\ndriver', u'/FPR1': u'FPR1/FPR1', u'/TOR1': u'TOR1/TOR1',\n",
    "              u'no driver/no driver': u'no\\ndriver'}\n",
    "        xstart, xend, xlabels = set_custom_labels(g.columns, 2)\n",
    "        ax.set_xticks([x+1 for x in xlabels.values()], minor=False)\n",
    "        xlabels = ['/'.join(k) if isinstance(k, tuple) else k for k in xlabels.keys()]\n",
    "        xlabels = [di[x] if x in di else x for x in xlabels]\n",
    "        ax.set_xticklabels(xlabels, minor=False, fontsize=4, va='bottom')\n",
    "\n",
    "        ## genotype\n",
    "        # labels\n",
    "        xlabels = g.columns.get_level_values('genotype_short')\n",
    "        ax.set_xticks([x+1 for x in np.arange(len(xlabels))], minor=True)\n",
    "        ax.set_xticklabels(['/'.join(x) if isinstance(x, tuple) else x for x in xlabels], \n",
    "                           minor=True, fontsize=5)\n",
    "                \n",
    "        ax.tick_params(axis='x', which='minor', size=0, pad=-6)\n",
    "        ax.tick_params(axis='x', which='major', size=0, pad=3)\n",
    "        ax.tick_params(axis='y', which='major', labelsize=5)\n",
    "    \n",
    "    # set spine and tick properties\n",
    "    for ax in fig.get_axes():        \n",
    "        \n",
    "        for loc in ['top','bottom','left','right']:\n",
    "            ax.spines[loc].set_linewidth(1)\n",
    "        \n",
    "# show legend\n",
    "leg1 = ax3.legend(bbox_to_anchor=(1.05, 0.5), frameon=False,\n",
    "                  loc='upper left', borderaxespad=0.,\n",
    "                  handlelength=0.75, title='Background', prop={'size':6})\n",
    "    \n",
    "leg2 = ax3.legend(bbox_to_anchor=(1.05, 0.5), frameon=False,\n",
    "                  loc='center left', borderaxespad=0.,\n",
    "                  handlelength=0.75, title='Genotype', prop={'size':6})\n",
    "\n",
    "fig.savefig(dir_supp+'figures/figureSX_pheno_cross/figureSX_pheno_cross_reduced.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_supp+'figures/figureSX_pheno_cross/figureSX_pheno_cross_reduced.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "groups_sp = spores_norm_df.groupby(['env_evo','env_test'])\n",
    "groups_hy = hybrids_norm_df.groupby(['env_evo','env_test'])\n",
    "groups_cl = clone_norm_df.groupby(['env_evo','env_test'])\n",
    "\n",
    "param='growth_rate'\n",
    "\n",
    "fig = plt.figure(figsize=(6, 6), dpi=300)\n",
    "\n",
    "import matplotlib.gridspec as gridspec\n",
    "grid = gridspec.GridSpec(2, 2, hspace=0.3, wspace=0.2)\n",
    "\n",
    "gs = {}\n",
    "gs[('HU','HU')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,0])\n",
    "gs[('HU','YNB')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,1])\n",
    "gs[('RM','RM')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,0])\n",
    "gs[('RM','YNB')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,1])\n",
    "\n",
    "for (ii, (env_evo, env_test)) in enumerate([('HU','HU'),('HU','YNB'),('RM','RM'),('RM','YNB')]):\n",
    "    \n",
    "    # group by evolution and test environment\n",
    "    gsp = groups_sp.get_group((env_evo, env_test))\n",
    "    ghy = groups_hy.get_group((env_evo, env_test))\n",
    "    gcl = groups_cl.get_group((env_evo, env_test))\n",
    "\n",
    "    # apply sorting ranks to reorder rows\n",
    "    gsp.loc[:,u'rank_background'] = gsp[u'background'].map(dict_background)\n",
    "    gsp.loc[:,u'rank_population'] = gsp[u'population'].map(dict_population)\n",
    "    gsp.loc[:,u'rank_genotype'] = gsp[u'genotype_short'].map(dict_genotype)\n",
    "    gsp.loc[:,u'rank_gene'] = gsp[u'gene'].map(dict_gene)\n",
    "    gsp.sort_values(by=[u'rank_background',u'rank_population',u'rank_genotype',u'rank_gene'],\n",
    "                    ascending=True,inplace=True)\n",
    "    \n",
    "    ghy.loc[:,u'rank_background_MATa'] = ghy[u'background_MATa'].map(dict_background)\n",
    "    ghy.loc[:,u'rank_background_MATα'] = ghy[u'background_MATα'].map(dict_background)\n",
    "    ghy.loc[:,u'rank_population_MATa'] = ghy[u'population_MATa'].map(dict_population)\n",
    "    ghy.loc[:,u'rank_population_MATα'] = ghy[u'population_MATα'].map(dict_population)\n",
    "    ghy.loc[:,u'rank_genotype_MATa'] = ghy[u'genotype_short_MATa'].map(dict_genotype)\n",
    "    ghy.loc[:,u'rank_genotype_MATα'] = ghy[u'genotype_short_MATα'].map(dict_genotype)\n",
    "    ghy.loc[:,u'rank_gene_MATa'] = ghy[u'gene_MATa'].map(dict_gene)\n",
    "    ghy.loc[:,u'rank_gene_MATα'] = ghy[u'gene_MATα'].map(dict_gene)\n",
    "    ghy.sort_values(by=[u'rank_background_MATa',u'rank_background_MATα',\n",
    "                        u'rank_population_MATa',u'rank_population_MATα',\n",
    "                        u'rank_gene_MATa',u'rank_gene_MATα',\n",
    "                        u'rank_genotype_MATa',u'rank_genotype_MATα'],\n",
    "                    ascending=True, inplace=True)\n",
    "\n",
    "    ### heatmap ###\n",
    "    ax1 = plt.subplot(gs[(env_evo, env_test)][0])\n",
    "    \n",
    "    ax1.text(-0.1, 1.2,  chr(ii + ord('A')), transform=ax1.transAxes,\n",
    "             fontsize=12, fontweight='bold', va='top', ha='right')\n",
    "    \n",
    "    # spores\n",
    "    S = gsp\n",
    "    S = S.groupby([u'mating',u'background',u'population',u'genotype_short',u'gene',u'tetrad',u'spore'],\n",
    "                    sort=False)[param].agg(np.mean).reset_index()\n",
    "    S = S.set_index('mating')[param]\n",
    "    S_ref = S.values.mean()\n",
    "        \n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    \n",
    "    heatmap_spores(S-S_ref, ax1, title, \n",
    "                   xlabel, ylabel, xticklabels, yticklabels, \n",
    "                   cmap=plt.cm.RdBu_r, radius=0.5)\n",
    "    \n",
    "    # hybrids    \n",
    "    H = ghy\n",
    "    H = H.groupby([u'background_MATa',u'population_MATa',u'gene_MATa',u'genotype_short_MATa',u'tetrad_MATa',u'spore_MATa',\n",
    "                   u'background_MATα',u'population_MATα',u'gene_MATα',u'genotype_short_MATα',u'tetrad_MATα',u'spore_MATα'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    H = H.unstack(level=[u'background_MATα',u'population_MATα',u'gene_MATα',u'genotype_short_MATα',u'tetrad_MATα',u'spore_MATα'])\n",
    "    H = H.fillna(0)\n",
    "    H_ref = H.loc['ancestral','ancestral'].loc['WAxNA','WAxNA'].values.mean()\n",
    "\n",
    "    title = environments[env_test]\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    \n",
    "    heatmap_hybrids(H-H_ref, ax1, title,\n",
    "                    xlabel, ylabel, xticklabels, yticklabels, \n",
    "                    fold=False, cmap=plt.cm.RdBu_r, pad=10)\n",
    "    \n",
    "    ax1.xaxis.set_ticks_position('top')\n",
    "    ax1.yaxis.set_ticks_position('left')\n",
    "    \n",
    "    ## genotype\n",
    "    # labels\n",
    "    xlabels = H.columns.get_level_values(u'genotype_short_MATα')\n",
    "    ax1.set_xticks([x+0.5 for x in np.arange(len(xlabels))], minor=True)\n",
    "    ax1.set_xticklabels(xlabels, minor=True, fontsize=5)\n",
    "    \n",
    "    ylabels = H.index.get_level_values(u'genotype_short_MATa')\n",
    "    ax1.set_yticks([y+0.5 for y in np.arange(len(ylabels))], minor=True)\n",
    "    ax1.set_yticklabels(ylabels, minor=True, fontsize=5, rotation=90)\n",
    "    \n",
    "    ## gene\n",
    "    # labels\n",
    "    xstart, xend, xlabels = set_custom_labels(H.columns, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(v+0.5, 1.08), xycoords=(\"data\", \"axes fraction\"),\n",
    "                     ha='center', va='top', annotation_clip=False, fontsize=4, style='italic')\n",
    "    ystart, yend, ylabels = set_custom_labels(H.index, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(-0.08, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                     ha='left', va='center', annotation_clip=False, fontsize=4, style='italic', rotation=90)\n",
    "        \n",
    "    ## population\n",
    "    # labels\n",
    "    xstart, xend, xlabels = set_custom_labels(H.columns, 1)\n",
    "    xlabels = {k: v for k,v in xlabels.items() if k != u'WAxNA'}\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(v+0.5, 1.08), xycoords=(\"data\", \"axes fraction\"),\n",
    "                     ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "    ystart, yend, ylabels = set_custom_labels(H.index, 1)\n",
    "    ylabels = {k: v for k,v in ylabels.items() if k != u'WAxNA'}\n",
    "    for k, v in ylabels.iteritems():\n",
    "        ax1.annotate(k, xy=(-0.08, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                     ha='left', va='center', annotation_clip=False, fontsize=4, rotation=90)\n",
    "    \n",
    "    ## background\n",
    "    xstart, xend, xlabels = set_custom_labels(H.columns, 0)\n",
    "    ax1.set_xticks([x+0.5 for x in xlabels.values()], minor=False)\n",
    "    ax1.set_xticklabels(xlabels.keys(), minor=False, fontsize=5)\n",
    "    xgrid=[xst+1. for xst in list(set(xstart.values()))]\n",
    "    [ax1.axvline(g, lw=1, ls=\"-\", color=\"black\") for g in xgrid]\n",
    "    \n",
    "    ystart, yend, ylabels= set_custom_labels(H.columns, 0)\n",
    "    ax1.set_yticks([y+0.5 for y in ylabels.values()], minor=False)\n",
    "    ax1.set_yticklabels(ylabels.keys(), minor=False, fontsize=5, rotation=90)\n",
    "    ygrid=[yst+1. for yst in list(set(ystart.values()))]\n",
    "    [ax1.axhline(g, lw=1, ls=\"-\", color=\"black\") for g in ygrid]\n",
    "    \n",
    "    ax1.tick_params(axis='both', which='minor', size=0, pad=2)\n",
    "    ax1.tick_params(axis='both', which='major', size=0, pad=15)\n",
    "\n",
    "fig.savefig(dir_supp+'figures/figureSX_pheno_cross/figureSX_pheno_cross_extended.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_supp+'figures/figureSX_pheno_cross/figureSX_pheno_cross_extended.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for (ii, (env_evo, env_test)) in enumerate([('HU','HU'),('RM','RM')]):#,('HU','YNB'),('RM','RM'),('RM','YNB')]):\n",
    "    \n",
    "    # group by evolution and test environment\n",
    "    gsp = groups_sp.get_group((env_evo, env_test))\n",
    "    ghy = groups_hy.get_group((env_evo, env_test))\n",
    "    gcl = groups_cl.get_group((env_evo, env_test))\n",
    "    \n",
    "    # filter by dictionary\n",
    "    gsp = gsp[(gsp['genotype'].isin([u'fpr1Δ',u'tor1Δ']))]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To examine the average mutational effect of beneficial variants, we considered both the (Fig. 5) and the (Fig. ) as an average over many possible backgrounds. To determine fitness effects of the evolved mutations in HU (*RNR2*, *RNR4*) and RM (*FPR1*, *TOR1*), we isolated individuals from ancestral and evolved populations. From these diploid cells, we sporulated and selected haploid segregants of each mating type (48 in HU and 52 in RM). We crossed the MATa and MATα versions to create hybrids. A number of independent spores and hybrids were used to measure biological variability of ancestral and evolved backgrounds. We found all de novo mutations to be unconditionally beneficial, yet their association to genetic backgrounds idiosyncratic effects.\n",
    "\n",
    "Figure 5. Average mutational effect in PSD95pdz3 when binding the wild-type ligand CRIPT ($\\langle\\Delta\\tau_i^x\\rangle$, top) or a class-switching T22F variant ($\\langle\\Delta\\tau\\rangle$, bottom). T22F contains a Thr to Phe mutation at position minus two of the peptide ligand (Fig. 1a).\n",
    "\n",
    "(A) block diagonal elements $(b,g)_a=(b,g)_\\alpha$ off-diagonal elements $(b,g)_a\\neq (b,g)_\\alpha$.\n",
    "\n",
    "(B-C) Average doubling time of the spores and the hybrids with respect to the ancestral WAxNA individuals, indicating the median and the 25% / 75% percentiles.\n",
    "\n",
    "(D-E) . (-- (homozygote); -+ (heterozygote); ++ (homozygote)). The data matrix shows the functional cost of every genotype $g$ at each background $b$ relative to wild-type PSD95pdz3\n",
    "\n",
    "Selected clones (circled) are. The difference, or epistasis, between mutational effects $\\langle\\Delta\\Delta\\tau_g\\rangle_k$\n",
    "\n",
    "Mutations rise or fall in frequency because of their own effect on fitness, their association to genetic backgrounds, and stochastic fluctuations (genetic drift)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Background-averaged measurements ancestral and evolved isolates in two environments. Color scale for all matrices is shown to the right of A and the dimensions of each matrix, in units of hours. (A) *RNR2* and *RNR4*, (B) *FPR1* and *TOR1*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now group the dataframe by environmental condition during the propagation and testing. We now average over the ensemble of . For a given tetrad, we compare "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis of variance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "labels = {\n",
    "#     'population':'background',\n",
    "    'gene':'de_novo_gene',\n",
    "    'genotype_short':'de_novo_genotype'\n",
    "}\n",
    "\n",
    "sdf = spores_norm_df\n",
    "sdf.rename(columns=labels, inplace=True)\n",
    "sdf = sdf[sdf.background.isin(['ancestral','evolved'])]\n",
    "sdf = sdf[np.isfinite(sdf['growth_rate'])]\n",
    "\n",
    "hdf = hybrids_norm_df\n",
    "hdf.rename(columns=labels, inplace=True)\n",
    "hdf = hdf[hdf.background.isin([('ancestral','ancestral'),('ancestral','evolved'),('evolved','evolved')])]\n",
    "hdf = hdf[np.isfinite(hdf['growth_rate'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ANOVA (frequentist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# fig, axes = plt.subplots(nrows=4, ncols=1, figsize=(2, 5), sharex=True, sharey=True)\n",
    "\n",
    "fig = plt.figure(figsize=(6, 1), dpi=300)\n",
    "\n",
    "import matplotlib.gridspec as gridspec\n",
    "    \n",
    "formulas = ['growth_rate ~ background + de_novo_gene',\n",
    "            'growth_rate ~ background + de_novo_gene + de_novo_genotype',\n",
    "            'growth_rate ~ background + de_novo_gene + de_novo_genotype',\n",
    "            'growth_rate ~ background + de_novo_gene + de_novo_genotype + auxotrophy',\n",
    "            'growth_rate ~ background + de_novo_gene + de_novo_genotype + auxotrophy + tetrad + C(spore)']\n",
    "\n",
    "grid = gridspec.GridSpec(nrows=1, ncols=5, hspace=0.2, wspace=0.22)\n",
    "\n",
    "gs = {}\n",
    "for ii in np.arange(len(formulas)):\n",
    "    gs[ii] = gridspec.GridSpecFromSubplotSpec(2, 1, subplot_spec=grid[ii], hspace=0)\n",
    "\n",
    "for ii, formula in enumerate(formulas):\n",
    "    \n",
    "    msdf = sdf.groupby(['env_evo','env_test']).apply(mean_vectors, formula) \n",
    "    mhdf = hdf.groupby(['env_evo','env_test']).apply(mean_vectors, formula)\n",
    "        \n",
    "    tsdf = sdf.groupby(['env_evo','env_test']).apply(variance_vectors, formula) \n",
    "    thdf = hdf.groupby(['env_evo','env_test']).apply(variance_vectors, formula)\n",
    "    df = pd.concat([tsdf, thdf], keys=['spores','hybrids'], names=['type'])\n",
    "    \n",
    "    # fix index\n",
    "    df = df.swaplevel('type', 'env_evo')\n",
    "    df = df.swaplevel('type', 'env_test')\n",
    "    df = df.sort_index(level=['env_evo','env_test','type'])\n",
    "    \n",
    "    # fix columns\n",
    "    df = df.rename(columns={'C(spore)':'spore'})\n",
    "  \n",
    "    df = df.sort_index(axis=1)\n",
    "#     order = ['background','de_novo_gene','de_novo_genotype','auxotrophy','tetrad','spore']\n",
    "#     order = pd.Series(list(set(order) & set(df.columns)))\n",
    "#     df = df[order]\n",
    "    \n",
    "    ax1 = plt.subplot(gs[ii][0])\n",
    "    utils.simple_axes(ax1)\n",
    "    \n",
    "    s = df.ix['HU']\n",
    "    s = s.sort_index(ascending=False)\n",
    "    s.plot(ax=ax1, kind='barh', stacked=True, \n",
    "           color=[factor_dict['color'][x] for x in s.columns],\n",
    "           align='center', legend=False, width=0.75)\n",
    "    \n",
    "    bars = ax1.patches\n",
    "    hatches = ''.join([factor_dict['hatch'][x] for x in s.index.get_level_values('type')]*len(s)*2)\n",
    "\n",
    "    for bar, hatch in zip(bars, hatches):\n",
    "        bar.set_hatch(hatch)\n",
    "    \n",
    "    ax1.set_title(formula, fontsize=4)\n",
    "    ax1.set_xlabel('')\n",
    "    ax1.set_xticks([])\n",
    "    ax1.set_xticklabels([])    \n",
    "    ystart, yend, ylabels = set_custom_labels(s.index, 0)\n",
    "    ax1.set_yticks(ylabels.values(), minor=False)\n",
    "    ax1.set_yticklabels(ylabels.keys(), minor=False, fontsize=5)\n",
    "        \n",
    "    ax2 = plt.subplot(gs[ii][1])\n",
    "    utils.simple_axes(ax2)\n",
    "    \n",
    "    h = df.ix['RM']\n",
    "    h = h.sort_index(ascending=False)\n",
    "    h.plot(ax=ax2, kind='barh', stacked=True, \n",
    "           color=[factor_dict['color'][x] for x in h.columns],\n",
    "           align='center', legend=False, width=0.75)\n",
    "    \n",
    "    bars = ax2.patches\n",
    "    hatches = ''.join([factor_dict['hatch'][x] for x in h.index.get_level_values('type')]*len(h)*2)\n",
    "\n",
    "    for bar, hatch in zip(bars, hatches):\n",
    "        bar.set_hatch(hatch)\n",
    "    \n",
    "    ax2.set_xlabel('Variance', fontsize=5)    \n",
    "    ystart, yend, ylabels = set_custom_labels(h.index, 0)\n",
    "    ax2.set_yticks(ylabels.values(), minor=False)\n",
    "    ax2.set_yticklabels(ylabels.keys(), minor=False, fontsize=5)\n",
    "    \n",
    "    for ax in [ax1,ax2]:\n",
    "        ax.set_xlim((0,1))\n",
    "#         ax.set_yticks([])\n",
    "#         ax.set_yticklabels([])\n",
    "        ax.set_ylabel('', fontsize=6)\n",
    "\n",
    "    leg2 = ax2.legend(bbox_to_anchor=(1.05, 1.0), frameon=False,\n",
    "                      loc='center left', borderaxespad=0.,\n",
    "                      handlelength=0.75, prop={'size':4})\n",
    "    \n",
    "for ax in fig.get_axes():        \n",
    "        \n",
    "    plt.setp(ax.get_xticklabels(), fontsize=5)\n",
    "    plt.setp(ax.get_yticklabels(), fontsize=5)\n",
    "    \n",
    "    ax.tick_params(size=2)\n",
    "        \n",
    "    for loc in ['top','bottom','left','right']:\n",
    "        ax.spines[loc].set_linewidth(1)\n",
    "\n",
    "fig.savefig(dir_supp+'figures/figureSX_pheno_cross/figureSX_pheno_cross_anova.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_supp+'figures/figureSX_pheno_cross/figureSX_pheno_cross_anova.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hierarchical model (Bayesian)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from pymc3 import  *\n",
    "\n",
    "with Model() as model_sat:\n",
    "    grp_mean = Normal('grp_mean', mu=0, sd=10)\n",
    "    grp_sd = Uniform('grp_sd', 0, 200)\n",
    "\n",
    "    # Define priors for intercept and regression coefficients.\n",
    "    priors = {'Intercept': Normal.dist(mu=test_df.doubling_time.mean(), sd=test_df.doubling_time.std()),\n",
    "          'env_evo': Normal.dist(mu=grp_mean, sd=grp_sd),\n",
    "          'env_test': Normal.dist(mu=grp_mean, sd=grp_sd),\n",
    "          'background': Normal.dist(mu=grp_mean, sd=grp_sd),\n",
    "          'genotype': Normal.dist(mu=grp_mean, sd=grp_sd)\n",
    "    }\n",
    "    glm.glm('growth_rate ~ env_evo + env_test + population + background + genotype', \n",
    "                    test_df,\n",
    "                priors=priors)\n",
    "    trace_sat = sample(10, NUTS(), progressbar=True)\n",
    "\n",
    "# scatter_matrix(trace_to_dataframe(trace_sat), figsize=(12,12));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Troubleshooting"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We focus on the block diagonal elements of the matrix to identify spores and hybrids that may have been mistyped for the mutation or the mating locus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# fig = plt.figure(figsize=(15, 36))\n",
    "\n",
    "# dict_panels = {\n",
    "# ('HU','HU','RNR2'):(0,0),('HU','YNB','RNR2'):(0,1),\n",
    "# ('HU','HU','RNR4'):(1,0),('HU','YNB','RNR4'):(1,1),\n",
    "# ('RM','RM','null'):(2,0),('RM','YNB','null'):(2,1),\n",
    "# ('RM','RM','FPR1'):(3,0),('RM','YNB','FPR1'):(3,1),\n",
    "# ('RM','RM','TOR1'):(4,0),('RM','YNB','TOR1'):(4,1)\n",
    "# }\n",
    "    \n",
    "# for ii, (env_evo,env_test,gene) in enumerate([('HU','HU','RNR2'),('HU','YNB','RNR2'),('HU','HU','RNR4'),('HU','YNB','RNR4'),\n",
    "#                                          ('RM','RM','null'),('RM','YNB','null'),('RM','RM','FPR1'),('RM','YNB','FPR1'),('RM','RM','TOR1'),('RM','YNB','TOR1')]):\n",
    "                    \n",
    "#     gs = spores_norm_df[(spores_norm_df.gene==gene)].groupby(['env_evo','env_test']).get_group((env_evo,env_test))\n",
    "#     gh = hybrids_norm_df[(hybrids_norm_df.gene==(gene,gene))].groupby(['env_evo','env_test']).get_group((env_evo,env_test))\n",
    "\n",
    "#     H_ref = gh[(gh.population==('ancestral','ancestral'))][param].dropna().median()\n",
    "    \n",
    "#     gh = gh.groupby([u'population_MATa',u'background_MATa',u'genotype_MATa',u'tetrad_MATa',u'spore_MATa',\n",
    "#                      u'population_MATα',u'background_MATα',u'genotype_MATα',u'tetrad_MATα',u'spore_MATα'], \n",
    "#                     sort=False).agg(np.mean)[param]#.dropna()\n",
    "        \n",
    "#     ax = plt.subplot2grid((5, 2), dict_panels[(env_evo,env_test,gene)])#dict_panels[(env_evo,env_test)])\n",
    "    \n",
    "#     S = {}\n",
    "#     S[u'MATa'] = gs[gs[u'mating']==u'MATa']\n",
    "#     S[u'MATα'] = gs[gs[u'mating']==u'MATα']\n",
    "\n",
    "#     H = gh.unstack(level=[u'population_MATα',u'background_MATα',\n",
    "#                           u'genotype_MATα',u'tetrad_MATα',u'spore_MATα'])\n",
    "            \n",
    "#     H = H#-H_ref\n",
    "      \n",
    "#     pheno.heatmap_cross(H, title=environments[env_test])\n",
    "        \n",
    "#     dict_mat = {u'MATα':{'x':range(0,len(H)), 'y':[.5+len(H)]*len(H)},\n",
    "#                 u'MATa':{'x':[-1.5]*len(H), 'y':range(0,len(H))}}\n",
    "    \n",
    "#     for mating in dict_mat.iterkeys():\n",
    "#         data = map(list, zip(*[dict_mat[mating]['x'], dict_mat[mating]['y']]))\n",
    "#         circles = [plt.Circle([x, y], 0.5) for (x,y) in data]\n",
    "#         col = PatchCollection(circles)\n",
    "\n",
    "#         s = S[mating].groupby([u'population',u'background',u'genotype',u'tetrad',u'spore'],sort=False)[param].median()\n",
    "#         s_ref = S[mating][(S[mating].population=='ancestral')][param].median()\n",
    "        \n",
    "#         col.set(array=s-s_ref, cmap='RdBu')\n",
    "#         col.set_clip_on(False)\n",
    "\n",
    "#         ax.add_collection(col)\n",
    "\n",
    "# fig.tight_layout()\n",
    "fig.subplots_adjust(left=0.18,bottom=0.1,right=0.95,top=0.98,hspace=0.80,wspace=0.85)\n",
    "fig.savefig(dir_paper+'figures/figure5/figure5_error.png', dpi=300)\n",
    "fig.savefig(dir_paper+'figures/figure5/figure5_error.svg', dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are potential genotyping errors for mutations in *FPR1* and *TOR1*. *TOR1* tetrads A, I and S. *FPR1* tetrad D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# fig = plt.figure(figsize=(15, 36))\n",
    "\n",
    "# dict_panels = {\n",
    "# ('RM','RM','TOR1','A'):(0,0),('RM','YNB','TOR1','A'):(0,1),\n",
    "# ('RM','RM','TOR1','I'):(1,0),('RM','YNB','TOR1','I'):(1,1),\n",
    "# ('RM','RM','TOR1','S'):(2,0),('RM','YNB','TOR1','S'):(2,1),\n",
    "# ('RM','RM','FPR1','D'):(3,0),('RM','YNB','FPR1','D'):(3,1)\n",
    "# }\n",
    "    \n",
    "# for ii, (env_evo,env_test,gene,tetrad) in enumerate([('RM','RM','TOR1','A'),('RM','YNB','TOR1','A'),\n",
    "#                                                      ('RM','RM','TOR1','I'),('RM','YNB','TOR1','I'),\n",
    "#                                                      ('RM','RM','TOR1','S'),('RM','YNB','TOR1','S'),\n",
    "#                                                      ('RM','RM','FPR1','D'),('RM','YNB','FPR1','D')]):\n",
    "                    \n",
    "#     gs = spores_norm_df[(spores_norm_df.gene==gene)&\\\n",
    "#                         (spores_norm_df.tetrad==tetrad)].groupby(['env_evo','env_test']).get_group((env_evo,env_test))\n",
    "#     gh = hybrids_norm_df[(hybrids_norm_df.gene==(gene,gene))&\\\n",
    "#                          (hybrids_norm_df.tetrad==(tetrad,tetrad))].groupby(['env_evo','env_test']).get_group((env_evo,env_test))\n",
    "\n",
    "#     S_ref = gs[(gs.population=='ancestral')][param].median()\n",
    "\n",
    "#     H_ref = gh[(gh.population==('ancestral','ancestral'))][param].dropna().median()\n",
    "    \n",
    "#     gs = gs.groupby([u'mating',u'population',u'background',u'genotype',u'tetrad',u'spore'], \n",
    "#                     sort=False).agg(np.mean)[param]\n",
    "    \n",
    "#     gh = gh.groupby([u'population_MATa',u'background_MATa',u'genotype_MATa',u'tetrad_MATa',u'spore_MATa',\n",
    "#                      u'population_MATα',u'background_MATα',u'genotype_MATα',u'tetrad_MATα',u'spore_MATα'], \n",
    "#                     sort=False).agg(np.mean)[param]#.dropna()\n",
    "        \n",
    "#     ax = plt.subplot2grid((4, 2), dict_panels[(env_evo,env_test,gene,tetrad)])\n",
    "    \n",
    "# #     S = {}\n",
    "# #     S[u'MATa'] = gs.loc[u'MATa']\n",
    "# #     S[u'MATα'] = gs.loc[u'MATα']\n",
    "#     S = gs\n",
    "    \n",
    "#     H = gh.unstack(level=[u'population_MATα',u'background_MATα',\n",
    "#                           u'genotype_MATα',u'tetrad_MATα',u'spore_MATα'])\n",
    "    \n",
    "#     print S[u'MATa'].shape, S[u'MATα'].shape, H.shape\n",
    "\n",
    "# #     H = H#-H_ref\n",
    "      \n",
    "#     pheno.heatmap_spores(S-S_ref)\n",
    "#     pheno.heatmap_hybrids(H-H_ref, title=environments[env_test])\n",
    "\n",
    "# # fig.tight_layout()\n",
    "# fig.subplots_adjust(left=0.18,bottom=0.1,right=0.95,top=0.98,hspace=0.80,wspace=0.85)\n",
    "# # fig.savefig(dir_paper+'figures/figure5/figure5_error.png', dpi=300)\n",
    "# # fig.savefig(dir_paper+'figures/figure5/figure5_error.svg', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# dict_panels = {\n",
    "# ('RM','RM','','A'):(0,0),('RM','YNB','','A'):(0,1),\n",
    "# ('RM','RM','','B'):(1,0),('RM','YNB','','B'):(1,1),\n",
    "# ('RM','RM','TOR1','A'):(2,0),('RM','YNB','TOR1','A'):(2,1),\n",
    "# ('RM','RM','TOR1','I'):(3,0),('RM','YNB','TOR1','I'):(3,1),\n",
    "# ('RM','RM','TOR1','S'):(4,0),('RM','YNB','TOR1','S'):(4,1),\n",
    "# ('RM','RM','FPR1','D'):(5,0),('RM','YNB','FPR1','D'):(5,1)\n",
    "# }\n",
    "\n",
    "# dpi=300\n",
    "# nrows=6\n",
    "# ncols=2\n",
    "\n",
    "# param='Generation Time'\n",
    "\n",
    "# height, width = np.array([nrows*3, ncols*6], dtype=float)\n",
    "\n",
    "# fig, axes = plt.subplots(nrows,ncols,figsize=(width, height),\n",
    "#                          sharex='col',sharey='row')\n",
    "    \n",
    "# for ii, (env_evo,env_test,gene,tetrad) in enumerate([('RM','RM','','A'),('RM','YNB','','A'),\n",
    "#                                                      ('RM','RM','','B'),('RM','YNB','','B'),\n",
    "#                                                      ('RM','RM','TOR1','A'),('RM','YNB','TOR1','A'),\n",
    "#                                                      ('RM','RM','TOR1','I'),('RM','YNB','TOR1','I'),\n",
    "#                                                      ('RM','RM','TOR1','S'),('RM','YNB','TOR1','S'),\n",
    "#                                                      ('RM','RM','FPR1','D'),('RM','YNB','FPR1','D')]):\n",
    "                    \n",
    "#     gs = spores_norm_df[(spores_norm_df.gene==gene)&\\\n",
    "#                         (spores_norm_df.tetrad==tetrad)].groupby(['env_evo','env_test']).get_group((env_evo,env_test))\n",
    "       \n",
    "#     ax = axes[dict_panels[(env_evo,env_test,gene,tetrad)]]\n",
    "    \n",
    "#     gs.boxplot(ax=ax, column=param, by=['population','genotype','mating','auxotrophy','tetrad','spore'], vert=False)\n",
    "    \n",
    "#     if ii<2:\n",
    "#         # set title\n",
    "#         ax.set_title(environments[env_test])\n",
    "#     else:\n",
    "#         ax.set_title('')\n",
    "\n",
    "# #     ax.annotate('%s %s (spores)'%(gene, tetrad),\n",
    "# #                 xy=(.8, .2), xycoords='axes fraction', fontsize=12,\n",
    "# #                 horizontalalignment='center', verticalalignment='right')\n",
    "\n",
    "# # set axes labels\n",
    "# for ax in [axes[(5,0)],axes[(5,1)]]:\n",
    "#     ax.set_xlabel(r'Rel. doubling time $\\langle\\tau\\rangle_{b,g,k}^{a/\\alpha}$', fontsize=10)\n",
    "\n",
    "# plt.suptitle('')\n",
    "\n",
    "# fig.tight_layout()\n",
    "\n",
    "# for fmt in ['pdf','png','svg']:\n",
    "#     fig.savefig(dir_paper+'figures/figure5/cross_label_errors_spores_boxplot.'+fmt, dpi=300)\n",
    "    \n",
    "# plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# dict_panels = {\n",
    "# ('RM','RM','','A'):(0,0),('RM','YNB','','A'):(0,1),\n",
    "# ('RM','RM','','B'):(1,0),('RM','YNB','','B'):(1,1),\n",
    "# ('RM','RM','TOR1','A'):(2,0),('RM','YNB','TOR1','A'):(2,1),\n",
    "# ('RM','RM','TOR1','I'):(3,0),('RM','YNB','TOR1','I'):(3,1),\n",
    "# ('RM','RM','TOR1','S'):(4,0),('RM','YNB','TOR1','S'):(4,1),\n",
    "# ('RM','RM','FPR1','D'):(5,0),('RM','YNB','FPR1','D'):(5,1)\n",
    "# }\n",
    "\n",
    "# dpi=300\n",
    "# nrows=6\n",
    "# ncols=2\n",
    "\n",
    "# param='Generation Time'\n",
    "\n",
    "# height, width = np.array([nrows*3, ncols*6], dtype=float)\n",
    "\n",
    "# fig, axes = plt.subplots(nrows,ncols,figsize=(width, height),\n",
    "#                          sharex='col',sharey='row')\n",
    "    \n",
    "# for ii, (env_evo,env_test,gene,tetrad) in enumerate([('RM','RM','','A'),('RM','YNB','','A'),\n",
    "#                                                      ('RM','RM','','B'),('RM','YNB','','B'),\n",
    "#                                                      ('RM','RM','TOR1','A'),('RM','YNB','TOR1','A'),\n",
    "#                                                      ('RM','RM','TOR1','I'),('RM','YNB','TOR1','I'),\n",
    "#                                                      ('RM','RM','TOR1','S'),('RM','YNB','TOR1','S'),\n",
    "#                                                      ('RM','RM','FPR1','D'),('RM','YNB','FPR1','D')]):\n",
    "                    \n",
    "#     gh = hybrids_norm_df[(hybrids_norm_df.gene==(gene,gene))&\\\n",
    "#                          (hybrids_norm_df.tetrad==(tetrad,tetrad))].groupby(['env_evo','env_test']).get_group((env_evo,env_test))\n",
    "       \n",
    "#     ax = axes[dict_panels[(env_evo,env_test,gene,tetrad)]]\n",
    "    \n",
    "#     gh.boxplot(ax=ax, column=param, by=['population','genotype','tetrad','spore'], vert=False)\n",
    "\n",
    "#     if ii<2:\n",
    "#         # set title\n",
    "#         ax.set_title(environments[env_test])\n",
    "#     else:\n",
    "#         ax.set_title('')\n",
    "        \n",
    "# #     ax.annotate('%s %s (hybrids)'%(gene, tetrad),\n",
    "# #                 xy=(.8, .2), xycoords='axes fraction', fontsize=12,\n",
    "# #                 horizontalalignment='center', verticalalignment='right')\n",
    "\n",
    "\n",
    "# # set axes labels\n",
    "# for ax in [axes[(5,0)],axes[(5,1)]]:\n",
    "#     ax.set_xlabel(r'Rel. doubling time $\\langle\\tau\\rangle_{b,g,k}^{a/\\alpha}$', fontsize=10)\n",
    "\n",
    "# plt.suptitle('')\n",
    "\n",
    "# fig.tight_layout()\n",
    "\n",
    "# for fmt in ['pdf','png','svg']:\n",
    "#     fig.savefig(dir_paper+'figures/figure5/cross_label_errors_hybrids_boxplot.'+fmt, dpi=300)\n",
    "\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "spores_abs_df = utils.load_data(dir_data+'pheno_spores10K_abs.pickle')\n",
    "spores_norm_df = utils.load_data(dir_data+'pheno_spores10K_norm.pickle')\n",
    "hybrids_abs_df = utils.load_data(dir_data+'pheno_hybrids10K_abs.pickle')\n",
    "hybrids_norm_df = utils.load_data(dir_data+'pheno_hybrids10K_norm.pickle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "dict_population = {'parents':0, 'ancestral':1, 'evolved':2}\n",
    "dict_background = {'WA':0, 'NA':1, 'WAxNA':2}\n",
    "dict_gene = {'RNR2':0, 'RNR4':1, 'no driver':2, 'FPR1':3, 'TOR1':4}\n",
    "dict_genotype = {'RNR2':0, 'RNR2*':2, 'RNR4':1, 'RNR4*':3,\n",
    "                 'no driver':0, 'FPR1*':2, 'TOR1':1, 'TOR1*':3}\n",
    "dict_panels = {\n",
    "'WAxNA': {'HU':(0,0),'RM':(0,1)},\n",
    "'WAxNA-WExSA': {'HU':(1,0),'RM':(1,1)}\n",
    "}\n",
    "\n",
    "# apply sorting ranks to reorder rows    \n",
    "hybrids_norm_df.loc[:,u'rank_background_MATa'] = hybrids_norm_df[u'background_MATa'].map(dict_background)\n",
    "hybrids_norm_df.loc[:,u'rank_background_MATα'] = hybrids_norm_df[u'background_MATα'].map(dict_background)\n",
    "hybrids_norm_df.loc[:,u'rank_genotype_MATa'] = hybrids_norm_df[u'genotype_MATa'].map(dict_genotype)\n",
    "hybrids_norm_df.loc[:,u'rank_genotype_MATα'] = hybrids_norm_df[u'genotype_MATα'].map(dict_genotype)\n",
    "hybrids_norm_df.loc[:,u'rank_population_MATa'] = hybrids_norm_df[u'population_MATa'].map(dict_population)\n",
    "hybrids_norm_df.loc[:,u'rank_population_MATα'] = hybrids_norm_df[u'population_MATα'].map(dict_population)\n",
    "hybrids_norm_df.sort(columns=[u'rank_population_MATa',u'rank_population_MATα',\n",
    "                              u'rank_background_MATa',u'rank_background_MATα',\n",
    "                              u'rank_genotype_MATa',u'rank_genotype_MATα',\n",
    "                              u'tetrad_MATa',u'tetrad_MATα',\n",
    "                              u'spore_MATa',u'spore_MATα'],\n",
    "                ascending=[True,True,True,True,True,True,True,True,True,True],inplace=True)\n",
    "\n",
    "# environment attributes\n",
    "environments = {u'HU': u'YNB+HU 10 mg/ml', \n",
    "                u'RM': u'YNB+RM 0.025 μg/ml', \n",
    "                u'YNB': u'YNB (control)'}\n",
    "\n",
    "param='growth_rate'\n",
    "\n",
    "fig = plt.figure(figsize=(13,13))\n",
    "\n",
    "for (background,env_test) in [('WAxNA','HU'),('WAxNA','RM'),('WAxNA-WExSA','HU'),('WAxNA-WExSA','RM')]:\n",
    "    ax = plt.subplot2grid((2, 2), dict_panels[background][env_test])\n",
    "    \n",
    "    # spores\n",
    "    gs = spores_norm_df.groupby(['background','Environment']).get_group((background,env_test))\n",
    "    S = gs.groupby([u'mating',u'spore'], sort=False).agg(np.mean)[param]\n",
    "\n",
    "    # hybrids\n",
    "    gh = hybrids_norm_df.groupby(['background','Environment']).get_group(((background,background),env_test))\n",
    "    gh = gh.groupby([u'spore_MATa',\n",
    "                     u'spore_MATα'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    H = gh.unstack(level=[u'spore_MATα'])\n",
    "\n",
    "    # plot\n",
    "    pheno.heatmap_spores(S, radius=0.5)\n",
    "    pheno.heatmap_hybrids(H, fold=False, title=environments[env_test], pad=100)\n",
    "\n",
    "# fig.subplots_adjust(left=0.2,bottom=0,right=0.8,top=1,hspace=0.01,wspace=0.01)\n",
    "fig.subplots_adjust(left=0.1,bottom=0.1,right=0.9,top=0.95,hspace=0.05,wspace=0.60)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
