{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Supplementary Materials\n",
    "\n",
    "\"Background-dependent selection of genetic variation in heterogeneous populations\"\n",
    "\n",
    "Ignacio Vázquez-García, Francisco Salinas, Jing Li, Andrej Fischer, Benjamin Barré, Johan Hallin, Anders Bergström, Elisa Alonso-Pérez, Jonas Warringer, Ville Mustonen, Gianni Liti\n",
    "\n",
    "## Figure 3 - Loss of heterozygosity (+ Supp. Figs.)\n",
    "\n",
    "This IPython notebook is provided for reproduction of Figure 3 of the paper. It can be viewed by copying its URL to nbviewer and it can be run by opening it in binder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# load external dependencies\n",
    "import matplotlib.gridspec as gridspec\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import matplotlib.lines as lines\n",
    "import matplotlib.patches as patches\n",
    "from matplotlib.collections import PatchCollection\n",
    "%matplotlib inline\n",
    "\n",
    "import json, matplotlib\n",
    "s = json.load( open(\"matplotlibrc.json\") )\n",
    "matplotlib.rcParams.update(s)\n",
    "\n",
    "import glob, os, re\n",
    "import numpy as np\n",
    "import pandas as pd # (*) Pandas for data manipulation\n",
    "from scipy import stats\n",
    "\n",
    "# load internal dependencies\n",
    "import colors,plot,model,utils,config\n",
    "from utils import sgd_length\n",
    "from paths import * # absolute file path for data directory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data import"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Length distribution of homozygosity tracts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HU</th>\n",
       "      <th>RM</th>\n",
       "      <th>YPD</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50kb_bin_center</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>0.732980</td>\n",
       "      <td>0.748970</td>\n",
       "      <td>0.750893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>0.139939</td>\n",
       "      <td>0.152117</td>\n",
       "      <td>0.159386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>0.052950</td>\n",
       "      <td>0.055077</td>\n",
       "      <td>0.052793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>0.033283</td>\n",
       "      <td>0.020607</td>\n",
       "      <td>0.021708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>0.012859</td>\n",
       "      <td>0.010116</td>\n",
       "      <td>0.008484</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       HU        RM       YPD\n",
       "50kb_bin_center                              \n",
       "25               0.732980  0.748970  0.750893\n",
       "75               0.139939  0.152117  0.159386\n",
       "125              0.052950  0.055077  0.052793\n",
       "175              0.033283  0.020607  0.021708\n",
       "225              0.012859  0.010116  0.008484"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load data\n",
    "loh_length_df = pd.read_csv(dir_data+'seq/loh/homozygosity_length.tsv', sep='\\t')\n",
    "\n",
    "loh_length_df = loh_length_df.set_index(\"50kb_bin_center\")\n",
    "\n",
    "loh_length_df = loh_length_df.reindex(columns=['HU','RM','YPD'])\n",
    "\n",
    "loh_length_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Examples of loss-of-heterozygosity in clonal genome sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>selection</th>\n",
       "      <th>population</th>\n",
       "      <th>set</th>\n",
       "      <th>clone</th>\n",
       "      <th>lineage</th>\n",
       "      <th>type</th>\n",
       "      <th>chr_arabic</th>\n",
       "      <th>chr_roman</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>pos</th>\n",
       "      <th>genotype</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HU</td>\n",
       "      <td>WAxNA_F12_1_HU_1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>subclone A</td>\n",
       "      <td>consensus</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>2718</td>\n",
       "      <td>32682</td>\n",
       "      <td>48</td>\n",
       "      <td>2718</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16528</th>\n",
       "      <td>RM</td>\n",
       "      <td>WAxNA_F12_1_RM_3</td>\n",
       "      <td>14</td>\n",
       "      <td>C3</td>\n",
       "      <td>subclone A</td>\n",
       "      <td>loh</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>2718</td>\n",
       "      <td>88212</td>\n",
       "      <td>327</td>\n",
       "      <td>2718</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16809</th>\n",
       "      <td>HU</td>\n",
       "      <td>WAxNA_F12_2_HU_1</td>\n",
       "      <td>1</td>\n",
       "      <td>C1</td>\n",
       "      <td>subclone A</td>\n",
       "      <td>loh</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>2718</td>\n",
       "      <td>32682</td>\n",
       "      <td>48</td>\n",
       "      <td>2718</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17098</th>\n",
       "      <td>HU</td>\n",
       "      <td>WAxNA_F12_2_HU_1</td>\n",
       "      <td>2</td>\n",
       "      <td>C2</td>\n",
       "      <td>bulk</td>\n",
       "      <td>loh</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>2718</td>\n",
       "      <td>56480</td>\n",
       "      <td>195</td>\n",
       "      <td>2718</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17401</th>\n",
       "      <td>HU</td>\n",
       "      <td>WAxNA_F12_2_HU_2</td>\n",
       "      <td>4</td>\n",
       "      <td>C1</td>\n",
       "      <td>bulk</td>\n",
       "      <td>loh</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>2718</td>\n",
       "      <td>58539</td>\n",
       "      <td>202</td>\n",
       "      <td>2718</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      selection        population  set clone     lineage       type  \\\n",
       "0            HU  WAxNA_F12_1_HU_1    1        subclone A  consensus   \n",
       "16528        RM  WAxNA_F12_1_RM_3   14    C3  subclone A        loh   \n",
       "16809        HU  WAxNA_F12_2_HU_1    1    C1  subclone A        loh   \n",
       "17098        HU  WAxNA_F12_2_HU_1    2    C2        bulk        loh   \n",
       "17401        HU  WAxNA_F12_2_HU_2    4    C1        bulk        loh   \n",
       "\n",
       "       chr_arabic chr_roman  start    end  length   pos  genotype gene  \n",
       "0               1         I   2718  32682      48  2718         2       \n",
       "16528           1         I   2718  88212     327  2718         0       \n",
       "16809           1         I   2718  32682      48  2718         0       \n",
       "17098           1         I   2718  56480     195  2718         0       \n",
       "17401           1         I   2718  58539     202  2718         0       "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "types = ('*consensus*.txt','*loh*.txt') # the tuple of file types'*consensus*.txt',*gt*.txt',\n",
    "files = []\n",
    "for t in types:\n",
    "    files.extend(glob.glob(dir_data + 'seq/loh/'+t))\n",
    "\n",
    "seq_loh_df = pd.DataFrame()\n",
    "\n",
    "for file_ in files:\n",
    "    \n",
    "    df = pd.read_csv(file_, index_col=None, names=['chr_arabic','start','length','end','genotype'], sep='\\t')\n",
    "    \n",
    "    f = re.split(r'[.-]+', os.path.basename(file_))\n",
    "\n",
    "    df['background'] = f[0].split('_')[0]\n",
    "    df['cross'] = f[0].split('_')[1]\n",
    "    df['cross_rep'] = f[0].split('_')[2]\n",
    "    df['selection'] = f[0].split('_')[3]\n",
    "    df['selection_rep'] = f[0].split('_')[4]\n",
    "    \n",
    "    df['type'] = f[1]\n",
    "    df['set'] = int(f[2].replace('set',''))#int(f[2].replace('set',''))\n",
    "\n",
    "    if f[1]=='consensus':\n",
    "        df['population'] = '_'.join(f[0].split('_')[:-1])\n",
    "        df['clone'] = \"\"\n",
    "    else:\n",
    "        df['population'] = '_'.join(f[0].split('_')[:-2])\n",
    "        df['clone'] = f[0].split('_')[6]\n",
    "    df['lineage'] = f[3]\n",
    "\n",
    "#     if f[1]!='gt':\n",
    "#         df['set'] = int(f[2].replace('set',''))\n",
    "#     else:\n",
    "#         df['set'] = \"\"\n",
    "    \n",
    "    seq_loh_df = seq_loh_df.append(df, ignore_index=True)\n",
    "\n",
    "seq_loh_df = utils.chr_to_gw(seq_loh_df)\n",
    "\n",
    "start = seq_loh_df[['selection','population','set','clone','lineage','type','chr_arabic','chr_roman','start','end','length','pos_start','genotype']]\\\n",
    ".rename(columns={'pos_start':'pos'})\n",
    "end = seq_loh_df[['selection','population','set','clone','lineage','type','chr_arabic','chr_roman','start','end','length','pos_end','genotype']]\\\n",
    ".rename(columns={'pos_end':'pos'})\n",
    "\n",
    "seq_loh_df = pd.concat([start, end]).sort_values('pos')\n",
    "\n",
    "# seq_loh_df = seq_loh_df[seq_loh_df.length > 10]\n",
    "\n",
    "seq_loh_df['lineage'] = seq_loh_df['lineage'].replace('A','subclone A')\n",
    "seq_loh_df['lineage'] = seq_loh_df['lineage'].replace('B','subclone B')\n",
    "seq_loh_df['lineage'] = seq_loh_df['lineage'].replace('C','subclone C')\n",
    "seq_loh_df['lineage'] = seq_loh_df['lineage'].replace('D','subclone D')\n",
    "\n",
    "# dummy\n",
    "seq_loh_df['gene'] = ''\n",
    "\n",
    "seq_loh_df['genotype'] = seq_loh_df['genotype'].astype(int)\n",
    "\n",
    "seq_loh_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ac</th>\n",
       "      <th>af</th>\n",
       "      <th>alt</th>\n",
       "      <th>an</th>\n",
       "      <th>allele</th>\n",
       "      <th>amino_acids</th>\n",
       "      <th>biotype</th>\n",
       "      <th>baseqranksum</th>\n",
       "      <th>cds_position</th>\n",
       "      <th>chr_arabic</th>\n",
       "      <th>...</th>\n",
       "      <th>population</th>\n",
       "      <th>coverage</th>\n",
       "      <th>missing</th>\n",
       "      <th>substitution</th>\n",
       "      <th>assignment</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>consequence_short</th>\n",
       "      <th>chr_roman</th>\n",
       "      <th>pos</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>[A]</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>R/I</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.31</td>\n",
       "      <td>101</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>WAxNA_F12_1_HU_1</td>\n",
       "      <td>28.795032</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>bulk</td>\n",
       "      <td>driver</td>\n",
       "      <td>non-synonymous</td>\n",
       "      <td>VII</td>\n",
       "      <td>4595191</td>\n",
       "      <td>de novo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[C]</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>R/G</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>None</td>\n",
       "      <td>100</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>WAxNA_F12_1_HU_1</td>\n",
       "      <td>27.116307</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>bulk</td>\n",
       "      <td>driver</td>\n",
       "      <td>non-synonymous</td>\n",
       "      <td>VII</td>\n",
       "      <td>4595192</td>\n",
       "      <td>de novo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>[A]</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>R/I</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0.3</td>\n",
       "      <td>101</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>WAxNA_F12_1_HU_2</td>\n",
       "      <td>28.993576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>bulk</td>\n",
       "      <td>driver</td>\n",
       "      <td>non-synonymous</td>\n",
       "      <td>VII</td>\n",
       "      <td>4595191</td>\n",
       "      <td>de novo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[A]</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>R/I</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>None</td>\n",
       "      <td>101</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>WAxNA_F12_1_HU_2</td>\n",
       "      <td>28.719772</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>bulk</td>\n",
       "      <td>driver</td>\n",
       "      <td>non-synonymous</td>\n",
       "      <td>VII</td>\n",
       "      <td>4595191</td>\n",
       "      <td>de novo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>[C]</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>Y/H</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0.289</td>\n",
       "      <td>505</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>WAxNA_F12_1_HU_3</td>\n",
       "      <td>31.647430</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>bulk</td>\n",
       "      <td>driver</td>\n",
       "      <td>non-synonymous</td>\n",
       "      <td>X</td>\n",
       "      <td>6225369</td>\n",
       "      <td>de novo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 80 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  ac   af  alt an allele amino_acids         biotype baseqranksum  \\\n",
       "0  1  0.5  [A]  2      A         R/I  protein_coding         1.31   \n",
       "1  2  1.0  [C]  2      C         R/G  protein_coding         None   \n",
       "2  1  0.5  [A]  2      A         R/I  protein_coding          0.3   \n",
       "3  2  1.0  [A]  2      A         R/I  protein_coding         None   \n",
       "4  1  0.5  [C]  2      C         Y/H  protein_coding        0.289   \n",
       "\n",
       "  cds_position  chr_arabic   ...           population   coverage missing  \\\n",
       "0          101           7   ...     WAxNA_F12_1_HU_1  28.795032     NaN   \n",
       "1          100           7   ...     WAxNA_F12_1_HU_1  27.116307     NaN   \n",
       "2          101           7   ...     WAxNA_F12_1_HU_2  28.993576     NaN   \n",
       "3          101           7   ...     WAxNA_F12_1_HU_2  28.719772     NaN   \n",
       "4          505          10   ...     WAxNA_F12_1_HU_3  31.647430     NaN   \n",
       "\n",
       "   substitution assignment mutation_type consequence_short chr_roman      pos  \\\n",
       "0           C>A       bulk        driver    non-synonymous       VII  4595191   \n",
       "1           T>C       bulk        driver    non-synonymous       VII  4595192   \n",
       "2           C>A       bulk        driver    non-synonymous       VII  4595191   \n",
       "3           C>A       bulk        driver    non-synonymous       VII  4595191   \n",
       "4           T>C       bulk        driver    non-synonymous         X  6225369   \n",
       "\n",
       "      type  \n",
       "0  de novo  \n",
       "1  de novo  \n",
       "2  de novo  \n",
       "3  de novo  \n",
       "4  de novo  \n",
       "\n",
       "[5 rows x 80 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load data\n",
    "seq_mut_df = utils.load_data(dir_data+'seq/de-novo/seq_dn_df_consensus.pickle')\n",
    "\n",
    "seq_mut_df = seq_mut_df.rename(columns={'chrom':'chr_arabic', 'pos':'site'})\n",
    "seq_mut_df['chr_roman'] = seq_mut_df['chr_arabic'].apply(utils.int_to_roman)\n",
    "seq_mut_df = utils.chr_to_gw(seq_mut_df)\n",
    "\n",
    "seq_mut_df = seq_mut_df[(seq_mut_df.time=='32')&\\\n",
    "                        (seq_mut_df.clone!='')].reset_index(drop=True)\n",
    "\n",
    "# Annotate zygosity\n",
    "seq_mut_df['genotype'] = seq_mut_df['genotype'].map({'0/0': 0, '0/1': 1, '1/0': 1, '1/1': 2})\n",
    "\n",
    "seq_mut_df['type'] = 'de novo'\n",
    "\n",
    "seq_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "columns=['selection','population','clone']\n",
    "seq_mut_fixed_df = pd.merge(seq_mut_df, \n",
    "                            seq_loh_df[['selection','population','clone','set','lineage']].drop_duplicates(columns), \n",
    "                            on=columns, how='left')\n",
    "seq_loh_fixed_df = pd.merge(seq_loh_df, \n",
    "                            seq_mut_df[['selection','population','clone']].drop_duplicates(columns), \n",
    "                            on=columns, how='left')\n",
    "\n",
    "seq_dn_df = pd.concat([seq_loh_fixed_df, seq_mut_fixed_df], axis=0).reset_index(drop=True)\n",
    "\n",
    "seq_dn_df['population'] = seq_dn_df['population']+'_T32'\n",
    "\n",
    "seq_dn_df = pd.pivot_table(seq_dn_df, \n",
    "                           index=['selection','population','set','clone','lineage','type'], \n",
    "                           columns=['chr_arabic','chr_roman','pos','gene'])['genotype']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fluctuation assay"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"6\" halign=\"left\">biological_rep</th>\n",
       "      <th colspan=\"4\" halign=\"left\">technical_rep</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"4\" halign=\"left\">Total 5-FOA</th>\n",
       "      <th colspan=\"6\" halign=\"left\">LOH rate</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"3\" halign=\"left\">mean</th>\n",
       "      <th colspan=\"3\" halign=\"left\">sem</th>\n",
       "      <th colspan=\"3\" halign=\"left\">mean</th>\n",
       "      <th>sem</th>\n",
       "      <th>...</th>\n",
       "      <th>mean</th>\n",
       "      <th colspan=\"3\" halign=\"left\">sem</th>\n",
       "      <th colspan=\"3\" halign=\"left\">mean</th>\n",
       "      <th colspan=\"3\" halign=\"left\">sem</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>background</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>...</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "      <th>WA/WA</th>\n",
       "      <th>NA/NA</th>\n",
       "      <th>WA/NA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>environment</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HU</th>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0.188982</td>\n",
       "      <td>...</td>\n",
       "      <td>237500</td>\n",
       "      <td>37796.447301</td>\n",
       "      <td>459910.705619</td>\n",
       "      <td>86473.571520</td>\n",
       "      <td>0.016617</td>\n",
       "      <td>0.002876</td>\n",
       "      <td>0.004080</td>\n",
       "      <td>0.004539</td>\n",
       "      <td>0.000354</td>\n",
       "      <td>0.000876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RM</th>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0.188982</td>\n",
       "      <td>...</td>\n",
       "      <td>3662500</td>\n",
       "      <td>329628.782009</td>\n",
       "      <td>434293.093511</td>\n",
       "      <td>302924.731340</td>\n",
       "      <td>0.000352</td>\n",
       "      <td>0.000990</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.000021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YPD</th>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>0.422577</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0.188982</td>\n",
       "      <td>...</td>\n",
       "      <td>4237500</td>\n",
       "      <td>957396.021956</td>\n",
       "      <td>316712.676795</td>\n",
       "      <td>342228.776865</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>0.000009</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            biological_rep                                            \\\n",
       "                      mean                   sem                       \n",
       "background           WA/WA NA/NA WA/NA     WA/WA     NA/NA     WA/NA   \n",
       "environment                                                            \n",
       "HU                     2.5   2.5   2.5  0.422577  0.422577  0.422577   \n",
       "RM                     2.5   2.5   2.5  0.422577  0.422577  0.422577   \n",
       "YPD                    2.5   2.5   2.5  0.422577  0.422577  0.422577   \n",
       "\n",
       "            technical_rep                          ...    Total 5-FOA  \\\n",
       "                     mean                   sem    ...           mean   \n",
       "background          WA/WA NA/NA WA/NA     WA/WA    ...          WA/NA   \n",
       "environment                                        ...                  \n",
       "HU                    1.5   1.5   1.5  0.188982    ...         237500   \n",
       "RM                    1.5   1.5   1.5  0.188982    ...        3662500   \n",
       "YPD                   1.5   1.5   1.5  0.188982    ...        4237500   \n",
       "\n",
       "                                                          LOH rate            \\\n",
       "                       sem                                    mean             \n",
       "background           WA/WA          NA/NA          WA/NA     WA/WA     NA/NA   \n",
       "environment                                                                    \n",
       "HU            37796.447301  459910.705619   86473.571520  0.016617  0.002876   \n",
       "RM           329628.782009  434293.093511  302924.731340  0.000352  0.000990   \n",
       "YPD          957396.021956  316712.676795  342228.776865  0.000090  0.000209   \n",
       "\n",
       "                                                     \n",
       "                            sem                      \n",
       "background      WA/NA     WA/WA     NA/NA     WA/NA  \n",
       "environment                                          \n",
       "HU           0.004080  0.004539  0.000354  0.000876  \n",
       "RM           0.000044  0.000181  0.000114  0.000021  \n",
       "YPD          0.000038  0.000022  0.000037  0.000009  \n",
       "\n",
       "[3 rows x 42 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read csv file containing the competition assay data\n",
    "# loh_fluctuation_df = pd.ExcelFile(dir_data+'fluctuation/LOH_assay_raw_data_IVG_new.xlsx')\n",
    "# loh_fluctuation_df = loh_fluctuation_df.parse('New', keep_default_na=False, na_values='NaN')\n",
    "loh_fluctuation_df = pd.read_csv(dir_data+'fluctuation/fluctuation_assay.csv')\n",
    "\n",
    "loh_fluctuation_df['5-FOA+'] = loh_fluctuation_df['5FOA']*200/10\n",
    "loh_fluctuation_df['Total 5-FOA'] = loh_fluctuation_df['YPD']*1000000*2/40\n",
    "\n",
    "loh_fluctuation_df['LOH rate'] = loh_fluctuation_df['5-FOA+']/loh_fluctuation_df['Total 5-FOA']\n",
    "\n",
    "loh_fluctuation_df = loh_fluctuation_df.groupby(['background','environment'])\\\n",
    ".agg([np.mean, stats.sem])\n",
    "\n",
    "loh_fluctuation_df = loh_fluctuation_df.ix[['WA/WA','NA/NA','WA/NA']]\n",
    "loh_fluctuation_df = loh_fluctuation_df.unstack('background')\n",
    "\n",
    "loh_fluctuation_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAUwAAAEWCAYAAAANe67OAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJztnXeYVOXVwH9nl2VpK0iRXhQBEUGJ5bMgECPBSFBjjAQN\nmqiogGIEBTUiggUF0UAENYY10RiCJYlgF6UoNkBERQQU6UjZpbos7ML5/njvwLBsuTM75c7u+T3P\nPDNzy3nPvfedM2857zmiqhiGYRhlk5ZsBQzDMFIFM5iGYRg+MYNpGIbhEzOYhmEYPjGDaRiG4RMz\nmIZhGD6J2GCKSLaIHBCRc+KhkFHxEZFuXh0KvVaKyPnJ1sswyiIigykiVYCLgc3Ar+OikVFZUOBY\noCkwDxiTXHUMo2wibWGGWgEjgUsjLUxErhCRwZGeZ1RYtgE/AFuBfZGcKCKnisgtcdHKMEqgSoTH\n/xp4BXgZmCQip6vq/KIHiUgrYDzQD9cibeTt2gIsFZGjVXVbtEobFQIBVuPqYA2g/2E7S65DXwAL\nVHWhiHRIlLKGARG0MEUkHbgEuBJYi6vwl5Vw+E5gvXfMiar6GNDEK6+NGUsD1yXvAnQEHgDGi4iE\n7S+2Dqnqu6q6I+HaGgaRtTDPA2oCpwD7gd/hjOfwogeqaq6IbPW+1vDeD6jqs+XQ1ahYCJDnvQDS\nvW0KJdehcAFWn4xEE8kY5mXAp6q6VFWXA68BrUTklFLOUWCxiAwHVpVVgIjovffeq7Nnz1bvXHsF\n6xVLFPgW2AD8AeivqgdKOM53HQKrRynySkkkSNGKRERnzZpF9+7dk62KUTxS9iHJx+pRSpASdako\ngXNcX5O7iZ/f8DuWrFqRbFUCy/jx4wGYM2cOCxcuPGzfww8/nAyVAocZy8hYtXU389dsP/hatXX3\nYftHjBjBnj176NWrFwcOHGD48OF06dLl4P5nn32Wjz76iPvvv5/33nuPsWPH8uOPP/LnP/+ZZ5+t\nOCMnkc6Sx52//vffzPvn/7gVePupfyZbnaSRV5jHotxFdK7bmRpVahy2b+vWrTz88MO899579OjR\ng3fffZdjjjmGnTt3sn37dgYMGEDfvn154403EBG6d+/O6tWrKSgoYMeOHbRr147c3FwuvfRSnnnm\nGQoLC2nevDlr166loKCAK6+8kn/9619s3bqVRx55hPT09CTdheiZPXu2Gc0I2JJXyIML8g5+v+u0\nGrQK21+jRg0+/PBD2rVrxyeffEKHDh1o27Yt8+fPJysrixNPPJFPPvmErKwsli5dyr59+6hZsyYi\nwurVqxN+PfEicC3Mp+4bS4/rr+SxO0cmW5Wksih3ETM3vsOi3EVH7GvQoAHDhw/nrrvuAiAjI4ON\nGzcyePBgqlevTuvWralevTo9evTgnHPOoVmzZmRnZ3P11VdTtWpVvv/+e6pWrYqqUlBQQHp6Oldc\ncQULFy6kYcOGrF69mmrVqrFjxw5+/PHHRF+6EUBOOOEEpk+fzrXXXsvjjz/OvHnz2L59O2+++SYL\nFy7k1FNPJScnh1q1agGQmZnJ119/zbJly/jqq6/YuXNnkq8gNgTOYG5ZtZ63n/onHVq1SbYqSaVz\n3c6c37gHnet2LvW4AwcOcPbZZ9O0aVMeffRRCgsLSUtLo3379rzxxht8+OGHrFy5kqFDh/L0009z\nxhlnkJ6eTrt27Zg8eTJr164lLc1Vg44dO7Ju3ToaNmzI5s2bqV27NmvXrk3E5cYca13Gli5durB1\n61Y6dOjA1q1b6dSpE0OHDqWgoID9+/cjIuTm5nLWWWeRkZFBmzZteOutt5g4cSIPPvgg06dPT/Yl\nxITATfoESZ+gk5+fz5gxYxg1apTvcyZMmMDll19O48aNoykyJQbqbdInclZt3c2WvMKD3xvUqEKr\n+rXiWWRK1KWiBM5gWkUPNClRye2PNyVIibpUlMAZzCDpYxxBSlRyq0cpQUrUpaIEbgxz9uzZyVbB\nqABYPTLigbUwU5CHHnqItLQ0tmzZwi233EKzZs3KJW/lypVkZmbStGnTsg5NiVaB1aPI2bZsBfnr\nNx78Xq1pY45ud2ji9frrr+faa68lPz8fEWH8+PG88sor5OfnM2PGDN5//30mTpzI+PHjGTp0KJde\neimPPvoorVq1KqnIlKhLRQmcH6bh2L5nD9O//IKLOnaiTvXqh+2rWrUqQ4YMYenSpWRnZ7Nt2zZ6\n9erFW2+9RVpaGldeeSXvvPMOVatWZdeuXdSqVYsvvviCrKwsOnTowCWXXMLf/vY3RIRt27Zxxhln\n0LJlS0aPHk3t2rUZO3Zskq7aSBb56zey7KpBB7+3e3YShBnMdu3aMWPGDHr06MHHH3/M+eefz6ZN\nm/j8889RVdLT01m6dCkiwubNm7nggguYPn06gwdXrGiO1iUPKNO//IIJc2cx/csvjti3b98+JkyY\nwIgRI0hLS2PHjh3k5eVxySWX0Lp1a0SEzMxM1q9fT9WqVRk8eDDt27enffv29OzZk0mTJnHCCScg\nItx///1kZWUhIvTo0SPa2fPAYfUo9vzmN7/h5ZdfZubMmezYsYNXXnmFH374gUaNGnHFFVcwdepU\nVJUZM2aQk5PD4sWLk61yzAmcwbQZcsdFHTtxS9efclHHTkfsq1q1KrfccgtDhgzh1VdfpXbt2qxZ\ns4aXX36ZtWvXsnLlSqpXr8727dsP+liGqFu3LsuWLePCCy+ksLCQu+++m7w8t8JDRKgoXVmrR5FR\nrWlj2j076eCrWtPD/zhFhJNPPpnly5fz05/+lLvvvpt169YRisiXnp5Oz5492bBhA1u2bOHOO+/k\nnHPOYd26dcm4nLhhY5gVhLlz51KrVi1+8pOflHrct99+y3vvvcf1118fTTEpMe5k9SglSIm6VBQz\nmEYkpEQlN3/elCAl6lJRAtclt7EnIxaYsTTiQeAMplX0svnoo48YNWoUN998M1u2bDli/z/+8Q9y\nc3OP2B4KC2cYRVm1fC3zZy8++Fq1/PAYAmWFdxs9ejQXX3wxcKie3XrrrRWuAZQSbkUr163jrocf\n5sHhwzmunD6HqUJOTg7Z2dlcc8011KtX77B9c+fOZeTIkezcuZMhQ4ZwzDHH0K1bN+bOnUvVqlUp\nLCzkq6++Yvz48YwcOZLXX3+datWqHTEBZBghtmzIZcy1Tx78fueUG2nVtvnB72WFd/vFL37Bli1b\neOeddxARCgsLOf7445k3b16FagSlxC9o+EMPMu3xxxn+0IPJViVhZGdnM2zYMLKzs4/Yl5GRAcA3\n33zDT37yE84++2xWrlzJpZdeSlZWFm3atKFDhw506dKFWbNmcccdd1BYWHiEHMPwS1nh3Ro0aED3\n7t358MMPKSws5J133mHjxo188cWRbnGpTMINZjT5pPv070nnq39On/4946VW4LjmmmsYO3Ys11xz\nzRH7zjjjDO6++25eeOEFFi1axEcffUSbNoecjOvXr8/ixYtJT0+na9eujBkzhmrVqiVS/aTjtys4\nbdo0unbtyrRp0+KrUIpTUni3wsJC9u/fT1paGiLCVVddxWeffcbixYu5//77uf766/nggw+SrX7M\niPksuY980jtE5KriMv6VNEteWvRxI6GkxMxmpN4Wl112GS+99FIcNQo+q5avZcuGQ+PeDZrUPaxL\nHgdSoi4VJR5jmEXzSY8QkXFeXumoqFGlBuccc07MFDQM43BatW0ebwNZIYi5wSxvPul7772X7t27\nV6iBYiN4hAfM3ZlfyPw124GEBM41Uph4zpJHnE8aMGNpJITwpF9rdu0/+Llo8i/DCCcuBlNVR3sf\nI86vacbSMIygEji3oorm6GrEBxG5SkSGiUj1so82jNgQOINpGOC8LUTkZRGpISJ9ReRW7/UzEakN\nvA1kUGR8PFK2fDyDgp05bPl4Rkz0Nio2gTOYfrvk2/fs4dlPP2H7nj3xVchIFkW9LR4Dmqjqu6q6\nAxgBfE0563CDM3vTacRLNDizd7kVNio+gTOYfrvkf5/7LneNHsXf574bX4WMpKCquUCJ3haqOkhV\n/6uqxf5j3nvvvTa8Y8SclFhLXhxL5rzK+tfeYEmnFtDzl8lWx4gfUXlb3HvvvfHSx6jEBM5g+u2S\n33PLn6iakc4dA++Ir0JG0iiPt8Xs2bPN48KIOYEzmH5p3rA5k0ZNSrYahmFUIlJ2DNNIXUSkS9lH\nlQ9rXRrxIHAGszwVPa8wj3mb55FXmBc7hYyYICINRGSoiLwCHBfv8uyP14gHgTOY5WFR7iJmbnyH\nRbmLkq2KcSTzgMXAu6XFEogV1sI04kHgDGZ5Wgad63bm/MY96Fy3c+wUMmJFZ6AlcKaI/C7ehVkL\n04gHgZv0KU/LwMLABRdV/VFElgALgPR4l2ctTCMeBK6FWR5ycnIYN24cOTk5yVbFKIKI3Acci2tp\nnh/v8uLRwrTo7EaFMpiTnnqKYcOGMempp5KtinEk+ao6VVX/DlSNd2HxaGH26dOHY445hj59+sRc\ntpEaBM5glqdlUP/002ja6xfUP/202ClkxIqssM9HJU0LwygHFWoM84ou51Krdm0u6tgpdgoZseIV\nEfkzbqnjf+JdWCxX+oTnu9m5fTfzZy8GEpL3xggYgTOY5aFO9epcdcb/JVsNo3haAtmqmpC8q7Hs\nkofn7F6zd8PBz0VzdxsVn8B1yWOJObIHinbABclWwjDKQzLykicsUrY5sgeKtsAJInKhiFyYbGXK\norgZ8S2FayjQfLYUrkmiZkYyiXmXvKy85LhI2ddSzkjZfgg5sJsjeyD4q/euJCAndbRjmLW3rGfj\nt1/QtUFj/plRla4NGrMqz4XcbFClBQ2qtIixpkYq4auFKSLNIpCZkEjZfthXIHy3qgr7ClIyZ3yF\nQlXnqOoc4FzvPa5EPYa5aRPLrhrEsqsGseebFSy7ahAH9u6LqW5G6uLXaPURkYf9dKWCFCl7+pdf\nMGHuLKZ/Wfo8g411JpT3k62AH+boXrZzgDm6N9mqGAHCr8F8H9gLdBWRR32ek/S85F2aNKXN6vV0\nadL04LbijKONdcYfEbnE+8P9tYjckmx9yqKbZPKI1KGbZCZbFSNA+B3DbAM8CNQGqpV1cHkiZceS\nl6dOZdrjj3NqixbcfvvtwCHjCBxcd25jnQmhHbBPVW9JBYNpGMXh12CeDrwA3KGqt8ZRn5j6z11z\nzTWHvUPxxtGCdiSE9bhIRS2AuvEuzFJUGPEgkomXU4G4j37HMmhCvXr1uP3226lXr97BbSHjWKNK\njVLONGKNqv5TVW9S1TWqOjLe5ZmxNOKBX4P5AK5bPjGOuhiGYQQav13y+4ENuPBco8s4tlxYy6Bi\nIiLjcBOBAKjqsCSqYxhR4beFuQh4BOeQHlfiHSl7+549PPvpJ2zfU6xXkxEnVPV24H/AEiA3yeoY\nRlT4NZgNgaHeK67Eu4Xp1zfTiAs9gOZAuf+tROQ6EbGk9EZC8WswtwL7gY1x1AWIfwuzON9MI2HU\nxPnzlrlyTERaicjLIlJDRPqKyK3e62ciUht4HmgQb4UNIxy/BjMNWAvUj6MuQPxbmCHfzJenTo1r\nOUaxTASeBKb7OLasJbZ/ALYkIoiLYYTwO+lTC/gl8EEcdUkIxflmRkteYR6LchfRuW5nc1MqA2/S\npxMu1a5QxhJJVc0VkdKW2E6Oh55BY9q0aUyaNIlBgwZZaowA4LeF+QWudbAwjroA8e+SF+ebGS22\npNI/3qTPCCAP+GckpxLFEttYxiRIBtuWrWDje3Pp2qAxtb2oSRvfm8u2ZSuSrVqlxm8LczcucVUv\n4PX4qRMct6Lte/Yw/csvuKhjJ+pUL77XZ0sqI+aXOE+LW3EtzVIpzxLbWMYkSAb56zey7KpBzNG9\nbGQPf+13Hd0kk3bPToJ2bZKtXqUlki45QKWJRBCaTQdKTHthSyojphbQEReTIK6ksrEMp5tk0q3y\n/OwCj1+Duct7vy9eigSNUCI1S6gWUx4D/g/wG/EqamwtuREP/I5h3gP8FnheRMbGUZ/AjDuFEqqV\n1B03IsOb9PkbzmAOTrI6hhEVfg3ms6o6EHgl3kvaktEqyMnJYdy4ceTk5CS87ErED8A7wCPeBFBc\nsdalEQ/8GsyqIvIwCcjFkgyys7MZNmwY2dnZB7cVDTRcXODh4pZZbtyVw+j3prFxV8nGtzJGeFfV\n8cBTwHkiMiTe5QWlp2JULPyOYf6IC76xIY66JI3ifDOLBhouLvBwcRNDT8+fySuffQ/M5J7ziveb\nK05WZUBVd4lIbVWN+ximtTCNeODXYDYCcoDW5S1QRK4D6qvqQ+WVFStCvpnhFHUZKs6FqLiJof6n\nnw/M9N6Lp7K6I3njmJ1F5DiLVnQk4U7qXRs0TrY6RjH47ZI3AH6Fj6WRqbgGuLiuddFAw8VloCxu\nYqhxVj3uOa8PjbNKdoyvrEGMvbHLGYkwlqnYJe/Tpw/HHHOMregJMH4N5qvAdap6m49jU24NsJ8I\nRhblKGZME5HfxbsQ65Ib8aDMLrmInAU8AcwUkf1ltQ5ScQ2wH59L88ssHyJyCtAX57gemOEYw4gE\nPy3MnwPPq+rQCLtSKbMG2I/Ppflllpu/AvOAt1R1brwLS8UuuRF8/Ez6VAX2i0gNAFUt0xemMq8B\nLo5ERzUKYhQlVT1DRHoAJ4jIcFV9OJ7lpVIdWrV1N1vyCgHYmV/I/DXbqV+wP8laGcXhx2DuA9KB\n23Gtxrjm9KmIJNqNKIhuSyKSCdQD3gQ+TbI6gWJLXiEPLnDtkDW79vPggjwerqZlnGUkgzINpqqO\nSoQiQSMnJ4fs7GyuueaacoeCS7QbUUDdlh4AXsYFcBmAW24bN2wtuREP/PphJoygVPLQ6h/gCB/N\nSEl0VKOARlHKVdWPAESkS7wLC0o9MioWgTOYQWkZxDIyuwFAXREJTTLWTaomhhElfv0wE0YQjCXE\nNjK7AbhZ8rHea0qSdTGiYNq0aXTt2pVp06YlW5WkETiDae4gFZbzcNGublPVJclWxogcW4kUwC55\nUFqYRsxpCBxFGcnPYkVQhnYqAtuWrSB/vcuwvXfbdja+59xoqzVtzNGVLF2GtTB9EsuYmcWtXS9u\nWwXjDOAsERkoIgPjXZgZy9gRyi+07KpB7PlmxcHPISNamQicwQxqRS8uZma0FLcuvRKsVR8I/BGY\nA7yWZF0CyZaPZ1CwM4ctH89ItirFMkf3sp0DzNG9yVYlaQSuSx5UYjlrXty69Iq+Vl1VVwOIyD1h\nK8GMMBqc2ZsGZ/b2vn2TVF2KwxKyBbCFGdQueSxnzYtbl16J1qpPT7YChhEtgWthBrVLbpQPEbka\nKABOE5GLrJVppCKBa2EaFZbGQANVHYKLmVouRCRNROKe6sIwwglcC9OosHwFnCMiLfwcLCKtgPFA\nP+BiXJoUgC+AhUBXYHnMtTSMUghcCzOoY5hG+VDVV1X1TlVdo6p/9nFKaZH7twN1gNYiUjOOaldK\nbEVPyQTOYNoYpgEucj9QWuT+Z1X1dlX9sbjzkxGIuqJgK3pKxrrkRtCJOnK/YcQaM5hGYClP5H5b\nGhk5xUV+Byz6exgJN5heiK9HvNlSwzACQnGR3wGL/h5GzA2mzW4aQcBal0Y8iMekj81uGknHJnyi\nJ+hr2pNJzA2mzW4aQaCitzDj6frT4MzedBrxUti6diNEPN2KoprdTKU0u7EM+WbElor+p2uuP8kh\nLpM+5ZndTCVimSjNiC2p8qdrpBaBcytKpYpuidKCi7kVGfEgcCt9UqkrZYnSgosZy8RRmZZSBq6F\naRhGyWxLq8a62YsB2Ll9N/O9zw2a1KVV2+ZJ0alPnz68+OKLlWI8NXAG01oGRiyoqF3ybdvyGX/b\nMwCs2buBMdc+CcCdU25MmsGsTATOYFbUim4kFqtD8SeIrd14EziDaRXdMFKDytjaDdykj2EYRlAJ\nnMFMpVlyI7hYPUocWwrXUKD5bClck2xV4o51yY0KidWjxNGgSgsaVPGVeSTlCVwLM9Wx5ZKGUXEJ\nnMFM9a5UaLlkdnZ2slUxKjCVqRscJKxLHmNsuaSRCCpTNzhIBK6FmerYcslgkOo9FSOYmMFMAjbO\nGX9SvadiBJPAGczK0DKwcU7DSE1sDDMJ2DinYaQmgTOYlYHQOKdhGKlFwrvkItJLRHolulyjYmH1\nyEgGyUizWxvYJyKiqpbw2CgWq0dGEElGmt2dQAugWhzKPkgsJ49MVlIIRD0yjHCSkWb3VVV9VFX3\nxLrscIJqTCqDrFgQlHpkGOFIPHozInIP8AhwGdAY2KmqT/g4z7pWAUdVJVFlWT2q2CSyLsWKuBhM\nwzCMikjgHNcNwzCCihlMwzAMnwTCcV1E+uK5jXizoZGefzHQCWgF5OImBx4EbgQUWKqqMyKQdyHO\nlSUX96cSlSwRORv4P+CEGMi6GmgCHA9sjlZWmLvOH4GBxckB8oGfAJmqOtqPfhUBERkJjFPVPBF5\nEZilqpM9f89dqjo3gXrsB/bivAR6qGoPEWkNvKqq7ROkw/NAATAA2A1UB74C2uJ+YwJsVNWn461P\nUAiEwQTaq+o9IjIuyvM3qep9IrIf52rSFOgFqKqO9eT6NUxneB8vAzp48qKSBfwK2AR8CbxUTlkA\ndbzzTymHrJ3ABtwfwoQS5OzxnsdwETlaVbdFoGOqM0RECoCqSdRBgUc9w/0xsFZEjgUuAWYlSIcH\ncRNuAuSF6fMIrg6N976PT5A+gSAoBrOm936g1KNKZoP3IO/APeB0YA+Q4e3fH4GsC4AdwFHe+eWR\n1VBVbxeR7cB/yynrVFUdLCI7cM8tKlmqmisiW3GO38XJOcCh5yER6lgRCBmGbKCWt606rlWfKAQY\nKCJZuOfzMvAbT4/diVBAVQtE5G3cPTgBGCQi1YFncQ2BEJWqfgTFYC4WkeHAqijPHw18C2QCQ3Bd\nmfuAk0XkTmCmX0GhLqiIdMR1W/dFKwv4n4jcgavwQ8op61PvHrUGbi2nLAX+QfH36h0gw3Pp2auq\nOyOQm+qEu4xsBWp696QO7g8vkXpM9gz3r3AtumOBV4DzEqjHrjB9JqlqHoCIXIJriRfiek+VBnMr\nMgzD8InNkhuGYfjEDKZhGIZPzGAahmH4xAymYRiGT8xgGoZh+CQlDKaIjBSRGt7naJ3boy37aBE5\nJwnljgv7fPD6/Z4rIt2Ki0guIgNE5Kfeaia8434dvr+8uhuH4z2/B7z33iUc01pEYr6CR0SuFpET\nYy23shIUP0w/hFZgtPUq1tU4B98ngL8D7+McrtOBscBQnLPxV0A9oCcwBzgRuB3ne7gb2IbzbZsA\n/BTnqDsB56O4C+eoewrQSUTOBc7wymqtqlNF5CJPZh7wlqdXGjARGAasxi1prImLFn4+brVGnld+\nS0+/N4Ff45YlvuCV10xV1+H84MRzzt8KzPfKmYnzz5sHnIlbHvoX7/0D4BIRuQC3KmOViByD81Xd\nD9QXkYe9e3KJiJwJ/BvQsHKN2LEN5zv7RxGpB5wELAM+Bbp5+3NFZCwwyduWg3vepwKf4VaezQD6\n4lYiLQGaAQ/gFm3k4erOfTifzTq4sHgKfJ2Ii6zopEQL0+NRVX0YWI5byvc48C+gO7AAmIozSPNw\nSyM/U9WHcAZqLrBWVZ/COcffALwI1FHVx4GNuKWCo3DG8GTc8sG9HFr1slhV3wcaAr/EOaMDfIOr\nkGlAb1wl34hzMP8QeNI7Lk9VnwS+V9VJQHvvOrYAhUAWzuC+ChzjlRcyWgJ0xq3O2YBbM79JVZ/1\n9LtAVe/HGeJc7z5sA/7jXWeo5dIaWOfJ6+fp/iPwP9yfxHne/WlX+qMwomCyqk4EFqnq33ErqqYC\nFwENcH+s4Or3AuBHVX0E92edr6rPeOco8Jp3zEu4td7dvPe1uGf8rVdGE5xRXZCA66sUpIrB1CKf\n3wZuAn6LW5kSWlIZeq8GnOatjHkdmAZsFpHzgX8Cl6jqAiBfRG7AGag3gRFAF+B74GjvVdMrc5eI\ndMAZQVXVfV5Zp+JahS08GU1wgURWAWcDg3CVOaRb+LW8DTT3Pu8K2ye4FmXzsHM+98pp7skOlzNX\nRIYCPXCtmKM83feHyQPXomnhnfsYbslbU9yP9iacsW4BrMCINcNE5E+4P2YAVHU37jktDDtuv6pu\nBY4WkVtwxq7o6pKiS4jn4XpGzYGVRY7fjOsVGTGg0q30EZE/AFtVdYaIXIsblihU1Sk+zk3DRfp5\nRFXXl3FsJnAz7k9pnqrOK7/2JZZ1Fq4F2gK4S1VLXJMvIgNKi1ouIjd6LWHDMIpQ6QymYRhGtKRK\nl9wwDCPpmME0DMPwiRlMwzAMn5jBNAzD8IkZTMMwDJ+YwTQMw/CJGUzDMAyfmME0DMPwiRlMwzAM\nn5jBNAzD8IkZTMMwDJ+YwTQMw/CJGUzDMAyf+DKYInKxiHwtIntE5CMR6RxvxQzDMIJGmQZTRFrj\nUhdMwQWcXQX8V0SktPOM1EBEsr33ySLySxE5WUQGe9s+FZFTijnnCxE5wfs8znsfJiIPFDnuNS+P\nzT0icpSI3BuWS+g+EblbRAaFHX++iEwqIiMkv2aYrCPyO4nIzSJye5Ftt4vIu165XUXkLyIySkR+\n5TUCRojI016cU8MoEz8V5XJgs6qOV9XVwPXeNqtkFYOlItICl1rjNFy6gzdFpAdwP9BXRGqJyJMi\ncpuXT+k94BYRScflAErDRXmvJiIZYbL3A3uA73DpPr4L2/cqLs1Im7Bt3+EixofTTkSG4XIsFSU8\nmOsk3J95zYM7VcfhchaNVdW5uLw3DbwyNqnqfbgo9r4TzBmVGz9GrzkuVwgAqrpLVT9V1f2lnGOk\nDm/jDMmH3vdWqrocuBSXO6gjLsXGYuB4VV3qfZ+ASzQHcCEuRUI6cFmY7G9UdayqPq+qe4E1Yfua\nAbcA94Q2qOr3uPxGFJUBPOp934PLfwSHUm8A9FTVF4EdRc4XABFpgktcNxA4F9jgJZV71ksVYRhl\n4sdgbsDlfQFARJqKyDXWjakYqOpiXCK5d3GZCdNFpBGwykva9RCuR3E88K03fq2q+g0u6yXA+ar6\nR1X9Iy6dw3yQAAAbrUlEQVSRWqmISE2csd0DXFBUpdLUxWVFHC4iI3DJ3kK09XI47SlBXg5wo4jc\njWt1jga2Az/19DGMMikzRYWItAW+BP6Ey5Q4AainqufEXz3DMIzgUGYr0eue/Rr4Pa5blonLi2wY\nhlGpsCRohmEYPrFxSMMwDJ+YwTQMw/CJGUzDMAyfmME0DMPwiRlMwzAMn5jBNAzD8IkZTMMwDJ+Y\nwTQMw/CJGUzDMAyfmME0DMPwSZVkFu7FThwNjALygNtx8RGfBuqq6jteXMaTcbES2+HCzT0O/Aw4\nBbjZO68Q2Kiqz5RDn6nA58AfcGG/HhSRN4H7gKuBbbhYkKuAAV50nkDjRfD5UFXfL+WY8OeQo6o1\nReQcXGzJP+DuydG4SEH34OJc7gDaAgO8c2sD33oidwP1gPrA71V1SwnlrgTW4eJwvgvciAvy0sgr\nryEumtEUoKqqdovuLvjDC4q9SFVPEZFpuBgK/8HF0FyPC2v3PfBLYD4uylNdYBmu/lXBhcWbCCzF\n1ZMFxLCueGHq/op7Dr2AOkC2qi4RkTm4UHvVcXX2NlX9sRRZDYFBQAfgIuBu3H2fhft97QBmAJfg\nnvV1qpofoZ5PAZ08uVOBwap6tYjc5slOid9RiKQaTFUtEJE9wK3AM0AG8APOEL7iHXaSqj7iBZFt\niwv+0RxoDHyK+5FNUNUdInJstLqISH3gE1X9s4hkAjtF5GxgKy4u5GhVXScio4BFOOMZd0TkIdwP\nry6uR/A/nIFaCdylqvtEpC/unnyFi/WowGvAT3FR8gtE5DicEQPYgjNm9VT17iLP4d+eke2KC7hy\nondP7sQZz7/gDFwbXODdy3Fh0hri/vTWAS1xIdT6lGIsWwIfAf2AmZ4+3XD39veqeo6IvAy0wMXN\nfLAct9EvbXFxMuvg4oBOwsXeLACW4O7teOAMb9tC3LXuxkXyOgGo6u0rAN7HBU6OZV25BBeG75fA\nc7hnpCJyCe4PrTXOqO8szVgCqOomEXkDF1Tn/bDn3BZX73fhwjtOwQUOPxCFnmcCH+D+QM4HfvB+\na6H7kpDfUawIQpd8Ie5feIGqPoR7KA1VdatnuPaLSCec0fgL7uH+Hy5OYyOgpmcsG+Iqa7R0xkX3\nfgj3wxec4V4MtPCMZQeckd4DbCpHWZGwRVWf9D4/hjNk83ABf88UkbuA5l6Q3d3AO96+83D3aIV3\nrgL/8D43UtXHcNHGQyzEXdsiVX0YF9LvFOAEERmD++G0xLWmtuJ+mAeA6cBg3B/dE8BJnn6DgHtL\nua5TgLNwhqgjzhD9CWcw24vIx971rMQZzGG+7lb5OAV4HddjeRkXB/Z9XEurJ+4edsW1LpsAP+L+\nqApwsTnP9WTc5wVOXoOrSzGpKyJyPK412AHY4wVc3g3sxLUsm+OCJO/CvyE6BRf4uV3Yc8bT/wnc\n7+AaYKqqFo2G70fPl3B/qINwf/pPACNxzz2Rv6OYEASD2RrXKjnXa9nMBk70DMHpuG54T1zU9764\nCvItcB3uR75SRO7D/QN+Xw49OuJ+lJNwlaQR8LG3r52IDMH9q7+O62J8UZyQOBAKJ1UFF3R3Zti2\nVar6IO5fezjOgP0CZ4jeAO4Cimt1bxORgbjWeojWuB9LqALvwd2T24HJuLqyBWdImuPu1UZcS/MB\nXOv1D7iu+e9xEdhXlXJdJwP9gYdxPYsZuF5DXdwz6IN7Dv/yju1SiqxY0RRn7Nt55bUEBuJaWt/g\n/ii64a6vGe5evQf8HJfO45+4P95+InKXiGQR27pyqaoOAv4OdBORP3rl/QzXnX4AaO2lDvFl3HD3\nuxnueYaeczdP/9/hejOFuJZ+pHr+A3gEdy/fBlDVVbj7+zmJ/R3FBAvvVgkRkV64H8B3qvp2svVJ\nRUQkQ1ULkq1HZUFEHgRygaaqemsx+4eq6vhyljHc612VSFLHMI3koKqvJVuHVMeMZVIoBI7y0owo\nbmjkFFzPMsvbvhA3LHQsrmcwB9fjElxr9jvcZN1WDo3p5+Amz2qXpUCgDaaIXMahhFf5wL81QU1i\nb5IkfFb2U1Vdkoiyo8Gb8Ooetmm+qn4VI9kdcBMdIeao6koRuZxDWRrzVHVaOco4Bde9DDFTVdeW\ndHwiEZGeuMmtENPCW+YiUg/ozaGkbMtU9UPijJdX6wrckAbANlX9X5SyOgGnhm161xuDLTci8hPc\nsEqIN1V1Y4RitnqTUiO8702AU1R1rIjUwmUVnYCbJB7nGc98nFFtBHzCoWGFDA6N6V8NNFHVMSJy\nD2UQaIOJm1W7Cjeu8rKqTk1g2YW4SYsWuLG9osm6gkYBbjC9JW5s8RcxlJ2GGx9rjMv8eK63vSfu\nBwswzXtFSzVgDG62fSVurDgoKO6HlY5zqfpXkf37gNtwkxw5uEmSROl1KW5sfR/OhScqg4mbMH0Q\nZ1y+x42Bx4pauAnIY3CTkDOikFHfG6fvjkvRnIFLET0Up++LOC+F77zx+Ua4erQVN2Z6HO7+3Ozt\n2xAme703R1G2PVTVQL9w6VWfwhtvTXDZLXEzx52SfR986tvc0/fkOMju5MluGbZNcJ4Lk2PxfHBd\np49xs/hJv59FdOuJm3DsUcL+o3Ddv4sTrFcabhLowRjIOh3XEmscBz274bw3GsT5fnTFTQD3iYd8\nm/QxDMPwSRDcigzDMFKCQI1hiog1dw2jkqCqUvZRR/Kr/2z4K241EsDy/17a5HoAEbkeN5a/AjdT\nXlNVx4vIRbgFEffgxk8/Af4GXK6qeyIpO1AGE+BXx91Y7HYdNwK5/b7Dtg3e+wETM7sc/BwitE3H\njeCWwT0Pfi9K+DlFzy2u/BDheoS2y+33+dIn/Nyi5U/M7HJwW1E9ipNX1rbitofrEk5J9yl0XGh/\n6BqLyivt++C9HzBh4lsHr72s+15Ut3AdizJh4lsHP4eegd/rLk5eSbJvGdyTCRPfOuy9NJ2Lu9aS\ndC1Jv3Adi7ue4n4TIRnhz6so4dtD1xKuX9HvpdXt8GsKv+4Q4eeF6/vflU9SDtpyuAdLiHzv1R64\nCZjueXh8jot58BVuccJRuJVsFwP/jqRgX11yERkiIqO8NcuGYRjJZDlugm2O9znEJlwLsx1ugklw\nq47W4AKoPIeLC3ERbsnzuUSI3xbmcTi3iac9n7TVOBeLz3BLFKfiFtln4VxaCorsOwO3Xvx7YDhw\nj6oujFRZwzCMUBe8KKoaau7+t5h9/b2Pw8M2R+wG53fSZw1ureyZOD80xYW9Ohbn63QWLsjAZ7g1\nyEX35eOseR4uKMDpkSpqGIaRbHy1MNVFwgHXygzns1JOK2nf7/yUaRiGETQCN+ljGIZRGrObdTxs\nlrz7ui9Ds+RTcI26XrgQkeNF5ELv2M2q+i8RmUEUs+MhzA/TMIxUIzRL3o1DhhNcaMjf4JYI1/Nm\nyOvj4pa2EpE2HJodj4pArfQxP0zDqDxE64dZSgvzCtyE9FxgLG7OZA4uQtGjuKW964Cu6uJ1RqV0\nYF5OncOZNWvWEdsipSQZsZBdHiItP9HXESv9yiMzVufH6x51795di6u30RDPuh40ud49S7rNifQV\n9zHMzs+t6IQLnlEIvIlzLWqzqF+bUfEu24g/3333HYWFhbRr167csnbu3Mk333xDp06dqFatWgy0\nM4JCRXm2iRjDvAu4bVG/NufiIiaP41CKBSOFeemll5gyZQpTp07lmWeiTtYJwK5duxg8eDDz589n\n4MCBHDgQSb4tI8hUpGebqFnyA52fWyE4f8w7cd72Rorz6aef0q9fP0SEv//97xx7bNRJO1m+fDmn\nnnoq3bp1Y/369Wzfvp26devGUFsjWSxbtozTTjuNbt26sW7dOnbs2FEueZe2HnDYGOZ/vnuirLXk\nMZspT4TBHMOh5Er/xWUXNCoAvXr1YuLEiWRkZNCrV69yyTrppJN44oknWL9+Pfn5+Rx99NEx0tJI\nNh07duTJJ59k3bp17N27lzp16pRXZKRryUuaKY9oHTkkwGAu6tdmMXBOvMsxEk+3bt0488wzOXDg\nANWrV2f27NlRy8rMzOTpp59m586dHHXUUYhENYFqBJA4PNvlJXzeBNSgyFpyDp8p34tbR96VIBpM\no2KTmZlZ9kE+ERFq1y4zD5WRgsTy2Ya64EXRUtaSe/QJ+xxVOhVzXDeMcjBy5MiQS5xRCQhcC3PT\npk2Hfd+9e/cR2yKlJBmxkF0eIi0/0dcRK/3KIzNW5wflHsVbVipcZyoTOIPZsGHDw77XqlXriG2R\nUpKMWMguD5GWn+jriJV+5ZEZq/ODco/iLSsVrjOVSbTj+ju4/M0Av17Ur01M8h4bFYOCggLWrFlD\nixYtyMjIKPsEw0gwiXZc3wP8AZdK9aQElG2kCAUFBdxwww288MIL3HDDDRQUFCRbJSOgiMhTInKW\niNQQkXlh24cWOe5qEaknIl1jVXaiHde3A1cBmxf1a/N6gso2UoA1a9bQpk0bevfuzYwZM1i7di3H\nHXdcstUygskK4DSgIS6oechYniEijXCZHhSoBpzobe+BczOaAVwD7FDVYZEWnIgW5hjgEZxfVD2g\nH/DTzs+t+EUCyjZShBYtWrBixQpmzJjBihUraN68ebJVMoKL4mxXM+B2b1sGLhtkOs5JvQnOsH7t\nHbsS53d5KvA2LpVOxCTDcf2heJdppB4ZGRk89dRTrF27lubNm9sYplEainNYPwFARG7FZYLMweUf\nE5wBzQVOBg7gnNlbcyh9TlTe84GbJTcqLxkZGdYNN8pEVR/1Pr5RwiHvF/n+XpHvpaXWKZXAGUzz\nw4z8+EjkjBo1ipEjR5a4L5wLL7ww5n6Yr732Gu3bt/ctM5oyYnleIuWaH2bwCZzBND/MyI+PRM4T\nTzzB5MmTS9wXzuWXXx5zP8wFCxaYH2YcZaXCdaYyifbDfAuXRP0Azg9zfbzLN1IH88M0SkNEsnDu\niF+p6q6w7b/HzZj3AO4B2uAmf7JwaSn2qupnYce3BFqp6hzv+5Wq+rwfHRLRwgz5Yc7r/NyK3+MC\nCP8FZ0ANAzjkh9mmTRtWrFjBU089ZUbTOIiIZJ1zzjnv9+3b98SpU6d+LSLnhhnN9Tjj+A5wBm6y\nJx9YA9QF6nhuRQosBVoCVUXkZ8DrRBDQPFHBN0J+mFWBWbjp/S4JKttIAcL9MNu0acPatWuTrZIR\nLE7q27fvid26dcvo27dve6BD2L4WuF5rBtDY2zZbVWd6nxVYi0uAVg3XWKvibUuPRIlEBxB+DXge\np+R9CSjbSBHMD9Mog6+mTp36NdB+6tSpS4EloR2qOqWkk1T1NR+y55V9iMMCCBuBwPwwjdJQ1V0i\ncu68efM6AEvCxzATSeBmyY3Ki/lhGqXhGcmPk6mDBRBOMCJy8BXJ8bGitGC3RXMwF/XLjAUl+YDG\nAktrUbERkb+KyGzv9dew7VeKSE8RuURErhORK0LbY61D4FqYFd1xfcCAARGVHzp+06ZNCb+O0047\nLaUCCA8YMKDEfang0G2O62VSUvIzgHpeRsg/ADVF5GLcZFBMCZzBrOiO6+HO4X4cw0PHT548OeHX\nEamTebIDCJfmlJ8KDt3muB41LYDVInI5bgZ8r6q+IiI1Y11Qoh3X3wS+B6Ys6temerzLNmKDqjJi\nxAh27drF0UcfzciRI2Pe/T1w4AB33HEHe/fupXHjxtxxxx0xlV8eVJV7772X7du3U7t2bUaNGmXd\n/+RQbLZIVR1T3MGq+q9YK5DoAMJbcF74UYVWMpLDt99+S40aNejfvz8HDhxg/frYL9BavHgxjRo1\non///uTk5LB9+/aYlxEta9euRUTo378/GRkZfP/998lWqVKiqteranfvVWzmyHiTaMf1mrjmc73O\nz63omaCyjXLSqFEjVqxYwcqVK1mzZg316tWLeRnNmzdnyZIlrFy5ks2bN1OrVq2YlxEt9evX5/vv\nv2flypWsXLmSY445JtkqGUki0QGEMxb1a3MtkLOoX5u3Sj/NCApZWVmMGjWKbdu28cADD1C9euxH\nU+rXr8+f/vQntm3bxrhx46hSJTjD6zVq1OCBBx5g27ZtjB49OlDG3EgsSXFcX9SvjTnbpRgtWrTg\n6quvjmsZxx13XGD9MJs1axb36zeCT3D+xiMgNOCuqojIYe9F9weNcJ1mz54d0fHxJvxeAsyaNSth\nZceCID5vI7aIyOm4iGdpwD9VdWkpx14NzFDV3FiVHziD6ccPM9w3ccCAAQwcOJAffvjh4HHh+0uS\nUdr2RJGMAMKlEbqXofsXK/0iPSYe56eCf6L5YfqiGy4OxY3AxSLya1x8ivG4vD2PefsUl4aig4jU\nx00218cF4MgFmuKCAT2nqqv8Fh44g+nHDzPcNzH8c3H7S5JR2vZEkYwAwqWRiADC8fTDjGe5iZBr\nfpi+UFXdJyLP4gzfDbhwbfNwLovnAc8AFwL7cUnQ/gAUANtVdaKIDAF+Asz0zl3lt/BE+2G+g8sa\nuQ64flG/NitKO7ewsJBXX32VxYsXM2DAAKpWrRpvdQ9jyZIlvPjii1xwwQWceeaZCS27MvLhhx/y\n9ttv06dPn3KlsYgHX3/9NS+88AI9e/bkrLPOSrY6B9m3bx9PPvkkaWlp3HjjjclWJxHMF5GRuBbk\nSFzmyJk4AwgwF2dE84EFuCRoabg8P5eJyC24sG41w871TaL9MAuA3biLLXNc4dFHH+Wss87i+OOP\nZ8yYYn1T40Z+fj6PPPIIvXv35rnnnuOHH35IaPmVjQ0bNjB16lR69+7N2LFj2bdvX7JVOsjevXsZ\nN24cvXv35vnnn2fDhg3JVukgDz30EMceeyxNmzZl3LhxyVYn7qjqXFUdpaqjVfU+VR2pqt+r6qOq\n+h9VXa2qd3n7X8d559RU1U+Ajao6QVVfDD83kvIT7Ye5F/g18A0lrwk9SE5ODi1btqRJkybs2LEj\n3joeRn5+PnXr1iUzM5PGjRuzbdu2hJZf2cjNzaVJkyZkZmZSp04d8vPzk63SQfbu3UudOnXIzMyk\nadOmgaoLO3bsoFmzZrRo0YLc3JjNbVQYVDVfVUd6nx8t6/iySLQfZk3gZaATLvJ6qQwePJjHHnuM\nKVOmcNNNN8VXyyLUqVOHE044gezsbKpXrx64LmJF46STTiI9PZ3s7GxOOukkjjrqqGSrdJCjjjqK\nDh06kJ2dTXp6Oh06dCj7pARx8803M2XKFJ555hkGDx6cbHUSjohkichZXr6fuJMMP0zfkdabNm1K\n//796d69e8z18kP//v2TUm5lZdiwYclWoUSuu+66ZKtQLK1ateLxxx8/+P27775LojaJpaQ8P+FJ\nzUSkNnCUqsYk50ngZskNwzB8cjDPD9Dei8b+MXCsiNyDs29v4PL8VEyDGY0fZtHzguaHGQrEO3Lk\nyMOC8l544YWB8sMsOrE1f/78iPUbOHBgqUGCzQ8zvrJS4TpjSEl5flYCrXBGciXQSERaq2r5m99F\no2wn8+XUOZxZs2YdsQ03y37E5+L2lySjtO2xpjh9gYjLT/R1RKNfcc+wPDJjdX5Q7lG8ZaXCdapq\nqJ7EwmZkAWcCWbGQV9YrGX6YP8eluvz5on5tbFovhqgqu3fvplatWikZrzHV9f/xxx+pVq0a6ekR\nZW4NDKl4/zXBeX4S0SUP+WHO6/zcihG43MHpQGYCyq40qCq33XYbVapUITc3l8cff5zMzNS5xfv2\n7WPQoEHUrVuXgoICxo8fnzI/WnA+w2vXriUnJ4dx48YFZVWMbwoLC7npppuoXbs2eXl5TJgwgbS0\n4Kf88mbHTwK+0gRkkkykH2YaMAB4GPgIODtBZVcKdu7cSUZGBv369ePkk09m8eLFyVYpIr788ks6\nduxIv379qFatWqB8Hf2wevVqrr32Wvr27cuMGTOSrU7ELF++nOOPP55+/fpRv359Nm4Mfoxvz1i+\nD8wB3g+5FsUj+VmIRLQwxwCTccZ5AnAHbgzvngSUXWnIyspi8+bNzJkzhwULFnDVVVclW6WIaNu2\nLRMnTiQ9PZ0ffviBOnXqJFuliKhWrRozZ87k66+/ZujQoclWJ2JatWrFV199RY0aNVi1alWqBEk+\nCTgR12ttDxSdJV8DVAd+BI4DVqjnbhQtyfDDfDjeZVZG0tLSmDx5Mp9//jm/+93vAuX47YesrCwm\nTpzI0qVLufbaa1OiOxjOmDFjWLhwIb/97W9p1KhRstWJmBo1avCXv/yFJUuW8Pvf/56MjIxkq+SH\nr3DBNdoDRWfJwS2UKcStF+8OICJtVXU5URI4tyIjeqpVq5bSQUJq166dsvqnpaVx+umnJ1uNcpGV\nlZVS91+dk/q5uJblktAYphaf/CwmCdHMYBqGkbIkepZcNEBRqkUkOMoYhhFXVDV13CA8AmUwDcMw\ngkxqjawbhmEkETOYhmEYPgmkwRSRq0XkRJ/HvRiBzBdFpIWIDCqnfoNF5B3vc28Ruaw88sLkThaR\nqiKSLiLZZew/4OcelXSeiPgOz13S+SLSsrh7KSLXisgLInKUiDwkIsf6KOMGEbnIi214SwS6hZ/3\nR7/3pAwZHSK5P37llXS/IpDTSUQe8/ZFXO+KyMsrra6VV56IXBXNswg6FWGWPJJBWC3yHi07AESk\nKtAT+GM55YV4AbgYl8Zjehn7h0Qpdwgum14k96Ck80OvohTgFim8CGSrjzQAqvqUiEzCjasP9KtY\n2HlpRDlbWoKMqOtIKfJKul9+5eyH6Otd+D0GelF6XSuvvNRyBPZJIFuYCSAWs3OjgCuAbapaGAN5\nqOps3JLRnwFHrK8L238+4HvdbAnn+b4HZZxfnBzB/alsA3wt2RG3cDwNqCoivqNXhJ2XQZQNgFjI\n8CGvtPvlV47iMhZEVe/C7zFuSWGJdS0G8lJuBtwPQW1hxnvqvrzyFWcMbsCl8Iwl64Hqqrq/hP3r\ngBq4VlwkFD0v0ntQ0vnFyamBa2H+HLhHRDqp6hdlyL8D+BtuZcYI4F6feoXO2w/8D5deNVJiIaMs\nedH0bo6Qo6ofeMMF0dS7ovd4HVCjlLoWtbyK6iJobkWGYRg+qaxdcsMwjIgxg2kYhuETM5iGYRg+\nMYNpGIbhEzOYhmEYPgmqW5GRAohIB+AqnKvRduANVV1S+lmGkbqYW5ERNSIyXlWHep+vB9riVngM\nAV7FJcC7DNiMi459BTAXOA9YBCwATgfycI7OeTgn6B9xK0aOBeoBbwG/8fYtAZoBD+D8APOAaWao\njURgXXKjPIQ7z2cA/wA+AFoAC4F9wGeq+hBuldAGVX0SWA1MAToCjVT1z0BL4DNcCmYFuqjqwzhH\n/mY4A7wAeMkrt5v3vhZoHd/LNAyHGUyjPEwVkUe8hFOFeGudca3F/bjW4OkiMhx4HTjg7dew43JF\n5CbgOw61NlsCC0RkKNAIZ2CLdoXmAbWA5kCZa9UNIxZYl9wwDMMn1sI0DMPwiRlMwzAMn5jBNAzD\n8IkZTMMwDJ+YwTQMw/CJGUzDMAyfmME0DMPwyf8Dm4JihBR5IQUAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x116f29350>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plt.figure(figsize=(4, 4), dpi=300)\n",
    "\n",
    "grid = gridspec.GridSpec(2, 2, height_ratios=[2, 1], hspace=0.7, wspace=0.4)\n",
    "\n",
    "gs = {}\n",
    "gs['length'] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,0])\n",
    "gs['fluctuation'] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,1])\n",
    "gs['loh'] = gridspec.GridSpecFromSubplotSpec(7, 1, subplot_spec=grid[1,:], hspace=0)\n",
    "\n",
    "### top left panel ###\n",
    "ax1 = plt.subplot(gs['length'][:])\n",
    "ax1.text(-0.18, 1.05, 'A', transform=ax1.transAxes,\n",
    "         fontsize=9, fontweight='bold', va='top', ha='right')\n",
    "\n",
    "colors = [config.dict_selection['color'][e] for e in loh_length_df.columns]\n",
    "\n",
    "loh_length_df.rename(columns=config.dict_selection['short_label'])\\\n",
    ".plot(ax=ax1, logy=True, color=colors, style='.', marker='o', ms=2., mec='none', legend=False)\n",
    "\n",
    "utils.simple_axes(ax1)\n",
    "ax1.set_xlim(0,1.1E3)\n",
    "ax1.set_xlabel('Homozygosity segment length (kb)')\n",
    "ax1.set_ylabel('Frequency')\n",
    "\n",
    "ax1.legend(frameon=False, loc='upper right', \n",
    "           borderaxespad=0., prop={'size':5},\n",
    "           handlelength=0.75)\n",
    "\n",
    "### top right panel ###\n",
    "ax2 = plt.subplot(gs['fluctuation'][:])\n",
    "ax2.text(-0.18, 1.05, 'B', transform=ax2.transAxes,\n",
    "         fontsize=9, fontweight='bold', va='top', ha='right')\n",
    "\n",
    "colors = [config.dict_background['color'][b] for b in loh_fluctuation_df['LOH rate','mean'].columns] \n",
    "loh_fluctuation_df['LOH rate','mean'].plot(ax=ax2, kind='bar', yerr = loh_fluctuation_df['LOH rate','sem'], \n",
    "                  color=colors, legend=False,\n",
    "                  error_kw=dict(ecolor='0.1', lw=.75, capsize=.75, capthick=.75))\n",
    "\n",
    "utils.simple_axes(ax2)\n",
    "ax2.set_yscale('log')\n",
    "ax2.set_xlabel('Environment')\n",
    "ax2.set_xticklabels(loh_fluctuation_df.index.get_level_values('environment'), minor=False, rotation=0)\n",
    "ax2.set_ylabel('5-FOA+/total 5-FOA')\n",
    "\n",
    "ax2.legend(frameon=False, loc='upper right', \n",
    "           borderaxespad=0., prop={'size':5},\n",
    "           handlelength=0.75)\n",
    "\n",
    "### bottom panel ###\n",
    "ax3 = plt.subplot(gs['loh'][:1])\n",
    "\n",
    "# make a color map of fixed colors\n",
    "cmap = mpl.colors.ListedColormap([config.dict_background['color']['NA'], \n",
    "                                  config.dict_background['color']['WAxNA'], \n",
    "                                  config.dict_background['color']['WA']])\n",
    "bounds=[0,1,2]\n",
    "norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "\n",
    "sdn = seq_dn_df.ix['HU','WAxNA_F12_1_HU_3_T32']\n",
    "\n",
    "indexer = [slice(None)]*len(sdn.index.names)\n",
    "indexer[sdn.index.names.index('type')] = ['consensus']\n",
    "gc = sdn.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "\n",
    "x = gc.columns.get_level_values('pos').values\n",
    "y = np.arange(len(gc.index))\n",
    "data = gc.values\n",
    "\n",
    "plot.heatmap(np.r_[x, x.max()+1], np.r_[y, y.max()+1], data, \n",
    "             ax3, '', '', '', [], [], cmap=cmap, vmin=0, vmax=2)\n",
    "\n",
    "ax3.set_title('WAxNA F12 1 HU 3 T32', fontsize=6, y=3.25)\n",
    "\n",
    "ylabels = ['Consensus']\n",
    "ax3.set_yticks(np.arange(len(ylabels)) + 0.5, minor=True)\n",
    "ax3.set_yticklabels(ylabels, fontweight='bold', va='center', minor=True)\n",
    "\n",
    "# show de novo mutations\n",
    "indexer = [slice(None)]*len(sdn.index.names)\n",
    "indexer[sdn.index.names.index('type')] = ['de novo']\n",
    "gn_all = sdn.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "        \n",
    "labels = gn_all.columns.get_level_values('gene')\n",
    "loc = zip(gn_all.columns.get_level_values('pos'), [-.25]*gn_all.shape[1])\n",
    "\n",
    "for l, xy in zip(labels, loc):\n",
    "    trans = ax3.get_xaxis_transform() # x in data units, y in axes fraction\n",
    "    ann = ax3.annotate(l, xy=xy, xytext=(0, 4), textcoords='offset points',\n",
    "                       arrowprops=dict(arrowstyle='wedge,tail_width=0.7', color='black'),\n",
    "                       fontsize=5, style=('italic' if l!='non-coding' else 'normal'), \n",
    "                       weight=('bold' if l in ['RNR2','RNR4','FPR1','TOR1'] else 'normal'),\n",
    "                       annotation_clip=False, va='bottom', ha='center') \n",
    "#                        bbox=(dict(facecolor='none', edgecolor='gray', boxstyle='round,pad=.2')\\\n",
    "#                              if l in ['RNR2','RNR4','FPR1','TOR1'] else \\\n",
    "#                              dict(facecolor='none', edgecolor='none', boxstyle='round,pad=.2')))\n",
    "\n",
    "ax4 = plt.subplot(gs['loh'][1:])\n",
    "\n",
    "ax4.text(-0.085, 1.75, 'C', transform=ax4.transAxes,\n",
    "         fontsize=9, fontweight='bold', va='top', ha='right')\n",
    "\n",
    "indexer = [slice(None)]*len(sdn.index.names)\n",
    "indexer[sdn.index.names.index('type')] = ['loh']\n",
    "gl = sdn.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "\n",
    "x = gl.columns.get_level_values('pos').values\n",
    "y = np.arange(len(gl.index.get_level_values('clone')))\n",
    "data = gl.values\n",
    "\n",
    "# make a color map of fixed colors\n",
    "# cmap = mpl.colors.ListedColormap([config.dict_background['color']['NA'],\n",
    "#                                   'w',\n",
    "#                                   config.dict_background['color']['WA']])\n",
    "cmap = mpl.colors.ListedColormap(['k','w','k'])\n",
    "bounds=[-1,0,1]\n",
    "norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "\n",
    "plot.heatmap(np.r_[x, x.max()+1], np.r_[y, y.max()+1], data, ax4, '', '', '', [], [], cmap=cmap, vmin=-1, vmax=1)\n",
    "\n",
    "lineage = gl.index.get_level_values('lineage').unique()[0]\n",
    "ax4.tick_params(axis='y', colors=config.dict_lineages[lineage]['fill'], width=5, which='both')\n",
    "\n",
    "# show de novo mutations\n",
    "indexer = [slice(None)]*len(sdn.index.names)\n",
    "indexer[sdn.index.names.index('type')] = ['de novo']\n",
    "gn = sdn.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "            \n",
    "for i,(k,g) in enumerate(gn.groupby(level='clone')):\n",
    "    g = g.dropna(axis=1)\n",
    "    x = g.columns.get_level_values('pos').values\n",
    "    y = [i+.5]*len(x)\n",
    "    data = g.values.flatten()\n",
    "    ax4.scatter(x, y, \n",
    "                facecolors=[config.dict_genotype['color'][int(gt)] for gt in data],\n",
    "                edgecolors='k', s=8, zorder=3)\n",
    "\n",
    "ylabels = gl.index.get_level_values('clone')\n",
    "ax4.set_yticks(np.arange(len(ylabels)) + 0.5)\n",
    "ax4.set_yticklabels(ylabels, fontweight='bold', va='center')\n",
    "[ax4.axhline(g, lw=0.5, ls=\"-\", color=\"lightgray\") for g in np.arange(len(ylabels))]\n",
    "\n",
    "xlabels = sgd_length.chr_start + (sgd_length.chr_end - sgd_length.chr_start)/2.\n",
    "xstart = sgd_length.chr_start\n",
    "ax4.set_xticks(xlabels)\n",
    "ax4.set_xticklabels(sgd_length.chr_arabic.apply(utils.int_to_roman))\n",
    "xgrid=[xst+1. for xst in list(set(xstart))]\n",
    "[ax4.axvline(g, lw=0.5, ls=\"-\", color=\"gray\") for g in xgrid]\n",
    "\n",
    "ax4.set_xlabel('Chromosome')\n",
    "\n",
    "# set legend    \n",
    "wt_artist = patches.Rectangle((0,0), width=1, height=1, color=config.dict_background['color']['WA/WA'])\n",
    "hom_artist = patches.Rectangle((0,0), width=1, height=1, color=config.dict_background['color']['NA/NA'])\n",
    "het_artist = patches.Rectangle((0,0), width=1, height=1, color=config.dict_background['color']['WA/NA'])\n",
    "\n",
    "leg1 = ax3.legend([wt_artist,hom_artist,het_artist], ['WA/WA','NA/NA','WA/NA'], \n",
    "                  bbox_to_anchor=(1.1, 0.5), ncol=1,\n",
    "                  frameon=False, loc='center',\n",
    "                  borderaxespad=0, handlelength=0.75, \n",
    "                  title='Background', prop={'size':4})\n",
    "\n",
    "snp_indel_artist = lines.Line2D((0,1),(0,0), mec='k', mfc=config.dict_genotype['color'][1], \n",
    "                                marker='o', markersize=3, linestyle='')\n",
    "loh_artist = patches.Rectangle((0,0), width=1, height=1, color='k')\n",
    "    \n",
    "leg2 = ax4.legend([snp_indel_artist,loh_artist], ['SNP/indel','LOH'], \n",
    "                  bbox_to_anchor=(1.1, 0.6), ncol=1,\n",
    "                  frameon=False, loc='center',\n",
    "                  borderaxespad=0, handlelength=0.75, \n",
    "                  title='Mutation', prop={'size':4})\n",
    "    \n",
    "het_artist = lines.Line2D((0,1),(0,0), mec='k', mfc=config.dict_genotype['color'][1], \n",
    "                          marker='o', markersize=3, linestyle='')\n",
    "hom_artist = lines.Line2D((0,1),(0,0), mec='k', mfc=config.dict_genotype['color'][2],\n",
    "                          marker='o', markersize=3, linestyle='')\n",
    "    \n",
    "leg3 = ax4.legend([het_artist,hom_artist], ['het.','hom.'], \n",
    "                  bbox_to_anchor=(1.1, 0), ncol=1,\n",
    "                  frameon=False, loc='lower center',\n",
    "                  borderaxespad=0, handlelength=0.75, \n",
    "                  title='Genotype', prop={'size':4})\n",
    "\n",
    "ax4.add_artist(leg2)\n",
    "        \n",
    "for leg in [leg1,leg2,leg3]:\n",
    "    plt.setp(leg.get_title(),fontsize=5)\n",
    "\n",
    "# axes limits\n",
    "for ax in fig.get_axes():\n",
    "    ax.xaxis.label.set_size(6)\n",
    "    ax.yaxis.label.set_size(6)\n",
    "    ax.tick_params(axis='both', which='major', size=2, labelsize=6)\n",
    "    ax.tick_params(axis='both', which='minor', size=1, labelsize=4)\n",
    "\n",
    "fig.savefig(dir_paper+'figures/figure3/figure3.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_paper+'figures/figure3/figure3.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fig. 4: Pervasive loss of heterozygosity. (A) The length distribution of homozygous segments, in bins corresponding to 25-kb increments, shows evidence of long homozygosity tracts above 300 kb in hydroxyurea and rapamycin. Ancestral haploid isolates are used to compare a set of *in silico* diploid genomes to evolved diploid isolates. Only unrelated isolate backgrounds were included. (B) Background- and environment-dependent rates of loss-of-heterozygosity were measured in a fluctuation assay by loss of the *URA3* marker. 5-FOA+ colonies indicate loss of the marker. The mean number of tandard error of the mean. (C) Parental genotype (WA/WA=blue, het=white, NA/NA=red) of six related isolates from population WAxNA F12 1 HU 3, carrying a shared *RNR2* heterozygous mutation and private regions with LOH (black). Chromosome II has undergone extensive LOH in isolates C5 and C6 and these individuals have 15% faster growth rate compared to C1-C4."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure S7 - "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(4, 14), dpi=300)\n",
    "\n",
    "indexer = [slice(None)]*len(seq_dn_df.index.names)\n",
    "indexer[seq_dn_df.index.names.index('type')] = ['consensus','loh']\n",
    "seq_dn_df_sliced = seq_dn_df.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "\n",
    "shape = pd.DataFrame({k: \n",
    "x.shape for k,x in seq_dn_df_sliced.groupby(level=['selection','population'])\n",
    "})\n",
    "\n",
    "ygrid = shape.ix[0].values\n",
    "grid = gridspec.GridSpec(len(ygrid), 1, height_ratios=ygrid, hspace=0.7)\n",
    "\n",
    "gs = {}\n",
    "\n",
    "###  ###\n",
    "\n",
    "for ii, ((e,p), sdf) in enumerate(seq_dn_df.groupby(level=['selection','population'])):\n",
    "\n",
    "    idx = 0\n",
    "    \n",
    "    nrows = ygrid[ii]\n",
    "    ncols = 1\n",
    "    gs[(e,p)] = gridspec.GridSpecFromSubplotSpec(nrows, ncols,\n",
    "                                                 subplot_spec=grid[ii],\n",
    "                                                 hspace=0, wspace=0)\n",
    "    \n",
    "    for jj, (s, jdf) in enumerate(sdf.groupby(level='set')):\n",
    "        \n",
    "        indexer = [slice(None)]*len(jdf.index.names)\n",
    "        indexer[jdf.index.names.index('type')] = ['consensus']\n",
    "        gc = jdf.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "        \n",
    "        print e,p,s\n",
    "        ax1 = plt.subplot(gs[(e,p)][idx:idx+len(gc)])\n",
    "\n",
    "        idx += len(gc)\n",
    "        \n",
    "        # set title\n",
    "        ax1.set_title(p)\n",
    "        \n",
    "        if len(gc) > 0:\n",
    "            x = gc.columns.get_level_values('pos').values\n",
    "            y = np.arange(len(gc.index))\n",
    "            data = gc.values\n",
    "            \n",
    "            # make a color map of fixed colors\n",
    "            cmap = mpl.colors.ListedColormap([config.dict_background['color']['NA'], \n",
    "                                              config.dict_background['color']['WAxNA'], \n",
    "                                              config.dict_background['color']['WA']])\n",
    "            bounds = [0,1,2]\n",
    "            norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "            \n",
    "            title = (p.replace('_',' ') if jj==0 else '')\n",
    "            xlabel= ''\n",
    "            ylabel= ''\n",
    "            xticklabels = []\n",
    "            yticklabels = []\n",
    "    \n",
    "            plot.heatmap(np.r_[x, x.max()+1], np.r_[y, y.max()+1], data, \n",
    "                         ax1, '', xlabel, ylabel, xticklabels, yticklabels, cmap=cmap, vmin=0, vmax=2)\n",
    "        \n",
    "            ax1.set_title(title, fontsize=6)\n",
    "\n",
    "        indexer = [slice(None)]*len(jdf.index.names)\n",
    "        indexer[jdf.index.names.index('type')] = ['loh']\n",
    "        gl = jdf.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "        \n",
    "        ax2 = plt.subplot(gs[(e,p)][idx:idx+len(gl)], sharex=ax1)\n",
    " \n",
    "        idx += len(gl)\n",
    "        \n",
    "        if len(gl) > 0:\n",
    "            x = gl.columns.get_level_values('pos').values\n",
    "            y = np.arange(len(gl.index))#.get_level_values('isolate')))\n",
    "            data = gl.values\n",
    "    \n",
    "            # make a color map of fixed colors\n",
    "#             cmap = mpl.colors.ListedColormap([config.dict_background['color']['NA'],\n",
    "#                                               'w',\n",
    "#                                               config.dict_background['color']['WA']])\n",
    "            cmap = mpl.colors.ListedColormap(['k','w','k'])\n",
    "            bounds=[-1,0,1]\n",
    "            norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "\n",
    "            plot.heatmap(np.r_[x, x.max()+1], np.r_[y, y.max()+1], data, \n",
    "                         ax2, '', '', '', [], [], cmap=cmap, vmin=-1, vmax=1)\n",
    "            \n",
    "            lineage = gl.index.get_level_values('lineage').unique()[0]\n",
    "            ax2.tick_params(axis='y', colors=config.dict_lineages[lineage]['fill'], width=5, which='both')\n",
    "\n",
    "        # show de novo mutations\n",
    "        indexer = [slice(None)]*len(jdf.index.names)\n",
    "        indexer[jdf.index.names.index('type')] = ['de novo']\n",
    "        gn = jdf.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "        \n",
    "        if len(gn) > 0:\n",
    "            \n",
    "            for i,(k,g) in enumerate(gn.groupby(level='clone')):\n",
    "                g = g.dropna(axis=1)\n",
    "                x = g.columns.get_level_values('pos').values\n",
    "                y = [i+.5]*len(x)\n",
    "                data = g.values.flatten()\n",
    "                ax2.scatter(x, y, \n",
    "                            facecolors=[config.dict_genotype['color'][gt] for gt in data],\n",
    "                            edgecolors='k', s=8, zorder=3)\n",
    "\n",
    "        # set grid\n",
    "        xlabels = sgd_length.chr_start + (sgd_length.chr_end - sgd_length.chr_start)/2.\n",
    "        xstart = sgd_length.chr_start\n",
    "        if ii<(len(seq_loh_df.index)-1):\n",
    "            ax2.set_xticks([])\n",
    "            ax2.set_xticklabels([])\n",
    "        else:\n",
    "            ax2.set_xticks(xlabels, minor=False)\n",
    "            ax2.set_xticklabels(sgd_length.chr_arabic.apply(utils.int_to_roman), minor=False)\n",
    "            ax2.set_xlabel('Chromosome')\n",
    "        xgrid=[xst+1. for xst in list(set(xstart))]\n",
    "        [ax2.axvline(g, lw=0.5, ls=\"-\", color=\"gray\") for g in xgrid]\n",
    "    \n",
    "        ylabels = gl.index.get_level_values('clone')\n",
    "        ax2.set_yticks(np.arange(len(ylabels)) + 0.5, minor=False)\n",
    "        ax2.set_yticklabels(ylabels, fontweight='bold', va='center', minor=False)\n",
    "        [ax2.axhline(g, lw=0.5, ls=\"-\", color=\"lightgray\") for g in np.arange(len(ylabels))]\n",
    "    \n",
    "    # show de novo mutations\n",
    "    indexer = [slice(None)]*len(sdf.index.names)\n",
    "    indexer[sdf.index.names.index('type')] = ['de novo']\n",
    "    gn_all = sdf.loc[tuple(indexer),:].dropna(axis=1, how='all')\n",
    "        \n",
    "    genes = gn_all.columns.get_level_values('gene')\n",
    "    loc = zip(gn_all.columns.get_level_values('pos'), [gn.shape[0]+.25]*gn_all.shape[1])\n",
    "\n",
    "    for l, xy in zip(genes, loc):     \n",
    "        trans = ax2.get_xaxis_transform() # x in data units, y in axes fraction\n",
    "        ann = ax2.annotate(l, xy=xy, xytext=(0, -2), textcoords='offset points',\n",
    "                           arrowprops=dict(arrowstyle='wedge,tail_width=0.7', color='black'),\n",
    "                           fontsize=5, style=('italic' if l!='non-coding' else 'normal'), \n",
    "                           annotation_clip=False, va='top', ha='center', \n",
    "                           bbox=(dict(facecolor='none', edgecolor='gray', boxstyle='round,pad=.2')\\\n",
    "                                if l in ['RNR2','RNR4','FPR1','TOR1'] else \\\n",
    "                                 dict(facecolor='none', edgecolor='none', boxstyle='round,pad=.2')))\n",
    "        \n",
    "#     # highlight haploids\n",
    "#     if p in ['WAxNA_F12_1_RM_1_T32','WAxNA_F12_2_RM_1_T32']:\n",
    "#         ax2 = plt.subplot(gs[('RM',p)][(1,0)], sharex=ax1)\n",
    "#         ax2.text(1.025, 0.5, 'aneuploidy', fontsize=5, va='center', ha='left')\n",
    "#     # highlight haploids\n",
    "#     if p in ['WAxNA_F12_1_RM_1_T32','WAxNA_F12_2_RM_1_T32']:\n",
    "# #     ax2 = plt.subplot(gs[('RM','WAxNA_F12_1_RM_1_T32')][(1,0)])\n",
    "#         ax2.text(1.025, 0.5, 'aneuploidy', fontsize=5, va='center', ha='left')\n",
    "#     ax2 = plt.subplot(gs[('RM','WAxNA_F12_2_RM_1_T32')][(1,0)])\n",
    "#     ax2.text(1.025, 0.5, 'aneuploidy', transform=ax2.transAxes,\n",
    "#              fontsize=5, va='center', ha='left')\n",
    "\n",
    "# set panel labels\n",
    "ax1 = plt.subplot(gs['HU','WAxNA_F12_1_HU_1_T32'][(0,0)])\n",
    "ax1.text(-0.05, 3,  chr(ord('A')), transform=ax1.transAxes,\n",
    "         fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "ax1.text(0, 3, config.dict_environment['short_label']['HU'], transform=ax1.transAxes,\n",
    "         fontsize=8, va='center', ha='left')\n",
    "\n",
    "ax1 = plt.subplot(gs['RM','WAxNA_F12_1_RM_1_T32'][(0,0)])\n",
    "ax1.text(-0.05, 3,  chr(ord('B')), transform=ax1.transAxes,\n",
    "         fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "ax1.text(0, 3, config.dict_environment['short_label']['RM'], transform=ax1.transAxes,\n",
    "         fontsize=8, va='center', ha='left')\n",
    "\n",
    "# highlight diploids\n",
    "# ax1 = plt.subplot(gs[('RM','WAxNA_F12_1_RM_1_T32')][(0,0)])\n",
    "# ax2 = plt.subplot(gs[('RM','WAxNA_F12_1_RM_1_T32')][1])\n",
    "# ax2.text(1.025, 0.5, 'aneuploidy', transform=ax2.transAxes, fontsize=5, va='center', ha='left')\n",
    "# ax2 = plt.subplot(gs[('RM','WAxNA_F12_2_RM_1_T32')][1])\n",
    "# ax2.text(1.025, 0.5, 'aneuploidy', transform=ax2.transAxes, fontsize=5, va='center', ha='left')\n",
    "\n",
    "# set legend\n",
    "for k, xy in zip([('HU','WAxNA_F12_2_HU_1_T32'),('RM','WAxNA_F12_1_RM_3_T32')],[(2,0),(2,0)]):\n",
    "    ax3 = plt.subplot(gs[k][(0,0)])\n",
    "    \n",
    "    wt_artist = patches.Rectangle((0,0), width=1, height=1, color=config.dict_background['color']['WA/WA'])\n",
    "    het_artist = patches.Rectangle((0,0), width=1, height=1, color=config.dict_background['color']['WA/NA'])\n",
    "    hom_artist = patches.Rectangle((0,0), width=1, height=1, color=config.dict_background['color']['NA/NA'])\n",
    "    \n",
    "    leg1 = ax3.legend([wt_artist,het_artist,hom_artist], ['WA/WA','WA/NA','NA/NA'], \n",
    "                      bbox_to_anchor=(1.1, 1.5), ncol=1,\n",
    "                      frameon=False, loc='center',\n",
    "                      borderaxespad=0, handlelength=0.75, \n",
    "                      title='Background', prop={'size':5})\n",
    "    \n",
    "    ax4 = plt.subplot(gs[k][(2,0)])\n",
    "    \n",
    "    snp_indel_artist = lines.Line2D((0,1),(0,0), mec='k', mfc=config.dict_genotype['color'][1], \n",
    "                                    marker='o', markersize=4, linestyle='')\n",
    "    loh_artist = lines.Line2D((0,1),(0,0), mec='k', mfc=config.dict_genotype['color'][2],\n",
    "                              marker='o', markersize=4, linestyle='')\n",
    "    \n",
    "    leg2 = ax4.legend([snp_indel_artist,loh_artist], \n",
    "                      ['SNP/indel','LOH'], \n",
    "                      bbox_to_anchor=(1.1, -1.5), ncol=1,\n",
    "                      frameon=False, loc='center',\n",
    "                      borderaxespad=0, handlelength=0.75, \n",
    "                      title='Mutation', prop={'size':5})\n",
    "    \n",
    "    het_artist = lines.Line2D((0,1),(0,0), mec='k', mfc=config.dict_genotype['color'][1], \n",
    "                              marker='o', markersize=4, linestyle='')\n",
    "    hom_artist = lines.Line2D((0,1),(0,0), mec='k', mfc=config.dict_genotype['color'][2],\n",
    "                              marker='o', markersize=4, linestyle='')\n",
    "    \n",
    "    leg3 = ax4.legend([het_artist,hom_artist], ['het.','hom.'], \n",
    "                      bbox_to_anchor=(1.1, -3), ncol=1,\n",
    "                      frameon=False, loc='center',\n",
    "                      borderaxespad=0, handlelength=0.75, \n",
    "                      title='Genotype', prop={'size':5})\n",
    "    \n",
    "    ax4.add_artist(leg2)\n",
    "    \n",
    "    for leg in [leg1,leg2,leg3]:\n",
    "        plt.setp(leg.get_title(),fontsize=6)\n",
    "    \n",
    "# axes limits\n",
    "for ax in fig.get_axes():\n",
    "    ax.xaxis.label.set_size(6)\n",
    "    ax.yaxis.label.set_size(6)\n",
    "    ax.tick_params(axis='both', which='major', size=2, labelsize=6)\n",
    "    ax.tick_params(axis='both', which='minor', size=1, labelsize=4)\n",
    "\n",
    "fig.savefig(dir_supp+'figures/figureSX_seq_loh/figureSX_seq_loh.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_supp+'figures/figureSX_seq_loh/figureSX_seq_loh.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
