{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 5 - Background-averaged fitness effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/matplotlib/__init__.py:872: UserWarning: axes.color_cycle is deprecated and replaced with axes.prop_cycle; please use the latter.\n",
      "  warnings.warn(self.msg_depr % (key, alt_key))\n",
      "/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/matplotlib/__init__.py:872: UserWarning: svg.embed_char_paths is deprecated and replaced with svg.fonttype; please use the latter.\n",
      "  warnings.warn(self.msg_depr % (key, alt_key))\n"
     ]
    }
   ],
   "source": [
    "# load external dependencies\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "%matplotlib inline\n",
    "\n",
    "import math\n",
    "\n",
    "import json, matplotlib\n",
    "s = json.load( open(\"matplotlibrc.json\") )\n",
    "matplotlib.rcParams.update(s)\n",
    "\n",
    "from itertools import *\n",
    "\n",
    "import random\n",
    "import pandas as pd # (*) Pandas for data manipulation\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import os,glob\n",
    "\n",
    "# load internal dependencies\n",
    "import colors,plot,stats,utils\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "# absolute file path for data directory\n",
    "from paths import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Spores"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this experiment, we choose a set of ancestral and evolved clones. All of these are diploid cells which undergo sporulation. We can import the plate layout and the growth parameters. These two dataframes can be merged by the plate index, and row and column indices within the plates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Layout</th>\n",
       "      <th>Plate</th>\n",
       "      <th>Row</th>\n",
       "      <th>Column</th>\n",
       "      <th>env_evo</th>\n",
       "      <th>env_test</th>\n",
       "      <th>ID_YJH</th>\n",
       "      <th>strain</th>\n",
       "      <th>ID</th>\n",
       "      <th>background</th>\n",
       "      <th>...</th>\n",
       "      <th>tetrad</th>\n",
       "      <th>spore</th>\n",
       "      <th>mating</th>\n",
       "      <th>auxotrophy</th>\n",
       "      <th>genotype_short</th>\n",
       "      <th>genotype_long</th>\n",
       "      <th>aminoacid</th>\n",
       "      <th>contamination</th>\n",
       "      <th>doubling_time</th>\n",
       "      <th>growth_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>HU</td>\n",
       "      <td>HU</td>\n",
       "      <td>1903</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA_F12_2_HU_1_T34_C1</td>\n",
       "      <td>evolved</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>MATα</td>\n",
       "      <td>lys2-</td>\n",
       "      <td>+</td>\n",
       "      <td>RNR4</td>\n",
       "      <td></td>\n",
       "      <td>no</td>\n",
       "      <td>0.197277</td>\n",
       "      <td>0.872196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>HU</td>\n",
       "      <td>HU</td>\n",
       "      <td>1903</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA_F12_2_HU_1_T34_C1</td>\n",
       "      <td>evolved</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>MATα</td>\n",
       "      <td>lys2-</td>\n",
       "      <td>+</td>\n",
       "      <td>RNR4</td>\n",
       "      <td></td>\n",
       "      <td>no</td>\n",
       "      <td>0.482655</td>\n",
       "      <td>0.715659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>HU</td>\n",
       "      <td>HU</td>\n",
       "      <td>1903</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA_F12_2_HU_1_T34_C1</td>\n",
       "      <td>evolved</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>MATα</td>\n",
       "      <td>lys2-</td>\n",
       "      <td>+</td>\n",
       "      <td>RNR4</td>\n",
       "      <td></td>\n",
       "      <td>no</td>\n",
       "      <td>0.517250</td>\n",
       "      <td>0.698703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>HU</td>\n",
       "      <td>HU</td>\n",
       "      <td>1903</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA_F12_2_HU_1_T34_C1</td>\n",
       "      <td>evolved</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>MATα</td>\n",
       "      <td>lys2-</td>\n",
       "      <td>+</td>\n",
       "      <td>RNR4</td>\n",
       "      <td></td>\n",
       "      <td>no</td>\n",
       "      <td>0.554730</td>\n",
       "      <td>0.680785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>HU</td>\n",
       "      <td>HU</td>\n",
       "      <td>1909*</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA_F12_2_HU_1_T34_C1</td>\n",
       "      <td>evolved</td>\n",
       "      <td>...</td>\n",
       "      <td>B</td>\n",
       "      <td>4</td>\n",
       "      <td>MATα</td>\n",
       "      <td>lys2-</td>\n",
       "      <td>-</td>\n",
       "      <td>RNR4*</td>\n",
       "      <td>Arg34Ile</td>\n",
       "      <td>no</td>\n",
       "      <td>0.276690</td>\n",
       "      <td>0.825483</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Layout  Plate  Row  Column env_evo env_test ID_YJH strain  \\\n",
       "0       0      0    0       0      HU       HU   1903  spore   \n",
       "1       0      0    0       1      HU       HU   1903  spore   \n",
       "2       0      0    0       2      HU       HU   1903  spore   \n",
       "3       0      0    0       3      HU       HU   1903  spore   \n",
       "4       0      0    0       4      HU       HU  1909*  spore   \n",
       "\n",
       "                        ID background     ...     tetrad spore mating  \\\n",
       "0  WAxNA_F12_2_HU_1_T34_C1    evolved     ...          A     2   MATα   \n",
       "1  WAxNA_F12_2_HU_1_T34_C1    evolved     ...          A     2   MATα   \n",
       "2  WAxNA_F12_2_HU_1_T34_C1    evolved     ...          A     2   MATα   \n",
       "3  WAxNA_F12_2_HU_1_T34_C1    evolved     ...          A     2   MATα   \n",
       "4  WAxNA_F12_2_HU_1_T34_C1    evolved     ...          B     4   MATα   \n",
       "\n",
       "  auxotrophy genotype_short genotype_long aminoacid contamination  \\\n",
       "0      lys2-              +          RNR4                      no   \n",
       "1      lys2-              +          RNR4                      no   \n",
       "2      lys2-              +          RNR4                      no   \n",
       "3      lys2-              +          RNR4                      no   \n",
       "4      lys2-              -         RNR4*  Arg34Ile            no   \n",
       "\n",
       "  doubling_time growth_rate  \n",
       "0      0.197277    0.872196  \n",
       "1      0.482655    0.715659  \n",
       "2      0.517250    0.698703  \n",
       "3      0.554730    0.680785  \n",
       "4      0.276690    0.825483  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load data\n",
    "spores_abs_df = utils.load_data(dir_data+'pheno/pheno_spores_abs.pickle')\n",
    "spores_norm_df = utils.load_data(dir_data+'pheno/pheno_spores_norm.pickle')\n",
    "\n",
    "# filter out blank positions, contamination and constructs\n",
    "spores_norm_df = spores_norm_df[(spores_norm_df['contamination']=='no')&\\\n",
    "                                ~(spores_norm_df['strain'].isin(['control','']))&\\\n",
    "                                ~(spores_norm_df['genotype_long'].isin([u'fpr1Δ',u'tor1Δ']))]\n",
    "\n",
    "   \n",
    "# remove NaNs\n",
    "# spores_norm_df = spores_norm_df[np.isfinite(spores_norm_df['growth_rate'])]\n",
    "\n",
    "spores_norm_df.head()  # show dataframe header to stdout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#spore_abs_df.to_csv(dir_data+'spores/20150402_spores_abs_pheno_filtered_with_labels.csv', encoding='utf-8', index=False, sep='\\t')\n",
    "#spore_norm_df.to_csv(dir_data+'spores/20150402_spores_norm_pheno_filtered_with_labels.csv', encoding='utf-8', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hybrids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Layout</th>\n",
       "      <th>Plate</th>\n",
       "      <th>Row</th>\n",
       "      <th>Column</th>\n",
       "      <th>env_evo</th>\n",
       "      <th>env_test</th>\n",
       "      <th>MATa</th>\n",
       "      <th>strain_MATa</th>\n",
       "      <th>ID_MATa</th>\n",
       "      <th>background_MATa</th>\n",
       "      <th>...</th>\n",
       "      <th>background</th>\n",
       "      <th>population</th>\n",
       "      <th>gene</th>\n",
       "      <th>genotype_short</th>\n",
       "      <th>genotype_long</th>\n",
       "      <th>auxotrophy</th>\n",
       "      <th>tetrad</th>\n",
       "      <th>spore</th>\n",
       "      <th>strain</th>\n",
       "      <th>contamination</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31488</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>20</td>\n",
       "      <td>RM</td>\n",
       "      <td>RM</td>\n",
       "      <td>lys_218_R1 a ura-</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA</td>\n",
       "      <td>ancestral</td>\n",
       "      <td>...</td>\n",
       "      <td>(ancestral, evolved)</td>\n",
       "      <td>(WAxNA, WAxNA)</td>\n",
       "      <td>(, TOR1)</td>\n",
       "      <td>(+, -)</td>\n",
       "      <td>(, TOR1*)</td>\n",
       "      <td>(lys2-, ura3-)</td>\n",
       "      <td>(A, A)</td>\n",
       "      <td>(1, 1)</td>\n",
       "      <td>(spore, spore)</td>\n",
       "      <td>(no, no)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31489</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>21</td>\n",
       "      <td>RM</td>\n",
       "      <td>RM</td>\n",
       "      <td>lys_218_R1 a ura-</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA</td>\n",
       "      <td>ancestral</td>\n",
       "      <td>...</td>\n",
       "      <td>(ancestral, evolved)</td>\n",
       "      <td>(WAxNA, WAxNA)</td>\n",
       "      <td>(, TOR1)</td>\n",
       "      <td>(+, -)</td>\n",
       "      <td>(, TOR1*)</td>\n",
       "      <td>(lys2-, ura3-)</td>\n",
       "      <td>(A, A)</td>\n",
       "      <td>(1, 1)</td>\n",
       "      <td>(spore, spore)</td>\n",
       "      <td>(no, no)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31490</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>RM</td>\n",
       "      <td>RM</td>\n",
       "      <td>lys_218_R1 a ura-</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA</td>\n",
       "      <td>ancestral</td>\n",
       "      <td>...</td>\n",
       "      <td>(ancestral, evolved)</td>\n",
       "      <td>(WAxNA, WAxNA)</td>\n",
       "      <td>(, TOR1)</td>\n",
       "      <td>(+, -)</td>\n",
       "      <td>(, TOR1*)</td>\n",
       "      <td>(lys2-, ura3-)</td>\n",
       "      <td>(A, A)</td>\n",
       "      <td>(1, 1)</td>\n",
       "      <td>(spore, spore)</td>\n",
       "      <td>(no, no)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31491</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>20</td>\n",
       "      <td>RM</td>\n",
       "      <td>YNB</td>\n",
       "      <td>lys_218_R1 a ura-</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA</td>\n",
       "      <td>ancestral</td>\n",
       "      <td>...</td>\n",
       "      <td>(ancestral, evolved)</td>\n",
       "      <td>(WAxNA, WAxNA)</td>\n",
       "      <td>(, TOR1)</td>\n",
       "      <td>(+, -)</td>\n",
       "      <td>(, TOR1*)</td>\n",
       "      <td>(lys2-, ura3-)</td>\n",
       "      <td>(A, A)</td>\n",
       "      <td>(1, 1)</td>\n",
       "      <td>(spore, spore)</td>\n",
       "      <td>(no, no)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31492</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>21</td>\n",
       "      <td>RM</td>\n",
       "      <td>YNB</td>\n",
       "      <td>lys_218_R1 a ura-</td>\n",
       "      <td>spore</td>\n",
       "      <td>WAxNA</td>\n",
       "      <td>ancestral</td>\n",
       "      <td>...</td>\n",
       "      <td>(ancestral, evolved)</td>\n",
       "      <td>(WAxNA, WAxNA)</td>\n",
       "      <td>(, TOR1)</td>\n",
       "      <td>(+, -)</td>\n",
       "      <td>(, TOR1*)</td>\n",
       "      <td>(lys2-, ura3-)</td>\n",
       "      <td>(A, A)</td>\n",
       "      <td>(1, 1)</td>\n",
       "      <td>(spore, spore)</td>\n",
       "      <td>(no, no)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Layout  Plate  Row  Column env_evo env_test               MATa  \\\n",
       "31488       1      0    8      20      RM       RM  lys_218_R1 a ura-   \n",
       "31489       1      0    8      21      RM       RM  lys_218_R1 a ura-   \n",
       "31490       1      0    9      20      RM       RM  lys_218_R1 a ura-   \n",
       "31491       1      1    8      20      RM      YNB  lys_218_R1 a ura-   \n",
       "31492       1      1    8      21      RM      YNB  lys_218_R1 a ura-   \n",
       "\n",
       "      strain_MATa ID_MATa background_MATa      ...                 background  \\\n",
       "31488       spore   WAxNA       ancestral      ...       (ancestral, evolved)   \n",
       "31489       spore   WAxNA       ancestral      ...       (ancestral, evolved)   \n",
       "31490       spore   WAxNA       ancestral      ...       (ancestral, evolved)   \n",
       "31491       spore   WAxNA       ancestral      ...       (ancestral, evolved)   \n",
       "31492       spore   WAxNA       ancestral      ...       (ancestral, evolved)   \n",
       "\n",
       "           population      gene genotype_short genotype_long      auxotrophy  \\\n",
       "31488  (WAxNA, WAxNA)  (, TOR1)         (+, -)     (, TOR1*)  (lys2-, ura3-)   \n",
       "31489  (WAxNA, WAxNA)  (, TOR1)         (+, -)     (, TOR1*)  (lys2-, ura3-)   \n",
       "31490  (WAxNA, WAxNA)  (, TOR1)         (+, -)     (, TOR1*)  (lys2-, ura3-)   \n",
       "31491  (WAxNA, WAxNA)  (, TOR1)         (+, -)     (, TOR1*)  (lys2-, ura3-)   \n",
       "31492  (WAxNA, WAxNA)  (, TOR1)         (+, -)     (, TOR1*)  (lys2-, ura3-)   \n",
       "\n",
       "       tetrad   spore          strain contamination  \n",
       "31488  (A, A)  (1, 1)  (spore, spore)      (no, no)  \n",
       "31489  (A, A)  (1, 1)  (spore, spore)      (no, no)  \n",
       "31490  (A, A)  (1, 1)  (spore, spore)      (no, no)  \n",
       "31491  (A, A)  (1, 1)  (spore, spore)      (no, no)  \n",
       "31492  (A, A)  (1, 1)  (spore, spore)      (no, no)  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load data\n",
    "hybrids_abs_df = utils.load_data(dir_data+'pheno/pheno_hybrids_abs.pickle')\n",
    "hybrids_norm_df = utils.load_data(dir_data+'pheno/pheno_hybrids_norm.pickle')\n",
    "\n",
    "# filter out contamination\n",
    "# hybrids_norm_df = hybrids_norm_df[hybrids_norm_df.contamination.isin([('no','no')])]\n",
    "# hybrids_norm_df['Generation Time'] = hybrids_norm_df['Generation Time']\\\n",
    "# .where(hybrids_norm_df.contamination.isin([('yes','no'),('no','yes'),('yes','yes')]), np.nan)\n",
    "\n",
    "# filter out blank positions, contamination and constructs\n",
    "hybrids_norm_df = hybrids_norm_df[(hybrids_norm_df['contamination']==('no','no'))&\\\n",
    "                                  ~(hybrids_norm_df['strain'].isin([('control','control'),('','')]))&\\\n",
    "                                  ~(hybrids_norm_df['genotype_long'].isin([u'fpr1Δ',u'tor1Δ']))]\n",
    "\n",
    "# remove NaNs\n",
    "# hybrids_norm_df = hybrids_norm_df[np.isfinite(hybrids_norm_df['growth_rate'])]\n",
    "\n",
    "hybrids_norm_df.head()  # show dataframe header to stdout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#hybrid_abs_df.to_csv(dir_data+'hybrids/20150402_hybrids_abs_pheno_filtered_with_labels.txt', index=False, sep='\\t')\n",
    "#hybrid_norm_df.to_csv(dir_data+'hybrids/20150402_hybrids_norm_pheno_filtered_with_labels.txt', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_evo</th>\n",
       "      <th>env_test</th>\n",
       "      <th>run</th>\n",
       "      <th>Index</th>\n",
       "      <th>Plate</th>\n",
       "      <th>Row</th>\n",
       "      <th>Column</th>\n",
       "      <th>ID</th>\n",
       "      <th>ID_plate</th>\n",
       "      <th>background</th>\n",
       "      <th>...</th>\n",
       "      <th>assignment</th>\n",
       "      <th>isolate</th>\n",
       "      <th>gene</th>\n",
       "      <th>genotype_short</th>\n",
       "      <th>genotype_long</th>\n",
       "      <th>aminoacid</th>\n",
       "      <th>mating</th>\n",
       "      <th>auxotrophy</th>\n",
       "      <th>doubling_time</th>\n",
       "      <th>growth_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RM</td>\n",
       "      <td>RM</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>P1_A1</td>\n",
       "      <td>1</td>\n",
       "      <td>(ancestral, ancestral)</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>(, )</td>\n",
       "      <td>(+, +)</td>\n",
       "      <td>(,)</td>\n",
       "      <td></td>\n",
       "      <td>MATa/α</td>\n",
       "      <td></td>\n",
       "      <td>-0.037316</td>\n",
       "      <td>1.026203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RM</td>\n",
       "      <td>RM</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>P2_A1</td>\n",
       "      <td>2</td>\n",
       "      <td>(ancestral, ancestral)</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>(, )</td>\n",
       "      <td>(+, +)</td>\n",
       "      <td>(,)</td>\n",
       "      <td></td>\n",
       "      <td>MATa/α</td>\n",
       "      <td></td>\n",
       "      <td>0.101343</td>\n",
       "      <td>0.932165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RM</td>\n",
       "      <td>RM</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>P6_A1</td>\n",
       "      <td>6</td>\n",
       "      <td>(evolved, evolved)</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>(, )</td>\n",
       "      <td>(+, +)</td>\n",
       "      <td>(,)</td>\n",
       "      <td></td>\n",
       "      <td>MATa/α</td>\n",
       "      <td></td>\n",
       "      <td>0.039118</td>\n",
       "      <td>0.973249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RM</td>\n",
       "      <td>RM</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>P7_A1</td>\n",
       "      <td>7</td>\n",
       "      <td>(evolved, evolved)</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>(, )</td>\n",
       "      <td>(+, +)</td>\n",
       "      <td>(,)</td>\n",
       "      <td></td>\n",
       "      <td>MATa/α</td>\n",
       "      <td></td>\n",
       "      <td>-0.802399</td>\n",
       "      <td>1.743999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RM</td>\n",
       "      <td>RM</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>P1_A2</td>\n",
       "      <td>1</td>\n",
       "      <td>(ancestral, ancestral)</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>(, )</td>\n",
       "      <td>(+, +)</td>\n",
       "      <td>(,)</td>\n",
       "      <td></td>\n",
       "      <td>MATa/α</td>\n",
       "      <td></td>\n",
       "      <td>0.235418</td>\n",
       "      <td>0.849439</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  env_evo env_test run  Index  Plate  Row  Column     ID  ID_plate  \\\n",
       "0      RM       RM   1      0      0    0       0  P1_A1         1   \n",
       "1      RM       RM   1      1      0    0       1  P2_A1         2   \n",
       "2      RM       RM   1      2      0    0       2  P6_A1         6   \n",
       "3      RM       RM   1      3      0    0       3  P7_A1         7   \n",
       "4      RM       RM   1      4      0    0       4  P1_A2         1   \n",
       "\n",
       "               background     ...     assignment isolate  gene genotype_short  \\\n",
       "0  (ancestral, ancestral)     ...                      1  (, )         (+, +)   \n",
       "1  (ancestral, ancestral)     ...                      1  (, )         (+, +)   \n",
       "2      (evolved, evolved)     ...                      1  (, )         (+, +)   \n",
       "3      (evolved, evolved)     ...                      1  (, )         (+, +)   \n",
       "4  (ancestral, ancestral)     ...                      2  (, )         (+, +)   \n",
       "\n",
       "  genotype_long  aminoacid  mating auxotrophy doubling_time growth_rate  \n",
       "0           (,)             MATa/α                -0.037316    1.026203  \n",
       "1           (,)             MATa/α                 0.101343    0.932165  \n",
       "2           (,)             MATa/α                 0.039118    0.973249  \n",
       "3           (,)             MATa/α                -0.802399    1.743999  \n",
       "4           (,)             MATa/α                 0.235418    0.849439  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load data\n",
    "clone_norm_df = utils.load_data(dir_data+'pheno/pheno_populations_norm.pickle')\n",
    "\n",
    "# filter out strains used for spatial control\n",
    "clone_norm_df = clone_norm_df[(clone_norm_df.background == 'ancestral')|\\\n",
    "                              (clone_norm_df.background == 'evolved')]\n",
    "\n",
    "clone_norm_df['genotype_short'] = map(lambda x: tuple(x.strip(\"/\").split(\"/\")),clone_norm_df['genotype_short'])\n",
    "clone_norm_df['genotype_long'] = map(lambda x: tuple(x.strip(\"/\").split(\"/\")),clone_norm_df['genotype_long'])\n",
    "\n",
    "clone_norm_df['population'] = zip(clone_norm_df.population, clone_norm_df.population)\n",
    "clone_norm_df['background'] = zip(clone_norm_df.background, clone_norm_df.background)\n",
    "clone_norm_df['gene'] = zip(clone_norm_df.gene, clone_norm_df.gene)\n",
    "# clone_norm_df['genotype'] = zip(clone_norm_df.genotype, clone_norm_df.genotype)\n",
    "\n",
    "# remove NaNs\n",
    "# clone_norm_df = clone_norm_df[np.isfinite(clone_norm_df['growth_rate'])]\n",
    "\n",
    "clone_norm_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summary of technical replicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# import colors\n",
    "\n",
    "# pt = pd.pivot_table(spore_norm_df,\n",
    "#                index=[\"env_evo\",\"env_test\"],\n",
    "#                values=[\"Generation Time\"],\n",
    "#                columns=['gene','genotype','tetrad','spore'],\n",
    "#                aggfunc=[np.mean,np.std])\n",
    "\n",
    "# groups = spore_norm_df.groupby(['env_evo','env_test',\n",
    "#                                 'gene','genotype','tetrad','spore'])\n",
    "\n",
    "# norm_filtered_avg_df = groups['Generation Time'].agg([np.mean, np.std, 'count'])\n",
    "\n",
    "\n",
    "# #pg.describe(include=['Lag','Yield'])\n",
    "# # norm_filtered_avg_df.groupby(['env_evo'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# population attributes\n",
    "sp_bg_dict = {\n",
    "'position':\n",
    "{'HU': \n",
    " {'parents': 0, 'ancestral': 1, 'evolved': 2},\n",
    "'RM':\n",
    " {'parents': 0, 'ancestral': 1, 'evolved': 2}\n",
    " }\n",
    "}\n",
    "\n",
    "# background attributes\n",
    "sp_cl_dict = {\n",
    "'position':\n",
    "{'HU': \n",
    " {'WA': 0, 'NA': 1, 'WAxNA': 2},\n",
    "'RM':\n",
    " {'WA': 0, 'NA': 1, 'WAxNA': 2},\n",
    " },\n",
    "'color':\n",
    "{'HU':\n",
    " {'WA': colors.fte_colors[0], 'NA': colors.fte_colors[1], 'WAxNA': colors.fte_colors[2]},\n",
    "'RM':\n",
    " {'WA': colors.fte_colors[0], 'NA': colors.fte_colors[1], 'WAxNA': colors.fte_colors[2]},\n",
    "}\n",
    "}\n",
    "\n",
    "# gene attributes\n",
    "sp_gn_dict = {\n",
    "'position':\n",
    "{'HU': \n",
    " {'RNR2': 0, 'RNR4': 1},\n",
    "'RM':\n",
    " {'no driver': 0, 'FPR1': 1, 'TOR1': 2},\n",
    " }\n",
    "}\n",
    "\n",
    "# genotype attributes\n",
    "sp_gt_long_dict = {\n",
    "'position':\n",
    " {'HU': \n",
    "  {'': 0, 'RNR2': 1, 'RNR2*': 2, 'RNR4': 3, 'RNR4*': 4},\n",
    "  'RM': \n",
    "  {'': 0, 'no driver': 1, 'FPR1*': 2, 'TOR1': 3, 'TOR1*': 4}\n",
    "  },\n",
    "'color':\n",
    "  {'HU': \n",
    "  {'': colors.fte_colors[0], 'RNR2': colors.fte_colors[1], 'RNR2*': colors.fte_colors[2], 'RNR4': colors.fte_colors[1], 'RNR4*': colors.fte_colors[2]},\n",
    "  'RM': \n",
    "  {'': colors.fte_colors[0], 'no driver': colors.fte_colors[1], 'FPR1*': colors.fte_colors[2], 'TOR1': colors.fte_colors[1], 'TOR1*': colors.fte_colors[2]}\n",
    "  }\n",
    "}\n",
    "\n",
    "# genotype attributes\n",
    "sp_gt_short_dict = {\n",
    "'position':\n",
    " {'HU': \n",
    "  {'+': 0, '-': 1},\n",
    "  'RM': \n",
    "  {'+': 0, '-': 1}\n",
    "  },\n",
    "'color':\n",
    "  {'HU': \n",
    "  {'+': colors.fte_colors[0], '-': colors.fte_colors[1]},\n",
    "  'RM': \n",
    "  {'+': colors.fte_colors[0], '-': colors.fte_colors[1]}\n",
    "  }\n",
    "}\n",
    "\n",
    "# collection attributes\n",
    "hy_bg_dict = {\n",
    "'position':\n",
    "{'HU': \n",
    " {('parents','parents'): 0, ('ancestral','ancestral'): 1, \n",
    "  ('ancestral','evolved'): 2, ('evolved','evolved'): 3},\n",
    "'RM':\n",
    " {('parents','parents'): 0, ('ancestral','ancestral'): 1, \n",
    "  ('ancestral','evolved'): 2, ('evolved','evolved'): 3}\n",
    " }\n",
    "}\n",
    "\n",
    "# background attributes\n",
    "hy_cl_dict = {\n",
    "'position':\n",
    "{'HU': \n",
    " {('WA','WA'): 0, ('NA','NA'): 1, ('WAxNA','WAxNA'): 2},\n",
    "'RM':\n",
    " {('WA','WA'): 0, ('NA','NA'): 1, ('WAxNA','WAxNA'): 2},\n",
    " },\n",
    "'color':\n",
    "{'HU':\n",
    " {('WA','WA'): colors.fte_colors[0], ('NA','NA'): colors.fte_colors[1], ('WAxNA','WAxNA'): colors.fte_colors[2]},\n",
    "'RM':\n",
    " {('WA','WA'): colors.fte_colors[0], ('NA','NA'): colors.fte_colors[1], ('WAxNA','WAxNA'): colors.fte_colors[2]},\n",
    "}\n",
    "}\n",
    "\n",
    "# gene attributes\n",
    "hy_gn_dict = {\n",
    "'position':\n",
    "{'HU': \n",
    " {('','RNR2'): 0, ('','RNR4'): 1,\n",
    "  ('RNR2','RNR2'): 0, ('RNR4','RNR4'): 1, ('RNR2','RNR4'): 2},\n",
    "'RM':\n",
    " {('','no driver'): 0, ('','FPR1'): 1, ('','TOR1'): 2,\n",
    "  ('no driver','no driver'): 0, ('no driver','TOR1'): 1, ('FPR1','FPR1'): 2, ('TOR1','TOR1'): 3, ('FPR1','TOR1'): 4},\n",
    " }\n",
    "}\n",
    "\n",
    "# genotype attributes\n",
    "hy_gt_long_dict = {\n",
    "'position':\n",
    "{'HU': \n",
    " {('',''): 0,\n",
    "  ('','RNR2'): 1, ('','RNR2*'): 2, ('','RNR4'): 3, ('','RNR4*'): 4,\n",
    "  ('RNR2','RNR2'): 5, ('RNR2','RNR2*'): 6, ('RNR2*','RNR2*'): 7, \n",
    "  ('RNR4','RNR4'): 8, ('RNR4','RNR4*'): 9, ('RNR4*','RNR4*'): 10, \n",
    "  ('RNR2','RNR4'): 11, ('RNR2*','RNR4'): 12, ('RNR2','RNR4*'): 13, ('RNR2*','RNR4*'): 14},\n",
    "'RM':\n",
    " {('',''): 0,\n",
    "  ('','FPR1'): 1, ('','FPR1*'): 2, ('','TOR1'): 3, ('','TOR1*'): 4,\n",
    "  ('no driver','no driver'): 5, ('FPR1','FPR1'): 6, ('FPR1*','FPR1*'): 7,\n",
    "  ('TOR1','TOR1'): 8, ('TOR1','TOR1*'): 9, ('TOR1*','TOR1*'): 10,\n",
    "  ('FPR1','TOR1'): 11, ('FPR1*','TOR1'): 12, ('FPR1','TOR1*'): 13, ('FPR1*','TOR1*'): 14},\n",
    " },\n",
    "'color':\n",
    "{'HU':\n",
    " {('',''): colors.fte_colors[0],\n",
    "  ('','RNR2'): colors.fte_colors[1], ('','RNR2*'): colors.fte_colors[2], ('','RNR4'): colors.fte_colors[1], ('','RNR4*'): colors.fte_colors[2],\n",
    "  ('RNR2','RNR2'): colors.fte_colors[1], ('RNR2','RNR2*'): colors.fte_colors[2], ('RNR2*','RNR2*'): colors.fte_colors[3],\n",
    "  ('RNR4','RNR4'): colors.fte_colors[1], ('RNR4','RNR4*'): colors.fte_colors[2], ('RNR4*','RNR4*'): colors.fte_colors[3],\n",
    "  ('RNR2','RNR4'): colors.fte_colors[1], ('RNR2*','RNR4'): colors.fte_colors[2], ('RNR2','RNR4*'): colors.fte_colors[2], ('RNR2*','RNR4*'): colors.fte_colors[3]},\n",
    "'RM':\n",
    " {('',''): colors.fte_colors[0],\n",
    "  ('','FPR1'): colors.fte_colors[1], ('','FPR1*'): colors.fte_colors[2], ('','TOR1'): colors.fte_colors[1], ('','TOR1*'): colors.fte_colors[2],\n",
    "  ('no driver','no driver'): colors.fte_colors[1], ('FPR1','FPR1'): colors.fte_colors[1], ('FPR1*','FPR1*'): colors.fte_colors[3],\n",
    "  ('TOR1','TOR1'): colors.fte_colors[1], ('TOR1','TOR1*'): colors.fte_colors[2], ('TOR1*','TOR1*'): colors.fte_colors[3],\n",
    "  ('FPR1','TOR1'): colors.fte_colors[1], ('FPR1*','TOR1'): colors.fte_colors[2], ('FPR1','TOR1*'): colors.fte_colors[2], ('FPR1*','TOR1*'): colors.fte_colors[3]},\n",
    "}\n",
    "}\n",
    "\n",
    "# genotype attributes\n",
    "hy_gt_short_dict = {\n",
    "'position':\n",
    "{'HU': \n",
    " {('+','+'): 0, ('+','-'): 1, ('-','-'): 2},\n",
    "'RM':\n",
    " {('+','+'): 0, ('+','-'): 1, ('-','-'): 2},\n",
    " },\n",
    "'color':\n",
    "{'HU':\n",
    " {('+','+'): colors.fte_colors[0], ('+','-'): colors.fte_colors[1], ('-','-'): colors.fte_colors[2]},\n",
    "'RM':\n",
    " {('+','+'): colors.fte_colors[0], ('+','-'): colors.fte_colors[1], ('-','-'): colors.fte_colors[2]},\n",
    "}\n",
    "}\n",
    "\n",
    "# factor attributes\n",
    "factor_dict = {\n",
    "    'color': {\n",
    "        'background': colors.fte_colors[0], \n",
    "        'de_novo_gene': colors.fte_colors[1], \n",
    "        'de_novo_genotype': colors.fte_colors[2],\n",
    "        'auxotrophy': colors.fte_colors[3],\n",
    "        'tetrad': colors.fte_colors[4],\n",
    "        'spore': colors.fte_colors[4]\n",
    "#         'background': \"#348ABD\", \n",
    "#         'de_novo_gene': \"#A60628\", \n",
    "#         'de_novo_genotype': \"#7A68A6\",\n",
    "#         'auxotrophy': \"#467821\",\n",
    "#         'tetrad': \"#CF4457\",\n",
    "#         'spore': \"#188487\"\n",
    "        },\n",
    "    'hatch': {\n",
    "        'spores': \" \", \n",
    "        'hybrids': \"/\"\n",
    "        },\n",
    "    'position': {\n",
    "        'background': 0, \n",
    "        'de_novo_gene': 1, \n",
    "        'de_novo_genotype': 2,\n",
    "        'auxotrophy': 3,\n",
    "        'tetrad': 4,\n",
    "        'spore': 5     \n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Spores and hybrids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this experiment, we cross the mutant allele to a sample of wild-derived lines with standing variation, and we evaluate phenotypes of F1 genotypes. Phenotypic variation can then be attributed to allelic complementation (that is, dominance effects) or to non-allelic complementation (that is, epistasis)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "dict_background = {'parents':0, 'ancestral':1, 'evolved':2}\n",
    "dict_population = {'WA':0, 'NA':1, 'WAxNA':2}\n",
    "dict_gene = {'RNR2':0, 'RNR4':1, 'no driver':0, 'FPR1':1, 'TOR1':2}\n",
    "dict_genotype = {'+':0,'-':1}\n",
    "\n",
    "labels = {\n",
    "    'Generation Time':'doubling_time',\n",
    "    'gene':'de_novo_gene',\n",
    "    'genotype_short':'de_novo_genotype'\n",
    "}\n",
    "\n",
    "# environment attributes\n",
    "environments = {u'HU': u'Hydroxyurea (YNB+HU 10 mg/ml)', \n",
    "                u'RM': u'Rapamycin (YNB+RM 0.025 μg/ml)', \n",
    "                u'YNB': u'Control (YNB)'}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ensemble by background and genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/pandas/core/indexing.py:266: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self.obj[key] = _infer_fill_value(value)\n",
      "/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/pandas/core/indexing.py:426: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self.obj[item] = s\n",
      "/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/ipykernel/__main__.py:117: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/ipykernel/__main__.py:131: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/matplotlib/font_manager.py:1287: UserWarning: findfont: Font family [u'sans-serif'] not found. Falling back to Bitstream Vera Sans\n",
      "  (prop.get_family(), self.defaultFamily[fontext]))\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAzUAAAFiCAYAAADP+PaKAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzs3XecXGXZ//HPNbM9bdM3CYQeCAkEAiGQ0FGKdBVQEFAE\nsTwWCIKKPiqCIgiICP589KEF5NHQDVXpCURDh4B0EpJs6qbuJrs7O9fvjzObDMuWKWfKTr7v12te\ne2bmnOtcM2d291xz3+e+zd0RERERERHprSKFTkBERERERCQbKmpERERERKRXU1EjIiIiIiK9mooa\nERERERHp1VTUiIiIiIhIr6aiRkREREREejUVNSIiIiKSU2Z2pplNSCxf0816B5nZcfnL7GP7/pSZ\nTTazq81soJmdnsjnQTP7fGKdaxKP/dnMLjaz85K2383Mokn3r0n8PMjMjk9+3R2Whybifidxf08z\n+4WZ/cbMqjN4HTuY2cndPH+NmY1pf02loqzQCYiIiIjIFuFUMzsYGGtmI4BTgd8BvwQeAaYCg4DH\nzOz/Ac8CzwNnAx8Bi4H3gX2AUcDvgW+6+4/N7FrgBeBlwIA9gP0SMeYA5wLzgXXA48CxwC7Aj9x9\nbSK/T7n7D8zsP8A1wGvuPt3MTgAOMLOZifU8cRsIvJf0+j4LLADWJO7vmShUdgQeS2xDUoxgwX15\nosg5IfHQF4EfAFOAw4H7AMzsDuAJYFLifdnP3b9sZpcAi4BT3P3QxGv7o5k9CdwLjAHmAUPc/efB\nLv1tMzsLuLOzA9UbqaVGRERERPLhdne/FnjD3euBwcBXgNsITsQvBR4kKEo2uvutQB/gX+5+DTDJ\n3ecCuwNL3H0xsCbRKvE4QaEQSWxPUox+QH9gNTAEqAHiQDOwW1J+5QDuvgb4EPjfpOeuAb7H5mLk\n7+5+AbCPmfU1s+8C+wJfN7NjEuu86O6/A+5K3O9uxntLytuTHkveZoW7/w/Q7O5/BD40s37A0MT9\nBYn1+rr7BoKi7LfBS/LrCYqw9ribXm+pUEuNiIiIiORaZyf0twO/cvf/SbSCfBPYCphNUHQAvAZ8\nwcxGAv82s88A/yBo7dkBuBF40N0nm9l44AxgFUHLRXuM14HlQFVieSywAYgmbu0au8rd3T80M0t6\n7NjE/t919/XAtWY2APhDUstPx9d/n5n9FGgF/tn+RKKL2VnAeDP7J/B/wM+AauCnSTFaEz9bEj/j\nBAXKcjP7GjDSzHYG3uywfnPS+u25kHgPSoa5d1c0drOh2YfA6MTdse7+VlhJieRDh89wnOAP3nPA\nBe7+fqHyEhERCYOZTSXoxrQfQYvHIoKWkPPdPVbg3CLATwi6Za1z91cyjHMusMrd/5bi+hM625eZ\nnUlwDtDf3Z/PJJewmdlBwAfuvqCL589091sSx3lXglaoR4F57r6xh9g7E7zWuWHnXSgZtdSY2QEE\nJ4PtFdHpwI/DSkokT9r7xM4EPgCOIOjPOgA4rIB5iYiIZMXMvgBMJ+iO9QowF9iW4NqSHwOdtSaE\nuf8fE3Thau9e1o/gOpDDCVpElgMvElxDc7SZ1QCHELR0rEp1P+7+RzM7ycz+SvA66wleWxOwMrG/\nfxNch/IIcIyZrQYuIHh/jgIeInGdSZYvOyNm9mWCFqonge0Jusq1tyr1NbO9gHEE71Uf4Blga4Lr\ninD32QStWykrxcaITK+pOZ3gZPAlgmavU0PLSCT//tfdvwd8P3F/50ImIyIiko1Ed6bfE5znTXf3\nie5+rrsfQXBxfFO+UgEeBnYAVrv7s8BGgovW4+7+d4LrPD4kuB5kBR+/xiVV7QXAQqCNzdeieCLu\nMoICan+CgQaMoMgblNhuTEavLjzlBPmdBDxN0HukvatYI8F599MEr+cVgm5j8U+G2bKl3f3MzCoI\nquBa4FDg7sTywe7+TOgZiuSImX1A0OL4AEFLzacJCpor3f0HacY6E3jZ3V8xs2vc/bwu1jsIGODu\n92eXffrM7FMEo76cAvwCOIbgj/1FwI3ufmdi9JV7Cf6AfgA0JS7OxMx2I7i4sy1x/xp3Py/xmtr/\nBpyX/FxieSjwI4Im9N+Z2Z4EI8RUAz9JXMyYzuvYAdirq64GidfwB2B3dy+ZUV1ERFKV+Hv/KMFJ\n8C7u/k4BcjgjcZG+SF5k0v3saIKqeom7P5W4sOtLiZuKGumNjk5ajhF8o5MJDVW5mYaqFBEpnGFJ\ny/MLkYAKGsm3TLqftXc9az8BuofgJOskMyupoeFki3GCu0cJhmIEuDLRGpAuDVXZNQ1VKSKSP8lf\nzm2T652Z2VfN7AQzO9LMLjSzGVnEusLMxpvZednEySczOz9x+7uZVVgw0ehliQk0ByQm3jyliPK6\nJPH8kWZ2Rr7zypW0iprEUHVHJe6ebWZxgu5nEFxcfWyIuYnkS/tJ8ItsHs4x3X8CXQ1VeUxilJX2\noSoPTaybPFTlPhbMSJw8VOWApKEqp7n7fQTXsJ1K8DuYHCOUoSrZ/D5AMFTleSSGqkwUa88RXMA5\ns7MYJIaqNLMf0flQlQea2Tg2D1V5XPJ6aKhKEZGwPEswrDHAj5OHIjaz0ZY0632IIgRf1DUCT5rZ\nThnGmUdw7UtTlnHyyQm+yHwOODnxWJzgMo81wFsEPSmKJa9IoiGisgA55Uxa19QkTiz+H0H3kyeS\nntqVoAvIfe5+YqgZiuRIh2tq3gf2Iugitg7Y3t1XZhF701CV7j4rizhpDVXZQ6wxFNFQlanSUJUi\nIukzs1OBWwiKjVcJRgAbRTC65/Au5lLJdF+HAjsBfQm+vLsR2Nbdr8og1hkEXzKeR9C7IaM4+ZTI\n+X42dwWvSDy1EFhP0GvjD+7+ahHltZxgpLXd3P3SfOaVK+kWNU8RVM9XuPsPkx4/kKDIaQFGpjMU\nn0ihJBU17VYTtJz81N2fKkxWIiIi4TCz/QkGg2mfp2YhQaEwrdDz1IiELePJN0VERERERIpBpvPU\niIiIiIhsYmY1Zja65zUVK2zFmlc+qagRERERkTAMJejqplj5V6x55U0m89SIiIiIiGxiZpMIJnUe\nbWaNXYyUqVgfj/c5YFm2k9cXa175pmtqRERERCRrie5P+4Y0YmfJxwpTseaVTypqRERERCRjZpGD\nKK/5IcHUAeD+Cq1Nv3SPpzWsvgXzIXy6L2UXRWC7YEI2n7uOtl+5+8tpxoqUlZUdNaKu7qKysuhW\nbW1x2uJtsxctWny5u89LM1a0qqrquKFDh15QFo2OiLW1eVtb2xOLFy++0t3fSjNWWXV11eeGDh58\nfrSsbFgsFovHYrFH6pcs/Y27v59mrIpy7JRKIt81bLDjbXGY2UTb1e6+oOcIH4tVVY6dVknkvwyr\ndTzWht+9gfi1icnAi5+766abbrp1eyOYvOvrwEHA8Wlsd00njx0LVHZcB/hJoV+nbqV7A35KMHdR\nZ891+bkGTu5w/xOf6aTnftzZuh1jdJdP8roEkwB/l2BekcmFfg91063jDYhQXnOb1W7XULbzCV6+\nx1e8bMKXPTrmGLd+I1dQXvObNGJF+xK9d/dI31Xnlm/lF1ds7z8q387PLBvpo61qRV+iP0sjVvmI\nurpHz/nqWavnvfyib1y7yjesafB/PDTTJ0+atHxEXd20NGJV1dXVPfPt//qvtf958w3f0LjeG9et\n9Qdn/t333GOPZcOHDftaGrH6jqgb/u/vf/e/1r3/+ove3FDvTcsX+v1/u93H7zp22ZDBg05NI1Zt\nDdFX9mJA4+ls5eeyjZ/DaD+CofFaypdUEjkujVhDa4i+uQ+1TWckYp3NaP80Q9oGUFZfjh1W6M9a\nKje11IiUEDP7BlAOvEtw0eBOBPPvPA3sAIxI3H4C3AncQTAp2+kEs08/CRxCMMHuK56YONTMpgE3\nAbsRzErcCDxCMHnXgYn9tE+UdhMw1t1/ZmbXAH8GTiWYxfgO4PPAtcA0YBHweXf/tJl9FXjC0/ym\nSiQVZvZToAHYm+Cz90N3n2ZmVxFMwHsawazfbwF7ACsJfh/OAq4Efgi8SfBZfwYYC7zria4eZrYX\nsI2735343L8JHOnun03cf5mP/z4eCfwW+E5iX7Xu/qvEujcS/E6uBda6++/M7HJ3/0Eu3yORdFl5\n9fcjg8b8NDpyrz4dn3N32hbMWuNr5n/d21r+r6dY/azssknRAd+bEq2t6SzWXbGlq97zDV9q9fiD\nPcUaNXLk7y+84Pyvfv2cs6s6PtfW1sZxnz2p4cmnnz4+Fov1ODn2VqNG3XbJJT8/6dQvfrGi43Ot\nra0cceRRK+bOnXt4ayz2Us+xRt537ZW/+syxRx3xiWvaN27cyEFHHrv85VdfP8BTaP3pa2WPHcLg\ng0dR/YlBv1qJczdLlq2mdW93/yiFWHM+xdB96qi0js81E+du6peuJbabuy/vKVYhafQzkdLyAkFR\nMwRw4B/AdcDxwN7ufnVinfEEJ2S3A4cDS4DFBIXPv4FKYFBS3Dp3b0gs/wP4JTCFYDK3ZqAWGAnU\nu/tfgUoza/+jfRTwO+B64DNJMYe4+x8JCjAS+98mhPdApCvTgXsICpKXzexE4AOgDXjG3a8gaLVx\n4DZ3fx0wgkmn7yIoNoxgkt5Wgi8O2m0DLDazYcBqd/8fYGnS8x1/H0nEGujufyBoDW13JMHvy4yk\nx8qze+ki4TKzKBb9r0jdhE8UNInniY6aNIBo+U9TiFUZxb68b2TAJwqa9liHlw0ZWEPk0hRi9a2u\nrv7c17561icKGoBoNMq1V105aERd3eUpxBo0oHbAYV/8whc+UdAAlJeXc83VVw+pG1H3qxRijawb\nPmxyZwUNQFVVFVdffunQrUaOvCSFWDsMoGxcZwUNQDkR9mPg0BqiP04h1u5DqNi+s4IGoJIIk6gd\nUkP0gp5iFZqKGpHSMghoAnYmOGH6DEGrzP3Av83sfGAvgpOyeGKbmUB/oIqg1aYW2AjskhR3npnt\nkFiOE5ygGbA7sB6IEvw9qTOzcwiu14sl1nuQ4NvobyX2ReLxejP7PEEhBcG32K+F8i6IfFLQPX/z\nZ/cO4McELYsA+5vZRQStKLD598MJWmaOAE5J3B9G8LlP/h15GdjF3ZcB1YnP9vCkGPDx38f2xxvM\n7JsErTXtjz0MfIlgNCMSXxCszeK1i+TCaKseVGmRrgfStbIqrKy61sw6LVaS7Lx1pKo8Yp2eVwPQ\n38qosMgIs25WCkw44tOfqoxEuj7F3XHHHaiurtquhzgAk44/7ri+3e1ywoTdqSiv2DWFWFNO/uwJ\ntd2uMHkfomXRST0FMjhoJ/oO626dramySNDzolsV2BE70Wdod+tsR03U4OieYhWahnQWKSHu/nD7\nspmdAdzu7q8mHvpXh9XPT2zzFvCzpMdf5ZNuB070j4+qkvxt0h2Jfb7p7n9Kyuf8xOLFnWz3i8TP\nOxM/V7n7ik72LZI1d2//3N0HYGafBW5090bgqcSt3b1J27V/hn+U+Hlr4uezHeK/b2ZTE8vt3cTu\nbI/Rze/jZR3itO9vU4FvZgcRtBKJFJM4KVzC4Lix+UuCLmPFSSFWaldMxGOxWI9rxuMpBYvHYm3d\nrujuxFNLLN7W1n2s9nipxEq8+V1WW/6xH93sD3pMKo5jPR/DglNLjUiJcvdbk06gso3V6ikME5l0\nQpbJPqZnuq1Iutz9bne/PuSYXX6Gs/l9dPenUukXL5JnC3zjqiZva+1yBW9tgljzEnff2EOs/yyM\nb2yOdXNC3+CttBL/wHs+63/pkX/8szkWi3W5wquvvUZzS3MqI6DNuefeexu72+Wcf/2L1tbWf6cQ\n6+nb/3rn6u5i/ePxJz0Wiz3eUyCHf/6H9d1e3/IBTW1t+N97itWKz3yT9cu6W+c9GltiFP9Q0Spq\nRERERCQt7u7EY79qW/SvNZ2dqLvHafvo2VXENv4whVixNvjdE20N6zqL1ebOg7HlKxtpuyiFWBtb\nWltv/NUVVzZ29vzGjRv51nfPW7mkfkkqsdY1NTXd+bvrrmvq7PnGxka+d955yxctWtTjtSvuvmL1\n6jX/uGn6Xzot8NasXcsFP/rv5Yvrl/yis+c7xFrYRNucd2ls6ez5DbTxb1Yv30D8ihRivb2G1tfm\n09RpFdhIjBdYs2Ij8et6ilVoGv1MRERERDJiFX1+bRV9z4rU7TnE+tYB4GsXetuSl1bQuuFyb226\nOqU4ZtaP6A2Drfzz+0cHDtnaqogD73pj/Om2VSubPP7j9R77nxRjRUaMqLt1wu67H3nRBecPnjxp\nEi0tLdw/84G2yy6/YuXKlSvPW75ixV9SjFU2oq7ub5MnTz5o2rTzB03cc082btzIPffeG/vV5b9e\n2dDQcG5DQ8N9KcaqGFE3/P6DD5i6z/e+9fWBu48fR2NjE3fee3/rr6/53coVK1aevnbdun+mGKu6\nhugjW1M1fnf6DxxIOS0479LY8jJrVm4kflKrx2enGKtfDdF/bkv1zuPpP6CWMpqJ8w6NG19l7coN\ntB0Xc38xlViFpKJGJARmlkKLeG6tW7eOfv36FTxGKTCzVPs154WObW7pvZEC6emC96JiZsXzR1GK\nhrsXzedYAwWIlIj169dnfWIWRgwJn45tbum9EUlN+R5fCSXOK3/4Mts3fRhKrKrDzmTjujXhxOo3\ngBVrO+21lrbG12cz5KMXso4TJ8LaiccwrCaadayWWJxHr7+T16+7JetY1XVDueCDOVnHCZOKGhER\nEdniuTu8mOitM3EqPY8cLCLFREWNiIiIyIuzid8TfIMdAdhr/4KmIyLpUVEjIiIiIphZXyJlpxCt\n3BNvW0ts41/d/ZVC5yWSCg3pLFKEXnklt/9Duor/t78Vbhj6p556igULFnT5/C23ZN8HuNAKdVyL\n9b3rbfnKZiX5+zpxKpETzyRy4pkwcWqhs8k7K6+5iMr+70Tq9vx9dJsDvxXdeuoPbcDof1hFnxfN\nbGSh8xPpiVpqRPLk0ksvZdttt+Wmm27iS1/6EuvWrWPvvffm0UcfpU+fPgwdOpTBgwfj7jz33HM0\nNTXxxBNP8I1vfIOBAwf2GP/vf/87jz32GBMmTGDEiBH079+fmpoaBg8ezKOPPso+++zD3LlzOeKI\nI5g5cya1tbX85je/4fTTT+ehhx7iqKOOyksf8ptvvpmFCxdy8MEH8/7777N27Vo2btzI+PHjWb9+\nPS+88ALz5s2joaGBxsZGDjjgAD766CMWLVqU89wykevjOmPGDKZPn86+++5b1Mc1Od8777yz6D+H\nkppS+33tjpltsV3OrLzmAus38ofR0fsPMNv8fXdkwOih3rRiaOyDx2bR2ulULSJFQy01Innk7hx5\n5JG899571NbWMmXKFKqqqhg3bhyRSIRjjz2WVatWse222zJ69GiGDBnCa6+9llLs5uZmxo8fz1Zb\nbUU0GsXdMTPMjNGjRzNs2DAmTpzIrFmz2H777XF3Jk2aRENDA+PHj+ftt9/O8asPtLa2MnHiRGbM\nmMGBBx5IJBIhEgn+FPXp04fp06dz4IEHYmZMmDCB6urqTc8Xq1we140bN7LLLrsU/XFNzrc3fA4l\nNb3h99XdmbWokVmLGotqKPbewsyqiETP61jQbHq+ZgiR4RPUUiNFT/PUiIQglXlqbr31Vs4444yc\n5VBfX8+IESMKHqMUpDNPTa6PK+jY5prem8Jxd5qXBsPCVg7fN6NWulmLGrll3ioAzhw3kP1H9Qk1\nxxzKeZOkme0EHOfuV3WzzjGR4RP+LzpiYpdvnMdjxF6driGd07ClDOmseWpEtkC5PvGVwtBxFclc\n89I5rH/ntk33q+r2K2A2pcXMZgBPAh928fwx7j4T6E95TXW3sSI6XZTiV9x9OkREeqn777+fa665\nhlgsVhL7CUtvy1e6VizHcurIGs4cN5Azxw1k6siaguZSZOYmfp6V/KCZjTazU4DPmdnJwDxft7ih\nu0C+MZyWEJFcUukt0gUzOwCYAix195vN7FhgDPCou6d2QYRssRobGykvLycej5fEfsLS2/KVroVx\nLCuH79vpcjrMrDd1OcunV4E+wOMdHl8BzAF2Bea4+wKr6FPvzWuHWGX/TwRxd9rqX1gF9DyyiUgB\nqagR6dp+7v5rM/t+4v5xwGtAawFzkl7ii1/8YkntJyy9LV/pWhjH0szU5SxEZjYQ2M3dnwaqAAfW\nJ6/j7k3AfDOb6e7BuNyxjSfF3n3ksejo/UdZ37pN1zZ5bCNti59f403LZwBfy+drEUmXihqRrrV/\n/dh+EVzM3X9nZj8DftZx5WnTpjFlyhSmTJmSp/Q+qb6+vihilIJiOJ7JdGxzS++N5FuOBqf4BkFL\nzNPA4MRj73S2orvP3bQcb3vLzCa3LXj6Cix6iFX0M+It5q1NK4nHfuWx5tvMTEWNFDUVNSJdm2Nm\nFwJRMxsF/NvMzgde6mzlq67qcnCZvNAIWeHKxfF85ZVXmDBhQtrbpXpcuop/yy23cPjhhxfFsX3q\nqafYbrvtGD16dKfP33LLLZx55pl5zUmf+/Q9+eSTbNe4ktF1w2Hi1E+MWlaI4ygAPAEsAHD3/01n\nQ3dfBJxmZuXesm4IsMHdV+cgR5GcUFEj0gV3nwXMSnropkLlUiraLj4bgOhlfy5wJuHR5Jud25Im\nbSxlXR3HcZXG2tefY26sjTdHP8yqAUN0HAvMzCYQdJP+EPhjpnHcvRVQ06X0Ohr9TEQkS5p885N6\nw6SN0rNuj2NFGbe/9g4HTpyg41gE3P0V4DlgY6FzESkETb4pEoJUJt/Mtd7Q/ay3tNRo8s0ti96b\n9Lk7vDg7uNNJ9zNJSehvmpkd6+5/DztuIrZvfdh3Qon10GWfZ/SGhaHE6n/IqaxtWBlOrEGDWdIQ\nTo+7preeZ9Di7AdKdYuydtdDGFKd/eSbrXHnmekP89aNM7KOVT18MN/61wOafFNEpFRo8k3ZEpkZ\n7LV/odOQTxpsZme4+625CP7spceGEqdpwFBWj9o+lFgAjWvCKWoA4h+8Ekqc5lZjebw26ziOYa0x\n1pH9fFCxuLP91HGM2fGTQ3ennVd5JZz0QNZxwqSiRkREpBdwd5qXzgGCOV3UOhKuEml9GgrMy1nw\n9R+FEmfx0O3oVzsolFgAg/tUhBerKZzLidavjVDx4lNZx4lHyqgYtxtDI81Zx2oBmpYtpOz5J7PP\nq0/2hVHYVNSIiIj0As1L57D+nds23df8LuHyF2bh068DwNyxvQ8ocEYZ6QPsDjxY6ERE8k1X84mI\niIi89wasWBLc3nuj0Nlk6lF3v7zQSYgUglpqRESk5JRIV6KPqRy+b6fLEpIdxsGQus3LvdPeZrZj\nJtfUWPBLsvfgSNk0YFfDNmzwttsaPX6Lu68NP1WRcKmoERGR0vPibOL33AIkuiSUwEXtZqYuZzlk\ne01lU+07cWpBc8nCOmBMuhuZWfnASNk9Y8qq9z2mZvDg7cqqaPI4zzavnfBA08ofVEeiX8hBriKh\nUlEjIiIiW7wwRnSLx+Osf+tmAPru/OVCzNnzJeDhdDcaFCn704k1Qw49vHpQdftjAyzCUdWDKqdW\n9h/5k9Uf/m1JW0uoiYqETUWNiIiUnolTN1802nu/dZc8CqPL4vq3bqbxvTs23e8/9qyw0kvVTaQ5\n/42ZDd46WnnEp6sGVnf2fP9IGaf3GT7syrXhjHwmkisqakRki9JbJgCV7JgZbXffDEC0BLqeSe75\nC7Px6b8DwBxs7175udkWoH2uGjPbCTjO3a/qZptDDqoaMLC7Im5iRV8NLCVFT0WNiIiIyHvzgpHP\n2pczKGr67vzlTpfzxd0vbV82sxnAk8CHna1rZse4+0wgWm6Rbqer7/3DbMiWQEWNiIhIDsXjcZjx\np+DOSecU4joLSYFvPxb69Nu8nIFIJFKILmddeRaIAZ8B7mp/0MxGA/sBR5pZDfDc7I1rVh5ZPWh4\nV4Heim3Ida4iWdNfVhERkVya8Sf8/tvx+2/fXNxI0TGLQHUfqO4TLPd+/yBoZLmmw+MrgDnAAmCO\nuy9YEW995ZWW9fHOgrR6nJvXL1mR21RFslcSv7UiIiIi2TAD+g2AfgMogWmNAM4ANgATkx909yZ3\nnw/MdPcFAA3x2Mk3rFv82t2Ny5vWxmMAxN15vaWRH6/+cHlDW+sF+U5eJF3qfiYiIpJD/vmzYfGC\nzctSnEpsxDx3v7CH5+cmLa8xs8n3bVh56mMbV38nCgPj4G34Ew3x2K/d/S0zuznnSYtkQUWNiIhI\nLr34LLz58ublSQcUNh/pVBjz1ACseOabAAw54IasY+WTuzcTDAl9U6FzEcmEihoRyam2E/bs9rHo\nvS/lMx2RtLg7sxc3ATB1ZE1Gc5cw62FYsXTzsoqaoqUh30V6LxU10quZ2TAg4u5LCp2LiJSe2Yub\nuGXeqk339x/VJ/0gI0ZDecXm5Sz01lYAEZFcU1Ejvd0UoByYkfygmVW7u8agFOllQmkZKTJ28jmb\nJ/o46ZyC5iKSjeV9RoUSp7m1lbWrV/W8YopWrm8OL1bVsFDixHAYPznrOG0WIdYWYTmV2ecUh9ah\no4hPmJJ1LC/PPp+wqaiR3q4ZqE5+wMzqgG3NbGt3n9H5Zj0zswMIiqal7n6zmZUDdwLnuPuybJIW\nkc6F0jISoqkjazpdTkckEqHt1eCa7Ogp54aSV6G5O7w4O7gzcWpJFJ/Ss62nHFHoFDo1YvudQ4s1\nctKhocWS/FJRI72SmQ0G6oD33f2tDk+PA4YADVnuZj93/7WZfT9x/zTgoSxjisgW6IF3FzB1q+FE\nV69m9uzZHH300YVOKTsvziZ+zy1AYm6IEC6wl+L3xzkfhhLnsGHOyBF1ocSqqa5mQ1NjKLGqa/rQ\nsvSDUGIt+s8bDHzn2azjxCMRmg49iyGDB2Udq6WlleVPPkjla7OyjtVW049tLrwq6zhhUlEjvZK7\nrwRWWmCXxGP/Sfx8zMxOBwab2YHu/nSGu2mfiKz9K8jdgOHAUuCejitPmzaNKVOmMGVK9s26maqv\nry+KGMl6asgPe389qW1pAWB1D/sthuOZrBiPbS5sb86xIzyxvIb6+rUZx0r1WEPX783cFTHumt8K\nQENDOZPv9nmmAAAgAElEQVSGZPZvc9KwQXzroVm0fHAWv/zlLzM+Ft7c0m2++VLR0EBN4v1tamig\nJYt8fN5PALBxvwglt2yk85npSqrHaMSIERnvozNmdibgAO5+a6jBRXoBFTXS250M1Lh7xyEo1wEx\nNhcmmZhjZhcCUTMb5e7TzOwMYHZnK191VWG/saivr8/6n2QYMTqy+17u9nm/N9z99aStIrhgu6fX\nWejjmaxYj20uuDuDPLimZsSI7K6pSfVYd/feDGpbz4YPVwbLAwcyYkTfjHJZWVFOJBKlurqauro6\namtrM4qz4t3gNQ0p8LH0umNhUPDNcVWW3c+K5TVB6p+Z7hTq9bj7LWa2DzAmrzsWKRIqaqS3Ww58\nrKO7mZ0KNAKr3T3jNtbEtrM6PKZvv0RyqNiuqfEultM1e+FSrjtiP6I/ua4kup+FNaeLhO4o4PVM\nNjSzPv1rB36juqbPOVVVVX1aW1uJxWIvrFy25JJYLPZCyHmKhE5FjfRq7v54+7KZneju97j7XwqZ\nk3xc7Pg9Cp2C5FGpzfNhZgyojG5aztTROwZDOUdra3t9QSNF7SEg7YsvzGzokOF1z3z5W+dvf+Rn\nTy7v07cf7s5br78y6reXXDx18LDhl+QgV5FQqaiRUrKDme3m7q8VOhEpLpoAtPeYMqKaeSs3blou\ntLDyKbmBAii9ArZEnA7MTXejoXUj7vzF7/93zIS9J2+q3M2MXXbbg9/fcd/gb55y7E8almvQTylu\nkUInIBKiN4CBZvaVQiciIpl5dnETj81fz2Pz1/NsYr6aguZTv4HnlwS3Z+szn/pq6lbD+fYjz/Gt\nb32LqVOnhpihyMf8BkhrcBwz22GrbbffJbmgSVZWVsZ3f3LpkFCyE8khFTVSMtz9QXd/upNBA0Sk\nl5jX0MySxhhLGmPMawhvQj2RUmdmE4CvAAemuP4xicVJhx51XLdd1sbtsVeW2YnknrqfiYhI0dh1\nUCV1fco2LRdaGJNvQukNFFBM3J3ZiVa9qSOzGzGvN3P3VxKTRHc7goOZjQb2A440sxqgZcOGxhjd\nnBPGYrFQcxXJBRU1UjLMbHtgR+Btd/+wwOlIESnraVjpPOUhPdt/VJ9NJ6XZFBFhMbNQRmDTQAG5\nU2wj5hXYEcCLPayzApgD7Jr4ufaBGXes/tK536nrqiB86pEH2oBomImKhE3dz6SUnAlsDxxZ6ERE\nJDPtRURycSMiKZsLdDvrp7s3uft8YKa7L3D31U2N6++e/odrGztbf/nSem64/OcaJUCKnlpqpJRE\ngI1AXaETCUNPJ3Tual9IlYaVFildYXURLBH7AAuA7punAXffNEra8iX135lx858GzJ391OFfOvc7\nQ3fadTyN69fx6H13Nt7/f9MbVjesPJ6eW4BECkpFjZSSh4BTgL8VOpFSdPajCwH48+FbFTgTCZuu\nSShei+/Zr8fHR574XL7SkeL3KBn0qHX3NuBLZrbzgvfe/W60rGw3oHFVw4pbNjY13e3uzfq7ILlk\nZoPdfWU2MVTUSEkws/7A7sB/A+cCswubUfbUEiP5omsSpDfT5/djBpPFZYLu/hbwzfDSEUlZHaCi\nRgQ4D9gJGAa8XeBcRERkC9JZi1rHx/LUolYDqElFio6Z9QV2cveXEvcPBYa6+1/NLALEzOwooBJ4\nDhjj7s+ksw8VNVIS3P3niV+KkWgADJG06JoE6c2mjqzh2hdXbFpOR9sJe/b4ePTelzJPLv/6A22F\nTkKkE3u7+5NJ92sSN4Cd3f1NM5sIxAgK83lmNt7dX091BypqpJRcA9QSDDv5pQLnItJrhDVscanS\n9WTFzcwYUBndtLyFGwEszFXwyf03hhIntraJ+sZVPa+YovoP3g0t1sIlK0KJ0+TVbNi682vi0uFA\nZOUyWtavzjpW3J2GmmGwewiDxEYiwFUZb+7uM5PvJh67I3kdM0trlB8VNVJK3gGagXWFTkREMlNK\ngxZ01grQ8bFe1gogxW8COZxPZo9dd8lV6KxsP25CeLEmTAotlgTMbDCw3Mw+QzA/0nvufk/78+7+\nny42TavVUUWNlJKngT4Ef9SlROlb89KWzUXfJdiVSCRddwJVuQrevCKcRqDFb7/FwDceDyVW7dmX\n0XTvb0OJVXPC91g3/WehxHq3eSvW/fnG7AOVl7HtZRcyeOm8rEO1lFXz0DMNzPntzVnH6jtiKL9c\nNLfnFQPj3f2pRHHziLu/luJ2b5vZLsAoYJ67L+luZRU1UkoOBG4Dsv/Nl9Akn0S2XXx28Nhlfy5U\nOhIyFRIixcPdZxQ6B5FOvG5m44Ed3P0mADM71d3/0sN2u7j7E4luaNuY2Rp339DVyipqpJQMBr5L\n0DfzkgLnIiIZ0KAFpa+Y5t4pu6/HOSpTGh951Gfn9Bynl4zSnxilaiugCfjINb+AZMndV5rZHu0F\nTUKjmZ0ArAa2dvfpnWza3v+4ARjt7v/qbj8qaqRkuPvPC52DiGRHgxZ8XCmdLEtxM7Mdt64bet2u\n24/eY/yO20RWrV3vb37wUeOwQbXXLl+15rpC5ye9XtzMLKlIXklwyUBfoLXjyomuassA3P3RVHag\nokZERCQHwmoFCFMpDcQQhtjxaQ2u1KVFd+8bSpxCKS8rm7D9qLqH7/j1D+r23GWHTY83btjIz/94\n+6V/ffjpKfUrGgqYoZSA54E9gRcB3H0WgJlNBdZ2sv6umqdGtliJqn4sEHH3pwudj0hvogEYilPY\nJ8vZDMQAPY/olsn1U/98fhWTxvYDYO6b6/jU3gPTjiHZqRsy8I5//PGXdVvXDf3Y432qq7jie1/t\nt2LV2iNueyCcC/tly+Tu60gUNB0en93F+mkVNKCiRkrL1cA9hPTlp5kdAEwBlrr7zWb2bWBH4A/d\nDD8oRUiDFUghJLcCrG6N8e1XFwFw3e6jqC0vzL9fd2dNc9um5WIwaWw/Lv6fDwC47GvbFTibLY+Z\n7X3y4QcO61jQJPvJ175Yq6JGwmRmJxHMLfg34HPAu8ASgtac+e7ec9/bDjTzupQEM+tP0LT5JPBE\nSGH3c/dfA8MA3P064F6Cyc1ERHod62JZtjxmdkxiccyBE8d32zy23ai6PGQkW5gq4HGCHjaPAebu\nbxPMN7gsk4BqqZFScR4wgGD0s9rE/WzFk++Y2XBgqrtf2tnK06ZNY8qUKUyZMiWEXWemvr4+ZzFa\nmpuz3kdtSwsAq7OIkW0eqeZQDMczWS6PLWT+vg7Lcr9hyiaXXL831767nEvH1m1a/unYERnF7Ew6\nMRpWxai2tsTyKuqj6c1V3NN7nEou3Q1+cM8zKwFYvDi996U3fH47xhkxojDfj5nZaGA/4EgzqwFW\nLViyrInggu1OrW/qchRdkU71szJf//G5M5e6e3J1vBE4DHgJ2APY1szaCFpqUp4AJ5mKGikVvwW+\nCNwBfD2kmHPM7EIgamZbAX8C/mFmY939zY4rX3XVVSHtNjP19fXd/pNM5ZqJ7mJUVAbbZ/OPuK2i\nIusY2eaRag6FPp7Jejq2YcTI9H1NZbrndGNmen1Pprnk470Z27+KC+ctBuBzI2u7zSfZ4hT2k05e\nxwyPU+/BNTXHjBtIJJJeh42e3uOwTtTTjZPx34UO1wB11j01lYhhH6ccWwHMIZjZfQ6w5K+PPL3m\nv889tW9FeXmnG/zvvY9uJIeTekrpWU8b50U3dye9pu2D4cnPdzOn0qxM96miRkrF9wiudxkGvB1G\nwMTIHMm/XEeFEbc32XP6O90+9tLpO+UzHSlCxTjCVzGaPLCGH84Lvqm/YtzIlLfrOF/Lime+CcCQ\nA27IKI9n6zfw/JLgW/dxg6sKMnx2x8EPSmGggOTjlO0xyjV3bwLmm9lMd18AUDdk0GWn/uDXl/3l\n8osGdixsZr80z6+efvcCYEwB0pVerCaa1Mk1lW+dsqSiRkqCu//czC4G/gn8u9D5SHg6K6w6Pq7i\nSsIU1mcueYCKOScewt7bbMDG783cz3+e7U85JftEM/SfhuaC7bszGiigMNx9UxefJSsa/jBsUG3b\n+M9946dfPOrg/pPGjem7trHJb7r30WXvfrT47frlDZ8FlhcwXemFasqSWoJbcr8/FTVSSn4LXATc\nTHDhmYjkWFjzfJS6vhXl/PEz+xP9yXXMnt3pCKZ5MXVkDXV9yjYtp6unljm1yvVeyxpW/4+Z3XzV\nrXcdPaBvn92aW1rXrW1sutfdPwC2+DmNJH0fa6nJAxU1UkpmAdOBCYVOREQk2dE7jgYgWlvL0Ucf\nXbA8zIwBldFNy8XgT3+v56LTtt60fMEXti5wRlsud28hmBrhnkLnIr1fv7L8DrKsokZKyf7AD4AX\ngN0KnIsUKc1PU7p0fU/u5aJlbqetqvnFLQsAOHq/QSlv19M1f7Bld001s1OAdcBCd3+10PnIlqcm\nmt+iRvPUSCn5MUFBUxxj8IrkwdmPLtw0WphIPj2wZA2rW2Osbo3xwJI1GceZOKYvr7y7nlfeXc/E\nMV2OKizp2wi8D3yj0InIlqkmGtl0ywe11EgpuQY4gGDYyfQmXxCRXk/X9+TX1MF9+PariwC4bvdR\nKW/XcUS3B392OHvuuhWVQyfyYe3nmXxi4QZRKDFPuvsaM7sprIAL33svlDjNrU7z1pNDiQXwQXSb\n0GK9Xx7OJbmxPlVUfi2EetJgxWpoaNs2+1geYYcpoxm9y8+zT6usjF+edVq361RU57fMUFEjpeQC\n4G7g+8CFBc5FRGSTB95dwNSthhNdvZrZs2cX9LqabCSP6Ma002H+BmzCZKLXX0+0trbrDbtRU1PO\nNRcezOApVxV0EIUSdIEFF049HlbA2ucfCCXO8m0nE33w/0KJBdB0w3WhxVp44WWhxBnwna+x4PLr\nsw9UUcF23/4yy3+XfdfpshHD6HPkoSz78x3Zxxo2pMd1ymtU1Ihkaq27zzGzwwqdiIj0Xi+fkcJ0\nHKend3XO1K2G8+1HnsOWfIvrrw/hRKcIzF64lOuO2G/TiG6ZFmqH7xd8y15b4EEUSlAEiAPpzWIr\nEpLy6s4nc80VFTVSStaa2W+AcL5KkqKQixNMCWhyVclGWCO6PfrcfCaPr6MszVYs/W3oUSgTUYtk\nqrym66LGzE4CaoG/AV8gmJ7zRWAc8Li7L0p3fypqpJTEgUsTPyUEPZ40bNknDEKH7khA28VnB48X\nYJS5sHLZ49bwzwXDatUoRZPH13HRNc9QeXvptGIViTuAk4FYqhskuqvtZ2Y7uXsT8Ji7N+QqQSlt\nZd13P6si6Bo5FmhKPDYVeBfok9H+MtlIpEgNBr6TWL6kkIlIeHJxgimSb8UyT00p0d+GHt0FPEfw\nDXiPavv2OX6rYYN/e9he4/tP2W3ngQ3r1rf85dFZq7YaNnjOouUNp+c2VSlF3bXUEIzOdxjwElBD\n8IX0y8AwYEcyaGlUUSOlpH2SilYzG+zuKwuaTQnQSYOI5Nr/m/EqF5+zD4MmX8Zvf/tbfvaznxU6\npVLxBMGXfVf2tOKQ2v4n77bj6Bvuvfz7g/vVVLc/XHn+F46pu3/W88d8++obZy1ZuTqXuUoJquhb\n0eVz7j4j6e7cMPanokZKyWEEf8Q/AxxKMBraFk3XTIiUlkyvPynmiSq/ftLuQfez+36k7mfhWwEc\nB9za1QpmFt1q2ODf/P2KiwbXVFV+4vnj9t+77PX3P9rx5/87o5OtRbpWXt11UZMLKmqklKwE7gd2\nA9K+wExECqeYT7qLia4/kVS5+9UprnrIiQdO6t9ZQdPuGyd+uo+KGklXWU3Xn6mc7C+vexPJrdnA\n1cBM4JUC5yKdyOTi8Y4nsmc/uhCAPx+uUUqzpYEgOqfPXH796/Ul/Pq8AzbNU6NrjvJu+JjRI/t1\nt8LAfn3zlYuUkLJqFTUiGXH3fwL/LHQeIrmmVo0tV6YFQDEPf5zpPDXJn3EVnukxswPc/ZnE3Q9f\n+M/7qzmeQV2tv3jFqjxlJqVERY2IiOSFBoLofTRRpYRkWNLys/+c++qqFavXDhpS27/Tla+8/f61\nQOdPinShrKYqv/vL695E8sDMxrn7vELnISIC4O6wbs2m5WAqkMwMOeCGjLZTAdszdyfy2S9vWs7m\nOBU7d78radlrqiq/8qnv/uKu+3594dBt6oZuWq+tLc7vZjzUdPdT//o38KlC5Cq9l1pqRLI3sNAJ\nlAp175B8KebuUVl7cTb0r928vNf+hc1HOvfibOL33AJABDI+TpkWntkyszOBUQDu/st0tm3a2PxM\neVn0iMO+fcmfRg8fstXeY3eoXL2use2JF+dtjLW13b5k5eqLgdZc5C2lS0WNSAbMbPeku2sLlohI\nHpR0ASDSi7k7zUvnAFA5fN+8tva4+y1mdhrBJIZpa421vQTsbWajZr/21g4Es7y/7O4xoKRbriQ3\notXqfiaSiT2BKMFnugV4NduAZnYAMAVY6u43m9lZBP2Qn3L357KNnw8a3Uq6U0wtcSXdPWri1OCb\n/8RyIRTTsS5aIRyn5qVzWP/ObZvuV9Xtl31e6fkI2JBNAHdfhKZFkBBEqvt0+ZyZnQTUAn8DvkBQ\njL8BjAQWZnKepaJGSkLiG6pLgHogrK8G9nP3X5vZ9xP3B7v75WZ2IdArihopLql80+nec6FZ0gVA\nCTIzdTnrBcI4Tu5OvGXz9VMFcACa0kCKhFXVdPd0FfA4MJagVRBgO3e/zcxOJ4PzLBU1UkqM4Buq\nupDitTfht5+JeoefHzNt2jSmTJnClClTQtp9+urr6z92v6eT347rd/UYQEtzc7fP50u+8ijU8ezq\ndaX7etM5thDe+1rb0gLA6hwcn3RzSyeXfLw3xaLYjrU3t2SdTyivyZ2Kec8H8cbtDRl0t/JVDdAa\nA2DVqgZW03U+I0aMyCzP7jUCO+QisEi6rLK6u6c3AocBLwE1BOdcbyZacN7LZH8qaqSUPAScQtCU\nGYY5iVaZqJmNAlaa2UXA052tfNVVV4W028zU19d38k/yk/OZJOu4fucxAhWVCzvdJt/ylUcujmem\n39x+8rh0f1whvWML4b2vbRUVGcV56fTNy2F1j0o1l3y9N8Wi0Me6oxXvBnGGZBEnjNfkL8wi/tTf\nAYgMGoRl0GqzwQaxrj44teo3aBDVdXn/zDTQxf+oMDROPTGUOJHK/lScn9ZYBl27fgbDLr0knFgP\nPc64v10bSqg1a5oZe8X3e16xJ2bU7LUHI489JOtQcYem5csYulcIdW9ZORzf/Uct0k1R4+4zku7O\nzT4hFTVSIsxsLLAj8BvgImB2tjHdfRYwK+mhm7KNKSJbHncPRj0DmDi1ZC64jl7250KnEKpPDL2d\nUZAulvNnHvC9xC10dS3LQolT338ow4cN63nFFI1oXhJarLp1H4QSZ81i2HDrzdkHKitj6H57Mbwm\n+78bLW2w8NUPic55LOtY8b4DelzHKrvtfhY6FTVSKr4B3AL8GTi1wLmIiGwW0lDBpapQQyB/knWx\nnEYEMyIVAzYt55OZjSbogn1NXncs0gWr6rb7WehU1Eip2Egwhv5cgnH6VxY2HckVjdgkqSq1lgTJ\nLTPwfgM2LWeicvi+nS7nyXZA+1nk/HzvXKSjHq6pCZ2KGikV8wiGdX4H2IMQhnSWj1MxIZKhIhjS\nWVIQwnEys0IM4wyAuz9lZhcQXHD9cCYxzGxYFZGzy4ns6fj6Rtpu9WAag4zmvpEtm0cr8ro/FTVS\nEtz9lkLnUIw0N0Vp0nHNr2zf22Ib0lmflc4V23HK0EYymHzTzKyG6JWDKT99d/oPHkJFtIU4b9N4\n/AI2LDezo3OQq5Q4L6vM6/5U1IiIiIiUhrVkMPpZDdFf7EDNOfsxsL8lXU9UR9XA1bQOfIClj6+n\nLcw8ZQvg5WFNG5gaFTUiIlJU1JIgkrH3gfOA76a6gZn16U/ZWR0Lmna1lLM3tXVP6lJVSZNaakQk\nJ3SiKCKlSn/fNjkYeCTNbT41hj4DOito2u1In3IVNZI2XVMjIiIiIhlYSDACaI/M7Bh3nwnUVBHt\n9iv1aIZDXMuWLZWWGjMbA0wGlrn7I2b2GWAg8C/gGGCOu89JZX8qakRERERKgLvf3NM6ifls9gOO\nNLMa4JX5NDWMo9/QrrZZSUt4ScoWI17WdUuNmZ0G7AscDowFTks8NRj4C3ASMAxIuQ+bihoREVH3\nHZEtxwpgDrArwbfgC/pa2YcraRk6mE+ehDrOHFY1AIPynKf0cs0x7/I5d78duD3RUnMqsNzMdiOY\nZ/ALwL+BrYGdgKdS2Z+KGhEREZEthLs3AfPNbKa7LwBopO3zD7Hs6SkM3HpbaiKRRHezdcR4loZV\nDbT+Abi4gGlLL9Tc1vPo4u7+NvB20kOvJS1fmc7+VNSIiIiIbGHcfW7S8gIz22s2qy5+jlUnVRIp\na8Mthr+/gbafx4JrHVTUSFqa27puqckFFTUiIhKa6GV/LnQKIpIBd18JnA+cb2aVQKu7pz2Rp0i7\n7rqf5YKKGhEREZESY2ZnJBZHu/ul6Wzr7s05SEm2MBti+a2JVdSIiPRiusBfRLqwjGDkqNd6WlEk\nF9a3tOV1fypqREREREqMuz9sZuXALmHFbIxUhRInFmujccOGUGIBNEaqw4tV3i+cQP0gstNOWYex\nsnJa26CxNfuUYg5tfQbAiG2yjuXVfXpcZ31LLOv9pENFjYiIiEiJMbNLgAjweFgx/+/aB0KJs//p\nRxGf9U4osQBevy2lEX9T8tRtz4YSZ+tP709rRfajYHtZlMaP5rPh3Teyj1VVTWvtCOJtNVnHsnjP\nMRrVUiMiIiIiWYoCcSC0PqovT78vlDj7f+4Ayt6Y2/OKKVpzz4OhxZr/15mhxBm5646sufeh7ANV\nVDBk0k7Yi7OzDuX9B+LbTqDl6SezjmUDB/a4jlpqRERERCRbb/e8ikju6JoakSJgZoOAC4Cm9lFj\nzOx4YALwjrvfUcj8REREenAHcDKQ36/LRRLWbszvRy+S172J9B4HA38CFptZLYC73wdcTYhN+dJ7\nmFmPNxGRInIXMBpI6apwM4uUWeTI4eWVjw0vr3xnWHnFGwPLyn9hZnW5TVNK1Zqmlk23rpjZGDM7\n3cyOSNzf1cwuMrMyMzvBzL6S6v7UUiOSkGiJOQFw4E3gZcCSno8AFwJXdLb9tGnTmDJlClOmTMlD\ntp2rr68vihilIJPjmcv3Tsc2t/TeSL6NGDEi17t4AhgMXNnTimZWPais/NGpfWt3++ygYQNGVVTR\nEo8ze/3qH05fsficftGyr+Y6WSk9azZ0PWSbmZ0G7AscDowFTgNw9zfMbJK7x8zsOeBTqe5PRY1I\nQqIl5j7Y1P1sGtAExBPfIBxIcOHlfsA/Om5/1VVXbVpO5Vt793Bn2q2vr8/6n2QYMUpF8vGE8I9X\nOnRsc0vvjZSwFcBxwK3drTSkrPzWrw3bavLB/QeVtz9WEYlwSP9B0cl9Bgz/7vz/3Jjv6yOk92tY\n33ULjbvfDtxuZmOAU4HlZrYbsArY3cyGAb8k6DWTEhU1Ip1w9wbg4qSHHkncREREip67X53KemZW\nt2Nlzf7JBU2ymmiUc4dvPexHH4U3BLNsGbprqWnn7m/zyUEtpiV+ptVCqKJGJAcK+a2+iIhIGg78\n1IBBg7tbYa+akCaklC1KUwpFTZg0UICIiIjIFsTMDih0DlL6WjbENt3yQUWNiIiIyJZlWNLyM4+t\naWjobuWXmtblOB0pRc0bY5tu+aCiRkRERGQL4u53JS3Xr2prfe7ptas67Su0Id7GH5Z+tCx/2Ump\naNnQuumWDypqRERERLZgK2Ktp92w7KO5Nyz9aO3ilmYAWuJxnlrbEP+vD99c2hBrPafAKUovlO+W\nGg0UIJIHTU1NrFixgtGjRxcsRjHkEIZiyEFEpJS4e5OZHfjA6uVHPbtu9fcNRsXxtpj73avaYte5\n+2JNMCzpatnQnNf9qaVGJA+WL1/Oc889V9AYxZBDGAqZQ1NTEwsWLADg+uuvB2DZsmWsXLkyrzGK\nLZeuYoURo1D5FFMu7ebPn89DDz2U8fZ6TenHykS2r6lQ3L2tJR6fubS1+aAlrc07Lmtt2bkh1vpD\nd19c6Nykd2ptWrPplg8qakRybO7cudx44408/PDDzJw5syAxiiEHgLvuuotnnnkmo22LIYfkgioa\njQLw5ptv8u677+Y1RrHl0lWsMGIUKp9iyqXdk08+ycEHH5zx9npN6cfKRLavSaRUtKxbtemWD+p+\nJpJjkyZNYvjw4cyZM4djjjmmIDGKIQeAz33ucxltVww5zJ07l5kzZzJ//nzefvttDjvsMB577DHu\nvfdeKioqaG5u5sADD8x5jGLLpWOsd955h0MPPZSbbrqJiRMnMmHChLRjFDqfYsql3V133cWCBQv4\n8MMPefrppzn33HNT3lavqfe8JpFS0tK0tsd1zGwMMBlY5u6PmNmuwDHAb4AfAEvc/cZU9qeiRiQP\nRo8enfU1INnGKIYcwlCoHJILqtWrV7PvvvsSiUQoLy+nqqqKffbZJy8xii2XrmJNmTIl5e2LLZ9i\nyqXd8ccfz6pVqxg7dixjx45Ne3u9pvRjFeI1FbsJpx4bTqA+fWnbda9wYjGdAccfGU6oG15nm5OO\nDiVU2ai6cPIqi+JDhhHZc7+sQ1lVDZFBo6g84ODs86qpgb/d0+0qrd0UNWZ2GrAvcDgwFjgNwN3f\nMLO93T1uZjVARaopmWY+F8membm76yrKEtHT8Uz8wX0+y31kHaPYcinFfIolFzMbBlS6+0eFzCPM\nWHpNvYuZ6YRRPqGr/5VmtgQYnvTQUnev62S9McA+wApgEbAKOA+4mqDQWe/uN6SSi4oaERERERHp\n1TRQgIiIiIiI9GoqakREREREpFdTUSMiIiIiIr2aRj8TkYyZ2fmJxUOAzwFfBMYAcYLhGA8Gqtz9\nrwXIoc3d/9vMjgSGufutucqht0vxOFYDLwKfBpoBBwYArwEjE883A+8AlcBEd78yxFySj+dw4LkU\nc7cX8kMAACAASURBVHkNuNzdv5FuLlnmm/f3Liy97RikIgfHqeCvSUQ+Ti01IpINBwYTnNycnHgs\nTjAIyRrgLWBdgXKImFk5wUmidC+V47gW2I3gZK4C2EgwWs22bP5fMgxYCjwBtIScS8fjmUouywhO\nUOdkmEs2+RbivQtLbzsGqQjzOBXLaxKRJCpqRHLEzL5qZieY2ZFmdqGZzcgwzhVmNt7Mzss0Rg6t\nBK4E2oAYwYnDO8DjZrYP8F1gYaFyAHZJrJP2xDY9Hb/E8weY2Wlmdo6ZRczs4S5idXkMzex8M/u0\nmZ3RXYwcS/U4tg+XOROoTSzfR/AtdxPBN96HAJcA9WHnwubjuXWKuRwM9AXGJ+Y7yIWO+R4A9APW\nmdl1wAN8/L2rI5ibYSpwP3ApH3/v7gNazOxHZnZW4jNxj5n1a9+hmR1kZj8wszMS9/9gZjvl8DUV\n+zFIRZif8YMpjtckIkk0pLNIjpjZVwnGWzeCExmAR939nTTjnEnQ7cEJuoymHUPS19XxIzjxbE7c\nyoH1BN8ALwK2A9529392iPWJY5gUxwm++Y4BS7qKIb1DCp+bTo93Yt1+wDxgEvAhMBCYDRxK0Gqw\nHNiBoCVhZ4KuUTcAXwHK3f2KXL8+EZFipZYakdz5ABhK0HXhMwQnMMdlEMeBWQSTU2UaQ9LX2fE7\ngeDvZgT4F7ArEAUaCE48PwAO6yRWx2OYHAc2fxPcXQzpHXr63EDnx7uSoDvUvwm6QDUk1j+OoNVr\nAlADtCYef4CgVehEgiJqhxy/LhGRoqaWGhERERER6dXUUiMiIiIiIr2aihoREREREenVVNSIiIiI\niEivpqJGRERERER6NRU1IiIiIiLSq6moERERERGRXk1FjYiIiIiI9GoqakREREREpFdTUSMiIiIi\nIr2aihoREREREenVygqdgIiIiIiIpM7Mfg9EAXf3bxY6n2KgokZEREREcsLMzgTGAVXAH9z9zRzv\nbz9ggbsv6uL57wE3Ar9w9++a2S+Ba4D/AHsA2wO1ieVmYGvgL+4+O7H9RHd/8f+zd+fxUZVXA8d/\nZyaTfSMhEBARN9wQFPcgitQFNyzWvSJtXVrr21qlbsW6o1VB39Ji27eL4lKruNeiuCCiKIprFVHA\nDYGwhpCNLDNz3j/uJJmELDOZO5lJcr6fTz65c+fOuc/MnWTm3Ps85wktHw1MBrYAL6jqQhHZjPN8\nPcB/VbUobN83AhXAJlV9OLQuG7gFqAKeBALAycAwYAYwAjgG+EpVZ4U9lbdU9Z8icl47z/NHwIuq\nur6N+44OPcccYJGqrm7n5exRLKkxxhhjjDHx9DDgAw4RkVKcRGBXYDZwFfA2sBdwJzABJ5HYBqwA\nzsP5wr8ZSAP+G7r/KWAUUAuMDm3zKU4yUCMidwH/Acao6mUAIiJAkapWiMjDIvI4cJ+qbhKRp4Ar\ngadDbdbQTy6wIey5TAY+CLv9HLAYuAxYCDwL/DDsPkL7zgcCqnqviNwTek0Ajg09l7eAO1T1GuBT\nEbkAGBR6XhWAT0REVbXVaysikgHcHXr+h6vqj4AhwF4icnMbr9+W0HN7HPgNcBO9gI2pMcYYY4wx\n8XQWcDuwAMgIrdsMHIHzXfRh4HXgUCA7dN9hOF+8XwL+ANSr6k3AGODvwI+B76nqK0Cxqt6nqosA\nCcXfqKr/BOrD2tEPqARQ1aXAWlVdGLqvCngHGB+2/Z+AXwFnicjeInI5cKCI/FJEDghtMyHUnnmh\n2xVAZuhnW6vXoTEhCbazXgFE5DggX1XfVNVXVfUG4CvgaNq2P/Ceqv459PgiYCPtv34aeg3qcV7v\nXsGSGmOMMcYYEy+NVwTOA64FdsH5/tn4EwB+BpyCk1TshdPtq7E3UUMoRkPotqhqFc6Vn6WhdetE\n5FIROYpWCULYb1S1jJZf4sOveqiq/ovmZArg58AvgMWq+rmq/h74QFVnqepHoW1eBM4GfiointBj\n78NJiMKDlwPeUPe3xu5rlwIvA6cDNwL/DHWfuyF0/34icrSIXA0cB3wcFlLClv+LcxXsp6HX8zTg\nmdA2O7x+TQFE+gPf0UvIjlexjOk7ROQbYGir1QocqKr/7f4WGWOMMbHrKZ9voa5Yc1T14043bn7M\nXjgJx5WqGohyf1cAf1HVmlbrp6jqnGhixVto7MvXbY15EZEXcBK/gKpOEJEzgcFAKbA+dNWqs/jf\nB15W1WqXm54QNqbG9HWNfWafB74MW7cpYS0yxhhjYpfQzzcRuR74Bqeb2MM4g9LfA44HqkPt2IIz\nDuUcEcnEGRD/J1Xd2lFsVf0CuDxsX2cCZ+BcySjF6QJWE4p/PPAucAgwH8gDikTk18BDwInAC7S8\natOtQoP6h+C8FrvhjONJxxkjky0iB+EUHygAsoA3cF7TXVT1dgBVnRvtflX1GReanzQsqTHG8XdV\nfa7zzUx3spKVxhgTs0R+vglO96zdgRWq+laoi9gyoL+q/jtUHe0bYDXOWJr9gU6vMrTSmACsCdtv\nY1K3Gmd8yQfAkThjUwSn61pB6HHDu/j83OLDad+ZOJXYTqJ53E01cAnwvzjdyj4GtrPjuJw+z7qf\nmT5NRL7GuTz/H5rPZKGqV0QZx0pWNu/btZKVInJeY8nK0IDP1s/zR/SxkpXGGBMJtz7fYtj/Bar6\nYHfsyxiwKzXGNDo5bFmBrvzTt5KVVrLSGGOSjRufb1GzhMZ0t6iqn4nINyISFJGAiFSKyNci8piI\nHBqvBhrTTb6vqt7QT1eTfStZ6bCSlcYYkzzc+HzrkIhcKCLfF5EJInK1iEQ9viMs1l0iMkJErogl\nTryEemY0LkfVbU1ErhSRy0TkdBH5gYjcIyLXikiBiNwbh7aKiJwsIv8Qkd+KyMUiso+IXBf6jB8q\nIre6vd9Eibakc/igs8dwzh6fCbwpIj9wuW3GdCfpfJMOWclKrGSlMcYkoVg/3yLlwRnEXg0sFJE9\nuxhnGc7Yl5oY43SJiFwlIj8RkX1F5Nci0tgL4hAR+SEwRESuEZEDgbEicp6ITBGRSSJyfejk4B9C\nnz2tKc7JxTqc7uCK83lfBnzUxvYxUcd/gOXA73DGKmXjHKtgqFv2Krf3myhRjakJ6585SVWfC325\neRg4B6erx86qWhuXlhoTB+31OQb+pqrLXNzPPap6ZZSPiaVk5SRgfuuSlT1NqGTlKmBPK1lpjDGR\na/2dLc77Gg/sifOFeTzOmM5hqjqzC7EuwDkpdgVO74Auxemq0P4X4owb3QeoVdX7Qr2SdsfpVr5G\nVR8MXbURoBynh0YFzgm4Y9rqfhfaXoGHVTUY2lc9TrIxFafU9Io4PKfrcIoopONUThuAc3L1KZzu\n2feq6pr2I/QMMSU1oXV74WSAChyvqq/Go6HGxEPYe7q1uH8IGGOMMfES+nzbGTjdPs9MX+BG38pv\nw5YHuBDPmG6jqrsmug3GGGOM2+zzzfQ10Y6pacsuYcsbXYhnjDHGGGN6GBHJFJG2ej9YrDhL1nZ1\np5iSGhHx0lwmdQtOyVdjjDHGGNP3FOFU/LRY3S9Z29Vtutr97CIROQ2nbOpwnIpDP7MiAcYYY4wx\nfY+IHIJT4XOoiFSr6vMWq9N4P8CZmuGNGOMkZbu6W1cLBSiwHdgELAHuUdX34tJCY4wxxhiT9ELd\nnw5X1cctVvdK1nZ1p6iu1NigM2OMMcYYE05E0vCknI4v6zggKB5vHRr8j6r6uxArIysz46xddhp0\ndFDV7/P5Kv1+/0vRTm0QipWFx3cuvuwxoHUinjLQBaraenLoSGLlpmXl/DB3wODDNBjc7vF6N2gw\nuEijuTrQHCs/Oy//gsys7IMCgUDl1k0b/xkMBt7uYqz+OR7vlAEpvlEBZauIfAW838VYA9Px/DgF\n2TcIm2sIzFHVjzt/ZHKI6kqNMaZvEpFCnIl2lwP5qvpshI+7V1WvaLXuVOAlVa0L30ZEfquqvWZm\nY5NcRORGnLkYKtq472jaeV+LyFnhZz7bek+H3Xe9qt7WetvWMTpqT/i2IrIL8H3gU6BKVd+J5jkb\n0x0kJW0SHt9sT+GeuZIzJAuC6LbVFcGtX1UQqJ+sgYaFkcYqKuh3fk521t0XnvuD3PFjDs1saPDz\n1AuvbHviPy+Xby7belZtXd27EbfLl/FTPL6bPf33zpWsgRloAC3/ujy4bfVWGrZ/XzX430hjZffr\nPzUtK+eq0aecl7/TXiPTGupqWbbw+a1fvrdoc0152akBf8MXEbVJRPoVDbghOyfv0lPOv7DfHvuN\nSt1eXcWCZx/f8tHbb2ws27j+ZFX9OtJYBd6Uu3M9Ked/P69/4a6pGSnVwQDzK8s2f1pbvW5LoOEk\nVV0bYSxPJt7ZmXgn7U9O/374vLUEWU7Vpg3UfVtD4GRVTfpiYG6UdDbGJAkRuRTw4UxYWYQzGVo5\nzsReuwODQj+/BZ4AHsWZKGwysBVnwrFjgG3Ax6r6Zij0j4D7gf2B40VkDDAfpwvqUaH9NE6Udj+w\nj6reFGrTfsB5QDC0vwOBRSIyDVgLjAjtY52I7KaqX7n7qhjTZIqIHIwzyd11qjpVRGbiTL47MTTv\n2hfAATjFbxYCh4vIG8B1hOZkC01uuw+wKiwBOQj4LLR8b2jbxuqgR4hIBi3/HjW07fWhfeWr6h2h\nbZfj/E1WABWq+qqI/A6wpMYkFfH6xkt6wV+9ux9fKF5f8x1ZA3M9A/bP9a968XERz7GRJBAF/fJO\nO3C/vX//7D/+UJCRnt60fswhB+ZdfelP8saf/ZPnRGSsqq7stF2+jMmSWXSHd9gx/cTjbb4juzjf\nM2D/fP+qF18UkcNU9bvOYmUXFF02dP9Drz/117/L93ibvzYP3f+QfltLV/f717SLXhGRgyL50l8w\nYOC1Bx917BWX3XRXnsfTXKtr1OFHFq5e9UXhjZect1BERqlqeWexCr2+O47Kyrvk4oLBOSLStP7A\njJz+q+pq+t+04etFIrJ/JJNwZ+GdvTfZ5x9EXrbQHGtnMorWU1f0MpsaYzV0FiuR3CjpbIxJHu/j\nJDX9cb40vQz8ATgNOFhV7wltMwLnC9kjwPHAemAdTuLzLpAGFITFLVbVstDyy8DtQAmQBdQB+cBg\noFRVHwPSRKTxv/+JwCxgNnBSWMz+qvoXnASM0P7DS8Qb47aHgKdxEpKPQsnJ10AAeENV7wKOpnnG\n709xZgs/EngSZ5Z0AT7BKZBTFBZ7F5zEfABQrqr/B2wIu7/13yOhWP1U9U9AYdi2E3D+XuaGrQv7\nxmhMkvCm3ePd9ZiWCU2IpKSTssvRRfgyZ3YWRkQkKyPjrn/dN6NFQtNoYFEhf59x68Ahgwb+LoJY\nHsR7i3eXo1smNI33p2bj3fmIQfgyb4kgVoo3xXftKVfe3iKhadRv0FDGX3TV4OyCousiiJXuS037\nn5/feGeLhKbR0D324oe/uHpQXkHhLyOIlZvh8VzQOqFptEdaJmfkDRic7fFeFEGsojQ83z+Y/BYJ\nTaNi0tiX7CE+5LzOYiWaJTXG9C4FQA2wF84XppNwrso8B7wrIlcCB+F8KWvsU/w8kAuk41y1yQdq\ngb3D4i4Tkd1Dy0GcL2gCjASqAC/O/5NiEbkYp2urP7TdPOCXwGWhfRFaXyoiZ+AkUuCcxf7ElVfB\nmB0pLd+7jwLX41xZBDhSRK7BuYoCzX8fCrwBnACcHbo9AOd9H/438hGwd+hsbUbovT0wLAa0/Hts\nXF8mIj/HuVrTuO5F4HycakaEThDs0G3OmEQSkWJJzS4WX2b722T0A0/K/iLSWVK++8h99sovyM9r\nd4PDR48kLTW1JIKmjZLsgdltJVpN7coeDOI5LoJYR+xx2Lgsry+13Q32PGy8x+P1TYog1vhxp5ye\n6/XumGg1OvrkSb7UtPTJnQUSOHFCTkFhWwlNoxNyCtIzxHNJZ7FSkB+MIKeoo232JScrDc//dBYr\n0az7mTG9iKq+2LgsIhcAj6hq42X/1l1Xrgw95gua55sCJ7Fp7RFgUqtxAeFnuR4N7XO5qv41rD1X\nhhantfG4xvEzT4R+b1XVzW3s25iYqWrj++5ZABE5HfiHqlYDr4d+Gj0T9rjG9/BvQr8fDP1+q1X8\nr0LdMlHVa0Orn2iM0cHf4/RWcRr315Tgh8b8/COyZ2pMt8kiZcerKq2JN1XVufrfUdel7AH9Czo9\n0Z6eliYiIp0Mgs+WlIwOv9+KCIi3/YwgLFZWXmH7GQ3g8Xrx+nwRxcor6J/W0Qa+1FQ8HWU9IV4k\nN8+T0mG7MjxexHndO4vVL52O95mGB3V6ZiQ1u1JjTC+lqg+GfYGKNVZDJGUiw76QdWUfD3X1scZE\nS1WfUtXZLsds9z0cy9+jqr4eSd9/Y7rZWt1eFuwov9CgH22oaQCqO4n19ZIPPvZ3FKuisoqKqqqq\nCKp6fRGsKq3raAP110LQv7WTOACfff3R21UdbVCxqZRAQ/2mCGJ9+v4bCzocK7Pm61UEA4HVnQXy\nox+/t71yS0fbrKqrQeHzzmLVE3x3Dds7fC3WUwdO1/WkZkmNMcYYY4yJiqrWEgw8o+VftVu2Obh5\neS3BwJ87S0RUdVtlVfVr/3l1Ubullmf85YHqquqaeyJo10b8te8HK0vb3Wdw/UeVBOpujyDWt9s2\nrP28dEX7PaMXPfSHbdXlm2+KINZn36xY/u23K9sulKaqPDBz+tYtG0p/G0Gsdz+trd6wrqHt3C2o\nyt/KSrdsCjR0HgsWfMv2siraPoxBlCVs3VxNoNMxSIlmSY0x7RCR00TkBhE5J3T7JyJyrYgckei2\nGWOMMQnnr7kqsHbpJ4FNy+s02PylWAMNBNZ/XBPcuOwdArV3RxJq7fqNl176m1uX/+Oxp/11dfVN\n68u3VXD9XX+ouv/xZxZt3Vbxl4ja1VAzJbB60cpg2ZeB8Clp1F9HYO07lcFt384n6P9XJKEqN68/\n78npl3/9+eKXgsFA83OsLt/CC3+4cds3H739dKCh4T+RxNqyofTMm3563up3F76sgUDztDtlmzYw\n8+rLyld++tFDgUDgjUhilQcaTr9u/Zdr36+pJBiWM27y13Pbxm+2ftdQ+ydV/aizOKqqNQROf44N\npeuoRWmOVUEDL7CxrJrAXZFUnUs0m6fGmA6ISDZwqareLSJXhX5fHaqSZIwxxvRpIpJGSvo1iOfH\n4stKRxX111QRDNxHoG5WNJNmikjmgP6F01J9KefvVDwwtaGhQTZuKSuvrtk+a+u2ij9HM2mmiOSS\nkvFbxHO2+DJ9qgHBX7eFYMNMAvX3RzM5pYj0yy4YcLPH652UXTggxV9XK9srt22sr6n6XV1N1aNR\nxioqHFh8q8ebcmrhgGJPbU2NVFWUr62u2Da9prrqqUjjhGIN7u/13eYVJhR6fVIdDEpNMLC6Iui/\nqTYYnBdlrF2y8E73IOMz8Uo9QWkg+NV2gr/1a/DVaGIlSlIkNZ2P+TJ9UCSD7uLbABEPzgD6u1S1\nSkR+raozGpObVtvqlVdeSUlJCSUlkRRnMcls8ODB2PE0xsTToEGDEv45Fw0RsS9qZgeqmjTvY0tq\nTLJK+B+JiEzHKVVcjVN56HicUq6LVPXtVtsm/D1cWlrKoEGDEh6jNxAREn08IfbjYcfTYa+DSVIJ\n/5yLhohofemqzjeMwIMLv+DSO5/ofMMINHx0P8Mu7rSOTUS++etZDD7nz67Emn7xoTy+YnvMcVI8\n8McTd2ZA//4xx6qvr+f9u2ez9s8PxxzLN7A/53z2RlIlNVbS2Zh2qOq0Vqvub3NDY4wxxhiTUEmT\n1KxYsQKA2vQ8tgWjnzh5eO160tLrwFvjdtOSRmWgiO31Ldfl5uZSXFzs+r4mTJjAu+++SzDYsvuq\nx+Ph0EMP5cUXX2znkcYYY4zpqURkADAM2A4si2YcizGJlDRJzXPPORMsH3DqD7nq3Q5Lb7dp3rAN\nBAu2U7vi1s437qEq95jFvPmvt1g3ceLEuCQ1tbW1jBw5koULF7ZYP27cOGpra13fn2np448/ZtSo\nUd0e//HHH+fyyy+P23478vrrr7PrrrsydOjQNu+fM2cOU6ZM6eZWuSeexzQZj2dr7bWxpx/XvqS3\n/432dSKe/fFl/kUy++8mGf19GqgLaPXGWvFl3o9/+82Jbp8xnYlrUiMi16jqnSJyGDAutHyNqt4Z\nz/0ak4xuu+02hg0bxv3338/5559PZWUlBx98MC+99BJZWVkUFRVRWFiIqvL2229TU1PDa6+9xqWX\nXkq/fv06jf/vf/+bV199lVGjRjFo0CByc3PJzMyksLCQl156iUMPPZSlS5dywgkn8Pzzz5Ofn8+M\nGTOYPHkyL7zwAieeeKIzy3KcPfDAA6xZs4Zx48bx1VdfUVFRQW1tLSNGjKCqqor333+fZcuWUVZW\nRnV1NWPHjuW7775j7dq1cW9btOJ5TOfOnctDDz3E4YcfnrTHc+7cuTzxxBNJ/54z0elNf6MmMuJJ\nOYK03GdSdjt2gKTlNq1XDRLc8MnU4Oblh+CPfXyIMfEU73lqvCLyS+BsaCp8nfjRt8YVZ5xxBj/8\n4Q8ZMGAAkyZN4sknn+S4447j2GOP5dFHH2Xffffluuuu47zzzmPSpElMnTq16TFff/11opufEKrK\nhAkT+PLLL8nPz6ekpIT09HT2228/PB4Pp556Klu3bmXYsGEMHTqU/v3788kn7U/6Fa6uro4RI0Yw\nZMgQvF4vqoqIICIMHTqUAQMGMHr0aN5880122203VJVDDjmEsrIyRowY0dQFNN4aGhoYPXo0c+fO\n5aijjsLj8eDxOP+KsrKyeOihhzjqqKMQEUaNGkVGRkbT/ckoXse0traWvffeO6mPZ+MX3WRuo4le\nb/sbNR0TESElbU7KHie2SGic+zx4i0dlefKGjklQ84yJWFyrn7W6UnO0qt7V1pUaEdG773Yq5Drd\nz6qi3te8YWtI6+Xdz8rb6X42fPhw1/c1btw4gDa7nzWuv+uuu3j11VdZt24dL7zwApdffjnl5eXM\nnTuXyZMnEwgEePHFF9lzzz055phjqKurY7/99mPBggXMnj2b3XffvaMm9KjTt5FUP3vwwQe54IIL\n4tYGq37mnkirnyX7MbXj6bDXwSSpuH/OiciewERVndnBNkdIvz2eT9llbEF722h9Nf7PHseqn0XO\nqp91v3ifWgn/VlAlIr8Cos9YTFLasGEDIsKaNWuYPHky5513HikpKZx11llMmTKl6czdEUccwZYt\nWzjkkEPYsGEDAMuXL09k0xMinl9+TWLYMTXGJCsRmYszFcE37dx/SmhxN8kqyu8wVmqWu40zJg7i\nOqamcdZ1VX0HeCee+zLuSk9P59133yU/v+X/ucbqZwAzZzonfk466STmzXMmrv3BD37QtO1ZZ50F\nOGezjTHGGNOtloZ+/wR4snGliAwFjgAmiEgmsJG6iiogd8cQDg3Ut3eXMUkjaaqfmeQSTcnmxoTG\nGON47rnn+PLLL/nFL35BSkp8/s12xz5i1RPaaCJnx7PH+S+QBSxotX4zsATYN/R7XbD8m22eQaNz\nxdP2cQ1u/rwGyIxjW42JmY3sM22aMGECBQUF5Ofnt/gpKChgwoQJiW6eMUmturoan8+3wzxPPW0f\nseoJbTSRs+PZ46TjDANo0e1fVWtU9VvgeVVdrap+NDAt8NUrZRr07xAkWLEmENz0Wd+s7mN6lKQ5\n1TJx4kQAatM9zBpbGPXjPbUNeNLryDx4httNSxqBQF7T69QoN7fdq8UxiWSemhNPPJEXXnihxf2/\n//3vyc7O5sILL2yx/rbbbuP6669v93HG9Cbnnntur9hHrHpCG3sDVYUPFjs3Ro+JW5lsO57JT0T6\nAfur6iKg8cvUyra2VdWlTcsN2x+SlLR6//Kn7vbk7pwjWQPyNVDvD5at3IJ/+wf4t58DbIv/MzCm\n65ImqYm9gle7RTt6jQ5H8SVAIBDgnHPO4YknnmD58uX84he/4MQTT+RPf/oTf/7zn3n55ZfZd999\nuemmm1i8eDE33HADn376KZ9//jnXXHMNa9as4YwzzmDSpEmJfirGxF1fnHzTJmvsJh8sJvj0HCDU\n/eKgI+OyGzuePcKlON3LFqnq36N5oPrrHhORucEtnx9N2cq90UAl8LKqbgBsTimT9JImqTE9j8fj\n4V//+hfjx4/n6quvJjs7G4/Hw6RJk6ipqeGVV15h5MiRXHLJJTz99NMsXryYV199lZKSEsaPH88f\n//hHKisrE/00jOmyvj75ZiObrLF3sePZo70GrO7qg1U1GIrxmmstMqabJE1Ss2LFCmrT89gW9AHg\nE8jzwbYGaAgVhh5eu56cyi0tHldflA3emqj3p+KDlFzwVyDa0On2lYEitrdR/CM3N5fi4uKo998b\nNJZsTktL4/PPP+fBBx/krbfe4umnnwbgmmuu4R//+AfgnOEZM2YMP/vZzygtLeXbb7+loKCA999/\n38rimh4tfPLN4cOHU1JSwqJFi9hvv/3YvHkzp556KnPmzNlh8s2jjjqqw7jhk2827idZJ7YMn6zx\niiuuYN68eS0ma/y///s/fvWrX/Hss8/aZI3xMHpM8wDZ0bHPkWjHs2cSkVHARJwSzn9JbGuM6X5x\nnXwz4kaEJt9snHhzVKGPG0fncPMHlXy8pTnhmDdsDQNmXtnisdumT496wk1P/v6k7vsb6j+7nWB5\nZLO1tzXxJcRv8stEi2TyzUYLFizgzjvvZP78+W42oUdd545k8s14s8k3IxOYdhEA3ul/a3cbm3yz\nZwgGgzD3r86NMy9u94t1b38dTI/l+ueciEwE+qnqnDjEVpt8M3I2+Wb3i+uVGhG5RlXvFJHDgHGh\n5WtU9c72HtNeQuOWriQ0fVEk89Q0Gj9+POPHj+/O5hmTFOwqY4LN/Sv63CNA6Nvh2T9NaHMSpbsK\nBZgeQeOR0DQq3VjmSpxBRXnMuvY8V2Jdes79/Pa0XVyJdeFf4dYfHuBKrBFF6YwclBN7IFU23KdS\nSQAAIABJREFUbatmY1Vd7LGCQfofczBFB+0ZeyyvF057I/Y4Lop39zOviPwSGAasD61r9/Rnqkcs\noUkS0cxTY4wxJoG6qVCA6REKReQCVY3LrNfF5e5cqXlt20Cuf3Bp5xtG6I4X3BvPdceL61yJ8+Qv\nByA5LhSxCgYp37qF7+rSYw6VSoD91n2Nd/FLMcfSnLyYY7gt3klNQFVnha7UHB1a1+YppOLiYvql\neyyhMcYYE5kzL27+QDnz4kS2xJhkUQQsS3QjjEmEeCc14VdlqkTkV7SaBKrR+PHjWVUdtITGGGNM\nRDweT5/tctaCy4UCEkVVqduwBIC0gYdbN7quyQJGAvMS3RBjultckxpVvSv0+x3gnY62XbBgAcWH\nfM/1NsQ7obGKL8YYYxJJRHpFl7O6DUuoWtk8gDm9+IgEtqbHeklV3+rqg0UkxeORY3OzMkfU1jdU\n1tbV/0dV17jZQGPiJWlKOq9fvx63CyPHO6EpLi4mLS3N9bjGGGOMMV1wsIjs0ZUxNUUF+RcMLS66\n/fRjj8w5fOQ+ueWVVcEHnn1505CB/T9du3HLmfForDFuiiip6UoVs0TrjoRm/Pjx1NREP0eO6RlE\n5GjgZ6p6buj2TJxJzR5V1Y0JbZwxxvQyaQMPb3PZRKUSiHqeiYGF/S48aJ89735i5vX90tNSG1d7\nLpw0YeCCdz8q+tH1MxaXbnan8pkx8RLplZqoqph1xcSJE6lN9zBrbGG723hqG6ibNqvluqxsMg+e\n0WJd08SaDRtJ32OKK+0LBPKYOHGis0+Ph7S0NGpqasjOznYlvkk+qvp6KJFvtBmnv3IwQU0yxpi4\nSIbxLCJiXc5idz4QVflSEfHtXFx085P3/LZfWqpvh/vHH3qA5/LzJw279n//7lYbjYmLSJOaiKuY\nddW6desYN66zkws7lsbLdLMRHcjvfBPTOzW9z1X1DhHJBS4A/th6w6lTp1JSUkJJSUl3tq+F0tLS\npIiRzPLr6wEo7+R5JsPxhNiPR28/npHq1a+DKqnL3gOgfr+DoQsJiW79AEqfcW6UlSH9RrvZQtOG\nOE0Iez/Rfz8bf9bxR+W2ldA0uvj0EzMsqTHJLtKkJuIqZl3VOFO9MclCREYCJSJyHfAgcAqwO/BI\nW9vPnDmzG1u3IzdmTe8LM68HUp2uFZ09z0QfT4j9ePSF4xmJ3v466PtvEnz93wB4CgqQLhQNqJUC\nqrY4fxvZBQWkF/fe16uXGwbQOFeNiOwJTFTVHf6hicgpqvo8ULjrkOKsjoLmZnfXKWRjui6ipCaa\nKmZdtWLFiuYbPj+1/uqoHq9pg6kIZjTdHl67nrT0OvA6Y14qA0Vsr+84Rm5uLsXFbpcrMD2Vqv4X\nOC1s1V8S1RZjjGmPqkLltqblrnSjsPEsvYOq3ta4LCJzgYXAN+HbiMhQ4AhggohkAl8u/vCzrT89\n4+R2+/+vLrVhpCb5JU31s+eee65peezxh/DIa/dG9fiTjvs9172/pen2vGEbCBZsp3bFrQBU7jGL\nefNf7zDGxIkTLakxxpgeQlXhg8XOjdFj+vC8JtLOchQRbDxLb/QW4AdOAp4MW78ZWALsG/r93aL3\nP9m8btOWwsFFbec10//66DYg+aaQNyaM65OsiMgtIpIiIkERmSIih4nI0SJyn4hMFZHr3d6nMcaY\nvif43hsE/3KH8/PeG4luTsKIADl5kJPXleE0pvd6GSfLbXGWWFVrVPVb4HlVXa2qurl82/nHXnzN\nhmVfftsiQG1dPTfc92DVC28ufa37mm1M18TjSs1DwGPAtcAQYGVofSC0v4o47NMYY0xf8+Z82Lyh\nefmQoxLbnkQZPab5DOXoMYlsiUkuFwDLgNHAp63vVNWljcu1dfXvpfpSxp36ixv+VJCbs/fI4bt5\nyyur9MPPV21v8Pvv27ClfAbO9zhjkpbrSY2qrhSRxj+Ux4Bf4lz2/E5V7xKRa9zepzHGmD5o0FDw\npTYvm5hsfuPnAPQfe1+CW2LcoKpXR7N9fYP/c+AYEen38YqvhgI1wCpVVaAPd+80PUW8xtQogKqu\nEpGs0O1dRGQqsD1O+zTGGNOHyFkXNw8hOfPihLYloT5YTPDpOUCoT3kXqp8Z00hVtwJbE90OY6IV\nl6RGVe8MW74wtLgoHvsyxhjTN3k8Hjj7p4luRq/x0tvfctiIYlLKy1m8eDEnn3xyoptkjDERS5rq\nZ8Z0hYgMADyquj7RbTHJK/D9Aztc533mw+5sjjHucmlMzWEjirnm3jdIe+QyZs+e7UrT+iKrymdM\nYlhSY3q6EsAHzA1fKSIZqmpdHZNMYNpFAHin/y3BLTGm9xARAk89AIDXup4lXi/uDph29HmJbkKb\nVs0+PelijbLzAt0uaZKaiRMnNt/w+bnopGlRPV7TUpg1Nqfptqe2AU96HZkHzwAgEMhruY825Obm\nRrVPkxTqgIzwFSJSDAwTkZ1VdW7bDzPG9AaWKLvnnU/Xc+cVYyksmZmQ7meqSt2GJYAzAahd4Ug+\n/uXujCRY5ylgUIa6Ess3dH+219S4EisjM5Oyyugmf29P9frVFGfG/h4OBoNsbkihKDst5lj1/iBv\n//ERvv3TwzHHSh3Ynws+fzPmOG5KmqRm+PDhLkcsaHEr3+XoJrFEpBAoBr5S1S9a3b0f0B8o6/aG\nGWNMD3X8EbsAkJ+fn5DxNHUbllC1svnLVo+dDDRBJbZFZArNhZoe7LYdG5MkXJ9805juoKpbVHUZ\nsEJE9haRvcPuexVIBQpFpI9OXGGMMSYRRAQ56EjnpxuvNqnqHOBzINhtOzUmiSTNlZoVK1Y03/D5\nKajz4832UeVPYXt9x5coc3NzKS4ujnMLTZI6C8hU1ftbra8E/Ng/dwOkPPtRh/e70wnCGBOLtIGH\nU7Hsj03LpktOpI2JNiMlIrsWFw/8n/S0tFFB1eqNGzc9UFtb+7yqNrjYRmPiImmSmueee65pefxJ\nYyjyZKPVpVTW5zBv/usdPnbixImW1PRdm4DM8BUich5QDZSranJ1+DTGGNMmEcGTmte0bLrkBVr3\nv4+AiHgGFRf/teTww0698vJf9B81cqRUVFTwr8fnHvPIY4+X+Xy+U+LQVmNc5XpSIyK3ALcA9cCP\ncS6FpgNnA18CGap6W3uPLy4uJiczHzZshOxaIKe9TY1BVRc0LovIJFV9WlX/mcg2meTjP+2ARDfB\nGGO6w2RgabQPGjx40MwfXzD5rOuvvTq7OaHciVtvuiHn4gt/knPcSae8+O3q79xtqTEui8eYmoeA\nx4BrgSGAN7Q+gJNEVbT3wOLiYsaPH09lTTnaUBeHpplebncR2T/RjTDGGGMSZAZRTnYuIrnZWVln\nt0xomg3deQi33niDdYcxSc/1pEZVV9J8luAx4JzQ8neqeietyu82akxoFixYQIO/3u1mmb7hM6Cf\niPw40Q0xxvQdm9/4OZvf+Hmim2H6OBEZhdNDJtoCOd87/9xz+nXU5W/Saad6273TmCQRr+pnjSUF\nVwFZodu7iMhUoM0JERsTmvXrbWJ40zWqOk9VF7VRNMAYY4zpNoFpFzXNodRdVPVj4HkiHFMjIo3j\nZNJzc3M7HI7g8/libJ0x8ReXQgGhKzKNyxeGFju8HGoJjTHGGGNMTE4APuhoAxEZChwBTBCRTOC9\nZ//9fNnPLr5wQHuP+eDDjitIGpMMkmaeGktoTKxEZDcROV5EhiW6LcYY09eoKvr+m86PWqH0BFkK\nlHayzWZgCbAaWKKqK1esXLly6fvvt7lxIBDg6t9cv8XdZhrjvqQp6WyMC6bg/DPfDfhzrMFE5Gjg\nZ6p6buj2T4ABwOuq+nas8Y0xJtHWPX1Eh+sGT4riX90Hiwk+PQcInTE96MgYW2e64FCcZKXdSyuq\nWgN8KyLPq+pqgLXrSn9w9g8veOOG31y361lnnJ6SmenMlPDpss/49bW/2bLqy6/uBO7qhvYb02VJ\nk9RMnDix+YbPT9BXT0qGkkNWy/vakJubG+fWmR7CA9QCrlRpUdXXReSwsFWFqvo7EbkasKTGGGPC\nBINB2LiuadlGlifES0Q4n7CqLg1b3iAiB11/862XT7/zrh9lZWZl1jfUa31Dwyelpetv8vv9S0TE\nkhoTNyJSqKoxXRFMmqRm+PDhba7P7+Z2mB7tBZz5kB53MWZ4ORht9dsYY0yjVcth6+bm5UOiLcJl\nXFBIFz+jVLUSuC30Y0x3KwZ6R1JjTCxEJBcYCdwA/BRY7ELMkUCJiFwHPAhsEZFraKfoxdSpUykp\nKaGkpCTWXXdZaWlnXam7J0a4AZeetMO6wPcPbFre+Kd5ru6vzTZ0cn9bzzkZjifEfjzcPp7JoCvv\nqXi+DlpXH/d9dCa/3mlDucttiOY55Xz9BekBPwC1X39BZRfakgyvpRsiOR6DBg2Kx64zaXkyzpik\nICLZwJ6q+mHo9nigSFUfExEP4BeRE4E0nN4ww1X1jWj2kTRJzYqvP6PWX93mfZo2mIpgm9PbtGt4\n7XpyKjtO+OqLssFbE1VcE5nKQBHbQ9MN5ebmUlwc93m7rgD2xPn+usKNgKr6X+C0sFUdloqeOXOm\nG7vtstLS0pg/JN2I0Vqgk/vj9MEecxsSfTwh9uMRj+OZDKI9nvF+HTavSgWgfwJf60Cq04Zon+e6\nTu6PJl7guEnwzRcAZBw3iewuvB7J8Fq6oavHwwW5dP4nYkwiHKyqC8NuZ4Z+APZS1eUiMhrw4yTm\ny0RkhKp+GukOkiap2Vq5iUdeu7fN+0467vdc9350V6TmDdtA2swrO9ymdvp0alfcGlVcE5nKPWYx\nb/7rgDNeKt5JjareHMr0B5NEVf1McvA+82HTcuPcEd7pf0tUc4xxRfjVqfbWhb/327LT6Us6vD+a\nImZy0Bj4KjQG9qAxkT/QuGkQsCZewbcF3Zmvpl6VbS7Os15ZWelarOoqd2LVBwJsq4v964hqkFq/\nn4rtsV+AawgEyRg2hILvxd4DwZeXB5+/2eXHq+rz4TdD6x4N30ZEDogmZtIkNca44F6cYVhe4PwE\nt8WEpDzb8fwGNkDJmN5BPnyL4CfvAeDZYz+rfpYYoyB+NRoK99uxWl4yGDBwoGuxdh4U954lfY6I\nFAKbROQkYF/gS1V9uvF+Vf28nYdGddUxbkmNiPwGpxIVwE9VdS8RuQn4TFXdHMhtTKOVQB3g3ikb\nY4zpxdY+dXiim2Dc9QSQ3nhDRPqr6ma3gte//aQrcf65Lp+pf3/HlVhb5v2G7LFTXYlV9cZMDpz2\ngiux/nDO7mwg24VIQXL91aysTY05UjpBDlz6b+r//UzMsaRfPw59/OnON3SMCFWULQTmq+onET5u\nhYjsDewELFPVDie1jOeVGr+q3hNKbpaKyDGE/aEZEweLgCycM1UmSfhPi+rqsTEmQV55byuH7JMD\nwNLllRx7cL/oAowe09z3d3Tk3c9cnSunj1PVua1W3SgiS1X1wYQ0yBjHpyIyAthdVe8HEJHzVPWf\nnTxub1V9LdQNbRcR2aaq29vbOJ5JTYqI/BbwAZ8AFwGPQNtpa21tbVurjYnGUcDDwLJEN8QYEz/J\n1qXxpbe/5bARxaSUl7N48WJOPvnkbm6BO2rqAlx135cAnFJSmODWGJfMBtquwmRMN1HVLSJyQGNC\nE1ItIt8HyoGdVfWhNh7aOJCoDBiqqh1e3ov3lZrbReRKIBWYhvNZc1hbG6en20UcE7NC4HKc99kt\nCW6LMaaPqKlp4Iq7FpJWdDFnnHFG1I9XVRavcypxjhmciUhkA4I7S+4gugTvwD2zmf7gagB+O2WX\nKB4Z2tf7i9GHZgEgCnJw3xpT01nhhs6KNsTJGcCHwHedbSgivszMzDMKCwquTElJ6e/3+4P+gH9+\naen6Gar6Vfybanq5oIiIalP5kS04vWuygYbWG4e6qm0EUNWXItlB3JIaVb0r9PueVnd9G699mr5N\nVW9OdBuMMfGXbF0aD9p3ALf8ZQmeb5YyY8aMqB+/eF0Nc5Ztbbp95E5ZbjYvYh+sqGLUHtlNyzsP\niPJk45fLYPP65uU+ltQkqWU4ZZ47JCI5xcUDF/5o8vl7XXLRhVmDBw0iEAjwyqsLLp12400/KCrq\n/6tuaKvp3d4DDgQ+AFDVNwFEZAxQ0cb2+/bYeWqMMcaYniJ8zMeSdzYzclcfUMm8Wcdy2pH9u2Uc\niBvJXXg781aezO+vL6awZCaLFy9mcLTd6HbfF/oXNy+bZJALbOtso50GD/7XH//3npEnTjih6Xuh\n1+vlhOOP4+ijxg446nvH/X7z5pgmezd9nKpWEkpoWq1vc7L0aBMaSKKkpl9OERedNK3N+zQthVlj\nc6KK56ltoG7arI63ycom8+Doz6qZzgUCeUyc6MxXkJvb6UkiV4QuVe4DeFR1Ubfs1BjT52Wle7n7\n57sBzgD7aI0ZnNnmck+jBxwBzzzYvGySQQPOVAftEpEhhxx80EHhCU249PR0Ztz5u6LjTzolLg00\nfZOInInz3nwcOA/4AqeC7WBgjapGfWYoaZKa4bu6fVanoNMteu5HR/Lr8D9o/NwDPI1NfWJMj3HR\nS848gX87fkiCW9J14VXCoq4YBohIwrqchTtsRDHX3PsGaY9cxuzZs6MPMOsGWPZB8/KVt7vbQBMV\nERkOZODM0N6RMeeceWaHX5rGjol9skZjWkkHFuCcjB4FfA0MVtWHRWQy0HOTGmNiISK5OP01FwLB\nxLbGGBNP4QOuA9MuctZN/1u3tmGn05d0eL9GcWqlq4UC3Pbau99R3xBA6uuZP38+Z599dnQByjaB\nBpuXI+Tma5lIyVaVT1VXiEiWqrZZoUBETgnN6h70B/wdNk97ykEwSSNHUrSq5dyZG1Q1fGbTWuB7\nOIUsPgf2BD4MXcH5siv7TJqk5r+rN7At6Gu67RPYvWELmZVbEP8ORRGiVl+UDd6aptsqPkjJBX8F\norHHby3Z41cGithev+P63Nxciot75Gy6VwB5ONXP8kO3jTEm6SVLoQAViCmf2nc0fPZh87JJBseL\nyP7h89SIyFDgCGCCiGQCrz38z3+VXf4/l7X74T//5VeU5vK6xnSqigCXybCm27P1m4Hh97eaU2mp\nG/tMmqRm4/YgV73rDEIbVejjxtE5BEs3kv67/3Elfu306dSuuBUAT/7+pO77G+o/u51geaSTmkau\nJ8Sv3GMW8+a/vsP6iRMn9tSk5n+Bc4FHgZ8luC3GmF5u7VOHNy1vq/Yz7f++BmD6JbuSl9U9H61u\nX7Eaf8jOzH/zG1JTUznhhBOiD+AR8KU2L0co/LXsyZKtKp+ITMH5nte6b+dmYAmwL7BEVTfttNPg\nV+c89PAZUyafn9Y6ztat5Vx93bSNwMDW9xnTkTRv2P+BzjpBusDT+SbdqzGhufmDSoINbVxKiFFP\nSDgSGt+TdG+JSP0KGINzpcbKhhtjus3S5ZVMv2RXpl+ya5cLBUzZrx9T9uuX0EIB73y6njuvGMvs\n2bNZvLjNgkQdkj1GtLlsEkNV5wDf4IxVCF9fo6rfAs+r6mqAdetKL7zp1umLfnLJz8o+/u8nBINB\nKioq+PsDc+qPOHpc6YYNG87t/mdgero0jzT9dIe4nU4Skd/g9JcD+Kmq7iUiNwGfqerjbT0mPKH5\neEsDRFfwrFM9PuGIc/zi4mLS0nY4SdMjqOrNIjINeAV4N9HtMc2SYfyDMfEUa6GAZHH8Ec6Em/n5\n+ZwcbTlnQA4agw7cqWnZJIUVtNNtTFWXhi3XicgJj819ouSNNxdfrarDRaitrq55pGzr1r+r6tZE\njfUyPVd3JTON4nmN3K+q94SSm6UicgxOpYM2pXqkZULjtpT0Hp1wdEdCM378eGpqajrfOHn9L3AN\n8ABONQ1jjEl6i9dWM+vD0BwgWsiRQ7IT26AuEhHIyWtejlD4XDmb3/g5AP3H3udu4/qo8MQlgm0V\nWAycFr8Wmb6ku5OaePY1ShGR3+IkMp8AFwHtzh3SL80Tt4TGv9coNLOoxyYc3ZXQLFiwgGCwRxcO\nexMoxykNaIyJI1VF33/T+bHKSDFZVlbH+mo/66v9LCurS3RzTC8iIpeJyM8T3Q7TN6V5pemnO8Qz\nqfGr6q04XzIFmAYsa2/jrXXBuCU02y+7EanZ1CMTju5MaNavX+96/G52JNAPeD/RDTGm1/tgMcGn\n5xB8eg58EP34C9Nsr7wUJ0lUZa+8pKnf0yOpKm+urebNtdWWbDsy6KCXjDHxlOHxNP10h7j991TV\nu0K/72l1V5uDuOuD7v/zaUxoMmbfTPUFZ7ke3xKapHM98A5wV6IbYkxvp6pQua1p2Xrbd91L31ZT\nHZom5KVvqxm3S16CW9RzJbI8dpKOH5wH7OJWsFJfkStx9hqczh9/fqQrsc6dB3/59SmuxPrhGzO5\n5nvuVIBNrypnp2B57IFUScvJJMeFceYeDSJ77kXq+T+OPZjXC48/3eEm3XWFplGvPSUUntCkfPEx\n4G5S09MTGq/X19sSGoB7gbE4Z6WiL0FkjImCtLPcN7g5DmRVeX3TxIyryt2v+mn6tMNxCui4Inel\nO1dldfRpjNx9sCuxAE4b51457dyB7lSulqWvUPfgw7HHSfUxaPbvKSzIjzlWg9/PypdeovKJZ2OO\n5elf2Ok2SVcooCtVzLpiQIaHWWNbvkCe2gbqps2KOpam+Ahk5+HbsoHAGRcSADxZ2WQePMOVtjZN\nfNmwkfQ9prgSs7vjq+ZTU1PDoYce2uK+3Nxc1/fXjX4NPAVcBVwdazARGQuU4MyC+4CIzARWA4+q\n6sZY4xvTk4mANg0KT3Bjerjd8nys3NbQtGy6LrwkdiLLYyeRQmAc8GAn20Ukxe/OmC8RSPG61yUp\nJcW9c/RBcaldDQ1oRUXMYTTVh8cjpKR4Y44VDCpaV0dwW+ztwtf5/6qkS2qIsopZV40c2lZmXOBa\nfPvX1lLPLTraoQpVXSIi33Mp3hGqeqeIXBW6vRnIAnp0NQVjXDF6TPOgzNFWvjcWJ+2aw9INtU3L\nputEpFu7nCU7Vb070W0wfVdaB4mriJwJ5AOPAf2BnwCfA2XAa6q6Pdr9RZLUNFYx89FcxewRwNWa\nkwsXLmTcuHFuhjR9T4WIzAD+41K8xuRFAFT1DhHJBS4A/th646lTp1JSUkJJSYlLu49eaWlpUsRo\nT36907WmPI77cKsNyXA8IfbjEc/jyeDdnd8xdGGtr3PO/Ha1nZEez3i+DlpXH9M+ysrqqAsEQ8tl\nlPqqo47hxt9WrM/DjXbE2gZV5b0tAQAOLvRGVVraTZG8DoMGDequ5kRFRLxAHlCrqj16ngeTWN60\nDq8upQMLgH2BoThzKq0BcnHef3FJavyqeruIXAmk4lQxU+CwaHfWkcGDB7NixYqWK31+av3R/3OP\nuS1+H5lVsQ3JqC/KBm/b/wsqA0Vs76DbdG5uLsXF7gxU62OCwG24dyVliYhcDXhFZCfgFGB3nKR+\nBzNnznRpt11TWloa84ekGzE6EkhNBRL8YX630wsjo5PNEn08IfbjEc/jqarNVc9Gj+nyl8fUtDVA\n198Tkbyn4v2+3rzKaUP/Lu7jP598R7Xf6X72n41evn9A9HFc+dsaFPug9ljbEetr+ebaav5d6hQK\nKCjI48jBiblqkxT/66IkIsVDBhdPHzZ0yIQhgwd7KquqZOiQweu3VVTeuq2i8slEt8/0PJ0kNbXA\n94APcRKZvYClwBBgGBD12bJOk5poq5h11XPPPbfDurHHH8Ijr93r5m4ictNBPydtxjUxxaidPp3a\nFbe2eV/lHrOYN//1dh87ceJES2q6phD4ZWj5lliDqeqbOHPfNPpLrDGN6TVCJZ0hNDfAQe5UMopW\nElSXillRhrepK19RRuz95vuyYDDIuqqGpmUTGRHZfeedBi3808w7hhx79Fg8oRK8GzZuGnj1TdP/\nttOg4qMS3ETTA6Wktv//TFXnht0MnyR2eZf319UHGpOEPgr9bhCRQlXdktDWmB30hi+gxrjt1pKB\niDi1R245YkCCW9OzLS+rY2tdoGn5qJ1tjJKIDFfVFR1ts9Og4if+89hDQ/bec/cW6wcOKOKB2ffm\n/2DKJZPXrd8Q13aa3qeTKzWucz2pEZFbcM6S1wM/xhn0kw6cDXwJZKjqbW7v1xicy5ivAScB43Gq\noRlj4sEKBbjG6/WyoSbQtNylGHbCAHD6xjfWx+7rU2+KyL04J/tGAlM72G7/U044dqfWCU3Y/dz+\n22v7PT/ftcrQpo9Iyejeayfx2NtDOJUMrsXpF7cytD4Q2p8LdeSMadMW4Dlgf2BtgttiTK8mIgnr\ncmZMe/YrSKNfurdpuY+7VVXLRKSzCWH2P2H80R2Wm20v4TGmIympPTypUdWVItLYN+4xnDEOTwLf\nqepdIhLbYBVj2rcYuAd4Hvg4wW0xcXLRS87A8r8dPyTBLTFJNGu66SW6OoFpI/F4GJzta1ru4y4R\np4LH18C/wu8QkbGq+kboZuWGTZvr6GD2i9pad+anMX1Lj+9+FuJcAVZdJSJZodu7iMhUulCizZhI\nqOoruDhzsjEmublVgS1WsX4RN+6xyTdbGIDz/autngvhg7deeWTu0+XTrvxFZnvdHx+e+1Q9TgVc\nYyLWK5IaVb0zbPnC0OKieOzLGGNMH5UkFdhM8rDJN1tYDHyuqsta36GqT4Ytbx9cPHD2ZVdNu/q+\nGbfneVpd4frks8+ZPnPWOpwyu8ZELKU3JDXGJJKI7NfWP/GeprOzzqp9fRis6SpVZfE6Zx6tMYMz\nE3aFw5hwqkrdhiUApA083N6XURKRKcBOAKp6O85VmmEislxVO6xvXbph4x3zXnnNd+DRJ1x66U8u\n6Dd65IjUiqoqHnh07pbFS5auW1u6/hRcnsrD9H7eDko6x0PSJDUTJ07ccaXPz0UnTev2tgT9Puqm\nzYophicrm8yDZ7R5XyCQ1/bzDcnNzY1p34Z+iW6AMcls8boa5izb2nS7x57ZtgpsvUo/PufiAAAZ\ntElEQVTdhiVUrXy46XZ68REJbE3Po6pzROSHNE9APQioAs4HHuzksQrcLCL/e8MdMybn5uQcGAgE\nKtaWrv8n8J6qqiWZJlreNF+37i9pkprhw4cnugmu6qgnb363taLvEJGRYTd7RYW9ZLsSYwP0jRsO\nfGhlp+s/nLxnRLGsApsxO/iO5rHLaTjjYMoifbCqbgP+GId2mT7Im9p+UiMiZ+J8JX4cOAg4AHgH\nGAysUdW3o91f0iQ1xsToQMCL856uB/6b2OYYk7xsMLVJRmkDD29z2URlLKHqn6p6T4LbYvq4Tq7U\npAMLgH1wpn8pAXZV1YdFZDLQc5OaFV9/xnZvHhXBDIbXrid7ewWB7Dy8VdsQfwMA9UXZ4K2hMlDE\n9vrY9ufxeEhLS6Ouro5gsO2uprm5uRQXF8e2I9MtQpfdbwFKcf5QjDHtsMHUxk2B7x/Y4TrvMx9G\nFEdErMtZ7KoBm1TGJAVveocF82pxJk3/ENgXZwzYN6ErOF92ZX9Jk9RsrdzE1pQsrnt/Cy/uuZWG\nwYPJmH0zKV80TzdSO306tStupXKPWcyb/3qX91VcXMz48eNZsGAB69evb3e7iRMnWlLTswjOZXc7\naL2Im92VjOmNkqW0tWmWwLmbyohjtdnSIYe5Eqd+ewM1W6pdiQVQ2sF3uWjtLJWuxEk9ZDQ5I++N\nOU4Q2F5XT+nGLbHHCgbx7LcfuVdfGXMsvB5YNL/jTTq4UqOqc1ut6jhYBJImqWk0qtBHSn7+DgmN\nWyJNaEyP9AJwNk7/TOOCthIKSyaMSTJW2rqFPj7+bxnwq9CP61Jf/FfnG0Wg7oJrqU4vdCUWwOB+\n2a7Fashyp139UvwMGhD7FfFgMMimQCr982MvIlXf4Ofrdz5k9Z8f7nzjTqQO7N/pNh2NqYkH16fb\nFZFbRCRFRIIiMkVEDhORo0XkPhGZKiLXt/fYtBQPN47OwV9ebgmNiYqI7APsAcwAzk1wc4wxxvRQ\ngWkXEZh2UaKbETURGYrTWyH2ywPGuMCb6mv66Q7xuFLzEPAYcC0wBGfwD0AgtL82K1P5UlLpn5HK\nzR9UckdOjANm2mAJTa93KTAH+BtwXoLbkpS6cvbyows6qUo4ObkqtMVTss0blKiz0W6NnzDuXQnV\nA0uQVcual91pnul5dgUyQss2p4xJuO4u6ez6lRpVXQksDd18DDgntPydqt5J8x9cC1npuWyo3M7H\nWxrcbpIlNH1DLdCA897bKcFtMcaYbiMfvoV+8h76yXvIh28lujkmQVT1dWAEzqDrLhORPUTkxFAv\nmzR3Wmf6Ik+qr+mnO8RrTI0CqOoqEckK3d5FRKbSXD+9heraCvy41yeykSU0fcYynLLOK3FqnVtJ\nZxcc8OCKRDchaSTDvEFujnHq4+MOjOmtammefDMqaR7PkbmelD+PTssu2ictM7MyGPC/u71ye/8U\n3zNbAv64jNExvVuvmHwzdEWmcfnC0GKH1Tga/PWut8YSmr5DVeckug3GGJMQo8c0d7sYPSaRLTGJ\nV0EXqp9lerzHDUlJ++fN/Yf17+dt/iJ6cZ7mv1hd9uN/Vmzcf2vQ72Y7TR/QXVdoGiVd9TO3WEJj\nTO+oXtbpuB7oM2N7kmGMU8qzH3V4f984EslFRPp8xTPT5CvgCuDySB8gItLf6/vL7UW79c/2eFvf\nx4nZheml/vpRT1VtdrmpprcTX/f2XkyapKZfThHp3hRmjc3BU9tA3bRZO2zjycom8+AZBAJ5TJw4\nsd1YjRNr1tTUcOihh3a5Tbm5sZfPMyaRkuFLsDHhLEntfZIh0Y31BE5nBTCgxxTBGEf0832MPTg9\nJ791QhNuUk7/HEtqTLQktXvnQk+apGb4ruHj2gra3CYz9Ds/hv0sXLiQcePGxRAh+fbVG5+TiIxT\n1YVx31HHbRgLlAAbVPUBEfkJMAB4XVXfTmTb+hIb19MsGV4L/2kHJLoJrkvU+CI76WDiYA0RFssR\nkVNU9XlgyG6+9LyOtg3vkmZMpMSX2q37S5qkprv0xgSgNz4nnLNNC7tjRx04QlXvFJGrQrcLVfV3\nInI10COSmli/BIef3bSB5cYNyZCYQe/ompksemOi21Op6gOdbROaz+YIYIKIZAKl3zbUVgLtJjbb\nAjaexkSvo+5nInImznWKx4EfAF8COwNlwGuq2mZhsY4kTVLT2RwQbrr55pt73b564XO6EbipO3bU\ngcYKMo1vTm31u4WpU6dSUlJCSUlJ3BvWntLS0pi3by9GfV1dl/bRHeLRpmQ4nhDdc+uW4/mneU2L\n+fdcA0D5lXc23+/CsYjH8XTjb8NtnSV4yfi31pYBndyfDM+jszZ09hwiiRGtQYMGuRovCpuBJTil\nn5cAa96prSyfEgzkZbbTBe25qs1VEIcStaZX6+RKTTqwANgHeAVnnqU1QA5Ogt0zkxpVtbnCTDJa\nEroq4xWRnYAtInIN7VSWmTlzZrc2rrXS0tI2PiR3PBsdrvX2bcdwpKatafMx8fbh5Ja3u+uKUaKP\nJ7R1PGI7nm1dnTjplaoWt6O5QhFITW1zvx3r+DlEH69zbvxtxEcytCF2gU7u7wmvpXQyLghAn+kZ\nx6MzqloDfCsiz6vqaoBsj/eX12/6+h83Fw0rzPGkhG/Lwu3l9S9Xb10OHJKgJpseSlI7LBRQC3wP\n+BBnSo5hwAvAoaHlqKt8JUVSY0wyUtU3gTfDVt2fqLYYA9YdsLfpLcczfAB9YNpFzrrpf+vWNtj4\npOipauNE6VQFA89lerzn/mLDqtnDUzPy903Nyq0KBhre3L6tslaDr5YF/ZcANQlsrumJOrhSo6pz\n27mry32ULakxxhjTbVpfCbJB+iYZ9OZxQSIyRFXXdLZdTTDwMjBcREa+tb1iOE4Ss0hVq0Jx/r+9\new+SqjzzOP595sJtEEFu4hVQB9EVS1YQMQJCjG7WVbNKTOKq2STGxM1eAmu2jJfaqFvZ9ZaUq3EX\n46JGa6s0lHhFUVkUMSpqBBFkMMKMWMJwGWAcmMFhnv3jPQ3NMJee6dM9Z3p+n6opmO4+7/uc9/R0\nv+95z3neHEcqhabHpnQWkfgVypngQtBeh8A9951opVOWQpWUBBRJYma3Ah8D44DZmW7n7iuAFbmK\nS3qOHpP9LF/pcluo5++B44H73f2juOpppa5S4PfA1e5encN6/gooBxa6+wdx1dNKXdcBjUC1uz8W\nc11TgR+5+7ej37ttCuVMzmjloxMrIi1TJ1h6gAfcvcrMBnd1INIz9aSZmnyly03V8zMAd/9PMzsH\nGAHEOqjh4H26nHDTU9ya13Mh8AHwZR7q2k3IULGsjW06xd1fNbMz0h7qdimURVqThEFsegyauZNC\nku2sdBLuC8qBi82sCWgAHujqYKTn8Tyvb9SVg5oOpcuNoZ5Qmdlw4Cx3vy3metLrSu3TKcBwYBPw\nZA7raXT3e8zsX4k/DXLzuga6+2wzu4EDb6KPS/oUR67eEzmXhA5sc+q8FhYdTxFpRxlQBCyJq8Dq\n46bGUk7TmrU0bHsnlrIAFj2+MLay6l58LpZyNgwZxLot22Ipq++wwWzYWpN9QSUllIw7icNvnpV1\nUcW9SuHadrqBPWhQ06F0uTHVcxThbMVLZjbW3VfnsK4jo87/lcDSXNYDvG1mswhp8eLWvK56M5tN\n/LNcmNk4YLKZXQ88Qu7eE9KNqTOfHHGcydbxFMmZ+YS1ZWrjKnD9jXfGUs7wm2ex/Oe3x1IWwPwf\n3RxbWX+cdWss5Yy98Se8f8s9WZdT1KsXp8/+AZV3z8m6rF4jhsN503jnnkeyLqtsxNB2X+NF+R1m\ndNmgJl/pcluo5y9yUU8rdeHu2b9z2q8nZ6mGW6grnk+0lutaAVyU9pBSKItITiiJRm4UwCVbBbEP\nka8QTgpOIQcnIkXa1YNmakREpIuoA79fEtoiCTEUCrXlPocDnxAWMxTJOy/uITM1IiIiIpIzo4B/\nogN9PTPrV4Jd2pui05rwnbtpetzdP8xdiFLQilqfqTGzmcBA4HHgW4R7uFcBRwAbOpP1VoMaEcmI\nzn6KiHQr89w947ve+1nxrAGUXDeW/ocOpVffPTgf8cWP+1tJZR17L8xloFKY2pmp6QMsAsYSFnoF\nGOXuj5rZFXQi660GNSIiIiKF5/lMX1hmJbOOpM9N5zBkYFFaEtJR9BtaTcPQF9n8+i725iRIKVzt\npHSuB2YQklz1I8zUrI5mcP7Umfo0qBEREREpMJ7h2gJm1vsQimc3H9CkDKM3p3HokUuJJz2x9CBt\nZD9z9yfSfo1l7cOiOAqRtpnZL6J//9LMxrfw/Pn5j0pERESEr55A/wEtDWhSxlCW36XhpSB4ca99\nP/mgmZr8+NzMhgB/DtSY2cXAg8ANwL3AqWa2CziTcJPUOOAz4CjgGeAC4G3gMGAEUOvuD+Z9L0RE\nRKTQHFpGcb+2XlCqc+DSCdvqm9p/UYz0Ls2PJ4HvAHsBJ2QkcWBRtDYLwCFANWHg0+juDwFfAucC\nd7v7S4Sc89ui14qIiIh0mJmdnfbrqg3Ut3lt2Q6+zHFEUoi21jfu+8kHDWrywN03AdOAxYTBTGoa\nN30IezzQCBRHryH6dyEw28zOJSyiNRioyHnQIiIiUqiGpf7j7u9X07BxZxsDl7fZXpOXqKSgVG2v\n3/eTD7r8LE/c/a+j/y4F7ov+XxU99x+tbHN79N8VLT0vIiIi0lHuPi/9993snfkMm16ZzpAjDqc3\nFt1f08Be3mL7zo00PAl8rytile6rasfuvNanQY2IiIhID7bX/SMzm7SILXcWY2cPoMT24PYFjdsa\n8TsaaHoIDWqkgyo21ua1Pl1+JpIHu3btoqqqqkvLSEIMceiqGNLrve++MNlaXV3N1q1b81ZGEmJo\nq5yeGANAZWUlCxYs6PB2SWiHdJ3djyTEkKS2jKst8s3dP631xst20DjqU+onbqLh5C+88aR63zs3\n0/TQIunWbKzd95MPGtSI5MHmzZv5wx86vDhurGUkIYZ58+axZMmSbhlDer3FxcUArF69mo8//jhv\nZSQhhrbK6YkxACxevJhp06Z1eLsktEO6zu5HEmJIUlvG1RZdxd0b3P1Td+/4CF8kzZbqun0/+aDL\nz0RaYGaHAf8M7HL326LH/pFwIuBVd38v07KWLVvGs88+S1VVFWVlZVxwwQUdjifbMpIQA8All1zS\n4W2SEEOq3srKSioqKpgxYwavvPIK8+fPp1evXjQ0NDBlypSclpGEGFoqZ+3atUyfPp25c+cyfvx4\nTj311B4TA4RBclVVFevXr+e1117jmmuuyWi7JLRDHPuRhBiS1JZxtYVIIaip/iKv9SVyULNmzZo5\ndXV15QBlZWUVY8aM+WFXx/TWuuqH+hTb2Rt3N23uX2orzho9bF9MZnaVuz/cfBszmwqsd/fKztTZ\nWrmZ2Lr+jTnesKUcwHoPqRg8cnJ6vMcCI9391c6U3RFRXdvdfUeu64rZNOAB4BwzG+ju24GtwBF0\n8O9mwoQJDB8+nDfffLNTnfA4ykhCDHHoqhjS692+fTuTJk2iqKiI0tJS+vTpw8SJE3NeRhJiaKuc\nyZMnd3rb7hoDwEUXXURNTQ1jx45l7NixGW+XhHaIYz+SEEOS2jKutkiqor594imouIjifjGVtRNK\n+/WNp6xdxBaXlZTEUlZx795YSXEsbV/ctzeUllISQ3uVZhBPzbqP2n2NmZUDZwDV7v6imZ0EXODu\nt5vZHcAD7p5R1l9L4mWS77333uJFixZNBZg+ffqr48ePnwZgZv9A6FCuAxqAk6NNqoHNQDnQD/iV\nu8eacmF51eY//Wpl3ejlW7/k5vGHvPqNkw+flnrOzG4E1gN/CzwK7AR2EdalOQF4GbgIOAZ4AhgT\nbVoKvB/t00jC+jOfEjrPZwKj3P3azsS7Zc3Ti/es+uVUgF7jbnlryHHnTopi/R7QB/gbYK67PxA9\n/pNmdQNUEhb9PCWKud7d741efxkhBXV/wro7dYRjkJquPjaq57OoLQYCWwjprEcA9VH7HEV4w27q\nzH7GycwuAi4mpNJeDcwDzgHmRYOa1OtucPd/a7atA78AFrv74rwFLTnR1vE0s9Pd/Z0sy8+qjCTE\nEEc5hRCDmQ0Derv7p1mUkYR2yHo/khBDEtoy7liSIvpcFDmAu1tLj5vZRmB42kOb3P3wtOcvByYB\nXwPGApe7+++i565y94ejfv+77r40k1gSOVPThj8CzwDfBYz967y4uz9vZicQ0h/HvkrUxt1Nm5dv\n/XJ0K097FM8LwHGEgYpHP/WEzn36a18AphMuZaoEZhIGB5Xu/pSZXcX+tWqyUjTwFKz38KHN6nfg\nKSD9YuOdaXWn2nUBIf30DwkDxpaWhk3t44fAEHd/Jq2MJmBltG0lcBrwLpB6U5cC7wF7st3POLj7\nU4R2SV1+NpswIGsys/MIawidDqxtYdsW/6ile2rreMbRScm2jCTEEEc5hRCDu1dns30cMcRRThz7\nkYQYktCWcceSFPqek45IH8C08vxjwGPRTM13gM1mdgpQA4wzs6MJfcexhOVQ2pXImZq2Lj8zsyvd\n/ZF8x7T0k+o51bubygGG9S2qSL/8LBNmdgYw3d1/mZMAm9m6/o05vrdunPUePtSb6pcMOWbid1uJ\nazJhcLssk9mttNevI8xirMzk/pLoUrw6wkzaenfv2hRaIiIiIlIwEjmoERERERERyZRSOouIiIiI\nSLeWyHtqVlQsm7OzrqYcYEDZoIpx5RMyvtTLzE519+VtPN+pjGQ7ly2dYzXV5QA+aFjFgAlnHRBT\nPjOKZWLt2rUP1dbWjoQDL+FLb5+22ipqp3W6TExEREREki6Rg5qddTXlj7/2m6kA35yyP/mXmc0E\nLgXmEjIqfAJMBh4jZBfbDRxtZtsJa4wsI9zE/gUhu8JnhBvXzzWzWsIN/FuAN9y9pq2YrKa6vN9v\nbpkKsOvamw98Li2jmJnlMqPYcWSY3a22tnZkWga59Ke+bWb/AiwGBplZPfB94F7gmujxMcAQwn0z\nIp1mZrMImQo/JyTTOIuQrXAOcJO7/zTL8g34OnAJ4f26EXidkMWuDpgPXO3uN2VTT0+UybEzs2mE\nBCBroufOA7YBhxG+X+4Cfgb8Nv0ESUeOW5SVcDdwPvBb4CuEz/WXCZ/zsb2f8t0OGdbfLdqqA+20\nh/A9/jtgAuH7rQL4FiEj5m3Afe5+dQwxWVT/sYR+wOeEYzSTcDPyDYTsm5PM7Abgfnff1kZ5c9w9\ndYLwa+6+MNsYRSQ+iRzUtKEPYVByIiH98UbChyiED8Mmwoe+EQY0mwmDAI+eS2X+WkhYh6QSOM3d\nn8syrnxlFHOyz+62HthByOA1hHAJ4krCIPFdQprsRmLKvtaTmdn3CV/Y9cA4YIK7z+xEObcDjwDn\nApM7U0a2zOw6wn70JSxIem+GmzqhM9FAeO9NJtzLt83M3s82Lg83BT4X5bX/NTAaGEB4Xze5e5WZ\ndXyJeNo/ftHzFYRU7f2AB4Hn3f38ZuW0evyijuAHhDTnj7a0fS5keDwzOXaDCW3USMgU2EhIhz6Z\ncHKmDnjloII7dtyK3X2hmY0AyoD/A6qAC6MYs3o/ZdAWHWmHGYTvoNHA0YSO/Qtp7fADM1tCK8c7\nOkH2MiHl/kbgIUJq+SuBVFtdBvwd4TP7VeB69p+AymlbtSOjdnL3eWZ2PLCB8P32kpldQfguLHL3\nPWb2eiwBubuZrSK8L5cR+jw7CH0Ed/cPzGxF9PKBwK/NrIqwLMOZ0etnEpYe+HfCSdNHCVk7TzSz\nS4Ht0f5+A3iaMDj7eVKu2hDpSRI5qBlQNqgiNUMzoGzQvgV3Uvmrm3kx+ve/mj2+Pv0XMxvaLGva\nI9FZoyWZxOSDhlWkZmh80LADFgFy97nN6kplCHsier75AppXRv/e1ez1b7A/o9gBGd5aKKNNZWVl\nFakZmrKysvQ2bN5OEDohED70JV5FhM5FHbDYzE5w94NSQrfjQ8IZ111ZlJGtTYQZxq/T8kC8NdsI\nnZ0F7t5kZsuBPWZ2BCFlY7lnuKhWO0oInYk+hL/pJcAYMxsc1XOUu2/oRLkHHT/C32hD9NhIQkeu\nL+Hs/NNm9lV3fzmtjIOOX1oZW6IyGtvYPhcyOZ7tHbuTCWfdy6KO9LGELIqLzewY4PeE/fwzQoex\npRmK9o7b0YTOPVEsK4ErCIOQlwjtnu37qb22yLgdgFXRPlURjm1fwkm0OwmDkM20cLwJae4PIbxX\nZhJmEwYRZs1fBr5JmN2fS5j92Ab0Irw/NwAXmNn/5qGt2pJJO6VOQuLum81slJldQlgGwYANUXrX\nce1dSt4BWwnHYyJhMLKcMBgcGsU03sxmEI7NAHe/0czmsH9Zgn0hRz917n63mf03oe1PjMpfRThR\nulIDGpGuoexnIjliZtMJi6/2J6xL9D+E+67u6mA5VxK+jH8KPN+ZMqTjWjl+o4FaQudsMWHG9V1C\n5+dMYBEwxd2vTyun+fFLL2Mv4Uz+HYTO6kHbS/eRwXumxeNNmOEvAR4mnKB7OCpnEGFWfRrhpNcQ\nwuxgBXA/8Diho32mu1+Th10UEUksDWpEpFWpBBjA5zGf1ZUuoOMZqB1ERAqPBjUi0qK0BBg/Jpw5\nXk44azyRcH1/KpHFJndf0FVxSmZ0PAO1g4hIYdI6NSLSmtQ15KlLZV4GbgWeY38iiyZ1/LoNHc9A\n7SAiUoA0UyMiIiIiIt2aZmpERERERKRb06BGRERERES6NQ1qRERERESkW9OgRkREREREujUNakRE\nREREpFvToEZERERERLo1DWpERERERKRb+3/mCyaddCjsTwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x11a9d2190>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# create groups by environment\n",
    "groups_sp = spores_norm_df.groupby(['env_evo','env_test'])\n",
    "groups_hy = hybrids_norm_df.groupby(['env_evo','env_test'])\n",
    "groups_cl = clone_norm_df.groupby(['env_evo','env_test'])\n",
    "\n",
    "param='growth_rate'\n",
    "\n",
    "fig = plt.figure(figsize=(12, 5), dpi=300)\n",
    "\n",
    "import matplotlib.gridspec as gridspec\n",
    "grid = gridspec.GridSpec(2, 3, width_ratios=[1,3,1], hspace=0.4, wspace=0.2)\n",
    "\n",
    "gs = {}\n",
    "gs['schematic'] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,0])\n",
    "gs['variance'] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,0])\n",
    "gs[('HU','heatmap')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,2])\n",
    "gs[('HU','barplot')] = gridspec.GridSpecFromSubplotSpec(1, 24, subplot_spec=grid[0,1], wspace=0)\n",
    "gs[('RM','heatmap')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,2])\n",
    "gs[('RM','barplot')] = gridspec.GridSpecFromSubplotSpec(1, 24, subplot_spec=grid[1,1], wspace=0)\n",
    "\n",
    "\n",
    "### analysis of variance ###\n",
    "ax = plt.subplot(gs['schematic'][:])\n",
    "\n",
    "ax.axis('off')\n",
    "ax.patch.set_visible(False)\n",
    "    \n",
    "ax.text(-0.1, 1.2, chr(ord('A')), transform=ax.transAxes,\n",
    "        fontsize=12, fontweight='bold', va='top', ha='right')\n",
    "\n",
    "### analysis of variance ###\n",
    "ax0 = plt.subplot(gs['variance'][:])\n",
    "    \n",
    "ax0.text(-0.1, 1.2, chr(ord('D')), transform=ax0.transAxes,\n",
    "         fontsize=12, fontweight='bold', va='top', ha='right')\n",
    "\n",
    "S = spores_norm_df[spores_norm_df.background.isin(['ancestral','evolved'])]\n",
    "S = S.rename(columns=labels, copy=True)\n",
    "S = S[np.isfinite(S['growth_rate'])]\n",
    "\n",
    "H = hybrids_norm_df[hybrids_norm_df.background.isin([('ancestral','ancestral'),('ancestral','evolved'),('evolved','evolved')])]\n",
    "H = H.rename(columns=labels, copy=True)\n",
    "H = H[np.isfinite(H['growth_rate'])]\n",
    "\n",
    "formula = 'growth_rate ~ background + de_novo_gene + de_novo_genotype + auxotrophy + tetrad + C(spore)'\n",
    "\n",
    "tsdf = S.groupby(['env_evo','env_test']).apply(stats.variance_vectors, formula)\n",
    "thdf = H.groupby(['env_evo','env_test']).apply(stats.variance_vectors, formula)\n",
    "    \n",
    "df = pd.concat([tsdf, thdf], keys=['spores','hybrids'], names=['type'])\n",
    "    \n",
    "# fix columns\n",
    "df = df.rename(columns={'C(spore)':'spore'})\n",
    "\n",
    "df = df.swaplevel('type', 'env_evo')\n",
    "df = df.sortlevel(['env_evo','env_test'], ascending=False, axis=0)\n",
    "    \n",
    "utils.simple_axes(ax0)\n",
    "    \n",
    "df.plot(ax=ax0, kind='barh', stacked=True, \n",
    "        color=[factor_dict['color'][x] for x in df.columns],\n",
    "        align='center', legend=False, width=0.75)\n",
    "\n",
    "bars = ax0.patches\n",
    "hatches = ''.join([factor_dict['hatch'][x] for x in df.index.get_level_values('type')]*len(df)*2)\n",
    "\n",
    "for bar, hatch in zip(bars, hatches):\n",
    "    bar.set_hatch(hatch)\n",
    "\n",
    "ax0.set_xlabel('Variance', fontsize=5, y=0.9)\n",
    "ax0.set_ylabel('')\n",
    "\n",
    "ystart, yend, ylabels = plot.set_custom_labels(df.index, 0)\n",
    "ax0.set_yticks(ylabels.values(), minor=True)\n",
    "ax0.set_yticklabels(ylabels.keys(), minor=True)\n",
    "\n",
    "ylabels = df.index.get_level_values('env_test')\n",
    "ax0.set_yticks(np.arange(len(ylabels)), minor=False)\n",
    "ax0.set_yticklabels(ylabels, minor=False)\n",
    "\n",
    "ax0.tick_params(axis='both', which='major', size=2, labelsize=5)\n",
    "ax0.tick_params(axis='both', which='minor', size=1, labelsize=5, pad=18)\n",
    "\n",
    "# Shrink current axis's height by 10% on the bottom\n",
    "box = ax0.get_position()\n",
    "ax0.set_position([box.x0, box.y0 + box.height * 0.2,\n",
    "                 box.width, box.height * 0.8])\n",
    "\n",
    "# Put a legend below current axis\n",
    "leg0_1 = ax0.legend(bbox_to_anchor=(0.5, -0.1), frameon=False, prop={'size':4}, \n",
    "                    handlelength=0.75, borderaxespad=3, loc='upper center', ncol=3)\n",
    "\n",
    "#Create custom artists\n",
    "s_artist = plt.Line2D((0,1),(0,0), color='k', marker='s', mfc='none', ls='')\n",
    "h_artist = plt.Line2D((0,1),(0,0), color='k', marker='s', mfc='none', ls='')\n",
    "\n",
    "#Create legend from custom artist/label lists\n",
    "leg0_2 = ax0.legend([s_artist, h_artist], ['spores', 'hybrids'], \n",
    "                    frameon=False, prop={'size':5}, handlelength=0.5, loc='upper right')\n",
    "\n",
    "ax0.add_artist(leg0_1)\n",
    "ax0.add_artist(leg0_2)\n",
    "\n",
    "for (ii, (env_evo, env_test)) in enumerate([('HU','HU'),('RM','RM')]):#,('HU','YNB'),('RM','RM'),('RM','YNB')]):\n",
    "    \n",
    "    # group by evolution and test environment\n",
    "    gsp = groups_sp.get_group((env_evo, env_test))\n",
    "    ghy = groups_hy.get_group((env_evo, env_test))\n",
    "    gcl = groups_cl.get_group((env_evo, env_test))\n",
    "\n",
    "    # apply sorting ranks to reorder rows\n",
    "    gsp.loc[:,u'rank_background'] = gsp[u'background'].map(dict_background)\n",
    "    gsp.loc[:,u'rank_population'] = gsp[u'population'].map(dict_population)\n",
    "    gsp.loc[:,u'rank_genotype'] = gsp[u'genotype_short'].map(dict_genotype)\n",
    "    gsp.loc[:,u'rank_gene'] = gsp[u'gene'].map(dict_gene)\n",
    "    gsp.sort_values(by=[u'rank_background',u'rank_population',u'rank_genotype',u'rank_gene'],\n",
    "                    ascending=True,inplace=True)\n",
    "    \n",
    "    ghy.loc[:,u'rank_background_MATa'] = ghy[u'background_MATa'].map(dict_background)\n",
    "    ghy.loc[:,u'rank_background_MATα'] = ghy[u'background_MATα'].map(dict_background)\n",
    "    ghy.loc[:,u'rank_population_MATa'] = ghy[u'population_MATa'].map(dict_population)\n",
    "    ghy.loc[:,u'rank_population_MATα'] = ghy[u'population_MATα'].map(dict_population)\n",
    "    ghy.loc[:,u'rank_genotype_MATa'] = ghy[u'genotype_short_MATa'].map(dict_genotype)\n",
    "    ghy.loc[:,u'rank_genotype_MATα'] = ghy[u'genotype_short_MATα'].map(dict_genotype)\n",
    "    ghy.loc[:,u'rank_gene_MATa'] = ghy[u'gene_MATa'].map(dict_gene)\n",
    "    ghy.loc[:,u'rank_gene_MATα'] = ghy[u'gene_MATα'].map(dict_gene)\n",
    "    ghy.sort_values(by=[u'rank_background_MATa',u'rank_background_MATα',\n",
    "                        u'rank_population_MATa',u'rank_population_MATα',\n",
    "                        u'rank_gene_MATa',u'rank_gene_MATα',\n",
    "                        u'rank_genotype_MATa',u'rank_genotype_MATα'],\n",
    "                    ascending=True, inplace=True)\n",
    "\n",
    "    ### heatmap ###\n",
    "    ax1 = plt.subplot(gs[(env_evo,'heatmap')][0])\n",
    "    \n",
    "    ax1.text(-0.1, 1.2,  chr(3*ii + ord('C')), transform=ax1.transAxes,\n",
    "             fontsize=12, fontweight='bold', va='top', ha='right')\n",
    "    \n",
    "    # spores\n",
    "    S = gsp\n",
    "    S = S.groupby([u'mating',u'background',u'population',u'genotype_short',u'gene'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    S_ref = np.nanmedian(S.ix[:,'ancestral'].values)\n",
    "    \n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    \n",
    "    plot.heatmap_spores(S-S_ref, ax1, title, \n",
    "                        xlabel, ylabel, xticklabels, yticklabels, \n",
    "                        cmap=plt.cm.RdBu_r, radius=0.25)\n",
    "    \n",
    "    # hybrids    \n",
    "    H = ghy\n",
    "    H = H.groupby([u'background_MATa',u'population_MATa',u'gene_MATa',u'genotype_short_MATa',\n",
    "                   u'background_MATα',u'population_MATα',u'gene_MATα',u'genotype_short_MATα'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    H = H.unstack(level=[u'background_MATα',u'population_MATα',u'gene_MATα',u'genotype_short_MATα'])\n",
    "    H_ref = np.nanmedian(H.loc['ancestral','ancestral'].values)\n",
    "    \n",
    "    title = environments[env_test]\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    legend_title = r'$\\langle r_{bg}^{a/\\alpha}\\rangle_k$'\n",
    "    \n",
    "    plot.heatmap_hybrids(H-H_ref, ax1, title,\n",
    "                         xlabel, ylabel, xticklabels, yticklabels,\n",
    "                         fold=False, cmap=plt.cm.RdBu_r, pad=30, \n",
    "                         legend_title=legend_title)\n",
    "    \n",
    "    ax1.xaxis.set_ticks_position('top')\n",
    "    ax1.yaxis.set_ticks_position('left')\n",
    "    \n",
    "    ## genotype\n",
    "    # labels\n",
    "    xlabels = H.columns.get_level_values(u'genotype_short_MATα')\n",
    "    ax1.set_xticks([x+0.5 for x in np.arange(len(xlabels))], minor=True)\n",
    "    ax1.set_xticklabels(xlabels, minor=True, fontsize=5)\n",
    "    \n",
    "    ylabels = H.index.get_level_values(u'genotype_short_MATa')\n",
    "    ax1.set_yticks([y+0.5 for y in np.arange(len(ylabels))], minor=True)\n",
    "    ax1.set_yticklabels(ylabels, minor=True, fontsize=5, rotation=90)\n",
    "    \n",
    "    ## gene\n",
    "    # labels\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(v+0.5, 1.15), xycoords=(\"data\", \"axes fraction\"),\n",
    "                     ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "    ystart, yend, ylabels = plot.set_custom_labels(H.index, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(-0.15, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                     ha='left', va='center', annotation_clip=False, fontsize=4, rotation=90)\n",
    "        \n",
    "    ## population\n",
    "    # labels\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, 1)\n",
    "    xlabels = {k: v for k,v in xlabels.items() if k != u'WAxNA'}\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(v+0.5, 1.15), xycoords=(\"data\", \"axes fraction\"),\n",
    "                     ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "    ystart, yend, ylabels = plot.set_custom_labels(H.index, 1)\n",
    "    ylabels = {k: v for k,v in ylabels.items() if k != u'WAxNA'}\n",
    "    for k, v in ylabels.iteritems():\n",
    "        ax1.annotate(k, xy=(-0.15, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                     ha='left', va='center', annotation_clip=False, fontsize=4, rotation=90)\n",
    "    \n",
    "    ## background\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, 0)\n",
    "    ax1.set_xticks([x+0.5 for x in xlabels.values()], minor=False)\n",
    "    ax1.set_xticklabels(xlabels.keys(), minor=False, fontsize=4)\n",
    "    xgrid=[xst+1. for xst in list(set(xstart.values()))]\n",
    "    [ax1.axvline(g, lw=1, ls=\"-\", color=\"black\") for g in xgrid]\n",
    "    \n",
    "    ystart, yend, ylabels= plot.set_custom_labels(H.columns, 0)\n",
    "    ax1.set_yticks([y+0.5 for y in ylabels.values()], minor=False)\n",
    "    ax1.set_yticklabels(ylabels.keys(), minor=False, fontsize=4, rotation=90)\n",
    "    ygrid=[yst+1. for yst in list(set(ystart.values()))]\n",
    "    [ax1.axhline(g, lw=1, ls=\"-\", color=\"black\") for g in ygrid]\n",
    "    \n",
    "    ax1.tick_params(axis='both', which='minor', size=0, pad=9)\n",
    "    ax1.tick_params(axis='both', which='major', size=0, pad=20)\n",
    "    \n",
    "    # filter by dictionary\n",
    "    gsp = gsp[(gsp['background'].isin(sp_bg_dict['position'][env_evo].keys())) &\n",
    "              (gsp['genotype_short'].isin(sp_gt_short_dict['position'][env_evo].keys())) &\n",
    "              (gsp['population'].isin(sp_cl_dict['position'][env_evo].keys()))]\n",
    "    ghy = ghy[(ghy['background'].isin(hy_bg_dict['position'][env_evo].keys())) &\n",
    "              (ghy['genotype_short'].isin(hy_gt_short_dict['position'][env_evo].keys())) &\n",
    "              (ghy['population'].isin(hy_cl_dict['position'][env_evo].keys()))]\n",
    "    \n",
    "    # apply sorting ranks to reorder rows\n",
    "    gsp.loc[:,'rank_background'] = gsp['background'].map(sp_bg_dict['position'][env_evo])\n",
    "    gsp.loc[:,'rank_population'] = gsp['population'].map(sp_cl_dict['position'][env_evo])\n",
    "    gsp.loc[:,'rank_gene'] = gsp['gene'].map(sp_gn_dict['position'][env_evo])\n",
    "    gsp.loc[:,'rank_genotype'] = gsp['genotype_short'].map(sp_gt_short_dict['position'][env_evo])\n",
    "    gsp.sort_values(['rank_background','rank_population','rank_gene','rank_genotype'],\n",
    "                    ascending=True,inplace=True)\n",
    "\n",
    "    ghy.loc[:,'rank_background'] = ghy['background'].map(hy_bg_dict['position'][env_evo])\n",
    "    ghy.loc[:,'rank_population'] = ghy['population'].map(hy_cl_dict['position'][env_evo])\n",
    "    ghy.loc[:,'rank_gene'] = ghy['gene'].map(hy_gn_dict['position'][env_evo])\n",
    "    ghy.loc[:,'rank_genotype'] = ghy['genotype_short'].map(hy_gt_short_dict['position'][env_evo])\n",
    "    ghy.sort_values(['rank_background','rank_population','rank_gene','rank_genotype'],\n",
    "                    ascending=True,inplace=True)\n",
    "    \n",
    "    ### spores barplot ###\n",
    "    ax2 = plt.subplot(gs[(env_evo,'barplot')][:7])\n",
    "\n",
    "    ax2.text(-0.1, 1.15,  chr(3*ii + ord('B')), transform=ax2.transAxes,\n",
    "             fontsize=12, fontweight='bold', va='center', ha='right')\n",
    "    \n",
    "    ax2.text(0, 1.15,  environments[env_test], transform=ax2.transAxes,\n",
    "             fontsize=6, va='center', ha='left')\n",
    "\n",
    "    # set reference as mean phenotype of the ancestral hybrid\n",
    "    ref = np.nanmedian(gsp.groupby(['background']).get_group('ancestral')[param].values)#median()\n",
    "    gsp.loc[:,param] = gsp[param] - ref\n",
    "    \n",
    "    gsp = gsp.groupby(['background','population','gene','genotype_short','tetrad','spore'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['background','population','gene','genotype_short'])\n",
    "    \n",
    "    ax2, bp = gsp.plot(ax=ax2, kind='box', by='population', return_type='both',\n",
    "                       labels=gsp.columns.get_level_values('population'))\n",
    "\n",
    "    plot.boxplot_custom(bp, ax2,\n",
    "                        colors=[sp_gt_short_dict['color'][env_evo][x] for x in gsp.columns.get_level_values('genotype_short')],\n",
    "                        hatches=['']*10)\n",
    "    \n",
    "    ax2.set_title('spores (haploid)', fontsize=6)\n",
    "    ax2.set_xlabel('', fontsize=6)\n",
    "    ax2.set_ylabel(r'Rel. growth rate $r_{bgk}^{a/\\alpha}$', fontsize=6)\n",
    "\n",
    "    ### hybrids barplot ###\n",
    "    ax3 = plt.subplot(gs[(env_evo,'barplot')][7:], sharey=ax2)\n",
    "    \n",
    "    # set reference as mean phenotype of the ancestral hybrid\n",
    "    ref = np.nanmedian(ghy.groupby(['background']).get_group(('ancestral','ancestral'))[param].values)\n",
    "    ghy.loc[:,param] = ghy[param] - ref\n",
    "    \n",
    "    ghy = ghy.groupby(['background','population','gene','genotype_short','tetrad','spore'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['background','population','gene','genotype_short'])\n",
    "    \n",
    "    ax3, bp = ghy.plot(ax=ax3, kind='box', by='population', return_type='both', \n",
    "                       labels=ghy.columns.get_level_values('population'))\n",
    "    \n",
    "    plot.boxplot_custom(bp, ax3, \n",
    "                        colors=[hy_gt_short_dict['color'][env_evo][x] for x in ghy.columns.get_level_values('genotype_short')],\n",
    "                        hatches=['']*30)\n",
    "    \n",
    "    # set reference as mean phenotype of the ancestral hybrid\n",
    "    ref = np.nanmedian(gcl.groupby(['background']).get_group(('ancestral','ancestral'))[param].values)\n",
    "    gcl.loc[:,param] = gcl[param] - ref\n",
    "    \n",
    "    gcl = gcl.groupby(['background','population','gene','genotype_short','isolate'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['background','population','gene','genotype_short'])\n",
    "\n",
    "    gcl = gcl.iloc[:, gcl.columns.get_level_values('background').isin([(u'evolved',u'evolved')])]\n",
    "\n",
    "    gcl = gcl.reindex(columns=ghy.columns)\n",
    "    \n",
    "    for i,d in enumerate(gcl):\n",
    "        y = gcl[d]\n",
    "        x = [i+1]*len(y)\n",
    "        ax3.plot(x, y, markersize=3, marker=\"x\", markeredgecolor='k', linestyle=\"None\")\n",
    "\n",
    "    ax3.set_title('hybrids (diploid)', fontsize=6)\n",
    "    ax3.set_xlabel('', fontsize=6)\n",
    "    ax3.set_ylabel('', fontsize=6)\n",
    "        \n",
    "    # Grid\n",
    "    for ax, g in zip([ax2,ax3],[gsp, ghy]):\n",
    "        \n",
    "        ## background\n",
    "        xstart, xend, xlabels = plot.set_custom_labels(g.columns, 0)\n",
    "        # labels\n",
    "        for k, v in xlabels.iteritems():\n",
    "            ax.annotate('\\nx\\n'.join(k) if isinstance(k, tuple) else k, \n",
    "                        xy=(v+1, 0.97), xycoords=(\"data\", \"axes fraction\"),\n",
    "                        ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "        # grid\n",
    "        xgrid=[xst+1.5 for xst in list(set(xstart.values()))]\n",
    "        [ax.axvline(x, lw=1.0, ls=\"-\", color=\"0.9\") for x in xgrid]\n",
    "        \n",
    "        ## population\n",
    "        # labels\n",
    "        di = {u'no driver': u'no\\ndriver'}\n",
    "        xstart, xend, xlabels = plot.set_custom_labels(g.columns, 1)\n",
    "        xlabels = {('x'.join(k) if isinstance(k, tuple) else k): v for k,v in xlabels.items() if k != u'WAxNA'}\n",
    "        for k, v in xlabels.iteritems():\n",
    "            ax.annotate(k, xy=(v+1, -0.05), xycoords=(\"data\", \"axes fraction\"),\n",
    "                        ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "\n",
    "        ## gene\n",
    "        di = {u'no driver': u'no\\ndriver', u'/': u'',\n",
    "              u'/RNR2': u'RNR2/RNR2', u'/RNR4': u'RNR4/RNR4', \n",
    "              u'/no driver': u'no\\ndriver', u'/FPR1': u'FPR1/FPR1', u'/TOR1': u'TOR1/TOR1',\n",
    "              u'no driver/no driver': u'no\\ndriver'}\n",
    "        xstart, xend, xlabels = plot.set_custom_labels(g.columns, 2)\n",
    "        ax.set_xticks([x+1 for x in xlabels.values()], minor=False)\n",
    "        xlabels = ['/'.join(k) if isinstance(k, tuple) else k for k in xlabels.keys()]\n",
    "        xlabels = [di[x] if x in di else x for x in xlabels]\n",
    "        ax.set_xticklabels(xlabels, minor=False, fontsize=4, va='top')\n",
    "\n",
    "        ## genotype\n",
    "        # labels\n",
    "        xlabels = g.columns.get_level_values('genotype_short')\n",
    "        ax.set_xticks([x+1 for x in np.arange(len(xlabels))], minor=True)\n",
    "        ax.set_xticklabels(['/'.join(x) if isinstance(x, tuple) else x for x in xlabels], \n",
    "                           minor=True, fontsize=5)\n",
    "                \n",
    "        ax.tick_params(axis='x', which='minor', size=0, pad=-5)\n",
    "        ax.tick_params(axis='x', which='major', size=0, pad=5)\n",
    "        ax.tick_params(axis='y', which='major', labelsize=5)\n",
    "    \n",
    "    # set spine and tick properties\n",
    "    for ax in fig.get_axes():\n",
    "        for loc in ['top','bottom','left','right']:\n",
    "            ax.spines[loc].set_linewidth(1)\n",
    "        \n",
    "# # show legend\n",
    "# leg1 = ax3.legend(bbox_to_anchor=(1.05, 0.5), frameon=False,\n",
    "#                   loc='upper left', borderaxespad=0.,\n",
    "#                   handlelength=0.75, title='Background', prop={'size':6})\n",
    "    \n",
    "# leg2 = ax3.legend(bbox_to_anchor=(1.05, 0.5), frameon=False,\n",
    "#                   loc='center left', borderaxespad=0.,\n",
    "#                   handlelength=0.75, title='Genotype', prop={'size':6})\n",
    "\n",
    "fig.savefig(dir_paper+'figures/figure5/figure5.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_paper+'figures/figure5/figure5.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "groups_sp = spores_norm_df.groupby(['env_evo','env_test'])\n",
    "groups_hy = hybrids_norm_df.groupby(['env_evo','env_test'])\n",
    "groups_cl = clone_norm_df.groupby(['env_evo','env_test'])\n",
    "\n",
    "param='growth_rate'\n",
    "\n",
    "fig = plt.figure(figsize=(12, 5), dpi=300)\n",
    "\n",
    "import matplotlib.gridspec as gridspec\n",
    "grid = gridspec.GridSpec(2, 3, width_ratios=[1,3,1], hspace=0.3, wspace=0.2)\n",
    "\n",
    "gs = {}\n",
    "gs['variance'] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,0])\n",
    "gs[('HU','heatmap')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,2])\n",
    "gs[('HU','barplot')] = gridspec.GridSpecFromSubplotSpec(1, 24, subplot_spec=grid[0,1], wspace=0)\n",
    "gs[('RM','heatmap')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,2])\n",
    "gs[('RM','barplot')] = gridspec.GridSpecFromSubplotSpec(1, 24, subplot_spec=grid[1,1], wspace=0)\n",
    "\n",
    "for (ii, (env_evo, env_test)) in enumerate([('HU','YNB'),('RM','YNB')]):#,('HU','YNB'),('RM','RM'),('RM','YNB')]):\n",
    "    \n",
    "    # group by evolution and test environment\n",
    "    gsp = groups_sp.get_group((env_evo, env_test))\n",
    "    ghy = groups_hy.get_group((env_evo, env_test))\n",
    "    gcl = groups_cl.get_group((env_evo, env_test))\n",
    "\n",
    "    # apply sorting ranks to reorder rows\n",
    "    gsp.loc[:,u'rank_background'] = gsp[u'background'].map(dict_background)\n",
    "    gsp.loc[:,u'rank_population'] = gsp[u'population'].map(dict_population)\n",
    "    gsp.loc[:,u'rank_genotype'] = gsp[u'genotype_short'].map(dict_genotype)\n",
    "    gsp.loc[:,u'rank_gene'] = gsp[u'gene'].map(dict_gene)\n",
    "    gsp.sort_values(by=[u'rank_background',u'rank_population',u'rank_genotype',u'rank_gene'],\n",
    "                    ascending=True,inplace=True)\n",
    "    \n",
    "    ghy.loc[:,u'rank_background_MATa'] = ghy[u'background_MATa'].map(dict_background)\n",
    "    ghy.loc[:,u'rank_background_MATα'] = ghy[u'background_MATα'].map(dict_background)\n",
    "    ghy.loc[:,u'rank_population_MATa'] = ghy[u'population_MATa'].map(dict_population)\n",
    "    ghy.loc[:,u'rank_population_MATα'] = ghy[u'population_MATα'].map(dict_population)\n",
    "    ghy.loc[:,u'rank_genotype_MATa'] = ghy[u'genotype_short_MATa'].map(dict_genotype)\n",
    "    ghy.loc[:,u'rank_genotype_MATα'] = ghy[u'genotype_short_MATα'].map(dict_genotype)\n",
    "    ghy.loc[:,u'rank_gene_MATa'] = ghy[u'gene_MATa'].map(dict_gene)\n",
    "    ghy.loc[:,u'rank_gene_MATα'] = ghy[u'gene_MATα'].map(dict_gene)\n",
    "    ghy.sort_values(by=[u'rank_background_MATa',u'rank_background_MATα',\n",
    "                        u'rank_population_MATa',u'rank_population_MATα',\n",
    "                        u'rank_gene_MATa',u'rank_gene_MATα',\n",
    "                        u'rank_genotype_MATa',u'rank_genotype_MATα'],\n",
    "                    ascending=True, inplace=True)\n",
    "\n",
    "    ### heatmap ###\n",
    "    ax1 = plt.subplot(gs[(env_evo,'heatmap')][0])\n",
    "    \n",
    "    ax1.text(-0.1, 1.2,  chr(ii + ord('C')), transform=ax1.transAxes,\n",
    "             fontsize=12, fontweight='bold', va='top', ha='right')\n",
    "    \n",
    "    # spores\n",
    "    S = gsp\n",
    "    S = S.groupby([u'mating',u'background',u'population',u'genotype_short',u'gene'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    S_ref = S.ix[:,'ancestral'].mean()\n",
    "    \n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    \n",
    "    heatmap_spores(S-S_ref, ax1, title, \n",
    "                   xlabel, ylabel, xticklabels, yticklabels, \n",
    "                   cmap=plt.cm.RdBu_r, radius=0.25)\n",
    "    \n",
    "    # hybrids    \n",
    "    H = ghy\n",
    "    H = H.groupby([u'background_MATa',u'population_MATa',u'gene_MATa',u'genotype_short_MATa',\n",
    "                   u'background_MATα',u'population_MATα',u'gene_MATα',u'genotype_short_MATα'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    H = H.unstack(level=[u'background_MATα',u'population_MATα',u'gene_MATα',u'genotype_short_MATα'])\n",
    "    H_ref = H.loc['ancestral','ancestral'].loc['WAxNA','WAxNA'].values.squeeze()#.loc['',''].loc['+','+']\n",
    "\n",
    "    title = environments[env_test]\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    \n",
    "    heatmap_hybrids(H-H_ref, ax1, title,\n",
    "                    xlabel, ylabel, xticklabels, yticklabels, \n",
    "                    fold=False, cmap=plt.cm.RdBu_r, pad=30)\n",
    "    \n",
    "    ax1.xaxis.set_ticks_position('top')\n",
    "    ax1.yaxis.set_ticks_position('left')\n",
    "    \n",
    "    ## genotype\n",
    "    # labels\n",
    "    xlabels = H.columns.get_level_values(u'genotype_short_MATα')\n",
    "    ax1.set_xticks([x+0.5 for x in np.arange(len(xlabels))], minor=True)\n",
    "    ax1.set_xticklabels(xlabels, minor=True, fontsize=5)\n",
    "    \n",
    "    ylabels = H.index.get_level_values(u'genotype_short_MATa')\n",
    "    ax1.set_yticks([y+0.5 for y in np.arange(len(ylabels))], minor=True)\n",
    "    ax1.set_yticklabels(ylabels, minor=True, fontsize=5, rotation=90)\n",
    "    \n",
    "    ## gene\n",
    "    # labels\n",
    "    xstart, xend, xlabels = set_custom_labels(H.columns, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(v+0.5, 1.15), xycoords=(\"data\", \"axes fraction\"),\n",
    "                     ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "    ystart, yend, ylabels = set_custom_labels(H.index, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(-0.15, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                     ha='left', va='center', annotation_clip=False, fontsize=4, rotation=90)\n",
    "        \n",
    "    ## population\n",
    "    # labels\n",
    "    xstart, xend, xlabels = set_custom_labels(H.columns, 1)\n",
    "    xlabels = {k: v for k,v in xlabels.items() if k != u'WAxNA'}\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(v+0.5, 1.15), xycoords=(\"data\", \"axes fraction\"),\n",
    "                     ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "    ystart, yend, ylabels = set_custom_labels(H.index, 1)\n",
    "    ylabels = {k: v for k,v in ylabels.items() if k != u'WAxNA'}\n",
    "    for k, v in ylabels.iteritems():\n",
    "        ax1.annotate(k, xy=(-0.15, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                     ha='left', va='center', annotation_clip=False, fontsize=4, rotation=90)\n",
    "    \n",
    "    ## background\n",
    "    xstart, xend, xlabels = set_custom_labels(H.columns, 0)\n",
    "    ax1.set_xticks([x+0.5 for x in xlabels.values()], minor=False)\n",
    "    ax1.set_xticklabels(xlabels.keys(), minor=False, fontsize=5)\n",
    "    xgrid=[xst+1. for xst in list(set(xstart.values()))]\n",
    "    [ax1.axvline(g, lw=1, ls=\"-\", color=\"black\") for g in xgrid]\n",
    "    \n",
    "    ystart, yend, ylabels= set_custom_labels(H.columns, 0)\n",
    "    ax1.set_yticks([y+0.5 for y in ylabels.values()], minor=False)\n",
    "    ax1.set_yticklabels(ylabels.keys(), minor=False, fontsize=5, rotation=90)\n",
    "    ygrid=[yst+1. for yst in list(set(ystart.values()))]\n",
    "    [ax1.axhline(g, lw=1, ls=\"-\", color=\"black\") for g in ygrid]\n",
    "    \n",
    "    ax1.tick_params(axis='both', which='minor', size=0, pad=9)\n",
    "    ax1.tick_params(axis='both', which='major', size=0, pad=20)\n",
    "    \n",
    "    # filter by dictionary\n",
    "    gsp = gsp[(gsp['background'].isin(sp_bg_dict['position'][env_evo].keys())) &\n",
    "              (gsp['genotype_short'].isin(sp_gt_short_dict['position'][env_evo].keys())) &\n",
    "              (gsp['population'].isin(sp_cl_dict['position'][env_evo].keys()))]\n",
    "    ghy = ghy[(ghy['background'].isin(hy_bg_dict['position'][env_evo].keys())) &\n",
    "              (ghy['genotype_short'].isin(hy_gt_short_dict['position'][env_evo].keys())) &\n",
    "              (ghy['population'].isin(hy_cl_dict['position'][env_evo].keys()))]\n",
    "    \n",
    "    # apply sorting ranks to reorder rows\n",
    "    gsp.loc[:,'rank_background'] = gsp['background'].map(sp_bg_dict['position'][env_evo])\n",
    "    gsp.loc[:,'rank_population'] = gsp['population'].map(sp_cl_dict['position'][env_evo])\n",
    "    gsp.loc[:,'rank_gene'] = gsp['gene'].map(sp_gn_dict['position'][env_evo])\n",
    "    gsp.loc[:,'rank_genotype'] = gsp['genotype_short'].map(sp_gt_short_dict['position'][env_evo])\n",
    "    gsp.sort_values(['rank_background','rank_population','rank_gene','rank_genotype'],\n",
    "                    ascending=True,inplace=True)\n",
    "\n",
    "    ghy.loc[:,'rank_background'] = ghy['background'].map(hy_bg_dict['position'][env_evo])\n",
    "    ghy.loc[:,'rank_population'] = ghy['population'].map(hy_cl_dict['position'][env_evo])\n",
    "    ghy.loc[:,'rank_gene'] = ghy['gene'].map(hy_gn_dict['position'][env_evo])\n",
    "    ghy.loc[:,'rank_genotype'] = ghy['genotype_short'].map(hy_gt_short_dict['position'][env_evo])\n",
    "    ghy.sort_values(['rank_background','rank_population','rank_gene','rank_genotype'],\n",
    "                    ascending=True,inplace=True)\n",
    "    \n",
    "    ### spores barplot ###\n",
    "    ax2 = plt.subplot(gs[(env_evo,'barplot')][:7])\n",
    "\n",
    "    ax2.text(-0.1, 1.2,  chr(ii + ord('A')), transform=ax2.transAxes,\n",
    "             fontsize=12, fontweight='bold', va='top', ha='right')\n",
    "    \n",
    "    gsp = gsp.groupby(['background','population','gene','genotype_short','tetrad','spore'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['background','population','gene','genotype_short'])\n",
    "    \n",
    "    ax2, bp = gsp.plot(ax=ax2, kind='box', by='population', return_type='both',\n",
    "                       labels=gsp.columns.get_level_values('population'))\n",
    "\n",
    "    boxplot_custom(bp, ax2,\n",
    "                   colors=[sp_gt_short_dict['color'][env_evo][x] for x in gsp.columns.get_level_values('genotype_short')],\n",
    "                   hatches=['/']*10)\n",
    "    \n",
    "    ax2.set_title('spores (haploid)', fontsize=6)\n",
    "    ax2.set_xlabel('', fontsize=6)\n",
    "    ax2.set_ylabel(r'Rel. doubling time $\\langle\\tau\\rangle_{b,g,k}^{a/\\alpha}$', fontsize=6)\n",
    "\n",
    "    ### hybrids barplot ###\n",
    "    ax3 = plt.subplot(gs[(env_evo,'barplot')][7:], sharey=ax2)\n",
    "    \n",
    "    ghy = ghy.groupby(['background','population','gene','genotype_short','tetrad','spore'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['background','population','gene','genotype_short'])\n",
    "    \n",
    "    ax3, bp = ghy.plot(ax=ax3, kind='box', by='population', return_type='both', \n",
    "                       labels=ghy.columns.get_level_values('population'))\n",
    "    \n",
    "    gcl = gcl.groupby(['background','population','gene','genotype_short','isolate'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['background','population','gene','genotype_short'])\n",
    "        \n",
    "#     ax3, bp = gcl.plot(ax=ax3, kind='box', by='population', return_type='both', \n",
    "#                        labels=gcl.columns.get_level_values('population'))\n",
    "\n",
    "    boxplot_custom(bp, ax3, \n",
    "                   colors=[hy_gt_short_dict['color'][env_evo][x] for x in ghy.columns.get_level_values('genotype_short')],\n",
    "                   hatches=['/']*30)\n",
    "    \n",
    "    ax3.set_title('hybrids (diploid)', fontsize=6)\n",
    "    ax3.set_xlabel('', fontsize=6)\n",
    "    ax3.set_ylabel('', fontsize=6)\n",
    "        \n",
    "    # Grid\n",
    "    for ax, g in zip([ax2,ax3],[gsp, ghy]):\n",
    "        \n",
    "        ## population\n",
    "        # labels\n",
    "#         xstart, xend, xlabels = set_custom_labels(g.columns, 1)\n",
    "#         ax.set_xticks([x+1 for x in xlabels.values()], minor=False)\n",
    "#         ax.set_xticklabels(xlabels.keys(), minor=False, fontsize=5)\n",
    "        \n",
    "        ## background\n",
    "        xstart, xend, xlabels = set_custom_labels(g.columns, 0)\n",
    "        # labels\n",
    "        for k, v in xlabels.iteritems():\n",
    "            ax.annotate('\\nx\\n'.join(k) if isinstance(k, tuple) else k, \n",
    "                        xy=(v+1, 0.97), xycoords=(\"data\", \"axes fraction\"),\n",
    "                        ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "        # grid\n",
    "        xgrid=[xst+1.5 for xst in list(set(xstart.values()))]\n",
    "        [ax.axvline(x, lw=1.0, ls=\"-\", color=\"0.9\") for x in xgrid]\n",
    "        \n",
    "        ## population\n",
    "        # labels\n",
    "        di = {u'no driver': u'no\\ndriver'}\n",
    "        xstart, xend, xlabels = set_custom_labels(g.columns, 1)\n",
    "#         xlabels = {k: v for k,v in xlabels.items() if k != u'WAxNA'}\n",
    "        xlabels = {('x'.join(k) if isinstance(k, tuple) else k): v for k,v in xlabels.items()}\n",
    "        for k, v in xlabels.iteritems():\n",
    "            ax.annotate(k, xy=(v+1, -0.1), xycoords=(\"data\", \"axes fraction\"),\n",
    "                        ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "\n",
    "        ## gene\n",
    "        di = {u'no driver': u'no\\ndriver', u'/': u'',\n",
    "              u'/RNR2': u'RNR2/RNR2', u'/RNR4': u'RNR4/RNR4', \n",
    "              u'/no driver': u'no\\ndriver', u'/FPR1': u'FPR1/FPR1', u'/TOR1': u'TOR1/TOR1',\n",
    "              u'no driver/no driver': u'no\\ndriver'}\n",
    "        xstart, xend, xlabels = set_custom_labels(g.columns, 2)\n",
    "        ax.set_xticks([x+1 for x in xlabels.values()], minor=False)\n",
    "        xlabels = ['/'.join(k) if isinstance(k, tuple) else k for k in xlabels.keys()]\n",
    "        xlabels = [di[x] if x in di else x for x in xlabels]\n",
    "        ax.set_xticklabels(xlabels, minor=False, fontsize=4, va='bottom')\n",
    "\n",
    "        ## genotype\n",
    "        # labels\n",
    "        xlabels = g.columns.get_level_values('genotype_short')\n",
    "        ax.set_xticks([x+1 for x in np.arange(len(xlabels))], minor=True)\n",
    "        ax.set_xticklabels(['/'.join(x) if isinstance(x, tuple) else x for x in xlabels], \n",
    "                           minor=True, fontsize=5)\n",
    "                \n",
    "        ax.tick_params(axis='x', which='minor', size=0, pad=-6)\n",
    "        ax.tick_params(axis='x', which='major', size=0, pad=3)\n",
    "        ax.tick_params(axis='y', which='major', labelsize=5)\n",
    "    \n",
    "    # set spine and tick properties\n",
    "    for ax in fig.get_axes():        \n",
    "        \n",
    "        for loc in ['top','bottom','left','right']:\n",
    "            ax.spines[loc].set_linewidth(1)\n",
    "        \n",
    "# show legend\n",
    "leg1 = ax3.legend(bbox_to_anchor=(1.05, 0.5), frameon=False,\n",
    "                  loc='upper left', borderaxespad=0.,\n",
    "                  handlelength=0.75, title='Background', prop={'size':6})\n",
    "    \n",
    "leg2 = ax3.legend(bbox_to_anchor=(1.05, 0.5), frameon=False,\n",
    "                  loc='center left', borderaxespad=0.,\n",
    "                  handlelength=0.75, title='Genotype', prop={'size':6})\n",
    "\n",
    "fig.savefig(dir_supp+'figures/figureSX_pheno_cross/figureSX_pheno_cross_reduced.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_supp+'figures/figureSX_pheno_cross/figureSX_pheno_cross_reduced.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "groups_sp = spores_norm_df.groupby(['env_evo','env_test'])\n",
    "groups_hy = hybrids_norm_df.groupby(['env_evo','env_test'])\n",
    "groups_cl = clone_norm_df.groupby(['env_evo','env_test'])\n",
    "\n",
    "param='growth_rate'\n",
    "\n",
    "fig = plt.figure(figsize=(6, 6), dpi=300)\n",
    "\n",
    "import matplotlib.gridspec as gridspec\n",
    "grid = gridspec.GridSpec(2, 2, hspace=0.3, wspace=0.2)\n",
    "\n",
    "gs = {}\n",
    "gs[('HU','HU')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,0])\n",
    "gs[('HU','YNB')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,1])\n",
    "gs[('RM','RM')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,0])\n",
    "gs[('RM','YNB')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,1])\n",
    "\n",
    "for (ii, (env_evo, env_test)) in enumerate([('HU','HU'),('HU','YNB'),('RM','RM'),('RM','YNB')]):\n",
    "    \n",
    "    # group by evolution and test environment\n",
    "    gsp = groups_sp.get_group((env_evo, env_test))\n",
    "    ghy = groups_hy.get_group((env_evo, env_test))\n",
    "    gcl = groups_cl.get_group((env_evo, env_test))\n",
    "\n",
    "    # apply sorting ranks to reorder rows\n",
    "    gsp.loc[:,u'rank_background'] = gsp[u'background'].map(dict_background)\n",
    "    gsp.loc[:,u'rank_population'] = gsp[u'population'].map(dict_population)\n",
    "    gsp.loc[:,u'rank_genotype'] = gsp[u'genotype_short'].map(dict_genotype)\n",
    "    gsp.loc[:,u'rank_gene'] = gsp[u'gene'].map(dict_gene)\n",
    "    gsp.sort_values(by=[u'rank_background',u'rank_population',u'rank_genotype',u'rank_gene'],\n",
    "                    ascending=True,inplace=True)\n",
    "    \n",
    "    ghy.loc[:,u'rank_background_MATa'] = ghy[u'background_MATa'].map(dict_background)\n",
    "    ghy.loc[:,u'rank_background_MATα'] = ghy[u'background_MATα'].map(dict_background)\n",
    "    ghy.loc[:,u'rank_population_MATa'] = ghy[u'population_MATa'].map(dict_population)\n",
    "    ghy.loc[:,u'rank_population_MATα'] = ghy[u'population_MATα'].map(dict_population)\n",
    "    ghy.loc[:,u'rank_genotype_MATa'] = ghy[u'genotype_short_MATa'].map(dict_genotype)\n",
    "    ghy.loc[:,u'rank_genotype_MATα'] = ghy[u'genotype_short_MATα'].map(dict_genotype)\n",
    "    ghy.loc[:,u'rank_gene_MATa'] = ghy[u'gene_MATa'].map(dict_gene)\n",
    "    ghy.loc[:,u'rank_gene_MATα'] = ghy[u'gene_MATα'].map(dict_gene)\n",
    "    ghy.sort_values(by=[u'rank_background_MATa',u'rank_background_MATα',\n",
    "                        u'rank_population_MATa',u'rank_population_MATα',\n",
    "                        u'rank_gene_MATa',u'rank_gene_MATα',\n",
    "                        u'rank_genotype_MATa',u'rank_genotype_MATα'],\n",
    "                    ascending=True, inplace=True)\n",
    "\n",
    "    ### heatmap ###\n",
    "    ax1 = plt.subplot(gs[(env_evo, env_test)][0])\n",
    "    \n",
    "    ax1.text(-0.1, 1.2,  chr(ii + ord('A')), transform=ax1.transAxes,\n",
    "             fontsize=12, fontweight='bold', va='top', ha='right')\n",
    "    \n",
    "    # spores\n",
    "    S = gsp\n",
    "    S = S.groupby([u'mating',u'background',u'population',u'genotype_short',u'gene',u'tetrad',u'spore'],\n",
    "                    sort=False)[param].agg(np.mean).reset_index()\n",
    "    S = S.set_index('mating')[param]\n",
    "    S_ref = S.values.mean()\n",
    "        \n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    \n",
    "    heatmap_spores(S-S_ref, ax1, title, \n",
    "                   xlabel, ylabel, xticklabels, yticklabels, \n",
    "                   cmap=plt.cm.RdBu_r, radius=0.5)\n",
    "    \n",
    "    # hybrids    \n",
    "    H = ghy\n",
    "    H = H.groupby([u'background_MATa',u'population_MATa',u'gene_MATa',u'genotype_short_MATa',u'tetrad_MATa',u'spore_MATa',\n",
    "                   u'background_MATα',u'population_MATα',u'gene_MATα',u'genotype_short_MATα',u'tetrad_MATα',u'spore_MATα'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    H = H.unstack(level=[u'background_MATα',u'population_MATα',u'gene_MATα',u'genotype_short_MATα',u'tetrad_MATα',u'spore_MATα'])\n",
    "    H = H.fillna(0)\n",
    "    H_ref = H.loc['ancestral','ancestral'].loc['WAxNA','WAxNA'].values.mean()\n",
    "\n",
    "    title = environments[env_test]\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    \n",
    "    heatmap_hybrids(H-H_ref, ax1, title,\n",
    "                    xlabel, ylabel, xticklabels, yticklabels, \n",
    "                    fold=False, cmap=plt.cm.RdBu_r, pad=10)\n",
    "    \n",
    "    ax1.xaxis.set_ticks_position('top')\n",
    "    ax1.yaxis.set_ticks_position('left')\n",
    "    \n",
    "    ## genotype\n",
    "    # labels\n",
    "    xlabels = H.columns.get_level_values(u'genotype_short_MATα')\n",
    "    ax1.set_xticks([x+0.5 for x in np.arange(len(xlabels))], minor=True)\n",
    "    ax1.set_xticklabels(xlabels, minor=True, fontsize=5)\n",
    "    \n",
    "    ylabels = H.index.get_level_values(u'genotype_short_MATa')\n",
    "    ax1.set_yticks([y+0.5 for y in np.arange(len(ylabels))], minor=True)\n",
    "    ax1.set_yticklabels(ylabels, minor=True, fontsize=5, rotation=90)\n",
    "    \n",
    "    ## gene\n",
    "    # labels\n",
    "    xstart, xend, xlabels = set_custom_labels(H.columns, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(v+0.5, 1.08), xycoords=(\"data\", \"axes fraction\"),\n",
    "                     ha='center', va='top', annotation_clip=False, fontsize=4, style='italic')\n",
    "    ystart, yend, ylabels = set_custom_labels(H.index, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(-0.08, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                     ha='left', va='center', annotation_clip=False, fontsize=4, style='italic', rotation=90)\n",
    "        \n",
    "    ## population\n",
    "    # labels\n",
    "    xstart, xend, xlabels = set_custom_labels(H.columns, 1)\n",
    "    xlabels = {k: v for k,v in xlabels.items() if k != u'WAxNA'}\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax1.annotate(k, xy=(v+0.5, 1.08), xycoords=(\"data\", \"axes fraction\"),\n",
    "                     ha='center', va='top', annotation_clip=False, fontsize=4)\n",
    "    ystart, yend, ylabels = set_custom_labels(H.index, 1)\n",
    "    ylabels = {k: v for k,v in ylabels.items() if k != u'WAxNA'}\n",
    "    for k, v in ylabels.iteritems():\n",
    "        ax1.annotate(k, xy=(-0.08, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                     ha='left', va='center', annotation_clip=False, fontsize=4, rotation=90)\n",
    "    \n",
    "    ## background\n",
    "    xstart, xend, xlabels = set_custom_labels(H.columns, 0)\n",
    "    ax1.set_xticks([x+0.5 for x in xlabels.values()], minor=False)\n",
    "    ax1.set_xticklabels(xlabels.keys(), minor=False, fontsize=5)\n",
    "    xgrid=[xst+1. for xst in list(set(xstart.values()))]\n",
    "    [ax1.axvline(g, lw=1, ls=\"-\", color=\"black\") for g in xgrid]\n",
    "    \n",
    "    ystart, yend, ylabels= set_custom_labels(H.columns, 0)\n",
    "    ax1.set_yticks([y+0.5 for y in ylabels.values()], minor=False)\n",
    "    ax1.set_yticklabels(ylabels.keys(), minor=False, fontsize=5, rotation=90)\n",
    "    ygrid=[yst+1. for yst in list(set(ystart.values()))]\n",
    "    [ax1.axhline(g, lw=1, ls=\"-\", color=\"black\") for g in ygrid]\n",
    "    \n",
    "    ax1.tick_params(axis='both', which='minor', size=0, pad=2)\n",
    "    ax1.tick_params(axis='both', which='major', size=0, pad=15)\n",
    "\n",
    "fig.savefig(dir_supp+'figures/figureSX_pheno_cross/figureSX_pheno_cross_extended.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_supp+'figures/figureSX_pheno_cross/figureSX_pheno_cross_extended.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for (ii, (env_evo, env_test)) in enumerate([('HU','HU'),('RM','RM')]):#,('HU','YNB'),('RM','RM'),('RM','YNB')]):\n",
    "    \n",
    "    # group by evolution and test environment\n",
    "    gsp = groups_sp.get_group((env_evo, env_test))\n",
    "    ghy = groups_hy.get_group((env_evo, env_test))\n",
    "    gcl = groups_cl.get_group((env_evo, env_test))\n",
    "    \n",
    "    # filter by dictionary\n",
    "    gsp = gsp[(gsp['genotype'].isin([u'fpr1Δ',u'tor1Δ']))]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To examine the average mutational effect of beneficial variants, we considered both the (Fig. 5) and the (Fig. ) as an average over many possible backgrounds. To determine fitness effects of the evolved mutations in HU (*RNR2*, *RNR4*) and RM (*FPR1*, *TOR1*), we isolated individuals from ancestral and evolved populations. From these diploid cells, we sporulated and selected haploid segregants of each mating type (48 in HU and 52 in RM). We crossed the MATa and MATα versions to create hybrids. A number of independent spores and hybrids were used to measure biological variability of ancestral and evolved backgrounds. We found all de novo mutations to be unconditionally beneficial, yet their association to genetic backgrounds idiosyncratic effects.\n",
    "\n",
    "Figure 5. Average mutational effect in PSD95pdz3 when binding the wild-type ligand CRIPT ($\\langle\\Delta\\tau_i^x\\rangle$, top) or a class-switching T22F variant ($\\langle\\Delta\\tau\\rangle$, bottom). T22F contains a Thr to Phe mutation at position minus two of the peptide ligand (Fig. 1a).\n",
    "\n",
    "(A) block diagonal elements $(b,g)_a=(b,g)_\\alpha$ off-diagonal elements $(b,g)_a\\neq (b,g)_\\alpha$.\n",
    "(B-C) Average growth rate of the spores and the hybrids with respect to the ancestral WAxNA individuals, indicating the median and the 25% / 75% percentiles.\n",
    "\n",
    "(D-E) . (-- (homozygote); -+ (heterozygote); ++ (homozygote)). The data matrix shows the functional cost of every genotype $g$ at each background $b$ relative to wild-type PSD95pdz3\n",
    "\n",
    "Selected clones (circled) are. The difference, or epistasis, between mutational effects $\\langle\\Delta\\Delta\\tau_g\\rangle_k$\n",
    "\n",
    "Mutations rise or fall in frequency because of their own effect on fitness, their association to genetic backgrounds, and stochastic fluctuations (genetic drift)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Background-averaged measurements ancestral and evolved isolates in two environments. Color scale for all matrices is shown to the right of A and the dimensions of each matrix, in units of hours. (A) *RNR2* and *RNR4*, (B) *FPR1* and *TOR1*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now group the dataframe by environmental condition during the propagation and testing. We now average over the ensemble of . For a given tetrad, we compare "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis of variance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "labels = {\n",
    "    'gene':'de_novo_gene',\n",
    "    'genotype_short':'de_novo_genotype'\n",
    "}\n",
    "\n",
    "sdf = spores_norm_df\n",
    "sdf.rename(columns=labels, inplace=True)\n",
    "sdf = sdf[sdf.background.isin(['ancestral','evolved'])]\n",
    "sdf = sdf[np.isfinite(sdf['growth_rate'])]\n",
    "\n",
    "hdf = hybrids_norm_df\n",
    "hdf.rename(columns=labels, inplace=True)\n",
    "hdf = hdf[hdf.background.isin([('ancestral','ancestral'),('ancestral','evolved'),('evolved','evolved')])]\n",
    "hdf = hdf[np.isfinite(hdf['growth_rate'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ANOVA (frequentist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# fig, axes = plt.subplots(nrows=4, ncols=1, figsize=(2, 5), sharex=True, sharey=True)\n",
    "\n",
    "fig = plt.figure(figsize=(6, 1), dpi=300)\n",
    "\n",
    "import matplotlib.gridspec as gridspec\n",
    "    \n",
    "formulas = ['growth_rate ~ background + de_novo_gene',\n",
    "            'growth_rate ~ background + de_novo_gene + de_novo_genotype',\n",
    "            'growth_rate ~ background + de_novo_gene + de_novo_genotype',\n",
    "            'growth_rate ~ background + de_novo_gene + de_novo_genotype + auxotrophy',\n",
    "            'growth_rate ~ background + de_novo_gene + de_novo_genotype + auxotrophy + tetrad + C(spore)']\n",
    "\n",
    "grid = gridspec.GridSpec(nrows=1, ncols=5, hspace=0.2, wspace=0.22)\n",
    "\n",
    "gs = {}\n",
    "for ii in np.arange(len(formulas)):\n",
    "    gs[ii] = gridspec.GridSpecFromSubplotSpec(2, 1, subplot_spec=grid[ii], hspace=0)\n",
    "\n",
    "for ii, formula in enumerate(formulas):\n",
    "    \n",
    "    msdf = sdf.groupby(['env_evo','env_test']).apply(mean_vectors, formula) \n",
    "    mhdf = hdf.groupby(['env_evo','env_test']).apply(mean_vectors, formula)\n",
    "        \n",
    "    tsdf = sdf.groupby(['env_evo','env_test']).apply(variance_vectors, formula) \n",
    "    thdf = hdf.groupby(['env_evo','env_test']).apply(variance_vectors, formula)\n",
    "    df = pd.concat([tsdf, thdf], keys=['spores','hybrids'], names=['type'])\n",
    "    \n",
    "    # fix index\n",
    "    df = df.swaplevel('type', 'env_evo')\n",
    "    df = df.swaplevel('type', 'env_test')\n",
    "    df = df.sort_index(level=['env_evo','env_test','type'])\n",
    "    \n",
    "    # fix columns\n",
    "    df = df.rename(columns={'C(spore)':'spore'})\n",
    "  \n",
    "    df = df.sort_index(axis=1)\n",
    "#     order = ['background','de_novo_gene','de_novo_genotype','auxotrophy','tetrad','spore']\n",
    "#     order = pd.Series(list(set(order) & set(df.columns)))\n",
    "#     df = df[order]\n",
    "    \n",
    "    ax1 = plt.subplot(gs[ii][0])\n",
    "    utils.simple_axes(ax1)\n",
    "    \n",
    "    s = df.ix['HU']\n",
    "    s = s.sort_index(ascending=False)\n",
    "    s.plot(ax=ax1, kind='barh', stacked=True, \n",
    "           color=[factor_dict['color'][x] for x in s.columns],\n",
    "           align='center', legend=False, width=0.75)\n",
    "    \n",
    "    bars = ax1.patches\n",
    "    hatches = ''.join([factor_dict['hatch'][x] for x in s.index.get_level_values('type')]*len(s)*2)\n",
    "\n",
    "    for bar, hatch in zip(bars, hatches):\n",
    "        bar.set_hatch(hatch)\n",
    "    \n",
    "    ax1.set_title(formula, fontsize=4)\n",
    "    ax1.set_xlabel('')\n",
    "    ax1.set_xticks([])\n",
    "    ax1.set_xticklabels([])    \n",
    "    ystart, yend, ylabels = set_custom_labels(s.index, 0)\n",
    "    ax1.set_yticks(ylabels.values(), minor=False)\n",
    "    ax1.set_yticklabels(ylabels.keys(), minor=False, fontsize=5)\n",
    "        \n",
    "    ax2 = plt.subplot(gs[ii][1])\n",
    "    utils.simple_axes(ax2)\n",
    "    \n",
    "    h = df.ix['RM']\n",
    "    h = h.sort_index(ascending=False)\n",
    "    h.plot(ax=ax2, kind='barh', stacked=True, \n",
    "           color=[factor_dict['color'][x] for x in h.columns],\n",
    "           align='center', legend=False, width=0.75)\n",
    "    \n",
    "    bars = ax2.patches\n",
    "    hatches = ''.join([factor_dict['hatch'][x] for x in h.index.get_level_values('type')]*len(h)*2)\n",
    "\n",
    "    for bar, hatch in zip(bars, hatches):\n",
    "        bar.set_hatch(hatch)\n",
    "    \n",
    "    ax2.set_xlabel('Variance', fontsize=5)    \n",
    "    ystart, yend, ylabels = set_custom_labels(h.index, 0)\n",
    "    ax2.set_yticks(ylabels.values(), minor=False)\n",
    "    ax2.set_yticklabels(ylabels.keys(), minor=False, fontsize=5)\n",
    "    \n",
    "    for ax in [ax1,ax2]:\n",
    "        ax.set_xlim((0,1))\n",
    "#         ax.set_yticks([])\n",
    "#         ax.set_yticklabels([])\n",
    "        ax.set_ylabel('', fontsize=6)\n",
    "\n",
    "    leg2 = ax2.legend(bbox_to_anchor=(1.05, 1.0), frameon=False,\n",
    "                      loc='center left', borderaxespad=0.,\n",
    "                      handlelength=0.75, prop={'size':4})\n",
    "    \n",
    "for ax in fig.get_axes():        \n",
    "        \n",
    "    plt.setp(ax.get_xticklabels(), fontsize=5)\n",
    "    plt.setp(ax.get_yticklabels(), fontsize=5)\n",
    "    \n",
    "    ax.tick_params(size=2)\n",
    "        \n",
    "    for loc in ['top','bottom','left','right']:\n",
    "        ax.spines[loc].set_linewidth(1)\n",
    "\n",
    "fig.savefig(dir_supp+'figures/figureSX_pheno_cross/figureSX_pheno_cross_anova.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_supp+'figures/figureSX_pheno_cross/figureSX_pheno_cross_anova.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hierarchical model (Bayesian)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from pymc3 import  *\n",
    "\n",
    "with Model() as model_sat:\n",
    "    grp_mean = Normal('grp_mean', mu=0, sd=10)\n",
    "    grp_sd = Uniform('grp_sd', 0, 200)\n",
    "\n",
    "    # Define priors for intercept and regression coefficients.\n",
    "    priors = {'Intercept': Normal.dist(mu=test_df.doubling_time.mean(), sd=test_df.doubling_time.std()),\n",
    "          'env_evo': Normal.dist(mu=grp_mean, sd=grp_sd),\n",
    "          'env_test': Normal.dist(mu=grp_mean, sd=grp_sd),\n",
    "          'background': Normal.dist(mu=grp_mean, sd=grp_sd),\n",
    "          'genotype': Normal.dist(mu=grp_mean, sd=grp_sd)\n",
    "    }\n",
    "    glm.glm('growth_rate ~ env_evo + env_test + population + background + genotype', \n",
    "                    test_df,\n",
    "                priors=priors)\n",
    "    trace_sat = sample(10, NUTS(), progressbar=True)\n",
    "\n",
    "# scatter_matrix(trace_to_dataframe(trace_sat), figsize=(12,12));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Troubleshooting"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We focus on the block diagonal elements of the matrix to identify spores and hybrids that may have been mistyped for the mutation or the mating locus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# fig = plt.figure(figsize=(15, 36))\n",
    "\n",
    "# dict_panels = {\n",
    "# ('HU','HU','RNR2'):(0,0),('HU','YNB','RNR2'):(0,1),\n",
    "# ('HU','HU','RNR4'):(1,0),('HU','YNB','RNR4'):(1,1),\n",
    "# ('RM','RM','null'):(2,0),('RM','YNB','null'):(2,1),\n",
    "# ('RM','RM','FPR1'):(3,0),('RM','YNB','FPR1'):(3,1),\n",
    "# ('RM','RM','TOR1'):(4,0),('RM','YNB','TOR1'):(4,1)\n",
    "# }\n",
    "    \n",
    "# for ii, (env_evo,env_test,gene) in enumerate([('HU','HU','RNR2'),('HU','YNB','RNR2'),('HU','HU','RNR4'),('HU','YNB','RNR4'),\n",
    "#                                          ('RM','RM','null'),('RM','YNB','null'),('RM','RM','FPR1'),('RM','YNB','FPR1'),('RM','RM','TOR1'),('RM','YNB','TOR1')]):\n",
    "                    \n",
    "#     gs = spores_norm_df[(spores_norm_df.gene==gene)].groupby(['env_evo','env_test']).get_group((env_evo,env_test))\n",
    "#     gh = hybrids_norm_df[(hybrids_norm_df.gene==(gene,gene))].groupby(['env_evo','env_test']).get_group((env_evo,env_test))\n",
    "\n",
    "#     H_ref = gh[(gh.population==('ancestral','ancestral'))][param].dropna().median()\n",
    "    \n",
    "#     gh = gh.groupby([u'population_MATa',u'background_MATa',u'genotype_MATa',u'tetrad_MATa',u'spore_MATa',\n",
    "#                      u'population_MATα',u'background_MATα',u'genotype_MATα',u'tetrad_MATα',u'spore_MATα'], \n",
    "#                     sort=False).agg(np.mean)[param]#.dropna()\n",
    "        \n",
    "#     ax = plt.subplot2grid((5, 2), dict_panels[(env_evo,env_test,gene)])#dict_panels[(env_evo,env_test)])\n",
    "    \n",
    "#     S = {}\n",
    "#     S[u'MATa'] = gs[gs[u'mating']==u'MATa']\n",
    "#     S[u'MATα'] = gs[gs[u'mating']==u'MATα']\n",
    "\n",
    "#     H = gh.unstack(level=[u'population_MATα',u'background_MATα',\n",
    "#                           u'genotype_MATα',u'tetrad_MATα',u'spore_MATα'])\n",
    "            \n",
    "#     H = H#-H_ref\n",
    "      \n",
    "#     pheno.heatmap_cross(H, title=environments[env_test])\n",
    "        \n",
    "#     dict_mat = {u'MATα':{'x':range(0,len(H)), 'y':[.5+len(H)]*len(H)},\n",
    "#                 u'MATa':{'x':[-1.5]*len(H), 'y':range(0,len(H))}}\n",
    "    \n",
    "#     for mating in dict_mat.iterkeys():\n",
    "#         data = map(list, zip(*[dict_mat[mating]['x'], dict_mat[mating]['y']]))\n",
    "#         circles = [plt.Circle([x, y], 0.5) for (x,y) in data]\n",
    "#         col = PatchCollection(circles)\n",
    "\n",
    "#         s = S[mating].groupby([u'population',u'background',u'genotype',u'tetrad',u'spore'],sort=False)[param].median()\n",
    "#         s_ref = S[mating][(S[mating].population=='ancestral')][param].median()\n",
    "        \n",
    "#         col.set(array=s-s_ref, cmap='RdBu')\n",
    "#         col.set_clip_on(False)\n",
    "\n",
    "#         ax.add_collection(col)\n",
    "\n",
    "# fig.tight_layout()\n",
    "fig.subplots_adjust(left=0.18,bottom=0.1,right=0.95,top=0.98,hspace=0.80,wspace=0.85)\n",
    "fig.savefig(dir_paper+'figures/figure5/figure5_error.png', dpi=300)\n",
    "fig.savefig(dir_paper+'figures/figure5/figure5_error.svg', dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are potential genotyping errors for mutations in *FPR1* and *TOR1*. *TOR1* tetrads A, I and S. *FPR1* tetrad D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# fig = plt.figure(figsize=(15, 36))\n",
    "\n",
    "# dict_panels = {\n",
    "# ('RM','RM','TOR1','A'):(0,0),('RM','YNB','TOR1','A'):(0,1),\n",
    "# ('RM','RM','TOR1','I'):(1,0),('RM','YNB','TOR1','I'):(1,1),\n",
    "# ('RM','RM','TOR1','S'):(2,0),('RM','YNB','TOR1','S'):(2,1),\n",
    "# ('RM','RM','FPR1','D'):(3,0),('RM','YNB','FPR1','D'):(3,1)\n",
    "# }\n",
    "    \n",
    "# for ii, (env_evo,env_test,gene,tetrad) in enumerate([('RM','RM','TOR1','A'),('RM','YNB','TOR1','A'),\n",
    "#                                                      ('RM','RM','TOR1','I'),('RM','YNB','TOR1','I'),\n",
    "#                                                      ('RM','RM','TOR1','S'),('RM','YNB','TOR1','S'),\n",
    "#                                                      ('RM','RM','FPR1','D'),('RM','YNB','FPR1','D')]):\n",
    "                    \n",
    "#     gs = spores_norm_df[(spores_norm_df.gene==gene)&\\\n",
    "#                         (spores_norm_df.tetrad==tetrad)].groupby(['env_evo','env_test']).get_group((env_evo,env_test))\n",
    "#     gh = hybrids_norm_df[(hybrids_norm_df.gene==(gene,gene))&\\\n",
    "#                          (hybrids_norm_df.tetrad==(tetrad,tetrad))].groupby(['env_evo','env_test']).get_group((env_evo,env_test))\n",
    "\n",
    "#     S_ref = gs[(gs.population=='ancestral')][param].median()\n",
    "\n",
    "#     H_ref = gh[(gh.population==('ancestral','ancestral'))][param].dropna().median()\n",
    "    \n",
    "#     gs = gs.groupby([u'mating',u'population',u'background',u'genotype',u'tetrad',u'spore'], \n",
    "#                     sort=False).agg(np.mean)[param]\n",
    "    \n",
    "#     gh = gh.groupby([u'population_MATa',u'background_MATa',u'genotype_MATa',u'tetrad_MATa',u'spore_MATa',\n",
    "#                      u'population_MATα',u'background_MATα',u'genotype_MATα',u'tetrad_MATα',u'spore_MATα'], \n",
    "#                     sort=False).agg(np.mean)[param]#.dropna()\n",
    "        \n",
    "#     ax = plt.subplot2grid((4, 2), dict_panels[(env_evo,env_test,gene,tetrad)])\n",
    "    \n",
    "# #     S = {}\n",
    "# #     S[u'MATa'] = gs.loc[u'MATa']\n",
    "# #     S[u'MATα'] = gs.loc[u'MATα']\n",
    "#     S = gs\n",
    "    \n",
    "#     H = gh.unstack(level=[u'population_MATα',u'background_MATα',\n",
    "#                           u'genotype_MATα',u'tetrad_MATα',u'spore_MATα'])\n",
    "    \n",
    "#     print S[u'MATa'].shape, S[u'MATα'].shape, H.shape\n",
    "\n",
    "# #     H = H#-H_ref\n",
    "      \n",
    "#     pheno.heatmap_spores(S-S_ref)\n",
    "#     pheno.heatmap_hybrids(H-H_ref, title=environments[env_test])\n",
    "\n",
    "# # fig.tight_layout()\n",
    "# fig.subplots_adjust(left=0.18,bottom=0.1,right=0.95,top=0.98,hspace=0.80,wspace=0.85)\n",
    "# # fig.savefig(dir_paper+'figures/figure5/figure5_error.png', dpi=300)\n",
    "# # fig.savefig(dir_paper+'figures/figure5/figure5_error.svg', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# dict_panels = {\n",
    "# ('RM','RM','','A'):(0,0),('RM','YNB','','A'):(0,1),\n",
    "# ('RM','RM','','B'):(1,0),('RM','YNB','','B'):(1,1),\n",
    "# ('RM','RM','TOR1','A'):(2,0),('RM','YNB','TOR1','A'):(2,1),\n",
    "# ('RM','RM','TOR1','I'):(3,0),('RM','YNB','TOR1','I'):(3,1),\n",
    "# ('RM','RM','TOR1','S'):(4,0),('RM','YNB','TOR1','S'):(4,1),\n",
    "# ('RM','RM','FPR1','D'):(5,0),('RM','YNB','FPR1','D'):(5,1)\n",
    "# }\n",
    "\n",
    "# dpi=300\n",
    "# nrows=6\n",
    "# ncols=2\n",
    "\n",
    "# param='Generation Time'\n",
    "\n",
    "# height, width = np.array([nrows*3, ncols*6], dtype=float)\n",
    "\n",
    "# fig, axes = plt.subplots(nrows,ncols,figsize=(width, height),\n",
    "#                          sharex='col',sharey='row')\n",
    "    \n",
    "# for ii, (env_evo,env_test,gene,tetrad) in enumerate([('RM','RM','','A'),('RM','YNB','','A'),\n",
    "#                                                      ('RM','RM','','B'),('RM','YNB','','B'),\n",
    "#                                                      ('RM','RM','TOR1','A'),('RM','YNB','TOR1','A'),\n",
    "#                                                      ('RM','RM','TOR1','I'),('RM','YNB','TOR1','I'),\n",
    "#                                                      ('RM','RM','TOR1','S'),('RM','YNB','TOR1','S'),\n",
    "#                                                      ('RM','RM','FPR1','D'),('RM','YNB','FPR1','D')]):\n",
    "                    \n",
    "#     gs = spores_norm_df[(spores_norm_df.gene==gene)&\\\n",
    "#                         (spores_norm_df.tetrad==tetrad)].groupby(['env_evo','env_test']).get_group((env_evo,env_test))\n",
    "       \n",
    "#     ax = axes[dict_panels[(env_evo,env_test,gene,tetrad)]]\n",
    "    \n",
    "#     gs.boxplot(ax=ax, column=param, by=['population','genotype','mating','auxotrophy','tetrad','spore'], vert=False)\n",
    "    \n",
    "#     if ii<2:\n",
    "#         # set title\n",
    "#         ax.set_title(environments[env_test])\n",
    "#     else:\n",
    "#         ax.set_title('')\n",
    "\n",
    "# #     ax.annotate('%s %s (spores)'%(gene, tetrad),\n",
    "# #                 xy=(.8, .2), xycoords='axes fraction', fontsize=12,\n",
    "# #                 horizontalalignment='center', verticalalignment='right')\n",
    "\n",
    "# # set axes labels\n",
    "# for ax in [axes[(5,0)],axes[(5,1)]]:\n",
    "#     ax.set_xlabel(r'Rel. doubling time $\\langle\\tau\\rangle_{b,g,k}^{a/\\alpha}$', fontsize=10)\n",
    "\n",
    "# plt.suptitle('')\n",
    "\n",
    "# fig.tight_layout()\n",
    "\n",
    "# for fmt in ['pdf','png','svg']:\n",
    "#     fig.savefig(dir_paper+'figures/figure5/cross_label_errors_spores_boxplot.'+fmt, dpi=300)\n",
    "    \n",
    "# plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# dict_panels = {\n",
    "# ('RM','RM','','A'):(0,0),('RM','YNB','','A'):(0,1),\n",
    "# ('RM','RM','','B'):(1,0),('RM','YNB','','B'):(1,1),\n",
    "# ('RM','RM','TOR1','A'):(2,0),('RM','YNB','TOR1','A'):(2,1),\n",
    "# ('RM','RM','TOR1','I'):(3,0),('RM','YNB','TOR1','I'):(3,1),\n",
    "# ('RM','RM','TOR1','S'):(4,0),('RM','YNB','TOR1','S'):(4,1),\n",
    "# ('RM','RM','FPR1','D'):(5,0),('RM','YNB','FPR1','D'):(5,1)\n",
    "# }\n",
    "\n",
    "# dpi=300\n",
    "# nrows=6\n",
    "# ncols=2\n",
    "\n",
    "# param='Generation Time'\n",
    "\n",
    "# height, width = np.array([nrows*3, ncols*6], dtype=float)\n",
    "\n",
    "# fig, axes = plt.subplots(nrows,ncols,figsize=(width, height),\n",
    "#                          sharex='col',sharey='row')\n",
    "    \n",
    "# for ii, (env_evo,env_test,gene,tetrad) in enumerate([('RM','RM','','A'),('RM','YNB','','A'),\n",
    "#                                                      ('RM','RM','','B'),('RM','YNB','','B'),\n",
    "#                                                      ('RM','RM','TOR1','A'),('RM','YNB','TOR1','A'),\n",
    "#                                                      ('RM','RM','TOR1','I'),('RM','YNB','TOR1','I'),\n",
    "#                                                      ('RM','RM','TOR1','S'),('RM','YNB','TOR1','S'),\n",
    "#                                                      ('RM','RM','FPR1','D'),('RM','YNB','FPR1','D')]):\n",
    "                    \n",
    "#     gh = hybrids_norm_df[(hybrids_norm_df.gene==(gene,gene))&\\\n",
    "#                          (hybrids_norm_df.tetrad==(tetrad,tetrad))].groupby(['env_evo','env_test']).get_group((env_evo,env_test))\n",
    "       \n",
    "#     ax = axes[dict_panels[(env_evo,env_test,gene,tetrad)]]\n",
    "    \n",
    "#     gh.boxplot(ax=ax, column=param, by=['population','genotype','tetrad','spore'], vert=False)\n",
    "\n",
    "#     if ii<2:\n",
    "#         # set title\n",
    "#         ax.set_title(environments[env_test])\n",
    "#     else:\n",
    "#         ax.set_title('')\n",
    "        \n",
    "# #     ax.annotate('%s %s (hybrids)'%(gene, tetrad),\n",
    "# #                 xy=(.8, .2), xycoords='axes fraction', fontsize=12,\n",
    "# #                 horizontalalignment='center', verticalalignment='right')\n",
    "\n",
    "\n",
    "# # set axes labels\n",
    "# for ax in [axes[(5,0)],axes[(5,1)]]:\n",
    "#     ax.set_xlabel(r'Rel. doubling time $\\langle\\tau\\rangle_{b,g,k}^{a/\\alpha}$', fontsize=10)\n",
    "\n",
    "# plt.suptitle('')\n",
    "\n",
    "# fig.tight_layout()\n",
    "\n",
    "# for fmt in ['pdf','png','svg']:\n",
    "#     fig.savefig(dir_paper+'figures/figure5/cross_label_errors_hybrids_boxplot.'+fmt, dpi=300)\n",
    "\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "spores_abs_df = utils.load_data(dir_data+'pheno/pheno_spores10K_abs.pickle')\n",
    "spores_norm_df = utils.load_data(dir_data+'pheno/pheno_spores10K_norm.pickle')\n",
    "hybrids_abs_df = utils.load_data(dir_data+'pheno/pheno_hybrids10K_abs.pickle')\n",
    "hybrids_norm_df = utils.load_data(dir_data+'pheno/pheno_hybrids10K_norm.pickle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/ipykernel/__main__.py:23: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'environment'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-23-e71c5f1028fc>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     36\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     37\u001b[0m     \u001b[0;31m# spores\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 38\u001b[0;31m     \u001b[0mgs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mspores_norm_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'background'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'environment'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_group\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbackground\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0menv_test\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     39\u001b[0m     \u001b[0mS\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34mu'mating'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34mu'spore'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msort\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0magg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmean\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mparam\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     40\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/pandas/core/generic.pyc\u001b[0m in \u001b[0;36mgroupby\u001b[0;34m(self, by, axis, level, as_index, sort, group_keys, squeeze)\u001b[0m\n\u001b[1;32m   3434\u001b[0m         \u001b[0maxis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_axis_number\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3435\u001b[0m         return groupby(self, by=by, axis=axis, level=level, as_index=as_index,\n\u001b[0;32m-> 3436\u001b[0;31m                        sort=sort, group_keys=group_keys, squeeze=squeeze)\n\u001b[0m\u001b[1;32m   3437\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3438\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0masfreq\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnormalize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/pandas/core/groupby.pyc\u001b[0m in \u001b[0;36mgroupby\u001b[0;34m(obj, by, **kwds)\u001b[0m\n\u001b[1;32m   1309\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'invalid type: %s'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1310\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1311\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mklass\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mby\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1312\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1313\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/pandas/core/groupby.pyc\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, obj, keys, axis, level, grouper, exclusions, selection, as_index, sort, group_keys, squeeze)\u001b[0m\n\u001b[1;32m    416\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mgrouper\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    417\u001b[0m             grouper, exclusions, obj = _get_grouper(obj, keys, axis=axis,\n\u001b[0;32m--> 418\u001b[0;31m                                                     level=level, sort=sort)\n\u001b[0m\u001b[1;32m    419\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    420\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/pandas/core/groupby.pyc\u001b[0m in \u001b[0;36m_get_grouper\u001b[0;34m(obj, key, axis, level, sort)\u001b[0m\n\u001b[1;32m   2262\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2263\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mis_in_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgpr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# df.groupby('name')\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2264\u001b[0;31m             \u001b[0min_axis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgpr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgpr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgpr\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2265\u001b[0m             \u001b[0mexclusions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2266\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/pandas/core/frame.pyc\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1967\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1968\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1969\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_column\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1970\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1971\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_getitem_column\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/pandas/core/frame.pyc\u001b[0m in \u001b[0;36m_getitem_column\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1974\u001b[0m         \u001b[0;31m# get column\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1975\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_unique\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1976\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_item_cache\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1977\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1978\u001b[0m         \u001b[0;31m# duplicate columns & possible reduce dimensionality\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/pandas/core/generic.pyc\u001b[0m in \u001b[0;36m_get_item_cache\u001b[0;34m(self, item)\u001b[0m\n\u001b[1;32m   1089\u001b[0m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcache\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1090\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mres\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1091\u001b[0;31m             \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1092\u001b[0m             \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_box_item_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1093\u001b[0m             \u001b[0mcache\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/pandas/core/internals.pyc\u001b[0m in \u001b[0;36mget\u001b[0;34m(self, item, fastpath)\u001b[0m\n\u001b[1;32m   3209\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3210\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misnull\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3211\u001b[0;31m                 \u001b[0mloc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3212\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3213\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0misnull\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/pandas/core/index.pyc\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   1757\u001b[0m                                  'backfill or nearest lookups')\n\u001b[1;32m   1758\u001b[0m             \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_values_from_object\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1759\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1760\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1761\u001b[0m         indexer = self.get_indexer([key], method=method,\n",
      "\u001b[0;32mpandas/index.pyx\u001b[0m in \u001b[0;36mpandas.index.IndexEngine.get_loc (pandas/index.c:3979)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/index.pyx\u001b[0m in \u001b[0;36mpandas.index.IndexEngine.get_loc (pandas/index.c:3843)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/hashtable.pyx\u001b[0m in \u001b[0;36mpandas.hashtable.PyObjectHashTable.get_item (pandas/hashtable.c:12265)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/hashtable.pyx\u001b[0m in \u001b[0;36mpandas.hashtable.PyObjectHashTable.get_item (pandas/hashtable.c:12216)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'environment'"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAaUAAAGICAYAAADh4P9GAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAE+lJREFUeJzt3HuMpXV9x/H3h1s0Kk0IdBeKF1QqdUmkIq1BkEFdU/CS\nRlIhERFiTQHjrdVouqjbakVjRKw3SqxSjDcIFU0VBZqOlwUq0KLSGjFgrSuw0XhZFRSEb/84Z+k4\nzOycPTNn5juz71cyYec5v+fM7/zYPe95nvOck6pCkqQO9ljpCUiStINRkiS1YZQkSW0YJUlSG0ZJ\nktSGUZIktWGUJEltjBSlJMcm+UySrUnuT3LaCPscnmQ6yV1Jvp/kjYufriRpLRv1SOnhwDeBVwJ3\nLTQ4ySOAq4A7gCOBVwGvS/KaMecpSdoNZFc/0SHJz4GXV9XFOxlzFnAu8LtVdc9w2ybgzKp65CLm\nK0lawyb1mtJTga/sCNLQF4GDkjx6Qj9TkrTKTSpK64Fts7ZtAzK8TZKkB/HqO0lSG3tN6H7vBNbN\n2rYOqOFtvyWJH1UuSWtAVWUx+0/qSOla4Ngk+8zY9mzg9qr63lw7VJVfC3y9+c1vXvE5rJYv18p1\ncp2W/2spjPo+pYcleVKSI4b7PGr4/SOHt5+b5OoZu3ycwaXjFyXZkOQFwOuBdy3JrCVJa9KoR0pP\nAf4TuBF4CPA3wH8M/wuDixcO2TG4qrYDG4GDgOuB9wLvrKrzl2bakqS1aKTXlKrqS+wkYFV1xhzb\n/guYGntmepCpqamVnsKq4VqNxnUajeu0fHb5zbMTmURSHeYhSRpfEqrphQ6SJO0yoyRJasMoSZLa\nMEqSpDaMkiSpDaMkSWrDKEmS2jBKkqQ2jJIkqQ2jJElqwyhJktowSpKkNoySJKkNoyRJasMoSZLa\nMEqSpDaMkiSpDaMkSWrDKEmS2jBKkqQ2jJIkqQ2jJElqwyhJktowSpKkNoySJKkNoyRJasMoSZLa\nMEqSpDaMkiSpDaMkSWrDKEmS2jBKkqQ2jJIkqQ2jJElqwyhJktowSpKkNoySJKkNoyRJasMoSZLa\nMEqSpDaMkiSpDaMkSWrDKEmS2jBKkqQ2jJIkqQ2jJElqwyhJktowSpKkNoySJKkNoyRJasMoSZLa\nMEqSpDaMkiSpDaMkSWrDKEmS2jBKkqQ2jJIkqQ2jJElqwyhJktoYOUpJzk5yW5K7k9yQ5JgFxp+Y\n5Nok25P8MMnlSQ5d/JQlSWvVSFFKcjJwPvBW4AjgGuCKJAfPM/5xwKeB6eH4ZwIPAT63+ClLktaq\nVNXCg5LrgJuq6swZ224BLq2qTXOMPwn4JLBPDX9AkingX4EDqurHs8bXKPOQJPWVhKrKYu5jwSOl\nJHsDRwJXzbrpSuDoeXa7BvgF8LIkeyR5BHA68LXZQZIkaYdRTt/tD+wJbJu1fRuwfq4dquoO4ETg\nLcCvgZ8CG4DnjT1TSdKat9ck7jTJIcDlwEeATwCPAP4WuBQ4fq59Nm/e/MCfp6ammJqamsTUJElL\nZHp6munp6SW9zwVfUxqevrsLOKWqLpux/X3Ahqp6UGSSvB3YWFVHztj2e8D3gWOq6ppZ431NSZJW\nuWV5Tamq7gVuBDbOumkjsGUn93vfrG33j/ozJUm7p1EDcR5wepKXJjksyXuAA4ELAJKcm+TqGeM/\nCzw5yRuTPD7JkxmcyvtfBoGTJOlBRnpNqaouSbIfsIlBjG4GTqiqrcMh64FDZoz/6vC9TW8AXsfg\n9N91wJ9U1d1LOH9J0hoy0vuUJj4JX1OSpFVvWV5TkiRpuRglSVIbRkmS1IZRkiS1YZQkSW0YJUlS\nG0ZJktSGUZIktWGUJEltGCVJUhtGSZLUhlGSJLVhlCRJbRglSVIbRkmS1IZRkiS1YZQkSW0YJUlS\nG0ZJktSGUZIktWGUJEltGCVJUhtGSZLUhlGSJLVhlCRJbRglSVIbRkmS1IZRkiS1YZQkSW0YJUlS\nG0ZJktSGUZIktWGUJEltGCVJUhtGSZLUhlGSJLVhlCRJbRglSVIbRkmS1IZRkiS1YZQkSW0YJUlS\nG0ZJktSGUZIktWGUJEltGCVJUhtGSZLUhlGSJLVhlCRJbRglSVIbRkmS1IZRkiS1YZQkSW0YJUlS\nG0ZJktSGUZIktWGUJEltGCVJUhtGSZLUxshRSnJ2ktuS3J3khiTHjLDPq5N8K8mvkvwgydsWN11J\n0lq21yiDkpwMnA+cCWwBXg5ckeQPqmrrPPucB5wIvBa4Gfgd4MClmLQkaW1KVS08KLkOuKmqzpyx\n7Rbg0qraNMf4JwDfBA6vqltGuP8aZR6SpL6SUFVZzH0sePouyd7AkcBVs266Ejh6nt2eD9wKnJjk\n1iTfTXJRkgMWM1lJ0to2ymtK+wN7Attmbd8GrJ9nn8cCjwFOBk4DTgUOAz471iwlSbuFkV5TGsMe\nwD7AqVV1K0CSFwPfTnJUVV0/oZ8rSVrFRonSj4D7gHWztq8D7pxnnzuA3+wIEkBVfSfJfcCjgAdF\nafPmzQ/8eWpqiqmpqRGmJklaKdPT00xPTy/pfS7mQodvM7jQ4Zw5xm8EvgA8vqq+O9z2OOA7wB9V\n1Q2zxnuhgyStcktxocOoUXohcDGDS8G3AGcBZwBPrKqtSc4FjqqqZw3HB/ga8AvgNUCAdwN7V9XT\n5rh/oyRJq9xSRGmk15Sq6pIk+wGbGLzX6GbghBnvUVoPHDJjfCV5LvD3wJeAuxlcrfdXi5msJGlt\nG+lIaeKT8EhJkla9ZXmfkiRJy8UoSZLaMEqSpDaMkiSpDaMkSWrDKEmS2jBKkqQ2jJIkqQ2jJElq\nwyhJktowSpKkNoySJKkNoyRJasMoSZLaMEqSpDaMkiSpDaMkSWrDKEmS2jBKkqQ2jJIkqQ2jJElq\nwyhJktowSpKkNoySJKkNoyRJasMoSZLaMEqSpDaMkiSpDaMkSWrDKEmS2jBKkqQ2jJIkqQ2jJElq\nwyhJktowSpKkNoySJKkNoyRJasMoSZLaMEqSpDaMkiSpDaMkSWrDKEmS2jBKkqQ2jJIkqQ2jJElq\nwyhJktowSpKkNoySJKkNoyRJasMoSZLaMEqSpDaMkiSpDaMkSWrDKEmS2jBKkqQ2jJIkqQ2jJElq\nwyhJktowSpKkNoySJKmNkaOU5OwktyW5O8kNSY4Zcb9Dk/w8yfbxpylJ2h2MFKUkJwPnA28FjgCu\nAa5IcvAC++0NfAKYXtw0JUm7g1TVwoOS64CbqurMGdtuAS6tqk072e/dwL7Al4H3VtW+84yrUeYh\nSeorCVWVxdzHgkdKw6OdI4GrZt10JXD0TvZ7DnAi8IrFTFCStPsY5fTd/sCewLZZ27cB6+faIclB\nwIXAi6rqrkXNUJK025jU1XcfBT5QVTcMv1/U4Zwkafew1whjfgTcB6ybtX0dcOc8+xwPHJtk8/D7\nAHskuQc4u6o+NHuHzZs3P/DnqakppqamRpiaJGmlTE9PMz09vaT3uZgLHb7N4EKHc+YY/8RZm/4U\n+GvgKOD2qvrZrPFe6CBJq9xSXOgwypESwHnAxUmuB7YAZwEHAhcMJ3IucFRVPQugqv571kSPAu6v\nqm8tZrKSpLVtpChV1SVJ9gM2MYjRzcAJVbV1OGQ9cMhkpihJ2l2MdPpu4pPw9J0krXrL8j4lSZKW\ni1GSJLVhlCRJbRglSVIbRkmS1IZRkiS1YZQkSW0YJUlSG0ZJktSGUZIktWGUJEltGCVJUhtGSZLU\nhlGSJLVhlCRJbRglSVIbRkmS1IZRkiS1YZQkSW0YJUlSG0ZJktSGUZIktWGUJEltGCVJUhtGSZLU\nhlGSJLVhlCRJbRglSVIbRkmS1IZRkiS1YZQkSW0YJUlSG0ZJktSGUZIktWGUJEltGCVJUhtGSZLU\nhlGSJLVhlCRJbRglSVIbRkmS1IZRkiS1YZQkSW0YJUlSG0ZJktSGUZIktWGUJEltGCVJUhtGSZLU\nhlGSJLVhlCRJbRglSVIbRkmS1IZRkiS1YZQkSW0YJUlSG0ZJktSGUZIktWGUJEltGCVJUhsjRynJ\n2UluS3J3khuSHLOTsccluTzJ7Ul+meTrSc5YmilLktaqkaKU5GTgfOCtwBHANcAVSQ6eZ5ejgW8A\nJwEbgA8CFyY5ZdEzliStWamqhQcl1wE3VdWZM7bdAlxaVZtG+kHJp4A9qurP5ritRpmHJKmvJFRV\nFnMfCx4pJdkbOBK4atZNVzI4IhrVvsBPdmG8JGk3s9cIY/YH9gS2zdq+DXjmKD8kyXOBZ7BrEZMk\n7WZGidKiJHka8DHgFVV143zjNm/e/MCfp6ammJqamvTUJEmLMD09zfT09JLe54KvKQ1P390FnFJV\nl83Y/j5gQ1Udv5N9jwE+B5xTVe/dyThfU5KkVW5ZXlOqqnuBG4GNs27aCGzZyeSeDnweeNPOgiRJ\n0g6jnr47D7g4yfUMQnQWcCBwAUCSc4GjqupZw++ngH8B3g98Msm64f3cV1U/WrrpS5LWkpGiVFWX\nJNkP2MQgRjcDJ1TV1uGQ9cAhM3Z5CfBQ4LXDrx2+Bzx2sZOWJK1NI71PaeKT8DUlSVr1luU1JUmS\nlotRkiS1YZQkSW0YJUlSG0ZJktSGUZIktWGUJEltGCVJUhtGSZLUhlGSJLVhlCRJbRglSVIbRkmS\n1IZRkiS1YZQkSW0YJUlSG0ZJktSGUZIktWGUJEltGCVJUhtGSZLUhlGSJLVhlCRJbRglSVIbRkmS\n1IZRkiS1YZQkSW0YJUlSG0ZJktSGUZIktWGUJEltGCVJUhtGSZLUhlGSJLVhlCRJbRglSVIbRkmS\n1IZRkiS1YZQkSW0YJUlSG0ZJktSGUZIktWGUJEltGCVJUhtGSZLUhlGSJLVhlCRJbRglSVIbRkmS\n1IZRkiS1YZQkSW0YJUlSG0ZJktSGUZIktWGUJEltGCVJUhtGSZLUhlGSJLVhlCRJbYwcpSRnJ7kt\nyd1JbkhyzALjD08yneSuJN9P8sbFT1eStJaNFKUkJwPnA28FjgCuAa5IcvA84x8BXAXcARwJvAp4\nXZLXLMWkJUlr06hHSq8BPlxVH66qb1fVKxkE56x5xp8KPBR4SVV9q6r+GXgH8JeLnvFubHp6eqWn\nsGq4VqNxnUbjOi2fBaOUZG8GRztXzbrpSuDoeXZ7KvCVqrpnxrYvAgclefQ4E5X/MHaFazUa12k0\nrtPyGeVIaX9gT2DbrO3bgPXz7LN+nvHZyT6SpN2cV99JktpIVe18wOD03V3AKVV12Yzt7wM2VNXx\nc+zzT8B+VfW8GdueAvw78Niq+t6s8TufhCRpVaiqLGb/vUb4AfcmuRHYCFw246aNwKXz7HYt8PYk\n+8x4XenZwO2zgzT8GYt6EJKktWHU03fnAacneWmSw5K8BzgQuAAgyblJrp4x/uMMjq4uSrIhyQuA\n1wPvWsK5S5LWmAWPlACq6pIk+wGbGMToZuCEqto6HLIeOGTG+O1JNgLvB64HfgK8s6rOX8rJS5LW\nlgVfU5Ikabksy9V3fkTRaHZlnZIcl+TyJLcn+WWSryc5Yznnu1J29e/TjP0OTfLzJNsnPccOxlmn\nJK9O8q0kv0rygyRvW465rqQxnp9OTHJtku1Jfjj8d3jocs13JSQ5NslnkmxNcn+S00bYZ6zn8YlH\nyY8oGs2urhODNy5/AzgJ2AB8ELgwySnLMN0VM8Y67dhvb+ATwPSk59jBOOuU5DzgTOB1wGHAicCX\nJz/blTPG89PjgE8z+Ht0BPBM4CHA55Zjvivo4cA3gVcyuF5gpxb1PF5VE/0CrgMumLXtFuDv5hl/\nFvBTYJ8Z2zYB35/0XFfya1fXaZ77+BRw6Uo/lo7rBLwb+EfgJcD2lX4c3dYJeAJwD/D7Kz335ut0\nEnAvw5c+htumgPsYvA1mxR/TMqzZz4HTFhgz9vP4RI+U/Iii0Yy5TnPZl8FFJWvSuOuU5DkMfut/\nxeRm18eY6/R84FbgxCS3JvlukouSHDDBqa6oMdfpGuAXwMuS7DE8Ijgd+FpV/XhSc12Fxn4en/Tp\nOz+iaDTjrNNvSfJc4BnAPyzt1FrZ5XVKchBwIfCiqlrwtMMaMc7fp8cCjwFOBk5j8KHKhwGfncwU\nW9jldaqqOxj8gvMW4NcMjgY2AM+ba/xubOzncT9maA1I8jTgY8ArqurGlZ5PMx8FPlBVNwy/943a\nc9sD2Ac4taq2VNUW4MXAHyc5amWn1keSQ4DLgY8ATwGOY3A6a74PEtAumnSUfsTgXOu6WdvXAXfO\ns8+d84yvneyz2o2zTgAMrxT6PHBOVV04mem1Mc46HQ+8Ocm9Se4FPgQ8PMk9Sf58clNdUeOs0x3A\nb6rq1h0bquo7w/t51CQm2cA46/QXwNaqekNVfb2qvsog3scl2ZVT7Wvd2M/jE41SVd0L7PiIopk2\nAlvm2e1a4Ngk+8zYNu9HFK0FY64TSZ7OIEhvqqr3Tm6GPYy5ToczuErqScOvNzG4euhJrNHfbsdc\npy3AXsMjAeCBK832BPx39//2YBCyme6fcZsGxn8eX4YrNV4I/Ap4KYNz1O8BtgMHD28/F7h6xvh9\ngdsZfFTRBuAFwM+AV6/0VSfN1mmKwQuu72DwG8iOr/1X+rF0Wqc59t9drr7b1b9PYfDpK//GIOJ/\nyOCy5y0r/ViardMxwG+ANwKPB54MfAH4H+ChK/14JrhOD2Pwi9wRwC+Bc4bfP3KedRr7eXy5HtCZ\nwG3A3cO/+E+bcdtHgFtnjd8w/AdxF/ADBqemVvx/TKd1Gn5/3xxft6304+i0TnPsu1tEaZx1YvBL\nzaeGTx53AhcDB6z042i4TicNx20frtPlwGEr/TgmvEbHMTginP188+GdrNNYz+N+zJAkqQ3PgUqS\n2jBKkqQ2jJIkqQ2jJElqwyhJktowSpKkNoySJKkNoyRJasMoSZLa+D/iLEcelqt/nAAAAABJRU5E\nrkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x11441c910>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dict_population = {'parents':0, 'ancestral':1, 'evolved':2}\n",
    "dict_background = {'WA':0, 'NA':1, 'WAxNA':2}\n",
    "dict_gene = {'RNR2':0, 'RNR4':1, 'no driver':2, 'FPR1':3, 'TOR1':4}\n",
    "dict_genotype = {'RNR2':0, 'RNR2*':2, 'RNR4':1, 'RNR4*':3,\n",
    "                 'no driver':0, 'FPR1*':2, 'TOR1':1, 'TOR1*':3}\n",
    "dict_panels = {\n",
    "'WAxNA': {'HU':(0,0),'RM':(0,1)},\n",
    "'WAxNA-WExSA': {'HU':(1,0),'RM':(1,1)}\n",
    "}\n",
    "\n",
    "# apply sorting ranks to reorder rows    \n",
    "hybrids_norm_df.loc[:,u'rank_background_MATa'] = hybrids_norm_df[u'background_MATa'].map(dict_background)\n",
    "hybrids_norm_df.loc[:,u'rank_background_MATα'] = hybrids_norm_df[u'background_MATα'].map(dict_background)\n",
    "hybrids_norm_df.loc[:,u'rank_genotype_MATa'] = hybrids_norm_df[u'genotype_MATa'].map(dict_genotype)\n",
    "hybrids_norm_df.loc[:,u'rank_genotype_MATα'] = hybrids_norm_df[u'genotype_MATα'].map(dict_genotype)\n",
    "hybrids_norm_df.loc[:,u'rank_population_MATa'] = hybrids_norm_df[u'population_MATa'].map(dict_population)\n",
    "hybrids_norm_df.loc[:,u'rank_population_MATα'] = hybrids_norm_df[u'population_MATα'].map(dict_population)\n",
    "hybrids_norm_df.sort(columns=[u'rank_population_MATa',u'rank_population_MATα',\n",
    "                              u'rank_background_MATa',u'rank_background_MATα',\n",
    "                              u'rank_genotype_MATa',u'rank_genotype_MATα',\n",
    "                              u'tetrad_MATa',u'tetrad_MATα',\n",
    "                              u'spore_MATa',u'spore_MATα'],\n",
    "                ascending=[True,True,True,True,True,True,True,True,True,True],inplace=True)\n",
    "\n",
    "# environment attributes\n",
    "environments = {u'HU': u'YNB+HU 10 mg/ml', \n",
    "                u'RM': u'YNB+RM 0.025 μg/ml', \n",
    "                u'YNB': u'YNB (control)'}\n",
    "\n",
    "param='growth_rate'\n",
    "\n",
    "fig = plt.figure(figsize=(13,13))\n",
    "\n",
    "for (background,env_test) in [('WAxNA','HU'),('WAxNA','RM'),('WAxNA-WExSA','HU'),('WAxNA-WExSA','RM')]:\n",
    "    ax = plt.subplot2grid((2, 2), dict_panels[background][env_test])\n",
    "    \n",
    "    # spores\n",
    "    gs = spores_norm_df.groupby(['background','Environment']).get_group((background,env_test))\n",
    "    S = gs.groupby([u'mating',u'spore'], sort=False).agg(np.mean)[param]\n",
    "\n",
    "    # hybrids\n",
    "    gh = hybrids_norm_df.groupby(['background','Environment']).get_group(((background,background),env_test))\n",
    "    gh = gh.groupby([u'spore_MATa',\n",
    "                     u'spore_MATα'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    H = gh.unstack(level=[u'spore_MATα'])\n",
    "\n",
    "    # plot\n",
    "    pheno.heatmap_spores(S, radius=0.5)\n",
    "    pheno.heatmap_hybrids(H, fold=False, title=environments[env_test], pad=100)\n",
    "\n",
    "# fig.subplots_adjust(left=0.2,bottom=0,right=0.8,top=1,hspace=0.01,wspace=0.01)\n",
    "fig.subplots_adjust(left=0.1,bottom=0.1,right=0.9,top=0.95,hspace=0.05,wspace=0.60)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
