{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Supplementary Information:\n",
    "\n",
    "> \"Background-dependent effects of selection on subclonal heterogeneity\"\n",
    "\n",
    "> Ignacio Vázquez-García, Francisco Salinas, Jing Li, Andrej Fischer, Benjamin Barré, Johan Hallin, Anders Bergström, Elisa Alonso-Pérez, Jonas Warringer, Ville Mustonen, Gianni Liti\n",
    "\n",
    "## Figure 4 (+ Supp. Figs.)\n",
    "\n",
    "This IPython notebook is provided for reproduction of Figures 4, S11, S12 and S13 of the paper. It can be viewed by copying its URL to nbviewer and it can be run by opening it in binder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ImportError",
     "evalue": "No module named statsmodels.formula.api",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-36be070edbf3>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;31m# load internal dependencies\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 18\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mcolors\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mconfig\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mplot\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mlmm\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mutils\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     19\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0;31m# absolute file path for root directory of git repo\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/projects/yeast-cancer-drugs/paper/science/submission/src/lmm.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mpandas\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mstatsmodels\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformula\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapi\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     14\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mstatsmodels\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstats\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0manova\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0manova_lm\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mImportError\u001b[0m: No module named statsmodels.formula.api"
     ]
    }
   ],
   "source": [
    "# load external dependencies\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "import matplotlib.lines as lines\n",
    "import matplotlib.patches as patches\n",
    "import matplotlib.ticker as ticker\n",
    "import matplotlib.transforms as transforms\n",
    "%matplotlib inline\n",
    "\n",
    "import json,matplotlib\n",
    "s = json.load( open(\"matplotlibrc.json\") )\n",
    "matplotlib.rcParams.update(s)\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd # (*) Pandas for data manipulation\n",
    "\n",
    "# load internal dependencies\n",
    "import colors,config,plot,lmm,utils\n",
    "\n",
    "# absolute file path for root directory of git repo\n",
    "dir_repo = utils.get_git_path()\n",
    "dir_data = dir_repo+\"/data/\"\n",
    "dir_paper = dir_repo+\"/manuscript/main/\"\n",
    "dir_supp = dir_repo+\"/manuscript/supp/\"\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import warnings\n",
    "\n",
    "with warnings.catch_warnings():\n",
    "    warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this experiment, we cross individuals of the ancestral population with pre-existing variation and mutant individuals from the evolved population. All of these are diploid cells which undergo sporulation (labelled *spores*) and mating (creating *hybrids*). We evaluate phenotypes of the spore and hybrid genotypes. Phenotypic variation can then be attributed to allelic complementation (that is, dominance effects) or to non-allelic complementation (that is, epistasis)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data import\n",
    "\n",
    "### Spores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# load data\n",
    "# spores_abs_df = utils.load_data(dir_data+'pheno/genetic-cross/pheno_spores_abs.pickle')\n",
    "spores_norm_df = utils.load_data(dir_data+'pheno/genetic-cross/pheno_spores_norm.pickle')\n",
    "\n",
    "# filter out blank positions and constructs\n",
    "spores_norm_df = spores_norm_df[~(spores_norm_df['strain'].isin(['control','']))&\\\n",
    "                                ~(spores_norm_df['genotype_long'].isin([u'fpr1Δ',u'tor1Δ']))]\n",
    "\n",
    "# set contaminated samples as missing (NaN)\n",
    "spores_norm_df.loc[(spores_norm_df['contamination']=='yes'),'growth_rate'] = np.nan\n",
    "\n",
    "# define individual and replicate\n",
    "spores_norm_df['individual'] = zip(spores_norm_df['group'],\n",
    "                                   spores_norm_df['background'],\n",
    "                                   spores_norm_df['gene'],\n",
    "                                   spores_norm_df['tetrad'],\n",
    "                                   spores_norm_df['spore'])\n",
    "spores_norm_df['replicate'] = zip(spores_norm_df['group'],\n",
    "                                  spores_norm_df['background'],\n",
    "                                  spores_norm_df['gene'],\n",
    "                                  spores_norm_df['tetrad'],\n",
    "                                  spores_norm_df['spore'],\n",
    "                                  spores_norm_df['plate'],\n",
    "                                  spores_norm_df['row'],\n",
    "                                  spores_norm_df['column'])\n",
    "\n",
    "# # filter out measurement replicates with >5% measurement error\n",
    "# spores_norm_df['pct'] = spores_norm_df.groupby(['selection','environment','replicate'])['growth_rate']\\\n",
    "# .apply(lambda x: (x-x.mean())/float(x.mean()))\n",
    "# spores_norm_df = spores_norm_df[abs(spores_norm_df['pct'])<0.05]\n",
    "\n",
    "spores_norm_df.head()  # show dataframe header to stdout"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hybrids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# load data\n",
    "# hybrids_abs_df = utils.load_data(dir_data+'pheno/genetic-cross/pheno_hybrids_abs.pickle')\n",
    "hybrids_norm_df = utils.load_data(dir_data+'pheno/genetic-cross/pheno_hybrids_norm.pickle')\n",
    "\n",
    "# filter out blank positions, contamination and constructs\n",
    "hybrids_norm_df = hybrids_norm_df[~(hybrids_norm_df['strain'].isin([('control','control'),('','')]))&\\\n",
    "                                  ~(hybrids_norm_df['genotype_long'].isin([u'fpr1Δ',u'tor1Δ']))]\n",
    "\n",
    "# set contaminated samples as missing (NaN)\n",
    "hybrids_norm_df.loc[hybrids_norm_df['contamination']\\\n",
    "                    .isin([('yes','yes'),('no','yes'),('yes','no')]),'growth_rate'] = np.nan\n",
    "\n",
    "# fix diploid genotypes for different genes\n",
    "hybrids_norm_df.loc[(hybrids_norm_df[u'gene_MATa']=='RNR4')&\\\n",
    "                    (hybrids_norm_df[u'gene_MATα']=='RNR2'),'genotype_short'] =\\\n",
    "hybrids_norm_df.loc[(hybrids_norm_df[u'gene_MATa']=='RNR4')&\\\n",
    "                    (hybrids_norm_df[u'gene_MATα']=='RNR2'),'genotype_short'].apply(lambda x: x[::-1])\n",
    "hybrids_norm_df.loc[(hybrids_norm_df[u'gene_MATa']=='TOR1')&\\\n",
    "                    (hybrids_norm_df[u'gene_MATα']=='FPR1'),'genotype_short'] =\\\n",
    "hybrids_norm_df.loc[(hybrids_norm_df[u'gene_MATa']=='TOR1')&\\\n",
    "                    (hybrids_norm_df[u'gene_MATα']=='FPR1'),'genotype_short'].apply(lambda x: x[::-1])\n",
    "\n",
    "# define individual\n",
    "hybrids_norm_df['individual'] = zip(hybrids_norm_df['group'],\n",
    "                                    hybrids_norm_df['background'],\n",
    "                                    hybrids_norm_df['gene'],\n",
    "                                    hybrids_norm_df['tetrad'],\n",
    "                                    hybrids_norm_df['spore'])\n",
    "hybrids_norm_df['replicate'] = zip(hybrids_norm_df['group'],\n",
    "                                   hybrids_norm_df['background'],\n",
    "                                   hybrids_norm_df['gene'],\n",
    "                                   hybrids_norm_df['tetrad'],\n",
    "                                   hybrids_norm_df['spore'],\n",
    "                                   hybrids_norm_df['plate'],\n",
    "                                   hybrids_norm_df['row'],\n",
    "                                   hybrids_norm_df['column'])\n",
    "\n",
    "# # filter out measurement replicates with >5% measurement error\n",
    "# hybrids_norm_df['pct'] = hybrids_norm_df.groupby(['selection','environment','replicate'])['growth_rate']\\\n",
    "# .apply(lambda x: (x-x.mean())/float(x.mean()))\n",
    "# hybrids_norm_df = hybrids_norm_df[abs(hybrids_norm_df['pct'])<0.05]\n",
    "\n",
    "hybrids_norm_df.head()  # show dataframe header to stdout"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# load data\n",
    "# clone_abs_df = utils.load_data(dir_data+'pheno/populations/pheno_populations_abs.pickle')\n",
    "clone_norm_df = utils.load_data(dir_data+'pheno/populations/pheno_populations_norm.pickle')\n",
    "\n",
    "# filter out strains used for spatial control\n",
    "clone_norm_df = clone_norm_df[(clone_norm_df.group == 'ancestral')|\\\n",
    "                              (clone_norm_df.group == 'evolved')]\n",
    "\n",
    "clone_norm_df['genotype_short'] = map(lambda x: tuple(x.strip(\"/\").split(\"/\")),clone_norm_df['genotype_short'])\n",
    "clone_norm_df['genotype_long'] = map(lambda x: tuple(x.strip(\"/\").split(\"/\")),clone_norm_df['genotype_long'])\n",
    "\n",
    "clone_norm_df['group'] = zip(clone_norm_df.group, clone_norm_df.group)\n",
    "clone_norm_df['background'] = zip(clone_norm_df.background, clone_norm_df.background)\n",
    "clone_norm_df['gene'] = zip(clone_norm_df.gene, clone_norm_df.gene)\n",
    "\n",
    "clone_norm_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Analysis of variance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "lmm_fit = pd.read_csv(dir_data+'pheno/genetic-cross/lmm_fit.csv')\n",
    "lmm_var_comp = pd.read_csv(dir_data+'pheno/genetic-cross/lmm_var_components.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We need to filter and sort the vector of spore phenotypes and matrix of hybrid phenotypes by environment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def filter_spores(S, env_evo):\n",
    "    # filter by dictionary\n",
    "    S = S[(S['group'].isin(config.spores_bg['position'][env_evo].keys())) &\n",
    "          (S['genotype_short'].isin(config.spores_gt_short['position'][env_evo].keys())) &\n",
    "          (S['background'].isin(config.spores_cl['position'][env_evo].keys()))]\n",
    "    return S\n",
    "          \n",
    "def filter_hybrids(H, env_evo):\n",
    "    # filter by dictionary\n",
    "    H = H[(H['group'].isin(config.hybrids_bg['position'][env_evo].keys())) &\n",
    "          (H['genotype_short'].isin(config.hybrids_gt_short['position'][env_evo].keys())) &\n",
    "          (H['background'].isin(config.hybrids_cl['position'][env_evo].keys()))]\n",
    "    return H\n",
    "    \n",
    "def sort_spores(S, env_evo):\n",
    "    # apply sorting ranks to reorder rows\n",
    "    S.loc[:,'rank_group'] = S['group'].map(config.spores_bg['position'][env_evo])\n",
    "    S.loc[:,'rank_background'] = S['background'].map(config.spores_cl['position'][env_evo])\n",
    "    S.loc[:,'rank_gene'] = S['gene'].map(config.spores_gn['position'][env_evo])\n",
    "    S.loc[:,'rank_genotype'] = S['genotype_short'].map(config.spores_gt_short['position'][env_evo])\n",
    "    S.sort_values(['rank_group','rank_background','rank_gene','rank_genotype'],\n",
    "                  ascending=True,inplace=True)\n",
    "    return S\n",
    "\n",
    "def sort_hybrids(H, env_evo):\n",
    "    # apply sorting ranks to reorder rows\n",
    "    H.loc[:,'rank_group'] = H['group'].map(config.hybrids_bg['position'][env_evo])\n",
    "    H.loc[:,'rank_background'] = H['background'].map(config.hybrids_cl['position'][env_evo])\n",
    "    H.loc[:,'rank_gene'] = H['gene'].map(config.hybrids_gn['position'][env_evo])\n",
    "    H.loc[:,'rank_genotype'] = H['genotype_short'].map(config.hybrids_gt_short['position'][env_evo])\n",
    "    H.sort_values(['rank_group','rank_background','rank_gene','rank_genotype'],\n",
    "                  ascending=True,inplace=True)\n",
    "    return H"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 4 - Background-averaged fitness effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from IPython.display import Image\n",
    "Image(filename=dir_paper+'figures/figure4/figure4A_schematic_submission.png', retina=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "param='growth_rate'\n",
    "\n",
    "fig = plt.figure(figsize=(12, 6), dpi=300)\n",
    "\n",
    "grid = gridspec.GridSpec(2, 3, width_ratios=[1.1,3,1.075], hspace=0.35, wspace=0.2)\n",
    "\n",
    "gs = {}\n",
    "gs['schematic'] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,0])\n",
    "gs['variance'] = gridspec.GridSpecFromSubplotSpec(3, 1, height_ratios=[5,5,2], subplot_spec=grid[1,0], hspace=0.1)\n",
    "gs[('HU','heatmap')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,2])\n",
    "gs[('HU','barplot')] = gridspec.GridSpecFromSubplotSpec(1, 24, subplot_spec=grid[0,1], wspace=0)\n",
    "gs[('RM','heatmap')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,2])\n",
    "gs[('RM','barplot')] = gridspec.GridSpecFromSubplotSpec(1, 24, subplot_spec=grid[1,1], wspace=0)\n",
    "\n",
    "### Schematic ###\n",
    "ax = plt.subplot(gs['schematic'][:])\n",
    "ax.text(-0.1, 1.125,  chr(ord('A')), transform=ax.transAxes,\n",
    "        fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "ax.axis('off')\n",
    "\n",
    "### Box plots ###\n",
    "groups_sp = spores_norm_df.groupby(['selection','environment'])\n",
    "groups_hy = hybrids_norm_df.groupby(['selection','environment'])\n",
    "groups_cl = clone_norm_df.groupby(['selection','environment'])\n",
    "\n",
    "for (ii, (env_evo, env_test)) in enumerate([('HU','HU'),('RM','RM')]):\n",
    "    \n",
    "    # group by evolution and test environment\n",
    "    S = groups_sp.get_group((env_evo, env_test))\n",
    "    H = groups_hy.get_group((env_evo, env_test))\n",
    "    C = groups_cl.get_group((env_evo, env_test))\n",
    "    \n",
    "    # filter by dictionary\n",
    "    S = filter_spores(S, env_evo)\n",
    "    H = filter_hybrids(H, env_evo)\n",
    "    \n",
    "    # apply sorting ranks to reorder rows\n",
    "    S = sort_spores(S, env_evo)\n",
    "    H = sort_hybrids(H, env_evo)\n",
    "    \n",
    "    ### spores barplot ###\n",
    "    ax1 = plt.subplot(gs[(env_evo,'barplot')][:7])\n",
    "\n",
    "    ax1.text(-0.2, 1.125,  chr(3*ii + ord('B')), transform=ax1.transAxes,\n",
    "             fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "    \n",
    "    ax1.text(0, 1.125, config.environment['long_label'][env_test], transform=ax1.transAxes,\n",
    "             fontsize=6, va='center', ha='left')\n",
    "    \n",
    "    S = S.groupby(['group','background','gene','genotype_short','tetrad','spore'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['group','background','gene','genotype_short'])\n",
    "    S_ref = S.ix[:,'ancestral'].median().values\n",
    "    S = S-S_ref\n",
    "    \n",
    "    bp = S.plot(ax=ax1, kind='box', by='group', return_type='dict',\n",
    "                labels=S.columns.get_level_values('group'))\n",
    "\n",
    "    colors = [config.spores_gt_short['color'][env_evo][x] for x in S.columns.get_level_values('genotype_short')]\n",
    "    plot.boxplot_custom(bp, ax1, colors=colors, hatches=[' ']*10)\n",
    "        \n",
    "    ax1.set_title('Spores (haploid)', fontsize=6, y=1)\n",
    "    ax1.set_xlabel('')\n",
    "    ax1.set_ylabel(r'Rel. growth rate, $\\lambda^{btd}_{a\\alpha}$', fontsize=6, labelpad=2)\n",
    "\n",
    "    ### hybrids barplot ###\n",
    "    ax2 = plt.subplot(gs[(env_evo,'barplot')][7:], sharey=ax1)\n",
    "    \n",
    "    H = H.groupby(['group','background','gene','genotype_short','tetrad','spore'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['group','background','gene','genotype_short'])\n",
    "    ix_use = pd.MultiIndex.from_tuples([[('ancestral', 'ancestral')]], names=['group'])\n",
    "    H_ref = H.loc[:,H.columns.get_level_values('group').isin(ix_use.get_level_values(0))].median().squeeze()\n",
    "    H = H-H_ref\n",
    "    \n",
    "    C = C.groupby(['group','background','gene','genotype_short','isolate'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\n",
    "    C_ref = C[('ancestral', 'ancestral')].median()\n",
    "\n",
    "    C = C-C_ref\n",
    "    C = C.unstack(level=['group','background','gene','genotype_short'])\n",
    "    C = C.loc[:,(C.columns.get_level_values('group').isin([('evolved','evolved')])\\\n",
    "                 &~(C.columns.get_level_values('gene').isin([('','')])))]\n",
    "    C = C.reindex(columns=H.columns)\n",
    "    C = C.dropna(how='all')\n",
    "\n",
    "    bp = H.plot(ax=ax2, kind='box', by='group', return_type='dict', \n",
    "                labels=H.columns.get_level_values('group'))\n",
    "\n",
    "    colors = [config.hybrids_gt_short['color'][env_evo][x] for x in H.columns.get_level_values('genotype_short')]\n",
    "    plot.boxplot_custom(bp, ax2, colors=colors, hatches=[' ']*30)\n",
    "\n",
    "    for i,d in enumerate(C):\n",
    "        y = C[d]\n",
    "        x = [i+1]*len(y)\n",
    "        ax2.plot(x, y, \n",
    "                 mfc=config.hybrids_gt_short['color'][env_evo][d[-1]], mec='k', \n",
    "                 ms=3, marker=\"D\", linestyle=\"None\", zorder=6)\n",
    "    \n",
    "    ax2.set_title('De novo', fontsize=6, fontstyle='italic')\n",
    "    ax2.set_xlabel('')\n",
    "    ax2.set_ylabel('', fontsize=6)\n",
    "    \n",
    "    wt_artist = patches.Rectangle((0,0), width=1, height=1, color=config.hybrids_gt_short['color'][env_evo][('+','+')])\n",
    "    het_artist = patches.Rectangle((0,0), width=1, height=1, color=config.hybrids_gt_short['color'][env_evo][('+','-')])\n",
    "    hom_artist = patches.Rectangle((0,0), width=1, height=1, color=config.hybrids_gt_short['color'][env_evo][('-','-')])\n",
    "    clone_artist = lines.Line2D((0,.75),(0,0), mec='k', mfc='w', marker='D', markersize=3, linestyle='')\n",
    "\n",
    "    leg = ax2.legend([wt_artist,het_artist,hom_artist,clone_artist], \n",
    "                     ['WT','het.','hom.','clone'], ncol=1, \n",
    "                     frameon=False, loc='upper right',\n",
    "                     borderaxespad=0, handlelength=0.75, \n",
    "                     prop={'size':5}, labelspacing=.32)\n",
    "        \n",
    "    # Grid\n",
    "    for ax, g in zip([ax1,ax2],[S,H]):\n",
    "        ### horizontal ###\n",
    "        ax.yaxis.grid(ls=\"-\", lw=.75, color=\"0.9\", zorder=0)\n",
    "        ax.axhline(y=0., c='k', ls=\"--\", lw=.75, zorder=1)\n",
    "        ax.set_axisbelow(True)\n",
    "        \n",
    "        ### vertical ###\n",
    "        ## background\n",
    "        xstart, xend, xlabels = plot.set_custom_labels(g.columns, 0)\n",
    "        # labels\n",
    "        for k, v in xlabels.iteritems():\n",
    "            ax.annotate('\\nx\\n'.join(k) if isinstance(k, tuple) else k, \n",
    "                        xy=(v+1, 0.97), xycoords=(\"data\", \"axes fraction\"),\n",
    "                        ha='center', va='top', annotation_clip=False, fontsize=5)\n",
    "        # grid\n",
    "        xgrid=[xst+1.5 for xst in list(set(xstart.values()))]\n",
    "        [ax.axvline(x, lw=1.0, ls='-', color='0.9', zorder=0) for x in xgrid]\n",
    "\n",
    "        ## gene\n",
    "        xend, xstart, xlabels = plot.set_custom_labels(g.columns, 2)\n",
    "        \n",
    "        transform = transforms.blended_transform_factory(ax.transData, ax.transAxes)\n",
    "        for k in xlabels:\n",
    "            if (k!='') & (k!=('','')):\n",
    "                if abs(xstart[k]-xend[k])>0:\n",
    "                    line = lines.Line2D([xstart[k]+.75,xend[k]+1.25], [-.05,-.05], \n",
    "                                        color='k', lw=1, transform=transform)\n",
    "                else:\n",
    "                    line = lines.Line2D([xstart[k]+.75,xend[k]+1.25], [-.05,-.05], \n",
    "                                        color='k', lw=1, transform=transform)\n",
    "                ax.add_line(line)\n",
    "                line.set_clip_on(False)\n",
    "        \n",
    "        ax.set_xticks([x+1 for x in xlabels.values()], minor=False)\n",
    "        di = {u'no driver': u'no\\ndriver', u'/': u'',\n",
    "              u'\\nRNR2': u'RNR2\\nRNR2', u'\\nRNR4': u'RNR4\\nRNR4',\n",
    "              u'\\nno driver': u'no\\ndriver', u'no driver\\nno driver': u'no\\ndriver',\n",
    "              u'\\nFPR1': u'FPR1\\nFPR1', u'\\nTOR1': u'TOR1\\nTOR1'}\n",
    "        xlabels = ['\\n'.join(k) if isinstance(k, tuple) else k for k in xlabels.keys()]\n",
    "        xlabels = [di[x] if x in di else x for x in xlabels]\n",
    "        ax.set_xticklabels(xlabels, minor=False, fontsize=5, style='italic', va='top')\n",
    "                \n",
    "        ax.tick_params(axis='x', which='minor', size=0, pad=-30)\n",
    "        ax.tick_params(axis='x', which='major', size=0, pad=10)\n",
    "        ax.tick_params(axis='y', which='major', labelsize=6)\n",
    "        \n",
    "        if env_evo=='HU':\n",
    "            ax.set_ylim(-0.25,0.75)\n",
    "            ax.yaxis.set_major_locator( ticker.MaxNLocator(nbins = 5) )\n",
    "            ax.yaxis.set_minor_locator( ticker.MaxNLocator(nbins = 5) )\n",
    "        elif env_evo=='RM':\n",
    "            ax.set_ylim(-0.5,2.0)\n",
    "            ax.yaxis.set_major_locator( ticker.MaxNLocator(nbins = 5) )\n",
    "            ax.yaxis.set_minor_locator( ticker.MaxNLocator(nbins = 5) )\n",
    "\n",
    "### Heatmaps ###\n",
    "for (ii, (env_evo, env_test)) in enumerate([('HU','HU'),('RM','RM')]):\n",
    "    \n",
    "    # group by evolution and test environment\n",
    "    S = groups_sp.get_group((env_evo, env_test))\n",
    "    H = groups_hy.get_group((env_evo, env_test))\n",
    "    \n",
    "    # apply sorting ranks to reorder rows\n",
    "    S = sort_spores(S, env_evo)\n",
    "    H = sort_hybrids(H, env_evo)\n",
    "\n",
    "    ax = plt.subplot(gs[(env_evo,'heatmap')][:])\n",
    "        \n",
    "    ax.text(-0.15, 1.175,  chr(3*ii + ord('C')), transform=ax.transAxes,\n",
    "             fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "    \n",
    "    # spores\n",
    "    S = S.groupby([u'mating',u'group',u'background',u'genotype_short',u'gene'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    S_ref = S.ix[:,'ancestral'].mean()\n",
    "    S = S-S_ref\n",
    "    \n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    vmax = max(S.min(), S.max())\n",
    "    vmin = -vmax\n",
    "\n",
    "    plot.heatmap_spores(S, ax, title, \n",
    "                        xlabel, ylabel, xticklabels, yticklabels, \n",
    "                        cmap=plt.cm.RdBu_r, vmin=vmin, vmax=vmax, radius=0.25)\n",
    "    \n",
    "    # hybrids    \n",
    "    H = H.groupby([u'group_MATa',u'background_MATa',u'gene_MATa',u'genotype_short_MATa',\n",
    "                   u'group_MATα',u'background_MATα',u'gene_MATα',u'genotype_short_MATα'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    H = H.unstack(level=[u'group_MATα',u'background_MATα',u'gene_MATα',u'genotype_short_MATα'])\n",
    "    H_ref = H.loc['ancestral','ancestral'].loc['WAxNA','WAxNA'].values.squeeze()\n",
    "    H = H-H_ref\n",
    "\n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    vmax = max(np.nanmin(H), np.nanmax(H), key=abs)#H.values.max()\n",
    "    vmin = -vmax\n",
    "    legend_title = r'$\\langle\\lambda^{td}\\rangle_{a\\alpha}^b$'\n",
    "    \n",
    "    plot.heatmap_hybrids(H, ax, title,\n",
    "                         xlabel, ylabel, xticklabels, yticklabels, \n",
    "                         fold=False, legend_title=legend_title, cmap=plt.cm.RdBu_r, \n",
    "                         vmin=vmin, vmax=vmax, pad=30)\n",
    "    \n",
    "    ax.xaxis.set_ticks_position('top')\n",
    "    ax.yaxis.set_ticks_position('left')\n",
    "    \n",
    "    ax.annotate(r'$\\mathbf{a}$', xy=(-0.1, 1.01), xycoords=('axes fraction', 'axes fraction'),\n",
    "                 ha='center', va='bottom', annotation_clip=False, weight='bold', fontsize=8)\n",
    "    ax.annotate(r'$\\mathbf{\\alpha}$', xy=(-0.01, 1.1), xycoords=('axes fraction', 'axes fraction'),\n",
    "                 ha='right', va='center', annotation_clip=False, weight='bold', fontsize=8)\n",
    "    \n",
    "    ## gene\n",
    "    # labels\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax.annotate(k, xy=(v+0.5, 1.12), xycoords=(\"data\", \"axes fraction\"),\n",
    "                    ha='center', va='top', annotation_clip=False, \n",
    "                    style=('italic' if k!='no driver' else 'normal'), fontsize=5)\n",
    "    ystart, yend, ylabels = plot.set_custom_labels(H.index, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax.annotate(k, xy=(-0.12, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                    ha='left', va='center', annotation_clip=False, \n",
    "                    style=('italic' if k!='no driver' else 'normal'), fontsize=5, rotation=90)\n",
    "    \n",
    "    ## background\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, 0)\n",
    "    ax.set_xticks([x+0.5 for x in xlabels.values()], minor=False)\n",
    "    ax.set_xticklabels(xlabels.keys(), minor=False, fontsize=6)\n",
    "    xgrid=[xst+1. for xst in list(set(xstart.values()))]\n",
    "    [ax.axvline(g, lw=1, ls=\"-\", color=\"black\") for g in xgrid]\n",
    "    \n",
    "    ystart, yend, ylabels= plot.set_custom_labels(H.columns, 0)\n",
    "    ax.set_yticks([y+0.5 for y in ylabels.values()], minor=False)\n",
    "    ax.set_yticklabels(ylabels.keys(), minor=False, fontsize=6, rotation=90)\n",
    "    ygrid=[yst+1. for yst in list(set(ystart.values()))]\n",
    "    [ax.axhline(g, lw=1, ls=\"-\", color=\"black\") for g in ygrid]\n",
    "\n",
    "    for x in ([-1]+xstart.values()):\n",
    "        line = lines.Line2D([x+1,x+1], [0.,-.5], color='k', lw=1)\n",
    "        ax.add_line(line)\n",
    "        line.set_clip_on(False)\n",
    "    for y in ([-1]+ystart.values()):\n",
    "        line = lines.Line2D([0.,-.5], [y+1,y+1], color='k', lw=1)\n",
    "        ax.add_line(line)\n",
    "        line.set_clip_on(False)\n",
    "    \n",
    "    ax.tick_params(axis='both', which='minor', size=0, pad=10)\n",
    "    ax.tick_params(axis='both', which='major', size=0, pad=18)\n",
    "\n",
    "### Variance components ###\n",
    "# sdf = spores_norm_df\n",
    "# sdf = sdf[sdf.group.isin(['ancestral','evolved'])]\n",
    "# sdf = sdf[np.isfinite(sdf[param])]\n",
    "\n",
    "# hdf = hybrids_norm_df\n",
    "# hdf = hdf[hdf.group.isin([('ancestral','ancestral'),('ancestral','evolved'),('evolved','evolved')])]\n",
    "# hdf = hdf[np.isfinite(hdf[param])]\n",
    "\n",
    "# formula = 'growth_rate ~ C(group) + C(genotype_long) + C(auxotrophy) + C(individual)'\n",
    "# tsdf = sdf.groupby(['selection','environment']).apply(lmm.variance_vectors, formula) \n",
    "# thdf = hdf.groupby(['selection','environment']).apply(lmm.variance_vectors, formula)\n",
    "# lmm_var_comp = pd.concat([tsdf['var_explained'], thdf['var_explained']], keys=['spores','hybrids'], names=['type'])\n",
    "\n",
    "# # fix columns\n",
    "# lmm_var_comp = lmm_var_comp.rename(\n",
    "#     columns={\n",
    "#         'C(group)':'Time',\n",
    "#         'C(genotype_long)':'De novo\\n(genotype)',\n",
    "#         'C(auxotrophy)':'Auxotrophy',\n",
    "#         'C(individual)':'Background\\n(genotype)',\n",
    "#     }\n",
    "# )\n",
    "\n",
    "# order = ['Background\\n(genotype)','De novo\\n(genotype)','Time','Auxotrophy']\n",
    "# order = [val for val in order if val in df.columns]\n",
    "# lmm_var_comp = lmm_var_comp[order]\n",
    "\n",
    "# # fix index\n",
    "# lmm_var_comp = lmm_var_comp.swaplevel('type', 'selection')\n",
    "# lmm_var_comp = lmm_var_comp.swaplevel('type', 'environment')\n",
    "# lmm_var_comp = lmm_var_comp.sort_index(level=['selection','environment','type'])\n",
    "# lmm_var_comp = lmm_var_comp.sort_index(ascending=False)\n",
    "\n",
    "var_comp = lmm_var_comp.set_index(['selection','environment','type'])\n",
    "var_comp = var_comp.sort_index(level=['selection','environment','type'])\n",
    "var_comp = var_comp.sort_index(ascending=[False,False,True])\n",
    "\n",
    "# fix columns\n",
    "var_comp = var_comp.rename(\n",
    "    columns={\n",
    "        'background (genotype)':'Background\\n(genotype)',\n",
    "        'de novo (genotype)':'De novo\\n(genotype)',\n",
    "        'time':'Time',\n",
    "        'auxotrophy':'Auxotrophy'\n",
    "    }\n",
    ")\n",
    "\n",
    "axes = {}\n",
    "\n",
    "for ii, (env_evo, ge) in enumerate(var_comp.groupby(level='selection')):\n",
    "        \n",
    "    if env_evo=='HU':\n",
    "        axes[env_evo] = plt.subplot(gs['variance'][ii])\n",
    "        axes[env_evo].text(-0.1, 1.35,  chr(ord('D')), transform=axes[env_evo].transAxes,\n",
    "                           fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "    else:\n",
    "        axes[env_evo] = plt.subplot(gs['variance'][ii], sharex=axes['HU'])\n",
    "        \n",
    "    ax = axes[env_evo]\n",
    "    utils.simple_axes(ax)\n",
    "    \n",
    "    ge.plot(ax=ax, kind='barh', stacked=True, \n",
    "            color=[config.factors['color'][x] for x in ge.columns],\n",
    "            align='center', legend=False, width=0.75)\n",
    "    \n",
    "    bars = ax.patches\n",
    "    hatches = [config.factors['hatch'][x] for x in ge.index.get_level_values('type')]*len(ge)\n",
    "\n",
    "    for bar, hatch in zip(bars, hatches):\n",
    "        bar.set_hatch(hatch)\n",
    "    \n",
    "    # x-axis\n",
    "    ax.set_xlabel('Variance (%)', fontsize=6, labelpad=2, visible=True)\n",
    "    xscale = 1e-2\n",
    "    xticks = ticker.FuncFormatter(lambda x, pos: '{0:g}'.format(x/xscale))\n",
    "    ax.xaxis.set_major_formatter(xticks)\n",
    "    # y-axis\n",
    "    ax.set_ylabel('')\n",
    "    ystart, yend, ylabels = plot.set_custom_labels(ge.index, 1)\n",
    "    ax.set_yticks(ylabels.values())\n",
    "    ax.set_yticklabels(ylabels.keys(), fontsize=6)\n",
    "    \n",
    "    # annotate panels and axis\n",
    "    ax.annotate(env_evo, xy=(1, 0.05), xycoords=('axes fraction', 'axes fraction'),\n",
    "                ha='right', va='bottom', fontsize=6)\n",
    "    \n",
    "    ax.tick_params(axis='both', which='major', labelsize=6, size=2)\n",
    "    \n",
    "# fix x-axis labels (missing due to boxplot styling)\n",
    "axes['RM'].xaxis.set_visible(True)\n",
    "plt.setp(axes['RM'].get_xticklabels(), visible=True)\n",
    "\n",
    "leg1 = ax.legend(bbox_to_anchor=(0.5, -0.65), ncol=4,\n",
    "                 frameon=False, loc='lower center',\n",
    "                 borderaxespad=0, handlelength=0.75, prop={'size':6})\n",
    "    \n",
    "s_artist = patches.Rectangle((0,0), width=1, height=1, \n",
    "                             facecolor='w', edgecolor='k')\n",
    "h_artist = patches.Rectangle((0,0), width=1, height=1, hatch='///', \n",
    "                             facecolor='w', edgecolor='k')\n",
    "\n",
    "leg2 = axes['HU'].legend([s_artist,h_artist], ['Spores','Hybrids'],\n",
    "                         bbox_to_anchor=(0.5, 1.2), ncol=2, \n",
    "                         frameon=False, loc='upper center',\n",
    "                         borderaxespad=0, handlelength=0.75, prop={'size':6})\n",
    "\n",
    "ax.add_artist(leg1)\n",
    "\n",
    "fig.savefig(dir_paper+'figures/figure4/figure4.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_paper+'figures/figure4/figure4.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fig. 4:** Fitness contribution of genetic background and *de novo* mutations. (*A*) To determine fitness effects of background variation and de novo mutations in hydroxyurea (*RNR2*, *RNR4*) and rapamycin (*FPR1*, *TOR1*), we isolated individuals from ancestral and evolved populations. From these diploid cells, we sporulated and selected haploid segregants of each mating type. Spores with mutations in *RNR2*, *RNR4* and *TOR1* were genotyped to test if they carry the wild-type or mutated allele. We crossed the *MAT*a and *MAT*α versions to create hybrids (48x48 in hydroxyurea and 56x56 in rapamycin). Independent segregants were used to measure biological variability of ancestral and evolved backgrounds. All panels follow this legend. (*B*, *E*) Relative growth rate, $\\lambda$, measured with respect to the ancestral population for multiple combinations (background genotype, $b$; *de novo* genotype, $d$; time of sampling during the selection phase, $t$; auxotrophy, $x$) and averaged over measurement replicates. Medians and 25%/75% percentiles across groups are shown, with medians as horizontal black lines and colored by de novo genotype [wild-type +/+ (red); heterozygote +/− (blue); homozygote −/− (green)]. Outliers (circles) and isolated, selected clones with matching genotypes (diamonds) are highlighted. (*D*) Variance decomposition of the growth rate of spores (solid) and hybrids (hatched) that can be attributed to different components using a linear mixed model. Estimates of variance components are obtained by restricted maximum likelihood (Fig. S13 and Table S7). (*C*, *F*) Ensemble average of the growth rate of spores  and hybrids . An extended version of the figure with all combinations and controls can be found in Figs. S11 and S12, respectively."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure S11 - Background-averaged fitness effects (full matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "param='growth_rate'\n",
    "\n",
    "fig = plt.figure(figsize=(8, 8), dpi=300)\n",
    "\n",
    "grid = gridspec.GridSpec(2, 2, hspace=0.35, wspace=0.25)\n",
    "\n",
    "gs = {}\n",
    "gs[('HU','HU')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,0])\n",
    "gs[('HU','YNB')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,1])\n",
    "gs[('RM','RM')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,0])\n",
    "gs[('RM','YNB')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,1])\n",
    "\n",
    "groups_sp = spores_norm_df.groupby(['selection','environment'])\n",
    "groups_hy = hybrids_norm_df.groupby(['selection','environment'])\n",
    "groups_cl = clone_norm_df.groupby(['selection','environment'])\n",
    "\n",
    "for (ii, (env_evo, env_test)) in enumerate([('HU','HU'),('HU','YNB'),('RM','RM'),('RM','YNB')]):\n",
    "    \n",
    "    # group by evolution and test environment\n",
    "    S = groups_sp.get_group((env_evo, env_test))\n",
    "    H = groups_hy.get_group((env_evo, env_test))\n",
    "    C = groups_cl.get_group((env_evo, env_test))\n",
    "    \n",
    "    # apply sorting ranks to reorder rows\n",
    "    S = sort_spores(S, env_evo)\n",
    "    H = sort_hybrids(H, env_evo)\n",
    "\n",
    "    ### heatmap ###\n",
    "    ax = plt.subplot(gs[(env_evo, env_test)][0])\n",
    "    \n",
    "    ax.text(-0.115, 1.2,  chr(ii + ord('A')), transform=ax.transAxes,\n",
    "            fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "    \n",
    "    ax.text(0.5, 1.2,  \n",
    "            'Selection: %s\\nMeasurement: %s' %\\\n",
    "            (config.selection['long_label'][env_evo], config.environment['long_label'][env_test]), \n",
    "            transform=ax.transAxes,\n",
    "            fontsize=6, va='center', ha='center')\n",
    "    \n",
    "    # spores\n",
    "    S = S.groupby([u'mating',u'group',u'background',u'gene',u'genotype_short',u'tetrad',u'spore'],\n",
    "                    sort=False)[param].agg(np.median).reset_index()\n",
    "    S = S.set_index('mating')[param]\n",
    "    S_ref = np.nanmean(S.values)\n",
    "    S = S-S_ref\n",
    "        \n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    vmax = max(S.min(), S.max())\n",
    "    vmin = -vmax\n",
    "    \n",
    "    plot.heatmap_spores(S, ax, title, \n",
    "                        xlabel, ylabel, xticklabels, yticklabels, \n",
    "                        cmap=plt.cm.RdBu_r, vmin=vmin, vmax=vmax, radius=0.5)\n",
    "    \n",
    "    # hybrids    \n",
    "    H = H.groupby([u'group_MATa',u'background_MATa',u'gene_MATa',u'genotype_short_MATa',u'tetrad_MATa',u'spore_MATa',\n",
    "                   u'group_MATα',u'background_MATα',u'gene_MATα',u'genotype_short_MATα',u'tetrad_MATα',u'spore_MATα'],\n",
    "                    sort=False).agg(np.median)[param]\n",
    "    H = H.unstack(level=[u'group_MATα',u'background_MATα',u'gene_MATα',u'genotype_short_MATα',u'tetrad_MATα',u'spore_MATα'])\n",
    "    H_ref = np.nanmean(H.loc['ancestral','ancestral'].loc['WAxNA','WAxNA'])\n",
    "    H = H-H_ref\n",
    "\n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    vmax = max(np.nanmin(H), np.nanmax(H), key=abs)\n",
    "    vmin = -vmax\n",
    "    legend_title = r'$\\lambda^{btd}_{a\\alpha}$'\n",
    "    \n",
    "    plot.heatmap_hybrids(H, ax, title,\n",
    "                         xlabel, ylabel, xticklabels, yticklabels, \n",
    "                         fold=False, legend_title=legend_title, cmap=plt.cm.RdBu_r, \n",
    "                         vmin=vmin, vmax=vmax, pad=30)\n",
    "    \n",
    "    print 'Spores\\nSelection: %s, Environment: %s, Pass: %4d, Fail: %4d' \\\n",
    "    % (env_evo, env_test, S.notnull().values.flatten().sum(), S.isnull().values.flatten().sum())\n",
    "    print 'Hybrids\\nSelection: %s, Environment: %s, Pass: %4d, Fail: %4d\\n' \\\n",
    "    % (env_evo, env_test, H.notnull().values.flatten().sum(), H.isnull().values.flatten().sum())\n",
    "    \n",
    "    ax.xaxis.set_ticks_position('top')\n",
    "    ax.yaxis.set_ticks_position('left')\n",
    "    \n",
    "    ax.annotate(r'$\\mathbf{a}$', xy=(-0.075, 1.01), xycoords=('axes fraction', 'axes fraction'),\n",
    "                 ha='center', va='bottom', annotation_clip=False, weight='bold', fontsize=8)\n",
    "    ax.annotate(r'$\\mathbf{\\alpha}$', xy=(-0.01, 1.075), xycoords=('axes fraction', 'axes fraction'),\n",
    "                 ha='right', va='center', annotation_clip=False, weight='bold', fontsize=8)\n",
    "    \n",
    "    ## gene + genotype\n",
    "    # labels and grid\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, slice(2, 4))\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax.annotate(k[0].replace('no driver','no\\ndri.'), \n",
    "                    xy=(v+0.5, 1.05), xycoords=(\"data\", \"axes fraction\"),\n",
    "                    ha='center', va='bottom', annotation_clip=False, \n",
    "                    style=('italic' if k[0]!='no driver' else 'normal'), fontsize=6)\n",
    "    xgrid=[xst+1. for xst in list(set(xstart.values()))]\n",
    "    [ax.axvline(g, lw=.75, ls=\"--\", color=\"gray\") for g in xgrid]\n",
    "    for x in ([-1]+xstart.values()):\n",
    "        line = lines.Line2D([x+1,x+1], [0.,-3], color='gray', ls='-', lw=.75)\n",
    "        ax.add_line(line)\n",
    "        line.set_clip_on(False)\n",
    "    \n",
    "    ystart, yend, ylabels = plot.set_custom_labels(H.index, slice(2, 4))\n",
    "    for k, v in ylabels.iteritems():\n",
    "        ax.annotate(k[0].replace('no driver','no\\ndri.'), \n",
    "                    xy=(-0.05, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                    ha='right', va='center', annotation_clip=False,  \n",
    "                    style=('italic' if k[0]!='no driver' else 'normal'), fontsize=6, rotation=90)\n",
    "    ygrid=[yst+1. for yst in list(set(ystart.values()))]\n",
    "    [ax.axhline(g, lw=.75, ls=\"--\", color=\"gray\") for g in ygrid]\n",
    "    for y in ([-1]+ystart.values()):\n",
    "        line = lines.Line2D([0.,-3], [y+1,y+1], color='gray', ls='-', lw=.75)\n",
    "        ax.add_line(line)\n",
    "        line.set_clip_on(False)\n",
    "    \n",
    "    ## background\n",
    "    # labels and grid\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, 0)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax.annotate(config.population['short_label'][k], \n",
    "                    xy=(v+0.5, config.population['pad'][k]+1), xycoords=(\"data\", \"axes fraction\"),\n",
    "                    ha='center', va='bottom', annotation_clip=False, fontsize=6)\n",
    "    xgrid=[xst+1. for xst in list(set(xstart.values()))]\n",
    "    [ax.axvline(g, lw=.75, ls=\"-\", color=\"k\") for g in xgrid]\n",
    "    for x in ([-1]+xstart.values()):\n",
    "        line = lines.Line2D([x+1,x+1], [0.,-4], color='k', lw=1)\n",
    "        ax.add_line(line)\n",
    "        line.set_clip_on(False)\n",
    "    \n",
    "    ystart, yend, ylabels = plot.set_custom_labels(H.index, 0)\n",
    "    for k, v in ylabels.iteritems():\n",
    "        ax.annotate(config.population['short_label'][k], \n",
    "                    xy=(-config.population['pad'][k], v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                    ha='right', va='center', annotation_clip=False, fontsize=6, rotation=90)\n",
    "    ygrid=[yst+1. for yst in list(set(ystart.values()))]\n",
    "    [ax.axhline(g, lw=.75, ls=\"-\", color=\"k\") for g in ygrid]\n",
    "    for y in ([-1]+ystart.values()):\n",
    "        line = lines.Line2D([0.,-4], [y+1,y+1], color='k', lw=1)\n",
    "        ax.add_line(line)\n",
    "        line.set_clip_on(False)\n",
    "    \n",
    "    ax.tick_params(axis='both', which='minor', size=0, pad=2)\n",
    "    ax.tick_params(axis='both', which='major', size=0, pad=15)\n",
    "\n",
    "fig.savefig(dir_supp+'figures/supp_figure_pheno_cross/supp_figure_pheno_cross_extended.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_supp+'figures/supp_figure_pheno_cross/supp_figure_pheno_cross_extended.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fig. S11**: Fitness contribution of genetic background and *de novo* mutations. Given an ensemble of $n_b$ unique genetic backgrounds (${n_b\\,{=}\\,48}$ in hydroxyurea and ${n_b\\,{=}\\,56}$ in rapamycin), we constructed a matrix of size $n_b \\times n_b$ where every unique haploid background is crossed against itself and all other haploid backgrounds, and the two must be of opposite mating type (*MAT*a or *MAT*α). Each matrix element is labeled by background genotype, $b$; *de novo* genotype, $d$; time, $t$; and auxotrophy, $x$. Measurements of relative growth rates of spores $\\lambda^{btd}_{\\{a,\\alpha\\}}$ and hybrids $\\lambda^{btd}_{a\\alpha}$ are shown, normalized with respect to the ancestral WAxNA crosses. Measurements are taken in (*A*) YNB+HU 10 $\\text{mg ml}^{-1}$ and (*B*) YNB; (*C*) YNB+RM 0.025 $\\mu\\text{g ml}^{-1}$ and (*D*) YNB, respectively. The color scale for all matrices to the right of each panel indicates the relative fold-change with respect to the ancestral WAxNA crosses. White boxes indicate missing data due to mating inefficiency and slow growth."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure S12 - Background-averaged fitness effects (control)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "param='growth_rate'\n",
    "\n",
    "fig = plt.figure(figsize=(9.5, 6), dpi=300)\n",
    "\n",
    "grid = gridspec.GridSpec(2, 2, width_ratios=[3,1], hspace=0.4, wspace=0.2)\n",
    "\n",
    "gs = {}\n",
    "gs[('HU','barplot')] = gridspec.GridSpecFromSubplotSpec(1, 24, subplot_spec=grid[0,0], wspace=0)\n",
    "gs[('HU','heatmap')] = gridspec.GridSpecFromSubplotSpec(1, 1,subplot_spec=grid[0,1])\n",
    "gs[('RM','barplot')] = gridspec.GridSpecFromSubplotSpec(1, 24, subplot_spec=grid[1,0], wspace=0)\n",
    "gs[('RM','heatmap')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,1])\n",
    "\n",
    "### Box plots ###\n",
    "groups_sp = spores_norm_df.groupby(['selection','environment'])\n",
    "groups_hy = hybrids_norm_df.groupby(['selection','environment'])\n",
    "groups_cl = clone_norm_df.groupby(['selection','environment'])\n",
    "\n",
    "for (ii, (env_evo, env_test)) in enumerate([('HU','YNB'),('RM','YNB')]):\n",
    "    \n",
    "    # group by evolution and test environment\n",
    "    S = groups_sp.get_group((env_evo, env_test))\n",
    "    H = groups_hy.get_group((env_evo, env_test))\n",
    "    C = groups_cl.get_group((env_evo, env_test))\n",
    "    \n",
    "    # filter by dictionary\n",
    "    S = filter_spores(S, env_evo)\n",
    "    H = filter_hybrids(H, env_evo)\n",
    "    \n",
    "    # apply sorting ranks to reorder rows\n",
    "    S = sort_spores(S, env_evo)\n",
    "    H = sort_hybrids(H, env_evo)\n",
    "    \n",
    "    ### spores barplot ###\n",
    "    ax1 = plt.subplot(gs[(env_evo,'barplot')][:7])\n",
    "\n",
    "    ax1.text(-0.2, 1.15,  chr(2*ii + ord('A')), transform=ax1.transAxes,\n",
    "             fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "    \n",
    "    ax1.text(0, 1.15,  \n",
    "             'Selection: %s\\nMeasurement: %s' %\\\n",
    "             (config.selection['long_label'][env_evo], config.environment['long_label'][env_test]), \n",
    "             transform=ax1.transAxes,\n",
    "             fontsize=6, va='center', ha='left')\n",
    "    \n",
    "    S = S.groupby(['group','background','gene','genotype_short','tetrad','spore'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['group','background','gene','genotype_short'])\n",
    "    S_ref = S.ix[:,'ancestral'].median().values\n",
    "    S = S-S_ref\n",
    "    \n",
    "    bp = S.plot(ax=ax1, kind='box', by='group', return_type='dict',\n",
    "                labels=S.columns.get_level_values('group'))\n",
    "\n",
    "    plot.boxplot_custom(bp, ax1,\n",
    "                        colors=[config.spores_gt_short['color'][env_evo][x] for x in S.columns.get_level_values('genotype_short')],\n",
    "                        hatches=[' ']*10)\n",
    "    \n",
    "    ax1.set_title('Spores (haploid)', fontsize=6)\n",
    "    ax1.set_xlabel('')\n",
    "    ax1.set_ylabel(r'Rel. growth rate, $\\lambda^{btd}_{a\\alpha}$', fontsize=6, labelpad=2)\n",
    "\n",
    "    ### hybrids barplot ###\n",
    "    ax2 = plt.subplot(gs[(env_evo,'barplot')][7:], sharey=ax1)\n",
    "    \n",
    "    H = H.groupby(['group','background','gene','genotype_short','tetrad','spore'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['group','background','gene','genotype_short'])\n",
    "    ix_use = pd.MultiIndex.from_tuples([[('ancestral', 'ancestral')]], names=['group'])\n",
    "    H_ref = H.loc[:,H.columns.get_level_values('group').isin(ix_use.get_level_values(0))].median().squeeze()\n",
    "    H = H-H_ref\n",
    "    \n",
    "    C = C.groupby(['group','background','gene','genotype_short','isolate'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\n",
    "    C_ref = C[('ancestral', 'ancestral')].median()\n",
    "\n",
    "    C = C-C_ref\n",
    "    C = C.unstack(level=['group','background','gene','genotype_short'])\n",
    "    C = C.loc[:,(C.columns.get_level_values('group').isin([('evolved','evolved')])\\\n",
    "                 &~(C.columns.get_level_values('gene').isin([('','')])))]\n",
    "    C = C.reindex(columns=H.columns)\n",
    "    C = C.dropna(how='all')\n",
    "\n",
    "    bp = H.plot(ax=ax2, kind='box', by='group', return_type='dict', \n",
    "                labels=H.columns.get_level_values('group'))\n",
    "\n",
    "    plot.boxplot_custom(bp, ax2, \n",
    "                        colors=[config.hybrids_gt_short['color'][env_evo][x] for x in H.columns.get_level_values('genotype_short')],\n",
    "                        hatches=[' ']*30)\n",
    "\n",
    "    for i,d in enumerate(C):\n",
    "        y = C[d]\n",
    "        x = [i+1]*len(y)\n",
    "        ax2.plot(x, y, \n",
    "                 mfc=config.hybrids_gt_short['color'][env_evo][d[-1]], mec='k', \n",
    "                 ms=3, marker='D', linestyle=\"None\", zorder=6)\n",
    "    \n",
    "    ax2.set_title('Hybrids (diploid)', fontsize=6)\n",
    "    ax2.set_xlabel('')\n",
    "    ax2.set_ylabel('', fontsize=6)\n",
    "    \n",
    "    wt_artist = patches.Rectangle((0,0), width=1, height=1, color=config.hybrids_gt_short['color'][env_evo][('+','+')])\n",
    "    het_artist = patches.Rectangle((0,0), width=1, height=1, color=config.hybrids_gt_short['color'][env_evo][('+','-')])\n",
    "    hom_artist = patches.Rectangle((0,0), width=1, height=1, color=config.hybrids_gt_short['color'][env_evo][('-','-')])\n",
    "    clone_artist = lines.Line2D((0,1),(0,0), mec='k', mfc='w', marker='D', markersize=3, linestyle='')\n",
    "\n",
    "    leg = ax2.legend([wt_artist,het_artist,hom_artist,clone_artist], \n",
    "                     ['WT','het.','hom.','clone'], ncol=1, \n",
    "                     frameon=False, loc='upper right',\n",
    "                     borderaxespad=0, handlelength=0.75, \n",
    "                     prop={'size':5}, labelspacing=.32)\n",
    "        \n",
    "    # Grid\n",
    "    for ax, g in zip([ax1,ax2],[S,H]):\n",
    "        ### horizontal ###\n",
    "        ax.yaxis.grid(ls=\"-\", lw=.75, color=\"0.9\", zorder=-1)\n",
    "        ax.axhline(y=0., c='k', ls=\"--\", lw=.75, zorder=1)\n",
    "        ax.set_axisbelow(True)\n",
    "        \n",
    "        ### vertical ###\n",
    "        ## background\n",
    "        xstart, xend, xlabels = plot.set_custom_labels(g.columns, 0)\n",
    "        # labels\n",
    "        for k, v in xlabels.iteritems():\n",
    "            ax.annotate('\\nx\\n'.join(k) if isinstance(k, tuple) else k, \n",
    "                        xy=(v+1, 0.97), xycoords=(\"data\", \"axes fraction\"),\n",
    "                        ha='center', va='top', annotation_clip=False, fontsize=5)\n",
    "        # grid\n",
    "        xgrid=[xst+1.5 for xst in list(set(xstart.values()))]\n",
    "        [ax.axvline(x, lw=1.0, ls='-', color=\"0.9\", zorder=0) for x in xgrid]\n",
    "\n",
    "        ## gene\n",
    "        xend, xstart, xlabels = plot.set_custom_labels(g.columns, 2)\n",
    "        \n",
    "        transform = transforms.blended_transform_factory(ax.transData, ax.transAxes)\n",
    "        for k in xlabels:\n",
    "            if (k!='') & (k!=('','')):\n",
    "                if abs(xstart[k]-xend[k])>0:\n",
    "                    line = lines.Line2D([xstart[k]+.75,xend[k]+1.25], [-.05,-.05], \n",
    "                                        color='k', lw=1, transform=transform)\n",
    "                else:\n",
    "                    line = lines.Line2D([xstart[k]+.75,xend[k]+1.25], [-.05,-.05], \n",
    "                                        color='k', lw=1, transform=transform)\n",
    "                ax.add_line(line)\n",
    "                line.set_clip_on(False)\n",
    "        \n",
    "        ax.set_xticks([x+1 for x in xlabels.values()], minor=False)\n",
    "        di = {u'no driver': u'no\\ndriver', u'/': u'',\n",
    "              u'\\nRNR2': u'RNR2\\nRNR2', u'\\nRNR4': u'RNR4\\nRNR4',\n",
    "              u'\\nno driver': u'no\\ndriver', u'no driver\\nno driver': u'no\\ndriver',\n",
    "              u'\\nFPR1': u'FPR1\\nFPR1', u'\\nTOR1': u'TOR1\\nTOR1'}\n",
    "        xlabels = ['\\n'.join(k) if isinstance(k, tuple) else k for k in xlabels.keys()]\n",
    "        xlabels = [di[x] if x in di else x for x in xlabels]\n",
    "        ax.set_xticklabels(xlabels, minor=False, fontsize=5, style='italic', va='top')\n",
    "                \n",
    "        ax.tick_params(axis='x', which='minor', size=0, pad=-30)\n",
    "        ax.tick_params(axis='x', which='major', size=0, pad=10)\n",
    "        ax.tick_params(axis='y', which='major', labelsize=6)\n",
    "        \n",
    "        if env_evo=='HU':\n",
    "            ax.set_ylim(-0.25,0.5)\n",
    "            ax.yaxis.set_major_locator( ticker.MaxNLocator(nbins = 4) )\n",
    "            ax.yaxis.set_minor_locator( ticker.MaxNLocator(nbins = 4) )\n",
    "        elif env_evo=='RM':\n",
    "            ax.set_ylim(-0.5,0.5)\n",
    "            ax.yaxis.set_major_locator( ticker.MaxNLocator(nbins = 4) )\n",
    "            ax.yaxis.set_minor_locator( ticker.MaxNLocator(nbins = 4) )\n",
    "\n",
    "### Heatmaps ###\n",
    "for (ii, (env_evo, env_test)) in enumerate([('HU','YNB'),('RM','YNB')]):\n",
    "    \n",
    "    # group by evolution and test environment\n",
    "    S = groups_sp.get_group((env_evo, env_test))\n",
    "    H = groups_hy.get_group((env_evo, env_test))\n",
    "    \n",
    "    # apply sorting ranks to reorder rows\n",
    "    S = sort_spores(S, env_evo)\n",
    "    H = sort_hybrids(H, env_evo)\n",
    "\n",
    "    ax = plt.subplot(gs[(env_evo,'heatmap')][:])\n",
    "        \n",
    "    ax.text(-0.15, 1.21,  chr(2*ii + ord('B')), transform=ax.transAxes,\n",
    "             fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "    \n",
    "    # spores\n",
    "    S = S.groupby([u'mating',u'group',u'background',u'genotype_short',u'gene'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    S_ref = S.ix[:,'ancestral'].mean()\n",
    "    S = S-S_ref\n",
    "    \n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    vmax = max(S.min(), S.max())\n",
    "    vmin = -vmax\n",
    "\n",
    "    plot.heatmap_spores(S, ax, title, \n",
    "                        xlabel, ylabel, xticklabels, yticklabels, \n",
    "                        cmap=plt.cm.RdBu_r, vmin=vmin, vmax=vmax, radius=0.25)\n",
    "    \n",
    "    # hybrids    \n",
    "    H = H.groupby([u'group_MATa',u'background_MATa',u'gene_MATa',u'genotype_short_MATa',\n",
    "                   u'group_MATα',u'background_MATα',u'gene_MATα',u'genotype_short_MATα'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    H = H.unstack(level=[u'group_MATα',u'background_MATα',u'gene_MATα',u'genotype_short_MATα'])\n",
    "    H_ref = H.loc['ancestral','ancestral'].loc['WAxNA','WAxNA'].values.squeeze()\n",
    "    H = H-H_ref\n",
    "\n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    vmax = max(np.nanmin(H), np.nanmax(H), key=abs)\n",
    "    vmin = -vmax\n",
    "    legend_title = r'$\\langle \\lambda^{td}\\rangle_{a\\alpha}^b$'\n",
    "    \n",
    "    plot.heatmap_hybrids(H, ax, title,\n",
    "                         xlabel, ylabel, xticklabels, yticklabels, \n",
    "                         fold=False, legend_title=legend_title, cmap=plt.cm.RdBu_r, \n",
    "                         vmin=vmin, vmax=vmax, pad=30)\n",
    "    \n",
    "    ax.xaxis.set_ticks_position('top')\n",
    "    ax.yaxis.set_ticks_position('left')\n",
    "    \n",
    "    ax.annotate(r'$\\mathbf{a}$', xy=(-0.1, 1.01), xycoords=('axes fraction', 'axes fraction'),\n",
    "                 ha='center', va='bottom', annotation_clip=False, weight='bold', fontsize=8)\n",
    "    ax.annotate(r'$\\mathbf{\\alpha}$', xy=(-0.01, 1.1), xycoords=('axes fraction', 'axes fraction'),\n",
    "                 ha='right', va='center', annotation_clip=False, weight='bold', fontsize=8)\n",
    "    \n",
    "    ## gene\n",
    "    # labels\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax.annotate(k, xy=(v+0.5, 1.12), xycoords=(\"data\", \"axes fraction\"),\n",
    "                    ha='center', va='top', annotation_clip=False, \n",
    "                    style=('italic' if k!='no driver' else 'normal'), fontsize=5)\n",
    "    ystart, yend, ylabels = plot.set_custom_labels(H.index, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax.annotate(k, xy=(-0.12, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                    ha='left', va='center', annotation_clip=False, \n",
    "                    style=('italic' if k!='no driver' else 'normal'), fontsize=5, rotation=90)\n",
    "    \n",
    "    ## background\n",
    "    # labels and grid\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, 0)\n",
    "    ax.set_xticks([x+0.5 for x in xlabels.values()], minor=False)\n",
    "    ax.set_xticklabels(xlabels.keys(), minor=False, fontsize=6)\n",
    "    xgrid=[xst+1. for xst in list(set(xstart.values()))]\n",
    "    [ax.axvline(g, lw=1, ls=\"-\", color=\"black\") for g in xgrid]\n",
    "    \n",
    "    ystart, yend, ylabels= plot.set_custom_labels(H.columns, 0)\n",
    "    ax.set_yticks([y+0.5 for y in ylabels.values()], minor=False)\n",
    "    ax.set_yticklabels(ylabels.keys(), minor=False, fontsize=6, rotation=90)\n",
    "    ygrid=[yst+1. for yst in list(set(ystart.values()))]\n",
    "    [ax.axhline(g, lw=1, ls=\"-\", color=\"black\") for g in ygrid]\n",
    "\n",
    "    for x in ([-1]+xstart.values()):\n",
    "        line = lines.Line2D([x+1,x+1], [0.,-.5], color='k', lw=1)\n",
    "        ax.add_line(line)\n",
    "        line.set_clip_on(False)\n",
    "    for y in ([-1]+ystart.values()):\n",
    "        line = lines.Line2D([0.,-.5], [y+1,y+1], color='k', lw=1)\n",
    "        ax.add_line(line)\n",
    "        line.set_clip_on(False)\n",
    "    \n",
    "    ax.tick_params(axis='both', which='minor', size=0, pad=10)\n",
    "    ax.tick_params(axis='both', which='major', size=0, pad=18)\n",
    "\n",
    "fig.savefig(dir_supp+'figures/supp_figure_pheno_cross/supp_figure_pheno_cross_reduced.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_supp+'figures/supp_figure_pheno_cross/supp_figure_pheno_cross_reduced.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fig. S12:** Ensemble average of fitness effects over genetic backgrounds. (*A*, *C*) Relative growth rate, $\\lambda$, measured in a control environment (YNB) for cells selected in (*A*) hydroxyurea and (*C*) rapamycin. Measurements are normalized with respect to the ancestral population for multiple combinations (background genotype, $b$; *de novo* genotype, $d$; sampling time during selection, $t$; auxotrophy, $x$) and averaged over measurement replicates. Medians and 25\\%/75\\% percentiles across groups are shown, with medians as horizontal lines and colored by *de novo* genotype [wild-type +/+ (red); heterozygote +/- (blue); homozygote -/- (green)]. Outliers (circles) and isolated, selected clones with matching genotypes (diamonds) are highlighted. (*B*, *D*) Ensemble average of the growth rate of spores, $\\langle \\lambda^{td} \\rangle^b_{\\{a,\\alpha\\}}$, and hybrids, $\\langle \\lambda^{td} \\rangle^b_{a\\alpha}$ over genetic backgrounds. The color scale for all matrices to the right of each panel indicates the relative fold-change with respect to the ancestral WAxNA crosses."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure S13 - Analysis of variance"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Linear mixed model (frequentist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import statsmodels.api as sm\n",
    "\n",
    "panels = {\n",
    "    'HU': {\n",
    "        'HU':(0,0),\n",
    "        'YNB':(1,0)\n",
    "    },\n",
    "    'RM': {\n",
    "        'RM':(0,1),\n",
    "        'YNB':(1,1)\n",
    "    }\n",
    "}\n",
    "\n",
    "fig = plt.figure(figsize=(8, 4), dpi=300)\n",
    "\n",
    "grid = gridspec.GridSpec(2, 2, hspace=0.5, wspace=0.1)\n",
    "\n",
    "gs = {}\n",
    "\n",
    "var_comp = lmm_var_comp.copy()\n",
    "var_comp['total_variance'] = var_comp.sum(axis=1)\n",
    "    \n",
    "for ii, (s, e) in enumerate([('HU','HU'),('HU','YNB'),('RM','RM'),('RM','YNB')]):\n",
    "    \n",
    "    gs[(s, e)] = gridspec.GridSpecFromSubplotSpec(1, 2, subplot_spec=grid[panels[s][e]], wspace=0.02)\n",
    "\n",
    "for ii, ((s, e), fit_all) in enumerate(lmm_fit.groupby(['selection','environment'], sort=False)):\n",
    "\n",
    "    for jj, (t, fit_type) in enumerate(fit_all.groupby('type', sort=False)):\n",
    "        \n",
    "        ax = plt.subplot(gs[(s,e)][jj])\n",
    "        utils.simple_axes(ax)\n",
    "        \n",
    "        if jj==0:\n",
    "            ax.text(-0.2, 1.125, chr(ii + ord('A')), transform=ax.transAxes,\n",
    "                    fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "            ax.annotate('Selection: %s\\nMeasurement: %s' %\\\n",
    "                         (config.selection['long_label'][s], \\\n",
    "                          config.environment['long_label'][e]), \n",
    "                        xy=(0, 1.125), xycoords=\"axes fraction\",\n",
    "                        ha='left', va='center', annotation_clip=False, fontsize=6)\n",
    "\n",
    "        nobs = len(fit_type.observed_growth_rate)\n",
    "        y = fit_type.observed_growth_rate\n",
    "        yhat = fit_type.fitted_growth_rate\n",
    "        ymin, ymax = y.quantile([.005, .995])\n",
    "        yrange = ymax - ymin\n",
    "        \n",
    "        # scatter\n",
    "        ax.scatter(y, yhat, facecolor='gray', s=1.5, lw=0, alpha=0.75)\n",
    "        \n",
    "        # fit\n",
    "        line_fit = sm.OLS(y, sm.add_constant(yhat, prepend=False)).fit().params\n",
    "        fit = lambda x: line_fit[1] + line_fit[0] * x \n",
    "        ax.plot(np.linspace(ymin-.1*yrange, ymax+.1*yrange, nobs),\n",
    "                fit(np.linspace(ymin-.1*yrange, ymax+.1*yrange, nobs)), \n",
    "                label=t, color='k', ls='--', lw=1)\n",
    "        \n",
    "        # variance explained\n",
    "        var_total = var_comp.groupby(['selection','environment','type']).get_group((s,e,t))['total_variance'].values[0]\n",
    "        ax.annotate('{}\\n$r^2={:.2f}$'.format(t.capitalize(), var_total),\n",
    "                    xy=(.1, .9), xycoords=\"axes fraction\",\n",
    "                    ha='left', va='center', fontsize=6)\n",
    "        \n",
    "        # x-axis\n",
    "        ax.set_xlim(ymin-.1*yrange, ymax+.1*yrange)\n",
    "        ax.xaxis.set_major_locator( ticker.MaxNLocator(nbins = 4) )\n",
    "\n",
    "        # y-axis\n",
    "        ax.set_ylim(ymin-.1*yrange, ymax+.1*yrange)\n",
    "        ax.yaxis.set_major_locator( ticker.MaxNLocator(nbins = 4) )\n",
    "        \n",
    "        ax.set_aspect('equal', adjustable='box')\n",
    "    \n",
    "        if jj==0:\n",
    "            ax.annotate(r'Rel. growth rate (observed), $\\lambda$',\n",
    "                        xy=(1.2, -0.175), xycoords=\"axes fraction\",\n",
    "                        ha='center', va='center', annotation_clip=False, fontsize=6)\n",
    "            ax.set_ylabel(r'Rel. growth rate (fitted), $\\hat{\\lambda}$', fontsize=6)\n",
    "    \n",
    "for ax in fig.get_axes():\n",
    "    ax.tick_params(axis='both', which='major', labelsize=6, size=2)\n",
    "    ax.tick_params(axis='both', which='minor', labelsize=6, size=0)\n",
    "        \n",
    "    for loc in ['top','bottom','left','right']:\n",
    "        ax.spines[loc].set_linewidth(.6)\n",
    "\n",
    "fig.savefig(dir_supp+'figures/supp_figure_pheno_cross/supp_figure_pheno_cross_lmm.png', dpi=300, bbox_inches='tight')\n",
    "fig.savefig(dir_supp+'figures/supp_figure_pheno_cross/supp_figure_pheno_cross_lmm.svg', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fig. S13:** Hierarchical analysis of variance in the genetic cross using linear mixed models. We model the growth rate of spores $\\lambda_{\\{a,\\alpha\\}}^{btd}$ and hybrids $\\lambda_{a\\alpha}^{btd}$ as a function of background genotype $b$, *de novo* genotype $d$, sampling time during selection $t$, and auxotrophy $x$. Relative growth rates are accurately fitted by this model (Model 4). Model fits are summarised in Table S7. Measurements are taken in YNB+HU 10 $\\text{mg ml}^{-1}$ and YNB only for populations selected in hydroxyurea (*A*, *B*); YNB+RM 0.025 $\\mu \\text{g ml}^{-1}$ and YNB only for populations selected in rapamycin (*C*, *D*). The scatter shows a set of measurements $\\lambda$ ($x$-axis) against the fitted rates $\\hat{\\lambda}$ ($y$-axis). The total variance explained, $r^2$, is separately computed for spores and hybrids by environment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# import statsmodels.api as sm\n",
    "\n",
    "# param='growth_rate'\n",
    "\n",
    "# sdf = spores_norm_df\n",
    "# sdf = sdf[sdf.group.isin(['ancestral','evolved'])]\n",
    "# sdf = sdf[np.isfinite(sdf[param])]\n",
    "\n",
    "# hdf = hybrids_norm_df\n",
    "# hdf = hdf[hdf.group.isin([('ancestral','ancestral'),('ancestral','evolved'),('evolved','evolved')])]\n",
    "# hdf = hdf[np.isfinite(hdf[param])]\n",
    "\n",
    "# formulas = {'Model 1A': 'growth_rate ~ C(group)',\n",
    "#             'Model 1B': 'growth_rate ~ C(group) + C(genotype_long)',\n",
    "#             'Model 1C': 'growth_rate ~ C(group) + C(genotype_long) + C(auxotrophy)',\n",
    "#             'Model 2': 'growth_rate ~ C(group) + C(genotype_long) + C(auxotrophy) + C(individual)'}\n",
    "\n",
    "# fig = plt.figure(figsize=(7, 8), dpi=300)\n",
    "\n",
    "# grid = gridspec.GridSpec(4, 2, width_ratios=[1,2], hspace=0.5, wspace=0.5)\n",
    "\n",
    "# gs = {}\n",
    "# for ii, formula in enumerate(['Model 1A', 'Model 1B', 'Model 1C', 'Model 2']):\n",
    "#     gs[('anova', formula)] = gridspec.GridSpecFromSubplotSpec(2, 1, subplot_spec=grid[ii,0], hspace=0.1)\n",
    "    \n",
    "# for ii, (env_evo, env_test) in enumerate([('HU','HU'),('HU','YNB'),('RM','RM'),('RM','YNB')]):\n",
    "#     gs[('fit', (env_evo, env_test))] = gridspec.GridSpecFromSubplotSpec(1, 2, subplot_spec=grid[ii,1], wspace=0.05)\n",
    "\n",
    "# model_s = {}\n",
    "# model_h = {}\n",
    "# anova_s = {}\n",
    "# anova_h = {}\n",
    "\n",
    "# for ii, formula in enumerate(['Model 1A', 'Model 1B', 'Model 1C', 'Model 2']):\n",
    "\n",
    "#     model_s[formula] = sdf.groupby(['selection','environment']).apply(lmm.run_anova, formula=formulas[formula]) \n",
    "#     model_h[formula] = hdf.groupby(['selection','environment']).apply(lmm.run_anova, formula=formulas[formula])\n",
    "#     anova_s[formula] = sdf.groupby(['selection','environment']).apply(lmm.variance_vectors, formula=formulas[formula])\n",
    "#     anova_h[formula] = hdf.groupby(['selection','environment']).apply(lmm.variance_vectors, formula=formulas[formula])\n",
    "#     df = pd.concat([anova_s[formula], anova_h[formula]], \n",
    "#                    keys=['spores','hybrids'], names=['type'])\n",
    "    \n",
    "#     df = df['var_explained']\n",
    "    \n",
    "#     # fix columns\n",
    "#     df = df.rename(\n",
    "#         columns={\n",
    "#             'C(group)':'Time',\n",
    "#             'C(genotype_long)':'De novo\\n(genotype)',\n",
    "#             'C(auxotrophy)':'Auxotrophy',\n",
    "#             'C(individual)':'Background\\n(genotype)',\n",
    "#             'C(replicate)':'Measurement\\nerror'\n",
    "#         }\n",
    "#     )\n",
    "    \n",
    "#     order = ['Background\\n(genotype)','De novo\\n(genotype)','Time','Auxotrophy']#,'Measurement\\nerror']\n",
    "#     order = [val for val in order if val in df.columns]\n",
    "#     df = df[order]\n",
    "    \n",
    "#     # fix index\n",
    "#     df = df.swaplevel('type', 'selection')\n",
    "#     df = df.swaplevel('type', 'environment')\n",
    "#     df = df.sort_index(level=['selection','environment','type'])\n",
    "#     df = df.sort_index(ascending=False)\n",
    "    \n",
    "#     for jj, (env_evo, ge) in enumerate(df.groupby(level='selection')):\n",
    "#         print ii,jj\n",
    "#         if jj==0:\n",
    "#             ax = plt.subplot(gs[('anova', formula)][jj])\n",
    "#             ax.text(-0.175, 1.285, chr(ii + ord('A')), transform=ax.transAxes,\n",
    "#                     fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "#             ax.text(.5, 1.285, formula, transform=ax.transAxes,\n",
    "#                     fontsize=6, va='top', ha='center')\n",
    "#         else:\n",
    "#             ax = plt.subplot(gs[('anova', formula)][jj], sharex=ax)\n",
    "#         utils.simple_axes(ax)\n",
    "\n",
    "#         ge.plot(ax=ax, kind='barh', stacked=True, \n",
    "#                 color=[config.factors['color'][x] for x in ge.columns],\n",
    "#                 align='center', legend=False, width=0.75)\n",
    "    \n",
    "#         bars = ax.patches\n",
    "#         hatches = [config.factors['hatch'][x] for x in ge.index.get_level_values('type')]*len(ge)\n",
    "\n",
    "#         for bar, hatch in zip(bars, hatches):\n",
    "#             bar.set_hatch(hatch)\n",
    "            \n",
    "#         # x-axis\n",
    "#         ax.set_xlabel('Variance (%)', fontsize=6)\n",
    "#         xscale = 1e-2\n",
    "#         xticks = ticker.FuncFormatter(lambda x, pos: '{0:g}'.format(x/xscale))\n",
    "#         ax.xaxis.set_major_formatter(xticks)\n",
    "#         ax.set_xlim((0,1))\n",
    "        \n",
    "#         # y-axis\n",
    "#         ax.set_ylabel('', fontsize=6)\n",
    "#         if jj==0:\n",
    "#             ax.annotate(r'Measurement', \n",
    "#                         xy=(-0.175, 0.), xycoords=\"axes fraction\",\n",
    "#                         va='center', ha='right', annotation_clip=False, fontsize=6, rotation=90)\n",
    "#         ystart, yend, ylabels = plot.set_custom_labels(ge.index, 1)\n",
    "#         ax.set_yticks(ylabels.values())\n",
    "#         ax.set_yticklabels(ylabels.keys(), fontsize=6)\n",
    "        \n",
    "#         # annotate panels and axis\n",
    "#         ax.annotate(env_evo, xy=(0.9, 0.05), xycoords=('axes fraction', 'axes fraction'),\n",
    "#                     ha='center', va='bottom', fontsize=6)\n",
    "        \n",
    "# ax = plt.subplot(gs[('anova', 'Model 2')][0])\n",
    "\n",
    "# leg1 = ax.legend(bbox_to_anchor=(1.1, 4.65), ncol=1, \n",
    "#                  frameon=False, loc='lower left', borderaxespad=0.,\n",
    "#                  handlelength=0.75, prop={'size':6})\n",
    "\n",
    "# ax = plt.subplot(gs[('anova', 'Model 2')][0])\n",
    "\n",
    "# s_artist = patches.Rectangle((0,0), width=1, height=1, \n",
    "#                              facecolor='w', edgecolor='k')\n",
    "# h_artist = patches.Rectangle((0,0), width=1, height=1, hatch='///', \n",
    "#                              facecolor='w', edgecolor='k')\n",
    "\n",
    "# leg2 = ax.legend([s_artist,h_artist], ['Spores','Hybrids'],\n",
    "#                  bbox_to_anchor=(1.1, 4.5), ncol=1, \n",
    "#                  frameon=False, loc='upper left', borderaxespad=0., \n",
    "#                  handlelength=0.75, prop={'size':6})\n",
    "\n",
    "# ax.add_artist(leg1)\n",
    "\n",
    "# formula = 'Model 2'\n",
    "\n",
    "# for jj, ((env_evo, env_test), g) in enumerate(model_s[formula].groupby(level=['selection','environment'])):\n",
    "        \n",
    "#     for kk, (label, model, data, marker) in enumerate(zip(['Spores', 'Hybrids'], \n",
    "#                                                           [model_s[formula], model_h[formula]],\n",
    "#                                                           [sdf, hdf],\n",
    "#                                                           ['o','x'])):\n",
    "        \n",
    "#         ax = plt.subplot(gs[('fit', (env_evo,env_test))][kk])\n",
    "#         utils.simple_axes(ax)\n",
    "        \n",
    "#         if kk==0:\n",
    "#             ax.text(-0.2, 1.1285, chr(jj + ord('E')), transform=ax.transAxes,\n",
    "#                 fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "#             ax.annotate('Selection: %s\\nMeasurement: %s' %\\\n",
    "#                          (config.selection['long_label'][env_evo], \\\n",
    "#                           config.environment['long_label'][env_test]), \n",
    "#                         xy=(0, 1.125), xycoords=\"axes fraction\",\n",
    "#                         ha='left', va='center', annotation_clip=False, fontsize=6)\n",
    "\n",
    "#         lm, anova = model.ix[(env_evo, env_test)]\n",
    "#         data = data.groupby(['selection','environment']).get_group((env_evo, env_test))\n",
    "        \n",
    "#         nobs = lm.nobs\n",
    "#         y = data.growth_rate\n",
    "#         yhat = lm.fittedvalues\n",
    "#         ymin, ymax = y.quantile([.005, .995])\n",
    "#         yrange = ymax - ymin\n",
    "        \n",
    "#         # scatter\n",
    "#         ax.scatter(y, yhat, facecolor='gray', s=1.5, lw=0, alpha=0.75)\n",
    "        \n",
    "#         # fit\n",
    "#         line_fit = sm.OLS(y, sm.add_constant(yhat, prepend=False)).fit().params\n",
    "#         fit = lambda x: line_fit[1] + line_fit[0] * x \n",
    "#         ax.plot(np.linspace(ymin-.1*yrange, ymax+.1*yrange, nobs), \n",
    "#                 fit(np.linspace(ymin-.1*yrange, ymax+.1*yrange, nobs)), \n",
    "#                 label=label, color='k', ls='--', lw=1)\n",
    "        \n",
    "#         # variance explained\n",
    "#         factors = [each for each in anova.index if each != 'Residual']\n",
    "#         var_total = np.sum([lmm.variance_explained(factor, anova) for factor in factors])\n",
    "#         ax.annotate('{}\\n$r^2={:.2f}$'.format(label, var_total), \n",
    "#                     xy=(.1, .9), xycoords=\"axes fraction\",\n",
    "#                     ha='left', va='center', fontsize=6)\n",
    "        \n",
    "#         # x-axis\n",
    "#         ax.set_xlim(ymin-.1*yrange, ymax+.1*yrange)\n",
    "#         ax.xaxis.set_major_locator( ticker.MaxNLocator(nbins = 4) )\n",
    "\n",
    "#         # y-axis\n",
    "#         ax.set_ylim(ymin-.1*yrange, ymax+.1*yrange)\n",
    "#         ax.yaxis.set_major_locator( ticker.MaxNLocator(nbins = 4) )\n",
    "        \n",
    "#         ax.set_aspect('equal', adjustable='box')\n",
    "    \n",
    "#         if kk==0:\n",
    "#             ax.annotate(r'Rel. growth rate (observed), $\\lambda$', \n",
    "#                         xy=(1.2, -0.175), xycoords=\"axes fraction\",\n",
    "#                         ha='center', va='center', annotation_clip=False, fontsize=6)\n",
    "#             ax.set_ylabel(r'Rel. growth rate (fitted), $\\hat{\\lambda}$', fontsize=6)\n",
    "    \n",
    "# for ax in fig.get_axes():\n",
    "#     ax.tick_params(axis='both', which='major', labelsize=6, size=2)\n",
    "#     ax.tick_params(axis='both', which='minor', labelsize=6, size=0)\n",
    "        \n",
    "#     for loc in ['top','bottom','left','right']:\n",
    "#         ax.spines[loc].set_linewidth(.6)\n",
    "\n",
    "# fig.savefig(dir_supp+'figures/supp_figure_pheno_cross/supp_figure_pheno_cross_lmm.png', dpi=300, bbox_inches='tight')\n",
    "# fig.savefig(dir_supp+'figures/supp_figure_pheno_cross/supp_figure_pheno_cross_lmm.svg', dpi=300, bbox_inches='tight')\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# class LM(object):\n",
    "#     from rpy2.robjects import pandas2ri\n",
    "#     pandas2ri.activate()\n",
    "#     from rpy2.robjects.packages import importr\n",
    "#     stats = importr('stats')\n",
    "#     def convert_fit_to_python(self, fit):\n",
    "#         coeffs_r = fit.rx2('coefficients')\n",
    "#         coeffs= pandas2ri.ri2py(coeffs_r)\n",
    "#         coeff_names =  pandas2ri.ri2py(coeffs_r.names).tolist()\n",
    "#         coeff_series = pd.Series({k:v for k,v in zip(coeff_names, coeffs)})\n",
    "#         fitted_values = pandas2ri.ri2py(fit.rx2('fitted.values'))\n",
    "#         return coeff_series\n",
    "        \n",
    "#     def fit(self, formula, df):\n",
    "#         \"\"\"\n",
    "#         formula: string of the form: a~b+c\n",
    "#         df: Pandas Dataframe\n",
    "#         \"\"\"\n",
    "#         # Get R dataframe\n",
    "#         r_df = pandas2ri.py2ri(df)\n",
    "#         # Create linear fit\n",
    "#         fit = stats.lm(formula, data=df)\n",
    "#         self.fit = fit\n",
    "#         python_fit = self.convert_fit_to_python(fit)\n",
    "#         return python_fit\n",
    "    \n",
    "#     def predict(self, df):\n",
    "#         pred_r = stats.predict(self.fit, newdata=df)\n",
    "#         pred_python = pandas2ri.ri2py(pred_r)\n",
    "#         return pred_python\n",
    "\n",
    "# class LMM(object):\n",
    "#     from rpy2.robjects import pandas2ri\n",
    "#     pandas2ri.activate()\n",
    "#     from rpy2.robjects.packages import importr\n",
    "#     lme4 = importr('lme4')\n",
    "#     def convert_fit_to_python(self, fit):\n",
    "#         coeffs_r = fit.rx2('coefficients')\n",
    "#         coeffs= pandas2ri.ri2py(coeffs_r)\n",
    "#         coeff_names =  pandas2ri.ri2py(coeffs_r.names).tolist()\n",
    "#         coeff_series = pd.Series({k:v for k,v in zip(coeff_names, coeffs)})\n",
    "#         fitted_values = pandas2ri.ri2py(fit.rx2('fitted.values'))\n",
    "#         return coeff_series\n",
    "        \n",
    "#     def fit(self, formula, df):\n",
    "#         \"\"\"\n",
    "#         formula: string of the form: a~b+c\n",
    "#         df: Pandas Dataframe\n",
    "#         \"\"\"\n",
    "#         # Get R dataframe\n",
    "#         r_df = pandas2ri.py2ri(df)\n",
    "#         # Create linear fit\n",
    "#         fit = lme4.lmer(formula, data=df)\n",
    "#         self.fit = fit\n",
    "# #         python_fit = self.convert_fit_to_python(fit)\n",
    "#         return fit\n",
    "    \n",
    "#     def predict(self, df):\n",
    "#         pred_r = stats.predict(self.fit, newdata=df)\n",
    "#         pred_python = pandas2ri.ri2py(pred_r)\n",
    "#         return pred_python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# from rpy2.robjects.packages import importr\n",
    "# r = rpy2.robjects.r\n",
    "# import rpy2.robjects as robjects\n",
    "# import rpy2.robjects\n",
    "# from rpy2.robjects.packages import importr\n",
    "# from rpy2.robjects import pandas2ri\n",
    "# pandas2ri.activate()\n",
    "# stats = importr('stats')\n",
    "# lme4 = importr('lme4')\n",
    "# from rpy2.robjects.packages import importr\n",
    "# from rpy2.robjects import pandas2ri\n",
    "# pandas2ri.activate()\n",
    "# stats = importr('stats')\n",
    "# lme4 = importr('lme4')\n",
    "# lm = LM()\n",
    "# lmm = LMM()\n",
    "# model_1a = lm.fit(formula='growth_rate~1+group', df=sdf)\n",
    "# model_1b = lm.fit(formula='growth_rate~1+group+genotype_long', df=sdf)\n",
    "# model_1c = lm.fit(formula='growth_rate~1+group+genotype_long+auxotrophy', df=sdf)\n",
    "# model_2 = lmm.fit(formula='growth_rate~1+(1|individual)+group+genotype_long+auxotrophy', df=sdf)\n",
    "# pred_x = lmm.predict(sdf)\n",
    "# plt.scatter(sdf['growth_rate'].values, pred_x, label='Fitted')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  },
  "latex_metadata": {
   "affiliation": "Wellcome Trust Sanger Institute, Hinxton, Cambridge CB10 1SA, United Kingdom; Department of Applied Mathematics and Theoretical Physics, University of Cambridge, Cambridge CB3 0WA, United Kingdom",
   "author": "Ignacio Vazquez-Garcia",
   "title": "Background-dependent effects of selection on subclonal heterogeneity"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
