{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Supplemental Information:\n",
    "\n",
    "> **\"Clonal heterogeneity influences the fate of new adaptive mutations\"**\n",
    "\n",
    "> Ignacio Vázquez-García, Francisco Salinas, Jing Li, Andrej Fischer, Benjamin Barré, Johan Hallin, Anders Bergström, Elisa Alonso-Pérez, Jonas Warringer, Ville Mustonen, Gianni Liti\n",
    "\n",
    "## Figure 6 (+ Supp. Figs.)\n",
    "\n",
    "This IPython notebook is provided for reproduction of Figures 6, S10, S11 and S12 of the paper. It can be viewed by copying its URL to nbviewer and it can be run by opening it in binder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "# Load external dependencies\n",
    "from setup import *\n",
    "# Load internal dependencies\n",
    "import config,plot,utils\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this experiment, we cross individuals of the ancestral population with pre-existing variation and mutant individuals from the evolved population. All of these are diploid cells which undergo sporulation (labelled *spores*) and mating (creating *hybrids*). We evaluate phenotypes of the spore and hybrid genotypes. Phenotypic variation can then be attributed to dominance or epistatic effects."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data import\n",
    "\n",
    "### Spores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "IOError",
     "evalue": "File /Users/ivg/projects/yeast-cancer-drugs/paper/science/submission/data/pheno/genetic-cross/pheno_spores.csv does not exist",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIOError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-d39636f56c32>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Load data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mspores_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdir_data\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'pheno/genetic-cross/pheno_spores.csv'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mencoding\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'utf-8'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeep_default_na\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna_values\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'NaN'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m# Filter out blank positions and constructs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mspores_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mspores_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m~\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspores_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'strain'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'control'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m''\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m&\u001b[0m                      \u001b[0;34m~\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspores_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'genotype_long'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34mu'fpr1Δ'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34mu'tor1Δ'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/.virtualenvs/genetic-variation/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, skipfooter, skip_footer, doublequote, delim_whitespace, as_recarray, compact_ints, use_unsigned, low_memory, buffer_lines, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    653\u001b[0m                     skip_blank_lines=skip_blank_lines)\n\u001b[1;32m    654\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 655\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    656\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    657\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/.virtualenvs/genetic-variation/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    403\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    404\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 405\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    406\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    407\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/.virtualenvs/genetic-variation/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    762\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'has_index_names'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'has_index_names'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    763\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 764\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    765\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    766\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/.virtualenvs/genetic-variation/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m    983\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'c'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    984\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'c'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 985\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    986\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    987\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'python'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/.virtualenvs/genetic-variation/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   1603\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'allow_leading_cols'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex_col\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1604\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1605\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparsers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1606\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1607\u001b[0m         \u001b[0;31m# XXX\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__ (pandas/_libs/parsers.c:4209)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._setup_parser_source (pandas/_libs/parsers.c:8873)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mIOError\u001b[0m: File /Users/ivg/projects/yeast-cancer-drugs/paper/science/submission/data/pheno/genetic-cross/pheno_spores.csv does not exist"
     ]
    }
   ],
   "source": [
    "# Load data\n",
    "spores_df = pd.read_csv(dir_data+'pheno/genetic-cross/pheno_spores.csv', encoding='utf-8', keep_default_na=False, na_values='NaN')\n",
    "\n",
    "# Filter out blank positions and constructs\n",
    "spores_df = spores_df[~(spores_df['strain'].isin(['control','']))&\\\n",
    "                      ~(spores_df['genotype_long'].isin([u'fpr1Δ',u'tor1Δ']))]\n",
    "\n",
    "# Set contaminated samples as missing (NaN)\n",
    "spores_df.loc[(spores_df['contamination']=='yes'),\n",
    "              ['abs_growth_rate','abs_doubling_time','norm_growth_rate','norm_doubling_time']] = np.nan\n",
    "\n",
    "# Define individual and replicate\n",
    "spores_df['individual'] = spores_df[['group','background','gene','tetrad','spore']].apply(tuple, axis=1)\n",
    "spores_df['replicate'] = spores_df[['group','background','gene','tetrad','spore','plate','row','column']].apply(tuple, axis=1)\n",
    "\n",
    "# # Filter out measurement replicates with >5% measurement error\n",
    "# spores_df['pct'] = spores_df.groupby(['selection','environment','replicate'])['growth_rate']\\\n",
    "# .apply(lambda x: (x-x.mean())/float(x.mean()))\n",
    "# spores_df = spores_df[abs(spores_df['pct'])<0.05]\n",
    "\n",
    "spores_df.head()  # show dataframe header to stdout"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hybrids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Load data\n",
    "hybrids_df = pd.read_csv(dir_data+'pheno/genetic-cross/pheno_hybrids.csv', encoding='utf-8', keep_default_na=False, na_values='NaN')\n",
    "\n",
    "# Combine MATa/α to create tuples\n",
    "for column in ['strain','group','background','gene','genotype_short','genotype_long','amino_acids']:\n",
    "    hybrids_df[column] = utils.combine_columns(hybrids_df,u'%s_MATa'%column,u'%s_MATα'%column)\n",
    "for column in ['auxotrophy','mating','tetrad','spore','strain','contamination']:\n",
    "    hybrids_df[column] = utils.combine_columns(hybrids_df,u'%s_MATa'%column,u'%s_MATα'%column,mirror=False)\n",
    "\n",
    "# Filter out blank positions, contamination and constructs\n",
    "hybrids_df = hybrids_df[~(hybrids_df['strain'].isin([('control','control'),('','')]))&\\\n",
    "                        ~(hybrids_df['genotype_long'].isin([u'fpr1Δ',u'tor1Δ']))]\n",
    "\n",
    "# Set contaminated samples as missing (NaN)\n",
    "hybrids_df.loc[hybrids_df['contamination'].isin([('yes','yes'),('no','yes'),('yes','no')]),\n",
    "              ['abs_growth_rate','abs_doubling_time','norm_growth_rate','norm_doubling_time']] = np.nan\n",
    "\n",
    "# Fix diploid genotypes for different genes\n",
    "hybrids_df.loc[(hybrids_df[u'gene_MATa']=='RNR4')&\\\n",
    "               (hybrids_df[u'gene_MATα']=='RNR2'),'genotype_short'] =\\\n",
    "hybrids_df.loc[(hybrids_df[u'gene_MATa']=='RNR4')&\\\n",
    "               (hybrids_df[u'gene_MATα']=='RNR2'),'genotype_short'].apply(lambda x: x[::-1])\n",
    "hybrids_df.loc[(hybrids_df[u'gene_MATa']=='TOR1')&\\\n",
    "               (hybrids_df[u'gene_MATα']=='FPR1'),'genotype_short'] =\\\n",
    "hybrids_df.loc[(hybrids_df[u'gene_MATa']=='TOR1')&\\\n",
    "               (hybrids_df[u'gene_MATα']=='FPR1'),'genotype_short'].apply(lambda x: x[::-1])\n",
    "\n",
    "# Define individual\n",
    "hybrids_df['individual'] = hybrids_df[['group','background','gene','tetrad','spore']].apply(tuple, axis=1)\n",
    "hybrids_df['replicate'] = hybrids_df[['group','background','gene','tetrad','spore','plate','row','column']].apply(tuple, axis=1)\n",
    "\n",
    "# # Filter out measurement replicates with >5% measurement error\n",
    "# hybrids_df['pct'] = hybrids_df.groupby(['selection','environment','replicate'])['growth_rate']\\\n",
    "# .apply(lambda x: (x-x.mean())/float(x.mean()))\n",
    "# hybrids_df = hybrids_df[abs(hybrids_df['pct'])<0.05]\n",
    "\n",
    "hybrids_df.head()  # show dataframe header to stdout"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Load data\n",
    "clone_df = pd.read_csv(dir_data+'pheno/populations/pheno_populations.csv', encoding='utf-8', keep_default_na=False, na_values='NaN')\n",
    "\n",
    "# Filter out strains used for spatial control\n",
    "clone_df = clone_df[(clone_df.group == 'ancestral')|\\\n",
    "                    (clone_df.group == 'evolved')]\n",
    "\n",
    "clone_df['genotype_short'] = map(lambda x: tuple(x.strip(\"/\").split(\"/\")),clone_df['genotype_short'])\n",
    "clone_df['genotype_long'] = map(lambda x: tuple(x.strip(\"/\").split(\"/\")),clone_df['genotype_long'])\n",
    "\n",
    "clone_df['group'] = zip(clone_df.group, clone_df.group)\n",
    "clone_df['background'] = zip(clone_df.background, clone_df.background)\n",
    "clone_df['gene'] = zip(clone_df.gene, clone_df.gene)\n",
    "\n",
    "clone_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Analysis of variance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "lmm_fit = pd.read_csv(dir_data+'pheno/genetic-cross/lmm_fit.csv')\n",
    "lmm_var_comp = pd.read_csv(dir_data+'pheno/genetic-cross/lmm_var_components.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We need to filter and sort the vector of spore phenotypes and matrix of hybrid phenotypes by environment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def filter_spores(S, env_evo):\n",
    "    # Filter by dictionary\n",
    "    S = S[(S['group'].isin(config.spores_bg['position'][env_evo].keys())) &\n",
    "          (S['genotype_short'].isin(config.spores_gt_short['position'][env_evo].keys())) &\n",
    "          (S['background'].isin(config.spores_cl['position'][env_evo].keys()))]\n",
    "    return S\n",
    "          \n",
    "def filter_hybrids(H, env_evo):\n",
    "    # Filter by dictionary\n",
    "    H = H[(H['group'].isin(config.hybrids_bg['position'][env_evo].keys())) &\n",
    "          (H['genotype_short'].isin(config.hybrids_gt_short['position'][env_evo].keys())) &\n",
    "          (H['background'].isin(config.hybrids_cl['position'][env_evo].keys()))]\n",
    "    return H\n",
    "    \n",
    "def sort_spores(S, env_evo):\n",
    "    # Apply sorting ranks to reorder rows\n",
    "    S.loc[:,'rank_group'] = S['group'].map(config.spores_bg['position'][env_evo])\n",
    "    S.loc[:,'rank_background'] = S['background'].map(config.spores_cl['position'][env_evo])\n",
    "    S.loc[:,'rank_gene'] = S['gene'].map(config.spores_gn['position'][env_evo])\n",
    "    S.loc[:,'rank_genotype'] = S['genotype_short'].map(config.spores_gt_short['position'][env_evo])\n",
    "    S.sort_values(['rank_group','rank_background','rank_gene','rank_genotype'],\n",
    "                  ascending=True,inplace=True)\n",
    "    return S\n",
    "\n",
    "def sort_hybrids(H, env_evo):\n",
    "    # Apply sorting ranks to reorder rows\n",
    "    H.loc[:,'rank_group'] = H['group'].map(config.hybrids_bg['position'][env_evo])\n",
    "    H.loc[:,'rank_background'] = H['background'].map(config.hybrids_cl['position'][env_evo])\n",
    "    H.loc[:,'rank_gene'] = H['gene'].map(config.hybrids_gn['position'][env_evo])\n",
    "    H.loc[:,'rank_genotype'] = H['genotype_short'].map(config.hybrids_gt_short['position'][env_evo])\n",
    "    H.sort_values(['rank_group','rank_background','rank_gene','rank_genotype'],\n",
    "                  ascending=True,inplace=True)\n",
    "    return H"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 6 - Background-averaged fitness effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from IPython.display import Image\n",
    "Image(filename=dir_paper+'figures/figure6/figure6A_schematic_publication.png', retina=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "param='norm_growth_rate'\n",
    "\n",
    "fig = plt.figure(figsize=(12, 6))\n",
    "\n",
    "grid = gridspec.GridSpec(2, 3, width_ratios=[1.1,3,1.075], hspace=0.35, wspace=0.2)\n",
    "\n",
    "gs = {}\n",
    "gs['schematic'] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,0])\n",
    "gs['variance'] = gridspec.GridSpecFromSubplotSpec(3, 1, height_ratios=[5,5,2], subplot_spec=grid[1,0], hspace=0.1)\n",
    "gs[('HU','heatmap')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,2])\n",
    "gs[('HU','barplot')] = gridspec.GridSpecFromSubplotSpec(1, 24, subplot_spec=grid[0,1], wspace=0)\n",
    "gs[('RM','heatmap')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,2])\n",
    "gs[('RM','barplot')] = gridspec.GridSpecFromSubplotSpec(1, 24, subplot_spec=grid[1,1], wspace=0)\n",
    "\n",
    "### Schematic ###\n",
    "ax = plt.subplot(gs['schematic'][:])\n",
    "ax.text(-0.1, 1.125,  chr(ord('A')), transform=ax.transAxes,\n",
    "        fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "ax.axis('off')\n",
    "\n",
    "### Box plots ###\n",
    "groups_sp = spores_df.groupby(['selection','environment'])\n",
    "groups_hy = hybrids_df.groupby(['selection','environment'])\n",
    "groups_cl = clone_df.groupby(['selection','environment'])\n",
    "\n",
    "for (ii, (env_evo, env_test)) in enumerate([('HU','HU'),('RM','RM')]):\n",
    "    \n",
    "    # Group by evolution and test environment\n",
    "    S = groups_sp.get_group((env_evo, env_test))\n",
    "    H = groups_hy.get_group((env_evo, env_test))\n",
    "    C = groups_cl.get_group((env_evo, env_test))\n",
    "    \n",
    "    # Filter by dictionary\n",
    "    S = filter_spores(S, env_evo)\n",
    "    H = filter_hybrids(H, env_evo)\n",
    "    \n",
    "    # Apply sorting ranks to reorder rows\n",
    "    S = sort_spores(S, env_evo)\n",
    "    H = sort_hybrids(H, env_evo)\n",
    "    \n",
    "    ### Spores barplot ###\n",
    "    ax1 = plt.subplot(gs[(env_evo,'barplot')][:7])\n",
    "\n",
    "    ax1.text(-0.2, 1.15,  chr(3*ii + ord('B')), transform=ax1.transAxes,\n",
    "             fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "    \n",
    "    ax1.text(0, 1.15,  \n",
    "             'Selection: %s\\nMeasurement: %s' %\\\n",
    "             (config.selection['long_label'][env_evo], config.environment['long_label'][env_test]), \n",
    "             transform=ax1.transAxes,\n",
    "             fontsize=6, va='center', ha='left')\n",
    "    \n",
    "    S = S.groupby(['group','background','gene','genotype_short','tetrad','spore'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['group','background','gene','genotype_short'])\n",
    "    S_ref = S.ix[:,'ancestral'].median().values[0]\n",
    "    S = S-S_ref\n",
    "    \n",
    "    bp = S.plot(\n",
    "        ax=ax1, kind='box', by='group', return_type='dict',\n",
    "        labels=S.columns.get_level_values('group')\n",
    "    )\n",
    "\n",
    "    colors = [config.spores_gt_short['color'][env_evo][x] for x in S.columns.get_level_values('genotype_short')]\n",
    "    plot.boxplot_custom(bp, ax1, colors=colors, hatches=[' ']*10)\n",
    "        \n",
    "    ax1.set_title('Spores (haploid)', fontsize=6, y=1, weight='bold')\n",
    "    ax1.set_xlabel('')\n",
    "    ax1.set_ylabel(r'Rel. growth rate, $\\lambda^{btd}_{a\\alpha}$', fontsize=6, labelpad=2)\n",
    "\n",
    "    ### Hybrids barplot ###\n",
    "    ax2 = plt.subplot(gs[(env_evo,'barplot')][7:], sharey=ax1)\n",
    "    \n",
    "    H = H.groupby(['group','background','gene','genotype_short','tetrad','spore'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['group','background','gene','genotype_short'])\n",
    "    ix_use = pd.MultiIndex.from_tuples([[('ancestral', 'ancestral')]], names=['group'])\n",
    "    H_ref = H.loc[:,H.columns.get_level_values('group').isin(ix_use.get_level_values(0))].median().squeeze()\n",
    "    H = H-H_ref\n",
    "    \n",
    "    C = C.groupby(['group','background','gene','genotype_short','isolate'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\n",
    "    C_ref = C[('ancestral', 'ancestral')].median()\n",
    "    C = C-C_ref\n",
    "    \n",
    "    C = C.unstack(level=['group','background','gene','genotype_short'])\n",
    "    C = C.loc[:,(C.columns.get_level_values('group').isin([('evolved','evolved')])\\\n",
    "                 &~(C.columns.get_level_values('gene').isin([('','')])))]\n",
    "    C = C.reindex(columns=H.columns)\n",
    "    C = C.dropna(how='all')\n",
    "\n",
    "    bp = H.plot(\n",
    "        ax=ax2, kind='box', by='group', return_type='dict', \n",
    "        labels=H.columns.get_level_values('group')\n",
    "    )\n",
    "\n",
    "    colors = [config.hybrids_gt_short['color'][env_evo][x] for x in H.columns.get_level_values('genotype_short')]\n",
    "    plot.boxplot_custom(bp, ax2, colors=colors, hatches=[' ']*30)\n",
    "\n",
    "    for i,d in enumerate(C):\n",
    "        y = C[d]\n",
    "        x = [i+1]*len(y)\n",
    "        ax2.plot(x, y, \n",
    "                 mfc=config.hybrids_gt_short['color'][env_evo][d[-1]], mec='k', \n",
    "                 ms=3, marker=\"D\", linestyle=\"None\", zorder=6)\n",
    "    \n",
    "    ax2.set_title('Hybrids (diploid)', fontsize=6, y=1, weight='bold')\n",
    "    ax2.set_xlabel('')\n",
    "    ax2.set_ylabel('', fontsize=6)\n",
    "    \n",
    "    wt_artist = lines.Line2D((0,.75),(0,0), mec='k', mfc=config.hybrids_gt_short['color'][env_evo][('+','+')], marker='s', markersize=4, linestyle='')\n",
    "    het_artist = lines.Line2D((0,.75),(0,0), mec='k', mfc=config.hybrids_gt_short['color'][env_evo][('+','-')], marker='s', markersize=4, linestyle='')\n",
    "    hom_artist = lines.Line2D((0,.75),(0,0), mec='k', mfc=config.hybrids_gt_short['color'][env_evo][('-','-')], marker='s', markersize=4, linestyle='')\n",
    "    clone_artist = lines.Line2D((0,.75),(0,0), mec='k', mfc='w', marker='D', markersize=3, linestyle='')\n",
    "\n",
    "    leg = ax2.legend([wt_artist,het_artist,hom_artist,clone_artist], \n",
    "                     ['WT','het.','hom.','clone'], ncol=1, \n",
    "                     frameon=False, loc='upper right',\n",
    "                     borderaxespad=0, handlelength=0.75, \n",
    "                     prop={'size':5}, labelspacing=.32)\n",
    "        \n",
    "    # Grid\n",
    "    for ax, g in zip([ax1,ax2],[S,H]):\n",
    "        ### horizontal ###\n",
    "        ax.yaxis.grid(ls=\"-\", lw=.8, color=\"0.9\", zorder=0)\n",
    "        ax.axhline(y=0., c='k', ls=\"--\", dashes=(7, 7), lw=.8, zorder=1)\n",
    "        ax.set_axisbelow(True)\n",
    "        \n",
    "        ### vertical ###\n",
    "        ## Background\n",
    "        xstart, xend, xlabels = plot.set_custom_labels(g.columns, 0)\n",
    "        # labels\n",
    "        for k, v in xlabels.iteritems():\n",
    "            ax.annotate('\\nx\\n'.join(k) if isinstance(k, tuple) else k, \n",
    "                        xy=(v+1, 0.97), xycoords=(\"data\", \"axes fraction\"),\n",
    "                        ha='center', va='top', annotation_clip=False, fontsize=5)\n",
    "        # grid\n",
    "        xgrid=[xst+1.5 for xst in list(set(xstart.values()))]\n",
    "        [ax.axvline(x, lw=1.0, ls='-', color='0.9', zorder=0) for x in xgrid]\n",
    "\n",
    "        ## gene\n",
    "        xend, xstart, xlabels = plot.set_custom_labels(g.columns, 2)\n",
    "        \n",
    "        transform = transforms.blended_transform_factory(ax.transData, ax.transAxes)\n",
    "        for k in xlabels:\n",
    "            if (k!='') & (k!=('','')):\n",
    "                if abs(xstart[k]-xend[k])>0:\n",
    "                    line = lines.Line2D([xstart[k]+.75,xend[k]+1.25], [-.05,-.05], \n",
    "                                        color='k', lw=1, transform=transform)\n",
    "                else:\n",
    "                    line = lines.Line2D([xstart[k]+.75,xend[k]+1.25], [-.05,-.05], \n",
    "                                        color='k', lw=1, transform=transform)\n",
    "                ax.add_line(line)\n",
    "                line.set_clip_on(False)\n",
    "        \n",
    "        ax.set_xticks([x+1 for x in xlabels.values()], minor=False)\n",
    "        di = {u'no driver': u'no\\ndriver', u'/': u'',\n",
    "              u'\\nRNR2': u'RNR2\\nRNR2', u'\\nRNR4': u'RNR4\\nRNR4',\n",
    "              u'\\nno driver': u'no\\ndriver', u'no driver\\nno driver': u'no\\ndriver',\n",
    "              u'\\nFPR1': u'FPR1\\nFPR1', u'\\nTOR1': u'TOR1\\nTOR1'}\n",
    "        xlabels = ['\\n'.join(k) if isinstance(k, tuple) else k for k in xlabels.keys()]\n",
    "        xlabels = [di[x] if x in di else x for x in xlabels]\n",
    "        ax.set_xticklabels(xlabels, minor=False, fontsize=5, style='italic', va='top')\n",
    "                \n",
    "        ax.tick_params(axis='x', which='minor', size=0, pad=-30)\n",
    "        ax.tick_params(axis='x', which='major', size=0, pad=10)\n",
    "        ax.tick_params(axis='y', which='major', labelsize=6)\n",
    "        \n",
    "        if env_evo=='HU':\n",
    "            ax.set_ylim(-0.25,0.75)\n",
    "            ax.yaxis.set_major_locator( ticker.MaxNLocator(nbins = 4) )\n",
    "            ax.yaxis.set_minor_locator( ticker.MaxNLocator(nbins = 4) )\n",
    "        elif env_evo=='RM':\n",
    "            ax.set_ylim(-0.5,2.0)\n",
    "            ax.yaxis.set_major_locator( ticker.MaxNLocator(nbins = 5) )\n",
    "            ax.yaxis.set_minor_locator( ticker.MaxNLocator(nbins = 5) )\n",
    "\n",
    "### Heatmaps ###\n",
    "for (ii, (env_evo, env_test)) in enumerate([('HU','HU'),('RM','RM')]):\n",
    "    \n",
    "    # Group by evolution and test environment\n",
    "    S = groups_sp.get_group((env_evo, env_test))\n",
    "    H = groups_hy.get_group((env_evo, env_test))\n",
    "    \n",
    "    # Apply sorting ranks to reorder rows\n",
    "    S = sort_spores(S, env_evo)\n",
    "    H = sort_hybrids(H, env_evo)\n",
    "\n",
    "    ax = plt.subplot(gs[(env_evo,'heatmap')][:])\n",
    "        \n",
    "    ax.text(-0.15, 1.175,  chr(3*ii + ord('C')), transform=ax.transAxes,\n",
    "             fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "    \n",
    "    ax.text(0.5, 1.175,  \n",
    "            'Selection: %s\\nMeasurement: %s' %\\\n",
    "            (config.selection['long_label'][env_evo], config.environment['long_label'][env_test]), \n",
    "            transform=ax.transAxes,\n",
    "            fontsize=6, va='center', ha='center')\n",
    "    \n",
    "    # Spores\n",
    "    S = S.groupby([u'mating',u'group',u'background',u'genotype_short',u'gene'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    S_ref = S.ix[:,'ancestral'].mean()\n",
    "    S = S-S_ref\n",
    "    \n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    vmax = max(S.min(), S.max())\n",
    "    vmin = -vmax\n",
    "\n",
    "    plot.heatmap_spores(\n",
    "        S, ax, title, \n",
    "        xlabel, ylabel, xticklabels, yticklabels, \n",
    "        cmap=plt.cm.RdBu_r, vmin=vmin, vmax=vmax, radius=0.25\n",
    "    )\n",
    "    \n",
    "    # Hybrids    \n",
    "    H = H.groupby([u'group_MATa',u'background_MATa',u'gene_MATa',u'genotype_short_MATa',\n",
    "                   u'group_MATα',u'background_MATα',u'gene_MATα',u'genotype_short_MATα'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    H = H.unstack(level=[u'group_MATα',u'background_MATα',u'gene_MATα',u'genotype_short_MATα'])\n",
    "    H_ref = H.loc['ancestral','ancestral'].loc['WAxNA','WAxNA'].values.squeeze()\n",
    "    H = H-H_ref\n",
    "\n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    vmax = max(np.nanmin(H), np.nanmax(H), key=abs)#H.values.max()\n",
    "    vmin = -vmax\n",
    "    legend_title = r'$\\langle\\lambda^{td}\\rangle_{a\\alpha}^b$'\n",
    "    \n",
    "    plot.heatmap_hybrids(\n",
    "        H, ax, title, \n",
    "        xlabel, ylabel, xticklabels, yticklabels, \n",
    "        fold=False, legend_title=legend_title, cmap=plt.cm.RdBu_r, \n",
    "        vmin=vmin, vmax=vmax, pad=30\n",
    "    )\n",
    "    \n",
    "    ax.xaxis.set_ticks_position('top')\n",
    "    ax.yaxis.set_ticks_position('left')\n",
    "    \n",
    "    ax.annotate(r'$\\mathbf{a}$', xy=(-0.1, 1.01), xycoords=('axes fraction', 'axes fraction'),\n",
    "                 ha='center', va='bottom', annotation_clip=False, weight='bold', fontsize=8)\n",
    "    ax.annotate(r'$\\mathbf{\\alpha}$', xy=(-0.01, 1.1), xycoords=('axes fraction', 'axes fraction'),\n",
    "                 ha='right', va='center', annotation_clip=False, weight='bold', fontsize=8)\n",
    "    \n",
    "    ## Gene\n",
    "    # labels\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax.annotate(k, xy=(v+0.5, 1.12), xycoords=(\"data\", \"axes fraction\"),\n",
    "                    ha='center', va='top', annotation_clip=False, \n",
    "                    style=('italic' if k!='no driver' else 'normal'), fontsize=5)\n",
    "    ystart, yend, ylabels = plot.set_custom_labels(H.index, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax.annotate(k, xy=(-0.12, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                    ha='left', va='center', annotation_clip=False, \n",
    "                    style=('italic' if k!='no driver' else 'normal'), fontsize=5, rotation=90)\n",
    "    \n",
    "    ## Background\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, 0)\n",
    "    ax.set_xticks([x+0.5 for x in xlabels.values()], minor=False)\n",
    "    ax.set_xticklabels(xlabels.keys(), minor=False, fontsize=6)\n",
    "    xgrid=[xst+1. for xst in list(set(xstart.values()))]\n",
    "    [ax.axvline(g, lw=1, ls=\"-\", color=\"black\") for g in xgrid]\n",
    "    \n",
    "    ystart, yend, ylabels= plot.set_custom_labels(H.columns, 0)\n",
    "    ax.set_yticks([y+0.5 for y in ylabels.values()], minor=False)\n",
    "    ax.set_yticklabels(ylabels.keys(), minor=False, fontsize=6, rotation=90)\n",
    "    ygrid=[yst+1. for yst in list(set(ystart.values()))]\n",
    "    [ax.axhline(g, lw=1, ls=\"-\", color=\"black\") for g in ygrid]\n",
    "\n",
    "    for x in ([-1]+xstart.values()):\n",
    "        line = lines.Line2D([x+1,x+1], [0.,-.5], color='k', lw=1)\n",
    "        ax.add_line(line)\n",
    "        line.set_clip_on(False)\n",
    "    for y in ([-1]+ystart.values()):\n",
    "        line = lines.Line2D([0.,-.5], [y+1,y+1], color='k', lw=1)\n",
    "        ax.add_line(line)\n",
    "        line.set_clip_on(False)\n",
    "    \n",
    "    ax.tick_params(axis='both', which='minor', size=0, pad=10)\n",
    "    ax.tick_params(axis='both', which='major', size=0, pad=18)\n",
    "\n",
    "### Variance components ###\n",
    "var_comp = lmm_var_comp.set_index(['selection','environment','type'])\n",
    "var_comp = var_comp.sort_index(level=['selection','environment','type'])\n",
    "var_comp = var_comp.sort_index(ascending=[False,False,True])\n",
    "\n",
    "# Fix columns\n",
    "var_comp = var_comp.rename(\n",
    "    columns={\n",
    "        'background (genotype)':'Background\\n(genotype)',\n",
    "        'de novo (genotype)':'De novo\\n(genotype)',\n",
    "        'time':'Time of\\nsampling',\n",
    "        'auxotrophy':'Auxotrophy'\n",
    "    }\n",
    ")\n",
    "\n",
    "axes = {}\n",
    "\n",
    "for ii, (env_evo, ge) in enumerate(var_comp.groupby(level='selection')):\n",
    "        \n",
    "    if env_evo=='HU':\n",
    "        axes[env_evo] = plt.subplot(gs['variance'][ii])\n",
    "        axes[env_evo].text(-0.1, 1.35,  chr(ord('D')), transform=axes[env_evo].transAxes,\n",
    "                           fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "    else:\n",
    "        axes[env_evo] = plt.subplot(gs['variance'][ii], sharex=axes['HU'])\n",
    "        \n",
    "    ax = axes[env_evo]\n",
    "    utils.simple_axes(ax)\n",
    "    \n",
    "    ge.plot(\n",
    "        ax=ax, kind='barh', stacked=True, \n",
    "        color=[config.factors['color'][x] for x in ge.columns], edgecolor='w',\n",
    "        align='center', legend=False, width=0.75, rasterized=True\n",
    "    )\n",
    "    \n",
    "    bars = ax.patches\n",
    "    hatches = [config.factors['hatch'][x] for x in ge.index.get_level_values('type')]*len(ge)\n",
    "\n",
    "    for bar, hatch in zip(bars, hatches):\n",
    "        bar.set_hatch(hatch)\n",
    "    \n",
    "    # x-axis\n",
    "    ax.set_xlabel('Variance (%)', fontsize=6, labelpad=2, visible=True)\n",
    "    xscale = 1e-2\n",
    "    xticks = ticker.FuncFormatter(lambda x, pos: '{0:g}'.format(x/xscale))\n",
    "    ax.xaxis.set_major_formatter(xticks)\n",
    "    # y-axis\n",
    "    ax.set_ylabel('')\n",
    "    ystart, yend, ylabels = plot.set_custom_labels(ge.index, 1)\n",
    "    ax.set_yticks(ylabels.values())\n",
    "    ax.set_yticklabels(ylabels.keys(), fontsize=6)\n",
    "    \n",
    "    # Annotate panels and axis\n",
    "    ax.annotate(env_evo, xy=(1, 0.05), xycoords=('axes fraction', 'axes fraction'),\n",
    "                ha='right', va='bottom', fontsize=6)\n",
    "    \n",
    "    ax.tick_params(axis='both', which='major', labelsize=6, size=2)\n",
    "    \n",
    "# Fix x-axis labels (missing due to boxplot styling)\n",
    "axes['RM'].xaxis.set_visible(True)\n",
    "plt.setp(axes['RM'].get_xticklabels(), visible=True)\n",
    "\n",
    "leg1 = ax.legend(bbox_to_anchor=(0.5, -0.65), ncol=4,\n",
    "                 frameon=False, loc='lower center',\n",
    "                 borderaxespad=0, handlelength=0.75, prop={'size':6})\n",
    "    \n",
    "s_artist = patches.Rectangle((0,0), width=1, height=1, \n",
    "                             facecolor='w', edgecolor='k')\n",
    "h_artist = patches.Rectangle((0,0), width=1, height=1, hatch='///', \n",
    "                             facecolor='w', edgecolor='k')\n",
    "\n",
    "leg2 = axes['HU'].legend([s_artist,h_artist], ['Spores','Hybrids'],\n",
    "                         bbox_to_anchor=(0.5, 1.2), ncol=2, \n",
    "                         frameon=False, loc='upper center',\n",
    "                         borderaxespad=0, handlelength=0.75, prop={'size':6})\n",
    "\n",
    "ax.add_artist(leg1)\n",
    "\n",
    "plot.save_figure(dir_paper+'figures/figure6/figure6')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fig. 6:** Fitness contribution of genetic background and *de novo* mutations. (**A**) To determine fitness effects of background variation and de novo mutations in hydroxyurea (*RNR2*, *RNR4*) and rapamycin (*FPR1*, *TOR1*), we isolated individuals from ancestral and evolved populations. From these diploid cells, we sporulated and selected haploid segregants of each mating type. Spores with mutations in *RNR2*, *RNR4* and *TOR1* were genotyped to test if they carry the wild-type or mutated allele. We crossed the *MAT*a and *MAT*α versions to create hybrids (48x48 in hydroxyurea and 56x56 in rapamycin). Independent segregants were used to measure biological variability of ancestral and evolved backgrounds. All panels follow this legend. (**B**, **E**) Relative growth rate, $\\lambda$, measured with respect to the ancestral population for multiple combinations (background genotype, $b$; *de novo* genotype, $d$; time of sampling during the selection phase, $t$; auxotrophy, $x$) and averaged over measurement replicates. Medians and 25%/75% percentiles across groups are shown, with medians as horizontal black lines and colored by de novo genotype [wild-type +/+ (blue); heterozygote +/− (cyan); homozygote −/− (green)]. Outliers (circles) and isolated, selected clones with matching genotypes (diamonds) are highlighted. (**D**) Variance decomposition of the growth rate of spores (solid) and hybrids (hatched) that can be attributed to different components using a linear mixed model. Estimates of variance components are obtained by restricted maximum likelihood (Fig. S12 and Table S6). (**C**, **F**) Ensemble average of the growth rate of spores  and hybrids . An extended version of the figure with all combinations and controls can be found in Figs. S11 and S12, respectively."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure S10 - Background-averaged fitness effects (full matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "param='norm_growth_rate'\n",
    "\n",
    "fig = plt.figure(figsize=(8, 8))\n",
    "\n",
    "grid = gridspec.GridSpec(2, 2, hspace=0.35, wspace=0.25)\n",
    "\n",
    "gs = {}\n",
    "gs[('HU','HU')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,0])\n",
    "gs[('HU','SC')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[0,1])\n",
    "gs[('RM','RM')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,0])\n",
    "gs[('RM','SC')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,1])\n",
    "\n",
    "groups_sp = spores_df.groupby(['selection','environment'])\n",
    "groups_hy = hybrids_df.groupby(['selection','environment'])\n",
    "groups_cl = clone_df.groupby(['selection','environment'])\n",
    "\n",
    "for (ii, (env_evo, env_test)) in enumerate([('HU','HU'),('HU','SC'),('RM','RM'),('RM','SC')]):\n",
    "    \n",
    "    # Group by evolution and test environment\n",
    "    S = groups_sp.get_group((env_evo, env_test))\n",
    "    H = groups_hy.get_group((env_evo, env_test))\n",
    "    C = groups_cl.get_group((env_evo, env_test))\n",
    "    \n",
    "    # Apply sorting ranks to reorder rows\n",
    "    S = sort_spores(S, env_evo)\n",
    "    H = sort_hybrids(H, env_evo)\n",
    "\n",
    "    ### Heatmap ###\n",
    "    ax = plt.subplot(gs[(env_evo, env_test)][0])\n",
    "    \n",
    "    ax.text(-0.115, 1.2,  chr(ii + ord('A')), transform=ax.transAxes,\n",
    "            fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "    \n",
    "    ax.text(0.5, 1.2,  \n",
    "            'Selection: %s\\nMeasurement: %s' %\\\n",
    "            (config.selection['long_label'][env_evo], config.environment['long_label'][env_test]), \n",
    "            transform=ax.transAxes,\n",
    "            fontsize=6, va='center', ha='center')\n",
    "    \n",
    "    # Spores\n",
    "    S = S.groupby([u'mating',u'group',u'background',u'gene',u'genotype_short',u'tetrad',u'spore'],\n",
    "                    sort=False)[param].agg(np.median).reset_index()\n",
    "    S = S.set_index('mating')[param]\n",
    "    S_ref = np.nanmean(S)\n",
    "    S = S-S_ref\n",
    "        \n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    vmax = max(S.min(), S.max())\n",
    "    vmin = -vmax\n",
    "    \n",
    "    plot.heatmap_spores(\n",
    "        S, ax, title, \n",
    "        xlabel, ylabel, xticklabels, yticklabels, \n",
    "        cmap=plt.cm.RdBu_r, vmin=vmin, vmax=vmax, radius=0.5\n",
    "    )\n",
    "    \n",
    "    # Hybrids    \n",
    "    H = H.groupby([u'group_MATa',u'background_MATa',u'gene_MATa',u'genotype_short_MATa',u'tetrad_MATa',u'spore_MATa',\n",
    "                   u'group_MATα',u'background_MATα',u'gene_MATα',u'genotype_short_MATα',u'tetrad_MATα',u'spore_MATα'],\n",
    "                    sort=False).agg(np.median)[param]\n",
    "    H = H.unstack(level=[u'group_MATα',u'background_MATα',u'gene_MATα',u'genotype_short_MATα',u'tetrad_MATα',u'spore_MATα'])\n",
    "    H_ref = np.nanmean(H.loc['ancestral','ancestral'].loc['WAxNA','WAxNA'])\n",
    "    H = H-H_ref\n",
    "\n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    vmax = max(np.nanmin(H), np.nanmax(H), key=abs)\n",
    "    vmin = -vmax\n",
    "    legend_title = r'$\\lambda^{btd}_{a\\alpha}$'\n",
    "    \n",
    "    plot.heatmap_hybrids(\n",
    "        H, ax, title,\n",
    "        xlabel, ylabel, xticklabels, yticklabels, \n",
    "        fold=False, legend_title=legend_title, cmap=plt.cm.RdBu_r, \n",
    "        vmin=vmin, vmax=vmax, pad=30\n",
    "    )\n",
    "    \n",
    "    print 'Spores\\nSelection: %s, Environment: %s, Pass: %4d, Fail: %4d' \\\n",
    "    % (env_evo, env_test, S.notnull().values.flatten().sum(), S.isnull().values.flatten().sum())\n",
    "    print 'Hybrids\\nSelection: %s, Environment: %s, Pass: %4d, Fail: %4d\\n' \\\n",
    "    % (env_evo, env_test, H.notnull().values.flatten().sum(), H.isnull().values.flatten().sum())\n",
    "    \n",
    "    ax.xaxis.set_ticks_position('top')\n",
    "    ax.yaxis.set_ticks_position('left')\n",
    "    \n",
    "    ax.annotate(r'$\\mathbf{a}$', xy=(-0.075, 1.01), xycoords=('axes fraction', 'axes fraction'),\n",
    "                 ha='center', va='bottom', annotation_clip=False, weight='bold', fontsize=8)\n",
    "    ax.annotate(r'$\\mathbf{\\alpha}$', xy=(-0.01, 1.075), xycoords=('axes fraction', 'axes fraction'),\n",
    "                 ha='right', va='center', annotation_clip=False, weight='bold', fontsize=8)\n",
    "    \n",
    "    ## Gene + genotype\n",
    "    # labels and grid\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, slice(2, 4))\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax.annotate(k[0].replace('no driver','no\\ndri.'), \n",
    "                    xy=(v+0.5, 1.05), xycoords=(\"data\", \"axes fraction\"),\n",
    "                    ha='center', va='bottom', annotation_clip=False, \n",
    "                    style=('italic' if k[0]!='no driver' else 'normal'), fontsize=6)\n",
    "    xgrid=[xst+1. for xst in list(set(xstart.values()))]\n",
    "    [ax.axvline(g, lw=.75, ls=\"--\", dashes=(5, 2), color=\"gray\") for g in xgrid]\n",
    "    for x in ([-1]+xstart.values()):\n",
    "        line = lines.Line2D([x+1,x+1], [0.,-3], color='gray', ls='-', lw=.75)\n",
    "        ax.add_line(line)\n",
    "        line.set_clip_on(False)\n",
    "    \n",
    "    ystart, yend, ylabels = plot.set_custom_labels(H.index, slice(2, 4))\n",
    "    for k, v in ylabels.iteritems():\n",
    "        ax.annotate(k[0].replace('no driver','no\\ndri.'), \n",
    "                    xy=(-0.05, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                    ha='right', va='center', annotation_clip=False,  \n",
    "                    style=('italic' if k[0]!='no driver' else 'normal'), fontsize=6, rotation=90)\n",
    "    ygrid=[yst+1. for yst in list(set(ystart.values()))]\n",
    "    [ax.axhline(g, lw=.75, ls=\"--\", dashes=(5, 2), color=\"gray\") for g in ygrid]\n",
    "    for y in ([-1]+ystart.values()):\n",
    "        line = lines.Line2D([0.,-3], [y+1,y+1], color='gray', ls='-', lw=.75)\n",
    "        ax.add_line(line)\n",
    "        line.set_clip_on(False)\n",
    "    \n",
    "    ## Background\n",
    "    # labels and grid\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, 0)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax.annotate(config.population['short_label'][k], \n",
    "                    xy=(v+0.5, config.population['pad'][k]+1), xycoords=(\"data\", \"axes fraction\"),\n",
    "                    ha='center', va='bottom', annotation_clip=False, fontsize=6)\n",
    "    xgrid=[xst+1. for xst in list(set(xstart.values()))]\n",
    "    [ax.axvline(g, lw=.75, ls=\"-\", color=\"k\") for g in xgrid]\n",
    "    for x in ([-1]+xstart.values()):\n",
    "        line = lines.Line2D([x+1,x+1], [0.,-4], color='k', lw=1)\n",
    "        ax.add_line(line)\n",
    "        line.set_clip_on(False)\n",
    "    \n",
    "    ystart, yend, ylabels = plot.set_custom_labels(H.index, 0)\n",
    "    for k, v in ylabels.iteritems():\n",
    "        ax.annotate(config.population['short_label'][k], \n",
    "                    xy=(-config.population['pad'][k], v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                    ha='right', va='center', annotation_clip=False, fontsize=6, rotation=90)\n",
    "    ygrid=[yst+1. for yst in list(set(ystart.values()))]\n",
    "    [ax.axhline(g, lw=.75, ls=\"-\", color=\"k\") for g in ygrid]\n",
    "    for y in ([-1]+ystart.values()):\n",
    "        line = lines.Line2D([0.,-4], [y+1,y+1], color='k', lw=1)\n",
    "        ax.add_line(line)\n",
    "        line.set_clip_on(False)\n",
    "    \n",
    "    ax.tick_params(axis='both', which='minor', size=0, pad=2)\n",
    "    ax.tick_params(axis='both', which='major', size=0, pad=15)\n",
    "\n",
    "plot.save_figure(dir_supp+'figures/supp_figure_pheno_cross/supp_figure_pheno_cross_extended')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fig. S10**: Fitness contribution of genetic background and *de novo* mutations. Given an ensemble of $n_b$ unique genetic backgrounds (${n_b\\,{=}\\,48}$ in hydroxyurea and ${n_b\\,{=}\\,56}$ in rapamycin), we constructed a matrix of size $n_b \\times n_b$ where every unique haploid background is crossed against itself and all other haploid backgrounds, and the two must be of opposite mating type (*MAT*a or *MAT*α). Each matrix element is labeled by background genotype, $b$; *de novo* genotype, $d$; time, $t$; and auxotrophy, $x$. Measurements of relative growth rates of spores $\\lambda^{btd}_{\\{a,\\alpha\\}}$ and hybrids $\\lambda^{btd}_{a\\alpha}$ are shown, normalized with respect to the ancestral WAxNA crosses. Measurements are taken in (**A**) SC+HU 10 $\\text{mg ml}^{-1}$ and (**B**) SC; (**C**) SC+RM 0.025 $\\mu\\text{g ml}^{-1}$ and (**D**) SC, respectively. The color scale for all matrices to the right of each panel indicates the relative fold-change with respect to the ancestral WAxNA crosses. White boxes indicate missing data due to mating inefficiency and slow growth."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure S11- Background-averaged fitness effects (control)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "param='norm_growth_rate'\n",
    "\n",
    "fig = plt.figure(figsize=(9.5, 6))\n",
    "\n",
    "grid = gridspec.GridSpec(2, 2, width_ratios=[3,1], hspace=0.4, wspace=0.2)\n",
    "\n",
    "gs = {}\n",
    "gs[('HU','barplot')] = gridspec.GridSpecFromSubplotSpec(1, 24, subplot_spec=grid[0,0], wspace=0)\n",
    "gs[('HU','heatmap')] = gridspec.GridSpecFromSubplotSpec(1, 1,subplot_spec=grid[0,1])\n",
    "gs[('RM','barplot')] = gridspec.GridSpecFromSubplotSpec(1, 24, subplot_spec=grid[1,0], wspace=0)\n",
    "gs[('RM','heatmap')] = gridspec.GridSpecFromSubplotSpec(1, 1, subplot_spec=grid[1,1])\n",
    "\n",
    "### Box plots ###\n",
    "groups_sp = spores_df.groupby(['selection','environment'])\n",
    "groups_hy = hybrids_df.groupby(['selection','environment'])\n",
    "groups_cl = clone_df.groupby(['selection','environment'])\n",
    "\n",
    "for (ii, (env_evo, env_test)) in enumerate([('HU','SC'),('RM','SC')]):\n",
    "    \n",
    "    # Group by evolution and test environment\n",
    "    S = groups_sp.get_group((env_evo, env_test))\n",
    "    H = groups_hy.get_group((env_evo, env_test))\n",
    "    C = groups_cl.get_group((env_evo, env_test))\n",
    "    \n",
    "    # Filter by dictionary\n",
    "    S = filter_spores(S, env_evo)\n",
    "    H = filter_hybrids(H, env_evo)\n",
    "    \n",
    "    # Apply sorting ranks to reorder rows\n",
    "    S = sort_spores(S, env_evo)\n",
    "    H = sort_hybrids(H, env_evo)\n",
    "    \n",
    "    ### Spores barplot ###\n",
    "    ax1 = plt.subplot(gs[(env_evo,'barplot')][:7])\n",
    "\n",
    "    ax1.text(-0.2, 1.15,  chr(2*ii + ord('A')), transform=ax1.transAxes,\n",
    "             fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "    \n",
    "    ax1.text(0, 1.15,  \n",
    "             'Selection: %s\\nMeasurement: %s' %\\\n",
    "             (config.selection['long_label'][env_evo], config.environment['long_label'][env_test]), \n",
    "             transform=ax1.transAxes,\n",
    "             fontsize=6, va='center', ha='left')\n",
    "    \n",
    "    S = S.groupby(['group','background','gene','genotype_short','tetrad','spore'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['group','background','gene','genotype_short'])\n",
    "    S_ref = S.ix[:,'ancestral'].mean().values[0]\n",
    "    S = S-S_ref\n",
    "    \n",
    "    bp = S.plot(\n",
    "        ax=ax1, kind='box', by='group', return_type='dict',\n",
    "        labels=S.columns.get_level_values('group')\n",
    "    )\n",
    "\n",
    "    plot.boxplot_custom(bp, ax1,\n",
    "                        colors=[config.spores_gt_short['color'][env_evo][x] for x in S.columns.get_level_values('genotype_short')],\n",
    "                        hatches=[' ']*10)\n",
    "    \n",
    "    ax1.set_title('Spores (haploid)', fontsize=6, weight='bold')\n",
    "    ax1.set_xlabel('')\n",
    "    ax1.set_ylabel(r'Rel. growth rate, $\\lambda^{btd}_{a\\alpha}$', fontsize=6, labelpad=2)\n",
    "\n",
    "    ### Hybrids barplot ###\n",
    "    ax2 = plt.subplot(gs[(env_evo,'barplot')][7:], sharey=ax1)\n",
    "    \n",
    "    H = H.groupby(['group','background','gene','genotype_short','tetrad','spore'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\\\n",
    "    .unstack(level=['group','background','gene','genotype_short'])\n",
    "    ix_use = pd.MultiIndex.from_tuples([[('ancestral', 'ancestral')]], names=['group'])\n",
    "    H_ref = H.loc[:,H.columns.get_level_values('group').isin(ix_use.get_level_values(0))].mean().squeeze()\n",
    "    H = H-H_ref\n",
    "    \n",
    "    C = C.groupby(['group','background','gene','genotype_short','isolate'], sort=False)[param]\\\n",
    "    .agg([np.mean, np.median, np.std, 'count'])['mean']\n",
    "    C_ref = C[('ancestral', 'ancestral')].mean()\n",
    "\n",
    "    C = C-C_ref\n",
    "    C = C.unstack(level=['group','background','gene','genotype_short'])\n",
    "    C = C.loc[:,(C.columns.get_level_values('group').isin([('evolved','evolved')])\\\n",
    "                 &~(C.columns.get_level_values('gene').isin([('','')])))]\n",
    "    C = C.reindex(columns=H.columns)\n",
    "    C = C.dropna(how='all')\n",
    "\n",
    "    bp = H.plot(\n",
    "        ax=ax2, kind='box', by='group', return_type='dict', \n",
    "        labels=H.columns.get_level_values('group')\n",
    "    )\n",
    "\n",
    "    plot.boxplot_custom(bp, ax2, \n",
    "                        colors=[config.hybrids_gt_short['color'][env_evo][x] for x in H.columns.get_level_values('genotype_short')],\n",
    "                        hatches=[' ']*30)\n",
    "\n",
    "    for i,d in enumerate(C):\n",
    "        y = C[d]\n",
    "        x = [i+1]*len(y)\n",
    "        ax2.plot(x, y, \n",
    "                 mfc=config.hybrids_gt_short['color'][env_evo][d[-1]], mec='k', \n",
    "                 ms=3, marker='D', linestyle=\"None\", zorder=6)\n",
    "    \n",
    "    ax2.set_title('Hybrids (diploid)', fontsize=6, weight='bold')\n",
    "    ax2.set_xlabel('')\n",
    "    ax2.set_ylabel('', fontsize=6)\n",
    "    \n",
    "    wt_artist = patches.Rectangle((0,0), width=1, height=1, color=config.hybrids_gt_short['color'][env_evo][('+','+')])\n",
    "    het_artist = patches.Rectangle((0,0), width=1, height=1, color=config.hybrids_gt_short['color'][env_evo][('+','-')])\n",
    "    hom_artist = patches.Rectangle((0,0), width=1, height=1, color=config.hybrids_gt_short['color'][env_evo][('-','-')])\n",
    "    \n",
    "    wt_artist = lines.Line2D((0,.75),(0,0), mec='k', mfc=config.hybrids_gt_short['color'][env_evo][('+','+')], marker='s', markersize=4, linestyle='')\n",
    "    het_artist = lines.Line2D((0,.75),(0,0), mec='k', mfc=config.hybrids_gt_short['color'][env_evo][('+','-')], marker='s', markersize=4, linestyle='')\n",
    "    hom_artist = lines.Line2D((0,.75),(0,0), mec='k', mfc=config.hybrids_gt_short['color'][env_evo][('-','-')], marker='s', markersize=4, linestyle='')\n",
    "    clone_artist = lines.Line2D((0,1),(0,0), mec='k', mfc='w', marker='D', markersize=3, linestyle='')\n",
    "\n",
    "    leg = ax2.legend([wt_artist,het_artist,hom_artist,clone_artist], \n",
    "                     ['WT','het.','hom.','clone'], ncol=1, \n",
    "                     frameon=False, loc='upper right',\n",
    "                     borderaxespad=0, handlelength=0.75, \n",
    "                     prop={'size':5}, labelspacing=.32)\n",
    "    \n",
    "    for ax, g in zip([ax1,ax2],[S,H]):\n",
    "        ### Horizontal ###\n",
    "        ax.yaxis.grid(ls=\"-\", lw=.8, color=\"0.9\", zorder=-1)\n",
    "        ax.axhline(y=0., c='k', ls=\"--\", dashes=(7, 7), lw=.8, zorder=1)\n",
    "        ax.set_axisbelow(True)\n",
    "        \n",
    "        ### Bertical ###\n",
    "        ## Background\n",
    "        xstart, xend, xlabels = plot.set_custom_labels(g.columns, 0)\n",
    "        # labels\n",
    "        for k, v in xlabels.iteritems():\n",
    "            ax.annotate('\\nx\\n'.join(k) if isinstance(k, tuple) else k, \n",
    "                        xy=(v+1, 0.97), xycoords=(\"data\", \"axes fraction\"),\n",
    "                        ha='center', va='top', annotation_clip=False, fontsize=5)\n",
    "        # grid\n",
    "        xgrid=[xst+1.5 for xst in list(set(xstart.values()))]\n",
    "        [ax.axvline(x, lw=1.0, ls='-', color=\"0.9\", zorder=0) for x in xgrid]\n",
    "\n",
    "        ## Gene\n",
    "        xend, xstart, xlabels = plot.set_custom_labels(g.columns, 2)\n",
    "        \n",
    "        transform = transforms.blended_transform_factory(ax.transData, ax.transAxes)\n",
    "        for k in xlabels:\n",
    "            if (k!='') & (k!=('','')):\n",
    "                if abs(xstart[k]-xend[k])>0:\n",
    "                    line = lines.Line2D([xstart[k]+.75,xend[k]+1.25], [-.05,-.05], \n",
    "                                        color='k', lw=1, transform=transform)\n",
    "                else:\n",
    "                    line = lines.Line2D([xstart[k]+.75,xend[k]+1.25], [-.05,-.05], \n",
    "                                        color='k', lw=1, transform=transform)\n",
    "                ax.add_line(line)\n",
    "                line.set_clip_on(False)\n",
    "        \n",
    "        ax.set_xticks([x+1 for x in xlabels.values()], minor=False)\n",
    "        di = {u'no driver': u'no\\ndriver', u'/': u'',\n",
    "              u'\\nRNR2': u'RNR2\\nRNR2', u'\\nRNR4': u'RNR4\\nRNR4',\n",
    "              u'\\nno driver': u'no\\ndriver', u'no driver\\nno driver': u'no\\ndriver',\n",
    "              u'\\nFPR1': u'FPR1\\nFPR1', u'\\nTOR1': u'TOR1\\nTOR1'}\n",
    "        xlabels = ['\\n'.join(k) if isinstance(k, tuple) else k for k in xlabels.keys()]\n",
    "        xlabels = [di[x] if x in di else x for x in xlabels]\n",
    "        ax.set_xticklabels(xlabels, minor=False, fontsize=5, style='italic', va='top')\n",
    "                \n",
    "        ax.tick_params(axis='x', which='minor', size=0, pad=-30)\n",
    "        ax.tick_params(axis='x', which='major', size=0, pad=10)\n",
    "        ax.tick_params(axis='y', which='major', labelsize=6)\n",
    "        \n",
    "        if env_evo=='HU':\n",
    "            ax.set_ylim(-0.25,0.5)\n",
    "            ax.yaxis.set_major_locator( ticker.MaxNLocator(nbins = 3) )\n",
    "            ax.yaxis.set_minor_locator( ticker.MaxNLocator(nbins = 3) )\n",
    "        elif env_evo=='RM':\n",
    "            ax.set_ylim(-0.5,0.5)\n",
    "            ax.yaxis.set_major_locator( ticker.MaxNLocator(nbins = 4) )\n",
    "            ax.yaxis.set_minor_locator( ticker.MaxNLocator(nbins = 4) )\n",
    "\n",
    "### Heatmaps ###\n",
    "for (ii, (env_evo, env_test)) in enumerate([('HU','SC'),('RM','SC')]):\n",
    "    \n",
    "    # Group by evolution and test environment\n",
    "    S = groups_sp.get_group((env_evo, env_test))\n",
    "    H = groups_hy.get_group((env_evo, env_test))\n",
    "    \n",
    "    # Apply sorting ranks to reorder rows\n",
    "    S = sort_spores(S, env_evo)\n",
    "    H = sort_hybrids(H, env_evo)\n",
    "\n",
    "    ax = plt.subplot(gs[(env_evo,'heatmap')][:])\n",
    "        \n",
    "    ax.text(-0.15, 1.21,  chr(2*ii + ord('B')), transform=ax.transAxes,\n",
    "             fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "    \n",
    "    ax.text(0.5, 1.21,  \n",
    "            'Selection: %s\\nMeasurement: %s' %\\\n",
    "            (config.selection['long_label'][env_evo], config.environment['long_label'][env_test]), \n",
    "            transform=ax.transAxes,\n",
    "            fontsize=6, va='center', ha='center')\n",
    "    \n",
    "    # Spores\n",
    "    S = S.groupby([u'mating',u'group',u'background',u'genotype_short',u'gene'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    S_ref = S.ix[:,'ancestral'].mean()\n",
    "    S = S-S_ref\n",
    "    \n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    vmax = max(S.min(), S.max())\n",
    "    vmin = -vmax\n",
    "\n",
    "    plot.heatmap_spores(S, ax, title, \n",
    "                        xlabel, ylabel, xticklabels, yticklabels, \n",
    "                        cmap=plt.cm.RdBu_r, vmin=vmin, vmax=vmax, radius=0.25)\n",
    "    \n",
    "    # Hybrids    \n",
    "    H = H.groupby([u'group_MATa',u'background_MATa',u'gene_MATa',u'genotype_short_MATa',\n",
    "                   u'group_MATα',u'background_MATα',u'gene_MATα',u'genotype_short_MATα'],\n",
    "                    sort=False).agg(np.mean)[param]\n",
    "    H = H.unstack(level=[u'group_MATα',u'background_MATα',u'gene_MATα',u'genotype_short_MATα'])\n",
    "    H_ref = H.loc['ancestral','ancestral'].loc['WAxNA','WAxNA'].values.squeeze()\n",
    "    H = H-H_ref\n",
    "\n",
    "    title = ''\n",
    "    xlabel= ''\n",
    "    ylabel= ''\n",
    "    xticklabels = []\n",
    "    yticklabels = []\n",
    "    vmax = max(np.nanmin(H), np.nanmax(H), key=abs)\n",
    "    vmin = -vmax\n",
    "    legend_title = r'$\\langle \\lambda^{td}\\rangle_{a\\alpha}^b$'\n",
    "    \n",
    "    plot.heatmap_hybrids(H, ax, title,\n",
    "                         xlabel, ylabel, xticklabels, yticklabels, \n",
    "                         fold=False, legend_title=legend_title, cmap=plt.cm.RdBu_r, \n",
    "                         vmin=vmin, vmax=vmax, pad=30)\n",
    "    \n",
    "    ax.xaxis.set_ticks_position('top')\n",
    "    ax.yaxis.set_ticks_position('left')\n",
    "    \n",
    "    ax.annotate(r'$\\mathbf{a}$', xy=(-0.1, 1.01), xycoords=('axes fraction', 'axes fraction'),\n",
    "                 ha='center', va='bottom', annotation_clip=False, weight='bold', fontsize=8)\n",
    "    ax.annotate(r'$\\mathbf{\\alpha}$', xy=(-0.01, 1.1), xycoords=('axes fraction', 'axes fraction'),\n",
    "                 ha='right', va='center', annotation_clip=False, weight='bold', fontsize=8)\n",
    "    \n",
    "    ## Gene\n",
    "    # labels\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax.annotate(k, xy=(v+0.5, 1.12), xycoords=(\"data\", \"axes fraction\"),\n",
    "                    ha='center', va='top', annotation_clip=False, \n",
    "                    style=('italic' if k!='no driver' else 'normal'), fontsize=5)\n",
    "    ystart, yend, ylabels = plot.set_custom_labels(H.index, 2)\n",
    "    for k, v in xlabels.iteritems():\n",
    "        ax.annotate(k, xy=(-0.12, v+0.5), xycoords=(\"axes fraction\", \"data\"),\n",
    "                    ha='left', va='center', annotation_clip=False, \n",
    "                    style=('italic' if k!='no driver' else 'normal'), fontsize=5, rotation=90)\n",
    "    \n",
    "    ## Background\n",
    "    # labels and grid\n",
    "    xstart, xend, xlabels = plot.set_custom_labels(H.columns, 0)\n",
    "    ax.set_xticks([x+0.5 for x in xlabels.values()], minor=False)\n",
    "    ax.set_xticklabels(xlabels.keys(), minor=False, fontsize=6)\n",
    "    xgrid=[xst+1. for xst in list(set(xstart.values()))]\n",
    "    [ax.axvline(g, lw=1, ls=\"-\", color=\"black\") for g in xgrid]\n",
    "    \n",
    "    ystart, yend, ylabels= plot.set_custom_labels(H.columns, 0)\n",
    "    ax.set_yticks([y+0.5 for y in ylabels.values()], minor=False)\n",
    "    ax.set_yticklabels(ylabels.keys(), minor=False, fontsize=6, rotation=90)\n",
    "    ygrid=[yst+1. for yst in list(set(ystart.values()))]\n",
    "    [ax.axhline(g, lw=1, ls=\"-\", color=\"black\") for g in ygrid]\n",
    "\n",
    "    for x in ([-1]+xstart.values()):\n",
    "        line = lines.Line2D([x+1,x+1], [0.,-.5], color='k', lw=1)\n",
    "        ax.add_line(line)\n",
    "        line.set_clip_on(False)\n",
    "    for y in ([-1]+ystart.values()):\n",
    "        line = lines.Line2D([0.,-.5], [y+1,y+1], color='k', lw=1)\n",
    "        ax.add_line(line)\n",
    "        line.set_clip_on(False)\n",
    "    \n",
    "    ax.tick_params(axis='both', which='minor', size=0, pad=10)\n",
    "    ax.tick_params(axis='both', which='major', size=0, pad=18)\n",
    "\n",
    "plot.save_figure(dir_supp+'figures/supp_figure_pheno_cross/supp_figure_pheno_cross_reduced')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fig. S11:** Ensemble average of fitness effects over genetic backgrounds. (**A**, **C**) Relative growth rate, $\\lambda$, measured in a control environment (SC) for cells selected in (A) hydroxyurea and (C) rapamycin. Measurements are normalized with respect to the ancestral population for multiple combinations (background genotype, $b$; *de novo* genotype, $d$; sampling time during selection, $t$; auxotrophy, $x$) and averaged over measurement replicates. Medians and 25\\%/75\\% percentiles across groups are shown, with medians as horizontal lines and colored by *de novo* genotype [wild-type +/+ (blue); heterozygote +/- (cyan); homozygote -/- (green)]. Outliers (circles) and isolated, selected clones with matching genotypes (diamonds) are highlighted. (**B**, **D**) Ensemble average of the growth rate of spores, $\\langle \\lambda^{td} \\rangle^b_{\\{a,\\alpha\\}}$, and hybrids, $\\langle \\lambda^{td} \\rangle^b_{a\\alpha}$ over genetic backgrounds. The color scale for all matrices to the right of each panel indicates the relative fold-change with respect to the ancestral WAxNA crosses."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure S12 - Analysis of variance"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Linear mixed model (frequentist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import statsmodels.api as sm\n",
    "\n",
    "panels = {\n",
    "    'HU': {\n",
    "        'HU':(0,0),\n",
    "        'SC':(1,0)\n",
    "    },\n",
    "    'RM': {\n",
    "        'RM':(0,1),\n",
    "        'SC':(1,1)\n",
    "    }\n",
    "}\n",
    "\n",
    "fig = plt.figure(figsize=(8, 4))\n",
    "\n",
    "grid = gridspec.GridSpec(2, 2, hspace=0.5, wspace=0.1)\n",
    "\n",
    "gs = {}\n",
    "\n",
    "var_comp = lmm_var_comp.copy()\n",
    "var_comp['total_variance'] = var_comp.sum(axis=1)\n",
    "    \n",
    "for ii, (s, e) in enumerate([('HU','HU'),('HU','SC'),('RM','RM'),('RM','SC')]):\n",
    "    \n",
    "    gs[(s, e)] = gridspec.GridSpecFromSubplotSpec(1, 2, subplot_spec=grid[panels[s][e]], wspace=0.02)\n",
    "\n",
    "for ii, ((s, e), fit_all) in enumerate(lmm_fit.groupby(['selection','environment'], sort=False)):\n",
    "\n",
    "    for jj, (t, fit_type) in enumerate(fit_all.groupby('type', sort=False)):\n",
    "        \n",
    "        ax = plt.subplot(gs[(s,e)][jj])\n",
    "        utils.simple_axes(ax)\n",
    "        \n",
    "        if jj==0:\n",
    "            ax.text(-0.2, 1.125, chr(ii + ord('A')), transform=ax.transAxes,\n",
    "                    fontsize=9, fontweight='bold', va='center', ha='right')\n",
    "            ax.annotate('Selection: %s\\nMeasurement: %s' %\\\n",
    "                         (config.selection['long_label'][s], \\\n",
    "                          config.environment['long_label'][e]), \n",
    "                        xy=(0, 1.125), xycoords=\"axes fraction\",\n",
    "                        ha='left', va='center', annotation_clip=False, fontsize=6)\n",
    "\n",
    "        nobs = len(fit_type.observed_growth_rate)\n",
    "        y = fit_type.observed_growth_rate\n",
    "        yhat = fit_type.fitted_growth_rate\n",
    "        ymin, ymax = y.quantile([.005, .995])\n",
    "        yrange = ymax - ymin\n",
    "        \n",
    "        # Scatter\n",
    "        ax.scatter(y, yhat, facecolor='gray', s=1.5, lw=0, alpha=0.75, rasterized=True)\n",
    "        \n",
    "        # Fit\n",
    "        line_fit = sm.OLS(y, sm.add_constant(yhat, prepend=False)).fit().params\n",
    "        fit = lambda x: line_fit[1] + line_fit[0] * x \n",
    "        ax.plot(np.linspace(ymin-.1*yrange, ymax+.1*yrange, nobs),\n",
    "                fit(np.linspace(ymin-.1*yrange, ymax+.1*yrange, nobs)), \n",
    "                label=t, color='k', ls='--', dashes=(5, 2), lw=1)\n",
    "        \n",
    "        # Variance explained\n",
    "        var_total = var_comp.groupby(['selection','environment','type']).get_group((s,e,t))['total_variance'].values[0]\n",
    "        ax.annotate('{}\\n' r'$r^2$ = {:.2f}'.format(t.capitalize(), var_total),\n",
    "                    xy=(.1, .9), xycoords=\"axes fraction\",\n",
    "                    ha='left', va='center', fontsize=6)\n",
    "        \n",
    "        # x-axis\n",
    "        ax.set_xlim(ymin-.1*yrange, ymax+.1*yrange)\n",
    "        ax.xaxis.set_major_locator( ticker.MaxNLocator(nbins = 4) )\n",
    "\n",
    "        # y-axis\n",
    "        ax.set_ylim(ymin-.1*yrange, ymax+.1*yrange)\n",
    "        ax.yaxis.set_major_locator( ticker.MaxNLocator(nbins = 4) )\n",
    "        \n",
    "        ax.set_aspect('equal', adjustable='box')\n",
    "    \n",
    "        if jj==0:\n",
    "            ax.annotate(r'Rel. growth rate (observed), $\\lambda$',\n",
    "                        xy=(1.2, -0.175), xycoords=\"axes fraction\",\n",
    "                        ha='center', va='center', annotation_clip=False, fontsize=6)\n",
    "            ax.set_ylabel(r'Rel. growth rate (fitted), $\\hat{\\lambda}$', fontsize=6)\n",
    "    \n",
    "for ax in fig.get_axes():\n",
    "    ax.tick_params(axis='both', which='major', labelsize=6, size=2)\n",
    "    ax.tick_params(axis='both', which='minor', labelsize=6, size=0)\n",
    "        \n",
    "    for loc in ['top','bottom','left','right']:\n",
    "        ax.spines[loc].set_linewidth(.6)\n",
    "\n",
    "plot.save_figure(dir_supp+'figures/supp_figure_pheno_cross/supp_figure_pheno_cross_lmm')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fig. S12:** Hierarchical analysis of variance in the genetic cross using linear mixed models. We model the growth rate of spores $\\lambda_{\\{a,\\alpha\\}}^{btd}$ and hybrids $\\lambda_{a\\alpha}^{btd}$ as a function of background genotype $b$, *de novo* genotype $d$, sampling time during selection $t$, and auxotrophy $x$. Relative growth rates are accurately fitted by this model (Model 4). Model fits are summarised in Table S6. Measurements are taken in SC+HU 10 $\\text{mg ml}^{-1}$ and SC only for populations selected in hydroxyurea (**A**, **B**); SC+RM 0.025 $\\mu \\text{g ml}^{-1}$ and SC only for populations selected in rapamycin (**C**, **D**). The scatter shows a set of measurements $\\lambda$ ($x$-axis) against the fitted rates $\\hat{\\lambda}$ ($y$-axis). The total variance explained, $r^2$, is separately computed for spores and hybrids by environment."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "genetic-variation",
   "language": "python",
   "name": "genetic-variation"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  },
  "latex_metadata": {
   "affiliation": "Wellcome Trust Sanger Institute, Hinxton, Cambridge CB10 1SA, United Kingdom; Department of Applied Mathematics and Theoretical Physics, University of Cambridge, Cambridge CB3 0WA, United Kingdom",
   "author": "Ignacio Vazquez-Garcia",
   "title": "Clonal heterogeneity influences the fate of new adaptive mutations"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
