{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Supplemental Information:\n",
    "\n",
    "> **\"Clonal heterogeneity influences the fate of new adaptive mutations\"**\n",
    "\n",
    "> Ignacio Vázquez-García, Francisco Salinas, Jing Li, Andrej Fischer, Benjamin Barré, Johan Hallin, Anders Bergström, Elisa Alonso-Pérez, Jonas Warringer, Ville Mustonen, Gianni Liti\n",
    "\n",
    "## Figures S6, S7 and S8\n",
    "\n",
    "This IPython notebook is provided for reproduction of Figures S6, S7 and S8 of the paper. It can be viewed by copying its URL to nbviewer and it can be run by opening it in binder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ivg/.virtualenvs/genetic-variation/lib/python2.7/site-packages/seaborn/apionly.py:6: UserWarning: As seaborn no longer sets a default style on import, the seaborn.apionly module is deprecated. It will be removed in a future version.\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "# Load external dependencies\n",
    "from setup import *\n",
    "# Load internal dependencies\n",
    "import colors,config,plot,utils\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "IOError",
     "evalue": "File /Users/ivg/projects/yeast-cancer-drugs/paper/science/submission/data/pheno/genetic-constructs/pheno_constructs.csv does not exist",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIOError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-f6d595100474>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Load data for genetic constructs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mpheno_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdir_data\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'pheno/genetic-constructs/pheno_constructs.csv'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mencoding\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'utf-8'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeep_default_na\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna_values\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'NaN'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mpheno_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpheno_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mpheno_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroup\u001b[0m\u001b[0;34m!=\u001b[0m\u001b[0;34m'control'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;31m# Filter out strains used for spatial control\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m# Load ancestral populations as control\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/.virtualenvs/genetic-variation/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, skipfooter, skip_footer, doublequote, delim_whitespace, as_recarray, compact_ints, use_unsigned, low_memory, buffer_lines, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    653\u001b[0m                     skip_blank_lines=skip_blank_lines)\n\u001b[1;32m    654\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 655\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    656\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    657\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/.virtualenvs/genetic-variation/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    403\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    404\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 405\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    406\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    407\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/.virtualenvs/genetic-variation/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    762\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'has_index_names'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'has_index_names'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    763\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 764\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    765\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    766\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/.virtualenvs/genetic-variation/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m    983\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'c'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    984\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'c'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 985\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    986\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    987\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'python'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/ivg/.virtualenvs/genetic-variation/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   1603\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'allow_leading_cols'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex_col\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1604\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1605\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparsers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1606\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1607\u001b[0m         \u001b[0;31m# XXX\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__ (pandas/_libs/parsers.c:4209)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._setup_parser_source (pandas/_libs/parsers.c:8873)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mIOError\u001b[0m: File /Users/ivg/projects/yeast-cancer-drugs/paper/science/submission/data/pheno/genetic-constructs/pheno_constructs.csv does not exist"
     ]
    }
   ],
   "source": [
    "# Load data for genetic constructs\n",
    "pheno_df = pd.read_csv(dir_data+'pheno/genetic-constructs/pheno_constructs.csv', encoding='utf-8', keep_default_na=False, na_values='NaN')\n",
    "pheno_df = pheno_df[pheno_df.group!='control'] # Filter out strains used for spatial control\n",
    "\n",
    "# Load ancestral populations as control\n",
    "control_df = pd.read_csv(dir_data+'pheno/populations/pheno_populations.csv', encoding='utf-8', keep_default_na=False, na_values='NaN')\n",
    "control_df = control_df[(control_df.group == 'ancestral')] # Filter out strains used for spatial control\n",
    "\n",
    "# Normalise by the ancestral population\n",
    "def norm(results, control, param='growth_rate'):\n",
    "\n",
    "    results['rel_' + param] = results['norm_' + param] - control['norm_' + param].mean()\n",
    "    \n",
    "    return results\n",
    "\n",
    "pheno_df = pheno_df.groupby(['selection','environment'], as_index=False).apply(\n",
    "    lambda results_df: norm(results_df, control_df, param='growth_rate')\n",
    ")\n",
    "\n",
    "pheno_df = pheno_df.groupby(['selection','environment'], as_index=False).apply(\n",
    "    lambda results_df: norm(results_df, control_df, param='doubling_time')\n",
    ")\n",
    "\n",
    "# Match shared controls by candidate gene\n",
    "groups_1 = pheno_df.groupby(['selection','environment','gene','background'])\n",
    "for (ii,((env_evo, env_test, gene, background),g1)) in enumerate(groups_1):\n",
    "    \n",
    "    if gene!='':\n",
    "        tmp = groups_1.get_group((env_evo, env_test, '', background))        \n",
    "        tmp.loc[:,'gene'] =  tmp['gene'].replace('', gene)\n",
    "        \n",
    "        pheno_df = pheno_df.append(tmp)\n",
    "\n",
    "pheno_df = pheno_df[pheno_df.gene != '']\n",
    "pheno_df = pheno_df.reset_index(drop=True)\n",
    "\n",
    "# # Filter out measurement replicates with >5% measurement error\n",
    "# pheno_df['pct'] = pheno_df.groupby(['selection','environment','gene','background','genotype_long'])['growth_rate']\\\n",
    "# .apply(lambda x: (x-x.mean())/float(x.mean()))\n",
    "# pheno_df = pheno_df[abs(pheno_df['pct'])<0.05]\n",
    "\n",
    "pheno_df.head() # Show dataframe header to stdout"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure S6 - Strategy for strain construction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import Image\n",
    "Image(filename=dir_supp+'figures/supp_figure_schematic_constructs/supp_figure_schematic_constructs_publication.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fig. S6**: Strategy for strain construction. (*A*-*C*) Gene deletion was mediated by homologous recombination between the terminals of the PCR product and the corresponding genomic sequence where the gene to be deleted (‘target’) is encoded. Blue and red lines indicate WA and NA chromosomes, respectively. Flanking regions in green indicate two different homologous sequences targeted for recombination, which are 30-40 bp long in *S. cerevisiae*. Genes of interest were individually deleted in both WA and NA haploids, resulting in *rnr4*∆, *fpr1*∆ and *tor1*∆ strains in both parental backgrounds. A similar strategy was used to delete genes in WA and NA homozygous diploids. *RNR2* and *RNR4* were only deleted in one allele while there is the wild-type gene remaining in the other allele. The primer sequences used are listed in Extended Data Table 6. (*C*) Evolved segregants with *de novo* mutations were isolated from the WAxNA $\\text{F}_\\text{12}$ populations. Using the same strategy, *RNR2* or *TOR1* could be rid of either the wild-type allele or the mutated allele. (*D*) We crossed the strain constructed in (*A*) with the parental strain with wild-type gene, to obtain strains with deleted genes in WA, NA homozygous diploid and WA/NA hybrid."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figures S7 and S8 - Validation tests for driver and passenger mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats\n",
    "\n",
    "param = 'rel_growth_rate'\n",
    "\n",
    "# shape = pd.DataFrame({k: \n",
    "#  pd.pivot_table(x, values='norm_growth_rate', columns=['background','genotype_long']).shape\n",
    "#  for k,x in pheno_df.groupby(['selection','environment','gene'])\n",
    "# })\n",
    "\n",
    "shape = pd.pivot_table(pheno_df, values='norm_growth_rate', columns=['selection','environment','gene'], index=['background','genotype_long']).count()\n",
    "\n",
    "for (ii, ((env_evo), gph1)) in enumerate(pheno_df.groupby(['selection'])):\n",
    "\n",
    "    # Remove NaNs\n",
    "    gph1 = gph1[np.isfinite(gph1[param])]\n",
    "    \n",
    "    nrows=len(gph1.groupby(['background','gene','genotype_long']))\n",
    "    ncols=len(gph1.groupby(['environment']))\n",
    "            \n",
    "    height, width = np.array([nrows*0.11, ncols*3], dtype=float)\n",
    "    \n",
    "    fig = plt.figure(figsize=(width, height))\n",
    "    \n",
    "    fig.subplots_adjust(left=0.07,bottom=0.01,right=0.85,top=0.99)\n",
    "\n",
    "    grid = gridspec.GridSpec(1, 2, hspace=0., wspace=0.)\n",
    "    \n",
    "    gs = {}\n",
    "        \n",
    "    for (jj, ((env_test), gph2)) in enumerate(gph1.groupby(['environment'])):\n",
    "        \n",
    "        nrows = len(gph2['gene'].unique())\n",
    "        ncols = 1\n",
    "        height_ratios = shape[env_evo,env_test].values.squeeze()\n",
    "        gs[env_test] = gridspec.GridSpecFromSubplotSpec(nrows, ncols, height_ratios=height_ratios, \n",
    "                                                        subplot_spec=grid[0,jj], hspace=0., wspace=0.)\n",
    "    \n",
    "        axes = {}\n",
    "        \n",
    "        for (kk, ((gene), gph3)) in enumerate(gph2.groupby('gene')):\n",
    "                        \n",
    "            gph3.loc[:,'rank_background'] = gph3['background'].map(config.construct_background['position'])\n",
    "            gph3.loc[:,'rank_genotype'] = gph3['genotype_long'].map(config.construct_genotype[env_evo][gene])\n",
    "            gph3.sort_values(['rank_background','rank_genotype'], \n",
    "                             ascending=[False,False], inplace=True)\n",
    "      \n",
    "            if kk == 0:\n",
    "                axes[env_test] = plt.subplot(gs[env_test][kk])\n",
    "                ax1 = axes[env_test]\n",
    "                ax1.set_title(config.environment['long_label'][env_test], fontsize=7)\n",
    "            else:\n",
    "                ax1 = plt.subplot(gs[env_test][kk], sharex=axes[env_test])\n",
    "            \n",
    "            gph3 = gph3.set_index(['background','genotype_long'], append=True)[param]\\\n",
    "            .unstack(['background','genotype_long'])\n",
    "                        \n",
    "            # boxplot\n",
    "            bp = gph3.plot(\n",
    "                ax=ax1, kind='box', \n",
    "                widths=0.65, vert=False, return_type='dict',\n",
    "                labels=gph3.columns.get_level_values('background')\n",
    "            )\n",
    "            \n",
    "            colors = [config.construct_background['color']['wt'][b] if g=='WT' \\\n",
    "                      else config.construct_background['color']['mut'][b] \\\n",
    "                      for b,g in zip(gph3.columns.get_level_values('background'),\\\n",
    "                                     gph3.columns.get_level_values('genotype_long'))]\n",
    "            \n",
    "            plot.boxplot_custom(bp, ax1, colors=colors, hatches=['']*30)\n",
    "            \n",
    "            for ll, x in enumerate(gph3.columns):\n",
    "                \n",
    "                if x in config.construct_tests[gene]:\n",
    "                    \n",
    "                    wt_data = gph3[config.construct_tests[gene][x]].dropna()\n",
    "                    mut_data = gph3[x].dropna()\n",
    "                    z_stat, p_val = stats.ranksums(wt_data,mut_data)\n",
    "                                        \n",
    "                    if p_val < 0.0001:\n",
    "                    \n",
    "                        x_min = min(wt_data.min(),mut_data.min())\n",
    "                        x_max = max(wt_data.max(),mut_data.max())\n",
    "                             \n",
    "                        ax1.annotate('', xy=(x_max, ll+1), xycoords='data',\n",
    "                                     xytext=(x_max, ll+2), textcoords='data',\n",
    "                                     arrowprops=dict(arrowstyle=\"-\", ec='#aaaaaa', linewidth=.75,\n",
    "                                                     connectionstyle=\"bar,fraction=-0.3\"))\n",
    "                        ax1.annotate('*',#utils.stars(p_val), \n",
    "                                     xy=((x_max+0.02 if env_evo=='HU' else x_max+0.04), ll+1.5), xycoords='data',\n",
    "                                     ha='center', va='center', fontsize=6, weight='bold', rotation=270)\n",
    "                        \n",
    "            if jj==0:\n",
    "                ax1.annotate(gene, \n",
    "                             xy=(-0.2, 0.5), xycoords=(\"axes fraction\", \"axes fraction\"),\n",
    "                             ha='right', va='center', annotation_clip=False, rotation=90, \n",
    "                             fontsize=6, fontstyle='italic', fontweight='bold')\n",
    "\n",
    "            # reset ticks\n",
    "            ax1.set_yticks([])\n",
    "            ax1.set_yticklabels([])\n",
    "            \n",
    "            ax1.set_axisbelow(False)\n",
    "            \n",
    "            ### vertical ###\n",
    "            ax1.xaxis.grid(ls=\"-\", lw=.75, color=\"0.9\", zorder=0)\n",
    "            ax1.axvline(x=0., color='k', ls=\"--\", dashes=(7, 7), lw=.75, zorder=3)\n",
    "            \n",
    "            ### horizontal ###\n",
    "            \n",
    "            ## background\n",
    "            ystart, yend, ylabels = plot.set_custom_labels(gph3.columns, 0)\n",
    "            \n",
    "            # grid\n",
    "            ygrid=[yst+1.5 for yst in list(set(ystart.values()))]\n",
    "            [ax1.axhline(g, lw=.75, ls=\"-\", color=\"0.9\", zorder=2) for g in ygrid]\n",
    "            \n",
    "            # labels\n",
    "            if jj==0:\n",
    "                # tick labels\n",
    "                ax1.set_yticks([y+1 for y in ylabels.values()], minor=True)\n",
    "                ax1.set_yticklabels(ylabels.keys(), minor=True)\n",
    "                ax1.get_yaxis().tick_left()\n",
    "                if kk==0:\n",
    "                    # axis label\n",
    "                    ax1.set_ylabel('Background', transform=ax1.transAxes, \n",
    "                                   weight='bold', rotation=0)\n",
    "                    ax1.yaxis.set_label_coords(-0.1, 1.025)\n",
    "            \n",
    "            ## genotype\n",
    "            # labels\n",
    "            if jj==1:\n",
    "                # tick labels\n",
    "                ylabels = gph3.columns.get_level_values('genotype_long')\n",
    "                ax1.set_yticks(np.arange(1, len(ylabels)+1), minor=True)\n",
    "                ax1.set_yticklabels(['/'.join(y) if isinstance(y, tuple) else y for y in ylabels], minor=True)\n",
    "                ax1.get_yaxis().tick_right()\n",
    "                if kk==0:\n",
    "                    # axis label\n",
    "                    ax1.set_ylabel('Genotype', transform=ax1.transAxes, \n",
    "                                   weight='bold', rotation=0)\n",
    "                    ax1.yaxis.set_label_coords(1.1, 1.025)\n",
    "            \n",
    "            # set axes labels\n",
    "            ax1.set_xlabel(r'Rel. growth rate, $\\lambda_{bg}$', fontsize=10)\n",
    "            \n",
    "            if jj==1 and kk==0:\n",
    "                wa_artist = patches.Rectangle((0,0), width=1, height=1, edgecolor='k',\n",
    "                                              facecolor=config.construct_background['color']['mut']['WA'])\n",
    "                na_artist = patches.Rectangle((0,0), width=1, height=1, edgecolor='k',\n",
    "                                              facecolor=config.construct_background['color']['mut']['NA'])\n",
    "                wana_artist = patches.Rectangle((0,0), width=1, height=1, edgecolor='k',\n",
    "                                                facecolor=config.construct_background['color']['mut']['WA/NA'])\n",
    "\n",
    "                leg1 = ax1.legend([wa_artist,na_artist,wana_artist], \n",
    "                                  ['WA; WA/WA','NA; NA/NA','WA/NA'], \n",
    "                                  ncol=1, loc='upper right',\n",
    "                                  borderaxespad=0, handlelength=0.75, \n",
    "                                  prop={'size':5}, title='Background', \n",
    "                                  labelspacing=.32)\n",
    "\n",
    "                wt_artist = patches.Rectangle((0,0), width=1, height=1, facecolor='0.9', edgecolor='k')\n",
    "                construct_artist = patches.Rectangle((0,0), width=1, height=1, facecolor='0.1', edgecolor='k')\n",
    "\n",
    "                leg2 = ax1.legend([wt_artist,construct_artist], \n",
    "                                  ['WT','construct'], \n",
    "                                  ncol=1, loc='lower right',\n",
    "                                  borderaxespad=0, handlelength=0.75, \n",
    "                                  prop={'size':5}, title='Genotype', \n",
    "                                  labelspacing=.32)\n",
    "                        \n",
    "                ax1.add_artist(leg1)\n",
    "        \n",
    "                for leg in [leg1,leg2]:\n",
    "                    plt.setp(leg.get_title(),fontsize=6)\n",
    "                    leg.set_zorder(2)\n",
    "                    leg.get_frame().set_edgecolor('none')\n",
    "                    leg.get_frame().set_facecolor('w')\n",
    "                    \n",
    "            if env_evo=='RM' and env_test=='RM' and gene=='FPR1':\n",
    "                transform = transforms.blended_transform_factory(ax1.transAxes, ax1.transData)\n",
    "                x0, x1, y0, y1 = (1.1, 1.9, .5, 7.5)\n",
    "                im = ax1.imshow(plt.imread(dir_supp+'figures/supp_figure_pheno_constructs/FPR1_LOH_inset.png'),\n",
    "                                aspect='auto', extent=(x0, x1, y0, y1), transform=transform, zorder=1)\n",
    "            \n",
    "            transform = transforms.blended_transform_factory(ax1.transAxes, ax1.transAxes)\n",
    "            if env_evo=='HU':\n",
    "                ax1.set_xlim(-0.45,0.49)\n",
    "                ax1.xaxis.set_major_locator( ticker.MaxNLocator(nbins = 4) )\n",
    "                ax1.xaxis.set_minor_locator( ticker.MaxNLocator(nbins = 4) )\n",
    "                ax1.xaxis.set_major_formatter(ticker.FormatStrFormatter('%.2f'))\n",
    "                if env_test=='HU':\n",
    "                    ax1.spines['left'].set_visible(False)\n",
    "                    ax1.spines['right'].set_visible(True)\n",
    "                    patch = patches.Rectangle((-.18,0), width=.18, height=1,\n",
    "                                              linewidth=.75, facecolor='0.9', edgecolor='w', \n",
    "                                              transform=transform, zorder=0)\n",
    "                if env_test=='SC':\n",
    "                    ax1.spines['left'].set_visible(True)\n",
    "                    ax1.spines['right'].set_visible(False)\n",
    "                    patch = patches.Rectangle((1,0), width=.4, height=1,\n",
    "                                              linewidth=.75, facecolor='0.9', edgecolor='w', \n",
    "                                              transform=transform, zorder=0)\n",
    "            elif env_evo=='RM':\n",
    "                ax1.set_xlim(-0.65,2)\n",
    "                ax1.xaxis.set_major_locator( ticker.MaxNLocator(nbins = 3) )\n",
    "                ax1.xaxis.set_minor_locator( ticker.MaxNLocator(nbins = 3) )\n",
    "                ax1.xaxis.set_major_formatter(ticker.FormatStrFormatter('%.2f'))\n",
    "                if env_test=='RM':\n",
    "                    ax1.spines['left'].set_visible(False)\n",
    "                    ax1.spines['right'].set_visible(True)\n",
    "                    patch = patches.Rectangle((-.18,0), width=.18, height=1,\n",
    "                                              linewidth=.75, facecolor='0.9', edgecolor='w',  \n",
    "                                              transform=transform, zorder=0)\n",
    "                if env_test=='SC':\n",
    "                    ax1.spines['left'].set_visible(True)\n",
    "                    ax1.spines['right'].set_visible(False)\n",
    "                    patch = patches.Rectangle((1,0), width=.525, height=1,\n",
    "                                              linewidth=.75, facecolor='0.9', edgecolor='w',  \n",
    "                                              transform=transform, zorder=0)\n",
    "                    patch.set_clip_on(False)\n",
    "            ax1.add_patch(patch)\n",
    "            patch.set_clip_on(False)\n",
    "\n",
    "        # Tweak axes\n",
    "        for ax in fig.get_axes():\n",
    "            ax.spines['top'].set_visible(True)\n",
    "            ax.spines['bottom'].set_visible(True)\n",
    "            \n",
    "            ax.xaxis.label.set_size(6)\n",
    "            ax.yaxis.label.set_size(6)\n",
    "            \n",
    "            ax.tick_params(axis='x', which='both', size=0, labelsize=6)\n",
    "            ax.tick_params(axis='y', which='major', size=0, labelsize=6)\n",
    "            ax.tick_params(axis='y', which='minor', size=0, labelsize=6)\n",
    "            for sp in ax.spines.values():\n",
    "                sp.set(color='k', linewidth=.5, linestyle='-')\n",
    "    \n",
    "    plot.save_figure(dir_supp+'figures/supp_figure_pheno_constructs/supp_figure_pheno_constructs_%s' % env_evo)\n",
    "    \n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fig. S7:** Validation tests for driver mutations in hydroxyurea, measured in SC+HU (left) and SC (right). Growth rate measurements, $\\lambda_{bg}$, are shown for measurement replicates of each construct ($n = 64$) and grouped by candidate gene and by background of the construct, where the background $b$ can be WA, NA (haploid); WA/WA, NA/NA (diploid); WA/NA (hybrid), and the genotype $g$ can be wild-type for the gene, deleted or hemizygous. Medians and 25%/75% percentiles across groups are shown, with medians as horizontal lines and outliers highlighted. The color of each of the boxes reflects the background, with WA and WA/WA (blue), NA and NA/NA (red) and WA/NA (purple). Lighter shades indicate a wild-type (WT) control for that specific background and darker colors are the candidate strains. Using a non-parametric Wilcoxon rank-sum test, we compared deletion strains against their respective WT control and hemizygous strains against each other. Significance tests with a $P$-value below $10^{−4}$ are highlighted."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fig. S8:** Validation tests for driver and passenger mutations in rapamycin, measured in SC+RM (left) and SC (right). Growth rate measurements, $\\lambda_{bg}$, are shown for measurement replicates of each construct ($n = 64$) and grouped by candidate gene and by background of the construct, where the background $b$ can be WA, NA (haploid); WA/WA, NA/NA (diploid); WA/NA (hybrid), and the genotype $g$ can be wild-type for the gene, deleted or hemizygous. Medians and 25%/75% percentiles across groups are shown, with medians as horizontal lines and outliers highlighted. The color of each of the boxes reflects the background, with WA and WA/WA (blue), NA and NA/NA (red) and WA/NA (purple). Lighter shades indicate a wild-type (WT) control for that specific background and darker colors are the candidate strains. Using a non-parametric Wilcoxon rank-sum test, we compared deletion strains against their respective WT control and hemizygous strains against each other. Significance tests with a $P$-value below $10^{−4}$ are highlighted. Visual inspection of *FPR1* heterozygous deletions using a spot assay (inset) manifests the immediate loss of the wild-type allele by LOH – validated by colony Sanger sequencing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import Image\n",
    "Image(filename=dir_supp+'figures/supp_figure_pheno_constructs/FPR1_LOH_inset.png')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "genetic-variation",
   "language": "python",
   "name": "genetic-variation"
  },
  "latex_metadata": {
   "affiliation": "Wellcome Trust Sanger Institute, Hinxton, Cambridge CB10 1SA, United Kingdom; Department of Applied Mathematics and Theoretical Physics, University of Cambridge, Cambridge CB3 0WA, United Kingdom",
   "author": "Ignacio Vazquez-Garcia",
   "title": "Clonal heterogeneity influences the fate of new adaptive mutations"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
