{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Supplementary Material\n",
    "\n",
    "> **\"Background-dependent selection of genetic variation in heterogeneous populations\"**\n",
    "\n",
    "> Ignacio Vázquez-García, Francisco Salinas, Jing Li, Andrej Fischer, Benjamin Barré, Johan Hallin, Anders Bergström, Elisa Alonso-Pérez, Jonas Warringer, Ville Mustonen, Gianni Liti\n",
    "\n",
    "## Supplementary Figures S9 and S10\n",
    "\n",
    "This IPython notebook is provided for reproduction of Figures S9 and S10 of the paper. It can be viewed by copying its URL to nbviewer and it can be run by opening it in binder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# load external dependencies\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "import matplotlib.lines as lines\n",
    "import matplotlib.patches as patches\n",
    "import matplotlib.transforms as transforms\n",
    "from matplotlib.ticker import FormatStrFormatter,MaxNLocator\n",
    "%matplotlib inline\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd # (*) Pandas for data manipulation\n",
    "from scipy import stats\n",
    "\n",
    "import json, matplotlib\n",
    "s = json.load( open(\"matplotlibrc.json\") )\n",
    "matplotlib.rcParams.update(s)\n",
    "\n",
    "# load internal dependencies\n",
    "import colors,config,plot,model,utils\n",
    "\n",
    "# absolute file path for root directory of git repo\n",
    "dir_repo = utils.get_git_path()\n",
    "dir_data = dir_repo+\"/data/\"\n",
    "dir_paper = dir_repo+\"/manuscript/main/\"\n",
    "dir_supp = dir_repo+\"/manuscript/supp/\"\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import warnings\n",
    "\n",
    "with warnings.catch_warnings():\n",
    "    warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>selection</th>\n",
       "      <th>environment</th>\n",
       "      <th>run</th>\n",
       "      <th>index</th>\n",
       "      <th>plate</th>\n",
       "      <th>row</th>\n",
       "      <th>column</th>\n",
       "      <th>group</th>\n",
       "      <th>population</th>\n",
       "      <th>background</th>\n",
       "      <th>...</th>\n",
       "      <th>time</th>\n",
       "      <th>assignment</th>\n",
       "      <th>isolate</th>\n",
       "      <th>gene</th>\n",
       "      <th>genotype_short</th>\n",
       "      <th>genotype_long</th>\n",
       "      <th>amino_acids</th>\n",
       "      <th>mating</th>\n",
       "      <th>auxotrophy</th>\n",
       "      <th>growth_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8000</th>\n",
       "      <td>HU</td>\n",
       "      <td>HU</td>\n",
       "      <td>1</td>\n",
       "      <td>282</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>42</td>\n",
       "      <td>parents</td>\n",
       "      <td></td>\n",
       "      <td>WA</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>RNR4</td>\n",
       "      <td></td>\n",
       "      <td>WT</td>\n",
       "      <td></td>\n",
       "      <td>MATa</td>\n",
       "      <td>ura3</td>\n",
       "      <td>-0.233068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7744</th>\n",
       "      <td>HU</td>\n",
       "      <td>HU</td>\n",
       "      <td>1</td>\n",
       "      <td>286</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>46</td>\n",
       "      <td>parents</td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>RNR4</td>\n",
       "      <td></td>\n",
       "      <td>WT</td>\n",
       "      <td></td>\n",
       "      <td>MATa</td>\n",
       "      <td>ura3</td>\n",
       "      <td>0.302884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8001</th>\n",
       "      <td>HU</td>\n",
       "      <td>HU</td>\n",
       "      <td>1</td>\n",
       "      <td>378</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>42</td>\n",
       "      <td>parents</td>\n",
       "      <td></td>\n",
       "      <td>WA</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>RNR4</td>\n",
       "      <td></td>\n",
       "      <td>WT</td>\n",
       "      <td></td>\n",
       "      <td>MATa</td>\n",
       "      <td>ura3</td>\n",
       "      <td>-0.251914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7745</th>\n",
       "      <td>HU</td>\n",
       "      <td>HU</td>\n",
       "      <td>1</td>\n",
       "      <td>382</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>46</td>\n",
       "      <td>parents</td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>RNR4</td>\n",
       "      <td></td>\n",
       "      <td>WT</td>\n",
       "      <td></td>\n",
       "      <td>MATa</td>\n",
       "      <td>ura3</td>\n",
       "      <td>0.350931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8002</th>\n",
       "      <td>HU</td>\n",
       "      <td>HU</td>\n",
       "      <td>1</td>\n",
       "      <td>434</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>parents</td>\n",
       "      <td></td>\n",
       "      <td>WA</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>RNR4</td>\n",
       "      <td></td>\n",
       "      <td>WT</td>\n",
       "      <td></td>\n",
       "      <td>MATα</td>\n",
       "      <td>lys2</td>\n",
       "      <td>-0.242307</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     selection environment run  index  plate  row  column    group population  \\\n",
       "8000        HU          HU   1    282      0    5      42  parents              \n",
       "7744        HU          HU   1    286      0    5      46  parents              \n",
       "8001        HU          HU   1    378      0    7      42  parents              \n",
       "7745        HU          HU   1    382      0    7      46  parents              \n",
       "8002        HU          HU   1    434      0    9       2  parents              \n",
       "\n",
       "     background     ...     time assignment isolate  gene genotype_short  \\\n",
       "8000         WA     ...                              RNR4                  \n",
       "7744         NA     ...                              RNR4                  \n",
       "8001         WA     ...                              RNR4                  \n",
       "7745         NA     ...                              RNR4                  \n",
       "8002         WA     ...                              RNR4                  \n",
       "\n",
       "     genotype_long amino_acids mating auxotrophy growth_rate  \n",
       "8000            WT               MATa       ura3   -0.233068  \n",
       "7744            WT               MATa       ura3    0.302884  \n",
       "8001            WT               MATa       ura3   -0.251914  \n",
       "7745            WT               MATa       ura3    0.350931  \n",
       "8002            WT               MATα       lys2   -0.242307  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load data\n",
    "pheno_df = utils.load_data(dir_data+'pheno/genetic-constructs/pheno_constructs_norm.pickle')\n",
    "\n",
    "# # filter out measurement replicates with >5% measurement error\n",
    "# pheno_df['pct'] = pheno_df.groupby(['selection','environment','gene','background','genotype_long'])['growth_rate']\\\n",
    "# .apply(lambda x: (x-x.mean())/float(x.mean()))\n",
    "# pheno_df = pheno_df[abs(pheno_df['pct'])<0.05]\n",
    "\n",
    "pheno_df.head() # show dataframe header to stdout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from scipy import stats \n",
    "\n",
    "stattest=[]\n",
    "\n",
    "param='growth_rate'\n",
    "\n",
    "for ((env_evo,env_test,gene),geeg) in pheno_df.groupby(['selection','environment','gene']):\n",
    "    for ((background,genotype),gbg) in geeg.groupby(['background','genotype_long']):\n",
    "        if (background,genotype) in config.construct_tests[gene]:\n",
    "            wt=config.construct_tests[gene][(background,genotype)]#\n",
    "            mut=(background,genotype)\n",
    "            wt_data = geeg.groupby(['background','genotype_long']).get_group(wt)[param]\n",
    "            mut_data = geeg.groupby(['background','genotype_long']).get_group(mut)[param]\n",
    "            wt_data = wt_data.dropna().tolist()\n",
    "            mut_data = mut_data.dropna().tolist()\n",
    "            x_min=min(wt_data+mut_data)\n",
    "            x_max=max(wt_data+mut_data)\n",
    "            z_stat,p_val = stats.ranksums(wt_data,mut_data)\n",
    "            stattest.append([env_evo,env_test,gene,\n",
    "                             wt,mut,x_min,x_max,\n",
    "                             z_stat,p_val,utils.stars(p_val)])\n",
    "\n",
    "stattest = pd.DataFrame(stattest,columns=('selection','environment','gene',\n",
    "                                          'wt','mut','min','max',\n",
    "                                          'z_stat','p_val','stars'))\n",
    "\n",
    "stattest_to_file = stattest.sort_values(['gene','selection', 'environment','p_val'])\n",
    "stattest_to_file['wt'] = pd.Series(','.join(k) for k in stattest_to_file['wt'])\n",
    "stattest_to_file['mut'] = pd.Series(','.join(k) for k in stattest_to_file['mut'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "param = 'growth_rate'\n",
    "\n",
    "shape = pd.DataFrame({k: \n",
    " pd.pivot_table(x, values='growth_rate', columns=['background','genotype_long']).shape\n",
    " for k,x in pheno_df.groupby(['selection','environment','gene'])\n",
    "})\n",
    "\n",
    "for (ii, ((env_evo), gph1)) in enumerate(pheno_df.groupby(['selection'])):\n",
    "\n",
    "    # remove NaNs\n",
    "    gph1 = gph1[np.isfinite(gph1[param])]\n",
    "    \n",
    "    nrows=len(gph1.groupby(['background','gene','genotype_long']))\n",
    "    ncols=len(gph1.groupby(['environment']))\n",
    "            \n",
    "    height, width = np.array([nrows*0.11, ncols*3], dtype=float)\n",
    "    \n",
    "    fig = plt.figure(figsize=(width, height), dpi=300)\n",
    "    \n",
    "    fig.subplots_adjust(left=0.07,bottom=0.01,right=0.85,top=0.99)\n",
    "\n",
    "    grid = gridspec.GridSpec(1, 2, hspace=0., wspace=0.)\n",
    "    \n",
    "    gs = {}\n",
    "        \n",
    "    for (jj, ((env_test), gph2)) in enumerate(gph1.groupby(['environment'])):\n",
    "        \n",
    "        nrows = len(gph2['gene'].unique())\n",
    "        ncols = 1\n",
    "        height_ratios = shape[env_evo,env_test].values.squeeze()\n",
    "        gs[env_test] = gridspec.GridSpecFromSubplotSpec(nrows, ncols, height_ratios=height_ratios, \n",
    "                                                        subplot_spec=grid[0,jj], hspace=0., wspace=0.)\n",
    "    \n",
    "        axes = {}\n",
    "        \n",
    "        for (kk, ((gene), gph3)) in enumerate(gph2.groupby('gene')):\n",
    "                        \n",
    "            gph3.loc[:,'rank_background'] = gph3['background'].map(config.construct_background['position'])\n",
    "            gph3.loc[:,'rank_genotype'] = gph3['genotype_long'].map(config.construct_genotype[env_evo][gene])\n",
    "            gph3.sort_values(['rank_background','rank_genotype'], \n",
    "                             ascending=[False,False], inplace=True)\n",
    "      \n",
    "            if kk == 0:\n",
    "                axes[env_test] = plt.subplot(gs[env_test][kk])\n",
    "                ax1 = axes[env_test]\n",
    "                ax1.set_title(config.environment['long_label'][env_test], fontsize=8)\n",
    "            else:\n",
    "                ax1 = plt.subplot(gs[env_test][kk], sharex=axes[env_test])\n",
    "            \n",
    "            gph3 = gph3.set_index(['background','genotype_long'], append=True)[param]\\\n",
    "            .unstack(['background','genotype_long'])\n",
    "            \n",
    "            # boxplot\n",
    "            bp = gph3.plot(ax=ax1, \n",
    "                           kind='box', \n",
    "                           widths=0.65, \n",
    "                           vert=False, \n",
    "                           return_type='dict',\n",
    "                           labels=gph3.columns.get_level_values('background'))\n",
    "            \n",
    "            colors = [config.construct_background['color']['wt'][b] if g=='WT' \\\n",
    "                      else config.construct_background['color']['mut'][b] \\\n",
    "                      for b,g in zip(gph3.columns.get_level_values('background'),\\\n",
    "                                     gph3.columns.get_level_values('genotype_long'))]\n",
    "            \n",
    "            plot.boxplot_custom(bp, ax1, colors=colors, hatches=['']*30)\n",
    "            \n",
    "            for ll, x in enumerate(gph3.columns):\n",
    "                \n",
    "                if x in config.construct_tests[gene]:\n",
    "                    \n",
    "                    wt_data = gph3[config.construct_tests[gene][x]].dropna()\n",
    "                    mut_data = gph3[x].dropna()\n",
    "                    z_stat, p_val = stats.ranksums(wt_data,mut_data)\n",
    "                                        \n",
    "                    if p_val < 0.0001:\n",
    "                    \n",
    "                        x_min = min(wt_data.min(),mut_data.min())\n",
    "                        x_max = max(wt_data.max(),mut_data.max())\n",
    "                             \n",
    "                        ax1.annotate('', xy=(x_max, ll+1), xycoords='data',\n",
    "                                     xytext=(x_max, ll+2), textcoords='data',\n",
    "                                     arrowprops=dict(arrowstyle=\"-\", ec='#aaaaaa', linewidth=.75,\n",
    "                                                     connectionstyle=\"bar,fraction=-0.3\"))\n",
    "                        ax1.annotate(utils.stars(p_val), \n",
    "                                     xy=((x_max+0.02 if env_evo=='HU' else x_max+0.04), ll+1.5), xycoords='data',\n",
    "                                     ha='center', va='center', fontsize=6, rotation=270)\n",
    "                        \n",
    "            if jj==0:\n",
    "                ax1.annotate(gene, \n",
    "                             xy=(-0.2, 0.5), xycoords=(\"axes fraction\", \"axes fraction\"),\n",
    "                             ha='right', va='center', annotation_clip=False, rotation=90, \n",
    "                             fontsize=8, fontstyle='italic')\n",
    "\n",
    "            # reset ticks\n",
    "            ax1.set_yticks([])\n",
    "            ax1.set_yticklabels([])\n",
    "            \n",
    "            ax1.set_axisbelow(False)\n",
    "            \n",
    "            ### vertical ###\n",
    "            ax1.xaxis.grid(ls=\"-\", lw=.75, color=\"0.9\", zorder=0)\n",
    "            ax1.axvline(x=0., color='k', ls=\"--\", lw=.75, zorder=3)\n",
    "            \n",
    "            ### horizontal ###\n",
    "            \n",
    "            ## background\n",
    "            ystart, yend, ylabels = plot.set_custom_labels(gph3.columns, 0)\n",
    "            \n",
    "            # grid\n",
    "            ygrid=[yst+1.5 for yst in list(set(ystart.values()))]\n",
    "            [ax1.axhline(g, lw=.75, ls=\"-\", color=\"0.9\", zorder=2) for g in ygrid]\n",
    "            \n",
    "            # labels\n",
    "            if jj==0:\n",
    "                # tick labels\n",
    "                ax1.set_yticks([y+1 for y in ylabels.values()], minor=True)\n",
    "                ax1.set_yticklabels(ylabels.keys(), minor=True)\n",
    "                ax1.get_yaxis().tick_left()\n",
    "                if kk==0:\n",
    "                    # axis label\n",
    "                    ax1.set_ylabel('Background', transform=ax1.transAxes, rotation=0)\n",
    "                    ax1.yaxis.set_label_coords(-0.125, 1.025)\n",
    "            \n",
    "            ## genotype\n",
    "            # labels\n",
    "            if jj==1:\n",
    "                # tick labels\n",
    "                ylabels = gph3.columns.get_level_values('genotype_long')\n",
    "                ax1.set_yticks(np.arange(1, len(ylabels)+1), minor=True)\n",
    "                ax1.set_yticklabels(['/'.join(y) if isinstance(y, tuple) else y for y in ylabels], minor=True)\n",
    "                ax1.get_yaxis().tick_right()\n",
    "                if kk==0:\n",
    "                    # axis label\n",
    "                    ax1.set_ylabel('Genotype', transform=ax1.transAxes, rotation=0)\n",
    "                    ax1.yaxis.set_label_coords(1.125, 1.025)\n",
    "            \n",
    "            # set axes labels\n",
    "            ax1.set_xlabel(r'Rel. growth rate, $\\lambda_{bg}$', fontsize=10)\n",
    "            \n",
    "            if jj==1 and kk==0:\n",
    "                wa_artist = patches.Rectangle((0,0), width=1, height=1, \n",
    "                                              color=config.construct_background['color']['mut']['WA'])\n",
    "                na_artist = patches.Rectangle((0,0), width=1, height=1, \n",
    "                                              color=config.construct_background['color']['mut']['NA'])\n",
    "                wana_artist = patches.Rectangle((0,0), width=1, height=1, \n",
    "                                                color=config.construct_background['color']['mut']['WA/NA'])\n",
    "\n",
    "                leg1 = ax1.legend([wa_artist,na_artist,wana_artist], \n",
    "                                  ['WA; WA/WA','NA; NA/NA','WA/NA'], \n",
    "                                  ncol=1, loc='upper right',\n",
    "                                  borderaxespad=0, handlelength=0.75, \n",
    "                                  prop={'size':5}, title='Background', \n",
    "                                  labelspacing=.32)\n",
    "\n",
    "                wt_artist = patches.Rectangle((0,0), width=1, height=1, facecolor='0.9', edgecolor='k')\n",
    "                construct_artist = patches.Rectangle((0,0), width=1, height=1, facecolor='0.1', edgecolor='k')\n",
    "\n",
    "                leg2 = ax1.legend([wt_artist,construct_artist], \n",
    "                                  ['WT','construct'], \n",
    "                                  ncol=1, loc='lower right',\n",
    "                                  borderaxespad=0, handlelength=0.75, \n",
    "                                  prop={'size':5}, title='Genotype', \n",
    "                                  labelspacing=.32)\n",
    "                        \n",
    "                ax1.add_artist(leg1)\n",
    "        \n",
    "                for leg in [leg1,leg2]:\n",
    "                    plt.setp(leg.get_title(),fontsize=6)\n",
    "                    leg.set_zorder(2)\n",
    "                    leg.get_frame().set_edgecolor('none')\n",
    "                    leg.get_frame().set_facecolor('w')\n",
    "                    \n",
    "            if env_evo=='RM' and env_test=='RM' and gene=='FPR1':\n",
    "                transform = transforms.blended_transform_factory(ax1.transAxes, ax1.transData)\n",
    "                x0, x1, y0, y1 = (1.125, 1.875, .5, 7.5)\n",
    "                im = ax1.imshow(plt.imread(dir_supp+'figures/supp_figure_pheno_constructs/FPR1_LOH_inset.png'),\n",
    "                                aspect='auto', extent=(x0, x1, y0, y1), transform=transform, zorder=1)\n",
    "            \n",
    "            transform = transforms.blended_transform_factory(ax1.transAxes, ax1.transAxes)\n",
    "            if env_evo=='HU':\n",
    "                ax1.set_xlim(-0.45,0.49)\n",
    "                ax1.xaxis.set_major_locator( MaxNLocator(nbins = 4) )\n",
    "                ax1.xaxis.set_minor_locator( MaxNLocator(nbins = 4) )\n",
    "                ax1.xaxis.set_major_formatter(FormatStrFormatter('%.2f'))\n",
    "                if env_test=='HU':\n",
    "                    ax1.spines['left'].set_visible(False)\n",
    "                    ax1.spines['right'].set_visible(True)\n",
    "                    patch = patches.Rectangle((-.25,0), width=.05, height=1,\n",
    "                                              linewidth=.75, facecolor='0.9', edgecolor='w', \n",
    "                                              transform=transform, zorder=0)\n",
    "                    patch = patches.Rectangle((-.18,0), width=.18, height=1,\n",
    "                                              linewidth=.75, facecolor='0.9', edgecolor='w', \n",
    "                                              transform=transform, zorder=0)\n",
    "                if env_test=='YNB':\n",
    "                    ax1.spines['left'].set_visible(True)\n",
    "                    ax1.spines['right'].set_visible(False)\n",
    "                    patch = patches.Rectangle((1,0), width=.45, height=1,\n",
    "                                              linewidth=.75, facecolor='0.9', edgecolor='w', \n",
    "                                              transform=transform, zorder=0)\n",
    "            elif env_evo=='RM':\n",
    "                ax1.set_xlim(-0.65,2)\n",
    "                ax1.xaxis.set_major_locator( MaxNLocator(nbins = 3) )\n",
    "                ax1.xaxis.set_minor_locator( MaxNLocator(nbins = 3) )\n",
    "                ax1.xaxis.set_major_formatter(FormatStrFormatter('%.2f'))\n",
    "                if env_test=='RM':\n",
    "                    ax1.spines['left'].set_visible(False)\n",
    "                    ax1.spines['right'].set_visible(True)\n",
    "                    patch = patches.Rectangle((-.18,0), width=.18, height=1,\n",
    "                                              linewidth=.75, facecolor='0.9', edgecolor='w',  \n",
    "                                              transform=transform, zorder=0)\n",
    "                if env_test=='YNB':\n",
    "                    ax1.spines['left'].set_visible(True)\n",
    "                    ax1.spines['right'].set_visible(False)\n",
    "                    patch = patches.Rectangle((1,0), width=.575, height=1,\n",
    "                                              linewidth=.75, facecolor='0.9', edgecolor='w',  \n",
    "                                              transform=transform, zorder=0)\n",
    "                    patch.set_clip_on(False)\n",
    "            ax1.add_patch(patch)\n",
    "            patch.set_clip_on(False)\n",
    "\n",
    "        # tweak axes\n",
    "        for ax in fig.get_axes():\n",
    "            ax.spines['top'].set_visible(True)\n",
    "            ax.spines['bottom'].set_visible(True)\n",
    "            \n",
    "            ax.xaxis.label.set_size(8)\n",
    "            ax.yaxis.label.set_size(8)\n",
    "            ax.tick_params(axis='x', which='both', size=0, labelsize=7)\n",
    "            ax.tick_params(axis='y', which='major', size=2, labelsize=7)\n",
    "            ax.tick_params(axis='y', which='minor', size=2, labelsize=7)\n",
    "            for sp in ax.spines.values():\n",
    "                sp.set(color='k', linewidth=.5, linestyle='-')\n",
    "\n",
    "    fig.savefig(dir_supp+'figures/supp_figure_pheno_constructs/supp_figure_pheno_constructs_%s.png' % env_evo, \n",
    "                dpi=300, bbox_inches='tight')\n",
    "    fig.savefig(dir_supp+'figures/supp_figure_pheno_constructs/supp_figure_pheno_constructs_%s.svg' % env_evo, \n",
    "                dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fig. S8:** Validation tests for driver mutations in hydroxyurea, measured in YNB+HU (left) and YNB (right). Growth rate measurements, $\\lambda_{bg}$, are averaged over measurement replicates ($n = 64$) and grouped by candidate gene and by background of the construct, where the background $b$ can be WA, NA (haploid); WA/WA, NA/NA (diploid); WA/NA (hybrid), the genotype $g$ can be wild-type for the gene, deleted or hemizygous. Medians and 25%/75% percentiles across groups are shown, with medians as horizontal lines and outliers highlighted. The color of each of the boxes reflects the background, with WA and WA/WA (blue), NA and NA/NA (red) and WA/NA (purple). Lighter shades indicate WT control for that specific background and darker colors are the candidate strains. Using a non-parametric Wilcoxon rank-sum test, we compared deletion strains against their respective WT control and hemizygous strains against each other. Significance tests with a $p$-value below 0.0001 are highlighted."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fig. S9:** Validation tests for driver and passenger mutations in rapamycin, measured in YNB+RM (left) and YNB (right). Growth rate measurements, $\\lambda_{bg}$, are averaged over measurement replicates ($n = 64$) and grouped by candidate gene and by background of the construct, where the background $b$ can be WA, NA (haploid); WA/WA, NA/NA (diploid); WA/NA (hybrid), the genotype $g$ can be wild-type for the gene, deleted or hemizygous. Medians and 25%/75% percentiles across groups are shown, with medians as horizontal lines and outliers highlighted. The color of each of the boxes reflects the background, with WA and WA/WA (blue), NA and NA/NA (red) and WA/NA (purple). Lighter shades indicate WT control for that specific background and darker colours are the candidate strains. Using a non-parametric Wilcoxon rank-sum test, we compared deletion strains against their respective WT control and hemizygous strains against each other. Significance tests with a $p$-value below 0.0001 are highlighted. Visual inspection of *FPR1* heterozygous deletions using a spot assay (inset) manifests the immediate loss of the wild-type allele by LOH -- validated by colony Sanger sequencing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from IPython.display import Image\n",
    "Image(filename=dir_supp+'figures/supp_figure_pheno_constructs/FPR1_LOH_inset.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  },
  "latex_metadata": {
   "affiliation": "Wellcome Trust Sanger Institute, Hinxton, Cambridge CB10 1SA, United Kingdom; Department of Applied Mathematics and Theoretical Physics, University of Cambridge, Cambridge CB3 0WA, United Kingdom",
   "author": "Ignacio Vazquez-Garcia",
   "title": "Background-dependent selection of genetic variation in heterogeneous populations"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
